// eslint/lib/rules/array-element-newline.js
/**
 * @fileoverview Rule to enforce line breaks after each array element
 * @author Jan Peer Stöcklmair <https://github.com/JPeer264>
 */

"use strict";

const astUtils = require("./utils/ast-utils");

//------------------------------------------------------------------------------
// Rule Definition
//------------------------------------------------------------------------------

module.exports = ***REMOVED***
    meta: ***REMOVED***
        type: "layout",

        docs: ***REMOVED***
            description: "enforce line breaks after each array element",
            category: "Stylistic Issues",
            recommended: false,
            url: "https://eslint.org/docs/rules/array-element-newline"
        ***REMOVED***,

        fixable: "whitespace",

        schema: [
            ***REMOVED***
                oneOf: [
                    ***REMOVED***
                        enum: ["always", "never", "consistent"]
                    ***REMOVED***,
                    ***REMOVED***
                        type: "object",
                        properties: ***REMOVED***
                            multiline: ***REMOVED***
                                type: "boolean"
                            ***REMOVED***,
                            minItems: ***REMOVED***
                                type: ["integer", "null"],
                                minimum: 0
                            ***REMOVED***
                        ***REMOVED***,
                        additionalProperties: false
                    ***REMOVED***
                ]
            ***REMOVED***
        ],

        messages: ***REMOVED***
            unexpectedLineBreak: "There should be no linebreak here.",
            missingLineBreak: "There should be a linebreak after this element."
        ***REMOVED***
    ***REMOVED***,

    create(context) ***REMOVED***
        const sourceCode = context.getSourceCode();

        //----------------------------------------------------------------------
        // Helpers
        //----------------------------------------------------------------------

        /**
         * Normalizes a given option value.
         *
         * @param ***REMOVED***string|Object|undefined***REMOVED*** providedOption - An option value to parse.
         * @returns ***REMOVED******REMOVED***multiline: boolean, minItems: number***REMOVED******REMOVED*** Normalized option object.
         */
        function normalizeOptionValue(providedOption) ***REMOVED***
            let consistent = false;
            let multiline = false;
            let minItems;

            const option = providedOption || "always";

            if (!option || option === "always" || option.minItems === 0) ***REMOVED***
                minItems = 0;
            ***REMOVED*** else if (option === "never") ***REMOVED***
                minItems = Number.POSITIVE_INFINITY;
            ***REMOVED*** else if (option === "consistent") ***REMOVED***
                consistent = true;
                minItems = Number.POSITIVE_INFINITY;
            ***REMOVED*** else ***REMOVED***
                multiline = Boolean(option.multiline);
                minItems = option.minItems || Number.POSITIVE_INFINITY;
            ***REMOVED***

            return ***REMOVED*** consistent, multiline, minItems ***REMOVED***;
        ***REMOVED***

        /**
         * Normalizes a given option value.
         *
         * @param ***REMOVED***string|Object|undefined***REMOVED*** options - An option value to parse.
         * @returns ***REMOVED******REMOVED***ArrayExpression: ***REMOVED***multiline: boolean, minItems: number***REMOVED***, ArrayPattern: ***REMOVED***multiline: boolean, minItems: number***REMOVED******REMOVED******REMOVED*** Normalized option object.
         */
        function normalizeOptions(options) ***REMOVED***
            const value = normalizeOptionValue(options);

            return ***REMOVED*** ArrayExpression: value, ArrayPattern: value ***REMOVED***;
        ***REMOVED***

        /**
         * Reports that there shouldn't be a line break after the first token
         * @param ***REMOVED***Token***REMOVED*** token - The token to use for the report.
         * @returns ***REMOVED***void***REMOVED***
         */
        function reportNoLineBreak(token) ***REMOVED***
            const tokenBefore = sourceCode.getTokenBefore(token, ***REMOVED*** includeComments: true ***REMOVED***);

            context.report(***REMOVED***
                loc: ***REMOVED***
                    start: tokenBefore.loc.end,
                    end: token.loc.start
                ***REMOVED***,
                messageId: "unexpectedLineBreak",
                fix(fixer) ***REMOVED***
                    if (astUtils.isCommentToken(tokenBefore)) ***REMOVED***
                        return null;
                    ***REMOVED***

                    if (!astUtils.isTokenOnSameLine(tokenBefore, token)) ***REMOVED***
                        return fixer.replaceTextRange([tokenBefore.range[1], token.range[0]], " ");
                    ***REMOVED***

                    /*
                     * This will check if the comma is on the same line as the next element
                     * Following array:
                     * [
                     *     1
                     *     , 2
                     *     , 3
                     * ]
                     *
                     * will be fixed to:
                     * [
                     *     1, 2, 3
                     * ]
                     */
                    const twoTokensBefore = sourceCode.getTokenBefore(tokenBefore, ***REMOVED*** includeComments: true ***REMOVED***);

                    if (astUtils.isCommentToken(twoTokensBefore)) ***REMOVED***
                        return null;
                    ***REMOVED***

                    return fixer.replaceTextRange([twoTokensBefore.range[1], tokenBefore.range[0]], "");

                ***REMOVED***
            ***REMOVED***);
        ***REMOVED***

        /**
         * Reports that there should be a line break after the first token
         * @param ***REMOVED***Token***REMOVED*** token - The token to use for the report.
         * @returns ***REMOVED***void***REMOVED***
         */
        function reportRequiredLineBreak(token) ***REMOVED***
            const tokenBefore = sourceCode.getTokenBefore(token, ***REMOVED*** includeComments: true ***REMOVED***);

            context.report(***REMOVED***
                loc: ***REMOVED***
                    start: tokenBefore.loc.end,
                    end: token.loc.start
                ***REMOVED***,
                messageId: "missingLineBreak",
                fix(fixer) ***REMOVED***
                    return fixer.replaceTextRange([tokenBefore.range[1], token.range[0]], "\n");
                ***REMOVED***
            ***REMOVED***);
        ***REMOVED***

        /**
         * Reports a given node if it violated this rule.
         *
         * @param ***REMOVED***ASTNode***REMOVED*** node - A node to check. This is an ObjectExpression node or an ObjectPattern node.
         * @returns ***REMOVED***void***REMOVED***
         */
        function check(node) ***REMOVED***
            const elements = node.elements;
            const normalizedOptions = normalizeOptions(context.options[0]);
            const options = normalizedOptions[node.type];

            let elementBreak = false;

            /*
             * MULTILINE: true
             * loop through every element and check
             * if at least one element has linebreaks inside
             * this ensures that following is not valid (due to elements are on the same line):
             *
             * [
             *      1,
             *      2,
             *      3
             * ]
             */
            if (options.multiline) ***REMOVED***
                elementBreak = elements
                    .filter(element => element !== null)
                    .some(element => element.loc.start.line !== element.loc.end.line);
            ***REMOVED***

            const linebreaksCount = node.elements.map((element, i) => ***REMOVED***
                const previousElement = elements[i - 1];

                if (i === 0 || element === null || previousElement === null) ***REMOVED***
                    return false;
                ***REMOVED***

                const commaToken = sourceCode.getFirstTokenBetween(previousElement, element, astUtils.isCommaToken);
                const lastTokenOfPreviousElement = sourceCode.getTokenBefore(commaToken);
                const firstTokenOfCurrentElement = sourceCode.getTokenAfter(commaToken);

                return !astUtils.isTokenOnSameLine(lastTokenOfPreviousElement, firstTokenOfCurrentElement);
            ***REMOVED***).filter(isBreak => isBreak === true).length;

            const needsLinebreaks = (
                elements.length >= options.minItems ||
                (
                    options.multiline &&
                    elementBreak
                ) ||
                (
                    options.consistent &&
                    linebreaksCount > 0 &&
                    linebreaksCount < node.elements.length
                )
            );

            elements.forEach((element, i) => ***REMOVED***
                const previousElement = elements[i - 1];

                if (i === 0 || element === null || previousElement === null) ***REMOVED***
                    return;
                ***REMOVED***

                const commaToken = sourceCode.getFirstTokenBetween(previousElement, element, astUtils.isCommaToken);
                const lastTokenOfPreviousElement = sourceCode.getTokenBefore(commaToken);
                const firstTokenOfCurrentElement = sourceCode.getTokenAfter(commaToken);

                if (needsLinebreaks) ***REMOVED***
                    if (astUtils.isTokenOnSameLine(lastTokenOfPreviousElement, firstTokenOfCurrentElement)) ***REMOVED***
                        reportRequiredLineBreak(firstTokenOfCurrentElement);
                    ***REMOVED***
                ***REMOVED*** else ***REMOVED***
                    if (!astUtils.isTokenOnSameLine(lastTokenOfPreviousElement, firstTokenOfCurrentElement)) ***REMOVED***
                        reportNoLineBreak(firstTokenOfCurrentElement);
                    ***REMOVED***
                ***REMOVED***
            ***REMOVED***);
        ***REMOVED***

        //----------------------------------------------------------------------
        // Public
        //----------------------------------------------------------------------

        return ***REMOVED***
            ArrayPattern: check,
            ArrayExpression: check
        ***REMOVED***;
    ***REMOVED***
***REMOVED***;
