// eslint/lib/rules/array-bracket-spacing.js
/**
 * @fileoverview Disallows or enforces spaces inside of array brackets.
 * @author Jamund Ferguson
 */
"use strict";

const astUtils = require("./utils/ast-utils");

//------------------------------------------------------------------------------
// Rule Definition
//------------------------------------------------------------------------------

module.exports = ***REMOVED***
    meta: ***REMOVED***
        type: "layout",

        docs: ***REMOVED***
            description: "enforce consistent spacing inside array brackets",
            category: "Stylistic Issues",
            recommended: false,
            url: "https://eslint.org/docs/rules/array-bracket-spacing"
        ***REMOVED***,

        fixable: "whitespace",

        schema: [
            ***REMOVED***
                enum: ["always", "never"]
            ***REMOVED***,
            ***REMOVED***
                type: "object",
                properties: ***REMOVED***
                    singleValue: ***REMOVED***
                        type: "boolean"
                    ***REMOVED***,
                    objectsInArrays: ***REMOVED***
                        type: "boolean"
                    ***REMOVED***,
                    arraysInArrays: ***REMOVED***
                        type: "boolean"
                    ***REMOVED***
                ***REMOVED***,
                additionalProperties: false
            ***REMOVED***
        ],

        messages: ***REMOVED***
            unexpectedSpaceAfter: "There should be no space after '***REMOVED******REMOVED***tokenValue***REMOVED******REMOVED***'.",
            unexpectedSpaceBefore: "There should be no space before '***REMOVED******REMOVED***tokenValue***REMOVED******REMOVED***'.",
            missingSpaceAfter: "A space is required after '***REMOVED******REMOVED***tokenValue***REMOVED******REMOVED***'.",
            missingSpaceBefore: "A space is required before '***REMOVED******REMOVED***tokenValue***REMOVED******REMOVED***'."
        ***REMOVED***
    ***REMOVED***,
    create(context) ***REMOVED***
        const spaced = context.options[0] === "always",
            sourceCode = context.getSourceCode();

        /**
         * Determines whether an option is set, relative to the spacing option.
         * If spaced is "always", then check whether option is set to false.
         * If spaced is "never", then check whether option is set to true.
         * @param ***REMOVED***Object***REMOVED*** option - The option to exclude.
         * @returns ***REMOVED***boolean***REMOVED*** Whether or not the property is excluded.
         */
        function isOptionSet(option) ***REMOVED***
            return context.options[1] ? context.options[1][option] === !spaced : false;
        ***REMOVED***

        const options = ***REMOVED***
            spaced,
            singleElementException: isOptionSet("singleValue"),
            objectsInArraysException: isOptionSet("objectsInArrays"),
            arraysInArraysException: isOptionSet("arraysInArrays")
        ***REMOVED***;

        //--------------------------------------------------------------------------
        // Helpers
        //--------------------------------------------------------------------------

        /**
         * Reports that there shouldn't be a space after the first token
         * @param ***REMOVED***ASTNode***REMOVED*** node - The node to report in the event of an error.
         * @param ***REMOVED***Token***REMOVED*** token - The token to use for the report.
         * @returns ***REMOVED***void***REMOVED***
         */
        function reportNoBeginningSpace(node, token) ***REMOVED***
            context.report(***REMOVED***
                node,
                loc: token.loc.start,
                messageId: "unexpectedSpaceAfter",
                data: ***REMOVED***
                    tokenValue: token.value
                ***REMOVED***,
                fix(fixer) ***REMOVED***
                    const nextToken = sourceCode.getTokenAfter(token);

                    return fixer.removeRange([token.range[1], nextToken.range[0]]);
                ***REMOVED***
            ***REMOVED***);
        ***REMOVED***

        /**
         * Reports that there shouldn't be a space before the last token
         * @param ***REMOVED***ASTNode***REMOVED*** node - The node to report in the event of an error.
         * @param ***REMOVED***Token***REMOVED*** token - The token to use for the report.
         * @returns ***REMOVED***void***REMOVED***
         */
        function reportNoEndingSpace(node, token) ***REMOVED***
            context.report(***REMOVED***
                node,
                loc: token.loc.start,
                messageId: "unexpectedSpaceBefore",
                data: ***REMOVED***
                    tokenValue: token.value
                ***REMOVED***,
                fix(fixer) ***REMOVED***
                    const previousToken = sourceCode.getTokenBefore(token);

                    return fixer.removeRange([previousToken.range[1], token.range[0]]);
                ***REMOVED***
            ***REMOVED***);
        ***REMOVED***

        /**
         * Reports that there should be a space after the first token
         * @param ***REMOVED***ASTNode***REMOVED*** node - The node to report in the event of an error.
         * @param ***REMOVED***Token***REMOVED*** token - The token to use for the report.
         * @returns ***REMOVED***void***REMOVED***
         */
        function reportRequiredBeginningSpace(node, token) ***REMOVED***
            context.report(***REMOVED***
                node,
                loc: token.loc.start,
                messageId: "missingSpaceAfter",
                data: ***REMOVED***
                    tokenValue: token.value
                ***REMOVED***,
                fix(fixer) ***REMOVED***
                    return fixer.insertTextAfter(token, " ");
                ***REMOVED***
            ***REMOVED***);
        ***REMOVED***

        /**
         * Reports that there should be a space before the last token
         * @param ***REMOVED***ASTNode***REMOVED*** node - The node to report in the event of an error.
         * @param ***REMOVED***Token***REMOVED*** token - The token to use for the report.
         * @returns ***REMOVED***void***REMOVED***
         */
        function reportRequiredEndingSpace(node, token) ***REMOVED***
            context.report(***REMOVED***
                node,
                loc: token.loc.start,
                messageId: "missingSpaceBefore",
                data: ***REMOVED***
                    tokenValue: token.value
                ***REMOVED***,
                fix(fixer) ***REMOVED***
                    return fixer.insertTextBefore(token, " ");
                ***REMOVED***
            ***REMOVED***);
        ***REMOVED***

        /**
         * Determines if a node is an object type
         * @param ***REMOVED***ASTNode***REMOVED*** node - The node to check.
         * @returns ***REMOVED***boolean***REMOVED*** Whether or not the node is an object type.
         */
        function isObjectType(node) ***REMOVED***
            return node && (node.type === "ObjectExpression" || node.type === "ObjectPattern");
        ***REMOVED***

        /**
         * Determines if a node is an array type
         * @param ***REMOVED***ASTNode***REMOVED*** node - The node to check.
         * @returns ***REMOVED***boolean***REMOVED*** Whether or not the node is an array type.
         */
        function isArrayType(node) ***REMOVED***
            return node && (node.type === "ArrayExpression" || node.type === "ArrayPattern");
        ***REMOVED***

        /**
         * Validates the spacing around array brackets
         * @param ***REMOVED***ASTNode***REMOVED*** node - The node we're checking for spacing
         * @returns ***REMOVED***void***REMOVED***
         */
        function validateArraySpacing(node) ***REMOVED***
            if (options.spaced && node.elements.length === 0) ***REMOVED***
                return;
            ***REMOVED***

            const first = sourceCode.getFirstToken(node),
                second = sourceCode.getFirstToken(node, 1),
                last = node.typeAnnotation
                    ? sourceCode.getTokenBefore(node.typeAnnotation)
                    : sourceCode.getLastToken(node),
                penultimate = sourceCode.getTokenBefore(last),
                firstElement = node.elements[0],
                lastElement = node.elements[node.elements.length - 1];

            const openingBracketMustBeSpaced =
                options.objectsInArraysException && isObjectType(firstElement) ||
                options.arraysInArraysException && isArrayType(firstElement) ||
                options.singleElementException && node.elements.length === 1
                    ? !options.spaced : options.spaced;

            const closingBracketMustBeSpaced =
                options.objectsInArraysException && isObjectType(lastElement) ||
                options.arraysInArraysException && isArrayType(lastElement) ||
                options.singleElementException && node.elements.length === 1
                    ? !options.spaced : options.spaced;

            if (astUtils.isTokenOnSameLine(first, second)) ***REMOVED***
                if (openingBracketMustBeSpaced && !sourceCode.isSpaceBetweenTokens(first, second)) ***REMOVED***
                    reportRequiredBeginningSpace(node, first);
                ***REMOVED***
                if (!openingBracketMustBeSpaced && sourceCode.isSpaceBetweenTokens(first, second)) ***REMOVED***
                    reportNoBeginningSpace(node, first);
                ***REMOVED***
            ***REMOVED***

            if (first !== penultimate && astUtils.isTokenOnSameLine(penultimate, last)) ***REMOVED***
                if (closingBracketMustBeSpaced && !sourceCode.isSpaceBetweenTokens(penultimate, last)) ***REMOVED***
                    reportRequiredEndingSpace(node, last);
                ***REMOVED***
                if (!closingBracketMustBeSpaced && sourceCode.isSpaceBetweenTokens(penultimate, last)) ***REMOVED***
                    reportNoEndingSpace(node, last);
                ***REMOVED***
            ***REMOVED***
        ***REMOVED***

        //--------------------------------------------------------------------------
        // Public
        //--------------------------------------------------------------------------

        return ***REMOVED***
            ArrayPattern: validateArraySpacing,
            ArrayExpression: validateArraySpacing
        ***REMOVED***;
    ***REMOVED***
***REMOVED***;
