// eslint/lib/rules/arrow-spacing.js
/**
 * @fileoverview Rule to define spacing before/after arrow function's arrow.
 * @author Jxck
 */
"use strict";

//------------------------------------------------------------------------------
// Requirements
//------------------------------------------------------------------------------

const astUtils = require("./utils/ast-utils");

//------------------------------------------------------------------------------
// Rule Definition
//------------------------------------------------------------------------------

module.exports = ***REMOVED***
    meta: ***REMOVED***
        type: "layout",

        docs: ***REMOVED***
            description: "enforce consistent spacing before and after the arrow in arrow functions",
            category: "ECMAScript 6",
            recommended: false,
            url: "https://eslint.org/docs/rules/arrow-spacing"
        ***REMOVED***,

        fixable: "whitespace",

        schema: [
            ***REMOVED***
                type: "object",
                properties: ***REMOVED***
                    before: ***REMOVED***
                        type: "boolean",
                        default: true
                    ***REMOVED***,
                    after: ***REMOVED***
                        type: "boolean",
                        default: true
                    ***REMOVED***
                ***REMOVED***,
                additionalProperties: false
            ***REMOVED***
        ],

        messages: ***REMOVED***
            expectedBefore: "Missing space before =>.",
            unexpectedBefore: "Unexpected space before =>.",

            expectedAfter: "Missing space after =>.",
            unexpectedAfter: "Unexpected space after =>."
        ***REMOVED***
    ***REMOVED***,

    create(context) ***REMOVED***

        // merge rules with default
        const rule = Object.assign(***REMOVED******REMOVED***, context.options[0]);

        rule.before = rule.before !== false;
        rule.after = rule.after !== false;

        const sourceCode = context.getSourceCode();

        /**
         * Get tokens of arrow(`=>`) and before/after arrow.
         * @param ***REMOVED***ASTNode***REMOVED*** node The arrow function node.
         * @returns ***REMOVED***Object***REMOVED*** Tokens of arrow and before/after arrow.
         */
        function getTokens(node) ***REMOVED***
            const arrow = sourceCode.getTokenBefore(node.body, astUtils.isArrowToken);

            return ***REMOVED***
                before: sourceCode.getTokenBefore(arrow),
                arrow,
                after: sourceCode.getTokenAfter(arrow)
            ***REMOVED***;
        ***REMOVED***

        /**
         * Count spaces before/after arrow(`=>`) token.
         * @param ***REMOVED***Object***REMOVED*** tokens Tokens before/after arrow.
         * @returns ***REMOVED***Object***REMOVED*** count of space before/after arrow.
         */
        function countSpaces(tokens) ***REMOVED***
            const before = tokens.arrow.range[0] - tokens.before.range[1];
            const after = tokens.after.range[0] - tokens.arrow.range[1];

            return ***REMOVED*** before, after ***REMOVED***;
        ***REMOVED***

        /**
         * Determines whether space(s) before after arrow(`=>`) is satisfy rule.
         * if before/after value is `true`, there should be space(s).
         * if before/after value is `false`, there should be no space.
         * @param ***REMOVED***ASTNode***REMOVED*** node The arrow function node.
         * @returns ***REMOVED***void***REMOVED***
         */
        function spaces(node) ***REMOVED***
            const tokens = getTokens(node);
            const countSpace = countSpaces(tokens);

            if (rule.before) ***REMOVED***

                // should be space(s) before arrow
                if (countSpace.before === 0) ***REMOVED***
                    context.report(***REMOVED***
                        node: tokens.before,
                        messageId: "expectedBefore",
                        fix(fixer) ***REMOVED***
                            return fixer.insertTextBefore(tokens.arrow, " ");
                        ***REMOVED***
                    ***REMOVED***);
                ***REMOVED***
            ***REMOVED*** else ***REMOVED***

                // should be no space before arrow
                if (countSpace.before > 0) ***REMOVED***
                    context.report(***REMOVED***
                        node: tokens.before,
                        messageId: "unexpectedBefore",
                        fix(fixer) ***REMOVED***
                            return fixer.removeRange([tokens.before.range[1], tokens.arrow.range[0]]);
                        ***REMOVED***
                    ***REMOVED***);
                ***REMOVED***
            ***REMOVED***

            if (rule.after) ***REMOVED***

                // should be space(s) after arrow
                if (countSpace.after === 0) ***REMOVED***
                    context.report(***REMOVED***
                        node: tokens.after,
                        messageId: "expectedAfter",
                        fix(fixer) ***REMOVED***
                            return fixer.insertTextAfter(tokens.arrow, " ");
                        ***REMOVED***
                    ***REMOVED***);
                ***REMOVED***
            ***REMOVED*** else ***REMOVED***

                // should be no space after arrow
                if (countSpace.after > 0) ***REMOVED***
                    context.report(***REMOVED***
                        node: tokens.after,
                        messageId: "unexpectedAfter",
                        fix(fixer) ***REMOVED***
                            return fixer.removeRange([tokens.arrow.range[1], tokens.after.range[0]]);
                        ***REMOVED***
                    ***REMOVED***);
                ***REMOVED***
            ***REMOVED***
        ***REMOVED***

        return ***REMOVED***
            ArrowFunctionExpression: spaces
        ***REMOVED***;
    ***REMOVED***
***REMOVED***;
