// eslint/lib/rules/comma-dangle.js
/**
 * @fileoverview Rule to forbid or enforce dangling commas.
 * @author Ian Christian Myers
 */

"use strict";

//------------------------------------------------------------------------------
// Requirements
//------------------------------------------------------------------------------

const lodash = require("lodash");
const astUtils = require("./utils/ast-utils");

//------------------------------------------------------------------------------
// Helpers
//------------------------------------------------------------------------------

const DEFAULT_OPTIONS = Object.freeze(***REMOVED***
    arrays: "never",
    objects: "never",
    imports: "never",
    exports: "never",
    functions: "never"
***REMOVED***);

/**
 * Checks whether or not a trailing comma is allowed in a given node.
 * If the `lastItem` is `RestElement` or `RestProperty`, it disallows trailing commas.
 *
 * @param ***REMOVED***ASTNode***REMOVED*** lastItem - The node of the last element in the given node.
 * @returns ***REMOVED***boolean***REMOVED*** `true` if a trailing comma is allowed.
 */
function isTrailingCommaAllowed(lastItem) ***REMOVED***
    return !(
        lastItem.type === "RestElement" ||
        lastItem.type === "RestProperty" ||
        lastItem.type === "ExperimentalRestProperty"
    );
***REMOVED***

/**
 * Normalize option value.
 *
 * @param ***REMOVED***string|Object|undefined***REMOVED*** optionValue - The 1st option value to normalize.
 * @returns ***REMOVED***Object***REMOVED*** The normalized option value.
 */
function normalizeOptions(optionValue) ***REMOVED***
    if (typeof optionValue === "string") ***REMOVED***
        return ***REMOVED***
            arrays: optionValue,
            objects: optionValue,
            imports: optionValue,
            exports: optionValue,

            // For backward compatibility, always ignore functions.
            functions: "ignore"
        ***REMOVED***;
    ***REMOVED***
    if (typeof optionValue === "object" && optionValue !== null) ***REMOVED***
        return ***REMOVED***
            arrays: optionValue.arrays || DEFAULT_OPTIONS.arrays,
            objects: optionValue.objects || DEFAULT_OPTIONS.objects,
            imports: optionValue.imports || DEFAULT_OPTIONS.imports,
            exports: optionValue.exports || DEFAULT_OPTIONS.exports,
            functions: optionValue.functions || DEFAULT_OPTIONS.functions
        ***REMOVED***;
    ***REMOVED***

    return DEFAULT_OPTIONS;
***REMOVED***

//------------------------------------------------------------------------------
// Rule Definition
//------------------------------------------------------------------------------

module.exports = ***REMOVED***
    meta: ***REMOVED***
        type: "layout",

        docs: ***REMOVED***
            description: "require or disallow trailing commas",
            category: "Stylistic Issues",
            recommended: false,
            url: "https://eslint.org/docs/rules/comma-dangle"
        ***REMOVED***,

        fixable: "code",

        schema: ***REMOVED***
            definitions: ***REMOVED***
                value: ***REMOVED***
                    enum: [
                        "always-multiline",
                        "always",
                        "never",
                        "only-multiline"
                    ]
                ***REMOVED***,
                valueWithIgnore: ***REMOVED***
                    enum: [
                        "always-multiline",
                        "always",
                        "ignore",
                        "never",
                        "only-multiline"
                    ]
                ***REMOVED***
            ***REMOVED***,
            type: "array",
            items: [
                ***REMOVED***
                    oneOf: [
                        ***REMOVED***
                            $ref: "#/definitions/value"
                        ***REMOVED***,
                        ***REMOVED***
                            type: "object",
                            properties: ***REMOVED***
                                arrays: ***REMOVED*** $ref: "#/definitions/valueWithIgnore" ***REMOVED***,
                                objects: ***REMOVED*** $ref: "#/definitions/valueWithIgnore" ***REMOVED***,
                                imports: ***REMOVED*** $ref: "#/definitions/valueWithIgnore" ***REMOVED***,
                                exports: ***REMOVED*** $ref: "#/definitions/valueWithIgnore" ***REMOVED***,
                                functions: ***REMOVED*** $ref: "#/definitions/valueWithIgnore" ***REMOVED***
                            ***REMOVED***,
                            additionalProperties: false
                        ***REMOVED***
                    ]
                ***REMOVED***
            ]
        ***REMOVED***,

        messages: ***REMOVED***
            unexpected: "Unexpected trailing comma.",
            missing: "Missing trailing comma."
        ***REMOVED***
    ***REMOVED***,

    create(context) ***REMOVED***
        const options = normalizeOptions(context.options[0]);
        const sourceCode = context.getSourceCode();

        /**
         * Gets the last item of the given node.
         * @param ***REMOVED***ASTNode***REMOVED*** node - The node to get.
         * @returns ***REMOVED***ASTNode|null***REMOVED*** The last node or null.
         */
        function getLastItem(node) ***REMOVED***
            switch (node.type) ***REMOVED***
                case "ObjectExpression":
                case "ObjectPattern":
                    return lodash.last(node.properties);
                case "ArrayExpression":
                case "ArrayPattern":
                    return lodash.last(node.elements);
                case "ImportDeclaration":
                case "ExportNamedDeclaration":
                    return lodash.last(node.specifiers);
                case "FunctionDeclaration":
                case "FunctionExpression":
                case "ArrowFunctionExpression":
                    return lodash.last(node.params);
                case "CallExpression":
                case "NewExpression":
                    return lodash.last(node.arguments);
                default:
                    return null;
            ***REMOVED***
        ***REMOVED***

        /**
         * Gets the trailing comma token of the given node.
         * If the trailing comma does not exist, this returns the token which is
         * the insertion point of the trailing comma token.
         *
         * @param ***REMOVED***ASTNode***REMOVED*** node - The node to get.
         * @param ***REMOVED***ASTNode***REMOVED*** lastItem - The last item of the node.
         * @returns ***REMOVED***Token***REMOVED*** The trailing comma token or the insertion point.
         */
        function getTrailingToken(node, lastItem) ***REMOVED***
            switch (node.type) ***REMOVED***
                case "ObjectExpression":
                case "ArrayExpression":
                case "CallExpression":
                case "NewExpression":
                    return sourceCode.getLastToken(node, 1);
                default: ***REMOVED***
                    const nextToken = sourceCode.getTokenAfter(lastItem);

                    if (astUtils.isCommaToken(nextToken)) ***REMOVED***
                        return nextToken;
                    ***REMOVED***
                    return sourceCode.getLastToken(lastItem);
                ***REMOVED***
            ***REMOVED***
        ***REMOVED***

        /**
         * Checks whether or not a given node is multiline.
         * This rule handles a given node as multiline when the closing parenthesis
         * and the last element are not on the same line.
         *
         * @param ***REMOVED***ASTNode***REMOVED*** node - A node to check.
         * @returns ***REMOVED***boolean***REMOVED*** `true` if the node is multiline.
         */
        function isMultiline(node) ***REMOVED***
            const lastItem = getLastItem(node);

            if (!lastItem) ***REMOVED***
                return false;
            ***REMOVED***

            const penultimateToken = getTrailingToken(node, lastItem);
            const lastToken = sourceCode.getTokenAfter(penultimateToken);

            return lastToken.loc.end.line !== penultimateToken.loc.end.line;
        ***REMOVED***

        /**
         * Reports a trailing comma if it exists.
         *
         * @param ***REMOVED***ASTNode***REMOVED*** node - A node to check. Its type is one of
         *   ObjectExpression, ObjectPattern, ArrayExpression, ArrayPattern,
         *   ImportDeclaration, and ExportNamedDeclaration.
         * @returns ***REMOVED***void***REMOVED***
         */
        function forbidTrailingComma(node) ***REMOVED***
            const lastItem = getLastItem(node);

            if (!lastItem || (node.type === "ImportDeclaration" && lastItem.type !== "ImportSpecifier")) ***REMOVED***
                return;
            ***REMOVED***

            const trailingToken = getTrailingToken(node, lastItem);

            if (astUtils.isCommaToken(trailingToken)) ***REMOVED***
                context.report(***REMOVED***
                    node: lastItem,
                    loc: trailingToken.loc.start,
                    messageId: "unexpected",
                    fix(fixer) ***REMOVED***
                        return fixer.remove(trailingToken);
                    ***REMOVED***
                ***REMOVED***);
            ***REMOVED***
        ***REMOVED***

        /**
         * Reports the last element of a given node if it does not have a trailing
         * comma.
         *
         * If a given node is `ArrayPattern` which has `RestElement`, the trailing
         * comma is disallowed, so report if it exists.
         *
         * @param ***REMOVED***ASTNode***REMOVED*** node - A node to check. Its type is one of
         *   ObjectExpression, ObjectPattern, ArrayExpression, ArrayPattern,
         *   ImportDeclaration, and ExportNamedDeclaration.
         * @returns ***REMOVED***void***REMOVED***
         */
        function forceTrailingComma(node) ***REMOVED***
            const lastItem = getLastItem(node);

            if (!lastItem || (node.type === "ImportDeclaration" && lastItem.type !== "ImportSpecifier")) ***REMOVED***
                return;
            ***REMOVED***
            if (!isTrailingCommaAllowed(lastItem)) ***REMOVED***
                forbidTrailingComma(node);
                return;
            ***REMOVED***

            const trailingToken = getTrailingToken(node, lastItem);

            if (trailingToken.value !== ",") ***REMOVED***
                context.report(***REMOVED***
                    node: lastItem,
                    loc: trailingToken.loc.end,
                    messageId: "missing",
                    fix(fixer) ***REMOVED***
                        return fixer.insertTextAfter(trailingToken, ",");
                    ***REMOVED***
                ***REMOVED***);
            ***REMOVED***
        ***REMOVED***

        /**
         * If a given node is multiline, reports the last element of a given node
         * when it does not have a trailing comma.
         * Otherwise, reports a trailing comma if it exists.
         *
         * @param ***REMOVED***ASTNode***REMOVED*** node - A node to check. Its type is one of
         *   ObjectExpression, ObjectPattern, ArrayExpression, ArrayPattern,
         *   ImportDeclaration, and ExportNamedDeclaration.
         * @returns ***REMOVED***void***REMOVED***
         */
        function forceTrailingCommaIfMultiline(node) ***REMOVED***
            if (isMultiline(node)) ***REMOVED***
                forceTrailingComma(node);
            ***REMOVED*** else ***REMOVED***
                forbidTrailingComma(node);
            ***REMOVED***
        ***REMOVED***

        /**
         * Only if a given node is not multiline, reports the last element of a given node
         * when it does not have a trailing comma.
         * Otherwise, reports a trailing comma if it exists.
         *
         * @param ***REMOVED***ASTNode***REMOVED*** node - A node to check. Its type is one of
         *   ObjectExpression, ObjectPattern, ArrayExpression, ArrayPattern,
         *   ImportDeclaration, and ExportNamedDeclaration.
         * @returns ***REMOVED***void***REMOVED***
         */
        function allowTrailingCommaIfMultiline(node) ***REMOVED***
            if (!isMultiline(node)) ***REMOVED***
                forbidTrailingComma(node);
            ***REMOVED***
        ***REMOVED***

        const predicate = ***REMOVED***
            always: forceTrailingComma,
            "always-multiline": forceTrailingCommaIfMultiline,
            "only-multiline": allowTrailingCommaIfMultiline,
            never: forbidTrailingComma,
            ignore: lodash.noop
        ***REMOVED***;

        return ***REMOVED***
            ObjectExpression: predicate[options.objects],
            ObjectPattern: predicate[options.objects],

            ArrayExpression: predicate[options.arrays],
            ArrayPattern: predicate[options.arrays],

            ImportDeclaration: predicate[options.imports],

            ExportNamedDeclaration: predicate[options.exports],

            FunctionDeclaration: predicate[options.functions],
            FunctionExpression: predicate[options.functions],
            ArrowFunctionExpression: predicate[options.functions],
            CallExpression: predicate[options.functions],
            NewExpression: predicate[options.functions]
        ***REMOVED***;
    ***REMOVED***
***REMOVED***;
