// eslint/lib/rules/computed-property-spacing.js
/**
 * @fileoverview Disallows or enforces spaces inside computed properties.
 * @author Jamund Ferguson
 */
"use strict";

const astUtils = require("./utils/ast-utils");

//------------------------------------------------------------------------------
// Rule Definition
//------------------------------------------------------------------------------

module.exports = ***REMOVED***
    meta: ***REMOVED***
        type: "layout",

        docs: ***REMOVED***
            description: "enforce consistent spacing inside computed property brackets",
            category: "Stylistic Issues",
            recommended: false,
            url: "https://eslint.org/docs/rules/computed-property-spacing"
        ***REMOVED***,

        fixable: "whitespace",

        schema: [
            ***REMOVED***
                enum: ["always", "never"]
            ***REMOVED***,
            ***REMOVED***
                type: "object",
                properties: ***REMOVED***
                    enforceForClassMembers: ***REMOVED***
                        type: "boolean",
                        default: false
                    ***REMOVED***
                ***REMOVED***,
                additionalProperties: false
            ***REMOVED***
        ],

        messages: ***REMOVED***
            unexpectedSpaceBefore: "There should be no space before '***REMOVED******REMOVED***tokenValue***REMOVED******REMOVED***'.",
            unexpectedSpaceAfter: "There should be no space after '***REMOVED******REMOVED***tokenValue***REMOVED******REMOVED***'.",

            missingSpaceBefore: "A space is required before '***REMOVED******REMOVED***tokenValue***REMOVED******REMOVED***'.",
            missingSpaceAfter: "A space is required after '***REMOVED******REMOVED***tokenValue***REMOVED******REMOVED***'."
        ***REMOVED***
    ***REMOVED***,

    create(context) ***REMOVED***
        const sourceCode = context.getSourceCode();
        const propertyNameMustBeSpaced = context.options[0] === "always"; // default is "never"
        const enforceForClassMembers = context.options[1] && context.options[1].enforceForClassMembers;

        //--------------------------------------------------------------------------
        // Helpers
        //--------------------------------------------------------------------------

        /**
         * Reports that there shouldn't be a space after the first token
         * @param ***REMOVED***ASTNode***REMOVED*** node - The node to report in the event of an error.
         * @param ***REMOVED***Token***REMOVED*** token - The token to use for the report.
         * @param ***REMOVED***Token***REMOVED*** tokenAfter - The token after `token`.
         * @returns ***REMOVED***void***REMOVED***
         */
        function reportNoBeginningSpace(node, token, tokenAfter) ***REMOVED***
            context.report(***REMOVED***
                node,
                loc: token.loc.start,
                messageId: "unexpectedSpaceAfter",
                data: ***REMOVED***
                    tokenValue: token.value
                ***REMOVED***,
                fix(fixer) ***REMOVED***
                    return fixer.removeRange([token.range[1], tokenAfter.range[0]]);
                ***REMOVED***
            ***REMOVED***);
        ***REMOVED***

        /**
         * Reports that there shouldn't be a space before the last token
         * @param ***REMOVED***ASTNode***REMOVED*** node - The node to report in the event of an error.
         * @param ***REMOVED***Token***REMOVED*** token - The token to use for the report.
         * @param ***REMOVED***Token***REMOVED*** tokenBefore - The token before `token`.
         * @returns ***REMOVED***void***REMOVED***
         */
        function reportNoEndingSpace(node, token, tokenBefore) ***REMOVED***
            context.report(***REMOVED***
                node,
                loc: token.loc.start,
                messageId: "unexpectedSpaceBefore",
                data: ***REMOVED***
                    tokenValue: token.value
                ***REMOVED***,
                fix(fixer) ***REMOVED***
                    return fixer.removeRange([tokenBefore.range[1], token.range[0]]);
                ***REMOVED***
            ***REMOVED***);
        ***REMOVED***

        /**
         * Reports that there should be a space after the first token
         * @param ***REMOVED***ASTNode***REMOVED*** node - The node to report in the event of an error.
         * @param ***REMOVED***Token***REMOVED*** token - The token to use for the report.
         * @returns ***REMOVED***void***REMOVED***
         */
        function reportRequiredBeginningSpace(node, token) ***REMOVED***
            context.report(***REMOVED***
                node,
                loc: token.loc.start,
                messageId: "missingSpaceAfter",
                data: ***REMOVED***
                    tokenValue: token.value
                ***REMOVED***,
                fix(fixer) ***REMOVED***
                    return fixer.insertTextAfter(token, " ");
                ***REMOVED***
            ***REMOVED***);
        ***REMOVED***

        /**
         * Reports that there should be a space before the last token
         * @param ***REMOVED***ASTNode***REMOVED*** node - The node to report in the event of an error.
         * @param ***REMOVED***Token***REMOVED*** token - The token to use for the report.
         * @returns ***REMOVED***void***REMOVED***
         */
        function reportRequiredEndingSpace(node, token) ***REMOVED***
            context.report(***REMOVED***
                node,
                loc: token.loc.start,
                messageId: "missingSpaceBefore",
                data: ***REMOVED***
                    tokenValue: token.value
                ***REMOVED***,
                fix(fixer) ***REMOVED***
                    return fixer.insertTextBefore(token, " ");
                ***REMOVED***
            ***REMOVED***);
        ***REMOVED***

        /**
         * Returns a function that checks the spacing of a node on the property name
         * that was passed in.
         * @param ***REMOVED***string***REMOVED*** propertyName The property on the node to check for spacing
         * @returns ***REMOVED***Function***REMOVED*** A function that will check spacing on a node
         */
        function checkSpacing(propertyName) ***REMOVED***
            return function(node) ***REMOVED***
                if (!node.computed) ***REMOVED***
                    return;
                ***REMOVED***

                const property = node[propertyName];

                const before = sourceCode.getTokenBefore(property),
                    first = sourceCode.getFirstToken(property),
                    last = sourceCode.getLastToken(property),
                    after = sourceCode.getTokenAfter(property);

                if (astUtils.isTokenOnSameLine(before, first)) ***REMOVED***
                    if (propertyNameMustBeSpaced) ***REMOVED***
                        if (!sourceCode.isSpaceBetweenTokens(before, first) && astUtils.isTokenOnSameLine(before, first)) ***REMOVED***
                            reportRequiredBeginningSpace(node, before);
                        ***REMOVED***
                    ***REMOVED*** else ***REMOVED***
                        if (sourceCode.isSpaceBetweenTokens(before, first)) ***REMOVED***
                            reportNoBeginningSpace(node, before, first);
                        ***REMOVED***
                    ***REMOVED***
                ***REMOVED***

                if (astUtils.isTokenOnSameLine(last, after)) ***REMOVED***
                    if (propertyNameMustBeSpaced) ***REMOVED***
                        if (!sourceCode.isSpaceBetweenTokens(last, after) && astUtils.isTokenOnSameLine(last, after)) ***REMOVED***
                            reportRequiredEndingSpace(node, after);
                        ***REMOVED***
                    ***REMOVED*** else ***REMOVED***
                        if (sourceCode.isSpaceBetweenTokens(last, after)) ***REMOVED***
                            reportNoEndingSpace(node, after, last);
                        ***REMOVED***
                    ***REMOVED***
                ***REMOVED***
            ***REMOVED***;
        ***REMOVED***


        //--------------------------------------------------------------------------
        // Public
        //--------------------------------------------------------------------------

        const listeners = ***REMOVED***
            Property: checkSpacing("key"),
            MemberExpression: checkSpacing("property")
        ***REMOVED***;

        if (enforceForClassMembers) ***REMOVED***
            listeners.MethodDefinition = checkSpacing("key");
        ***REMOVED***

        return listeners;

    ***REMOVED***
***REMOVED***;
