// eslint/lib/rules/dot-notation.js
/**
 * @fileoverview Rule to warn about using dot notation instead of square bracket notation when possible.
 * @author Josh Perez
 */
"use strict";

//------------------------------------------------------------------------------
// Requirements
//------------------------------------------------------------------------------

const astUtils = require("./utils/ast-utils");
const keywords = require("./utils/keywords");

//------------------------------------------------------------------------------
// Rule Definition
//------------------------------------------------------------------------------

const validIdentifier = /^[a-zA-Z_$][a-zA-Z0-9_$]*$/u;

// `null` literal must be handled separately.
const literalTypesToCheck = new Set(["string", "boolean"]);

module.exports = ***REMOVED***
    meta: ***REMOVED***
        type: "suggestion",

        docs: ***REMOVED***
            description: "enforce dot notation whenever possible",
            category: "Best Practices",
            recommended: false,
            url: "https://eslint.org/docs/rules/dot-notation"
        ***REMOVED***,

        schema: [
            ***REMOVED***
                type: "object",
                properties: ***REMOVED***
                    allowKeywords: ***REMOVED***
                        type: "boolean",
                        default: true
                    ***REMOVED***,
                    allowPattern: ***REMOVED***
                        type: "string",
                        default: ""
                    ***REMOVED***
                ***REMOVED***,
                additionalProperties: false
            ***REMOVED***
        ],

        fixable: "code",

        messages: ***REMOVED***
            useDot: "[***REMOVED******REMOVED***key***REMOVED******REMOVED***] is better written in dot notation.",
            useBrackets: ".***REMOVED******REMOVED***key***REMOVED******REMOVED*** is a syntax error."
        ***REMOVED***
    ***REMOVED***,

    create(context) ***REMOVED***
        const options = context.options[0] || ***REMOVED******REMOVED***;
        const allowKeywords = options.allowKeywords === void 0 || options.allowKeywords;
        const sourceCode = context.getSourceCode();

        let allowPattern;

        if (options.allowPattern) ***REMOVED***
            allowPattern = new RegExp(options.allowPattern, "u");
        ***REMOVED***

        /**
         * Check if the property is valid dot notation
         * @param ***REMOVED***ASTNode***REMOVED*** node The dot notation node
         * @param ***REMOVED***string***REMOVED*** value Value which is to be checked
         * @returns ***REMOVED***void***REMOVED***
         */
        function checkComputedProperty(node, value) ***REMOVED***
            if (
                validIdentifier.test(value) &&
                (allowKeywords || keywords.indexOf(String(value)) === -1) &&
                !(allowPattern && allowPattern.test(value))
            ) ***REMOVED***
                const formattedValue = node.property.type === "Literal" ? JSON.stringify(value) : `\`$***REMOVED***value***REMOVED***\``;

                context.report(***REMOVED***
                    node: node.property,
                    messageId: "useDot",
                    data: ***REMOVED***
                        key: formattedValue
                    ***REMOVED***,
                    fix(fixer) ***REMOVED***
                        const leftBracket = sourceCode.getTokenAfter(node.object, astUtils.isOpeningBracketToken);
                        const rightBracket = sourceCode.getLastToken(node);

                        if (sourceCode.getFirstTokenBetween(leftBracket, rightBracket, ***REMOVED*** includeComments: true, filter: astUtils.isCommentToken ***REMOVED***)) ***REMOVED***

                            // Don't perform any fixes if there are comments inside the brackets.
                            return null;
                        ***REMOVED***

                        const tokenAfterProperty = sourceCode.getTokenAfter(rightBracket);
                        const needsSpaceAfterProperty = tokenAfterProperty &&
                            rightBracket.range[1] === tokenAfterProperty.range[0] &&
                            !astUtils.canTokensBeAdjacent(String(value), tokenAfterProperty);

                        const textBeforeDot = astUtils.isDecimalInteger(node.object) ? " " : "";
                        const textAfterProperty = needsSpaceAfterProperty ? " " : "";

                        return fixer.replaceTextRange(
                            [leftBracket.range[0], rightBracket.range[1]],
                            `$***REMOVED***textBeforeDot***REMOVED***.$***REMOVED***value***REMOVED***$***REMOVED***textAfterProperty***REMOVED***`
                        );
                    ***REMOVED***
                ***REMOVED***);
            ***REMOVED***
        ***REMOVED***

        return ***REMOVED***
            MemberExpression(node) ***REMOVED***
                if (
                    node.computed &&
                    node.property.type === "Literal" &&
                    (literalTypesToCheck.has(typeof node.property.value) || astUtils.isNullLiteral(node.property))
                ) ***REMOVED***
                    checkComputedProperty(node, node.property.value);
                ***REMOVED***
                if (
                    node.computed &&
                    node.property.type === "TemplateLiteral" &&
                    node.property.expressions.length === 0
                ) ***REMOVED***
                    checkComputedProperty(node, node.property.quasis[0].value.cooked);
                ***REMOVED***
                if (
                    !allowKeywords &&
                    !node.computed &&
                    keywords.indexOf(String(node.property.name)) !== -1
                ) ***REMOVED***
                    context.report(***REMOVED***
                        node: node.property,
                        messageId: "useBrackets",
                        data: ***REMOVED***
                            key: node.property.name
                        ***REMOVED***,
                        fix(fixer) ***REMOVED***
                            const dot = sourceCode.getTokenBefore(node.property);
                            const textAfterDot = sourceCode.text.slice(dot.range[1], node.property.range[0]);

                            if (textAfterDot.trim()) ***REMOVED***

                                // Don't perform any fixes if there are comments between the dot and the property name.
                                return null;
                            ***REMOVED***

                            if (node.object.type === "Identifier" && node.object.name === "let") ***REMOVED***

                                /*
                                 * A statement that starts with `let[` is parsed as a destructuring variable declaration, not
                                 * a MemberExpression.
                                 */
                                return null;
                            ***REMOVED***

                            return fixer.replaceTextRange(
                                [dot.range[0], node.property.range[1]],
                                `[$***REMOVED***textAfterDot***REMOVED***"$***REMOVED***node.property.name***REMOVED***"]`
                            );
                        ***REMOVED***
                    ***REMOVED***);
                ***REMOVED***
            ***REMOVED***
        ***REMOVED***;
    ***REMOVED***
***REMOVED***;
