// eslint/lib/rules/func-call-spacing.js
/**
 * @fileoverview Rule to control spacing within function calls
 * @author Matt DuVall <http://www.mattduvall.com>
 */

"use strict";

//------------------------------------------------------------------------------
// Requirements
//------------------------------------------------------------------------------

const astUtils = require("./utils/ast-utils");

//------------------------------------------------------------------------------
// Rule Definition
//------------------------------------------------------------------------------

module.exports = ***REMOVED***
    meta: ***REMOVED***
        type: "layout",

        docs: ***REMOVED***
            description: "require or disallow spacing between function identifiers and their invocations",
            category: "Stylistic Issues",
            recommended: false,
            url: "https://eslint.org/docs/rules/func-call-spacing"
        ***REMOVED***,

        fixable: "whitespace",

        schema: ***REMOVED***
            anyOf: [
                ***REMOVED***
                    type: "array",
                    items: [
                        ***REMOVED***
                            enum: ["never"]
                        ***REMOVED***
                    ],
                    minItems: 0,
                    maxItems: 1
                ***REMOVED***,
                ***REMOVED***
                    type: "array",
                    items: [
                        ***REMOVED***
                            enum: ["always"]
                        ***REMOVED***,
                        ***REMOVED***
                            type: "object",
                            properties: ***REMOVED***
                                allowNewlines: ***REMOVED***
                                    type: "boolean"
                                ***REMOVED***
                            ***REMOVED***,
                            additionalProperties: false
                        ***REMOVED***
                    ],
                    minItems: 0,
                    maxItems: 2
                ***REMOVED***
            ]
        ***REMOVED***,

        messages: ***REMOVED***
            unexpected: "Unexpected newline between function name and paren.",
            missing: "Missing space between function name and paren."
        ***REMOVED***
    ***REMOVED***,

    create(context) ***REMOVED***

        const never = context.options[0] !== "always";
        const allowNewlines = !never && context.options[1] && context.options[1].allowNewlines;
        const sourceCode = context.getSourceCode();
        const text = sourceCode.getText();

        /**
         * Check if open space is present in a function name
         * @param ***REMOVED***ASTNode***REMOVED*** node node to evaluate
         * @param ***REMOVED***Token***REMOVED*** leftToken The last token of the callee. This may be the closing parenthesis that encloses the callee.
         * @param ***REMOVED***Token***REMOVED*** rightToken Tha first token of the arguments. this is the opening parenthesis that encloses the arguments.
         * @returns ***REMOVED***void***REMOVED***
         * @private
         */
        function checkSpacing(node, leftToken, rightToken) ***REMOVED***
            const textBetweenTokens = text.slice(leftToken.range[1], rightToken.range[0]).replace(/\/\*.*?\*\//gu, "");
            const hasWhitespace = /\s/u.test(textBetweenTokens);
            const hasNewline = hasWhitespace && astUtils.LINEBREAK_MATCHER.test(textBetweenTokens);

            /*
             * never allowNewlines hasWhitespace hasNewline message
             * F     F             F             F          Missing space between function name and paren.
             * F     F             F             T          (Invalid `!hasWhitespace && hasNewline`)
             * F     F             T             T          Unexpected newline between function name and paren.
             * F     F             T             F          (OK)
             * F     T             T             F          (OK)
             * F     T             T             T          (OK)
             * F     T             F             T          (Invalid `!hasWhitespace && hasNewline`)
             * F     T             F             F          Missing space between function name and paren.
             * T     T             F             F          (Invalid `never && allowNewlines`)
             * T     T             F             T          (Invalid `!hasWhitespace && hasNewline`)
             * T     T             T             T          (Invalid `never && allowNewlines`)
             * T     T             T             F          (Invalid `never && allowNewlines`)
             * T     F             T             F          Unexpected space between function name and paren.
             * T     F             T             T          Unexpected space between function name and paren.
             * T     F             F             T          (Invalid `!hasWhitespace && hasNewline`)
             * T     F             F             F          (OK)
             *
             * T                   T                        Unexpected space between function name and paren.
             * F                   F                        Missing space between function name and paren.
             * F     F                           T          Unexpected newline between function name and paren.
             */

            if (never && hasWhitespace) ***REMOVED***
                context.report(***REMOVED***
                    node,
                    loc: leftToken.loc.start,
                    messageId: "unexpected",
                    fix(fixer) ***REMOVED***

                        /*
                         * Only autofix if there is no newline
                         * https://github.com/eslint/eslint/issues/7787
                         */
                        if (!hasNewline) ***REMOVED***
                            return fixer.removeRange([leftToken.range[1], rightToken.range[0]]);
                        ***REMOVED***

                        return null;
                    ***REMOVED***
                ***REMOVED***);
            ***REMOVED*** else if (!never && !hasWhitespace) ***REMOVED***
                context.report(***REMOVED***
                    node,
                    loc: leftToken.loc.start,
                    messageId: "missing",
                    fix(fixer) ***REMOVED***
                        return fixer.insertTextBefore(rightToken, " ");
                    ***REMOVED***
                ***REMOVED***);
            ***REMOVED*** else if (!never && !allowNewlines && hasNewline) ***REMOVED***
                context.report(***REMOVED***
                    node,
                    loc: leftToken.loc.start,
                    messageId: "unexpected",
                    fix(fixer) ***REMOVED***
                        return fixer.replaceTextRange([leftToken.range[1], rightToken.range[0]], " ");
                    ***REMOVED***
                ***REMOVED***);
            ***REMOVED***
        ***REMOVED***

        return ***REMOVED***
            "CallExpression, NewExpression"(node) ***REMOVED***
                const lastToken = sourceCode.getLastToken(node);
                const lastCalleeToken = sourceCode.getLastToken(node.callee);
                const parenToken = sourceCode.getFirstTokenBetween(lastCalleeToken, lastToken, astUtils.isOpeningParenToken);
                const prevToken = parenToken && sourceCode.getTokenBefore(parenToken);

                // Parens in NewExpression are optional
                if (!(parenToken && parenToken.range[1] < node.range[1])) ***REMOVED***
                    return;
                ***REMOVED***

                checkSpacing(node, prevToken, parenToken);
            ***REMOVED***,

            ImportExpression(node) ***REMOVED***
                const leftToken = sourceCode.getFirstToken(node);
                const rightToken = sourceCode.getTokenAfter(leftToken);

                checkSpacing(node, leftToken, rightToken);
            ***REMOVED***
        ***REMOVED***;

    ***REMOVED***
***REMOVED***;
