// eslint/lib/rules/func-style.js
/**
 * @fileoverview Rule to enforce a particular function style
 * @author Nicholas C. Zakas
 */
"use strict";

//------------------------------------------------------------------------------
// Rule Definition
//------------------------------------------------------------------------------

module.exports = ***REMOVED***
    meta: ***REMOVED***
        type: "suggestion",

        docs: ***REMOVED***
            description: "enforce the consistent use of either `function` declarations or expressions",
            category: "Stylistic Issues",
            recommended: false,
            url: "https://eslint.org/docs/rules/func-style"
        ***REMOVED***,

        schema: [
            ***REMOVED***
                enum: ["declaration", "expression"]
            ***REMOVED***,
            ***REMOVED***
                type: "object",
                properties: ***REMOVED***
                    allowArrowFunctions: ***REMOVED***
                        type: "boolean",
                        default: false
                    ***REMOVED***
                ***REMOVED***,
                additionalProperties: false
            ***REMOVED***
        ],

        messages: ***REMOVED***
            expression: "Expected a function expression.",
            declaration: "Expected a function declaration."
        ***REMOVED***
    ***REMOVED***,

    create(context) ***REMOVED***

        const style = context.options[0],
            allowArrowFunctions = context.options[1] && context.options[1].allowArrowFunctions,
            enforceDeclarations = (style === "declaration"),
            stack = [];

        const nodesToCheck = ***REMOVED***
            FunctionDeclaration(node) ***REMOVED***
                stack.push(false);

                if (!enforceDeclarations && node.parent.type !== "ExportDefaultDeclaration") ***REMOVED***
                    context.report(***REMOVED*** node, messageId: "expression" ***REMOVED***);
                ***REMOVED***
            ***REMOVED***,
            "FunctionDeclaration:exit"() ***REMOVED***
                stack.pop();
            ***REMOVED***,

            FunctionExpression(node) ***REMOVED***
                stack.push(false);

                if (enforceDeclarations && node.parent.type === "VariableDeclarator") ***REMOVED***
                    context.report(***REMOVED*** node: node.parent, messageId: "declaration" ***REMOVED***);
                ***REMOVED***
            ***REMOVED***,
            "FunctionExpression:exit"() ***REMOVED***
                stack.pop();
            ***REMOVED***,

            ThisExpression() ***REMOVED***
                if (stack.length > 0) ***REMOVED***
                    stack[stack.length - 1] = true;
                ***REMOVED***
            ***REMOVED***
        ***REMOVED***;

        if (!allowArrowFunctions) ***REMOVED***
            nodesToCheck.ArrowFunctionExpression = function() ***REMOVED***
                stack.push(false);
            ***REMOVED***;

            nodesToCheck["ArrowFunctionExpression:exit"] = function(node) ***REMOVED***
                const hasThisExpr = stack.pop();

                if (enforceDeclarations && !hasThisExpr && node.parent.type === "VariableDeclarator") ***REMOVED***
                    context.report(***REMOVED*** node: node.parent, messageId: "declaration" ***REMOVED***);
                ***REMOVED***
            ***REMOVED***;
        ***REMOVED***

        return nodesToCheck;

    ***REMOVED***
***REMOVED***;
