// eslint/lib/rules/handle-callback-err.js
/**
 * @fileoverview Ensure handling of errors when we know they exist.
 * @author Jamund Ferguson
 */

"use strict";

//------------------------------------------------------------------------------
// Rule Definition
//------------------------------------------------------------------------------

module.exports = ***REMOVED***
    meta: ***REMOVED***
        type: "suggestion",

        docs: ***REMOVED***
            description: "require error handling in callbacks",
            category: "Node.js and CommonJS",
            recommended: false,
            url: "https://eslint.org/docs/rules/handle-callback-err"
        ***REMOVED***,

        schema: [
            ***REMOVED***
                type: "string"
            ***REMOVED***
        ],
        messages: ***REMOVED***
            expected: "Expected error to be handled."
        ***REMOVED***
    ***REMOVED***,

    create(context) ***REMOVED***

        const errorArgument = context.options[0] || "err";

        /**
         * Checks if the given argument should be interpreted as a regexp pattern.
         * @param ***REMOVED***string***REMOVED*** stringToCheck The string which should be checked.
         * @returns ***REMOVED***boolean***REMOVED*** Whether or not the string should be interpreted as a pattern.
         */
        function isPattern(stringToCheck) ***REMOVED***
            const firstChar = stringToCheck[0];

            return firstChar === "^";
        ***REMOVED***

        /**
         * Checks if the given name matches the configured error argument.
         * @param ***REMOVED***string***REMOVED*** name The name which should be compared.
         * @returns ***REMOVED***boolean***REMOVED*** Whether or not the given name matches the configured error variable name.
         */
        function matchesConfiguredErrorName(name) ***REMOVED***
            if (isPattern(errorArgument)) ***REMOVED***
                const regexp = new RegExp(errorArgument, "u");

                return regexp.test(name);
            ***REMOVED***
            return name === errorArgument;
        ***REMOVED***

        /**
         * Get the parameters of a given function scope.
         * @param ***REMOVED***Object***REMOVED*** scope The function scope.
         * @returns ***REMOVED***Array***REMOVED*** All parameters of the given scope.
         */
        function getParameters(scope) ***REMOVED***
            return scope.variables.filter(variable => variable.defs[0] && variable.defs[0].type === "Parameter");
        ***REMOVED***

        /**
         * Check to see if we're handling the error object properly.
         * @param ***REMOVED***ASTNode***REMOVED*** node The AST node to check.
         * @returns ***REMOVED***void***REMOVED***
         */
        function checkForError(node) ***REMOVED***
            const scope = context.getScope(),
                parameters = getParameters(scope),
                firstParameter = parameters[0];

            if (firstParameter && matchesConfiguredErrorName(firstParameter.name)) ***REMOVED***
                if (firstParameter.references.length === 0) ***REMOVED***
                    context.report(***REMOVED*** node, messageId: "expected" ***REMOVED***);
                ***REMOVED***
            ***REMOVED***
        ***REMOVED***

        return ***REMOVED***
            FunctionDeclaration: checkForError,
            FunctionExpression: checkForError,
            ArrowFunctionExpression: checkForError
        ***REMOVED***;

    ***REMOVED***
***REMOVED***;
