// eslint/lib/rules/id-length.js
/**
 * @fileoverview Rule that warns when identifier names are shorter or longer
 * than the values provided in configuration.
 * @author Burak Yigit Kaya aka BYK
 */

"use strict";

//------------------------------------------------------------------------------
// Rule Definition
//------------------------------------------------------------------------------

module.exports = ***REMOVED***
    meta: ***REMOVED***
        type: "suggestion",

        docs: ***REMOVED***
            description: "enforce minimum and maximum identifier lengths",
            category: "Stylistic Issues",
            recommended: false,
            url: "https://eslint.org/docs/rules/id-length"
        ***REMOVED***,

        schema: [
            ***REMOVED***
                type: "object",
                properties: ***REMOVED***
                    min: ***REMOVED***
                        type: "integer",
                        default: 2
                    ***REMOVED***,
                    max: ***REMOVED***
                        type: "integer"
                    ***REMOVED***,
                    exceptions: ***REMOVED***
                        type: "array",
                        uniqueItems: true,
                        items: ***REMOVED***
                            type: "string"
                        ***REMOVED***
                    ***REMOVED***,
                    properties: ***REMOVED***
                        enum: ["always", "never"]
                    ***REMOVED***
                ***REMOVED***,
                additionalProperties: false
            ***REMOVED***
        ],
        messages: ***REMOVED***
            tooShort: "Identifier name '***REMOVED******REMOVED***name***REMOVED******REMOVED***' is too short (< ***REMOVED******REMOVED***min***REMOVED******REMOVED***).",
            tooLong: "Identifier name '***REMOVED******REMOVED***name***REMOVED******REMOVED***' is too long (> ***REMOVED******REMOVED***max***REMOVED******REMOVED***)."
        ***REMOVED***
    ***REMOVED***,

    create(context) ***REMOVED***
        const options = context.options[0] || ***REMOVED******REMOVED***;
        const minLength = typeof options.min !== "undefined" ? options.min : 2;
        const maxLength = typeof options.max !== "undefined" ? options.max : Infinity;
        const properties = options.properties !== "never";
        const exceptions = (options.exceptions ? options.exceptions : [])
            .reduce((obj, item) => ***REMOVED***
                obj[item] = true;

                return obj;
            ***REMOVED***, ***REMOVED******REMOVED***);

        const SUPPORTED_EXPRESSIONS = ***REMOVED***
            MemberExpression: properties && function(parent) ***REMOVED***
                return !parent.computed && (

                    // regular property assignment
                    (parent.parent.left === parent && parent.parent.type === "AssignmentExpression" ||

                    // or the last identifier in an ObjectPattern destructuring
                    parent.parent.type === "Property" && parent.parent.value === parent &&
                    parent.parent.parent.type === "ObjectPattern" && parent.parent.parent.parent.left === parent.parent.parent)
                );
            ***REMOVED***,
            AssignmentPattern(parent, node) ***REMOVED***
                return parent.left === node;
            ***REMOVED***,
            VariableDeclarator(parent, node) ***REMOVED***
                return parent.id === node;
            ***REMOVED***,
            Property: properties && function(parent, node) ***REMOVED***
                return parent.key === node;
            ***REMOVED***,
            ImportDefaultSpecifier: true,
            RestElement: true,
            FunctionExpression: true,
            ArrowFunctionExpression: true,
            ClassDeclaration: true,
            FunctionDeclaration: true,
            MethodDefinition: true,
            CatchClause: true
        ***REMOVED***;

        return ***REMOVED***
            Identifier(node) ***REMOVED***
                const name = node.name;
                const parent = node.parent;

                const isShort = name.length < minLength;
                const isLong = name.length > maxLength;

                if (!(isShort || isLong) || exceptions[name]) ***REMOVED***
                    return; // Nothing to report
                ***REMOVED***

                const isValidExpression = SUPPORTED_EXPRESSIONS[parent.type];

                if (isValidExpression && (isValidExpression === true || isValidExpression(parent, node))) ***REMOVED***
                    context.report(***REMOVED***
                        node,
                        messageId: isShort ? "tooShort" : "tooLong",
                        data: ***REMOVED*** name, min: minLength, max: maxLength ***REMOVED***
                    ***REMOVED***);
                ***REMOVED***
            ***REMOVED***
        ***REMOVED***;
    ***REMOVED***
***REMOVED***;
