// eslint/lib/rules/implicit-arrow-linebreak.js
/**
 * @fileoverview enforce the location of arrow function bodies
 * @author Sharmila Jesupaul
 */
"use strict";

const ***REMOVED*** isCommentToken, isNotOpeningParenToken ***REMOVED*** = require("./utils/ast-utils");

//------------------------------------------------------------------------------
// Rule Definition
//------------------------------------------------------------------------------
module.exports = ***REMOVED***
    meta: ***REMOVED***
        type: "layout",

        docs: ***REMOVED***
            description: "enforce the location of arrow function bodies",
            category: "Stylistic Issues",
            recommended: false,
            url: "https://eslint.org/docs/rules/implicit-arrow-linebreak"
        ***REMOVED***,

        fixable: "whitespace",

        schema: [
            ***REMOVED***
                enum: ["beside", "below"]
            ***REMOVED***
        ],
        messages: ***REMOVED***
            expected: "Expected a linebreak before this expression.",
            unexpected: "Expected no linebreak before this expression."
        ***REMOVED***
    ***REMOVED***,

    create(context) ***REMOVED***
        const sourceCode = context.getSourceCode();
        const option = context.options[0] || "beside";

        /**
         * Validates the location of an arrow function body
         * @param ***REMOVED***ASTNode***REMOVED*** node The arrow function body
         * @returns ***REMOVED***void***REMOVED***
         */
        function validateExpression(node) ***REMOVED***
            if (node.body.type === "BlockStatement") ***REMOVED***
                return;
            ***REMOVED***

            const arrowToken = sourceCode.getTokenBefore(node.body, isNotOpeningParenToken);
            const firstTokenOfBody = sourceCode.getTokenAfter(arrowToken);

            if (arrowToken.loc.end.line === firstTokenOfBody.loc.start.line && option === "below") ***REMOVED***
                context.report(***REMOVED***
                    node: firstTokenOfBody,
                    messageId: "expected",
                    fix: fixer => fixer.insertTextBefore(firstTokenOfBody, "\n")
                ***REMOVED***);
            ***REMOVED*** else if (arrowToken.loc.end.line !== firstTokenOfBody.loc.start.line && option === "beside") ***REMOVED***
                context.report(***REMOVED***
                    node: firstTokenOfBody,
                    messageId: "unexpected",
                    fix(fixer) ***REMOVED***
                        if (sourceCode.getFirstTokenBetween(arrowToken, firstTokenOfBody, ***REMOVED*** includeComments: true, filter: isCommentToken ***REMOVED***)) ***REMOVED***
                            return null;
                        ***REMOVED***

                        return fixer.replaceTextRange([arrowToken.range[1], firstTokenOfBody.range[0]], " ");
                    ***REMOVED***
                ***REMOVED***);
            ***REMOVED***
        ***REMOVED***

        //----------------------------------------------------------------------
        // Public
        //----------------------------------------------------------------------
        return ***REMOVED***
            ArrowFunctionExpression: node => validateExpression(node)
        ***REMOVED***;
    ***REMOVED***
***REMOVED***;
