// eslint/lib/rules/init-declarations.js
/**
 * @fileoverview A rule to control the style of variable initializations.
 * @author Colin Ihrig
 */

"use strict";

//------------------------------------------------------------------------------
// Helpers
//------------------------------------------------------------------------------

/**
 * Checks whether or not a given node is a for loop.
 * @param ***REMOVED***ASTNode***REMOVED*** block - A node to check.
 * @returns ***REMOVED***boolean***REMOVED*** `true` when the node is a for loop.
 */
function isForLoop(block) ***REMOVED***
    return block.type === "ForInStatement" ||
    block.type === "ForOfStatement" ||
    block.type === "ForStatement";
***REMOVED***

/**
 * Checks whether or not a given declarator node has its initializer.
 * @param ***REMOVED***ASTNode***REMOVED*** node - A declarator node to check.
 * @returns ***REMOVED***boolean***REMOVED*** `true` when the node has its initializer.
 */
function isInitialized(node) ***REMOVED***
    const declaration = node.parent;
    const block = declaration.parent;

    if (isForLoop(block)) ***REMOVED***
        if (block.type === "ForStatement") ***REMOVED***
            return block.init === declaration;
        ***REMOVED***
        return block.left === declaration;
    ***REMOVED***
    return Boolean(node.init);
***REMOVED***

//------------------------------------------------------------------------------
// Rule Definition
//------------------------------------------------------------------------------

module.exports = ***REMOVED***
    meta: ***REMOVED***
        type: "suggestion",

        docs: ***REMOVED***
            description: "require or disallow initialization in variable declarations",
            category: "Variables",
            recommended: false,
            url: "https://eslint.org/docs/rules/init-declarations"
        ***REMOVED***,

        schema: ***REMOVED***
            anyOf: [
                ***REMOVED***
                    type: "array",
                    items: [
                        ***REMOVED***
                            enum: ["always"]
                        ***REMOVED***
                    ],
                    minItems: 0,
                    maxItems: 1
                ***REMOVED***,
                ***REMOVED***
                    type: "array",
                    items: [
                        ***REMOVED***
                            enum: ["never"]
                        ***REMOVED***,
                        ***REMOVED***
                            type: "object",
                            properties: ***REMOVED***
                                ignoreForLoopInit: ***REMOVED***
                                    type: "boolean"
                                ***REMOVED***
                            ***REMOVED***,
                            additionalProperties: false
                        ***REMOVED***
                    ],
                    minItems: 0,
                    maxItems: 2
                ***REMOVED***
            ]
        ***REMOVED***,
        messages: ***REMOVED***
            initialized: "Variable '***REMOVED******REMOVED***idName***REMOVED******REMOVED***' should be initialized on declaration.",
            notInitialized: "Variable '***REMOVED******REMOVED***idName***REMOVED******REMOVED***' should not be initialized on declaration."
        ***REMOVED***
    ***REMOVED***,

    create(context) ***REMOVED***

        const MODE_ALWAYS = "always",
            MODE_NEVER = "never";

        const mode = context.options[0] || MODE_ALWAYS;
        const params = context.options[1] || ***REMOVED******REMOVED***;

        //--------------------------------------------------------------------------
        // Public API
        //--------------------------------------------------------------------------

        return ***REMOVED***
            "VariableDeclaration:exit"(node) ***REMOVED***

                const kind = node.kind,
                    declarations = node.declarations;

                for (let i = 0; i < declarations.length; ++i) ***REMOVED***
                    const declaration = declarations[i],
                        id = declaration.id,
                        initialized = isInitialized(declaration),
                        isIgnoredForLoop = params.ignoreForLoopInit && isForLoop(node.parent);
                    let messageId = "";

                    if (mode === MODE_ALWAYS && !initialized) ***REMOVED***
                        messageId = "initialized";
                    ***REMOVED*** else if (mode === MODE_NEVER && kind !== "const" && initialized && !isIgnoredForLoop) ***REMOVED***
                        messageId = "notInitialized";
                    ***REMOVED***

                    if (id.type === "Identifier" && messageId) ***REMOVED***
                        context.report(***REMOVED***
                            node: declaration,
                            messageId,
                            data: ***REMOVED***
                                idName: id.name
                            ***REMOVED***
                        ***REMOVED***);
                    ***REMOVED***
                ***REMOVED***
            ***REMOVED***
        ***REMOVED***;
    ***REMOVED***
***REMOVED***;
