// eslint/lib/rules/keyword-spacing.js
/**
 * @fileoverview Rule to enforce spacing before and after keywords.
 * @author Toru Nagashima
 */

"use strict";

//------------------------------------------------------------------------------
// Requirements
//------------------------------------------------------------------------------

const astUtils = require("./utils/ast-utils"),
    keywords = require("./utils/keywords");

//------------------------------------------------------------------------------
// Constants
//------------------------------------------------------------------------------

const PREV_TOKEN = /^[)\]***REMOVED***>]$/u;
const NEXT_TOKEN = /^(?:[([***REMOVED***<~!]|\+\+?|--?)$/u;
const PREV_TOKEN_M = /^[)\]***REMOVED***>*]$/u;
const NEXT_TOKEN_M = /^[***REMOVED****]$/u;
const TEMPLATE_OPEN_PAREN = /\$\***REMOVED***$/u;
const TEMPLATE_CLOSE_PAREN = /^\***REMOVED***/u;
const CHECK_TYPE = /^(?:JSXElement|RegularExpression|String|Template)$/u;
const KEYS = keywords.concat(["as", "async", "await", "from", "get", "let", "of", "set", "yield"]);

// check duplications.
(function() ***REMOVED***
    KEYS.sort();
    for (let i = 1; i < KEYS.length; ++i) ***REMOVED***
        if (KEYS[i] === KEYS[i - 1]) ***REMOVED***
            throw new Error(`Duplication was found in the keyword list: $***REMOVED***KEYS[i]***REMOVED***`);
        ***REMOVED***
    ***REMOVED***
***REMOVED***());

//------------------------------------------------------------------------------
// Helpers
//------------------------------------------------------------------------------

/**
 * Checks whether or not a given token is a "Template" token ends with "$***REMOVED***".
 *
 * @param ***REMOVED***Token***REMOVED*** token - A token to check.
 * @returns ***REMOVED***boolean***REMOVED*** `true` if the token is a "Template" token ends with "$***REMOVED***".
 */
function isOpenParenOfTemplate(token) ***REMOVED***
    return token.type === "Template" && TEMPLATE_OPEN_PAREN.test(token.value);
***REMOVED***

/**
 * Checks whether or not a given token is a "Template" token starts with "***REMOVED***".
 *
 * @param ***REMOVED***Token***REMOVED*** token - A token to check.
 * @returns ***REMOVED***boolean***REMOVED*** `true` if the token is a "Template" token starts with "***REMOVED***".
 */
function isCloseParenOfTemplate(token) ***REMOVED***
    return token.type === "Template" && TEMPLATE_CLOSE_PAREN.test(token.value);
***REMOVED***

//------------------------------------------------------------------------------
// Rule Definition
//------------------------------------------------------------------------------

module.exports = ***REMOVED***
    meta: ***REMOVED***
        type: "layout",

        docs: ***REMOVED***
            description: "enforce consistent spacing before and after keywords",
            category: "Stylistic Issues",
            recommended: false,
            url: "https://eslint.org/docs/rules/keyword-spacing"
        ***REMOVED***,

        fixable: "whitespace",

        schema: [
            ***REMOVED***
                type: "object",
                properties: ***REMOVED***
                    before: ***REMOVED*** type: "boolean", default: true ***REMOVED***,
                    after: ***REMOVED*** type: "boolean", default: true ***REMOVED***,
                    overrides: ***REMOVED***
                        type: "object",
                        properties: KEYS.reduce((retv, key) => ***REMOVED***
                            retv[key] = ***REMOVED***
                                type: "object",
                                properties: ***REMOVED***
                                    before: ***REMOVED*** type: "boolean", default: true ***REMOVED***,
                                    after: ***REMOVED*** type: "boolean", default: true ***REMOVED***
                                ***REMOVED***,
                                additionalProperties: false
                            ***REMOVED***;
                            return retv;
                        ***REMOVED***, ***REMOVED******REMOVED***),
                        additionalProperties: false
                    ***REMOVED***
                ***REMOVED***,
                additionalProperties: false
            ***REMOVED***
        ],
        messages: ***REMOVED***
            expectedBefore: "Expected space(s) before \"***REMOVED******REMOVED***value***REMOVED******REMOVED***\".",
            expectedAfter: "Expected space(s) after \"***REMOVED******REMOVED***value***REMOVED******REMOVED***\".",
            unexpectedBefore: "Unexpected space(s) before \"***REMOVED******REMOVED***value***REMOVED******REMOVED***\".",
            unexpectedAfter: "Unexpected space(s) after \"***REMOVED******REMOVED***value***REMOVED******REMOVED***\"."
        ***REMOVED***
    ***REMOVED***,

    create(context) ***REMOVED***
        const sourceCode = context.getSourceCode();

        /**
         * Reports a given token if there are not space(s) before the token.
         *
         * @param ***REMOVED***Token***REMOVED*** token - A token to report.
         * @param ***REMOVED***RegExp***REMOVED*** pattern - A pattern of the previous token to check.
         * @returns ***REMOVED***void***REMOVED***
         */
        function expectSpaceBefore(token, pattern) ***REMOVED***
            const prevToken = sourceCode.getTokenBefore(token);

            if (prevToken &&
                (CHECK_TYPE.test(prevToken.type) || pattern.test(prevToken.value)) &&
                !isOpenParenOfTemplate(prevToken) &&
                astUtils.isTokenOnSameLine(prevToken, token) &&
                !sourceCode.isSpaceBetweenTokens(prevToken, token)
            ) ***REMOVED***
                context.report(***REMOVED***
                    loc: token.loc.start,
                    messageId: "expectedBefore",
                    data: token,
                    fix(fixer) ***REMOVED***
                        return fixer.insertTextBefore(token, " ");
                    ***REMOVED***
                ***REMOVED***);
            ***REMOVED***
        ***REMOVED***

        /**
         * Reports a given token if there are space(s) before the token.
         *
         * @param ***REMOVED***Token***REMOVED*** token - A token to report.
         * @param ***REMOVED***RegExp***REMOVED*** pattern - A pattern of the previous token to check.
         * @returns ***REMOVED***void***REMOVED***
         */
        function unexpectSpaceBefore(token, pattern) ***REMOVED***
            const prevToken = sourceCode.getTokenBefore(token);

            if (prevToken &&
                (CHECK_TYPE.test(prevToken.type) || pattern.test(prevToken.value)) &&
                !isOpenParenOfTemplate(prevToken) &&
                astUtils.isTokenOnSameLine(prevToken, token) &&
                sourceCode.isSpaceBetweenTokens(prevToken, token)
            ) ***REMOVED***
                context.report(***REMOVED***
                    loc: token.loc.start,
                    messageId: "unexpectedBefore",
                    data: token,
                    fix(fixer) ***REMOVED***
                        return fixer.removeRange([prevToken.range[1], token.range[0]]);
                    ***REMOVED***
                ***REMOVED***);
            ***REMOVED***
        ***REMOVED***

        /**
         * Reports a given token if there are not space(s) after the token.
         *
         * @param ***REMOVED***Token***REMOVED*** token - A token to report.
         * @param ***REMOVED***RegExp***REMOVED*** pattern - A pattern of the next token to check.
         * @returns ***REMOVED***void***REMOVED***
         */
        function expectSpaceAfter(token, pattern) ***REMOVED***
            const nextToken = sourceCode.getTokenAfter(token);

            if (nextToken &&
                (CHECK_TYPE.test(nextToken.type) || pattern.test(nextToken.value)) &&
                !isCloseParenOfTemplate(nextToken) &&
                astUtils.isTokenOnSameLine(token, nextToken) &&
                !sourceCode.isSpaceBetweenTokens(token, nextToken)
            ) ***REMOVED***
                context.report(***REMOVED***
                    loc: token.loc.start,
                    messageId: "expectedAfter",
                    data: token,
                    fix(fixer) ***REMOVED***
                        return fixer.insertTextAfter(token, " ");
                    ***REMOVED***
                ***REMOVED***);
            ***REMOVED***
        ***REMOVED***

        /**
         * Reports a given token if there are space(s) after the token.
         *
         * @param ***REMOVED***Token***REMOVED*** token - A token to report.
         * @param ***REMOVED***RegExp***REMOVED*** pattern - A pattern of the next token to check.
         * @returns ***REMOVED***void***REMOVED***
         */
        function unexpectSpaceAfter(token, pattern) ***REMOVED***
            const nextToken = sourceCode.getTokenAfter(token);

            if (nextToken &&
                (CHECK_TYPE.test(nextToken.type) || pattern.test(nextToken.value)) &&
                !isCloseParenOfTemplate(nextToken) &&
                astUtils.isTokenOnSameLine(token, nextToken) &&
                sourceCode.isSpaceBetweenTokens(token, nextToken)
            ) ***REMOVED***
                context.report(***REMOVED***
                    loc: token.loc.start,
                    messageId: "unexpectedAfter",
                    data: token,
                    fix(fixer) ***REMOVED***
                        return fixer.removeRange([token.range[1], nextToken.range[0]]);
                    ***REMOVED***
                ***REMOVED***);
            ***REMOVED***
        ***REMOVED***

        /**
         * Parses the option object and determines check methods for each keyword.
         *
         * @param ***REMOVED***Object|undefined***REMOVED*** options - The option object to parse.
         * @returns ***REMOVED***Object***REMOVED*** - Normalized option object.
         *      Keys are keywords (there are for every keyword).
         *      Values are instances of `***REMOVED***"before": function, "after": function***REMOVED***`.
         */
        function parseOptions(options = ***REMOVED******REMOVED***) ***REMOVED***
            const before = options.before !== false;
            const after = options.after !== false;
            const defaultValue = ***REMOVED***
                before: before ? expectSpaceBefore : unexpectSpaceBefore,
                after: after ? expectSpaceAfter : unexpectSpaceAfter
            ***REMOVED***;
            const overrides = (options && options.overrides) || ***REMOVED******REMOVED***;
            const retv = Object.create(null);

            for (let i = 0; i < KEYS.length; ++i) ***REMOVED***
                const key = KEYS[i];
                const override = overrides[key];

                if (override) ***REMOVED***
                    const thisBefore = ("before" in override) ? override.before : before;
                    const thisAfter = ("after" in override) ? override.after : after;

                    retv[key] = ***REMOVED***
                        before: thisBefore ? expectSpaceBefore : unexpectSpaceBefore,
                        after: thisAfter ? expectSpaceAfter : unexpectSpaceAfter
                    ***REMOVED***;
                ***REMOVED*** else ***REMOVED***
                    retv[key] = defaultValue;
                ***REMOVED***
            ***REMOVED***

            return retv;
        ***REMOVED***

        const checkMethodMap = parseOptions(context.options[0]);

        /**
         * Reports a given token if usage of spacing followed by the token is
         * invalid.
         *
         * @param ***REMOVED***Token***REMOVED*** token - A token to report.
         * @param ***REMOVED***RegExp|undefined***REMOVED*** pattern - Optional. A pattern of the previous
         *      token to check.
         * @returns ***REMOVED***void***REMOVED***
         */
        function checkSpacingBefore(token, pattern) ***REMOVED***
            checkMethodMap[token.value].before(token, pattern || PREV_TOKEN);
        ***REMOVED***

        /**
         * Reports a given token if usage of spacing preceded by the token is
         * invalid.
         *
         * @param ***REMOVED***Token***REMOVED*** token - A token to report.
         * @param ***REMOVED***RegExp|undefined***REMOVED*** pattern - Optional. A pattern of the next
         *      token to check.
         * @returns ***REMOVED***void***REMOVED***
         */
        function checkSpacingAfter(token, pattern) ***REMOVED***
            checkMethodMap[token.value].after(token, pattern || NEXT_TOKEN);
        ***REMOVED***

        /**
         * Reports a given token if usage of spacing around the token is invalid.
         *
         * @param ***REMOVED***Token***REMOVED*** token - A token to report.
         * @returns ***REMOVED***void***REMOVED***
         */
        function checkSpacingAround(token) ***REMOVED***
            checkSpacingBefore(token);
            checkSpacingAfter(token);
        ***REMOVED***

        /**
         * Reports the first token of a given node if the first token is a keyword
         * and usage of spacing around the token is invalid.
         *
         * @param ***REMOVED***ASTNode|null***REMOVED*** node - A node to report.
         * @returns ***REMOVED***void***REMOVED***
         */
        function checkSpacingAroundFirstToken(node) ***REMOVED***
            const firstToken = node && sourceCode.getFirstToken(node);

            if (firstToken && firstToken.type === "Keyword") ***REMOVED***
                checkSpacingAround(firstToken);
            ***REMOVED***
        ***REMOVED***

        /**
         * Reports the first token of a given node if the first token is a keyword
         * and usage of spacing followed by the token is invalid.
         *
         * This is used for unary operators (e.g. `typeof`), `function`, and `super`.
         * Other rules are handling usage of spacing preceded by those keywords.
         *
         * @param ***REMOVED***ASTNode|null***REMOVED*** node - A node to report.
         * @returns ***REMOVED***void***REMOVED***
         */
        function checkSpacingBeforeFirstToken(node) ***REMOVED***
            const firstToken = node && sourceCode.getFirstToken(node);

            if (firstToken && firstToken.type === "Keyword") ***REMOVED***
                checkSpacingBefore(firstToken);
            ***REMOVED***
        ***REMOVED***

        /**
         * Reports the previous token of a given node if the token is a keyword and
         * usage of spacing around the token is invalid.
         *
         * @param ***REMOVED***ASTNode|null***REMOVED*** node - A node to report.
         * @returns ***REMOVED***void***REMOVED***
         */
        function checkSpacingAroundTokenBefore(node) ***REMOVED***
            if (node) ***REMOVED***
                const token = sourceCode.getTokenBefore(node, astUtils.isKeywordToken);

                checkSpacingAround(token);
            ***REMOVED***
        ***REMOVED***

        /**
         * Reports `async` or `function` keywords of a given node if usage of
         * spacing around those keywords is invalid.
         *
         * @param ***REMOVED***ASTNode***REMOVED*** node - A node to report.
         * @returns ***REMOVED***void***REMOVED***
         */
        function checkSpacingForFunction(node) ***REMOVED***
            const firstToken = node && sourceCode.getFirstToken(node);

            if (firstToken &&
                ((firstToken.type === "Keyword" && firstToken.value === "function") ||
                firstToken.value === "async")
            ) ***REMOVED***
                checkSpacingBefore(firstToken);
            ***REMOVED***
        ***REMOVED***

        /**
         * Reports `class` and `extends` keywords of a given node if usage of
         * spacing around those keywords is invalid.
         *
         * @param ***REMOVED***ASTNode***REMOVED*** node - A node to report.
         * @returns ***REMOVED***void***REMOVED***
         */
        function checkSpacingForClass(node) ***REMOVED***
            checkSpacingAroundFirstToken(node);
            checkSpacingAroundTokenBefore(node.superClass);
        ***REMOVED***

        /**
         * Reports `if` and `else` keywords of a given node if usage of spacing
         * around those keywords is invalid.
         *
         * @param ***REMOVED***ASTNode***REMOVED*** node - A node to report.
         * @returns ***REMOVED***void***REMOVED***
         */
        function checkSpacingForIfStatement(node) ***REMOVED***
            checkSpacingAroundFirstToken(node);
            checkSpacingAroundTokenBefore(node.alternate);
        ***REMOVED***

        /**
         * Reports `try`, `catch`, and `finally` keywords of a given node if usage
         * of spacing around those keywords is invalid.
         *
         * @param ***REMOVED***ASTNode***REMOVED*** node - A node to report.
         * @returns ***REMOVED***void***REMOVED***
         */
        function checkSpacingForTryStatement(node) ***REMOVED***
            checkSpacingAroundFirstToken(node);
            checkSpacingAroundFirstToken(node.handler);
            checkSpacingAroundTokenBefore(node.finalizer);
        ***REMOVED***

        /**
         * Reports `do` and `while` keywords of a given node if usage of spacing
         * around those keywords is invalid.
         *
         * @param ***REMOVED***ASTNode***REMOVED*** node - A node to report.
         * @returns ***REMOVED***void***REMOVED***
         */
        function checkSpacingForDoWhileStatement(node) ***REMOVED***
            checkSpacingAroundFirstToken(node);
            checkSpacingAroundTokenBefore(node.test);
        ***REMOVED***

        /**
         * Reports `for` and `in` keywords of a given node if usage of spacing
         * around those keywords is invalid.
         *
         * @param ***REMOVED***ASTNode***REMOVED*** node - A node to report.
         * @returns ***REMOVED***void***REMOVED***
         */
        function checkSpacingForForInStatement(node) ***REMOVED***
            checkSpacingAroundFirstToken(node);
            checkSpacingAroundTokenBefore(node.right);
        ***REMOVED***

        /**
         * Reports `for` and `of` keywords of a given node if usage of spacing
         * around those keywords is invalid.
         *
         * @param ***REMOVED***ASTNode***REMOVED*** node - A node to report.
         * @returns ***REMOVED***void***REMOVED***
         */
        function checkSpacingForForOfStatement(node) ***REMOVED***
            if (node.await) ***REMOVED***
                checkSpacingBefore(sourceCode.getFirstToken(node, 0));
                checkSpacingAfter(sourceCode.getFirstToken(node, 1));
            ***REMOVED*** else ***REMOVED***
                checkSpacingAroundFirstToken(node);
            ***REMOVED***
            checkSpacingAround(sourceCode.getTokenBefore(node.right, astUtils.isNotOpeningParenToken));
        ***REMOVED***

        /**
         * Reports `import`, `export`, `as`, and `from` keywords of a given node if
         * usage of spacing around those keywords is invalid.
         *
         * This rule handles the `*` token in module declarations.
         *
         *     import*as A from "./a"; /*error Expected space(s) after "import".
         *                               error Expected space(s) before "as".
         *
         * @param ***REMOVED***ASTNode***REMOVED*** node - A node to report.
         * @returns ***REMOVED***void***REMOVED***
         */
        function checkSpacingForModuleDeclaration(node) ***REMOVED***
            const firstToken = sourceCode.getFirstToken(node);

            checkSpacingBefore(firstToken, PREV_TOKEN_M);
            checkSpacingAfter(firstToken, NEXT_TOKEN_M);

            if (node.type === "ExportDefaultDeclaration") ***REMOVED***
                checkSpacingAround(sourceCode.getTokenAfter(firstToken));
            ***REMOVED***

            if (node.source) ***REMOVED***
                const fromToken = sourceCode.getTokenBefore(node.source);

                checkSpacingBefore(fromToken, PREV_TOKEN_M);
                checkSpacingAfter(fromToken, NEXT_TOKEN_M);
            ***REMOVED***
        ***REMOVED***

        /**
         * Reports `as` keyword of a given node if usage of spacing around this
         * keyword is invalid.
         *
         * @param ***REMOVED***ASTNode***REMOVED*** node - A node to report.
         * @returns ***REMOVED***void***REMOVED***
         */
        function checkSpacingForImportNamespaceSpecifier(node) ***REMOVED***
            const asToken = sourceCode.getFirstToken(node, 1);

            checkSpacingBefore(asToken, PREV_TOKEN_M);
        ***REMOVED***

        /**
         * Reports `static`, `get`, and `set` keywords of a given node if usage of
         * spacing around those keywords is invalid.
         *
         * @param ***REMOVED***ASTNode***REMOVED*** node - A node to report.
         * @returns ***REMOVED***void***REMOVED***
         */
        function checkSpacingForProperty(node) ***REMOVED***
            if (node.static) ***REMOVED***
                checkSpacingAroundFirstToken(node);
            ***REMOVED***
            if (node.kind === "get" ||
                node.kind === "set" ||
                (
                    (node.method || node.type === "MethodDefinition") &&
                    node.value.async
                )
            ) ***REMOVED***
                const token = sourceCode.getTokenBefore(
                    node.key,
                    tok => ***REMOVED***
                        switch (tok.value) ***REMOVED***
                            case "get":
                            case "set":
                            case "async":
                                return true;
                            default:
                                return false;
                        ***REMOVED***
                    ***REMOVED***
                );

                if (!token) ***REMOVED***
                    throw new Error("Failed to find token get, set, or async beside method name");
                ***REMOVED***


                checkSpacingAround(token);
            ***REMOVED***
        ***REMOVED***

        /**
         * Reports `await` keyword of a given node if usage of spacing before
         * this keyword is invalid.
         *
         * @param ***REMOVED***ASTNode***REMOVED*** node - A node to report.
         * @returns ***REMOVED***void***REMOVED***
         */
        function checkSpacingForAwaitExpression(node) ***REMOVED***
            checkSpacingBefore(sourceCode.getFirstToken(node));
        ***REMOVED***

        return ***REMOVED***

            // Statements
            DebuggerStatement: checkSpacingAroundFirstToken,
            WithStatement: checkSpacingAroundFirstToken,

            // Statements - Control flow
            BreakStatement: checkSpacingAroundFirstToken,
            ContinueStatement: checkSpacingAroundFirstToken,
            ReturnStatement: checkSpacingAroundFirstToken,
            ThrowStatement: checkSpacingAroundFirstToken,
            TryStatement: checkSpacingForTryStatement,

            // Statements - Choice
            IfStatement: checkSpacingForIfStatement,
            SwitchStatement: checkSpacingAroundFirstToken,
            SwitchCase: checkSpacingAroundFirstToken,

            // Statements - Loops
            DoWhileStatement: checkSpacingForDoWhileStatement,
            ForInStatement: checkSpacingForForInStatement,
            ForOfStatement: checkSpacingForForOfStatement,
            ForStatement: checkSpacingAroundFirstToken,
            WhileStatement: checkSpacingAroundFirstToken,

            // Statements - Declarations
            ClassDeclaration: checkSpacingForClass,
            ExportNamedDeclaration: checkSpacingForModuleDeclaration,
            ExportDefaultDeclaration: checkSpacingForModuleDeclaration,
            ExportAllDeclaration: checkSpacingForModuleDeclaration,
            FunctionDeclaration: checkSpacingForFunction,
            ImportDeclaration: checkSpacingForModuleDeclaration,
            VariableDeclaration: checkSpacingAroundFirstToken,

            // Expressions
            ArrowFunctionExpression: checkSpacingForFunction,
            AwaitExpression: checkSpacingForAwaitExpression,
            ClassExpression: checkSpacingForClass,
            FunctionExpression: checkSpacingForFunction,
            NewExpression: checkSpacingBeforeFirstToken,
            Super: checkSpacingBeforeFirstToken,
            ThisExpression: checkSpacingBeforeFirstToken,
            UnaryExpression: checkSpacingBeforeFirstToken,
            YieldExpression: checkSpacingBeforeFirstToken,

            // Others
            ImportNamespaceSpecifier: checkSpacingForImportNamespaceSpecifier,
            MethodDefinition: checkSpacingForProperty,
            Property: checkSpacingForProperty
        ***REMOVED***;
    ***REMOVED***
***REMOVED***;
