// eslint/lib/rules/linebreak-style.js
/**
 * @fileoverview Rule to enforce a single linebreak style.
 * @author Erik Mueller
 */

"use strict";

//------------------------------------------------------------------------------
// Requirements
//------------------------------------------------------------------------------

const astUtils = require("./utils/ast-utils");

//------------------------------------------------------------------------------
// Rule Definition
//------------------------------------------------------------------------------

module.exports = ***REMOVED***
    meta: ***REMOVED***
        type: "layout",

        docs: ***REMOVED***
            description: "enforce consistent linebreak style",
            category: "Stylistic Issues",
            recommended: false,
            url: "https://eslint.org/docs/rules/linebreak-style"
        ***REMOVED***,

        fixable: "whitespace",

        schema: [
            ***REMOVED***
                enum: ["unix", "windows"]
            ***REMOVED***
        ],
        messages: ***REMOVED***
            expectedLF: "Expected linebreaks to be 'LF' but found 'CRLF'.",
            expectedCRLF: "Expected linebreaks to be 'CRLF' but found 'LF'."
        ***REMOVED***
    ***REMOVED***,

    create(context) ***REMOVED***
        const sourceCode = context.getSourceCode();

        //--------------------------------------------------------------------------
        // Helpers
        //--------------------------------------------------------------------------

        /**
         * Builds a fix function that replaces text at the specified range in the source text.
         * @param ***REMOVED***int[]***REMOVED*** range The range to replace
         * @param ***REMOVED***string***REMOVED*** text The text to insert.
         * @returns ***REMOVED***Function***REMOVED*** Fixer function
         * @private
         */
        function createFix(range, text) ***REMOVED***
            return function(fixer) ***REMOVED***
                return fixer.replaceTextRange(range, text);
            ***REMOVED***;
        ***REMOVED***

        //--------------------------------------------------------------------------
        // Public
        //--------------------------------------------------------------------------

        return ***REMOVED***
            Program: function checkForlinebreakStyle(node) ***REMOVED***
                const linebreakStyle = context.options[0] || "unix",
                    expectedLF = linebreakStyle === "unix",
                    expectedLFChars = expectedLF ? "\n" : "\r\n",
                    source = sourceCode.getText(),
                    pattern = astUtils.createGlobalLinebreakMatcher();
                let match;

                let i = 0;

                while ((match = pattern.exec(source)) !== null) ***REMOVED***
                    i++;
                    if (match[0] === expectedLFChars) ***REMOVED***
                        continue;
                    ***REMOVED***

                    const index = match.index;
                    const range = [index, index + match[0].length];

                    context.report(***REMOVED***
                        node,
                        loc: ***REMOVED***
                            line: i,
                            column: sourceCode.lines[i - 1].length
                        ***REMOVED***,
                        messageId: expectedLF ? "expectedLF" : "expectedCRLF",
                        fix: createFix(range, expectedLFChars)
                    ***REMOVED***);
                ***REMOVED***
            ***REMOVED***
        ***REMOVED***;
    ***REMOVED***
***REMOVED***;
