// eslint/lib/rules/max-classes-per-file.js
/**
 * @fileoverview Enforce a maximum number of classes per file
 * @author James Garbutt <https://github.com/43081j>
 */
"use strict";

//------------------------------------------------------------------------------
// Requirements
//------------------------------------------------------------------------------


//------------------------------------------------------------------------------
// Rule Definition
//------------------------------------------------------------------------------

module.exports = ***REMOVED***
    meta: ***REMOVED***
        type: "suggestion",

        docs: ***REMOVED***
            description: "enforce a maximum number of classes per file",
            category: "Best Practices",
            recommended: false,
            url: "https://eslint.org/docs/rules/max-classes-per-file"
        ***REMOVED***,

        schema: [
            ***REMOVED***
                type: "integer",
                minimum: 1
            ***REMOVED***
        ],

        messages: ***REMOVED***
            maximumExceeded: "File has too many classes (***REMOVED******REMOVED*** classCount ***REMOVED******REMOVED***). Maximum allowed is ***REMOVED******REMOVED*** max ***REMOVED******REMOVED***."
        ***REMOVED***
    ***REMOVED***,
    create(context) ***REMOVED***

        const maxClasses = context.options[0] || 1;

        let classCount = 0;

        return ***REMOVED***
            Program() ***REMOVED***
                classCount = 0;
            ***REMOVED***,
            "Program:exit"(node) ***REMOVED***
                if (classCount > maxClasses) ***REMOVED***
                    context.report(***REMOVED***
                        node,
                        messageId: "maximumExceeded",
                        data: ***REMOVED***
                            classCount,
                            max: maxClasses
                        ***REMOVED***
                    ***REMOVED***);
                ***REMOVED***
            ***REMOVED***,
            "ClassDeclaration, ClassExpression"() ***REMOVED***
                classCount++;
            ***REMOVED***
        ***REMOVED***;
    ***REMOVED***
***REMOVED***;
