// eslint/lib/rules/new-cap.js
/**
 * @fileoverview Rule to flag use of constructors without capital letters
 * @author Nicholas C. Zakas
 */

"use strict";

//------------------------------------------------------------------------------
// Requirements
//------------------------------------------------------------------------------

//------------------------------------------------------------------------------
// Helpers
//------------------------------------------------------------------------------

const CAPS_ALLOWED = [
    "Array",
    "Boolean",
    "Date",
    "Error",
    "Function",
    "Number",
    "Object",
    "RegExp",
    "String",
    "Symbol",
    "BigInt"
];

/**
 * Ensure that if the key is provided, it must be an array.
 * @param ***REMOVED***Object***REMOVED*** obj Object to check with `key`.
 * @param ***REMOVED***string***REMOVED*** key Object key to check on `obj`.
 * @param ***REMOVED*******REMOVED*** fallback If obj[key] is not present, this will be returned.
 * @returns ***REMOVED***string[]***REMOVED*** Returns obj[key] if it's an Array, otherwise `fallback`
 */
function checkArray(obj, key, fallback) ***REMOVED***

    /* istanbul ignore if */
    if (Object.prototype.hasOwnProperty.call(obj, key) && !Array.isArray(obj[key])) ***REMOVED***
        throw new TypeError(`$***REMOVED***key***REMOVED***, if provided, must be an Array`);
    ***REMOVED***
    return obj[key] || fallback;
***REMOVED***

/**
 * A reducer function to invert an array to an Object mapping the string form of the key, to `true`.
 * @param ***REMOVED***Object***REMOVED*** map Accumulator object for the reduce.
 * @param ***REMOVED***string***REMOVED*** key Object key to set to `true`.
 * @returns ***REMOVED***Object***REMOVED*** Returns the updated Object for further reduction.
 */
function invert(map, key) ***REMOVED***
    map[key] = true;
    return map;
***REMOVED***

/**
 * Creates an object with the cap is new exceptions as its keys and true as their values.
 * @param ***REMOVED***Object***REMOVED*** config Rule configuration
 * @returns ***REMOVED***Object***REMOVED*** Object with cap is new exceptions.
 */
function calculateCapIsNewExceptions(config) ***REMOVED***
    let capIsNewExceptions = checkArray(config, "capIsNewExceptions", CAPS_ALLOWED);

    if (capIsNewExceptions !== CAPS_ALLOWED) ***REMOVED***
        capIsNewExceptions = capIsNewExceptions.concat(CAPS_ALLOWED);
    ***REMOVED***

    return capIsNewExceptions.reduce(invert, ***REMOVED******REMOVED***);
***REMOVED***

//------------------------------------------------------------------------------
// Rule Definition
//------------------------------------------------------------------------------

module.exports = ***REMOVED***
    meta: ***REMOVED***
        type: "suggestion",

        docs: ***REMOVED***
            description: "require constructor names to begin with a capital letter",
            category: "Stylistic Issues",
            recommended: false,
            url: "https://eslint.org/docs/rules/new-cap"
        ***REMOVED***,

        schema: [
            ***REMOVED***
                type: "object",
                properties: ***REMOVED***
                    newIsCap: ***REMOVED***
                        type: "boolean",
                        default: true
                    ***REMOVED***,
                    capIsNew: ***REMOVED***
                        type: "boolean",
                        default: true
                    ***REMOVED***,
                    newIsCapExceptions: ***REMOVED***
                        type: "array",
                        items: ***REMOVED***
                            type: "string"
                        ***REMOVED***
                    ***REMOVED***,
                    newIsCapExceptionPattern: ***REMOVED***
                        type: "string"
                    ***REMOVED***,
                    capIsNewExceptions: ***REMOVED***
                        type: "array",
                        items: ***REMOVED***
                            type: "string"
                        ***REMOVED***
                    ***REMOVED***,
                    capIsNewExceptionPattern: ***REMOVED***
                        type: "string"
                    ***REMOVED***,
                    properties: ***REMOVED***
                        type: "boolean",
                        default: true
                    ***REMOVED***
                ***REMOVED***,
                additionalProperties: false
            ***REMOVED***
        ],
        messages: ***REMOVED***
            upper: "A function with a name starting with an uppercase letter should only be used as a constructor.",
            lower: "A constructor name should not start with a lowercase letter."
        ***REMOVED***
    ***REMOVED***,

    create(context) ***REMOVED***

        const config = Object.assign(***REMOVED******REMOVED***, context.options[0]);

        config.newIsCap = config.newIsCap !== false;
        config.capIsNew = config.capIsNew !== false;
        const skipProperties = config.properties === false;

        const newIsCapExceptions = checkArray(config, "newIsCapExceptions", []).reduce(invert, ***REMOVED******REMOVED***);
        const newIsCapExceptionPattern = config.newIsCapExceptionPattern ? new RegExp(config.newIsCapExceptionPattern, "u") : null;

        const capIsNewExceptions = calculateCapIsNewExceptions(config);
        const capIsNewExceptionPattern = config.capIsNewExceptionPattern ? new RegExp(config.capIsNewExceptionPattern, "u") : null;

        const listeners = ***REMOVED******REMOVED***;

        const sourceCode = context.getSourceCode();

        //--------------------------------------------------------------------------
        // Helpers
        //--------------------------------------------------------------------------

        /**
         * Get exact callee name from expression
         * @param ***REMOVED***ASTNode***REMOVED*** node CallExpression or NewExpression node
         * @returns ***REMOVED***string***REMOVED*** name
         */
        function extractNameFromExpression(node) ***REMOVED***

            let name = "";

            if (node.callee.type === "MemberExpression") ***REMOVED***
                const property = node.callee.property;

                if (property.type === "Literal" && (typeof property.value === "string")) ***REMOVED***
                    name = property.value;
                ***REMOVED*** else if (property.type === "Identifier" && !node.callee.computed) ***REMOVED***
                    name = property.name;
                ***REMOVED***
            ***REMOVED*** else ***REMOVED***
                name = node.callee.name;
            ***REMOVED***
            return name;
        ***REMOVED***

        /**
         * Returns the capitalization state of the string -
         * Whether the first character is uppercase, lowercase, or non-alphabetic
         * @param ***REMOVED***string***REMOVED*** str String
         * @returns ***REMOVED***string***REMOVED*** capitalization state: "non-alpha", "lower", or "upper"
         */
        function getCap(str) ***REMOVED***
            const firstChar = str.charAt(0);

            const firstCharLower = firstChar.toLowerCase();
            const firstCharUpper = firstChar.toUpperCase();

            if (firstCharLower === firstCharUpper) ***REMOVED***

                // char has no uppercase variant, so it's non-alphabetic
                return "non-alpha";
            ***REMOVED***
            if (firstChar === firstCharLower) ***REMOVED***
                return "lower";
            ***REMOVED***
            return "upper";

        ***REMOVED***

        /**
         * Check if capitalization is allowed for a CallExpression
         * @param ***REMOVED***Object***REMOVED*** allowedMap Object mapping calleeName to a Boolean
         * @param ***REMOVED***ASTNode***REMOVED*** node CallExpression node
         * @param ***REMOVED***string***REMOVED*** calleeName Capitalized callee name from a CallExpression
         * @param ***REMOVED***Object***REMOVED*** pattern RegExp object from options pattern
         * @returns ***REMOVED***boolean***REMOVED*** Returns true if the callee may be capitalized
         */
        function isCapAllowed(allowedMap, node, calleeName, pattern) ***REMOVED***
            const sourceText = sourceCode.getText(node.callee);

            if (allowedMap[calleeName] || allowedMap[sourceText]) ***REMOVED***
                return true;
            ***REMOVED***

            if (pattern && pattern.test(sourceText)) ***REMOVED***
                return true;
            ***REMOVED***

            if (calleeName === "UTC" && node.callee.type === "MemberExpression") ***REMOVED***

                // allow if callee is Date.UTC
                return node.callee.object.type === "Identifier" &&
                    node.callee.object.name === "Date";
            ***REMOVED***

            return skipProperties && node.callee.type === "MemberExpression";
        ***REMOVED***

        /**
         * Reports the given messageId for the given node. The location will be the start of the property or the callee.
         * @param ***REMOVED***ASTNode***REMOVED*** node CallExpression or NewExpression node.
         * @param ***REMOVED***string***REMOVED*** messageId The messageId to report.
         * @returns ***REMOVED***void***REMOVED***
         */
        function report(node, messageId) ***REMOVED***
            let callee = node.callee;

            if (callee.type === "MemberExpression") ***REMOVED***
                callee = callee.property;
            ***REMOVED***

            context.report(***REMOVED*** node, loc: callee.loc.start, messageId ***REMOVED***);
        ***REMOVED***

        //--------------------------------------------------------------------------
        // Public
        //--------------------------------------------------------------------------

        if (config.newIsCap) ***REMOVED***
            listeners.NewExpression = function(node) ***REMOVED***

                const constructorName = extractNameFromExpression(node);

                if (constructorName) ***REMOVED***
                    const capitalization = getCap(constructorName);
                    const isAllowed = capitalization !== "lower" || isCapAllowed(newIsCapExceptions, node, constructorName, newIsCapExceptionPattern);

                    if (!isAllowed) ***REMOVED***
                        report(node, "lower");
                    ***REMOVED***
                ***REMOVED***
            ***REMOVED***;
        ***REMOVED***

        if (config.capIsNew) ***REMOVED***
            listeners.CallExpression = function(node) ***REMOVED***

                const calleeName = extractNameFromExpression(node);

                if (calleeName) ***REMOVED***
                    const capitalization = getCap(calleeName);
                    const isAllowed = capitalization !== "upper" || isCapAllowed(capIsNewExceptions, node, calleeName, capIsNewExceptionPattern);

                    if (!isAllowed) ***REMOVED***
                        report(node, "upper");
                    ***REMOVED***
                ***REMOVED***
            ***REMOVED***;
        ***REMOVED***

        return listeners;
    ***REMOVED***
***REMOVED***;
