// eslint/lib/rules/no-array-constructor.js
/**
 * @fileoverview Disallow construction of dense arrays using the Array constructor
 * @author Matt DuVall <http://www.mattduvall.com/>
 */

"use strict";

//------------------------------------------------------------------------------
// Rule Definition
//------------------------------------------------------------------------------

module.exports = ***REMOVED***
    meta: ***REMOVED***
        type: "suggestion",

        docs: ***REMOVED***
            description: "disallow `Array` constructors",
            category: "Stylistic Issues",
            recommended: false,
            url: "https://eslint.org/docs/rules/no-array-constructor"
        ***REMOVED***,

        schema: [],

        messages: ***REMOVED***
            preferLiteral: "The array literal notation [] is preferable."
        ***REMOVED***
    ***REMOVED***,

    create(context) ***REMOVED***

        /**
         * Disallow construction of dense arrays using the Array constructor
         * @param ***REMOVED***ASTNode***REMOVED*** node node to evaluate
         * @returns ***REMOVED***void***REMOVED***
         * @private
         */
        function check(node) ***REMOVED***
            if (
                node.arguments.length !== 1 &&
                node.callee.type === "Identifier" &&
                node.callee.name === "Array"
            ) ***REMOVED***
                context.report(***REMOVED*** node, messageId: "preferLiteral" ***REMOVED***);
            ***REMOVED***
        ***REMOVED***

        return ***REMOVED***
            CallExpression: check,
            NewExpression: check
        ***REMOVED***;

    ***REMOVED***
***REMOVED***;
