// eslint/lib/rules/no-async-promise-executor.js
/**
 * @fileoverview disallow using an async function as a Promise executor
 * @author Teddy Katz
 */
"use strict";

//------------------------------------------------------------------------------
// Rule Definition
//------------------------------------------------------------------------------

module.exports = ***REMOVED***
    meta: ***REMOVED***
        type: "problem",

        docs: ***REMOVED***
            description: "disallow using an async function as a Promise executor",
            category: "Possible Errors",
            recommended: true,
            url: "https://eslint.org/docs/rules/no-async-promise-executor"
        ***REMOVED***,

        fixable: null,
        schema: [],
        messages: ***REMOVED***
            async: "Promise executor functions should not be async."
        ***REMOVED***
    ***REMOVED***,

    create(context) ***REMOVED***
        return ***REMOVED***
            "NewExpression[callee.name='Promise'][arguments.0.async=true]"(node) ***REMOVED***
                context.report(***REMOVED***
                    node: context.getSourceCode().getFirstToken(node.arguments[0], token => token.value === "async"),
                    messageId: "async"
                ***REMOVED***);
            ***REMOVED***
        ***REMOVED***;
    ***REMOVED***
***REMOVED***;
