// eslint/lib/rules/no-alert.js
/**
 * @fileoverview Rule to flag use of alert, confirm, prompt
 * @author Nicholas C. Zakas
 */
"use strict";

//------------------------------------------------------------------------------
// Requirements
//------------------------------------------------------------------------------

const getPropertyName = require("./utils/ast-utils").getStaticPropertyName;

//------------------------------------------------------------------------------
// Helpers
//------------------------------------------------------------------------------

/**
 * Checks if the given name is a prohibited identifier.
 * @param ***REMOVED***string***REMOVED*** name The name to check
 * @returns ***REMOVED***boolean***REMOVED*** Whether or not the name is prohibited.
 */
function isProhibitedIdentifier(name) ***REMOVED***
    return /^(alert|confirm|prompt)$/u.test(name);
***REMOVED***

/**
 * Finds the eslint-scope reference in the given scope.
 * @param ***REMOVED***Object***REMOVED*** scope The scope to search.
 * @param ***REMOVED***ASTNode***REMOVED*** node The identifier node.
 * @returns ***REMOVED***Reference|null***REMOVED*** Returns the found reference or null if none were found.
 */
function findReference(scope, node) ***REMOVED***
    const references = scope.references.filter(reference => reference.identifier.range[0] === node.range[0] &&
            reference.identifier.range[1] === node.range[1]);

    if (references.length === 1) ***REMOVED***
        return references[0];
    ***REMOVED***
    return null;
***REMOVED***

/**
 * Checks if the given identifier node is shadowed in the given scope.
 * @param ***REMOVED***Object***REMOVED*** scope The current scope.
 * @param ***REMOVED***string***REMOVED*** node The identifier node to check
 * @returns ***REMOVED***boolean***REMOVED*** Whether or not the name is shadowed.
 */
function isShadowed(scope, node) ***REMOVED***
    const reference = findReference(scope, node);

    return reference && reference.resolved && reference.resolved.defs.length > 0;
***REMOVED***

/**
 * Checks if the given identifier node is a ThisExpression in the global scope or the global window property.
 * @param ***REMOVED***Object***REMOVED*** scope The current scope.
 * @param ***REMOVED***string***REMOVED*** node The identifier node to check
 * @returns ***REMOVED***boolean***REMOVED*** Whether or not the node is a reference to the global object.
 */
function isGlobalThisReferenceOrGlobalWindow(scope, node) ***REMOVED***
    if (scope.type === "global" && node.type === "ThisExpression") ***REMOVED***
        return true;
    ***REMOVED***
    if (node.name === "window") ***REMOVED***
        return !isShadowed(scope, node);
    ***REMOVED***

    return false;
***REMOVED***

//------------------------------------------------------------------------------
// Rule Definition
//------------------------------------------------------------------------------

module.exports = ***REMOVED***
    meta: ***REMOVED***
        type: "suggestion",

        docs: ***REMOVED***
            description: "disallow the use of `alert`, `confirm`, and `prompt`",
            category: "Best Practices",
            recommended: false,
            url: "https://eslint.org/docs/rules/no-alert"
        ***REMOVED***,

        schema: [],

        messages: ***REMOVED***
            unexpected: "Unexpected ***REMOVED******REMOVED***name***REMOVED******REMOVED***."
        ***REMOVED***
    ***REMOVED***,

    create(context) ***REMOVED***
        return ***REMOVED***
            CallExpression(node) ***REMOVED***
                const callee = node.callee,
                    currentScope = context.getScope();

                // without window.
                if (callee.type === "Identifier") ***REMOVED***
                    const name = callee.name;

                    if (!isShadowed(currentScope, callee) && isProhibitedIdentifier(callee.name)) ***REMOVED***
                        context.report(***REMOVED***
                            node,
                            messageId: "unexpected",
                            data: ***REMOVED*** name ***REMOVED***
                        ***REMOVED***);
                    ***REMOVED***

                ***REMOVED*** else if (callee.type === "MemberExpression" && isGlobalThisReferenceOrGlobalWindow(currentScope, callee.object)) ***REMOVED***
                    const name = getPropertyName(callee);

                    if (isProhibitedIdentifier(name)) ***REMOVED***
                        context.report(***REMOVED***
                            node,
                            messageId: "unexpected",
                            data: ***REMOVED*** name ***REMOVED***
                        ***REMOVED***);
                    ***REMOVED***
                ***REMOVED***

            ***REMOVED***
        ***REMOVED***;

    ***REMOVED***
***REMOVED***;
