// eslint/lib/rules/no-buffer-constructor.js
/**
 * @fileoverview disallow use of the Buffer() constructor
 * @author Teddy Katz
 */
"use strict";

//------------------------------------------------------------------------------
// Rule Definition
//------------------------------------------------------------------------------

module.exports = ***REMOVED***
    meta: ***REMOVED***
        type: "problem",

        docs: ***REMOVED***
            description: "disallow use of the `Buffer()` constructor",
            category: "Node.js and CommonJS",
            recommended: false,
            url: "https://eslint.org/docs/rules/no-buffer-constructor"
        ***REMOVED***,

        schema: [],

        messages: ***REMOVED***
            deprecated: "***REMOVED******REMOVED***expr***REMOVED******REMOVED*** is deprecated. Use Buffer.from(), Buffer.alloc(), or Buffer.allocUnsafe() instead."
        ***REMOVED***
    ***REMOVED***,

    create(context) ***REMOVED***

        //----------------------------------------------------------------------
        // Public
        //----------------------------------------------------------------------

        return ***REMOVED***
            "CallExpression[callee.name='Buffer'], NewExpression[callee.name='Buffer']"(node) ***REMOVED***
                context.report(***REMOVED***
                    node,
                    messageId: "deprecated",
                    data: ***REMOVED*** expr: node.type === "CallExpression" ? "Buffer()" : "new Buffer()" ***REMOVED***
                ***REMOVED***);
            ***REMOVED***
        ***REMOVED***;
    ***REMOVED***
***REMOVED***;
