// eslint/lib/rules/no-control-regex.js
/**
 * @fileoverview Rule to forbid control charactes from regular expressions.
 * @author Nicholas C. Zakas
 */

"use strict";

const RegExpValidator = require("regexpp").RegExpValidator;
const collector = new (class ***REMOVED***
    constructor() ***REMOVED***
        this.ecmaVersion = 2018;
        this._source = "";
        this._controlChars = [];
        this._validator = new RegExpValidator(this);
    ***REMOVED***

    onPatternEnter() ***REMOVED***
        this._controlChars = [];
    ***REMOVED***

    onCharacter(start, end, cp) ***REMOVED***
        if (cp >= 0x00 &&
            cp <= 0x1F &&
            (
                this._source.codePointAt(start) === cp ||
                this._source.slice(start, end).startsWith("\\x") ||
                this._source.slice(start, end).startsWith("\\u")
            )
        ) ***REMOVED***
            this._controlChars.push(`\\x$***REMOVED***`0$***REMOVED***cp.toString(16)***REMOVED***`.slice(-2)***REMOVED***`);
        ***REMOVED***
    ***REMOVED***

    collectControlChars(regexpStr) ***REMOVED***
        try ***REMOVED***
            this._source = regexpStr;
            this._validator.validatePattern(regexpStr); // Call onCharacter hook
        ***REMOVED*** catch (err) ***REMOVED***

            // Ignore syntax errors in RegExp.
        ***REMOVED***
        return this._controlChars;
    ***REMOVED***
***REMOVED***)();

//------------------------------------------------------------------------------
// Rule Definition
//------------------------------------------------------------------------------

module.exports = ***REMOVED***
    meta: ***REMOVED***
        type: "problem",

        docs: ***REMOVED***
            description: "disallow control characters in regular expressions",
            category: "Possible Errors",
            recommended: true,
            url: "https://eslint.org/docs/rules/no-control-regex"
        ***REMOVED***,

        schema: [],

        messages: ***REMOVED***
            unexpected: "Unexpected control character(s) in regular expression: ***REMOVED******REMOVED***controlChars***REMOVED******REMOVED***."
        ***REMOVED***
    ***REMOVED***,

    create(context) ***REMOVED***

        /**
         * Get the regex expression
         * @param ***REMOVED***ASTNode***REMOVED*** node node to evaluate
         * @returns ***REMOVED***RegExp|null***REMOVED*** Regex if found else null
         * @private
         */
        function getRegExpPattern(node) ***REMOVED***
            if (node.regex) ***REMOVED***
                return node.regex.pattern;
            ***REMOVED***
            if (typeof node.value === "string" &&
                (node.parent.type === "NewExpression" || node.parent.type === "CallExpression") &&
                node.parent.callee.type === "Identifier" &&
                node.parent.callee.name === "RegExp" &&
                node.parent.arguments[0] === node
            ) ***REMOVED***
                return node.value;
            ***REMOVED***

            return null;
        ***REMOVED***

        return ***REMOVED***
            Literal(node) ***REMOVED***
                const pattern = getRegExpPattern(node);

                if (pattern) ***REMOVED***
                    const controlCharacters = collector.collectControlChars(pattern);

                    if (controlCharacters.length > 0) ***REMOVED***
                        context.report(***REMOVED***
                            node,
                            messageId: "unexpected",
                            data: ***REMOVED***
                                controlChars: controlCharacters.join(", ")
                            ***REMOVED***
                        ***REMOVED***);
                    ***REMOVED***
                ***REMOVED***
            ***REMOVED***
        ***REMOVED***;

    ***REMOVED***
***REMOVED***;
