// eslint/lib/rules/no-class-assign.js
/**
 * @fileoverview A rule to disallow modifying variables of class declarations
 * @author Toru Nagashima
 */

"use strict";

const astUtils = require("./utils/ast-utils");

//------------------------------------------------------------------------------
// Rule Definition
//------------------------------------------------------------------------------

module.exports = ***REMOVED***
    meta: ***REMOVED***
        type: "problem",

        docs: ***REMOVED***
            description: "disallow reassigning class members",
            category: "ECMAScript 6",
            recommended: true,
            url: "https://eslint.org/docs/rules/no-class-assign"
        ***REMOVED***,

        schema: [],

        messages: ***REMOVED***
            class: "'***REMOVED******REMOVED***name***REMOVED******REMOVED***' is a class."
        ***REMOVED***
    ***REMOVED***,

    create(context) ***REMOVED***

        /**
         * Finds and reports references that are non initializer and writable.
         * @param ***REMOVED***Variable***REMOVED*** variable - A variable to check.
         * @returns ***REMOVED***void***REMOVED***
         */
        function checkVariable(variable) ***REMOVED***
            astUtils.getModifyingReferences(variable.references).forEach(reference => ***REMOVED***
                context.report(***REMOVED*** node: reference.identifier, messageId: "class", data: ***REMOVED*** name: reference.identifier.name ***REMOVED*** ***REMOVED***);

            ***REMOVED***);
        ***REMOVED***

        /**
         * Finds and reports references that are non initializer and writable.
         * @param ***REMOVED***ASTNode***REMOVED*** node - A ClassDeclaration/ClassExpression node to check.
         * @returns ***REMOVED***void***REMOVED***
         */
        function checkForClass(node) ***REMOVED***
            context.getDeclaredVariables(node).forEach(checkVariable);
        ***REMOVED***

        return ***REMOVED***
            ClassDeclaration: checkForClass,
            ClassExpression: checkForClass
        ***REMOVED***;

    ***REMOVED***
***REMOVED***;
