// eslint/lib/rules/no-console.js
/**
 * @fileoverview Rule to flag use of console object
 * @author Nicholas C. Zakas
 */

"use strict";

//------------------------------------------------------------------------------
// Requirements
//------------------------------------------------------------------------------

const astUtils = require("./utils/ast-utils");

//------------------------------------------------------------------------------
// Rule Definition
//------------------------------------------------------------------------------

module.exports = ***REMOVED***
    meta: ***REMOVED***
        type: "suggestion",

        docs: ***REMOVED***
            description: "disallow the use of `console`",
            category: "Possible Errors",
            recommended: false,
            url: "https://eslint.org/docs/rules/no-console"
        ***REMOVED***,

        schema: [
            ***REMOVED***
                type: "object",
                properties: ***REMOVED***
                    allow: ***REMOVED***
                        type: "array",
                        items: ***REMOVED***
                            type: "string"
                        ***REMOVED***,
                        minItems: 1,
                        uniqueItems: true
                    ***REMOVED***
                ***REMOVED***,
                additionalProperties: false
            ***REMOVED***
        ],

        messages: ***REMOVED***
            unexpected: "Unexpected console statement."
        ***REMOVED***
    ***REMOVED***,

    create(context) ***REMOVED***
        const options = context.options[0] || ***REMOVED******REMOVED***;
        const allowed = options.allow || [];

        /**
         * Checks whether the given reference is 'console' or not.
         *
         * @param ***REMOVED***eslint-scope.Reference***REMOVED*** reference - The reference to check.
         * @returns ***REMOVED***boolean***REMOVED*** `true` if the reference is 'console'.
         */
        function isConsole(reference) ***REMOVED***
            const id = reference.identifier;

            return id && id.name === "console";
        ***REMOVED***

        /**
         * Checks whether the property name of the given MemberExpression node
         * is allowed by options or not.
         *
         * @param ***REMOVED***ASTNode***REMOVED*** node - The MemberExpression node to check.
         * @returns ***REMOVED***boolean***REMOVED*** `true` if the property name of the node is allowed.
         */
        function isAllowed(node) ***REMOVED***
            const propertyName = astUtils.getStaticPropertyName(node);

            return propertyName && allowed.indexOf(propertyName) !== -1;
        ***REMOVED***

        /**
         * Checks whether the given reference is a member access which is not
         * allowed by options or not.
         *
         * @param ***REMOVED***eslint-scope.Reference***REMOVED*** reference - The reference to check.
         * @returns ***REMOVED***boolean***REMOVED*** `true` if the reference is a member access which
         *      is not allowed by options.
         */
        function isMemberAccessExceptAllowed(reference) ***REMOVED***
            const node = reference.identifier;
            const parent = node.parent;

            return (
                parent.type === "MemberExpression" &&
                parent.object === node &&
                !isAllowed(parent)
            );
        ***REMOVED***

        /**
         * Reports the given reference as a violation.
         *
         * @param ***REMOVED***eslint-scope.Reference***REMOVED*** reference - The reference to report.
         * @returns ***REMOVED***void***REMOVED***
         */
        function report(reference) ***REMOVED***
            const node = reference.identifier.parent;

            context.report(***REMOVED***
                node,
                loc: node.loc,
                messageId: "unexpected"
            ***REMOVED***);
        ***REMOVED***

        return ***REMOVED***
            "Program:exit"() ***REMOVED***
                const scope = context.getScope();
                const consoleVar = astUtils.getVariableByName(scope, "console");
                const shadowed = consoleVar && consoleVar.defs.length > 0;

                /*
                 * 'scope.through' includes all references to undefined
                 * variables. If the variable 'console' is not defined, it uses
                 * 'scope.through'.
                 */
                const references = consoleVar
                    ? consoleVar.references
                    : scope.through.filter(isConsole);

                if (!shadowed) ***REMOVED***
                    references
                        .filter(isMemberAccessExceptAllowed)
                        .forEach(report);
                ***REMOVED***
            ***REMOVED***
        ***REMOVED***;
    ***REMOVED***
***REMOVED***;
