// eslint/lib/rules/no-compare-neg-zero.js
/**
 * @fileoverview The rule should warn against code that tries to compare against -0.
 * @author Aladdin-ADD <hh_2013@foxmail.com>
 */
"use strict";

//------------------------------------------------------------------------------
// Rule Definition
//------------------------------------------------------------------------------

module.exports = ***REMOVED***
    meta: ***REMOVED***
        type: "problem",

        docs: ***REMOVED***
            description: "disallow comparing against -0",
            category: "Possible Errors",
            recommended: true,
            url: "https://eslint.org/docs/rules/no-compare-neg-zero"
        ***REMOVED***,

        fixable: null,
        schema: [],

        messages: ***REMOVED***
            unexpected: "Do not use the '***REMOVED******REMOVED***operator***REMOVED******REMOVED***' operator to compare against -0."
        ***REMOVED***
    ***REMOVED***,

    create(context) ***REMOVED***

        //--------------------------------------------------------------------------
        // Helpers
        //--------------------------------------------------------------------------

        /**
         * Checks a given node is -0
         *
         * @param ***REMOVED***ASTNode***REMOVED*** node - A node to check.
         * @returns ***REMOVED***boolean***REMOVED*** `true` if the node is -0.
         */
        function isNegZero(node) ***REMOVED***
            return node.type === "UnaryExpression" && node.operator === "-" && node.argument.type === "Literal" && node.argument.value === 0;
        ***REMOVED***
        const OPERATORS_TO_CHECK = new Set([">", ">=", "<", "<=", "==", "===", "!=", "!=="]);

        return ***REMOVED***
            BinaryExpression(node) ***REMOVED***
                if (OPERATORS_TO_CHECK.has(node.operator)) ***REMOVED***
                    if (isNegZero(node.left) || isNegZero(node.right)) ***REMOVED***
                        context.report(***REMOVED***
                            node,
                            messageId: "unexpected",
                            data: ***REMOVED*** operator: node.operator ***REMOVED***
                        ***REMOVED***);
                    ***REMOVED***
                ***REMOVED***
            ***REMOVED***
        ***REMOVED***;
    ***REMOVED***
***REMOVED***;
