// eslint/lib/rules/no-dupe-keys.js
/**
 * @fileoverview Rule to flag use of duplicate keys in an object.
 * @author Ian Christian Myers
 */

"use strict";

//------------------------------------------------------------------------------
// Requirements
//------------------------------------------------------------------------------

const astUtils = require("./utils/ast-utils");

//------------------------------------------------------------------------------
// Helpers
//------------------------------------------------------------------------------

const GET_KIND = /^(?:init|get)$/u;
const SET_KIND = /^(?:init|set)$/u;

/**
 * The class which stores properties' information of an object.
 */
class ObjectInfo ***REMOVED***

    /**
     * @param ***REMOVED***ObjectInfo|null***REMOVED*** upper - The information of the outer object.
     * @param ***REMOVED***ASTNode***REMOVED*** node - The ObjectExpression node of this information.
     */
    constructor(upper, node) ***REMOVED***
        this.upper = upper;
        this.node = node;
        this.properties = new Map();
    ***REMOVED***

    /**
     * Gets the information of the given Property node.
     * @param ***REMOVED***ASTNode***REMOVED*** node - The Property node to get.
     * @returns ***REMOVED******REMOVED***get: boolean, set: boolean***REMOVED******REMOVED*** The information of the property.
     */
    getPropertyInfo(node) ***REMOVED***
        const name = astUtils.getStaticPropertyName(node);

        if (!this.properties.has(name)) ***REMOVED***
            this.properties.set(name, ***REMOVED*** get: false, set: false ***REMOVED***);
        ***REMOVED***
        return this.properties.get(name);
    ***REMOVED***

    /**
     * Checks whether the given property has been defined already or not.
     * @param ***REMOVED***ASTNode***REMOVED*** node - The Property node to check.
     * @returns ***REMOVED***boolean***REMOVED*** `true` if the property has been defined.
     */
    isPropertyDefined(node) ***REMOVED***
        const entry = this.getPropertyInfo(node);

        return (
            (GET_KIND.test(node.kind) && entry.get) ||
            (SET_KIND.test(node.kind) && entry.set)
        );
    ***REMOVED***

    /**
     * Defines the given property.
     * @param ***REMOVED***ASTNode***REMOVED*** node - The Property node to define.
     * @returns ***REMOVED***void***REMOVED***
     */
    defineProperty(node) ***REMOVED***
        const entry = this.getPropertyInfo(node);

        if (GET_KIND.test(node.kind)) ***REMOVED***
            entry.get = true;
        ***REMOVED***
        if (SET_KIND.test(node.kind)) ***REMOVED***
            entry.set = true;
        ***REMOVED***
    ***REMOVED***
***REMOVED***

//------------------------------------------------------------------------------
// Rule Definition
//------------------------------------------------------------------------------

module.exports = ***REMOVED***
    meta: ***REMOVED***
        type: "problem",

        docs: ***REMOVED***
            description: "disallow duplicate keys in object literals",
            category: "Possible Errors",
            recommended: true,
            url: "https://eslint.org/docs/rules/no-dupe-keys"
        ***REMOVED***,

        schema: [],

        messages: ***REMOVED***
            unexpected: "Duplicate key '***REMOVED******REMOVED***name***REMOVED******REMOVED***'."
        ***REMOVED***
    ***REMOVED***,

    create(context) ***REMOVED***
        let info = null;

        return ***REMOVED***
            ObjectExpression(node) ***REMOVED***
                info = new ObjectInfo(info, node);
            ***REMOVED***,
            "ObjectExpression:exit"() ***REMOVED***
                info = info.upper;
            ***REMOVED***,

            Property(node) ***REMOVED***
                const name = astUtils.getStaticPropertyName(node);

                // Skip destructuring.
                if (node.parent.type !== "ObjectExpression") ***REMOVED***
                    return;
                ***REMOVED***

                // Skip if the name is not static.
                if (name === null) ***REMOVED***
                    return;
                ***REMOVED***

                // Reports if the name is defined already.
                if (info.isPropertyDefined(node)) ***REMOVED***
                    context.report(***REMOVED***
                        node: info.node,
                        loc: node.key.loc,
                        messageId: "unexpected",
                        data: ***REMOVED*** name ***REMOVED***
                    ***REMOVED***);
                ***REMOVED***

                // Update info.
                info.defineProperty(node);
            ***REMOVED***
        ***REMOVED***;
    ***REMOVED***
***REMOVED***;
