// eslint/lib/rules/no-duplicate-case.js
/**
 * @fileoverview Rule to disallow a duplicate case label.
 * @author Dieter Oberkofler
 * @author Burak Yigit Kaya
 */

"use strict";

//------------------------------------------------------------------------------
// Rule Definition
//------------------------------------------------------------------------------

module.exports = ***REMOVED***
    meta: ***REMOVED***
        type: "problem",

        docs: ***REMOVED***
            description: "disallow duplicate case labels",
            category: "Possible Errors",
            recommended: true,
            url: "https://eslint.org/docs/rules/no-duplicate-case"
        ***REMOVED***,

        schema: [],

        messages: ***REMOVED***
            unexpected: "Duplicate case label."
        ***REMOVED***
    ***REMOVED***,

    create(context) ***REMOVED***
        const sourceCode = context.getSourceCode();

        return ***REMOVED***
            SwitchStatement(node) ***REMOVED***
                const previousKeys = new Set();

                for (const switchCase of node.cases) ***REMOVED***
                    if (switchCase.test) ***REMOVED***
                        const key = sourceCode.getText(switchCase.test);

                        if (previousKeys.has(key)) ***REMOVED***
                            context.report(***REMOVED*** node: switchCase, messageId: "unexpected" ***REMOVED***);
                        ***REMOVED*** else ***REMOVED***
                            previousKeys.add(key);
                        ***REMOVED***
                    ***REMOVED***
                ***REMOVED***
            ***REMOVED***
        ***REMOVED***;
    ***REMOVED***
***REMOVED***;
