// eslint/lib/rules/no-div-regex.js
/**
 * @fileoverview Rule to check for ambiguous div operator in regexes
 * @author Matt DuVall <http://www.mattduvall.com>
 */

"use strict";

//------------------------------------------------------------------------------
// Rule Definition
//------------------------------------------------------------------------------

module.exports = ***REMOVED***
    meta: ***REMOVED***
        type: "suggestion",

        docs: ***REMOVED***
            description: "disallow division operators explicitly at the beginning of regular expressions",
            category: "Best Practices",
            recommended: false,
            url: "https://eslint.org/docs/rules/no-div-regex"
        ***REMOVED***,

        fixable: "code",

        schema: [],

        messages: ***REMOVED***
            unexpected: "A regular expression literal can be confused with '/='."
        ***REMOVED***
    ***REMOVED***,

    create(context) ***REMOVED***
        const sourceCode = context.getSourceCode();

        return ***REMOVED***

            Literal(node) ***REMOVED***
                const token = sourceCode.getFirstToken(node);

                if (token.type === "RegularExpression" && token.value[1] === "=") ***REMOVED***
                    context.report(***REMOVED***
                        node,
                        messageId: "unexpected",
                        fix(fixer) ***REMOVED***
                            return fixer.replaceTextRange([token.range[0] + 1, token.range[0] + 2], "[=]");
                        ***REMOVED***
                    ***REMOVED***);
                ***REMOVED***
            ***REMOVED***
        ***REMOVED***;

    ***REMOVED***
***REMOVED***;
