// eslint/lib/rules/no-dupe-class-members.js
/**
 * @fileoverview A rule to disallow duplicate name in class members.
 * @author Toru Nagashima
 */

"use strict";

//------------------------------------------------------------------------------
// Rule Definition
//------------------------------------------------------------------------------

module.exports = ***REMOVED***
    meta: ***REMOVED***
        type: "problem",

        docs: ***REMOVED***
            description: "disallow duplicate class members",
            category: "ECMAScript 6",
            recommended: true,
            url: "https://eslint.org/docs/rules/no-dupe-class-members"
        ***REMOVED***,

        schema: [],

        messages: ***REMOVED***
            unexpected: "Duplicate name '***REMOVED******REMOVED***name***REMOVED******REMOVED***'."
        ***REMOVED***
    ***REMOVED***,

    create(context) ***REMOVED***
        let stack = [];

        /**
         * Gets state of a given member name.
         * @param ***REMOVED***string***REMOVED*** name - A name of a member.
         * @param ***REMOVED***boolean***REMOVED*** isStatic - A flag which specifies that is a static member.
         * @returns ***REMOVED***Object***REMOVED*** A state of a given member name.
         *   - retv.init ***REMOVED***boolean***REMOVED*** A flag which shows the name is declared as normal member.
         *   - retv.get ***REMOVED***boolean***REMOVED*** A flag which shows the name is declared as getter.
         *   - retv.set ***REMOVED***boolean***REMOVED*** A flag which shows the name is declared as setter.
         */
        function getState(name, isStatic) ***REMOVED***
            const stateMap = stack[stack.length - 1];
            const key = `$$***REMOVED***name***REMOVED***`; // to avoid "__proto__".

            if (!stateMap[key]) ***REMOVED***
                stateMap[key] = ***REMOVED***
                    nonStatic: ***REMOVED*** init: false, get: false, set: false ***REMOVED***,
                    static: ***REMOVED*** init: false, get: false, set: false ***REMOVED***
                ***REMOVED***;
            ***REMOVED***

            return stateMap[key][isStatic ? "static" : "nonStatic"];
        ***REMOVED***

        /**
         * Gets the name text of a given node.
         *
         * @param ***REMOVED***ASTNode***REMOVED*** node - A node to get the name.
         * @returns ***REMOVED***string***REMOVED*** The name text of the node.
         */
        function getName(node) ***REMOVED***
            switch (node.type) ***REMOVED***
                case "Identifier": return node.name;
                case "Literal": return String(node.value);

                /* istanbul ignore next: syntax error */
                default: return "";
            ***REMOVED***
        ***REMOVED***

        return ***REMOVED***

            // Initializes the stack of state of member declarations.
            Program() ***REMOVED***
                stack = [];
            ***REMOVED***,

            // Initializes state of member declarations for the class.
            ClassBody() ***REMOVED***
                stack.push(Object.create(null));
            ***REMOVED***,

            // Disposes the state for the class.
            "ClassBody:exit"() ***REMOVED***
                stack.pop();
            ***REMOVED***,

            // Reports the node if its name has been declared already.
            MethodDefinition(node) ***REMOVED***
                if (node.computed) ***REMOVED***
                    return;
                ***REMOVED***

                const name = getName(node.key);
                const state = getState(name, node.static);
                let isDuplicate = false;

                if (node.kind === "get") ***REMOVED***
                    isDuplicate = (state.init || state.get);
                    state.get = true;
                ***REMOVED*** else if (node.kind === "set") ***REMOVED***
                    isDuplicate = (state.init || state.set);
                    state.set = true;
                ***REMOVED*** else ***REMOVED***
                    isDuplicate = (state.init || state.get || state.set);
                    state.init = true;
                ***REMOVED***

                if (isDuplicate) ***REMOVED***
                    context.report(***REMOVED*** node, messageId: "unexpected", data: ***REMOVED*** name ***REMOVED*** ***REMOVED***);
                ***REMOVED***
            ***REMOVED***
        ***REMOVED***;
    ***REMOVED***
***REMOVED***;
