// eslint/lib/rules/no-delete-var.js
/**
 * @fileoverview Rule to flag when deleting variables
 * @author Ilya Volodin
 */

"use strict";

//------------------------------------------------------------------------------
// Rule Definition
//------------------------------------------------------------------------------

module.exports = ***REMOVED***
    meta: ***REMOVED***
        type: "suggestion",

        docs: ***REMOVED***
            description: "disallow deleting variables",
            category: "Variables",
            recommended: true,
            url: "https://eslint.org/docs/rules/no-delete-var"
        ***REMOVED***,

        schema: [],

        messages: ***REMOVED***
            unexpected: "Variables should not be deleted."
        ***REMOVED***
    ***REMOVED***,

    create(context) ***REMOVED***

        return ***REMOVED***

            UnaryExpression(node) ***REMOVED***
                if (node.operator === "delete" && node.argument.type === "Identifier") ***REMOVED***
                    context.report(***REMOVED*** node, messageId: "unexpected" ***REMOVED***);
                ***REMOVED***
            ***REMOVED***
        ***REMOVED***;

    ***REMOVED***
***REMOVED***;
