// eslint/lib/rules/no-debugger.js
/**
 * @fileoverview Rule to flag use of a debugger statement
 * @author Nicholas C. Zakas
 */

"use strict";

//------------------------------------------------------------------------------
// Rule Definition
//------------------------------------------------------------------------------

module.exports = ***REMOVED***
    meta: ***REMOVED***
        type: "problem",

        docs: ***REMOVED***
            description: "disallow the use of `debugger`",
            category: "Possible Errors",
            recommended: true,
            url: "https://eslint.org/docs/rules/no-debugger"
        ***REMOVED***,

        fixable: null,
        schema: [],

        messages: ***REMOVED***
            unexpected: "Unexpected 'debugger' statement."
        ***REMOVED***
    ***REMOVED***,

    create(context) ***REMOVED***

        return ***REMOVED***
            DebuggerStatement(node) ***REMOVED***
                context.report(***REMOVED***
                    node,
                    messageId: "unexpected"
                ***REMOVED***);
            ***REMOVED***
        ***REMOVED***;

    ***REMOVED***
***REMOVED***;
