// eslint/lib/rules/no-empty-pattern.js
/**
 * @fileoverview Rule to disallow an empty pattern
 * @author Alberto Rodríguez
 */
"use strict";

//------------------------------------------------------------------------------
// Rule Definition
//------------------------------------------------------------------------------

module.exports = ***REMOVED***
    meta: ***REMOVED***
        type: "problem",

        docs: ***REMOVED***
            description: "disallow empty destructuring patterns",
            category: "Best Practices",
            recommended: true,
            url: "https://eslint.org/docs/rules/no-empty-pattern"
        ***REMOVED***,

        schema: [],

        messages: ***REMOVED***
            unexpected: "Unexpected empty ***REMOVED******REMOVED***type***REMOVED******REMOVED*** pattern."
        ***REMOVED***
    ***REMOVED***,

    create(context) ***REMOVED***
        return ***REMOVED***
            ObjectPattern(node) ***REMOVED***
                if (node.properties.length === 0) ***REMOVED***
                    context.report(***REMOVED*** node, messageId: "unexpected", data: ***REMOVED*** type: "object" ***REMOVED*** ***REMOVED***);
                ***REMOVED***
            ***REMOVED***,
            ArrayPattern(node) ***REMOVED***
                if (node.elements.length === 0) ***REMOVED***
                    context.report(***REMOVED*** node, messageId: "unexpected", data: ***REMOVED*** type: "array" ***REMOVED*** ***REMOVED***);
                ***REMOVED***
            ***REMOVED***
        ***REMOVED***;
    ***REMOVED***
***REMOVED***;
