// eslint/lib/rules/no-extend-native.js
/**
 * @fileoverview Rule to flag adding properties to native object's prototypes.
 * @author David Nelson
 */

"use strict";

//------------------------------------------------------------------------------
// Requirements
//------------------------------------------------------------------------------

const astUtils = require("./utils/ast-utils");
const globals = require("globals");

//------------------------------------------------------------------------------
// Helpers
//------------------------------------------------------------------------------

const propertyDefinitionMethods = new Set(["defineProperty", "defineProperties"]);

//------------------------------------------------------------------------------
// Rule Definition
//------------------------------------------------------------------------------

module.exports = ***REMOVED***
    meta: ***REMOVED***
        type: "suggestion",

        docs: ***REMOVED***
            description: "disallow extending native types",
            category: "Best Practices",
            recommended: false,
            url: "https://eslint.org/docs/rules/no-extend-native"
        ***REMOVED***,

        schema: [
            ***REMOVED***
                type: "object",
                properties: ***REMOVED***
                    exceptions: ***REMOVED***
                        type: "array",
                        items: ***REMOVED***
                            type: "string"
                        ***REMOVED***,
                        uniqueItems: true
                    ***REMOVED***
                ***REMOVED***,
                additionalProperties: false
            ***REMOVED***
        ],

        messages: ***REMOVED***
            unexpected: "***REMOVED******REMOVED***builtin***REMOVED******REMOVED*** prototype is read only, properties should not be added."
        ***REMOVED***
    ***REMOVED***,

    create(context) ***REMOVED***

        const config = context.options[0] || ***REMOVED******REMOVED***;
        const exceptions = new Set(config.exceptions || []);
        const modifiedBuiltins = new Set(
            Object.keys(globals.builtin)
                .filter(builtin => builtin[0].toUpperCase() === builtin[0])
                .filter(builtin => !exceptions.has(builtin))
        );

        /**
         * Reports a lint error for the given node.
         * @param ***REMOVED***ASTNode***REMOVED*** node The node to report.
         * @param ***REMOVED***string***REMOVED*** builtin The name of the native builtin being extended.
         * @returns ***REMOVED***void***REMOVED***
         */
        function reportNode(node, builtin) ***REMOVED***
            context.report(***REMOVED***
                node,
                messageId: "unexpected",
                data: ***REMOVED***
                    builtin
                ***REMOVED***
            ***REMOVED***);
        ***REMOVED***

        /**
         * Check to see if the `prototype` property of the given object
         * identifier node is being accessed.
         * @param ***REMOVED***ASTNode***REMOVED*** identifierNode The Identifier representing the object
         * to check.
         * @returns ***REMOVED***boolean***REMOVED*** True if the identifier is the object of a
         * MemberExpression and its `prototype` property is being accessed,
         * false otherwise.
         */
        function isPrototypePropertyAccessed(identifierNode) ***REMOVED***
            return Boolean(
                identifierNode &&
                identifierNode.parent &&
                identifierNode.parent.type === "MemberExpression" &&
                identifierNode.parent.object === identifierNode &&
                astUtils.getStaticPropertyName(identifierNode.parent) === "prototype"
            );
        ***REMOVED***

        /**
         * Checks that an identifier is an object of a prototype whose member
         * is being assigned in an AssignmentExpression.
         * Example: Object.prototype.foo = "bar"
         * @param ***REMOVED***ASTNode***REMOVED*** identifierNode The identifier to check.
         * @returns ***REMOVED***boolean***REMOVED*** True if the identifier's prototype is modified.
         */
        function isInPrototypePropertyAssignment(identifierNode) ***REMOVED***
            return Boolean(
                isPrototypePropertyAccessed(identifierNode) &&
                identifierNode.parent.parent.type === "MemberExpression" &&
                identifierNode.parent.parent.parent.type === "AssignmentExpression" &&
                identifierNode.parent.parent.parent.left === identifierNode.parent.parent
            );
        ***REMOVED***

        /**
         * Checks that an identifier is an object of a prototype whose member
         * is being extended via the Object.defineProperty() or
         * Object.defineProperties() methods.
         * Example: Object.defineProperty(Array.prototype, "foo", ...)
         * Example: Object.defineProperties(Array.prototype, ...)
         * @param ***REMOVED***ASTNode***REMOVED*** identifierNode The identifier to check.
         * @returns ***REMOVED***boolean***REMOVED*** True if the identifier's prototype is modified.
         */
        function isInDefinePropertyCall(identifierNode) ***REMOVED***
            return Boolean(
                isPrototypePropertyAccessed(identifierNode) &&
                identifierNode.parent.parent.type === "CallExpression" &&
                identifierNode.parent.parent.arguments[0] === identifierNode.parent &&
                identifierNode.parent.parent.callee.type === "MemberExpression" &&
                identifierNode.parent.parent.callee.object.type === "Identifier" &&
                identifierNode.parent.parent.callee.object.name === "Object" &&
                identifierNode.parent.parent.callee.property.type === "Identifier" &&
                propertyDefinitionMethods.has(identifierNode.parent.parent.callee.property.name)
            );
        ***REMOVED***

        /**
         * Check to see if object prototype access is part of a prototype
         * extension. There are three ways a prototype can be extended:
         * 1. Assignment to prototype property (Object.prototype.foo = 1)
         * 2. Object.defineProperty()/Object.defineProperties() on a prototype
         * If prototype extension is detected, report the AssignmentExpression
         * or CallExpression node.
         * @param ***REMOVED***ASTNode***REMOVED*** identifierNode The Identifier representing the object
         * which prototype is being accessed and possibly extended.
         * @returns ***REMOVED***void***REMOVED***
         */
        function checkAndReportPrototypeExtension(identifierNode) ***REMOVED***
            if (isInPrototypePropertyAssignment(identifierNode)) ***REMOVED***

                // Identifier --> MemberExpression --> MemberExpression --> AssignmentExpression
                reportNode(identifierNode.parent.parent.parent, identifierNode.name);
            ***REMOVED*** else if (isInDefinePropertyCall(identifierNode)) ***REMOVED***

                // Identifier --> MemberExpression --> CallExpression
                reportNode(identifierNode.parent.parent, identifierNode.name);
            ***REMOVED***
        ***REMOVED***

        return ***REMOVED***

            "Program:exit"() ***REMOVED***
                const globalScope = context.getScope();

                modifiedBuiltins.forEach(builtin => ***REMOVED***
                    const builtinVar = globalScope.set.get(builtin);

                    if (builtinVar && builtinVar.references) ***REMOVED***
                        builtinVar.references
                            .map(ref => ref.identifier)
                            .forEach(checkAndReportPrototypeExtension);
                    ***REMOVED***
                ***REMOVED***);
            ***REMOVED***
        ***REMOVED***;

    ***REMOVED***
***REMOVED***;
