// eslint/lib/rules/no-ex-assign.js
/**
 * @fileoverview Rule to flag assignment of the exception parameter
 * @author Stephen Murray <spmurrayzzz>
 */

"use strict";

const astUtils = require("./utils/ast-utils");

//------------------------------------------------------------------------------
// Rule Definition
//------------------------------------------------------------------------------

module.exports = ***REMOVED***
    meta: ***REMOVED***
        type: "problem",

        docs: ***REMOVED***
            description: "disallow reassigning exceptions in `catch` clauses",
            category: "Possible Errors",
            recommended: true,
            url: "https://eslint.org/docs/rules/no-ex-assign"
        ***REMOVED***,

        schema: [],

        messages: ***REMOVED***
            unexpected: "Do not assign to the exception parameter."
        ***REMOVED***
    ***REMOVED***,

    create(context) ***REMOVED***

        /**
         * Finds and reports references that are non initializer and writable.
         * @param ***REMOVED***Variable***REMOVED*** variable - A variable to check.
         * @returns ***REMOVED***void***REMOVED***
         */
        function checkVariable(variable) ***REMOVED***
            astUtils.getModifyingReferences(variable.references).forEach(reference => ***REMOVED***
                context.report(***REMOVED*** node: reference.identifier, messageId: "unexpected" ***REMOVED***);
            ***REMOVED***);
        ***REMOVED***

        return ***REMOVED***
            CatchClause(node) ***REMOVED***
                context.getDeclaredVariables(node).forEach(checkVariable);
            ***REMOVED***
        ***REMOVED***;

    ***REMOVED***
***REMOVED***;
