// eslint/lib/rules/no-extra-semi.js
/**
 * @fileoverview Rule to flag use of unnecessary semicolons
 * @author Nicholas C. Zakas
 */

"use strict";

//------------------------------------------------------------------------------
// Requirements
//------------------------------------------------------------------------------

const FixTracker = require("./utils/fix-tracker");
const astUtils = require("./utils/ast-utils");

//------------------------------------------------------------------------------
// Rule Definition
//------------------------------------------------------------------------------

module.exports = ***REMOVED***
    meta: ***REMOVED***
        type: "suggestion",

        docs: ***REMOVED***
            description: "disallow unnecessary semicolons",
            category: "Possible Errors",
            recommended: true,
            url: "https://eslint.org/docs/rules/no-extra-semi"
        ***REMOVED***,

        fixable: "code",
        schema: [],

        messages: ***REMOVED***
            unexpected: "Unnecessary semicolon."
        ***REMOVED***
    ***REMOVED***,

    create(context) ***REMOVED***
        const sourceCode = context.getSourceCode();

        /**
         * Reports an unnecessary semicolon error.
         * @param ***REMOVED***Node|Token***REMOVED*** nodeOrToken - A node or a token to be reported.
         * @returns ***REMOVED***void***REMOVED***
         */
        function report(nodeOrToken) ***REMOVED***
            context.report(***REMOVED***
                node: nodeOrToken,
                messageId: "unexpected",
                fix(fixer) ***REMOVED***

                    /*
                     * Expand the replacement range to include the surrounding
                     * tokens to avoid conflicting with semi.
                     * https://github.com/eslint/eslint/issues/7928
                     */
                    return new FixTracker(fixer, context.getSourceCode())
                        .retainSurroundingTokens(nodeOrToken)
                        .remove(nodeOrToken);
                ***REMOVED***
            ***REMOVED***);
        ***REMOVED***

        /**
         * Checks for a part of a class body.
         * This checks tokens from a specified token to a next MethodDefinition or the end of class body.
         *
         * @param ***REMOVED***Token***REMOVED*** firstToken - The first token to check.
         * @returns ***REMOVED***void***REMOVED***
         */
        function checkForPartOfClassBody(firstToken) ***REMOVED***
            for (let token = firstToken;
                token.type === "Punctuator" && !astUtils.isClosingBraceToken(token);
                token = sourceCode.getTokenAfter(token)
            ) ***REMOVED***
                if (astUtils.isSemicolonToken(token)) ***REMOVED***
                    report(token);
                ***REMOVED***
            ***REMOVED***
        ***REMOVED***

        return ***REMOVED***

            /**
             * Reports this empty statement, except if the parent node is a loop.
             * @param ***REMOVED***Node***REMOVED*** node - A EmptyStatement node to be reported.
             * @returns ***REMOVED***void***REMOVED***
             */
            EmptyStatement(node) ***REMOVED***
                const parent = node.parent,
                    allowedParentTypes = [
                        "ForStatement",
                        "ForInStatement",
                        "ForOfStatement",
                        "WhileStatement",
                        "DoWhileStatement",
                        "IfStatement",
                        "LabeledStatement",
                        "WithStatement"
                    ];

                if (allowedParentTypes.indexOf(parent.type) === -1) ***REMOVED***
                    report(node);
                ***REMOVED***
            ***REMOVED***,

            /**
             * Checks tokens from the head of this class body to the first MethodDefinition or the end of this class body.
             * @param ***REMOVED***Node***REMOVED*** node - A ClassBody node to check.
             * @returns ***REMOVED***void***REMOVED***
             */
            ClassBody(node) ***REMOVED***
                checkForPartOfClassBody(sourceCode.getFirstToken(node, 1)); // 0 is `***REMOVED***`.
            ***REMOVED***,

            /**
             * Checks tokens from this MethodDefinition to the next MethodDefinition or the end of this class body.
             * @param ***REMOVED***Node***REMOVED*** node - A MethodDefinition node of the start point.
             * @returns ***REMOVED***void***REMOVED***
             */
            MethodDefinition(node) ***REMOVED***
                checkForPartOfClassBody(sourceCode.getTokenAfter(node));
            ***REMOVED***
        ***REMOVED***;

    ***REMOVED***
***REMOVED***;
