// eslint/lib/rules/no-mixed-requires.js
/**
 * @fileoverview Rule to enforce grouped require statements for Node.JS
 * @author Raphael Pigulla
 */

"use strict";

//------------------------------------------------------------------------------
// Rule Definition
//------------------------------------------------------------------------------

module.exports = ***REMOVED***
    meta: ***REMOVED***
        type: "suggestion",

        docs: ***REMOVED***
            description: "disallow `require` calls to be mixed with regular variable declarations",
            category: "Node.js and CommonJS",
            recommended: false,
            url: "https://eslint.org/docs/rules/no-mixed-requires"
        ***REMOVED***,

        schema: [
            ***REMOVED***
                oneOf: [
                    ***REMOVED***
                        type: "boolean"
                    ***REMOVED***,
                    ***REMOVED***
                        type: "object",
                        properties: ***REMOVED***
                            grouping: ***REMOVED***
                                type: "boolean"
                            ***REMOVED***,
                            allowCall: ***REMOVED***
                                type: "boolean"
                            ***REMOVED***
                        ***REMOVED***,
                        additionalProperties: false
                    ***REMOVED***
                ]
            ***REMOVED***
        ]
    ***REMOVED***,

    create(context) ***REMOVED***

        const options = context.options[0];
        let grouping = false,
            allowCall = false;

        if (typeof options === "object") ***REMOVED***
            grouping = options.grouping;
            allowCall = options.allowCall;
        ***REMOVED*** else ***REMOVED***
            grouping = !!options;
        ***REMOVED***

        /**
         * Returns the list of built-in modules.
         *
         * @returns ***REMOVED***string[]***REMOVED*** An array of built-in Node.js modules.
         */
        function getBuiltinModules() ***REMOVED***

            /*
             * This list is generated using:
             * `require("repl")._builtinLibs.concat('repl').sort()`
             * This particular list is as per nodejs v0.12.2 and iojs v0.7.1
             */
            return [
                "assert", "buffer", "child_process", "cluster", "crypto",
                "dgram", "dns", "domain", "events", "fs", "http", "https",
                "net", "os", "path", "punycode", "querystring", "readline",
                "repl", "smalloc", "stream", "string_decoder", "tls", "tty",
                "url", "util", "v8", "vm", "zlib"
            ];
        ***REMOVED***

        const BUILTIN_MODULES = getBuiltinModules();

        const DECL_REQUIRE = "require",
            DECL_UNINITIALIZED = "uninitialized",
            DECL_OTHER = "other";

        const REQ_CORE = "core",
            REQ_FILE = "file",
            REQ_MODULE = "module",
            REQ_COMPUTED = "computed";

        /**
         * Determines the type of a declaration statement.
         * @param ***REMOVED***ASTNode***REMOVED*** initExpression The init node of the VariableDeclarator.
         * @returns ***REMOVED***string***REMOVED*** The type of declaration represented by the expression.
         */
        function getDeclarationType(initExpression) ***REMOVED***
            if (!initExpression) ***REMOVED***

                // "var x;"
                return DECL_UNINITIALIZED;
            ***REMOVED***

            if (initExpression.type === "CallExpression" &&
                initExpression.callee.type === "Identifier" &&
                initExpression.callee.name === "require"
            ) ***REMOVED***

                // "var x = require('util');"
                return DECL_REQUIRE;
            ***REMOVED***
            if (allowCall &&
                initExpression.type === "CallExpression" &&
                initExpression.callee.type === "CallExpression"
            ) ***REMOVED***

                // "var x = require('diagnose')('sub-module');"
                return getDeclarationType(initExpression.callee);
            ***REMOVED***
            if (initExpression.type === "MemberExpression") ***REMOVED***

                // "var x = require('glob').Glob;"
                return getDeclarationType(initExpression.object);
            ***REMOVED***

            // "var x = 42;"
            return DECL_OTHER;
        ***REMOVED***

        /**
         * Determines the type of module that is loaded via require.
         * @param ***REMOVED***ASTNode***REMOVED*** initExpression The init node of the VariableDeclarator.
         * @returns ***REMOVED***string***REMOVED*** The module type.
         */
        function inferModuleType(initExpression) ***REMOVED***
            if (initExpression.type === "MemberExpression") ***REMOVED***

                // "var x = require('glob').Glob;"
                return inferModuleType(initExpression.object);
            ***REMOVED***
            if (initExpression.arguments.length === 0) ***REMOVED***

                // "var x = require();"
                return REQ_COMPUTED;
            ***REMOVED***

            const arg = initExpression.arguments[0];

            if (arg.type !== "Literal" || typeof arg.value !== "string") ***REMOVED***

                // "var x = require(42);"
                return REQ_COMPUTED;
            ***REMOVED***

            if (BUILTIN_MODULES.indexOf(arg.value) !== -1) ***REMOVED***

                // "var fs = require('fs');"
                return REQ_CORE;
            ***REMOVED***
            if (/^\.***REMOVED***0,2***REMOVED***\//u.test(arg.value)) ***REMOVED***

                // "var utils = require('./utils');"
                return REQ_FILE;
            ***REMOVED***

            // "var async = require('async');"
            return REQ_MODULE;

        ***REMOVED***

        /**
         * Check if the list of variable declarations is mixed, i.e. whether it
         * contains both require and other declarations.
         * @param ***REMOVED***ASTNode***REMOVED*** declarations The list of VariableDeclarators.
         * @returns ***REMOVED***boolean***REMOVED*** True if the declarations are mixed, false if not.
         */
        function isMixed(declarations) ***REMOVED***
            const contains = ***REMOVED******REMOVED***;

            declarations.forEach(declaration => ***REMOVED***
                const type = getDeclarationType(declaration.init);

                contains[type] = true;
            ***REMOVED***);

            return !!(
                contains[DECL_REQUIRE] &&
                (contains[DECL_UNINITIALIZED] || contains[DECL_OTHER])
            );
        ***REMOVED***

        /**
         * Check if all require declarations in the given list are of the same
         * type.
         * @param ***REMOVED***ASTNode***REMOVED*** declarations The list of VariableDeclarators.
         * @returns ***REMOVED***boolean***REMOVED*** True if the declarations are grouped, false if not.
         */
        function isGrouped(declarations) ***REMOVED***
            const found = ***REMOVED******REMOVED***;

            declarations.forEach(declaration => ***REMOVED***
                if (getDeclarationType(declaration.init) === DECL_REQUIRE) ***REMOVED***
                    found[inferModuleType(declaration.init)] = true;
                ***REMOVED***
            ***REMOVED***);

            return Object.keys(found).length <= 1;
        ***REMOVED***


        return ***REMOVED***

            VariableDeclaration(node) ***REMOVED***

                if (isMixed(node.declarations)) ***REMOVED***
                    context.report(***REMOVED*** node, message: "Do not mix 'require' and other declarations." ***REMOVED***);
                ***REMOVED*** else if (grouping && !isGrouped(node.declarations)) ***REMOVED***
                    context.report(***REMOVED*** node, message: "Do not mix core, module, file and computed requires." ***REMOVED***);
                ***REMOVED***
            ***REMOVED***
        ***REMOVED***;

    ***REMOVED***
***REMOVED***;
