// eslint/lib/rules/no-negated-in-lhs.js
/**
 * @fileoverview A rule to disallow negated left operands of the `in` operator
 * @author Michael Ficarra
 * @deprecated in ESLint v3.3.0
 */

"use strict";

//------------------------------------------------------------------------------
// Rule Definition
//------------------------------------------------------------------------------

module.exports = ***REMOVED***
    meta: ***REMOVED***
        type: "problem",

        docs: ***REMOVED***
            description: "disallow negating the left operand in `in` expressions",
            category: "Possible Errors",
            recommended: false,
            url: "https://eslint.org/docs/rules/no-negated-in-lhs"
        ***REMOVED***,

        replacedBy: ["no-unsafe-negation"],

        deprecated: true,
        schema: []
    ***REMOVED***,

    create(context) ***REMOVED***

        return ***REMOVED***

            BinaryExpression(node) ***REMOVED***
                if (node.operator === "in" && node.left.type === "UnaryExpression" && node.left.operator === "!") ***REMOVED***
                    context.report(***REMOVED*** node, message: "The 'in' expression's left operand is negated." ***REMOVED***);
                ***REMOVED***
            ***REMOVED***
        ***REMOVED***;

    ***REMOVED***
***REMOVED***;
