// eslint/lib/rules/no-process-env.js
/**
 * @fileoverview Disallow the use of process.env()
 * @author Vignesh Anand
 */
"use strict";

//------------------------------------------------------------------------------
// Rule Definition
//------------------------------------------------------------------------------

module.exports = ***REMOVED***
    meta: ***REMOVED***
        type: "suggestion",

        docs: ***REMOVED***
            description: "disallow the use of `process.env`",
            category: "Node.js and CommonJS",
            recommended: false,
            url: "https://eslint.org/docs/rules/no-process-env"
        ***REMOVED***,

        schema: []
    ***REMOVED***,

    create(context) ***REMOVED***

        return ***REMOVED***

            MemberExpression(node) ***REMOVED***
                const objectName = node.object.name,
                    propertyName = node.property.name;

                if (objectName === "process" && !node.computed && propertyName && propertyName === "env") ***REMOVED***
                    context.report(***REMOVED*** node, message: "Unexpected use of process.env." ***REMOVED***);
                ***REMOVED***

            ***REMOVED***

        ***REMOVED***;

    ***REMOVED***
***REMOVED***;
