// eslint/lib/rules/no-plusplus.js
/**
 * @fileoverview Rule to flag use of unary increment and decrement operators.
 * @author Ian Christian Myers
 * @author Brody McKee (github.com/mrmckeb)
 */

"use strict";

//------------------------------------------------------------------------------
// Rule Definition
//------------------------------------------------------------------------------

module.exports = ***REMOVED***
    meta: ***REMOVED***
        type: "suggestion",

        docs: ***REMOVED***
            description: "disallow the unary operators `++` and `--`",
            category: "Stylistic Issues",
            recommended: false,
            url: "https://eslint.org/docs/rules/no-plusplus"
        ***REMOVED***,

        schema: [
            ***REMOVED***
                type: "object",
                properties: ***REMOVED***
                    allowForLoopAfterthoughts: ***REMOVED***
                        type: "boolean",
                        default: false
                    ***REMOVED***
                ***REMOVED***,
                additionalProperties: false
            ***REMOVED***
        ]
    ***REMOVED***,

    create(context) ***REMOVED***

        const config = context.options[0];
        let allowInForAfterthought = false;

        if (typeof config === "object") ***REMOVED***
            allowInForAfterthought = config.allowForLoopAfterthoughts === true;
        ***REMOVED***

        return ***REMOVED***

            UpdateExpression(node) ***REMOVED***
                if (allowInForAfterthought && node.parent.type === "ForStatement") ***REMOVED***
                    return;
                ***REMOVED***
                context.report(***REMOVED***
                    node,
                    message: "Unary operator '***REMOVED******REMOVED***operator***REMOVED******REMOVED***' used.",
                    data: ***REMOVED***
                        operator: node.operator
                    ***REMOVED***
                ***REMOVED***);
            ***REMOVED***

        ***REMOVED***;

    ***REMOVED***
***REMOVED***;
