// eslint/lib/rules/no-redeclare.js
/**
 * @fileoverview Rule to flag when the same variable is declared more then once.
 * @author Ilya Volodin
 */

"use strict";

//------------------------------------------------------------------------------
// Requirements
//------------------------------------------------------------------------------

const astUtils = require("./utils/ast-utils");

//------------------------------------------------------------------------------
// Rule Definition
//------------------------------------------------------------------------------

module.exports = ***REMOVED***
    meta: ***REMOVED***
        type: "suggestion",

        docs: ***REMOVED***
            description: "disallow variable redeclaration",
            category: "Best Practices",
            recommended: true,
            url: "https://eslint.org/docs/rules/no-redeclare"
        ***REMOVED***,

        messages: ***REMOVED***
            redeclared: "'***REMOVED******REMOVED***id***REMOVED******REMOVED***' is already defined.",
            redeclaredAsBuiltin: "'***REMOVED******REMOVED***id***REMOVED******REMOVED***' is already defined as a built-in global variable.",
            redeclaredBySyntax: "'***REMOVED******REMOVED***id***REMOVED******REMOVED***' is already defined by a variable declaration."
        ***REMOVED***,

        schema: [
            ***REMOVED***
                type: "object",
                properties: ***REMOVED***
                    builtinGlobals: ***REMOVED*** type: "boolean", default: true ***REMOVED***
                ***REMOVED***,
                additionalProperties: false
            ***REMOVED***
        ]
    ***REMOVED***,

    create(context) ***REMOVED***
        const options = ***REMOVED***
            builtinGlobals: Boolean(
                context.options.length === 0 ||
                context.options[0].builtinGlobals
            )
        ***REMOVED***;
        const sourceCode = context.getSourceCode();

        /**
         * Iterate declarations of a given variable.
         * @param ***REMOVED***escope.variable***REMOVED*** variable The variable object to iterate declarations.
         * @returns ***REMOVED***IterableIterator<***REMOVED***type:string,node:ASTNode,loc:SourceLocation***REMOVED***>***REMOVED*** The declarations.
         */
        function *iterateDeclarations(variable) ***REMOVED***
            if (options.builtinGlobals && (
                variable.eslintImplicitGlobalSetting === "readonly" ||
                variable.eslintImplicitGlobalSetting === "writable"
            )) ***REMOVED***
                yield ***REMOVED*** type: "builtin" ***REMOVED***;
            ***REMOVED***

            for (const id of variable.identifiers) ***REMOVED***
                yield ***REMOVED*** type: "syntax", node: id, loc: id.loc ***REMOVED***;
            ***REMOVED***

            if (variable.eslintExplicitGlobalComments) ***REMOVED***
                for (const comment of variable.eslintExplicitGlobalComments) ***REMOVED***
                    yield ***REMOVED***
                        type: "comment",
                        node: comment,
                        loc: astUtils.getNameLocationInGlobalDirectiveComment(
                            sourceCode,
                            comment,
                            variable.name
                        )
                    ***REMOVED***;
                ***REMOVED***
            ***REMOVED***
        ***REMOVED***

        /**
         * Find variables in a given scope and flag redeclared ones.
         * @param ***REMOVED***Scope***REMOVED*** scope - An eslint-scope scope object.
         * @returns ***REMOVED***void***REMOVED***
         * @private
         */
        function findVariablesInScope(scope) ***REMOVED***
            for (const variable of scope.variables) ***REMOVED***
                const [
                    declaration,
                    ...extraDeclarations
                ] = iterateDeclarations(variable);

                if (extraDeclarations.length === 0) ***REMOVED***
                    continue;
                ***REMOVED***

                /*
                 * If the type of a declaration is different from the type of
                 * the first declaration, it shows the location of the first
                 * declaration.
                 */
                const detailMessageId = declaration.type === "builtin"
                    ? "redeclaredAsBuiltin"
                    : "redeclaredBySyntax";
                const data = ***REMOVED*** id: variable.name ***REMOVED***;

                // Report extra declarations.
                for (const ***REMOVED*** type, node, loc ***REMOVED*** of extraDeclarations) ***REMOVED***
                    const messageId = type === declaration.type
                        ? "redeclared"
                        : detailMessageId;

                    context.report(***REMOVED*** node, loc, messageId, data ***REMOVED***);
                ***REMOVED***
            ***REMOVED***
        ***REMOVED***

        /**
         * Find variables in the current scope.
         * @param ***REMOVED***ASTNode***REMOVED*** node The node of the current scope.
         * @returns ***REMOVED***void***REMOVED***
         * @private
         */
        function checkForBlock(node) ***REMOVED***
            const scope = context.getScope();

            /*
             * In ES5, some node type such as `BlockStatement` doesn't have that scope.
             * `scope.block` is a different node in such a case.
             */
            if (scope.block === node) ***REMOVED***
                findVariablesInScope(scope);
            ***REMOVED***
        ***REMOVED***

        return ***REMOVED***
            Program() ***REMOVED***
                const scope = context.getScope();

                findVariablesInScope(scope);

                // Node.js or ES modules has a special scope.
                if (
                    scope.type === "global" &&
                    scope.childScopes[0] &&

                    // The special scope's block is the Program node.
                    scope.block === scope.childScopes[0].block
                ) ***REMOVED***
                    findVariablesInScope(scope.childScopes[0]);
                ***REMOVED***
            ***REMOVED***,

            FunctionDeclaration: checkForBlock,
            FunctionExpression: checkForBlock,
            ArrowFunctionExpression: checkForBlock,

            BlockStatement: checkForBlock,
            ForStatement: checkForBlock,
            ForInStatement: checkForBlock,
            ForOfStatement: checkForBlock,
            SwitchStatement: checkForBlock
        ***REMOVED***;
    ***REMOVED***
***REMOVED***;
