// eslint/lib/rules/no-return-assign.js
/**
 * @fileoverview Rule to flag when return statement contains assignment
 * @author Ilya Volodin
 */
"use strict";

//------------------------------------------------------------------------------
// Requirements
//------------------------------------------------------------------------------

const astUtils = require("./utils/ast-utils");

//------------------------------------------------------------------------------
// Helpers
//------------------------------------------------------------------------------

const SENTINEL_TYPE = /^(?:[a-zA-Z]+?Statement|ArrowFunctionExpression|FunctionExpression|ClassExpression)$/u;

//------------------------------------------------------------------------------
// Rule Definition
//------------------------------------------------------------------------------

module.exports = ***REMOVED***
    meta: ***REMOVED***
        type: "suggestion",

        docs: ***REMOVED***
            description: "disallow assignment operators in `return` statements",
            category: "Best Practices",
            recommended: false,
            url: "https://eslint.org/docs/rules/no-return-assign"
        ***REMOVED***,

        schema: [
            ***REMOVED***
                enum: ["except-parens", "always"]
            ***REMOVED***
        ]
    ***REMOVED***,

    create(context) ***REMOVED***
        const always = (context.options[0] || "except-parens") !== "except-parens";
        const sourceCode = context.getSourceCode();

        return ***REMOVED***
            AssignmentExpression(node) ***REMOVED***
                if (!always && astUtils.isParenthesised(sourceCode, node)) ***REMOVED***
                    return;
                ***REMOVED***

                let currentChild = node;
                let parent = currentChild.parent;

                // Find ReturnStatement or ArrowFunctionExpression in ancestors.
                while (parent && !SENTINEL_TYPE.test(parent.type)) ***REMOVED***
                    currentChild = parent;
                    parent = parent.parent;
                ***REMOVED***

                // Reports.
                if (parent && parent.type === "ReturnStatement") ***REMOVED***
                    context.report(***REMOVED***
                        node: parent,
                        message: "Return statement should not contain assignment."
                    ***REMOVED***);
                ***REMOVED*** else if (parent && parent.type === "ArrowFunctionExpression" && parent.body === currentChild) ***REMOVED***
                    context.report(***REMOVED***
                        node: parent,
                        message: "Arrow function should not return assignment."
                    ***REMOVED***);
                ***REMOVED***
            ***REMOVED***
        ***REMOVED***;
    ***REMOVED***
***REMOVED***;
