// eslint/lib/rules/no-shadow-restricted-names.js
/**
 * @fileoverview Disallow shadowing of NaN, undefined, and Infinity (ES5 section 15.1.1)
 * @author Michael Ficarra
 */
"use strict";

/**
 * Determines if a variable safely shadows undefined.
 * This is the case when a variable named `undefined` is never assigned to a value (i.e. it always shares the same value
 * as the global).
 * @param ***REMOVED***eslintScope.Variable***REMOVED*** variable The variable to check
 * @returns ***REMOVED***boolean***REMOVED*** true if this variable safely shadows `undefined`
 */
function safelyShadowsUndefined(variable) ***REMOVED***
    return variable.name === "undefined" &&
        variable.references.every(ref => !ref.isWrite()) &&
        variable.defs.every(def => def.node.type === "VariableDeclarator" && def.node.init === null);
***REMOVED***

//------------------------------------------------------------------------------
// Rule Definition
//------------------------------------------------------------------------------

module.exports = ***REMOVED***
    meta: ***REMOVED***
        type: "suggestion",

        docs: ***REMOVED***
            description: "disallow identifiers from shadowing restricted names",
            category: "Variables",
            recommended: true,
            url: "https://eslint.org/docs/rules/no-shadow-restricted-names"
        ***REMOVED***,

        schema: []
    ***REMOVED***,

    create(context) ***REMOVED***


        const RESTRICTED = new Set(["undefined", "NaN", "Infinity", "arguments", "eval"]);

        return ***REMOVED***
            "VariableDeclaration, :function, CatchClause"(node) ***REMOVED***
                for (const variable of context.getDeclaredVariables(node)) ***REMOVED***
                    if (variable.defs.length > 0 && RESTRICTED.has(variable.name) && !safelyShadowsUndefined(variable)) ***REMOVED***
                        context.report(***REMOVED***
                            node: variable.defs[0].name,
                            message: "Shadowing of global property '***REMOVED******REMOVED***idName***REMOVED******REMOVED***'.",
                            data: ***REMOVED***
                                idName: variable.name
                            ***REMOVED***
                        ***REMOVED***);
                    ***REMOVED***
                ***REMOVED***
            ***REMOVED***
        ***REMOVED***;

    ***REMOVED***
***REMOVED***;
