// eslint/lib/rules/no-sparse-arrays.js
/**
 * @fileoverview Disallow sparse arrays
 * @author Nicholas C. Zakas
 */
"use strict";

//------------------------------------------------------------------------------
// Rule Definition
//------------------------------------------------------------------------------

module.exports = ***REMOVED***
    meta: ***REMOVED***
        type: "problem",

        docs: ***REMOVED***
            description: "disallow sparse arrays",
            category: "Possible Errors",
            recommended: true,
            url: "https://eslint.org/docs/rules/no-sparse-arrays"
        ***REMOVED***,

        schema: []
    ***REMOVED***,

    create(context) ***REMOVED***


        //--------------------------------------------------------------------------
        // Public
        //--------------------------------------------------------------------------

        return ***REMOVED***

            ArrayExpression(node) ***REMOVED***

                const emptySpot = node.elements.indexOf(null) > -1;

                if (emptySpot) ***REMOVED***
                    context.report(***REMOVED*** node, message: "Unexpected comma in middle of array." ***REMOVED***);
                ***REMOVED***
            ***REMOVED***

        ***REMOVED***;

    ***REMOVED***
***REMOVED***;
