// eslint/lib/rules/no-self-compare.js
/**
 * @fileoverview Rule to flag comparison where left part is the same as the right
 * part.
 * @author Ilya Volodin
 */

"use strict";

//------------------------------------------------------------------------------
// Rule Definition
//------------------------------------------------------------------------------

module.exports = ***REMOVED***
    meta: ***REMOVED***
        type: "problem",

        docs: ***REMOVED***
            description: "disallow comparisons where both sides are exactly the same",
            category: "Best Practices",
            recommended: false,
            url: "https://eslint.org/docs/rules/no-self-compare"
        ***REMOVED***,

        schema: []
    ***REMOVED***,

    create(context) ***REMOVED***
        const sourceCode = context.getSourceCode();

        /**
         * Determines whether two nodes are composed of the same tokens.
         * @param ***REMOVED***ASTNode***REMOVED*** nodeA The first node
         * @param ***REMOVED***ASTNode***REMOVED*** nodeB The second node
         * @returns ***REMOVED***boolean***REMOVED*** true if the nodes have identical token representations
         */
        function hasSameTokens(nodeA, nodeB) ***REMOVED***
            const tokensA = sourceCode.getTokens(nodeA);
            const tokensB = sourceCode.getTokens(nodeB);

            return tokensA.length === tokensB.length &&
                tokensA.every((token, index) => token.type === tokensB[index].type && token.value === tokensB[index].value);
        ***REMOVED***

        return ***REMOVED***

            BinaryExpression(node) ***REMOVED***
                const operators = new Set(["===", "==", "!==", "!=", ">", "<", ">=", "<="]);

                if (operators.has(node.operator) && hasSameTokens(node.left, node.right)) ***REMOVED***
                    context.report(***REMOVED*** node, message: "Comparing to itself is potentially pointless." ***REMOVED***);
                ***REMOVED***
            ***REMOVED***
        ***REMOVED***;

    ***REMOVED***
***REMOVED***;
