// eslint/lib/rules/no-tabs.js
/**
 * @fileoverview Rule to check for tabs inside a file
 * @author Gyandeep Singh
 */

"use strict";

//------------------------------------------------------------------------------
// Helpers
//------------------------------------------------------------------------------

const tabRegex = /\t+/gu;
const anyNonWhitespaceRegex = /\S/u;

//------------------------------------------------------------------------------
// Public Interface
//------------------------------------------------------------------------------

module.exports = ***REMOVED***
    meta: ***REMOVED***
        type: "layout",

        docs: ***REMOVED***
            description: "disallow all tabs",
            category: "Stylistic Issues",
            recommended: false,
            url: "https://eslint.org/docs/rules/no-tabs"
        ***REMOVED***,
        schema: [***REMOVED***
            type: "object",
            properties: ***REMOVED***
                allowIndentationTabs: ***REMOVED***
                    type: "boolean",
                    default: false
                ***REMOVED***
            ***REMOVED***,
            additionalProperties: false
        ***REMOVED***]
    ***REMOVED***,

    create(context) ***REMOVED***
        const sourceCode = context.getSourceCode();
        const allowIndentationTabs = context.options && context.options[0] && context.options[0].allowIndentationTabs;

        return ***REMOVED***
            Program(node) ***REMOVED***
                sourceCode.getLines().forEach((line, index) => ***REMOVED***
                    let match;

                    while ((match = tabRegex.exec(line)) !== null) ***REMOVED***
                        if (allowIndentationTabs && !anyNonWhitespaceRegex.test(line.slice(0, match.index))) ***REMOVED***
                            continue;
                        ***REMOVED***

                        context.report(***REMOVED***
                            node,
                            loc: ***REMOVED***
                                line: index + 1,
                                column: match.index
                            ***REMOVED***,
                            message: "Unexpected tab character."
                        ***REMOVED***);
                    ***REMOVED***
                ***REMOVED***);
            ***REMOVED***
        ***REMOVED***;
    ***REMOVED***
***REMOVED***;
