// eslint/lib/rules/no-throw-literal.js
/**
 * @fileoverview Rule to restrict what can be thrown as an exception.
 * @author Dieter Oberkofler
 */

"use strict";

const astUtils = require("./utils/ast-utils");

//------------------------------------------------------------------------------
// Rule Definition
//------------------------------------------------------------------------------

module.exports = ***REMOVED***
    meta: ***REMOVED***
        type: "suggestion",

        docs: ***REMOVED***
            description: "disallow throwing literals as exceptions",
            category: "Best Practices",
            recommended: false,
            url: "https://eslint.org/docs/rules/no-throw-literal"
        ***REMOVED***,

        schema: [],

        messages: ***REMOVED***
            object: "Expected an error object to be thrown.",
            undef: "Do not throw undefined."
        ***REMOVED***
    ***REMOVED***,

    create(context) ***REMOVED***

        return ***REMOVED***

            ThrowStatement(node) ***REMOVED***
                if (!astUtils.couldBeError(node.argument)) ***REMOVED***
                    context.report(***REMOVED*** node, messageId: "object" ***REMOVED***);
                ***REMOVED*** else if (node.argument.type === "Identifier") ***REMOVED***
                    if (node.argument.name === "undefined") ***REMOVED***
                        context.report(***REMOVED*** node, messageId: "undef" ***REMOVED***);
                    ***REMOVED***
                ***REMOVED***

            ***REMOVED***

        ***REMOVED***;

    ***REMOVED***
***REMOVED***;
