// eslint/lib/rules/no-template-curly-in-string.js
/**
 * @fileoverview Warn when using template string syntax in regular strings
 * @author Jeroen Engels
 */
"use strict";

//------------------------------------------------------------------------------
// Rule Definition
//------------------------------------------------------------------------------

module.exports = ***REMOVED***
    meta: ***REMOVED***
        type: "problem",

        docs: ***REMOVED***
            description: "disallow template literal placeholder syntax in regular strings",
            category: "Possible Errors",
            recommended: false,
            url: "https://eslint.org/docs/rules/no-template-curly-in-string"
        ***REMOVED***,

        schema: []
    ***REMOVED***,

    create(context) ***REMOVED***
        const regex = /\$\***REMOVED***[^***REMOVED***]+\***REMOVED***/u;

        return ***REMOVED***
            Literal(node) ***REMOVED***
                if (typeof node.value === "string" && regex.test(node.value)) ***REMOVED***
                    context.report(***REMOVED***
                        node,
                        message: "Unexpected template string expression."
                    ***REMOVED***);
                ***REMOVED***
            ***REMOVED***
        ***REMOVED***;

    ***REMOVED***
***REMOVED***;
