// eslint/lib/rules/no-useless-catch.js
/**
 * @fileoverview Reports useless `catch` clauses that just rethrow their error.
 * @author Teddy Katz
 */

"use strict";

//------------------------------------------------------------------------------
// Rule Definition
//------------------------------------------------------------------------------

module.exports = ***REMOVED***
    meta: ***REMOVED***
        type: "suggestion",

        docs: ***REMOVED***
            description: "disallow unnecessary `catch` clauses",
            category: "Best Practices",
            recommended: true,
            url: "https://eslint.org/docs/rules/no-useless-catch"
        ***REMOVED***,

        schema: []
    ***REMOVED***,

    create(context) ***REMOVED***
        return ***REMOVED***
            CatchClause(node) ***REMOVED***
                if (
                    node.param &&
                    node.param.type === "Identifier" &&
                    node.body.body.length &&
                    node.body.body[0].type === "ThrowStatement" &&
                    node.body.body[0].argument.type === "Identifier" &&
                    node.body.body[0].argument.name === node.param.name
                ) ***REMOVED***
                    if (node.parent.finalizer) ***REMOVED***
                        context.report(***REMOVED***
                            node,
                            message: "Unnecessary catch clause."
                        ***REMOVED***);
                    ***REMOVED*** else ***REMOVED***
                        context.report(***REMOVED***
                            node: node.parent,
                            message: "Unnecessary try/catch wrapper."
                        ***REMOVED***);
                    ***REMOVED***
                ***REMOVED***
            ***REMOVED***
        ***REMOVED***;
    ***REMOVED***
***REMOVED***;
