// eslint/lib/rules/no-unexpected-multiline.js
/**
 * @fileoverview Rule to spot scenarios where a newline looks like it is ending a statement, but is not.
 * @author Glen Mailer
 */
"use strict";

//------------------------------------------------------------------------------
// Requirements
//------------------------------------------------------------------------------

const astUtils = require("./utils/ast-utils");

//------------------------------------------------------------------------------
// Rule Definition
//------------------------------------------------------------------------------

module.exports = ***REMOVED***
    meta: ***REMOVED***
        type: "problem",

        docs: ***REMOVED***
            description: "disallow confusing multiline expressions",
            category: "Possible Errors",
            recommended: true,
            url: "https://eslint.org/docs/rules/no-unexpected-multiline"
        ***REMOVED***,

        schema: [],
        messages: ***REMOVED***
            function: "Unexpected newline between function and ( of function call.",
            property: "Unexpected newline between object and [ of property access.",
            taggedTemplate: "Unexpected newline between template tag and template literal.",
            division: "Unexpected newline between numerator and division operator."
        ***REMOVED***
    ***REMOVED***,

    create(context) ***REMOVED***

        const REGEX_FLAG_MATCHER = /^[gimsuy]+$/u;

        const sourceCode = context.getSourceCode();

        /**
         * Check to see if there is a newline between the node and the following open bracket
         * line's expression
         * @param ***REMOVED***ASTNode***REMOVED*** node The node to check.
         * @param ***REMOVED***string***REMOVED*** messageId The error messageId to use.
         * @returns ***REMOVED***void***REMOVED***
         * @private
         */
        function checkForBreakAfter(node, messageId) ***REMOVED***
            const openParen = sourceCode.getTokenAfter(node, astUtils.isNotClosingParenToken);
            const nodeExpressionEnd = sourceCode.getTokenBefore(openParen);

            if (openParen.loc.start.line !== nodeExpressionEnd.loc.end.line) ***REMOVED***
                context.report(***REMOVED*** node, loc: openParen.loc.start, messageId, data: ***REMOVED*** char: openParen.value ***REMOVED*** ***REMOVED***);
            ***REMOVED***
        ***REMOVED***

        //--------------------------------------------------------------------------
        // Public API
        //--------------------------------------------------------------------------

        return ***REMOVED***

            MemberExpression(node) ***REMOVED***
                if (!node.computed) ***REMOVED***
                    return;
                ***REMOVED***
                checkForBreakAfter(node.object, "property");
            ***REMOVED***,

            TaggedTemplateExpression(node) ***REMOVED***
                if (node.tag.loc.end.line === node.quasi.loc.start.line) ***REMOVED***
                    return;
                ***REMOVED***
                context.report(***REMOVED*** node, loc: node.loc.start, messageId: "taggedTemplate" ***REMOVED***);
            ***REMOVED***,

            CallExpression(node) ***REMOVED***
                if (node.arguments.length === 0) ***REMOVED***
                    return;
                ***REMOVED***
                checkForBreakAfter(node.callee, "function");
            ***REMOVED***,

            "BinaryExpression[operator='/'] > BinaryExpression[operator='/'].left"(node) ***REMOVED***
                const secondSlash = sourceCode.getTokenAfter(node, token => token.value === "/");
                const tokenAfterOperator = sourceCode.getTokenAfter(secondSlash);

                if (
                    tokenAfterOperator.type === "Identifier" &&
                    REGEX_FLAG_MATCHER.test(tokenAfterOperator.value) &&
                    secondSlash.range[1] === tokenAfterOperator.range[0]
                ) ***REMOVED***
                    checkForBreakAfter(node.left, "division");
                ***REMOVED***
            ***REMOVED***
        ***REMOVED***;

    ***REMOVED***
***REMOVED***;
