// eslint/lib/rules/no-underscore-dangle.js
/**
 * @fileoverview Rule to flag trailing underscores in variable declarations.
 * @author Matt DuVall <http://www.mattduvall.com>
 */

"use strict";

//------------------------------------------------------------------------------
// Rule Definition
//------------------------------------------------------------------------------

module.exports = ***REMOVED***
    meta: ***REMOVED***
        type: "suggestion",

        docs: ***REMOVED***
            description: "disallow dangling underscores in identifiers",
            category: "Stylistic Issues",
            recommended: false,
            url: "https://eslint.org/docs/rules/no-underscore-dangle"
        ***REMOVED***,

        schema: [
            ***REMOVED***
                type: "object",
                properties: ***REMOVED***
                    allow: ***REMOVED***
                        type: "array",
                        items: ***REMOVED***
                            type: "string"
                        ***REMOVED***
                    ***REMOVED***,
                    allowAfterThis: ***REMOVED***
                        type: "boolean",
                        default: false
                    ***REMOVED***,
                    allowAfterSuper: ***REMOVED***
                        type: "boolean",
                        default: false
                    ***REMOVED***,
                    enforceInMethodNames: ***REMOVED***
                        type: "boolean",
                        default: false
                    ***REMOVED***
                ***REMOVED***,
                additionalProperties: false
            ***REMOVED***
        ]
    ***REMOVED***,

    create(context) ***REMOVED***

        const options = context.options[0] || ***REMOVED******REMOVED***;
        const ALLOWED_VARIABLES = options.allow ? options.allow : [];
        const allowAfterThis = typeof options.allowAfterThis !== "undefined" ? options.allowAfterThis : false;
        const allowAfterSuper = typeof options.allowAfterSuper !== "undefined" ? options.allowAfterSuper : false;
        const enforceInMethodNames = typeof options.enforceInMethodNames !== "undefined" ? options.enforceInMethodNames : false;

        //-------------------------------------------------------------------------
        // Helpers
        //-------------------------------------------------------------------------

        /**
         * Check if identifier is present inside the allowed option
         * @param ***REMOVED***string***REMOVED*** identifier name of the node
         * @returns ***REMOVED***boolean***REMOVED*** true if its is present
         * @private
         */
        function isAllowed(identifier) ***REMOVED***
            return ALLOWED_VARIABLES.some(ident => ident === identifier);
        ***REMOVED***

        /**
         * Check if identifier has a underscore at the end
         * @param ***REMOVED***ASTNode***REMOVED*** identifier node to evaluate
         * @returns ***REMOVED***boolean***REMOVED*** true if its is present
         * @private
         */
        function hasTrailingUnderscore(identifier) ***REMOVED***
            const len = identifier.length;

            return identifier !== "_" && (identifier[0] === "_" || identifier[len - 1] === "_");
        ***REMOVED***

        /**
         * Check if identifier is a special case member expression
         * @param ***REMOVED***ASTNode***REMOVED*** identifier node to evaluate
         * @returns ***REMOVED***boolean***REMOVED*** true if its is a special case
         * @private
         */
        function isSpecialCaseIdentifierForMemberExpression(identifier) ***REMOVED***
            return identifier === "__proto__";
        ***REMOVED***

        /**
         * Check if identifier is a special case variable expression
         * @param ***REMOVED***ASTNode***REMOVED*** identifier node to evaluate
         * @returns ***REMOVED***boolean***REMOVED*** true if its is a special case
         * @private
         */
        function isSpecialCaseIdentifierInVariableExpression(identifier) ***REMOVED***

            // Checks for the underscore library usage here
            return identifier === "_";
        ***REMOVED***

        /**
         * Check if function has a underscore at the end
         * @param ***REMOVED***ASTNode***REMOVED*** node node to evaluate
         * @returns ***REMOVED***void***REMOVED***
         * @private
         */
        function checkForTrailingUnderscoreInFunctionDeclaration(node) ***REMOVED***
            if (node.id) ***REMOVED***
                const identifier = node.id.name;

                if (typeof identifier !== "undefined" && hasTrailingUnderscore(identifier) && !isAllowed(identifier)) ***REMOVED***
                    context.report(***REMOVED***
                        node,
                        message: "Unexpected dangling '_' in '***REMOVED******REMOVED***identifier***REMOVED******REMOVED***'.",
                        data: ***REMOVED***
                            identifier
                        ***REMOVED***
                    ***REMOVED***);
                ***REMOVED***
            ***REMOVED***
        ***REMOVED***

        /**
         * Check if variable expression has a underscore at the end
         * @param ***REMOVED***ASTNode***REMOVED*** node node to evaluate
         * @returns ***REMOVED***void***REMOVED***
         * @private
         */
        function checkForTrailingUnderscoreInVariableExpression(node) ***REMOVED***
            const identifier = node.id.name;

            if (typeof identifier !== "undefined" && hasTrailingUnderscore(identifier) &&
                !isSpecialCaseIdentifierInVariableExpression(identifier) && !isAllowed(identifier)) ***REMOVED***
                context.report(***REMOVED***
                    node,
                    message: "Unexpected dangling '_' in '***REMOVED******REMOVED***identifier***REMOVED******REMOVED***'.",
                    data: ***REMOVED***
                        identifier
                    ***REMOVED***
                ***REMOVED***);
            ***REMOVED***
        ***REMOVED***

        /**
         * Check if member expression has a underscore at the end
         * @param ***REMOVED***ASTNode***REMOVED*** node node to evaluate
         * @returns ***REMOVED***void***REMOVED***
         * @private
         */
        function checkForTrailingUnderscoreInMemberExpression(node) ***REMOVED***
            const identifier = node.property.name,
                isMemberOfThis = node.object.type === "ThisExpression",
                isMemberOfSuper = node.object.type === "Super";

            if (typeof identifier !== "undefined" && hasTrailingUnderscore(identifier) &&
                !(isMemberOfThis && allowAfterThis) &&
                !(isMemberOfSuper && allowAfterSuper) &&
                !isSpecialCaseIdentifierForMemberExpression(identifier) && !isAllowed(identifier)) ***REMOVED***
                context.report(***REMOVED***
                    node,
                    message: "Unexpected dangling '_' in '***REMOVED******REMOVED***identifier***REMOVED******REMOVED***'.",
                    data: ***REMOVED***
                        identifier
                    ***REMOVED***
                ***REMOVED***);
            ***REMOVED***
        ***REMOVED***

        /**
         * Check if method declaration or method property has a underscore at the end
         * @param ***REMOVED***ASTNode***REMOVED*** node node to evaluate
         * @returns ***REMOVED***void***REMOVED***
         * @private
         */
        function checkForTrailingUnderscoreInMethod(node) ***REMOVED***
            const identifier = node.key.name;
            const isMethod = node.type === "MethodDefinition" || node.type === "Property" && node.method;

            if (typeof identifier !== "undefined" && enforceInMethodNames && isMethod && hasTrailingUnderscore(identifier)) ***REMOVED***
                context.report(***REMOVED***
                    node,
                    message: "Unexpected dangling '_' in '***REMOVED******REMOVED***identifier***REMOVED******REMOVED***'.",
                    data: ***REMOVED***
                        identifier
                    ***REMOVED***
                ***REMOVED***);
            ***REMOVED***
        ***REMOVED***

        //--------------------------------------------------------------------------
        // Public API
        //--------------------------------------------------------------------------

        return ***REMOVED***
            FunctionDeclaration: checkForTrailingUnderscoreInFunctionDeclaration,
            VariableDeclarator: checkForTrailingUnderscoreInVariableExpression,
            MemberExpression: checkForTrailingUnderscoreInMemberExpression,
            MethodDefinition: checkForTrailingUnderscoreInMethod,
            Property: checkForTrailingUnderscoreInMethod
        ***REMOVED***;

    ***REMOVED***
***REMOVED***;
