// eslint/lib/rules/no-unused-labels.js
/**
 * @fileoverview Rule to disallow unused labels.
 * @author Toru Nagashima
 */

"use strict";

//------------------------------------------------------------------------------
// Rule Definition
//------------------------------------------------------------------------------

module.exports = ***REMOVED***
    meta: ***REMOVED***
        type: "suggestion",

        docs: ***REMOVED***
            description: "disallow unused labels",
            category: "Best Practices",
            recommended: true,
            url: "https://eslint.org/docs/rules/no-unused-labels"
        ***REMOVED***,

        schema: [],

        fixable: "code",

        messages: ***REMOVED***
            unused: "'***REMOVED******REMOVED***name***REMOVED******REMOVED***:' is defined but never used."
        ***REMOVED***
    ***REMOVED***,

    create(context) ***REMOVED***
        const sourceCode = context.getSourceCode();
        let scopeInfo = null;

        /**
         * Adds a scope info to the stack.
         *
         * @param ***REMOVED***ASTNode***REMOVED*** node - A node to add. This is a LabeledStatement.
         * @returns ***REMOVED***void***REMOVED***
         */
        function enterLabeledScope(node) ***REMOVED***
            scopeInfo = ***REMOVED***
                label: node.label.name,
                used: false,
                upper: scopeInfo
            ***REMOVED***;
        ***REMOVED***

        /**
         * Removes the top of the stack.
         * At the same time, this reports the label if it's never used.
         *
         * @param ***REMOVED***ASTNode***REMOVED*** node - A node to report. This is a LabeledStatement.
         * @returns ***REMOVED***void***REMOVED***
         */
        function exitLabeledScope(node) ***REMOVED***
            if (!scopeInfo.used) ***REMOVED***
                context.report(***REMOVED***
                    node: node.label,
                    messageId: "unused",
                    data: node.label,
                    fix(fixer) ***REMOVED***

                        /*
                         * Only perform a fix if there are no comments between the label and the body. This will be the case
                         * when there is exactly one token/comment (the ":") between the label and the body.
                         */
                        if (sourceCode.getTokenAfter(node.label, ***REMOVED*** includeComments: true ***REMOVED***) ===
                                sourceCode.getTokenBefore(node.body, ***REMOVED*** includeComments: true ***REMOVED***)) ***REMOVED***
                            return fixer.removeRange([node.range[0], node.body.range[0]]);
                        ***REMOVED***

                        return null;
                    ***REMOVED***
                ***REMOVED***);
            ***REMOVED***

            scopeInfo = scopeInfo.upper;
        ***REMOVED***

        /**
         * Marks the label of a given node as used.
         *
         * @param ***REMOVED***ASTNode***REMOVED*** node - A node to mark. This is a BreakStatement or
         *      ContinueStatement.
         * @returns ***REMOVED***void***REMOVED***
         */
        function markAsUsed(node) ***REMOVED***
            if (!node.label) ***REMOVED***
                return;
            ***REMOVED***

            const label = node.label.name;
            let info = scopeInfo;

            while (info) ***REMOVED***
                if (info.label === label) ***REMOVED***
                    info.used = true;
                    break;
                ***REMOVED***
                info = info.upper;
            ***REMOVED***
        ***REMOVED***

        return ***REMOVED***
            LabeledStatement: enterLabeledScope,
            "LabeledStatement:exit": exitLabeledScope,
            BreakStatement: markAsUsed,
            ContinueStatement: markAsUsed
        ***REMOVED***;
    ***REMOVED***
***REMOVED***;
