// eslint/lib/rules/no-unsafe-finally.js
/**
 * @fileoverview Rule to flag unsafe statements in finally block
 * @author Onur Temizkan
 */

"use strict";

//------------------------------------------------------------------------------
// Helpers
//------------------------------------------------------------------------------

const SENTINEL_NODE_TYPE_RETURN_THROW = /^(?:Program|(?:Function|Class)(?:Declaration|Expression)|ArrowFunctionExpression)$/u;
const SENTINEL_NODE_TYPE_BREAK = /^(?:Program|(?:Function|Class)(?:Declaration|Expression)|ArrowFunctionExpression|DoWhileStatement|WhileStatement|ForOfStatement|ForInStatement|ForStatement|SwitchStatement)$/u;
const SENTINEL_NODE_TYPE_CONTINUE = /^(?:Program|(?:Function|Class)(?:Declaration|Expression)|ArrowFunctionExpression|DoWhileStatement|WhileStatement|ForOfStatement|ForInStatement|ForStatement)$/u;


//------------------------------------------------------------------------------
// Rule Definition
//------------------------------------------------------------------------------

module.exports = ***REMOVED***
    meta: ***REMOVED***
        type: "problem",

        docs: ***REMOVED***
            description: "disallow control flow statements in `finally` blocks",
            category: "Possible Errors",
            recommended: true,
            url: "https://eslint.org/docs/rules/no-unsafe-finally"
        ***REMOVED***,

        schema: []
    ***REMOVED***,
    create(context) ***REMOVED***

        /**
         * Checks if the node is the finalizer of a TryStatement
         *
         * @param ***REMOVED***ASTNode***REMOVED*** node - node to check.
         * @returns ***REMOVED***boolean***REMOVED*** - true if the node is the finalizer of a TryStatement
         */
        function isFinallyBlock(node) ***REMOVED***
            return node.parent.type === "TryStatement" && node.parent.finalizer === node;
        ***REMOVED***

        /**
         * Climbs up the tree if the node is not a sentinel node
         *
         * @param ***REMOVED***ASTNode***REMOVED*** node - node to check.
         * @param ***REMOVED***string***REMOVED*** label - label of the break or continue statement
         * @returns ***REMOVED***boolean***REMOVED*** - return whether the node is a finally block or a sentinel node
         */
        function isInFinallyBlock(node, label) ***REMOVED***
            let labelInside = false;
            let sentinelNodeType;

            if (node.type === "BreakStatement" && !node.label) ***REMOVED***
                sentinelNodeType = SENTINEL_NODE_TYPE_BREAK;
            ***REMOVED*** else if (node.type === "ContinueStatement") ***REMOVED***
                sentinelNodeType = SENTINEL_NODE_TYPE_CONTINUE;
            ***REMOVED*** else ***REMOVED***
                sentinelNodeType = SENTINEL_NODE_TYPE_RETURN_THROW;
            ***REMOVED***

            for (
                let currentNode = node;
                currentNode && !sentinelNodeType.test(currentNode.type);
                currentNode = currentNode.parent
            ) ***REMOVED***
                if (currentNode.parent.label && label && (currentNode.parent.label.name === label.name)) ***REMOVED***
                    labelInside = true;
                ***REMOVED***
                if (isFinallyBlock(currentNode)) ***REMOVED***
                    if (label && labelInside) ***REMOVED***
                        return false;
                    ***REMOVED***
                    return true;
                ***REMOVED***
            ***REMOVED***
            return false;
        ***REMOVED***

        /**
         * Checks whether the possibly-unsafe statement is inside a finally block.
         *
         * @param ***REMOVED***ASTNode***REMOVED*** node - node to check.
         * @returns ***REMOVED***void***REMOVED***
         */
        function check(node) ***REMOVED***
            if (isInFinallyBlock(node, node.label)) ***REMOVED***
                context.report(***REMOVED***
                    message: "Unsafe usage of ***REMOVED******REMOVED***nodeType***REMOVED******REMOVED***.",
                    data: ***REMOVED***
                        nodeType: node.type
                    ***REMOVED***,
                    node,
                    line: node.loc.line,
                    column: node.loc.column
                ***REMOVED***);
            ***REMOVED***
        ***REMOVED***

        return ***REMOVED***
            ReturnStatement: check,
            ThrowStatement: check,
            BreakStatement: check,
            ContinueStatement: check
        ***REMOVED***;
    ***REMOVED***
***REMOVED***;
