// eslint/lib/rules/no-useless-concat.js
/**
 * @fileoverview disallow unncessary concatenation of template strings
 * @author Henry Zhu
 */
"use strict";

//------------------------------------------------------------------------------
// Requirements
//------------------------------------------------------------------------------

const astUtils = require("./utils/ast-utils");

//------------------------------------------------------------------------------
// Helpers
//------------------------------------------------------------------------------

/**
 * Checks whether or not a given node is a concatenation.
 * @param ***REMOVED***ASTNode***REMOVED*** node - A node to check.
 * @returns ***REMOVED***boolean***REMOVED*** `true` if the node is a concatenation.
 */
function isConcatenation(node) ***REMOVED***
    return node.type === "BinaryExpression" && node.operator === "+";
***REMOVED***

/**
 * Checks if the given token is a `+` token or not.
 * @param ***REMOVED***Token***REMOVED*** token - The token to check.
 * @returns ***REMOVED***boolean***REMOVED*** `true` if the token is a `+` token.
 */
function isConcatOperatorToken(token) ***REMOVED***
    return token.value === "+" && token.type === "Punctuator";
***REMOVED***

/**
 * Get's the right most node on the left side of a BinaryExpression with + operator.
 * @param ***REMOVED***ASTNode***REMOVED*** node - A BinaryExpression node to check.
 * @returns ***REMOVED***ASTNode***REMOVED*** node
 */
function getLeft(node) ***REMOVED***
    let left = node.left;

    while (isConcatenation(left)) ***REMOVED***
        left = left.right;
    ***REMOVED***
    return left;
***REMOVED***

/**
 * Get's the left most node on the right side of a BinaryExpression with + operator.
 * @param ***REMOVED***ASTNode***REMOVED*** node - A BinaryExpression node to check.
 * @returns ***REMOVED***ASTNode***REMOVED*** node
 */
function getRight(node) ***REMOVED***
    let right = node.right;

    while (isConcatenation(right)) ***REMOVED***
        right = right.left;
    ***REMOVED***
    return right;
***REMOVED***

//------------------------------------------------------------------------------
// Rule Definition
//------------------------------------------------------------------------------

module.exports = ***REMOVED***
    meta: ***REMOVED***
        type: "suggestion",

        docs: ***REMOVED***
            description: "disallow unnecessary concatenation of literals or template literals",
            category: "Best Practices",
            recommended: false,
            url: "https://eslint.org/docs/rules/no-useless-concat"
        ***REMOVED***,

        schema: []
    ***REMOVED***,

    create(context) ***REMOVED***
        const sourceCode = context.getSourceCode();

        return ***REMOVED***
            BinaryExpression(node) ***REMOVED***

                // check if not concatenation
                if (node.operator !== "+") ***REMOVED***
                    return;
                ***REMOVED***

                // account for the `foo + "a" + "b"` case
                const left = getLeft(node);
                const right = getRight(node);

                if (astUtils.isStringLiteral(left) &&
                    astUtils.isStringLiteral(right) &&
                    astUtils.isTokenOnSameLine(left, right)
                ) ***REMOVED***
                    const operatorToken = sourceCode.getFirstTokenBetween(left, right, isConcatOperatorToken);

                    context.report(***REMOVED***
                        node,
                        loc: operatorToken.loc.start,
                        message: "Unexpected string concatenation of literals."
                    ***REMOVED***);
                ***REMOVED***
            ***REMOVED***
        ***REMOVED***;
    ***REMOVED***
***REMOVED***;
