// eslint/lib/rules/require-yield.js
/**
 * @fileoverview Rule to flag the generator functions that does not have yield.
 * @author Toru Nagashima
 */

"use strict";

//------------------------------------------------------------------------------
// Rule Definition
//------------------------------------------------------------------------------

module.exports = ***REMOVED***
    meta: ***REMOVED***
        type: "suggestion",

        docs: ***REMOVED***
            description: "require generator functions to contain `yield`",
            category: "ECMAScript 6",
            recommended: true,
            url: "https://eslint.org/docs/rules/require-yield"
        ***REMOVED***,

        schema: []
    ***REMOVED***,

    create(context) ***REMOVED***
        const stack = [];

        /**
         * If the node is a generator function, start counting `yield` keywords.
         * @param ***REMOVED***Node***REMOVED*** node - A function node to check.
         * @returns ***REMOVED***void***REMOVED***
         */
        function beginChecking(node) ***REMOVED***
            if (node.generator) ***REMOVED***
                stack.push(0);
            ***REMOVED***
        ***REMOVED***

        /**
         * If the node is a generator function, end counting `yield` keywords, then
         * reports result.
         * @param ***REMOVED***Node***REMOVED*** node - A function node to check.
         * @returns ***REMOVED***void***REMOVED***
         */
        function endChecking(node) ***REMOVED***
            if (!node.generator) ***REMOVED***
                return;
            ***REMOVED***

            const countYield = stack.pop();

            if (countYield === 0 && node.body.body.length > 0) ***REMOVED***
                context.report(***REMOVED*** node, message: "This generator function does not have 'yield'." ***REMOVED***);
            ***REMOVED***
        ***REMOVED***

        return ***REMOVED***
            FunctionDeclaration: beginChecking,
            "FunctionDeclaration:exit": endChecking,
            FunctionExpression: beginChecking,
            "FunctionExpression:exit": endChecking,

            // Increases the count of `yield` keyword.
            YieldExpression() ***REMOVED***

                /* istanbul ignore else */
                if (stack.length > 0) ***REMOVED***
                    stack[stack.length - 1] += 1;
                ***REMOVED***
            ***REMOVED***
        ***REMOVED***;
    ***REMOVED***
***REMOVED***;
