// eslint/lib/rules/space-before-function-paren.js
/**
 * @fileoverview Rule to validate spacing before function paren.
 * @author Mathias Schreck <https://github.com/lo1tuma>
 */
"use strict";

//------------------------------------------------------------------------------
// Requirements
//------------------------------------------------------------------------------

const astUtils = require("./utils/ast-utils");

//------------------------------------------------------------------------------
// Rule Definition
//------------------------------------------------------------------------------

module.exports = ***REMOVED***
    meta: ***REMOVED***
        type: "layout",

        docs: ***REMOVED***
            description: "enforce consistent spacing before `function` definition opening parenthesis",
            category: "Stylistic Issues",
            recommended: false,
            url: "https://eslint.org/docs/rules/space-before-function-paren"
        ***REMOVED***,

        fixable: "whitespace",

        schema: [
            ***REMOVED***
                oneOf: [
                    ***REMOVED***
                        enum: ["always", "never"]
                    ***REMOVED***,
                    ***REMOVED***
                        type: "object",
                        properties: ***REMOVED***
                            anonymous: ***REMOVED***
                                enum: ["always", "never", "ignore"]
                            ***REMOVED***,
                            named: ***REMOVED***
                                enum: ["always", "never", "ignore"]
                            ***REMOVED***,
                            asyncArrow: ***REMOVED***
                                enum: ["always", "never", "ignore"]
                            ***REMOVED***
                        ***REMOVED***,
                        additionalProperties: false
                    ***REMOVED***
                ]
            ***REMOVED***
        ]
    ***REMOVED***,

    create(context) ***REMOVED***
        const sourceCode = context.getSourceCode();
        const baseConfig = typeof context.options[0] === "string" ? context.options[0] : "always";
        const overrideConfig = typeof context.options[0] === "object" ? context.options[0] : ***REMOVED******REMOVED***;

        /**
         * Determines whether a function has a name.
         * @param ***REMOVED***ASTNode***REMOVED*** node The function node.
         * @returns ***REMOVED***boolean***REMOVED*** Whether the function has a name.
         */
        function isNamedFunction(node) ***REMOVED***
            if (node.id) ***REMOVED***
                return true;
            ***REMOVED***

            const parent = node.parent;

            return parent.type === "MethodDefinition" ||
                (parent.type === "Property" &&
                    (
                        parent.kind === "get" ||
                        parent.kind === "set" ||
                        parent.method
                    )
                );
        ***REMOVED***

        /**
         * Gets the config for a given function
         * @param ***REMOVED***ASTNode***REMOVED*** node The function node
         * @returns ***REMOVED***string***REMOVED*** "always", "never", or "ignore"
         */
        function getConfigForFunction(node) ***REMOVED***
            if (node.type === "ArrowFunctionExpression") ***REMOVED***

                // Always ignore non-async functions and arrow functions without parens, e.g. async foo => bar
                if (node.async && astUtils.isOpeningParenToken(sourceCode.getFirstToken(node, ***REMOVED*** skip: 1 ***REMOVED***))) ***REMOVED***
                    return overrideConfig.asyncArrow || baseConfig;
                ***REMOVED***
            ***REMOVED*** else if (isNamedFunction(node)) ***REMOVED***
                return overrideConfig.named || baseConfig;

            // `generator-star-spacing` should warn anonymous generators. E.g. `function* () ***REMOVED******REMOVED***`
            ***REMOVED*** else if (!node.generator) ***REMOVED***
                return overrideConfig.anonymous || baseConfig;
            ***REMOVED***

            return "ignore";
        ***REMOVED***

        /**
         * Checks the parens of a function node
         * @param ***REMOVED***ASTNode***REMOVED*** node A function node
         * @returns ***REMOVED***void***REMOVED***
         */
        function checkFunction(node) ***REMOVED***
            const functionConfig = getConfigForFunction(node);

            if (functionConfig === "ignore") ***REMOVED***
                return;
            ***REMOVED***

            const rightToken = sourceCode.getFirstToken(node, astUtils.isOpeningParenToken);
            const leftToken = sourceCode.getTokenBefore(rightToken);
            const hasSpacing = sourceCode.isSpaceBetweenTokens(leftToken, rightToken);

            if (hasSpacing && functionConfig === "never") ***REMOVED***
                context.report(***REMOVED***
                    node,
                    loc: leftToken.loc.end,
                    message: "Unexpected space before function parentheses.",
                    fix(fixer) ***REMOVED***
                        const comments = sourceCode.getCommentsBefore(rightToken);

                        // Don't fix anything if there's a single line comment between the left and the right token
                        if (comments.some(comment => comment.type === "Line")) ***REMOVED***
                            return null;
                        ***REMOVED***
                        return fixer.replaceTextRange(
                            [leftToken.range[1], rightToken.range[0]],
                            comments.reduce((text, comment) => text + sourceCode.getText(comment), "")
                        );
                    ***REMOVED***
                ***REMOVED***);
            ***REMOVED*** else if (!hasSpacing && functionConfig === "always") ***REMOVED***
                context.report(***REMOVED***
                    node,
                    loc: leftToken.loc.end,
                    message: "Missing space before function parentheses.",
                    fix: fixer => fixer.insertTextAfter(leftToken, " ")
                ***REMOVED***);
            ***REMOVED***
        ***REMOVED***

        return ***REMOVED***
            ArrowFunctionExpression: checkFunction,
            FunctionDeclaration: checkFunction,
            FunctionExpression: checkFunction
        ***REMOVED***;
    ***REMOVED***
***REMOVED***;
