// eslint/lib/rules/space-unary-ops.js
/**
 * @fileoverview This rule shoud require or disallow spaces before or after unary operations.
 * @author Marcin Kumorek
 */
"use strict";

//------------------------------------------------------------------------------
// Requirements
//------------------------------------------------------------------------------

const astUtils = require("./utils/ast-utils");

//------------------------------------------------------------------------------
// Rule Definition
//------------------------------------------------------------------------------

module.exports = ***REMOVED***
    meta: ***REMOVED***
        type: "layout",

        docs: ***REMOVED***
            description: "enforce consistent spacing before or after unary operators",
            category: "Stylistic Issues",
            recommended: false,
            url: "https://eslint.org/docs/rules/space-unary-ops"
        ***REMOVED***,

        fixable: "whitespace",

        schema: [
            ***REMOVED***
                type: "object",
                properties: ***REMOVED***
                    words: ***REMOVED***
                        type: "boolean",
                        default: true
                    ***REMOVED***,
                    nonwords: ***REMOVED***
                        type: "boolean",
                        default: false
                    ***REMOVED***,
                    overrides: ***REMOVED***
                        type: "object",
                        additionalProperties: ***REMOVED***
                            type: "boolean"
                        ***REMOVED***
                    ***REMOVED***
                ***REMOVED***,
                additionalProperties: false
            ***REMOVED***
        ],
        messages: ***REMOVED***
            unexpectedBefore: "Unexpected space before unary operator '***REMOVED******REMOVED***operator***REMOVED******REMOVED***'.",
            unexpectedAfter: "Unexpected space after unary operator '***REMOVED******REMOVED***operator***REMOVED******REMOVED***'.",
            unexpectedAfterWord: "Unexpected space after unary word operator '***REMOVED******REMOVED***word***REMOVED******REMOVED***'.",
            wordOperator: "Unary word operator '***REMOVED******REMOVED***word***REMOVED******REMOVED***' must be followed by whitespace.",
            operator: "Unary operator '***REMOVED******REMOVED***operator***REMOVED******REMOVED***' must be followed by whitespace.",
            beforeUnaryExpressions: "Space is required before unary expressions '***REMOVED******REMOVED***token***REMOVED******REMOVED***'."
        ***REMOVED***
    ***REMOVED***,

    create(context) ***REMOVED***
        const options = context.options[0] || ***REMOVED*** words: true, nonwords: false ***REMOVED***;

        const sourceCode = context.getSourceCode();

        //--------------------------------------------------------------------------
        // Helpers
        //--------------------------------------------------------------------------

        /**
         * Check if the node is the first "!" in a "!!" convert to Boolean expression
         * @param ***REMOVED***ASTnode***REMOVED*** node AST node
         * @returns ***REMOVED***boolean***REMOVED*** Whether or not the node is first "!" in "!!"
         */
        function isFirstBangInBangBangExpression(node) ***REMOVED***
            return node && node.type === "UnaryExpression" && node.argument.operator === "!" &&
                node.argument && node.argument.type === "UnaryExpression" && node.argument.operator === "!";
        ***REMOVED***

        /**
         * Checks if an override exists for a given operator.
         * @param ***REMOVED***string***REMOVED*** operator Operator
         * @returns ***REMOVED***boolean***REMOVED*** Whether or not an override has been provided for the operator
         */
        function overrideExistsForOperator(operator) ***REMOVED***
            return options.overrides && Object.prototype.hasOwnProperty.call(options.overrides, operator);
        ***REMOVED***

        /**
         * Gets the value that the override was set to for this operator
         * @param ***REMOVED***string***REMOVED*** operator Operator
         * @returns ***REMOVED***boolean***REMOVED*** Whether or not an override enforces a space with this operator
         */
        function overrideEnforcesSpaces(operator) ***REMOVED***
            return options.overrides[operator];
        ***REMOVED***

        /**
         * Verify Unary Word Operator has spaces after the word operator
         * @param ***REMOVED***ASTnode***REMOVED*** node AST node
         * @param ***REMOVED***Object***REMOVED*** firstToken first token from the AST node
         * @param ***REMOVED***Object***REMOVED*** secondToken second token from the AST node
         * @param ***REMOVED***string***REMOVED*** word The word to be used for reporting
         * @returns ***REMOVED***void***REMOVED***
         */
        function verifyWordHasSpaces(node, firstToken, secondToken, word) ***REMOVED***
            if (secondToken.range[0] === firstToken.range[1]) ***REMOVED***
                context.report(***REMOVED***
                    node,
                    messageId: "wordOperator",
                    data: ***REMOVED***
                        word
                    ***REMOVED***,
                    fix(fixer) ***REMOVED***
                        return fixer.insertTextAfter(firstToken, " ");
                    ***REMOVED***
                ***REMOVED***);
            ***REMOVED***
        ***REMOVED***

        /**
         * Verify Unary Word Operator doesn't have spaces after the word operator
         * @param ***REMOVED***ASTnode***REMOVED*** node AST node
         * @param ***REMOVED***Object***REMOVED*** firstToken first token from the AST node
         * @param ***REMOVED***Object***REMOVED*** secondToken second token from the AST node
         * @param ***REMOVED***string***REMOVED*** word The word to be used for reporting
         * @returns ***REMOVED***void***REMOVED***
         */
        function verifyWordDoesntHaveSpaces(node, firstToken, secondToken, word) ***REMOVED***
            if (astUtils.canTokensBeAdjacent(firstToken, secondToken)) ***REMOVED***
                if (secondToken.range[0] > firstToken.range[1]) ***REMOVED***
                    context.report(***REMOVED***
                        node,
                        messageId: "unexpectedAfterWord",
                        data: ***REMOVED***
                            word
                        ***REMOVED***,
                        fix(fixer) ***REMOVED***
                            return fixer.removeRange([firstToken.range[1], secondToken.range[0]]);
                        ***REMOVED***
                    ***REMOVED***);
                ***REMOVED***
            ***REMOVED***
        ***REMOVED***

        /**
         * Check Unary Word Operators for spaces after the word operator
         * @param ***REMOVED***ASTnode***REMOVED*** node AST node
         * @param ***REMOVED***Object***REMOVED*** firstToken first token from the AST node
         * @param ***REMOVED***Object***REMOVED*** secondToken second token from the AST node
         * @param ***REMOVED***string***REMOVED*** word The word to be used for reporting
         * @returns ***REMOVED***void***REMOVED***
         */
        function checkUnaryWordOperatorForSpaces(node, firstToken, secondToken, word) ***REMOVED***
            if (overrideExistsForOperator(word)) ***REMOVED***
                if (overrideEnforcesSpaces(word)) ***REMOVED***
                    verifyWordHasSpaces(node, firstToken, secondToken, word);
                ***REMOVED*** else ***REMOVED***
                    verifyWordDoesntHaveSpaces(node, firstToken, secondToken, word);
                ***REMOVED***
            ***REMOVED*** else if (options.words) ***REMOVED***
                verifyWordHasSpaces(node, firstToken, secondToken, word);
            ***REMOVED*** else ***REMOVED***
                verifyWordDoesntHaveSpaces(node, firstToken, secondToken, word);
            ***REMOVED***
        ***REMOVED***

        /**
         * Verifies YieldExpressions satisfy spacing requirements
         * @param ***REMOVED***ASTnode***REMOVED*** node AST node
         * @returns ***REMOVED***void***REMOVED***
         */
        function checkForSpacesAfterYield(node) ***REMOVED***
            const tokens = sourceCode.getFirstTokens(node, 3),
                word = "yield";

            if (!node.argument || node.delegate) ***REMOVED***
                return;
            ***REMOVED***

            checkUnaryWordOperatorForSpaces(node, tokens[0], tokens[1], word);
        ***REMOVED***

        /**
         * Verifies AwaitExpressions satisfy spacing requirements
         * @param ***REMOVED***ASTNode***REMOVED*** node AwaitExpression AST node
         * @returns ***REMOVED***void***REMOVED***
         */
        function checkForSpacesAfterAwait(node) ***REMOVED***
            const tokens = sourceCode.getFirstTokens(node, 3);

            checkUnaryWordOperatorForSpaces(node, tokens[0], tokens[1], "await");
        ***REMOVED***

        /**
         * Verifies UnaryExpression, UpdateExpression and NewExpression have spaces before or after the operator
         * @param ***REMOVED***ASTnode***REMOVED*** node AST node
         * @param ***REMOVED***Object***REMOVED*** firstToken First token in the expression
         * @param ***REMOVED***Object***REMOVED*** secondToken Second token in the expression
         * @returns ***REMOVED***void***REMOVED***
         */
        function verifyNonWordsHaveSpaces(node, firstToken, secondToken) ***REMOVED***
            if (node.prefix) ***REMOVED***
                if (isFirstBangInBangBangExpression(node)) ***REMOVED***
                    return;
                ***REMOVED***
                if (firstToken.range[1] === secondToken.range[0]) ***REMOVED***
                    context.report(***REMOVED***
                        node,
                        messageId: "operator",
                        data: ***REMOVED***
                            operator: firstToken.value
                        ***REMOVED***,
                        fix(fixer) ***REMOVED***
                            return fixer.insertTextAfter(firstToken, " ");
                        ***REMOVED***
                    ***REMOVED***);
                ***REMOVED***
            ***REMOVED*** else ***REMOVED***
                if (firstToken.range[1] === secondToken.range[0]) ***REMOVED***
                    context.report(***REMOVED***
                        node,
                        messageId: "beforeUnaryExpressions",
                        data: ***REMOVED***
                            token: secondToken.value
                        ***REMOVED***,
                        fix(fixer) ***REMOVED***
                            return fixer.insertTextBefore(secondToken, " ");
                        ***REMOVED***
                    ***REMOVED***);
                ***REMOVED***
            ***REMOVED***
        ***REMOVED***

        /**
         * Verifies UnaryExpression, UpdateExpression and NewExpression don't have spaces before or after the operator
         * @param ***REMOVED***ASTnode***REMOVED*** node AST node
         * @param ***REMOVED***Object***REMOVED*** firstToken First token in the expression
         * @param ***REMOVED***Object***REMOVED*** secondToken Second token in the expression
         * @returns ***REMOVED***void***REMOVED***
         */
        function verifyNonWordsDontHaveSpaces(node, firstToken, secondToken) ***REMOVED***
            if (node.prefix) ***REMOVED***
                if (secondToken.range[0] > firstToken.range[1]) ***REMOVED***
                    context.report(***REMOVED***
                        node,
                        messageId: "unexpectedAfter",
                        data: ***REMOVED***
                            operator: firstToken.value
                        ***REMOVED***,
                        fix(fixer) ***REMOVED***
                            if (astUtils.canTokensBeAdjacent(firstToken, secondToken)) ***REMOVED***
                                return fixer.removeRange([firstToken.range[1], secondToken.range[0]]);
                            ***REMOVED***
                            return null;
                        ***REMOVED***
                    ***REMOVED***);
                ***REMOVED***
            ***REMOVED*** else ***REMOVED***
                if (secondToken.range[0] > firstToken.range[1]) ***REMOVED***
                    context.report(***REMOVED***
                        node,
                        messageId: "unexpectedBefore",
                        data: ***REMOVED***
                            operator: secondToken.value
                        ***REMOVED***,
                        fix(fixer) ***REMOVED***
                            return fixer.removeRange([firstToken.range[1], secondToken.range[0]]);
                        ***REMOVED***
                    ***REMOVED***);
                ***REMOVED***
            ***REMOVED***
        ***REMOVED***

        /**
         * Verifies UnaryExpression, UpdateExpression and NewExpression satisfy spacing requirements
         * @param ***REMOVED***ASTnode***REMOVED*** node AST node
         * @returns ***REMOVED***void***REMOVED***
         */
        function checkForSpaces(node) ***REMOVED***
            const tokens = node.type === "UpdateExpression" && !node.prefix
                ? sourceCode.getLastTokens(node, 2)
                : sourceCode.getFirstTokens(node, 2);
            const firstToken = tokens[0];
            const secondToken = tokens[1];

            if ((node.type === "NewExpression" || node.prefix) && firstToken.type === "Keyword") ***REMOVED***
                checkUnaryWordOperatorForSpaces(node, firstToken, secondToken, firstToken.value);
                return;
            ***REMOVED***

            const operator = node.prefix ? tokens[0].value : tokens[1].value;

            if (overrideExistsForOperator(operator)) ***REMOVED***
                if (overrideEnforcesSpaces(operator)) ***REMOVED***
                    verifyNonWordsHaveSpaces(node, firstToken, secondToken);
                ***REMOVED*** else ***REMOVED***
                    verifyNonWordsDontHaveSpaces(node, firstToken, secondToken);
                ***REMOVED***
            ***REMOVED*** else if (options.nonwords) ***REMOVED***
                verifyNonWordsHaveSpaces(node, firstToken, secondToken);
            ***REMOVED*** else ***REMOVED***
                verifyNonWordsDontHaveSpaces(node, firstToken, secondToken);
            ***REMOVED***
        ***REMOVED***

        //--------------------------------------------------------------------------
        // Public
        //--------------------------------------------------------------------------

        return ***REMOVED***
            UnaryExpression: checkForSpaces,
            UpdateExpression: checkForSpaces,
            NewExpression: checkForSpaces,
            YieldExpression: checkForSpacesAfterYield,
            AwaitExpression: checkForSpacesAfterAwait
        ***REMOVED***;

    ***REMOVED***
***REMOVED***;
