// eslint/lib/rules/template-curly-spacing.js
/**
 * @fileoverview Rule to enforce spacing around embedded expressions of template strings
 * @author Toru Nagashima
 */

"use strict";

//------------------------------------------------------------------------------
// Requirements
//------------------------------------------------------------------------------

const astUtils = require("./utils/ast-utils");

//------------------------------------------------------------------------------
// Helpers
//------------------------------------------------------------------------------

const OPEN_PAREN = /\$\***REMOVED***$/u;
const CLOSE_PAREN = /^\***REMOVED***/u;

//------------------------------------------------------------------------------
// Rule Definition
//------------------------------------------------------------------------------

module.exports = ***REMOVED***
    meta: ***REMOVED***
        type: "layout",

        docs: ***REMOVED***
            description: "require or disallow spacing around embedded expressions of template strings",
            category: "ECMAScript 6",
            recommended: false,
            url: "https://eslint.org/docs/rules/template-curly-spacing"
        ***REMOVED***,

        fixable: "whitespace",

        schema: [
            ***REMOVED*** enum: ["always", "never"] ***REMOVED***
        ],
        messages: ***REMOVED***
            expectedBefore: "Expected space(s) before '***REMOVED***'.",
            expectedAfter: "Expected space(s) after '$***REMOVED***'.",
            unexpectedBefore: "Unexpected space(s) before '***REMOVED***'.",
            unexpectedAfter: "Unexpected space(s) after '$***REMOVED***'."
        ***REMOVED***
    ***REMOVED***,

    create(context) ***REMOVED***
        const sourceCode = context.getSourceCode();
        const always = context.options[0] === "always";
        const prefix = always ? "expected" : "unexpected";

        /**
         * Checks spacing before `***REMOVED***` of a given token.
         * @param ***REMOVED***Token***REMOVED*** token - A token to check. This is a Template token.
         * @returns ***REMOVED***void***REMOVED***
         */
        function checkSpacingBefore(token) ***REMOVED***
            const prevToken = sourceCode.getTokenBefore(token);

            if (prevToken &&
                CLOSE_PAREN.test(token.value) &&
                astUtils.isTokenOnSameLine(prevToken, token) &&
                sourceCode.isSpaceBetweenTokens(prevToken, token) !== always
            ) ***REMOVED***
                context.report(***REMOVED***
                    loc: token.loc.start,
                    messageId: `$***REMOVED***prefix***REMOVED***Before`,
                    fix(fixer) ***REMOVED***
                        if (always) ***REMOVED***
                            return fixer.insertTextBefore(token, " ");
                        ***REMOVED***
                        return fixer.removeRange([
                            prevToken.range[1],
                            token.range[0]
                        ]);
                    ***REMOVED***
                ***REMOVED***);
            ***REMOVED***
        ***REMOVED***

        /**
         * Checks spacing after `$***REMOVED***` of a given token.
         * @param ***REMOVED***Token***REMOVED*** token - A token to check. This is a Template token.
         * @returns ***REMOVED***void***REMOVED***
         */
        function checkSpacingAfter(token) ***REMOVED***
            const nextToken = sourceCode.getTokenAfter(token);

            if (nextToken &&
                OPEN_PAREN.test(token.value) &&
                astUtils.isTokenOnSameLine(token, nextToken) &&
                sourceCode.isSpaceBetweenTokens(token, nextToken) !== always
            ) ***REMOVED***
                context.report(***REMOVED***
                    loc: ***REMOVED***
                        line: token.loc.end.line,
                        column: token.loc.end.column - 2
                    ***REMOVED***,
                    messageId: `$***REMOVED***prefix***REMOVED***After`,
                    fix(fixer) ***REMOVED***
                        if (always) ***REMOVED***
                            return fixer.insertTextAfter(token, " ");
                        ***REMOVED***
                        return fixer.removeRange([
                            token.range[1],
                            nextToken.range[0]
                        ]);
                    ***REMOVED***
                ***REMOVED***);
            ***REMOVED***
        ***REMOVED***

        return ***REMOVED***
            TemplateElement(node) ***REMOVED***
                const token = sourceCode.getFirstToken(node);

                checkSpacingBefore(token);
                checkSpacingAfter(token);
            ***REMOVED***
        ***REMOVED***;
    ***REMOVED***
***REMOVED***;
