// eslint/lib/rules/vars-on-top.js
/**
 * @fileoverview Rule to enforce var declarations are only at the top of a function.
 * @author Danny Fritz
 * @author Gyandeep Singh
 */
"use strict";

//------------------------------------------------------------------------------
// Rule Definition
//------------------------------------------------------------------------------

module.exports = ***REMOVED***
    meta: ***REMOVED***
        type: "suggestion",

        docs: ***REMOVED***
            description: "require `var` declarations be placed at the top of their containing scope",
            category: "Best Practices",
            recommended: false,
            url: "https://eslint.org/docs/rules/vars-on-top"
        ***REMOVED***,

        schema: [],
        messages: ***REMOVED***
            top: "All 'var' declarations must be at the top of the function scope."
        ***REMOVED***
    ***REMOVED***,

    create(context) ***REMOVED***

        //--------------------------------------------------------------------------
        // Helpers
        //--------------------------------------------------------------------------

        /**
         * @param ***REMOVED***ASTNode***REMOVED*** node - any node
         * @returns ***REMOVED***boolean***REMOVED*** whether the given node structurally represents a directive
         */
        function looksLikeDirective(node) ***REMOVED***
            return node.type === "ExpressionStatement" &&
                node.expression.type === "Literal" && typeof node.expression.value === "string";
        ***REMOVED***

        /**
         * Check to see if its a ES6 import declaration
         * @param ***REMOVED***ASTNode***REMOVED*** node - any node
         * @returns ***REMOVED***boolean***REMOVED*** whether the given node represents a import declaration
         */
        function looksLikeImport(node) ***REMOVED***
            return node.type === "ImportDeclaration" || node.type === "ImportSpecifier" ||
                node.type === "ImportDefaultSpecifier" || node.type === "ImportNamespaceSpecifier";
        ***REMOVED***

        /**
         * Checks whether a given node is a variable declaration or not.
         *
         * @param ***REMOVED***ASTNode***REMOVED*** node - any node
         * @returns ***REMOVED***boolean***REMOVED*** `true` if the node is a variable declaration.
         */
        function isVariableDeclaration(node) ***REMOVED***
            return (
                node.type === "VariableDeclaration" ||
                (
                    node.type === "ExportNamedDeclaration" &&
                    node.declaration &&
                    node.declaration.type === "VariableDeclaration"
                )
            );
        ***REMOVED***

        /**
         * Checks whether this variable is on top of the block body
         * @param ***REMOVED***ASTNode***REMOVED*** node - The node to check
         * @param ***REMOVED***ASTNode[]***REMOVED*** statements - collection of ASTNodes for the parent node block
         * @returns ***REMOVED***boolean***REMOVED*** True if var is on top otherwise false
         */
        function isVarOnTop(node, statements) ***REMOVED***
            const l = statements.length;
            let i = 0;

            // skip over directives
            for (; i < l; ++i) ***REMOVED***
                if (!looksLikeDirective(statements[i]) && !looksLikeImport(statements[i])) ***REMOVED***
                    break;
                ***REMOVED***
            ***REMOVED***

            for (; i < l; ++i) ***REMOVED***
                if (!isVariableDeclaration(statements[i])) ***REMOVED***
                    return false;
                ***REMOVED***
                if (statements[i] === node) ***REMOVED***
                    return true;
                ***REMOVED***
            ***REMOVED***

            return false;
        ***REMOVED***

        /**
         * Checks whether variable is on top at the global level
         * @param ***REMOVED***ASTNode***REMOVED*** node - The node to check
         * @param ***REMOVED***ASTNode***REMOVED*** parent - Parent of the node
         * @returns ***REMOVED***void***REMOVED***
         */
        function globalVarCheck(node, parent) ***REMOVED***
            if (!isVarOnTop(node, parent.body)) ***REMOVED***
                context.report(***REMOVED*** node, messageId: "top" ***REMOVED***);
            ***REMOVED***
        ***REMOVED***

        /**
         * Checks whether variable is on top at functional block scope level
         * @param ***REMOVED***ASTNode***REMOVED*** node - The node to check
         * @param ***REMOVED***ASTNode***REMOVED*** parent - Parent of the node
         * @param ***REMOVED***ASTNode***REMOVED*** grandParent - Parent of the node's parent
         * @returns ***REMOVED***void***REMOVED***
         */
        function blockScopeVarCheck(node, parent, grandParent) ***REMOVED***
            if (!(/Function/u.test(grandParent.type) &&
                    parent.type === "BlockStatement" &&
                    isVarOnTop(node, parent.body))) ***REMOVED***
                context.report(***REMOVED*** node, messageId: "top" ***REMOVED***);
            ***REMOVED***
        ***REMOVED***

        //--------------------------------------------------------------------------
        // Public API
        //--------------------------------------------------------------------------

        return ***REMOVED***
            "VariableDeclaration[kind='var']"(node) ***REMOVED***
                if (node.parent.type === "ExportNamedDeclaration") ***REMOVED***
                    globalVarCheck(node.parent, node.parent.parent);
                ***REMOVED*** else if (node.parent.type === "Program") ***REMOVED***
                    globalVarCheck(node, node.parent);
                ***REMOVED*** else ***REMOVED***
                    blockScopeVarCheck(node, node.parent, node.parent.parent);
                ***REMOVED***
            ***REMOVED***
        ***REMOVED***;

    ***REMOVED***
***REMOVED***;
