// rxjs/_esm5/internal/operators/bufferCount.js
/** PURE_IMPORTS_START tslib,_Subscriber PURE_IMPORTS_END */
import * as tslib_1 from "tslib";
import ***REMOVED*** Subscriber ***REMOVED*** from '../Subscriber';
export function bufferCount(bufferSize, startBufferEvery) ***REMOVED***
    if (startBufferEvery === void 0) ***REMOVED***
        startBufferEvery = null;
    ***REMOVED***
    return function bufferCountOperatorFunction(source) ***REMOVED***
        return source.lift(new BufferCountOperator(bufferSize, startBufferEvery));
    ***REMOVED***;
***REMOVED***
var BufferCountOperator = /*@__PURE__*/ (function () ***REMOVED***
    function BufferCountOperator(bufferSize, startBufferEvery) ***REMOVED***
        this.bufferSize = bufferSize;
        this.startBufferEvery = startBufferEvery;
        if (!startBufferEvery || bufferSize === startBufferEvery) ***REMOVED***
            this.subscriberClass = BufferCountSubscriber;
        ***REMOVED***
        else ***REMOVED***
            this.subscriberClass = BufferSkipCountSubscriber;
        ***REMOVED***
    ***REMOVED***
    BufferCountOperator.prototype.call = function (subscriber, source) ***REMOVED***
        return source.subscribe(new this.subscriberClass(subscriber, this.bufferSize, this.startBufferEvery));
    ***REMOVED***;
    return BufferCountOperator;
***REMOVED***());
var BufferCountSubscriber = /*@__PURE__*/ (function (_super) ***REMOVED***
    tslib_1.__extends(BufferCountSubscriber, _super);
    function BufferCountSubscriber(destination, bufferSize) ***REMOVED***
        var _this = _super.call(this, destination) || this;
        _this.bufferSize = bufferSize;
        _this.buffer = [];
        return _this;
    ***REMOVED***
    BufferCountSubscriber.prototype._next = function (value) ***REMOVED***
        var buffer = this.buffer;
        buffer.push(value);
        if (buffer.length == this.bufferSize) ***REMOVED***
            this.destination.next(buffer);
            this.buffer = [];
        ***REMOVED***
    ***REMOVED***;
    BufferCountSubscriber.prototype._complete = function () ***REMOVED***
        var buffer = this.buffer;
        if (buffer.length > 0) ***REMOVED***
            this.destination.next(buffer);
        ***REMOVED***
        _super.prototype._complete.call(this);
    ***REMOVED***;
    return BufferCountSubscriber;
***REMOVED***(Subscriber));
var BufferSkipCountSubscriber = /*@__PURE__*/ (function (_super) ***REMOVED***
    tslib_1.__extends(BufferSkipCountSubscriber, _super);
    function BufferSkipCountSubscriber(destination, bufferSize, startBufferEvery) ***REMOVED***
        var _this = _super.call(this, destination) || this;
        _this.bufferSize = bufferSize;
        _this.startBufferEvery = startBufferEvery;
        _this.buffers = [];
        _this.count = 0;
        return _this;
    ***REMOVED***
    BufferSkipCountSubscriber.prototype._next = function (value) ***REMOVED***
        var _a = this, bufferSize = _a.bufferSize, startBufferEvery = _a.startBufferEvery, buffers = _a.buffers, count = _a.count;
        this.count++;
        if (count % startBufferEvery === 0) ***REMOVED***
            buffers.push([]);
        ***REMOVED***
        for (var i = buffers.length; i--;) ***REMOVED***
            var buffer = buffers[i];
            buffer.push(value);
            if (buffer.length === bufferSize) ***REMOVED***
                buffers.splice(i, 1);
                this.destination.next(buffer);
            ***REMOVED***
        ***REMOVED***
    ***REMOVED***;
    BufferSkipCountSubscriber.prototype._complete = function () ***REMOVED***
        var _a = this, buffers = _a.buffers, destination = _a.destination;
        while (buffers.length > 0) ***REMOVED***
            var buffer = buffers.shift();
            if (buffer.length > 0) ***REMOVED***
                destination.next(buffer);
            ***REMOVED***
        ***REMOVED***
        _super.prototype._complete.call(this);
    ***REMOVED***;
    return BufferSkipCountSubscriber;
***REMOVED***(Subscriber));
