// rxjs/_esm5/internal/operators/delay.js
/** PURE_IMPORTS_START tslib,_scheduler_async,_util_isDate,_Subscriber,_Notification PURE_IMPORTS_END */
import * as tslib_1 from "tslib";
import ***REMOVED*** async ***REMOVED*** from '../scheduler/async';
import ***REMOVED*** isDate ***REMOVED*** from '../util/isDate';
import ***REMOVED*** Subscriber ***REMOVED*** from '../Subscriber';
import ***REMOVED*** Notification ***REMOVED*** from '../Notification';
export function delay(delay, scheduler) ***REMOVED***
    if (scheduler === void 0) ***REMOVED***
        scheduler = async;
    ***REMOVED***
    var absoluteDelay = isDate(delay);
    var delayFor = absoluteDelay ? (+delay - scheduler.now()) : Math.abs(delay);
    return function (source) ***REMOVED*** return source.lift(new DelayOperator(delayFor, scheduler)); ***REMOVED***;
***REMOVED***
var DelayOperator = /*@__PURE__*/ (function () ***REMOVED***
    function DelayOperator(delay, scheduler) ***REMOVED***
        this.delay = delay;
        this.scheduler = scheduler;
    ***REMOVED***
    DelayOperator.prototype.call = function (subscriber, source) ***REMOVED***
        return source.subscribe(new DelaySubscriber(subscriber, this.delay, this.scheduler));
    ***REMOVED***;
    return DelayOperator;
***REMOVED***());
var DelaySubscriber = /*@__PURE__*/ (function (_super) ***REMOVED***
    tslib_1.__extends(DelaySubscriber, _super);
    function DelaySubscriber(destination, delay, scheduler) ***REMOVED***
        var _this = _super.call(this, destination) || this;
        _this.delay = delay;
        _this.scheduler = scheduler;
        _this.queue = [];
        _this.active = false;
        _this.errored = false;
        return _this;
    ***REMOVED***
    DelaySubscriber.dispatch = function (state) ***REMOVED***
        var source = state.source;
        var queue = source.queue;
        var scheduler = state.scheduler;
        var destination = state.destination;
        while (queue.length > 0 && (queue[0].time - scheduler.now()) <= 0) ***REMOVED***
            queue.shift().notification.observe(destination);
        ***REMOVED***
        if (queue.length > 0) ***REMOVED***
            var delay_1 = Math.max(0, queue[0].time - scheduler.now());
            this.schedule(state, delay_1);
        ***REMOVED***
        else ***REMOVED***
            this.unsubscribe();
            source.active = false;
        ***REMOVED***
    ***REMOVED***;
    DelaySubscriber.prototype._schedule = function (scheduler) ***REMOVED***
        this.active = true;
        var destination = this.destination;
        destination.add(scheduler.schedule(DelaySubscriber.dispatch, this.delay, ***REMOVED***
            source: this, destination: this.destination, scheduler: scheduler
        ***REMOVED***));
    ***REMOVED***;
    DelaySubscriber.prototype.scheduleNotification = function (notification) ***REMOVED***
        if (this.errored === true) ***REMOVED***
            return;
        ***REMOVED***
        var scheduler = this.scheduler;
        var message = new DelayMessage(scheduler.now() + this.delay, notification);
        this.queue.push(message);
        if (this.active === false) ***REMOVED***
            this._schedule(scheduler);
        ***REMOVED***
    ***REMOVED***;
    DelaySubscriber.prototype._next = function (value) ***REMOVED***
        this.scheduleNotification(Notification.createNext(value));
    ***REMOVED***;
    DelaySubscriber.prototype._error = function (err) ***REMOVED***
        this.errored = true;
        this.queue = [];
        this.destination.error(err);
        this.unsubscribe();
    ***REMOVED***;
    DelaySubscriber.prototype._complete = function () ***REMOVED***
        this.scheduleNotification(Notification.createComplete());
        this.unsubscribe();
    ***REMOVED***;
    return DelaySubscriber;
***REMOVED***(Subscriber));
var DelayMessage = /*@__PURE__*/ (function () ***REMOVED***
    function DelayMessage(time, notification) ***REMOVED***
        this.time = time;
        this.notification = notification;
    ***REMOVED***
    return DelayMessage;
***REMOVED***());
