// rxjs/_esm5/internal/operators/distinct.js
/** PURE_IMPORTS_START tslib,_OuterSubscriber,_util_subscribeToResult PURE_IMPORTS_END */
import * as tslib_1 from "tslib";
import ***REMOVED*** OuterSubscriber ***REMOVED*** from '../OuterSubscriber';
import ***REMOVED*** subscribeToResult ***REMOVED*** from '../util/subscribeToResult';
export function distinct(keySelector, flushes) ***REMOVED***
    return function (source) ***REMOVED*** return source.lift(new DistinctOperator(keySelector, flushes)); ***REMOVED***;
***REMOVED***
var DistinctOperator = /*@__PURE__*/ (function () ***REMOVED***
    function DistinctOperator(keySelector, flushes) ***REMOVED***
        this.keySelector = keySelector;
        this.flushes = flushes;
    ***REMOVED***
    DistinctOperator.prototype.call = function (subscriber, source) ***REMOVED***
        return source.subscribe(new DistinctSubscriber(subscriber, this.keySelector, this.flushes));
    ***REMOVED***;
    return DistinctOperator;
***REMOVED***());
var DistinctSubscriber = /*@__PURE__*/ (function (_super) ***REMOVED***
    tslib_1.__extends(DistinctSubscriber, _super);
    function DistinctSubscriber(destination, keySelector, flushes) ***REMOVED***
        var _this = _super.call(this, destination) || this;
        _this.keySelector = keySelector;
        _this.values = new Set();
        if (flushes) ***REMOVED***
            _this.add(subscribeToResult(_this, flushes));
        ***REMOVED***
        return _this;
    ***REMOVED***
    DistinctSubscriber.prototype.notifyNext = function (outerValue, innerValue, outerIndex, innerIndex, innerSub) ***REMOVED***
        this.values.clear();
    ***REMOVED***;
    DistinctSubscriber.prototype.notifyError = function (error, innerSub) ***REMOVED***
        this._error(error);
    ***REMOVED***;
    DistinctSubscriber.prototype._next = function (value) ***REMOVED***
        if (this.keySelector) ***REMOVED***
            this._useKeySelector(value);
        ***REMOVED***
        else ***REMOVED***
            this._finalizeNext(value, value);
        ***REMOVED***
    ***REMOVED***;
    DistinctSubscriber.prototype._useKeySelector = function (value) ***REMOVED***
        var key;
        var destination = this.destination;
        try ***REMOVED***
            key = this.keySelector(value);
        ***REMOVED***
        catch (err) ***REMOVED***
            destination.error(err);
            return;
        ***REMOVED***
        this._finalizeNext(key, value);
    ***REMOVED***;
    DistinctSubscriber.prototype._finalizeNext = function (key, value) ***REMOVED***
        var values = this.values;
        if (!values.has(key)) ***REMOVED***
            values.add(key);
            this.destination.next(value);
        ***REMOVED***
    ***REMOVED***;
    return DistinctSubscriber;
***REMOVED***(OuterSubscriber));
export ***REMOVED*** DistinctSubscriber ***REMOVED***;
