// rxjs/_esm5/internal/operators/every.js
/** PURE_IMPORTS_START tslib,_Subscriber PURE_IMPORTS_END */
import * as tslib_1 from "tslib";
import ***REMOVED*** Subscriber ***REMOVED*** from '../Subscriber';
export function every(predicate, thisArg) ***REMOVED***
    return function (source) ***REMOVED*** return source.lift(new EveryOperator(predicate, thisArg, source)); ***REMOVED***;
***REMOVED***
var EveryOperator = /*@__PURE__*/ (function () ***REMOVED***
    function EveryOperator(predicate, thisArg, source) ***REMOVED***
        this.predicate = predicate;
        this.thisArg = thisArg;
        this.source = source;
    ***REMOVED***
    EveryOperator.prototype.call = function (observer, source) ***REMOVED***
        return source.subscribe(new EverySubscriber(observer, this.predicate, this.thisArg, this.source));
    ***REMOVED***;
    return EveryOperator;
***REMOVED***());
var EverySubscriber = /*@__PURE__*/ (function (_super) ***REMOVED***
    tslib_1.__extends(EverySubscriber, _super);
    function EverySubscriber(destination, predicate, thisArg, source) ***REMOVED***
        var _this = _super.call(this, destination) || this;
        _this.predicate = predicate;
        _this.thisArg = thisArg;
        _this.source = source;
        _this.index = 0;
        _this.thisArg = thisArg || _this;
        return _this;
    ***REMOVED***
    EverySubscriber.prototype.notifyComplete = function (everyValueMatch) ***REMOVED***
        this.destination.next(everyValueMatch);
        this.destination.complete();
    ***REMOVED***;
    EverySubscriber.prototype._next = function (value) ***REMOVED***
        var result = false;
        try ***REMOVED***
            result = this.predicate.call(this.thisArg, value, this.index++, this.source);
        ***REMOVED***
        catch (err) ***REMOVED***
            this.destination.error(err);
            return;
        ***REMOVED***
        if (!result) ***REMOVED***
            this.notifyComplete(false);
        ***REMOVED***
    ***REMOVED***;
    EverySubscriber.prototype._complete = function () ***REMOVED***
        this.notifyComplete(true);
    ***REMOVED***;
    return EverySubscriber;
***REMOVED***(Subscriber));
