// rxjs/_esm5/internal/operators/filter.js
/** PURE_IMPORTS_START tslib,_Subscriber PURE_IMPORTS_END */
import * as tslib_1 from "tslib";
import ***REMOVED*** Subscriber ***REMOVED*** from '../Subscriber';
export function filter(predicate, thisArg) ***REMOVED***
    return function filterOperatorFunction(source) ***REMOVED***
        return source.lift(new FilterOperator(predicate, thisArg));
    ***REMOVED***;
***REMOVED***
var FilterOperator = /*@__PURE__*/ (function () ***REMOVED***
    function FilterOperator(predicate, thisArg) ***REMOVED***
        this.predicate = predicate;
        this.thisArg = thisArg;
    ***REMOVED***
    FilterOperator.prototype.call = function (subscriber, source) ***REMOVED***
        return source.subscribe(new FilterSubscriber(subscriber, this.predicate, this.thisArg));
    ***REMOVED***;
    return FilterOperator;
***REMOVED***());
var FilterSubscriber = /*@__PURE__*/ (function (_super) ***REMOVED***
    tslib_1.__extends(FilterSubscriber, _super);
    function FilterSubscriber(destination, predicate, thisArg) ***REMOVED***
        var _this = _super.call(this, destination) || this;
        _this.predicate = predicate;
        _this.thisArg = thisArg;
        _this.count = 0;
        return _this;
    ***REMOVED***
    FilterSubscriber.prototype._next = function (value) ***REMOVED***
        var result;
        try ***REMOVED***
            result = this.predicate.call(this.thisArg, value, this.count++);
        ***REMOVED***
        catch (err) ***REMOVED***
            this.destination.error(err);
            return;
        ***REMOVED***
        if (result) ***REMOVED***
            this.destination.next(value);
        ***REMOVED***
    ***REMOVED***;
    return FilterSubscriber;
***REMOVED***(Subscriber));
