// rxjs/_esm5/internal/operators/find.js
/** PURE_IMPORTS_START tslib,_Subscriber PURE_IMPORTS_END */
import * as tslib_1 from "tslib";
import ***REMOVED*** Subscriber ***REMOVED*** from '../Subscriber';
export function find(predicate, thisArg) ***REMOVED***
    if (typeof predicate !== 'function') ***REMOVED***
        throw new TypeError('predicate is not a function');
    ***REMOVED***
    return function (source) ***REMOVED*** return source.lift(new FindValueOperator(predicate, source, false, thisArg)); ***REMOVED***;
***REMOVED***
var FindValueOperator = /*@__PURE__*/ (function () ***REMOVED***
    function FindValueOperator(predicate, source, yieldIndex, thisArg) ***REMOVED***
        this.predicate = predicate;
        this.source = source;
        this.yieldIndex = yieldIndex;
        this.thisArg = thisArg;
    ***REMOVED***
    FindValueOperator.prototype.call = function (observer, source) ***REMOVED***
        return source.subscribe(new FindValueSubscriber(observer, this.predicate, this.source, this.yieldIndex, this.thisArg));
    ***REMOVED***;
    return FindValueOperator;
***REMOVED***());
export ***REMOVED*** FindValueOperator ***REMOVED***;
var FindValueSubscriber = /*@__PURE__*/ (function (_super) ***REMOVED***
    tslib_1.__extends(FindValueSubscriber, _super);
    function FindValueSubscriber(destination, predicate, source, yieldIndex, thisArg) ***REMOVED***
        var _this = _super.call(this, destination) || this;
        _this.predicate = predicate;
        _this.source = source;
        _this.yieldIndex = yieldIndex;
        _this.thisArg = thisArg;
        _this.index = 0;
        return _this;
    ***REMOVED***
    FindValueSubscriber.prototype.notifyComplete = function (value) ***REMOVED***
        var destination = this.destination;
        destination.next(value);
        destination.complete();
        this.unsubscribe();
    ***REMOVED***;
    FindValueSubscriber.prototype._next = function (value) ***REMOVED***
        var _a = this, predicate = _a.predicate, thisArg = _a.thisArg;
        var index = this.index++;
        try ***REMOVED***
            var result = predicate.call(thisArg || this, value, index, this.source);
            if (result) ***REMOVED***
                this.notifyComplete(this.yieldIndex ? index : value);
            ***REMOVED***
        ***REMOVED***
        catch (err) ***REMOVED***
            this.destination.error(err);
        ***REMOVED***
    ***REMOVED***;
    FindValueSubscriber.prototype._complete = function () ***REMOVED***
        this.notifyComplete(this.yieldIndex ? -1 : undefined);
    ***REMOVED***;
    return FindValueSubscriber;
***REMOVED***(Subscriber));
export ***REMOVED*** FindValueSubscriber ***REMOVED***;
