// rxjs/_esm5/internal/operators/multicast.js
/** PURE_IMPORTS_START _observable_ConnectableObservable PURE_IMPORTS_END */
import ***REMOVED*** connectableObservableDescriptor ***REMOVED*** from '../observable/ConnectableObservable';
export function multicast(subjectOrSubjectFactory, selector) ***REMOVED***
    return function multicastOperatorFunction(source) ***REMOVED***
        var subjectFactory;
        if (typeof subjectOrSubjectFactory === 'function') ***REMOVED***
            subjectFactory = subjectOrSubjectFactory;
        ***REMOVED***
        else ***REMOVED***
            subjectFactory = function subjectFactory() ***REMOVED***
                return subjectOrSubjectFactory;
            ***REMOVED***;
        ***REMOVED***
        if (typeof selector === 'function') ***REMOVED***
            return source.lift(new MulticastOperator(subjectFactory, selector));
        ***REMOVED***
        var connectable = Object.create(source, connectableObservableDescriptor);
        connectable.source = source;
        connectable.subjectFactory = subjectFactory;
        return connectable;
    ***REMOVED***;
***REMOVED***
var MulticastOperator = /*@__PURE__*/ (function () ***REMOVED***
    function MulticastOperator(subjectFactory, selector) ***REMOVED***
        this.subjectFactory = subjectFactory;
        this.selector = selector;
    ***REMOVED***
    MulticastOperator.prototype.call = function (subscriber, source) ***REMOVED***
        var selector = this.selector;
        var subject = this.subjectFactory();
        var subscription = selector(subject).subscribe(subscriber);
        subscription.add(source.subscribe(subject));
        return subscription;
    ***REMOVED***;
    return MulticastOperator;
***REMOVED***());
export ***REMOVED*** MulticastOperator ***REMOVED***;
