// rxjs/_esm5/internal/operators/pairwise.js
/** PURE_IMPORTS_START tslib,_Subscriber PURE_IMPORTS_END */
import * as tslib_1 from "tslib";
import ***REMOVED*** Subscriber ***REMOVED*** from '../Subscriber';
export function pairwise() ***REMOVED***
    return function (source) ***REMOVED*** return source.lift(new PairwiseOperator()); ***REMOVED***;
***REMOVED***
var PairwiseOperator = /*@__PURE__*/ (function () ***REMOVED***
    function PairwiseOperator() ***REMOVED***
    ***REMOVED***
    PairwiseOperator.prototype.call = function (subscriber, source) ***REMOVED***
        return source.subscribe(new PairwiseSubscriber(subscriber));
    ***REMOVED***;
    return PairwiseOperator;
***REMOVED***());
var PairwiseSubscriber = /*@__PURE__*/ (function (_super) ***REMOVED***
    tslib_1.__extends(PairwiseSubscriber, _super);
    function PairwiseSubscriber(destination) ***REMOVED***
        var _this = _super.call(this, destination) || this;
        _this.hasPrev = false;
        return _this;
    ***REMOVED***
    PairwiseSubscriber.prototype._next = function (value) ***REMOVED***
        var pair;
        if (this.hasPrev) ***REMOVED***
            pair = [this.prev, value];
        ***REMOVED***
        else ***REMOVED***
            this.hasPrev = true;
        ***REMOVED***
        this.prev = value;
        if (pair) ***REMOVED***
            this.destination.next(pair);
        ***REMOVED***
    ***REMOVED***;
    return PairwiseSubscriber;
***REMOVED***(Subscriber));
