// rxjs/_esm5/internal/operators/repeat.js
/** PURE_IMPORTS_START tslib,_Subscriber,_observable_empty PURE_IMPORTS_END */
import * as tslib_1 from "tslib";
import ***REMOVED*** Subscriber ***REMOVED*** from '../Subscriber';
import ***REMOVED*** empty ***REMOVED*** from '../observable/empty';
export function repeat(count) ***REMOVED***
    if (count === void 0) ***REMOVED***
        count = -1;
    ***REMOVED***
    return function (source) ***REMOVED***
        if (count === 0) ***REMOVED***
            return empty();
        ***REMOVED***
        else if (count < 0) ***REMOVED***
            return source.lift(new RepeatOperator(-1, source));
        ***REMOVED***
        else ***REMOVED***
            return source.lift(new RepeatOperator(count - 1, source));
        ***REMOVED***
    ***REMOVED***;
***REMOVED***
var RepeatOperator = /*@__PURE__*/ (function () ***REMOVED***
    function RepeatOperator(count, source) ***REMOVED***
        this.count = count;
        this.source = source;
    ***REMOVED***
    RepeatOperator.prototype.call = function (subscriber, source) ***REMOVED***
        return source.subscribe(new RepeatSubscriber(subscriber, this.count, this.source));
    ***REMOVED***;
    return RepeatOperator;
***REMOVED***());
var RepeatSubscriber = /*@__PURE__*/ (function (_super) ***REMOVED***
    tslib_1.__extends(RepeatSubscriber, _super);
    function RepeatSubscriber(destination, count, source) ***REMOVED***
        var _this = _super.call(this, destination) || this;
        _this.count = count;
        _this.source = source;
        return _this;
    ***REMOVED***
    RepeatSubscriber.prototype.complete = function () ***REMOVED***
        if (!this.isStopped) ***REMOVED***
            var _a = this, source = _a.source, count = _a.count;
            if (count === 0) ***REMOVED***
                return _super.prototype.complete.call(this);
            ***REMOVED***
            else if (count > -1) ***REMOVED***
                this.count = count - 1;
            ***REMOVED***
            source.subscribe(this._unsubscribeAndRecycle());
        ***REMOVED***
    ***REMOVED***;
    return RepeatSubscriber;
***REMOVED***(Subscriber));
