// rxjs/_esm5/internal/operators/retryWhen.js
/** PURE_IMPORTS_START tslib,_Subject,_OuterSubscriber,_util_subscribeToResult PURE_IMPORTS_END */
import * as tslib_1 from "tslib";
import ***REMOVED*** Subject ***REMOVED*** from '../Subject';
import ***REMOVED*** OuterSubscriber ***REMOVED*** from '../OuterSubscriber';
import ***REMOVED*** subscribeToResult ***REMOVED*** from '../util/subscribeToResult';
export function retryWhen(notifier) ***REMOVED***
    return function (source) ***REMOVED*** return source.lift(new RetryWhenOperator(notifier, source)); ***REMOVED***;
***REMOVED***
var RetryWhenOperator = /*@__PURE__*/ (function () ***REMOVED***
    function RetryWhenOperator(notifier, source) ***REMOVED***
        this.notifier = notifier;
        this.source = source;
    ***REMOVED***
    RetryWhenOperator.prototype.call = function (subscriber, source) ***REMOVED***
        return source.subscribe(new RetryWhenSubscriber(subscriber, this.notifier, this.source));
    ***REMOVED***;
    return RetryWhenOperator;
***REMOVED***());
var RetryWhenSubscriber = /*@__PURE__*/ (function (_super) ***REMOVED***
    tslib_1.__extends(RetryWhenSubscriber, _super);
    function RetryWhenSubscriber(destination, notifier, source) ***REMOVED***
        var _this = _super.call(this, destination) || this;
        _this.notifier = notifier;
        _this.source = source;
        return _this;
    ***REMOVED***
    RetryWhenSubscriber.prototype.error = function (err) ***REMOVED***
        if (!this.isStopped) ***REMOVED***
            var errors = this.errors;
            var retries = this.retries;
            var retriesSubscription = this.retriesSubscription;
            if (!retries) ***REMOVED***
                errors = new Subject();
                try ***REMOVED***
                    var notifier = this.notifier;
                    retries = notifier(errors);
                ***REMOVED***
                catch (e) ***REMOVED***
                    return _super.prototype.error.call(this, e);
                ***REMOVED***
                retriesSubscription = subscribeToResult(this, retries);
            ***REMOVED***
            else ***REMOVED***
                this.errors = null;
                this.retriesSubscription = null;
            ***REMOVED***
            this._unsubscribeAndRecycle();
            this.errors = errors;
            this.retries = retries;
            this.retriesSubscription = retriesSubscription;
            errors.next(err);
        ***REMOVED***
    ***REMOVED***;
    RetryWhenSubscriber.prototype._unsubscribe = function () ***REMOVED***
        var _a = this, errors = _a.errors, retriesSubscription = _a.retriesSubscription;
        if (errors) ***REMOVED***
            errors.unsubscribe();
            this.errors = null;
        ***REMOVED***
        if (retriesSubscription) ***REMOVED***
            retriesSubscription.unsubscribe();
            this.retriesSubscription = null;
        ***REMOVED***
        this.retries = null;
    ***REMOVED***;
    RetryWhenSubscriber.prototype.notifyNext = function (outerValue, innerValue, outerIndex, innerIndex, innerSub) ***REMOVED***
        var _unsubscribe = this._unsubscribe;
        this._unsubscribe = null;
        this._unsubscribeAndRecycle();
        this._unsubscribe = _unsubscribe;
        this.source.subscribe(this);
    ***REMOVED***;
    return RetryWhenSubscriber;
***REMOVED***(OuterSubscriber));
