// rxjs/_esm5/internal/operators/sampleTime.js
/** PURE_IMPORTS_START tslib,_Subscriber,_scheduler_async PURE_IMPORTS_END */
import * as tslib_1 from "tslib";
import ***REMOVED*** Subscriber ***REMOVED*** from '../Subscriber';
import ***REMOVED*** async ***REMOVED*** from '../scheduler/async';
export function sampleTime(period, scheduler) ***REMOVED***
    if (scheduler === void 0) ***REMOVED***
        scheduler = async;
    ***REMOVED***
    return function (source) ***REMOVED*** return source.lift(new SampleTimeOperator(period, scheduler)); ***REMOVED***;
***REMOVED***
var SampleTimeOperator = /*@__PURE__*/ (function () ***REMOVED***
    function SampleTimeOperator(period, scheduler) ***REMOVED***
        this.period = period;
        this.scheduler = scheduler;
    ***REMOVED***
    SampleTimeOperator.prototype.call = function (subscriber, source) ***REMOVED***
        return source.subscribe(new SampleTimeSubscriber(subscriber, this.period, this.scheduler));
    ***REMOVED***;
    return SampleTimeOperator;
***REMOVED***());
var SampleTimeSubscriber = /*@__PURE__*/ (function (_super) ***REMOVED***
    tslib_1.__extends(SampleTimeSubscriber, _super);
    function SampleTimeSubscriber(destination, period, scheduler) ***REMOVED***
        var _this = _super.call(this, destination) || this;
        _this.period = period;
        _this.scheduler = scheduler;
        _this.hasValue = false;
        _this.add(scheduler.schedule(dispatchNotification, period, ***REMOVED*** subscriber: _this, period: period ***REMOVED***));
        return _this;
    ***REMOVED***
    SampleTimeSubscriber.prototype._next = function (value) ***REMOVED***
        this.lastValue = value;
        this.hasValue = true;
    ***REMOVED***;
    SampleTimeSubscriber.prototype.notifyNext = function () ***REMOVED***
        if (this.hasValue) ***REMOVED***
            this.hasValue = false;
            this.destination.next(this.lastValue);
        ***REMOVED***
    ***REMOVED***;
    return SampleTimeSubscriber;
***REMOVED***(Subscriber));
function dispatchNotification(state) ***REMOVED***
    var subscriber = state.subscriber, period = state.period;
    subscriber.notifyNext();
    this.schedule(state, period);
***REMOVED***
