// rxjs/_esm5/internal/operators/switchMap.js
/** PURE_IMPORTS_START tslib,_OuterSubscriber,_InnerSubscriber,_util_subscribeToResult,_map,_observable_from PURE_IMPORTS_END */
import * as tslib_1 from "tslib";
import ***REMOVED*** OuterSubscriber ***REMOVED*** from '../OuterSubscriber';
import ***REMOVED*** InnerSubscriber ***REMOVED*** from '../InnerSubscriber';
import ***REMOVED*** subscribeToResult ***REMOVED*** from '../util/subscribeToResult';
import ***REMOVED*** map ***REMOVED*** from './map';
import ***REMOVED*** from ***REMOVED*** from '../observable/from';
export function switchMap(project, resultSelector) ***REMOVED***
    if (typeof resultSelector === 'function') ***REMOVED***
        return function (source) ***REMOVED*** return source.pipe(switchMap(function (a, i) ***REMOVED*** return from(project(a, i)).pipe(map(function (b, ii) ***REMOVED*** return resultSelector(a, b, i, ii); ***REMOVED***)); ***REMOVED***)); ***REMOVED***;
    ***REMOVED***
    return function (source) ***REMOVED*** return source.lift(new SwitchMapOperator(project)); ***REMOVED***;
***REMOVED***
var SwitchMapOperator = /*@__PURE__*/ (function () ***REMOVED***
    function SwitchMapOperator(project) ***REMOVED***
        this.project = project;
    ***REMOVED***
    SwitchMapOperator.prototype.call = function (subscriber, source) ***REMOVED***
        return source.subscribe(new SwitchMapSubscriber(subscriber, this.project));
    ***REMOVED***;
    return SwitchMapOperator;
***REMOVED***());
var SwitchMapSubscriber = /*@__PURE__*/ (function (_super) ***REMOVED***
    tslib_1.__extends(SwitchMapSubscriber, _super);
    function SwitchMapSubscriber(destination, project) ***REMOVED***
        var _this = _super.call(this, destination) || this;
        _this.project = project;
        _this.index = 0;
        return _this;
    ***REMOVED***
    SwitchMapSubscriber.prototype._next = function (value) ***REMOVED***
        var result;
        var index = this.index++;
        try ***REMOVED***
            result = this.project(value, index);
        ***REMOVED***
        catch (error) ***REMOVED***
            this.destination.error(error);
            return;
        ***REMOVED***
        this._innerSub(result, value, index);
    ***REMOVED***;
    SwitchMapSubscriber.prototype._innerSub = function (result, value, index) ***REMOVED***
        var innerSubscription = this.innerSubscription;
        if (innerSubscription) ***REMOVED***
            innerSubscription.unsubscribe();
        ***REMOVED***
        var innerSubscriber = new InnerSubscriber(this, value, index);
        var destination = this.destination;
        destination.add(innerSubscriber);
        this.innerSubscription = subscribeToResult(this, result, undefined, undefined, innerSubscriber);
        if (this.innerSubscription !== innerSubscriber) ***REMOVED***
            destination.add(this.innerSubscription);
        ***REMOVED***
    ***REMOVED***;
    SwitchMapSubscriber.prototype._complete = function () ***REMOVED***
        var innerSubscription = this.innerSubscription;
        if (!innerSubscription || innerSubscription.closed) ***REMOVED***
            _super.prototype._complete.call(this);
        ***REMOVED***
        this.unsubscribe();
    ***REMOVED***;
    SwitchMapSubscriber.prototype._unsubscribe = function () ***REMOVED***
        this.innerSubscription = null;
    ***REMOVED***;
    SwitchMapSubscriber.prototype.notifyComplete = function (innerSub) ***REMOVED***
        var destination = this.destination;
        destination.remove(innerSub);
        this.innerSubscription = null;
        if (this.isStopped) ***REMOVED***
            _super.prototype._complete.call(this);
        ***REMOVED***
    ***REMOVED***;
    SwitchMapSubscriber.prototype.notifyNext = function (outerValue, innerValue, outerIndex, innerIndex, innerSub) ***REMOVED***
        this.destination.next(innerValue);
    ***REMOVED***;
    return SwitchMapSubscriber;
***REMOVED***(OuterSubscriber));
