// rxjs/_esm5/internal/operators/tap.js
/** PURE_IMPORTS_START tslib,_Subscriber,_util_noop,_util_isFunction PURE_IMPORTS_END */
import * as tslib_1 from "tslib";
import ***REMOVED*** Subscriber ***REMOVED*** from '../Subscriber';
import ***REMOVED*** noop ***REMOVED*** from '../util/noop';
import ***REMOVED*** isFunction ***REMOVED*** from '../util/isFunction';
export function tap(nextOrObserver, error, complete) ***REMOVED***
    return function tapOperatorFunction(source) ***REMOVED***
        return source.lift(new DoOperator(nextOrObserver, error, complete));
    ***REMOVED***;
***REMOVED***
var DoOperator = /*@__PURE__*/ (function () ***REMOVED***
    function DoOperator(nextOrObserver, error, complete) ***REMOVED***
        this.nextOrObserver = nextOrObserver;
        this.error = error;
        this.complete = complete;
    ***REMOVED***
    DoOperator.prototype.call = function (subscriber, source) ***REMOVED***
        return source.subscribe(new TapSubscriber(subscriber, this.nextOrObserver, this.error, this.complete));
    ***REMOVED***;
    return DoOperator;
***REMOVED***());
var TapSubscriber = /*@__PURE__*/ (function (_super) ***REMOVED***
    tslib_1.__extends(TapSubscriber, _super);
    function TapSubscriber(destination, observerOrNext, error, complete) ***REMOVED***
        var _this = _super.call(this, destination) || this;
        _this._tapNext = noop;
        _this._tapError = noop;
        _this._tapComplete = noop;
        _this._tapError = error || noop;
        _this._tapComplete = complete || noop;
        if (isFunction(observerOrNext)) ***REMOVED***
            _this._context = _this;
            _this._tapNext = observerOrNext;
        ***REMOVED***
        else if (observerOrNext) ***REMOVED***
            _this._context = observerOrNext;
            _this._tapNext = observerOrNext.next || noop;
            _this._tapError = observerOrNext.error || noop;
            _this._tapComplete = observerOrNext.complete || noop;
        ***REMOVED***
        return _this;
    ***REMOVED***
    TapSubscriber.prototype._next = function (value) ***REMOVED***
        try ***REMOVED***
            this._tapNext.call(this._context, value);
        ***REMOVED***
        catch (err) ***REMOVED***
            this.destination.error(err);
            return;
        ***REMOVED***
        this.destination.next(value);
    ***REMOVED***;
    TapSubscriber.prototype._error = function (err) ***REMOVED***
        try ***REMOVED***
            this._tapError.call(this._context, err);
        ***REMOVED***
        catch (err) ***REMOVED***
            this.destination.error(err);
            return;
        ***REMOVED***
        this.destination.error(err);
    ***REMOVED***;
    TapSubscriber.prototype._complete = function () ***REMOVED***
        try ***REMOVED***
            this._tapComplete.call(this._context);
        ***REMOVED***
        catch (err) ***REMOVED***
            this.destination.error(err);
            return;
        ***REMOVED***
        return this.destination.complete();
    ***REMOVED***;
    return TapSubscriber;
***REMOVED***(Subscriber));
