// rxjs/_esm5/internal/operators/window.js
/** PURE_IMPORTS_START tslib,_Subject,_OuterSubscriber,_util_subscribeToResult PURE_IMPORTS_END */
import * as tslib_1 from "tslib";
import ***REMOVED*** Subject ***REMOVED*** from '../Subject';
import ***REMOVED*** OuterSubscriber ***REMOVED*** from '../OuterSubscriber';
import ***REMOVED*** subscribeToResult ***REMOVED*** from '../util/subscribeToResult';
export function window(windowBoundaries) ***REMOVED***
    return function windowOperatorFunction(source) ***REMOVED***
        return source.lift(new WindowOperator(windowBoundaries));
    ***REMOVED***;
***REMOVED***
var WindowOperator = /*@__PURE__*/ (function () ***REMOVED***
    function WindowOperator(windowBoundaries) ***REMOVED***
        this.windowBoundaries = windowBoundaries;
    ***REMOVED***
    WindowOperator.prototype.call = function (subscriber, source) ***REMOVED***
        var windowSubscriber = new WindowSubscriber(subscriber);
        var sourceSubscription = source.subscribe(windowSubscriber);
        if (!sourceSubscription.closed) ***REMOVED***
            windowSubscriber.add(subscribeToResult(windowSubscriber, this.windowBoundaries));
        ***REMOVED***
        return sourceSubscription;
    ***REMOVED***;
    return WindowOperator;
***REMOVED***());
var WindowSubscriber = /*@__PURE__*/ (function (_super) ***REMOVED***
    tslib_1.__extends(WindowSubscriber, _super);
    function WindowSubscriber(destination) ***REMOVED***
        var _this = _super.call(this, destination) || this;
        _this.window = new Subject();
        destination.next(_this.window);
        return _this;
    ***REMOVED***
    WindowSubscriber.prototype.notifyNext = function (outerValue, innerValue, outerIndex, innerIndex, innerSub) ***REMOVED***
        this.openWindow();
    ***REMOVED***;
    WindowSubscriber.prototype.notifyError = function (error, innerSub) ***REMOVED***
        this._error(error);
    ***REMOVED***;
    WindowSubscriber.prototype.notifyComplete = function (innerSub) ***REMOVED***
        this._complete();
    ***REMOVED***;
    WindowSubscriber.prototype._next = function (value) ***REMOVED***
        this.window.next(value);
    ***REMOVED***;
    WindowSubscriber.prototype._error = function (err) ***REMOVED***
        this.window.error(err);
        this.destination.error(err);
    ***REMOVED***;
    WindowSubscriber.prototype._complete = function () ***REMOVED***
        this.window.complete();
        this.destination.complete();
    ***REMOVED***;
    WindowSubscriber.prototype._unsubscribe = function () ***REMOVED***
        this.window = null;
    ***REMOVED***;
    WindowSubscriber.prototype.openWindow = function () ***REMOVED***
        var prevWindow = this.window;
        if (prevWindow) ***REMOVED***
            prevWindow.complete();
        ***REMOVED***
        var destination = this.destination;
        var newWindow = this.window = new Subject();
        destination.next(newWindow);
    ***REMOVED***;
    return WindowSubscriber;
***REMOVED***(OuterSubscriber));
