// discord.js/src/client/rest/RequestHandlers/Burst.js
const RequestHandler = require('./RequestHandler');
const DiscordAPIError = require('../DiscordAPIError');
const ***REMOVED*** Events: ***REMOVED*** RATE_LIMIT ***REMOVED*** ***REMOVED*** = require('../../../util/Constants');

class BurstRequestHandler extends RequestHandler ***REMOVED***
  constructor(restManager, endpoint) ***REMOVED***
    super(restManager, endpoint);

    this.client = restManager.client;

    this.limit = Infinity;
    this.resetTime = null;
    this.remaining = 1;
    this.timeDifference = 0;

    this.resetTimeout = null;
  ***REMOVED***

  push(request) ***REMOVED***
    super.push(request);
    this.handle();
  ***REMOVED***

  execute(item) ***REMOVED***
    if (!item) return;
    item.request.gen().end((err, res) => ***REMOVED***
      if (res && res.headers) ***REMOVED***
        this.limit = Number(res.headers['x-ratelimit-limit']);
        this.resetTime = Number(res.headers['x-ratelimit-reset']) * 1000;
        this.remaining = Number(res.headers['x-ratelimit-remaining']);
        this.timeDifference = Date.now() - new Date(res.headers.date).getTime();
      ***REMOVED***
      if (err) ***REMOVED***
        if (err.status === 429) ***REMOVED***
          this.queue.unshift(item);
          if (res.headers['x-ratelimit-global']) this.globalLimit = true;
          if (this.resetTimeout) return;
          this.resetTimeout = this.client.setTimeout(() => ***REMOVED***
            this.remaining = this.limit;
            this.globalLimit = false;
            this.handle();
            this.resetTimeout = null;
          ***REMOVED***, Number(res.headers['retry-after']) + this.client.options.restTimeOffset);
        ***REMOVED*** else if (err.status >= 500 && err.status < 600) ***REMOVED***
          if (item.retries === this.client.options.retryLimit) ***REMOVED***
            item.reject(err);
            this.handle();
          ***REMOVED*** else ***REMOVED***
            item.retries++;
            this.queue.unshift(item);
            this.resetTimeout = this.client.setTimeout(() => ***REMOVED***
              this.handle();
              this.resetTimeout = null;
            ***REMOVED***, 1e3 + this.client.options.restTimeOffset);
          ***REMOVED***
        ***REMOVED*** else ***REMOVED***
          item.reject(err.status >= 400 && err.status < 500 ?
            new DiscordAPIError(res.request.path, res.body, res.request.method) : err);
          this.handle();
        ***REMOVED***
      ***REMOVED*** else ***REMOVED***
        if (this.remaining === 0) ***REMOVED***
          if (this.client.listenerCount(RATE_LIMIT)) ***REMOVED***
            this.client.emit(RATE_LIMIT, ***REMOVED***
              limit: this.limit,
              timeDifference: this.timeDifference,
              path: item.request.path,
              method: item.request.method,
            ***REMOVED***);
          ***REMOVED***
        ***REMOVED***
        this.globalLimit = false;
        const data = res && res.body ? res.body : ***REMOVED******REMOVED***;
        item.resolve(data);
        this.handle();
      ***REMOVED***
    ***REMOVED***);
  ***REMOVED***

  handle() ***REMOVED***
    super.handle();
    if (this.queue.length === 0) return;
    if ((this.remaining <= 0 || this.globalLimit) && Date.now() - this.timeDifference < this.resetTime) return;
    this.execute(this.queue.shift());
    this.remaining--;
    this.handle();
  ***REMOVED***
***REMOVED***

module.exports = BurstRequestHandler;
