// discord.js/src/client/rest/APIRequest.js
const snekfetch = require('snekfetch');
const Constants = require('../../util/Constants');

class APIRequest ***REMOVED***
  constructor(rest, method, path, auth, data, files, reason) ***REMOVED***
    this.rest = rest;
    this.client = rest.client;
    this.method = method;
    this.path = path.toString();
    this.auth = auth;
    this.data = data;
    this.files = files;
    this.route = this.getRoute(this.path);
    this.reason = reason;
  ***REMOVED***

  getRoute(url) ***REMOVED***
    const route = url.split('?')[0].split('/');
    const routeBucket = [];
    for (let i = 0; i < route.length; i++) ***REMOVED***
      // Reactions routes and sub-routes all share the same bucket
      if (route[i - 1] === 'reactions') break;
      // Literal IDs should only be taken account if they are the Major ID (the Channel/Guild ID)
      if (/\d***REMOVED***16,19***REMOVED***/g.test(route[i]) && !/channels|guilds/.test(route[i - 1])) routeBucket.push(':id');
      // All other parts of the route should be considered as part of the bucket identifier
      else routeBucket.push(route[i]);
    ***REMOVED***
    return routeBucket.join('/');
  ***REMOVED***

  getAuth() ***REMOVED***
    if (this.client.token && this.client.user && this.client.user.bot) ***REMOVED***
      return `Bot $***REMOVED***this.client.token***REMOVED***`;
    ***REMOVED*** else if (this.client.token) ***REMOVED***
      return this.client.token;
    ***REMOVED***
    throw new Error(Constants.Errors.NO_TOKEN);
  ***REMOVED***

  gen() ***REMOVED***
    const API = `$***REMOVED***this.client.options.http.host***REMOVED***/api/v$***REMOVED***this.client.options.http.version***REMOVED***`;
    const request = snekfetch[this.method](`$***REMOVED***API***REMOVED***$***REMOVED***this.path***REMOVED***`);
    if (this.auth) request.set('Authorization', this.getAuth());
    if (this.reason) request.set('X-Audit-Log-Reason', encodeURIComponent(this.reason));
    if (!this.rest.client.browser) request.set('User-Agent', this.rest.userAgentManager.userAgent);
    if (this.files) ***REMOVED***
      for (const file of this.files) if (file && file.file) request.attach(file.name, file.file, file.name);
      if (typeof this.data !== 'undefined') request.attach('payload_json', JSON.stringify(this.data));
    ***REMOVED*** else if (this.data) ***REMOVED***
      request.send(this.data);
    ***REMOVED***
    return request;
  ***REMOVED***
***REMOVED***

module.exports = APIRequest;
