// rxjs/src/internal/operators/buffer.ts
import ***REMOVED*** Operator ***REMOVED*** from '../Operator';
import ***REMOVED*** Subscriber ***REMOVED*** from '../Subscriber';
import ***REMOVED*** Observable ***REMOVED*** from '../Observable';
import ***REMOVED*** OuterSubscriber ***REMOVED*** from '../OuterSubscriber';
import ***REMOVED*** InnerSubscriber ***REMOVED*** from '../InnerSubscriber';
import ***REMOVED*** subscribeToResult ***REMOVED*** from '../util/subscribeToResult';
import ***REMOVED*** OperatorFunction ***REMOVED*** from '../types';

/**
 * Buffers the source Observable values until `closingNotifier` emits.
 *
 * <span class="informal">Collects values from the past as an array, and emits
 * that array only when another Observable emits.</span>
 *
 * ![](buffer.png)
 *
 * Buffers the incoming Observable values until the given `closingNotifier`
 * Observable emits a value, at which point it emits the buffer on the output
 * Observable and starts a new buffer internally, awaiting the next time
 * `closingNotifier` emits.
 *
 * ## Example
 *
 * On every click, emit array of most recent interval events
 *
 * ```ts
 * import ***REMOVED*** fromEvent, interval ***REMOVED*** from 'rxjs';
 * import ***REMOVED*** buffer ***REMOVED*** from 'rxjs/operators';
 *
 * const clicks = fromEvent(document, 'click');
 * const intervalEvents = interval(1000);
 * const buffered = intervalEvents.pipe(buffer(clicks));
 * buffered.subscribe(x => console.log(x));
 * ```
 *
 * @see ***REMOVED***@link bufferCount***REMOVED***
 * @see ***REMOVED***@link bufferTime***REMOVED***
 * @see ***REMOVED***@link bufferToggle***REMOVED***
 * @see ***REMOVED***@link bufferWhen***REMOVED***
 * @see ***REMOVED***@link window***REMOVED***
 *
 * @param ***REMOVED***Observable<any>***REMOVED*** closingNotifier An Observable that signals the
 * buffer to be emitted on the output Observable.
 * @return ***REMOVED***Observable<T[]>***REMOVED*** An Observable of buffers, which are arrays of
 * values.
 * @method buffer
 * @owner Observable
 */
export function buffer<T>(closingNotifier: Observable<any>): OperatorFunction<T, T[]> ***REMOVED***
  return function bufferOperatorFunction(source: Observable<T>) ***REMOVED***
    return source.lift(new BufferOperator<T>(closingNotifier));
  ***REMOVED***;
***REMOVED***

class BufferOperator<T> implements Operator<T, T[]> ***REMOVED***

  constructor(private closingNotifier: Observable<any>) ***REMOVED***
  ***REMOVED***

  call(subscriber: Subscriber<T[]>, source: any): any ***REMOVED***
    return source.subscribe(new BufferSubscriber(subscriber, this.closingNotifier));
  ***REMOVED***
***REMOVED***

/**
 * We need this JSDoc comment for affecting ESDoc.
 * @ignore
 * @extends ***REMOVED***Ignored***REMOVED***
 */
class BufferSubscriber<T> extends OuterSubscriber<T, any> ***REMOVED***
  private buffer: T[] = [];

  constructor(destination: Subscriber<T[]>, closingNotifier: Observable<any>) ***REMOVED***
    super(destination);
    this.add(subscribeToResult(this, closingNotifier));
  ***REMOVED***

  protected _next(value: T) ***REMOVED***
    this.buffer.push(value);
  ***REMOVED***

  notifyNext(outerValue: T, innerValue: any,
             outerIndex: number, innerIndex: number,
             innerSub: InnerSubscriber<T, any>): void ***REMOVED***
    const buffer = this.buffer;
    this.buffer = [];
    this.destination.next(buffer);
  ***REMOVED***
***REMOVED***
