// rxjs/src/internal/operators/delay.ts
import ***REMOVED*** async ***REMOVED*** from '../scheduler/async';
import ***REMOVED*** isDate ***REMOVED*** from '../util/isDate';
import ***REMOVED*** Operator ***REMOVED*** from '../Operator';
import ***REMOVED*** Subscriber ***REMOVED*** from '../Subscriber';
import ***REMOVED*** Subscription ***REMOVED*** from '../Subscription';
import ***REMOVED*** Notification ***REMOVED*** from '../Notification';
import ***REMOVED*** Observable ***REMOVED*** from '../Observable';
import ***REMOVED*** MonoTypeOperatorFunction, PartialObserver, SchedulerAction, SchedulerLike, TeardownLogic ***REMOVED*** from '../types';

/**
 * Delays the emission of items from the source Observable by a given timeout or
 * until a given Date.
 *
 * <span class="informal">Time shifts each item by some specified amount of
 * milliseconds.</span>
 *
 * ![](delay.png)
 *
 * If the delay argument is a Number, this operator time shifts the source
 * Observable by that amount of time expressed in milliseconds. The relative
 * time intervals between the values are preserved.
 *
 * If the delay argument is a Date, this operator time shifts the start of the
 * Observable execution until the given date occurs.
 *
 * ## Examples
 * Delay each click by one second
 * ```ts
 * import ***REMOVED*** fromEvent ***REMOVED*** from 'rxjs';
 * import ***REMOVED*** delay ***REMOVED*** from 'rxjs/operators';
 *
 * const clicks = fromEvent(document, 'click');
 * const delayedClicks = clicks.pipe(delay(1000)); // each click emitted after 1 second
 * delayedClicks.subscribe(x => console.log(x));
 * ```
 *
 * Delay all clicks until a future date happens
 * ```ts
 * import ***REMOVED*** fromEvent ***REMOVED*** from 'rxjs';
 * import ***REMOVED*** delay ***REMOVED*** from 'rxjs/operators';
 *
 * const clicks = fromEvent(document, 'click');
 * const date = new Date('March 15, 2050 12:00:00'); // in the future
 * const delayedClicks = clicks.pipe(delay(date)); // click emitted only after that date
 * delayedClicks.subscribe(x => console.log(x));
 * ```
 *
 * @see ***REMOVED***@link debounceTime***REMOVED***
 * @see ***REMOVED***@link delayWhen***REMOVED***
 *
 * @param ***REMOVED***number|Date***REMOVED*** delay The delay duration in milliseconds (a `number`) or
 * a `Date` until which the emission of the source items is delayed.
 * @param ***REMOVED***SchedulerLike***REMOVED*** [scheduler=async] The ***REMOVED***@link SchedulerLike***REMOVED*** to use for
 * managing the timers that handle the time-shift for each item.
 * @return ***REMOVED***Observable***REMOVED*** An Observable that delays the emissions of the source
 * Observable by the specified timeout or Date.
 * @method delay
 * @owner Observable
 */
export function delay<T>(delay: number|Date,
                         scheduler: SchedulerLike = async): MonoTypeOperatorFunction<T> ***REMOVED***
  const absoluteDelay = isDate(delay);
  const delayFor = absoluteDelay ? (+delay - scheduler.now()) : Math.abs(<number>delay);
  return (source: Observable<T>) => source.lift(new DelayOperator(delayFor, scheduler));
***REMOVED***

class DelayOperator<T> implements Operator<T, T> ***REMOVED***
  constructor(private delay: number,
              private scheduler: SchedulerLike) ***REMOVED***
  ***REMOVED***

  call(subscriber: Subscriber<T>, source: any): TeardownLogic ***REMOVED***
    return source.subscribe(new DelaySubscriber(subscriber, this.delay, this.scheduler));
  ***REMOVED***
***REMOVED***

interface DelayState<T> ***REMOVED***
  source: DelaySubscriber<T>;
  destination: PartialObserver<T>;
  scheduler: SchedulerLike;
***REMOVED***

/**
 * We need this JSDoc comment for affecting ESDoc.
 * @ignore
 * @extends ***REMOVED***Ignored***REMOVED***
 */
class DelaySubscriber<T> extends Subscriber<T> ***REMOVED***
  private queue: Array<DelayMessage<T>> = [];
  private active: boolean = false;
  private errored: boolean = false;

  private static dispatch<T>(this: SchedulerAction<DelayState<T>>, state: DelayState<T>): void ***REMOVED***
    const source = state.source;
    const queue = source.queue;
    const scheduler = state.scheduler;
    const destination = state.destination;

    while (queue.length > 0 && (queue[0].time - scheduler.now()) <= 0) ***REMOVED***
      queue.shift().notification.observe(destination);
    ***REMOVED***

    if (queue.length > 0) ***REMOVED***
      const delay = Math.max(0, queue[0].time - scheduler.now());
      this.schedule(state, delay);
    ***REMOVED*** else ***REMOVED***
      this.unsubscribe();
      source.active = false;
    ***REMOVED***
  ***REMOVED***

  constructor(destination: Subscriber<T>,
              private delay: number,
              private scheduler: SchedulerLike) ***REMOVED***
    super(destination);
  ***REMOVED***

  private _schedule(scheduler: SchedulerLike): void ***REMOVED***
    this.active = true;
    const destination = this.destination as Subscription;
    destination.add(scheduler.schedule<DelayState<T>>(DelaySubscriber.dispatch, this.delay, ***REMOVED***
      source: this, destination: this.destination, scheduler: scheduler
    ***REMOVED***));
  ***REMOVED***

  private scheduleNotification(notification: Notification<T>): void ***REMOVED***
    if (this.errored === true) ***REMOVED***
      return;
    ***REMOVED***

    const scheduler = this.scheduler;
    const message = new DelayMessage(scheduler.now() + this.delay, notification);
    this.queue.push(message);

    if (this.active === false) ***REMOVED***
      this._schedule(scheduler);
    ***REMOVED***
  ***REMOVED***

  protected _next(value: T) ***REMOVED***
    this.scheduleNotification(Notification.createNext(value));
  ***REMOVED***

  protected _error(err: any) ***REMOVED***
    this.errored = true;
    this.queue = [];
    this.destination.error(err);
    this.unsubscribe();
  ***REMOVED***

  protected _complete() ***REMOVED***
    this.scheduleNotification(Notification.createComplete());
    this.unsubscribe();
  ***REMOVED***
***REMOVED***

class DelayMessage<T> ***REMOVED***
  constructor(public readonly time: number,
              public readonly notification: Notification<T>) ***REMOVED***
  ***REMOVED***
***REMOVED***
