// rxjs/src/internal/operators/distinct.ts
import ***REMOVED*** Observable ***REMOVED*** from '../Observable';
import ***REMOVED*** Operator ***REMOVED*** from '../Operator';
import ***REMOVED*** Subscriber ***REMOVED*** from '../Subscriber';
import ***REMOVED*** OuterSubscriber ***REMOVED*** from '../OuterSubscriber';
import ***REMOVED*** InnerSubscriber ***REMOVED*** from '../InnerSubscriber';
import ***REMOVED*** subscribeToResult ***REMOVED*** from '../util/subscribeToResult';
import ***REMOVED*** MonoTypeOperatorFunction, TeardownLogic ***REMOVED*** from '../types';

/**
 * Returns an Observable that emits all items emitted by the source Observable that are distinct by comparison from previous items.
 *
 * If a keySelector function is provided, then it will project each value from the source observable into a new value that it will
 * check for equality with previously projected values. If a keySelector function is not provided, it will use each value from the
 * source observable directly with an equality check against previous values.
 *
 * In JavaScript runtimes that support `Set`, this operator will use a `Set` to improve performance of the distinct value checking.
 *
 * In other runtimes, this operator will use a minimal implementation of `Set` that relies on an `Array` and `indexOf` under the
 * hood, so performance will degrade as more values are checked for distinction. Even in newer browsers, a long-running `distinct`
 * use might result in memory leaks. To help alleviate this in some scenarios, an optional `flushes` parameter is also provided so
 * that the internal `Set` can be "flushed", basically clearing it of values.
 *
 * ## Examples
 * A simple example with numbers
 * ```ts
 * import ***REMOVED*** of ***REMOVED*** from 'rxjs';
 * import ***REMOVED*** distinct ***REMOVED*** from 'rxjs/operators';
 *
 * of(1, 1, 2, 2, 2, 1, 2, 3, 4, 3, 2, 1).pipe(
 *     distinct(),
 *   )
 *   .subscribe(x => console.log(x)); // 1, 2, 3, 4
 * ```
 *
 * An example using a keySelector function
 * ```typescript
 * import ***REMOVED*** of ***REMOVED*** from 'rxjs';
 * import ***REMOVED*** distinct ***REMOVED*** from 'rxjs/operators';
 *
 * interface Person ***REMOVED***
 *    age: number,
 *    name: string
 * ***REMOVED***
 *
 * of<Person>(
 *     ***REMOVED*** age: 4, name: 'Foo'***REMOVED***,
 *     ***REMOVED*** age: 7, name: 'Bar'***REMOVED***,
 *     ***REMOVED*** age: 5, name: 'Foo'***REMOVED***,
 *   ).pipe(
 *     distinct((p: Person) => p.name),
 *   )
 *   .subscribe(x => console.log(x));
 *
 * // displays:
 * // ***REMOVED*** age: 4, name: 'Foo' ***REMOVED***
 * // ***REMOVED*** age: 7, name: 'Bar' ***REMOVED***
 * ```
 * @see ***REMOVED***@link distinctUntilChanged***REMOVED***
 * @see ***REMOVED***@link distinctUntilKeyChanged***REMOVED***
 *
 * @param ***REMOVED***function***REMOVED*** [keySelector] Optional function to select which value you want to check as distinct.
 * @param ***REMOVED***Observable***REMOVED*** [flushes] Optional Observable for flushing the internal HashSet of the operator.
 * @return ***REMOVED***Observable***REMOVED*** An Observable that emits items from the source Observable with distinct values.
 * @method distinct
 * @owner Observable
 */
export function distinct<T, K>(keySelector?: (value: T) => K,
                               flushes?: Observable<any>): MonoTypeOperatorFunction<T> ***REMOVED***
  return (source: Observable<T>) => source.lift(new DistinctOperator(keySelector, flushes));
***REMOVED***

class DistinctOperator<T, K> implements Operator<T, T> ***REMOVED***
  constructor(private keySelector: (value: T) => K, private flushes: Observable<any>) ***REMOVED***
  ***REMOVED***

  call(subscriber: Subscriber<T>, source: any): TeardownLogic ***REMOVED***
    return source.subscribe(new DistinctSubscriber(subscriber, this.keySelector, this.flushes));
  ***REMOVED***
***REMOVED***

/**
 * We need this JSDoc comment for affecting ESDoc.
 * @ignore
 * @extends ***REMOVED***Ignored***REMOVED***
 */
export class DistinctSubscriber<T, K> extends OuterSubscriber<T, T> ***REMOVED***
  private values = new Set<K>();

  constructor(destination: Subscriber<T>, private keySelector: (value: T) => K, flushes: Observable<any>) ***REMOVED***
    super(destination);

    if (flushes) ***REMOVED***
      this.add(subscribeToResult(this, flushes));
    ***REMOVED***
  ***REMOVED***

  notifyNext(outerValue: T, innerValue: T,
             outerIndex: number, innerIndex: number,
             innerSub: InnerSubscriber<T, T>): void ***REMOVED***
    this.values.clear();
  ***REMOVED***

  notifyError(error: any, innerSub: InnerSubscriber<T, T>): void ***REMOVED***
    this._error(error);
  ***REMOVED***

  protected _next(value: T): void ***REMOVED***
    if (this.keySelector) ***REMOVED***
      this._useKeySelector(value);
    ***REMOVED*** else ***REMOVED***
      this._finalizeNext(value, value);
    ***REMOVED***
  ***REMOVED***

  private _useKeySelector(value: T): void ***REMOVED***
    let key: K;
    const ***REMOVED*** destination ***REMOVED*** = this;
    try ***REMOVED***
      key = this.keySelector(value);
    ***REMOVED*** catch (err) ***REMOVED***
      destination.error(err);
      return;
    ***REMOVED***
    this._finalizeNext(key, value);
  ***REMOVED***

  private _finalizeNext(key: K|T, value: T) ***REMOVED***
    const ***REMOVED*** values ***REMOVED*** = this;
    if (!values.has(<K>key)) ***REMOVED***
      values.add(<K>key);
      this.destination.next(value);
    ***REMOVED***
  ***REMOVED***

***REMOVED***
