// rxjs/src/internal/operators/mergeMap.ts
import ***REMOVED*** Observable ***REMOVED*** from '../Observable';
import ***REMOVED*** Operator ***REMOVED*** from '../Operator';
import ***REMOVED*** Subscriber ***REMOVED*** from '../Subscriber';
import ***REMOVED*** Subscription ***REMOVED*** from '../Subscription';
import ***REMOVED*** subscribeToResult ***REMOVED*** from '../util/subscribeToResult';
import ***REMOVED*** OuterSubscriber ***REMOVED*** from '../OuterSubscriber';
import ***REMOVED*** InnerSubscriber ***REMOVED*** from '../InnerSubscriber';
import ***REMOVED*** ObservableInput, OperatorFunction, ObservedValueOf ***REMOVED*** from '../types';
import ***REMOVED*** map ***REMOVED*** from './map';
import ***REMOVED*** from ***REMOVED*** from '../observable/from';

/* tslint:disable:max-line-length */
export function mergeMap<T, O extends ObservableInput<any>>(project: (value: T, index: number) => O, concurrent?: number): OperatorFunction<T, ObservedValueOf<O>>;
/** @deprecated resultSelector no longer supported, use inner map instead */
export function mergeMap<T, O extends ObservableInput<any>>(project: (value: T, index: number) => O, resultSelector: undefined, concurrent?: number): OperatorFunction<T, ObservedValueOf<O>>;
/** @deprecated resultSelector no longer supported, use inner map instead */
export function mergeMap<T, R, O extends ObservableInput<any>>(project: (value: T, index: number) => O, resultSelector: (outerValue: T, innerValue: ObservedValueOf<O>, outerIndex: number, innerIndex: number) => R, concurrent?: number): OperatorFunction<T, R>;
/* tslint:enable:max-line-length */

/**
 * Projects each source value to an Observable which is merged in the output
 * Observable.
 *
 * <span class="informal">Maps each value to an Observable, then flattens all of
 * these inner Observables using ***REMOVED***@link mergeAll***REMOVED***.</span>
 *
 * ![](mergeMap.png)
 *
 * Returns an Observable that emits items based on applying a function that you
 * supply to each item emitted by the source Observable, where that function
 * returns an Observable, and then merging those resulting Observables and
 * emitting the results of this merger.
 *
 * ## Example
 * Map and flatten each letter to an Observable ticking every 1 second
 * ```ts
 * import ***REMOVED*** of, interval ***REMOVED*** from 'rxjs';
 * import ***REMOVED*** mergeMap, map ***REMOVED*** from 'rxjs/operators';
 *
 * const letters = of('a', 'b', 'c');
 * const result = letters.pipe(
 *   mergeMap(x => interval(1000).pipe(map(i => x+i))),
 * );
 * result.subscribe(x => console.log(x));
 *
 * // Results in the following:
 * // a0
 * // b0
 * // c0
 * // a1
 * // b1
 * // c1
 * // continues to list a,b,c with respective ascending integers
 * ```
 *
 * @see ***REMOVED***@link concatMap***REMOVED***
 * @see ***REMOVED***@link exhaustMap***REMOVED***
 * @see ***REMOVED***@link merge***REMOVED***
 * @see ***REMOVED***@link mergeAll***REMOVED***
 * @see ***REMOVED***@link mergeMapTo***REMOVED***
 * @see ***REMOVED***@link mergeScan***REMOVED***
 * @see ***REMOVED***@link switchMap***REMOVED***
 *
 * @param ***REMOVED***function(value: T, ?index: number): ObservableInput***REMOVED*** project A function
 * that, when applied to an item emitted by the source Observable, returns an
 * Observable.
 * @param ***REMOVED***number***REMOVED*** [concurrent=Number.POSITIVE_INFINITY] Maximum number of input
 * Observables being subscribed to concurrently.
 * @return ***REMOVED***Observable***REMOVED*** An Observable that emits the result of applying the
 * projection function (and the optional deprecated `resultSelector`) to each item
 * emitted by the source Observable and merging the results of the Observables
 * obtained from this transformation.
 * @method mergeMap
 * @owner Observable
 */
export function mergeMap<T, R, O extends ObservableInput<any>>(
  project: (value: T, index: number) => O,
  resultSelector?: ((outerValue: T, innerValue: ObservedValueOf<O>, outerIndex: number, innerIndex: number) => R) | number,
  concurrent: number = Number.POSITIVE_INFINITY
): OperatorFunction<T, ObservedValueOf<O>|R> ***REMOVED***
  if (typeof resultSelector === 'function') ***REMOVED***
    // DEPRECATED PATH
    return (source: Observable<T>) => source.pipe(
      mergeMap((a, i) => from(project(a, i)).pipe(
        map((b: any, ii: number) => resultSelector(a, b, i, ii)),
      ), concurrent)
    );
  ***REMOVED*** else if (typeof resultSelector === 'number') ***REMOVED***
    concurrent = resultSelector;
  ***REMOVED***
  return (source: Observable<T>) => source.lift(new MergeMapOperator(project, concurrent));
***REMOVED***

export class MergeMapOperator<T, R> implements Operator<T, R> ***REMOVED***
  constructor(private project: (value: T, index: number) => ObservableInput<R>,
              private concurrent: number = Number.POSITIVE_INFINITY) ***REMOVED***
  ***REMOVED***

  call(observer: Subscriber<R>, source: any): any ***REMOVED***
    return source.subscribe(new MergeMapSubscriber(
      observer, this.project, this.concurrent
    ));
  ***REMOVED***
***REMOVED***

/**
 * We need this JSDoc comment for affecting ESDoc.
 * @ignore
 * @extends ***REMOVED***Ignored***REMOVED***
 */
export class MergeMapSubscriber<T, R> extends OuterSubscriber<T, R> ***REMOVED***
  private hasCompleted: boolean = false;
  private buffer: T[] = [];
  private active: number = 0;
  protected index: number = 0;

  constructor(destination: Subscriber<R>,
              private project: (value: T, index: number) => ObservableInput<R>,
              private concurrent: number = Number.POSITIVE_INFINITY) ***REMOVED***
    super(destination);
  ***REMOVED***

  protected _next(value: T): void ***REMOVED***
    if (this.active < this.concurrent) ***REMOVED***
      this._tryNext(value);
    ***REMOVED*** else ***REMOVED***
      this.buffer.push(value);
    ***REMOVED***
  ***REMOVED***

  protected _tryNext(value: T) ***REMOVED***
    let result: ObservableInput<R>;
    const index = this.index++;
    try ***REMOVED***
      result = this.project(value, index);
    ***REMOVED*** catch (err) ***REMOVED***
      this.destination.error(err);
      return;
    ***REMOVED***
    this.active++;
    this._innerSub(result, value, index);
  ***REMOVED***

  private _innerSub(ish: ObservableInput<R>, value: T, index: number): void ***REMOVED***
    const innerSubscriber = new InnerSubscriber(this, value, index);
    const destination = this.destination as Subscription;
    destination.add(innerSubscriber);
    const innerSubscription = subscribeToResult<T, R>(this, ish, undefined, undefined, innerSubscriber);
    // The returned subscription will usually be the subscriber that was
    // passed. However, interop subscribers will be wrapped and for
    // unsubscriptions to chain correctly, the wrapper needs to be added, too.
    if (innerSubscription !== innerSubscriber) ***REMOVED***
      destination.add(innerSubscription);
    ***REMOVED***
  ***REMOVED***

  protected _complete(): void ***REMOVED***
    this.hasCompleted = true;
    if (this.active === 0 && this.buffer.length === 0) ***REMOVED***
      this.destination.complete();
    ***REMOVED***
    this.unsubscribe();
  ***REMOVED***

  notifyNext(outerValue: T, innerValue: R,
             outerIndex: number, innerIndex: number,
             innerSub: InnerSubscriber<T, R>): void ***REMOVED***
    this.destination.next(innerValue);
  ***REMOVED***

  notifyComplete(innerSub: Subscription): void ***REMOVED***
    const buffer = this.buffer;
    this.remove(innerSub);
    this.active--;
    if (buffer.length > 0) ***REMOVED***
      this._next(buffer.shift());
    ***REMOVED*** else if (this.active === 0 && this.hasCompleted) ***REMOVED***
      this.destination.complete();
    ***REMOVED***
  ***REMOVED***
***REMOVED***
