// rxjs/src/internal/operators/skipUntil.ts
import ***REMOVED*** Operator ***REMOVED*** from '../Operator';
import ***REMOVED*** Subscriber ***REMOVED*** from '../Subscriber';
import ***REMOVED*** Observable ***REMOVED*** from '../Observable';
import ***REMOVED*** OuterSubscriber ***REMOVED*** from '../OuterSubscriber';
import ***REMOVED*** InnerSubscriber ***REMOVED*** from '../InnerSubscriber';
import ***REMOVED*** subscribeToResult ***REMOVED*** from '../util/subscribeToResult';
import ***REMOVED*** MonoTypeOperatorFunction, TeardownLogic, ObservableInput ***REMOVED*** from '../types';
import ***REMOVED*** Subscription ***REMOVED*** from '../Subscription';

/**
 * Returns an Observable that skips items emitted by the source Observable until a second Observable emits an item.
 *
 * The `skipUntil` operator causes the observable stream to skip the emission of values ​​until the passed in observable emits the first value.
 * This can be particularly useful in combination with user interactions, responses of http requests or waiting for specific times to pass by.
 *
 * ![](skipUntil.png)
 *
 * Internally the `skipUntil` operator subscribes to the passed in observable (in the following called *notifier*) in order to recognize the emission
 * of its first value. When this happens, the operator unsubscribes from the *notifier* and starts emitting the values of the *source*
 * observable. It will never let the *source* observable emit any values if the *notifier* completes or throws an error without emitting
 * a value before.
 *
 * ## Example
 *
 * In the following example, all emitted values ​​of the interval observable are skipped until the user clicks anywhere within the page.
 *
 * ```ts
 * import ***REMOVED*** interval, fromEvent ***REMOVED*** from 'rxjs';
 * import ***REMOVED*** skipUntil ***REMOVED*** from 'rxjs/operators';
 *
 * const intervalObservable = interval(1000);
 * const click = fromEvent(document, 'click');
 *
 * const emitAfterClick = intervalObservable.pipe(
 *   skipUntil(click)
 * );
 * // clicked at 4.6s. output: 5...6...7...8........ or
 * // clicked at 7.3s. output: 8...9...10..11.......
 * const subscribe = emitAfterClick.subscribe(value => console.log(value));
 * ```
 *
 * @param ***REMOVED***Observable***REMOVED*** notifier - The second Observable that has to emit an item before the source Observable's elements begin to
 * be mirrored by the resulting Observable.
 * @return ***REMOVED***Observable<T>***REMOVED*** An Observable that skips items from the source Observable until the second Observable emits
 * an item, then emits the remaining items.
 * @method skipUntil
 * @owner Observable
 */
export function skipUntil<T>(notifier: Observable<any>): MonoTypeOperatorFunction<T> ***REMOVED***
  return (source: Observable<T>) => source.lift(new SkipUntilOperator(notifier));
***REMOVED***

class SkipUntilOperator<T> implements Operator<T, T> ***REMOVED***
  constructor(private notifier: Observable<any>) ***REMOVED***
  ***REMOVED***

  call(destination: Subscriber<T>, source: any): TeardownLogic ***REMOVED***
    return source.subscribe(new SkipUntilSubscriber(destination, this.notifier));
  ***REMOVED***
***REMOVED***

/**
 * We need this JSDoc comment for affecting ESDoc.
 * @ignore
 * @extends ***REMOVED***Ignored***REMOVED***
 */
class SkipUntilSubscriber<T, R> extends OuterSubscriber<T, R> ***REMOVED***

  private hasValue: boolean = false;
  private innerSubscription: Subscription;

  constructor(destination: Subscriber<R>, notifier: ObservableInput<any>) ***REMOVED***
    super(destination);
    const innerSubscriber = new InnerSubscriber(this, undefined, undefined);
    this.add(innerSubscriber);
    this.innerSubscription = innerSubscriber;
    const innerSubscription = subscribeToResult(this, notifier, undefined, undefined, innerSubscriber);
    // The returned subscription will usually be the subscriber that was
    // passed. However, interop subscribers will be wrapped and for
    // unsubscriptions to chain correctly, the wrapper needs to be added, too.
    if (innerSubscription !== innerSubscriber) ***REMOVED***
      this.add(innerSubscription);
      this.innerSubscription = innerSubscription;
    ***REMOVED***
  ***REMOVED***

  protected _next(value: T) ***REMOVED***
    if (this.hasValue) ***REMOVED***
      super._next(value);
    ***REMOVED***
  ***REMOVED***

  notifyNext(outerValue: T, innerValue: R,
             outerIndex: number, innerIndex: number,
             innerSub: InnerSubscriber<T, R>): void ***REMOVED***
    this.hasValue = true;
    if (this.innerSubscription) ***REMOVED***
      this.innerSubscription.unsubscribe();
    ***REMOVED***
  ***REMOVED***

  notifyComplete() ***REMOVED***
    /* do nothing */
  ***REMOVED***
***REMOVED***
