// rxjs/src/internal/operators/switchMap.ts
import ***REMOVED*** Operator ***REMOVED*** from '../Operator';
import ***REMOVED*** Observable ***REMOVED*** from '../Observable';
import ***REMOVED*** Subscriber ***REMOVED*** from '../Subscriber';
import ***REMOVED*** Subscription ***REMOVED*** from '../Subscription';
import ***REMOVED*** OuterSubscriber ***REMOVED*** from '../OuterSubscriber';
import ***REMOVED*** InnerSubscriber ***REMOVED*** from '../InnerSubscriber';
import ***REMOVED*** subscribeToResult ***REMOVED*** from '../util/subscribeToResult';
import ***REMOVED*** ObservableInput, OperatorFunction, ObservedValueOf ***REMOVED*** from '../types';
import ***REMOVED*** map ***REMOVED*** from './map';
import ***REMOVED*** from ***REMOVED*** from '../observable/from';

/* tslint:disable:max-line-length */
export function switchMap<T, O extends ObservableInput<any>>(project: (value: T, index: number) => O): OperatorFunction<T, ObservedValueOf<O>>;
/** @deprecated resultSelector is no longer supported, use inner map instead */
export function switchMap<T, O extends ObservableInput<any>>(project: (value: T, index: number) => O, resultSelector: undefined): OperatorFunction<T, ObservedValueOf<O>>;
/** @deprecated resultSelector is no longer supported, use inner map instead */
export function switchMap<T, R, O extends ObservableInput<any>>(project: (value: T, index: number) => O, resultSelector: (outerValue: T, innerValue: ObservedValueOf<O>, outerIndex: number, innerIndex: number) => R): OperatorFunction<T, R>;
/* tslint:enable:max-line-length */

/**
 * Projects each source value to an Observable which is merged in the output
 * Observable, emitting values only from the most recently projected Observable.
 *
 * <span class="informal">Maps each value to an Observable, then flattens all of
 * these inner Observables.</span>
 *
 * ![](switchMap.png)
 *
 * Returns an Observable that emits items based on applying a function that you
 * supply to each item emitted by the source Observable, where that function
 * returns an (so-called "inner") Observable. Each time it observes one of these
 * inner Observables, the output Observable begins emitting the items emitted by
 * that inner Observable. When a new inner Observable is emitted, `switchMap`
 * stops emitting items from the earlier-emitted inner Observable and begins
 * emitting items from the new one. It continues to behave like this for
 * subsequent inner Observables.
 *
 * ## Example
 * Generate new Observable according to source Observable values
 * ```typescript
 * import ***REMOVED*** of ***REMOVED*** from 'rxjs';
 * import ***REMOVED*** switchMap ***REMOVED*** from 'rxjs/operators';
 *
 * const switched = of(1, 2, 3).pipe(switchMap((x: number) => of(x, x ** 2, x ** 3)));
 * switched.subscribe(x => console.log(x));
 * // outputs
 * // 1
 * // 1
 * // 1
 * // 2
 * // 4
 * // 8
 * // ... and so on
 * ```
 *
 * Rerun an interval Observable on every click event
 * ```ts
 * import ***REMOVED*** fromEvent, interval ***REMOVED*** from 'rxjs';
 * import ***REMOVED*** switchMap ***REMOVED*** from 'rxjs/operators';
 *
 * const clicks = fromEvent(document, 'click');
 * const result = clicks.pipe(switchMap((ev) => interval(1000)));
 * result.subscribe(x => console.log(x));
 * ```
 *
 * @see ***REMOVED***@link concatMap***REMOVED***
 * @see ***REMOVED***@link exhaustMap***REMOVED***
 * @see ***REMOVED***@link mergeMap***REMOVED***
 * @see ***REMOVED***@link switchAll***REMOVED***
 * @see ***REMOVED***@link switchMapTo***REMOVED***
 *
 * @param ***REMOVED***function(value: T, ?index: number): ObservableInput***REMOVED*** project A function
 * that, when applied to an item emitted by the source Observable, returns an
 * Observable.
 * @return ***REMOVED***Observable***REMOVED*** An Observable that emits the result of applying the
 * projection function (and the optional deprecated `resultSelector`) to each item
 * emitted by the source Observable and taking only the values from the most recently
 * projected inner Observable.
 * @method switchMap
 * @owner Observable
 */
export function switchMap<T, R, O extends ObservableInput<any>>(
  project: (value: T, index: number) => O,
  resultSelector?: (outerValue: T, innerValue: ObservedValueOf<O>, outerIndex: number, innerIndex: number) => R,
): OperatorFunction<T, ObservedValueOf<O>|R> ***REMOVED***
  if (typeof resultSelector === 'function') ***REMOVED***
    return (source: Observable<T>) => source.pipe(
      switchMap((a, i) => from(project(a, i)).pipe(
        map((b, ii) => resultSelector(a, b, i, ii))
      ))
    );
  ***REMOVED***
  return (source: Observable<T>) => source.lift(new SwitchMapOperator(project));
***REMOVED***

class SwitchMapOperator<T, R> implements Operator<T, R> ***REMOVED***
  constructor(private project: (value: T, index: number) => ObservableInput<R>) ***REMOVED***
  ***REMOVED***

  call(subscriber: Subscriber<R>, source: any): any ***REMOVED***
    return source.subscribe(new SwitchMapSubscriber(subscriber, this.project));
  ***REMOVED***
***REMOVED***

/**
 * We need this JSDoc comment for affecting ESDoc.
 * @ignore
 * @extends ***REMOVED***Ignored***REMOVED***
 */
class SwitchMapSubscriber<T, R> extends OuterSubscriber<T, R> ***REMOVED***
  private index: number = 0;
  private innerSubscription: Subscription;

  constructor(destination: Subscriber<R>,
              private project: (value: T, index: number) => ObservableInput<R>) ***REMOVED***
    super(destination);
  ***REMOVED***

  protected _next(value: T) ***REMOVED***
    let result: ObservableInput<R>;
    const index = this.index++;
    try ***REMOVED***
      result = this.project(value, index);
    ***REMOVED*** catch (error) ***REMOVED***
      this.destination.error(error);
      return;
    ***REMOVED***
    this._innerSub(result, value, index);
  ***REMOVED***

  private _innerSub(result: ObservableInput<R>, value: T, index: number) ***REMOVED***
    const innerSubscription = this.innerSubscription;
    if (innerSubscription) ***REMOVED***
      innerSubscription.unsubscribe();
    ***REMOVED***
    const innerSubscriber = new InnerSubscriber(this, value, index);
    const destination = this.destination as Subscription;
    destination.add(innerSubscriber);
    this.innerSubscription = subscribeToResult(this, result, undefined, undefined, innerSubscriber);
    // The returned subscription will usually be the subscriber that was
    // passed. However, interop subscribers will be wrapped and for
    // unsubscriptions to chain correctly, the wrapper needs to be added, too.
    if (this.innerSubscription !== innerSubscriber) ***REMOVED***
      destination.add(this.innerSubscription);
    ***REMOVED***
  ***REMOVED***

  protected _complete(): void ***REMOVED***
    const ***REMOVED***innerSubscription***REMOVED*** = this;
    if (!innerSubscription || innerSubscription.closed) ***REMOVED***
      super._complete();
    ***REMOVED***
    this.unsubscribe();
  ***REMOVED***

  protected _unsubscribe() ***REMOVED***
    this.innerSubscription = null;
  ***REMOVED***

  notifyComplete(innerSub: Subscription): void ***REMOVED***
    const destination = this.destination as Subscription;
    destination.remove(innerSub);
    this.innerSubscription = null;
    if (this.isStopped) ***REMOVED***
      super._complete();
    ***REMOVED***
  ***REMOVED***

  notifyNext(outerValue: T, innerValue: R,
             outerIndex: number, innerIndex: number,
             innerSub: InnerSubscriber<T, R>): void ***REMOVED***
      this.destination.next(innerValue);
  ***REMOVED***
***REMOVED***
