// rxjs/src/internal/operators/tap.ts
import ***REMOVED*** Operator ***REMOVED*** from '../Operator';
import ***REMOVED*** Subscriber ***REMOVED*** from '../Subscriber';
import ***REMOVED*** Observable ***REMOVED*** from '../Observable';
import ***REMOVED*** MonoTypeOperatorFunction, PartialObserver, TeardownLogic ***REMOVED*** from '../types';
import ***REMOVED*** noop ***REMOVED*** from '../util/noop';
import ***REMOVED*** isFunction ***REMOVED*** from '../util/isFunction';

/* tslint:disable:max-line-length */
/** @deprecated Use an observer instead of a complete callback */
export function tap<T>(next: null | undefined, error: null | undefined, complete: () => void): MonoTypeOperatorFunction<T>;
/** @deprecated Use an observer instead of an error callback */
export function tap<T>(next: null | undefined, error: (error: any) => void, complete?: () => void): MonoTypeOperatorFunction<T>;
/** @deprecated Use an observer instead of a complete callback */
export function tap<T>(next: (value: T) => void, error: null | undefined, complete: () => void): MonoTypeOperatorFunction<T>;
export function tap<T>(next?: (x: T) => void, error?: (e: any) => void, complete?: () => void): MonoTypeOperatorFunction<T>;
export function tap<T>(observer: PartialObserver<T>): MonoTypeOperatorFunction<T>;
/* tslint:enable:max-line-length */

/**
 * Perform a side effect for every emission on the source Observable, but return
 * an Observable that is identical to the source.
 *
 * <span class="informal">Intercepts each emission on the source and runs a
 * function, but returns an output which is identical to the source as long as errors don't occur.</span>
 *
 * ![](do.png)
 *
 * Returns a mirrored Observable of the source Observable, but modified so that
 * the provided Observer is called to perform a side effect for every value,
 * error, and completion emitted by the source. Any errors that are thrown in
 * the aforementioned Observer or handlers are safely sent down the error path
 * of the output Observable.
 *
 * This operator is useful for debugging your Observables for the correct values
 * or performing other side effects.
 *
 * Note: this is different to a `subscribe` on the Observable. If the Observable
 * returned by `tap` is not subscribed, the side effects specified by the
 * Observer will never happen. `tap` therefore simply spies on existing
 * execution, it does not trigger an execution to happen like `subscribe` does.
 *
 * ## Example
 * Map every click to the clientX position of that click, while also logging the click event
 * ```ts
 * import ***REMOVED*** fromEvent ***REMOVED*** from 'rxjs';
 * import ***REMOVED*** tap, map ***REMOVED*** from 'rxjs/operators';
 *
 * const clicks = fromEvent(document, 'click');
 * const positions = clicks.pipe(
 *   tap(ev => console.log(ev)),
 *   map(ev => ev.clientX),
 * );
 * positions.subscribe(x => console.log(x));
 * ```
 *
 * @see ***REMOVED***@link map***REMOVED***
 * @see ***REMOVED***@link Observable#subscribe***REMOVED***
 *
 * @param ***REMOVED***Observer|function***REMOVED*** [nextOrObserver] A normal Observer object or a
 * callback for `next`.
 * @param ***REMOVED***function***REMOVED*** [error] Callback for errors in the source.
 * @param ***REMOVED***function***REMOVED*** [complete] Callback for the completion of the source.
 * @return ***REMOVED***Observable***REMOVED*** An Observable identical to the source, but runs the
 * specified Observer or callback(s) for each item.
 * @name tap
 */
export function tap<T>(nextOrObserver?: PartialObserver<T> | ((x: T) => void),
                       error?: (e: any) => void,
                       complete?: () => void): MonoTypeOperatorFunction<T> ***REMOVED***
  return function tapOperatorFunction(source: Observable<T>): Observable<T> ***REMOVED***
    return source.lift(new DoOperator(nextOrObserver, error, complete));
  ***REMOVED***;
***REMOVED***

class DoOperator<T> implements Operator<T, T> ***REMOVED***
  constructor(private nextOrObserver?: PartialObserver<T> | ((x: T) => void),
              private error?: (e: any) => void,
              private complete?: () => void) ***REMOVED***
  ***REMOVED***
  call(subscriber: Subscriber<T>, source: any): TeardownLogic ***REMOVED***
    return source.subscribe(new TapSubscriber(subscriber, this.nextOrObserver, this.error, this.complete));
  ***REMOVED***
***REMOVED***

/**
 * We need this JSDoc comment for affecting ESDoc.
 * @ignore
 * @extends ***REMOVED***Ignored***REMOVED***
 */

class TapSubscriber<T> extends Subscriber<T> ***REMOVED***
  private _context: any;

  private _tapNext: ((value: T) => void) = noop;

  private _tapError: ((err: any) => void) = noop;

  private _tapComplete: (() => void) = noop;

  constructor(destination: Subscriber<T>,
              observerOrNext?: PartialObserver<T> | ((value: T) => void),
              error?: (e?: any) => void,
              complete?: () => void) ***REMOVED***
      super(destination);
      this._tapError = error || noop;
      this._tapComplete = complete || noop;
      if (isFunction(observerOrNext)) ***REMOVED***
        this._context = this;
        this._tapNext = observerOrNext;
      ***REMOVED*** else if (observerOrNext) ***REMOVED***
        this._context = observerOrNext;
        this._tapNext = observerOrNext.next || noop;
        this._tapError = observerOrNext.error || noop;
        this._tapComplete = observerOrNext.complete || noop;
      ***REMOVED***
    ***REMOVED***

  _next(value: T) ***REMOVED***
    try ***REMOVED***
      this._tapNext.call(this._context, value);
    ***REMOVED*** catch (err) ***REMOVED***
      this.destination.error(err);
      return;
    ***REMOVED***
    this.destination.next(value);
  ***REMOVED***

  _error(err: any) ***REMOVED***
    try ***REMOVED***
      this._tapError.call(this._context, err);
    ***REMOVED*** catch (err) ***REMOVED***
      this.destination.error(err);
      return;
    ***REMOVED***
    this.destination.error(err);
  ***REMOVED***

  _complete() ***REMOVED***
    try ***REMOVED***
      this._tapComplete.call(this._context, );
    ***REMOVED*** catch (err) ***REMOVED***
      this.destination.error(err);
      return;
    ***REMOVED***
    return this.destination.complete();
  ***REMOVED***
***REMOVED***
