// eslint-plugin-promise/rules/no-native.js
// Borrowed from here:
// https://github.com/colonyamerican/eslint-plugin-cah/issues/3

'use strict'

const getDocsUrl = require('./lib/get-docs-url')

function isDeclared(scope, ref) ***REMOVED***
  return scope.variables.some(variable => ***REMOVED***
    if (variable.name !== ref.identifier.name) ***REMOVED***
      return false
    ***REMOVED***

    if (!variable.defs || !variable.defs.length) ***REMOVED***
      return false
    ***REMOVED***

    return true
  ***REMOVED***)
***REMOVED***

module.exports = ***REMOVED***
  meta: ***REMOVED***
    docs: ***REMOVED***
      url: getDocsUrl('no-native')
    ***REMOVED***,
    messages: ***REMOVED***
      name: '"***REMOVED******REMOVED***name***REMOVED******REMOVED***" is not defined.'
    ***REMOVED***
  ***REMOVED***,
  create(context) ***REMOVED***
    /**
     * Checks for and reports reassigned constants
     *
     * @param ***REMOVED***Scope***REMOVED*** scope - an escope Scope object
     * @returns ***REMOVED***void***REMOVED***
     * @private
     */
    return ***REMOVED***
      'Program:exit'() ***REMOVED***
        const scope = context.getScope()

        scope.implicit.left.forEach(ref => ***REMOVED***
          if (ref.identifier.name !== 'Promise') ***REMOVED***
            return
          ***REMOVED***

          if (!isDeclared(scope, ref)) ***REMOVED***
            context.report(***REMOVED***
              node: ref.identifier,
              messageId: 'name',
              data: ***REMOVED*** name: ref.identifier.name ***REMOVED***
            ***REMOVED***)
          ***REMOVED***
        ***REMOVED***)
      ***REMOVED***
    ***REMOVED***
  ***REMOVED***
***REMOVED***
