// eslint-plugin-promise/rules/no-new-statics.js
'use strict'

const PROMISE_STATICS = require('./lib/promise-statics')
const getDocsUrl = require('./lib/get-docs-url')

module.exports = ***REMOVED***
  meta: ***REMOVED***
    docs: ***REMOVED***
      url: getDocsUrl('no-new-statics')
    ***REMOVED***,
    fixable: 'code'
  ***REMOVED***,
  create(context) ***REMOVED***
    return ***REMOVED***
      NewExpression(node) ***REMOVED***
        if (
          node.callee.type === 'MemberExpression' &&
          node.callee.object.name === 'Promise' &&
          PROMISE_STATICS[node.callee.property.name]
        ) ***REMOVED***
          context.report(***REMOVED***
            node,
            message: "Avoid calling 'new' on 'Promise.***REMOVED******REMOVED*** name ***REMOVED******REMOVED***()'",
            data: ***REMOVED*** name: node.callee.property.name ***REMOVED***,
            fix(fixer) ***REMOVED***
              return fixer.replaceTextRange(
                [node.start, node.start + 'new '.length],
                ''
              )
            ***REMOVED***
          ***REMOVED***)
        ***REMOVED***
      ***REMOVED***
    ***REMOVED***
  ***REMOVED***
***REMOVED***
