// eslint-plugin-promise/rules/prefer-await-to-then.js
/**
 * Rule: prefer-await-to-then
 * Discourage using then() and instead use async/await.
 */

'use strict'

const getDocsUrl = require('./lib/get-docs-url')

module.exports = ***REMOVED***
  meta: ***REMOVED***
    docs: ***REMOVED***
      url: getDocsUrl('prefer-await-to-then')
    ***REMOVED***
  ***REMOVED***,
  create(context) ***REMOVED***
    /** Returns true if node is inside yield or await expression. */
    function isInsideYieldOrAwait() ***REMOVED***
      return context.getAncestors().some(parent => ***REMOVED***
        return (
          parent.type === 'AwaitExpression' || parent.type === 'YieldExpression'
        )
      ***REMOVED***)
    ***REMOVED***

    /**
     * Returns true if node is created at the top-level scope.
     * Await statements are not allowed at the top level,
     * only within function declarations.
     */
    function isTopLevelScoped() ***REMOVED***
      return context.getScope().block.type === 'Program'
    ***REMOVED***

    return ***REMOVED***
      MemberExpression(node) ***REMOVED***
        if (isTopLevelScoped() || isInsideYieldOrAwait()) ***REMOVED***
          return
        ***REMOVED***

        // if you're a then expression then you're probably a promise
        if (node.property && node.property.name === 'then') ***REMOVED***
          context.report(***REMOVED***
            node: node.property,
            message: 'Prefer await to then().'
          ***REMOVED***)
        ***REMOVED***
      ***REMOVED***
    ***REMOVED***
  ***REMOVED***
***REMOVED***
