// cross-spawn/lib/enoent.js
'use strict';

const isWin = process.platform === 'win32';

function notFoundError(original, syscall) ***REMOVED***
    return Object.assign(new Error(`$***REMOVED***syscall***REMOVED*** $***REMOVED***original.command***REMOVED*** ENOENT`), ***REMOVED***
        code: 'ENOENT',
        errno: 'ENOENT',
        syscall: `$***REMOVED***syscall***REMOVED*** $***REMOVED***original.command***REMOVED***`,
        path: original.command,
        spawnargs: original.args,
    ***REMOVED***);
***REMOVED***

function hookChildProcess(cp, parsed) ***REMOVED***
    if (!isWin) ***REMOVED***
        return;
    ***REMOVED***

    const originalEmit = cp.emit;

    cp.emit = function (name, arg1) ***REMOVED***
        // If emitting "exit" event and exit code is 1, we need to check if
        // the command exists and emit an "error" instead
        // See https://github.com/IndigoUnited/node-cross-spawn/issues/16
        if (name === 'exit') ***REMOVED***
            const err = verifyENOENT(arg1, parsed, 'spawn');

            if (err) ***REMOVED***
                return originalEmit.call(cp, 'error', err);
            ***REMOVED***
        ***REMOVED***

        return originalEmit.apply(cp, arguments); // eslint-disable-line prefer-rest-params
    ***REMOVED***;
***REMOVED***

function verifyENOENT(status, parsed) ***REMOVED***
    if (isWin && status === 1 && !parsed.file) ***REMOVED***
        return notFoundError(parsed.original, 'spawn');
    ***REMOVED***

    return null;
***REMOVED***

function verifyENOENTSync(status, parsed) ***REMOVED***
    if (isWin && status === 1 && !parsed.file) ***REMOVED***
        return notFoundError(parsed.original, 'spawnSync');
    ***REMOVED***

    return null;
***REMOVED***

module.exports = ***REMOVED***
    hookChildProcess,
    verifyENOENT,
    verifyENOENTSync,
    notFoundError,
***REMOVED***;
