// inquirer/lib/ui/baseUI.js
'use strict';
var _ = require('lodash');
var MuteStream = require('mute-stream');
var readline = require('readline');

/**
 * Base interface class other can inherits from
 */

class UI ***REMOVED***
  constructor(opt) ***REMOVED***
    // Instantiate the Readline interface
    // @Note: Don't reassign if already present (allow test to override the Stream)
    if (!this.rl) ***REMOVED***
      this.rl = readline.createInterface(setupReadlineOptions(opt));
    ***REMOVED***

    this.rl.resume();

    this.onForceClose = this.onForceClose.bind(this);

    // Make sure new prompt start on a newline when closing
    process.on('exit', this.onForceClose);

    // Terminate process on SIGINT (which will call process.on('exit') in return)
    this.rl.on('SIGINT', this.onForceClose);
  ***REMOVED***

  /**
   * Handle the ^C exit
   * @return ***REMOVED***null***REMOVED***
   */

  onForceClose() ***REMOVED***
    this.close();
    process.kill(process.pid, 'SIGINT');
    console.log('');
  ***REMOVED***

  /**
   * Close the interface and cleanup listeners
   */

  close() ***REMOVED***
    // Remove events listeners
    this.rl.removeListener('SIGINT', this.onForceClose);
    process.removeListener('exit', this.onForceClose);

    this.rl.output.unmute();

    if (this.activePrompt && typeof this.activePrompt.close === 'function') ***REMOVED***
      this.activePrompt.close();
    ***REMOVED***

    // Close the readline
    this.rl.output.end();
    this.rl.pause();
    this.rl.close();
  ***REMOVED***
***REMOVED***

function setupReadlineOptions(opt) ***REMOVED***
  opt = opt || ***REMOVED******REMOVED***;

  // Default `input` to stdin
  var input = opt.input || process.stdin;

  // Add mute capabilities to the output
  var ms = new MuteStream();
  ms.pipe(opt.output || process.stdout);
  var output = ms;

  return _.extend(
    ***REMOVED***
      terminal: true,
      input: input,
      output: output
    ***REMOVED***,
    _.omit(opt, ['input', 'output'])
  );
***REMOVED***

module.exports = UI;
