// argparse/lib/argument_parser.js
/**
 * class ArgumentParser
 *
 * Object for parsing command line strings into js objects.
 *
 * Inherited from [[ActionContainer]]
 **/
'use strict';

var util    = require('util');
var format  = require('util').format;
var Path    = require('path');
var sprintf = require('sprintf-js').sprintf;

// Constants
var c = require('./const');

var $$ = require('./utils');

var ActionContainer = require('./action_container');

// Errors
var argumentErrorHelper = require('./argument/error');

var HelpFormatter = require('./help/formatter');

var Namespace = require('./namespace');


/**
 * new ArgumentParser(options)
 *
 * Create a new ArgumentParser object.
 *
 * ##### Options:
 * - `prog`  The name of the program (default: Path.basename(process.argv[1]))
 * - `usage`  A usage message (default: auto-generated from arguments)
 * - `description`  A description of what the program does
 * - `epilog`  Text following the argument descriptions
 * - `parents`  Parsers whose arguments should be copied into this one
 * - `formatterClass`  HelpFormatter class for printing help messages
 * - `prefixChars`  Characters that prefix optional arguments
 * - `fromfilePrefixChars` Characters that prefix files containing additional arguments
 * - `argumentDefault`  The default value for all arguments
 * - `addHelp`  Add a -h/-help option
 * - `conflictHandler`  Specifies how to handle conflicting argument names
 * - `debug`  Enable debug mode. Argument errors throw exception in
 *   debug mode and process.exit in normal. Used for development and
 *   testing (default: false)
 *
 * See also [original guide][1]
 *
 * [1]:http://docs.python.org/dev/library/argparse.html#argumentparser-objects
 **/
function ArgumentParser(options) ***REMOVED***
  if (!(this instanceof ArgumentParser)) ***REMOVED***
    return new ArgumentParser(options);
  ***REMOVED***
  var self = this;
  options = options || ***REMOVED******REMOVED***;

  options.description = (options.description || null);
  options.argumentDefault = (options.argumentDefault || null);
  options.prefixChars = (options.prefixChars || '-');
  options.conflictHandler = (options.conflictHandler || 'error');
  ActionContainer.call(this, options);

  options.addHelp = typeof options.addHelp === 'undefined' || !!options.addHelp;
  options.parents = options.parents || [];
  // default program name
  options.prog = (options.prog || Path.basename(process.argv[1]));
  this.prog = options.prog;
  this.usage = options.usage;
  this.epilog = options.epilog;
  this.version = options.version;

  this.debug = (options.debug === true);

  this.formatterClass = (options.formatterClass || HelpFormatter);
  this.fromfilePrefixChars = options.fromfilePrefixChars || null;
  this._positionals = this.addArgumentGroup(***REMOVED*** title: 'Positional arguments' ***REMOVED***);
  this._optionals = this.addArgumentGroup(***REMOVED*** title: 'Optional arguments' ***REMOVED***);
  this._subparsers = null;

  // register types
  function FUNCTION_IDENTITY(o) ***REMOVED***
    return o;
  ***REMOVED***
  this.register('type', 'auto', FUNCTION_IDENTITY);
  this.register('type', null, FUNCTION_IDENTITY);
  this.register('type', 'int', function (x) ***REMOVED***
    var result = parseInt(x, 10);
    if (isNaN(result)) ***REMOVED***
      throw new Error(x + ' is not a valid integer.');
    ***REMOVED***
    return result;
  ***REMOVED***);
  this.register('type', 'float', function (x) ***REMOVED***
    var result = parseFloat(x);
    if (isNaN(result)) ***REMOVED***
      throw new Error(x + ' is not a valid float.');
    ***REMOVED***
    return result;
  ***REMOVED***);
  this.register('type', 'string', function (x) ***REMOVED***
    return '' + x;
  ***REMOVED***);

  // add help and version arguments if necessary
  var defaultPrefix = (this.prefixChars.indexOf('-') > -1) ? '-' : this.prefixChars[0];
  if (options.addHelp) ***REMOVED***
    this.addArgument(
      [ defaultPrefix + 'h', defaultPrefix + defaultPrefix + 'help' ],
      ***REMOVED***
        action: 'help',
        defaultValue: c.SUPPRESS,
        help: 'Show this help message and exit.'
      ***REMOVED***
    );
  ***REMOVED***
  if (typeof this.version !== 'undefined') ***REMOVED***
    this.addArgument(
      [ defaultPrefix + 'v', defaultPrefix + defaultPrefix + 'version' ],
      ***REMOVED***
        action: 'version',
        version: this.version,
        defaultValue: c.SUPPRESS,
        help: "Show program's version number and exit."
      ***REMOVED***
    );
  ***REMOVED***

  // add parent arguments and defaults
  options.parents.forEach(function (parent) ***REMOVED***
    self._addContainerActions(parent);
    if (typeof parent._defaults !== 'undefined') ***REMOVED***
      for (var defaultKey in parent._defaults) ***REMOVED***
        if (parent._defaults.hasOwnProperty(defaultKey)) ***REMOVED***
          self._defaults[defaultKey] = parent._defaults[defaultKey];
        ***REMOVED***
      ***REMOVED***
    ***REMOVED***
  ***REMOVED***);
***REMOVED***

util.inherits(ArgumentParser, ActionContainer);

/**
 * ArgumentParser#addSubparsers(options) -> [[ActionSubparsers]]
 * - options (object): hash of options see [[ActionSubparsers.new]]
 *
 * See also [subcommands][1]
 *
 * [1]:http://docs.python.org/dev/library/argparse.html#sub-commands
 **/
ArgumentParser.prototype.addSubparsers = function (options) ***REMOVED***
  if (this._subparsers) ***REMOVED***
    this.error('Cannot have multiple subparser arguments.');
  ***REMOVED***

  options = options || ***REMOVED******REMOVED***;
  options.debug = (this.debug === true);
  options.optionStrings = [];
  options.parserClass = (options.parserClass || ArgumentParser);


  if (!!options.title || !!options.description) ***REMOVED***

    this._subparsers = this.addArgumentGroup(***REMOVED***
      title: (options.title || 'subcommands'),
      description: options.description
    ***REMOVED***);
    delete options.title;
    delete options.description;

  ***REMOVED*** else ***REMOVED***
    this._subparsers = this._positionals;
  ***REMOVED***

  // prog defaults to the usage message of this parser, skipping
  // optional arguments and with no "usage:" prefix
  if (!options.prog) ***REMOVED***
    var formatter = this._getFormatter();
    var positionals = this._getPositionalActions();
    var groups = this._mutuallyExclusiveGroups;
    formatter.addUsage(this.usage, positionals, groups, '');
    options.prog = formatter.formatHelp().trim();
  ***REMOVED***

  // create the parsers action and add it to the positionals list
  var ParsersClass = this._popActionClass(options, 'parsers');
  var action = new ParsersClass(options);
  this._subparsers._addAction(action);

  // return the created parsers action
  return action;
***REMOVED***;

ArgumentParser.prototype._addAction = function (action) ***REMOVED***
  if (action.isOptional()) ***REMOVED***
    this._optionals._addAction(action);
  ***REMOVED*** else ***REMOVED***
    this._positionals._addAction(action);
  ***REMOVED***
  return action;
***REMOVED***;

ArgumentParser.prototype._getOptionalActions = function () ***REMOVED***
  return this._actions.filter(function (action) ***REMOVED***
    return action.isOptional();
  ***REMOVED***);
***REMOVED***;

ArgumentParser.prototype._getPositionalActions = function () ***REMOVED***
  return this._actions.filter(function (action) ***REMOVED***
    return action.isPositional();
  ***REMOVED***);
***REMOVED***;


/**
 * ArgumentParser#parseArgs(args, namespace) -> Namespace|Object
 * - args (array): input elements
 * - namespace (Namespace|Object): result object
 *
 * Parsed args and throws error if some arguments are not recognized
 *
 * See also [original guide][1]
 *
 * [1]:http://docs.python.org/dev/library/argparse.html#the-parse-args-method
 **/
ArgumentParser.prototype.parseArgs = function (args, namespace) ***REMOVED***
  var argv;
  var result = this.parseKnownArgs(args, namespace);

  args = result[0];
  argv = result[1];
  if (argv && argv.length > 0) ***REMOVED***
    this.error(
      format('Unrecognized arguments: %s.', argv.join(' '))
    );
  ***REMOVED***
  return args;
***REMOVED***;

/**
 * ArgumentParser#parseKnownArgs(args, namespace) -> array
 * - args (array): input options
 * - namespace (Namespace|Object): result object
 *
 * Parse known arguments and return tuple of result object
 * and unknown args
 *
 * See also [original guide][1]
 *
 * [1]:http://docs.python.org/dev/library/argparse.html#partial-parsing
 **/
ArgumentParser.prototype.parseKnownArgs = function (args, namespace) ***REMOVED***
  var self = this;

  // args default to the system args
  args = args || process.argv.slice(2);

  // default Namespace built from parser defaults
  namespace = namespace || new Namespace();

  self._actions.forEach(function (action) ***REMOVED***
    if (action.dest !== c.SUPPRESS) ***REMOVED***
      if (!$$.has(namespace, action.dest)) ***REMOVED***
        if (action.defaultValue !== c.SUPPRESS) ***REMOVED***
          var defaultValue = action.defaultValue;
          if (typeof action.defaultValue === 'string') ***REMOVED***
            defaultValue = self._getValue(action, defaultValue);
          ***REMOVED***
          namespace[action.dest] = defaultValue;
        ***REMOVED***
      ***REMOVED***
    ***REMOVED***
  ***REMOVED***);

  Object.keys(self._defaults).forEach(function (dest) ***REMOVED***
    namespace[dest] = self._defaults[dest];
  ***REMOVED***);

  // parse the arguments and exit if there are any errors
  try ***REMOVED***
    var res = this._parseKnownArgs(args, namespace);

    namespace = res[0];
    args = res[1];
    if ($$.has(namespace, c._UNRECOGNIZED_ARGS_ATTR)) ***REMOVED***
      args = $$.arrayUnion(args, namespace[c._UNRECOGNIZED_ARGS_ATTR]);
      delete namespace[c._UNRECOGNIZED_ARGS_ATTR];
    ***REMOVED***
    return [ namespace, args ];
  ***REMOVED*** catch (e) ***REMOVED***
    this.error(e);
  ***REMOVED***
***REMOVED***;

ArgumentParser.prototype._parseKnownArgs = function (argStrings, namespace) ***REMOVED***
  var self = this;

  var extras = [];

  // replace arg strings that are file references
  if (this.fromfilePrefixChars !== null) ***REMOVED***
    argStrings = this._readArgsFromFiles(argStrings);
  ***REMOVED***
  // map all mutually exclusive arguments to the other arguments
  // they can't occur with
  // Python has 'conflicts = action_conflicts.setdefault(mutex_action, [])'
  // though I can't conceive of a way in which an action could be a member
  // of two different mutually exclusive groups.

  function actionHash(action) ***REMOVED***
    // some sort of hashable key for this action
    // action itself cannot be a key in actionConflicts
    // I think getName() (join of optionStrings) is unique enough
    return action.getName();
  ***REMOVED***

  var conflicts, key;
  var actionConflicts = ***REMOVED******REMOVED***;

  this._mutuallyExclusiveGroups.forEach(function (mutexGroup) ***REMOVED***
    mutexGroup._groupActions.forEach(function (mutexAction, i, groupActions) ***REMOVED***
      key = actionHash(mutexAction);
      if (!$$.has(actionConflicts, key)) ***REMOVED***
        actionConflicts[key] = [];
      ***REMOVED***
      conflicts = actionConflicts[key];
      conflicts.push.apply(conflicts, groupActions.slice(0, i));
      conflicts.push.apply(conflicts, groupActions.slice(i + 1));
    ***REMOVED***);
  ***REMOVED***);

  // find all option indices, and determine the arg_string_pattern
  // which has an 'O' if there is an option at an index,
  // an 'A' if there is an argument, or a '-' if there is a '--'
  var optionStringIndices = ***REMOVED******REMOVED***;

  var argStringPatternParts = [];

  argStrings.forEach(function (argString, argStringIndex) ***REMOVED***
    if (argString === '--') ***REMOVED***
      argStringPatternParts.push('-');
      while (argStringIndex < argStrings.length) ***REMOVED***
        argStringPatternParts.push('A');
        argStringIndex++;
      ***REMOVED***
    ***REMOVED*** else ***REMOVED***
      // otherwise, add the arg to the arg strings
      // and note the index if it was an option
      var pattern;
      var optionTuple = self._parseOptional(argString);
      if (!optionTuple) ***REMOVED***
        pattern = 'A';
      ***REMOVED*** else ***REMOVED***
        optionStringIndices[argStringIndex] = optionTuple;
        pattern = 'O';
      ***REMOVED***
      argStringPatternParts.push(pattern);
    ***REMOVED***
  ***REMOVED***);
  var argStringsPattern = argStringPatternParts.join('');

  var seenActions = [];
  var seenNonDefaultActions = [];


  function takeAction(action, argumentStrings, optionString) ***REMOVED***
    seenActions.push(action);
    var argumentValues = self._getValues(action, argumentStrings);

    // error if this argument is not allowed with other previously
    // seen arguments, assuming that actions that use the default
    // value don't really count as "present"
    if (argumentValues !== action.defaultValue) ***REMOVED***
      seenNonDefaultActions.push(action);
      if (actionConflicts[actionHash(action)]) ***REMOVED***
        actionConflicts[actionHash(action)].forEach(function (actionConflict) ***REMOVED***
          if (seenNonDefaultActions.indexOf(actionConflict) >= 0) ***REMOVED***
            throw argumentErrorHelper(
              action,
              format('Not allowed with argument "%s".', actionConflict.getName())
            );
          ***REMOVED***
        ***REMOVED***);
      ***REMOVED***
    ***REMOVED***

    if (argumentValues !== c.SUPPRESS) ***REMOVED***
      action.call(self, namespace, argumentValues, optionString);
    ***REMOVED***
  ***REMOVED***

  function consumeOptional(startIndex) ***REMOVED***
    // get the optional identified at this index
    var optionTuple = optionStringIndices[startIndex];
    var action = optionTuple[0];
    var optionString = optionTuple[1];
    var explicitArg = optionTuple[2];

    // identify additional optionals in the same arg string
    // (e.g. -xyz is the same as -x -y -z if no args are required)
    var actionTuples = [];

    var args, argCount, start, stop;

    for (;;) ***REMOVED***
      if (!action) ***REMOVED***
        extras.push(argStrings[startIndex]);
        return startIndex + 1;
      ***REMOVED***
      if (explicitArg) ***REMOVED***
        argCount = self._matchArgument(action, 'A');

        // if the action is a single-dash option and takes no
        // arguments, try to parse more single-dash options out
        // of the tail of the option string
        var chars = self.prefixChars;
        if (argCount === 0 && chars.indexOf(optionString[1]) < 0) ***REMOVED***
          actionTuples.push([ action, [], optionString ]);
          optionString = optionString[0] + explicitArg[0];
          var newExplicitArg = explicitArg.slice(1) || null;
          var optionalsMap = self._optionStringActions;

          if (Object.keys(optionalsMap).indexOf(optionString) >= 0) ***REMOVED***
            action = optionalsMap[optionString];
            explicitArg = newExplicitArg;
          ***REMOVED*** else ***REMOVED***
            throw argumentErrorHelper(action, sprintf('ignored explicit argument %r', explicitArg));
          ***REMOVED***
        ***REMOVED*** else if (argCount === 1) ***REMOVED***
          // if the action expect exactly one argument, we've
          // successfully matched the option; exit the loop
          stop = startIndex + 1;
          args = [ explicitArg ];
          actionTuples.push([ action, args, optionString ]);
          break;
        ***REMOVED*** else ***REMOVED***
          // error if a double-dash option did not use the
          // explicit argument
          throw argumentErrorHelper(action, sprintf('ignored explicit argument %r', explicitArg));
        ***REMOVED***
      ***REMOVED*** else ***REMOVED***
        // if there is no explicit argument, try to match the
        // optional's string arguments with the following strings
        // if successful, exit the loop

        start = startIndex + 1;
        var selectedPatterns = argStringsPattern.substr(start);

        argCount = self._matchArgument(action, selectedPatterns);
        stop = start + argCount;


        args = argStrings.slice(start, stop);

        actionTuples.push([ action, args, optionString ]);
        break;
      ***REMOVED***

    ***REMOVED***

    // add the Optional to the list and return the index at which
    // the Optional's string args stopped
    if (actionTuples.length < 1) ***REMOVED***
      throw new Error('length should be > 0');
    ***REMOVED***
    for (var i = 0; i < actionTuples.length; i++) ***REMOVED***
      takeAction.apply(self, actionTuples[i]);
    ***REMOVED***
    return stop;
  ***REMOVED***

  // the list of Positionals left to be parsed; this is modified
  // by consume_positionals()
  var positionals = self._getPositionalActions();

  function consumePositionals(startIndex) ***REMOVED***
    // match as many Positionals as possible
    var selectedPattern = argStringsPattern.substr(startIndex);
    var argCounts = self._matchArgumentsPartial(positionals, selectedPattern);

    // slice off the appropriate arg strings for each Positional
    // and add the Positional and its args to the list
    for (var i = 0; i < positionals.length; i++) ***REMOVED***
      var action = positionals[i];
      var argCount = argCounts[i];
      if (typeof argCount === 'undefined') ***REMOVED***
        continue;
      ***REMOVED***
      var args = argStrings.slice(startIndex, startIndex + argCount);

      startIndex += argCount;
      takeAction(action, args);
    ***REMOVED***

    // slice off the Positionals that we just parsed and return the
    // index at which the Positionals' string args stopped
    positionals = positionals.slice(argCounts.length);
    return startIndex;
  ***REMOVED***

  // consume Positionals and Optionals alternately, until we have
  // passed the last option string
  var startIndex = 0;
  var position;

  var maxOptionStringIndex = -1;

  Object.keys(optionStringIndices).forEach(function (position) ***REMOVED***
    maxOptionStringIndex = Math.max(maxOptionStringIndex, parseInt(position, 10));
  ***REMOVED***);

  var positionalsEndIndex, nextOptionStringIndex;

  while (startIndex <= maxOptionStringIndex) ***REMOVED***
    // consume any Positionals preceding the next option
    nextOptionStringIndex = null;
    for (position in optionStringIndices) ***REMOVED***
      if (!optionStringIndices.hasOwnProperty(position)) ***REMOVED*** continue; ***REMOVED***

      position = parseInt(position, 10);
      if (position >= startIndex) ***REMOVED***
        if (nextOptionStringIndex !== null) ***REMOVED***
          nextOptionStringIndex = Math.min(nextOptionStringIndex, position);
        ***REMOVED*** else ***REMOVED***
          nextOptionStringIndex = position;
        ***REMOVED***
      ***REMOVED***
    ***REMOVED***

    if (startIndex !== nextOptionStringIndex) ***REMOVED***
      positionalsEndIndex = consumePositionals(startIndex);
      // only try to parse the next optional if we didn't consume
      // the option string during the positionals parsing
      if (positionalsEndIndex > startIndex) ***REMOVED***
        startIndex = positionalsEndIndex;
        continue;
      ***REMOVED*** else ***REMOVED***
        startIndex = positionalsEndIndex;
      ***REMOVED***
    ***REMOVED***

    // if we consumed all the positionals we could and we're not
    // at the index of an option string, there were extra arguments
    if (!optionStringIndices[startIndex]) ***REMOVED***
      var strings = argStrings.slice(startIndex, nextOptionStringIndex);
      extras = extras.concat(strings);
      startIndex = nextOptionStringIndex;
    ***REMOVED***
    // consume the next optional and any arguments for it
    startIndex = consumeOptional(startIndex);
  ***REMOVED***

  // consume any positionals following the last Optional
  var stopIndex = consumePositionals(startIndex);

  // if we didn't consume all the argument strings, there were extras
  extras = extras.concat(argStrings.slice(stopIndex));

  // if we didn't use all the Positional objects, there were too few
  // arg strings supplied.
  if (positionals.length > 0) ***REMOVED***
    self.error('too few arguments');
  ***REMOVED***

  // make sure all required actions were present
  self._actions.forEach(function (action) ***REMOVED***
    if (action.required) ***REMOVED***
      if (seenActions.indexOf(action) < 0) ***REMOVED***
        self.error(format('Argument "%s" is required', action.getName()));
      ***REMOVED***
    ***REMOVED***
  ***REMOVED***);

  // make sure all required groups have one option present
  var actionUsed = false;
  self._mutuallyExclusiveGroups.forEach(function (group) ***REMOVED***
    if (group.required) ***REMOVED***
      actionUsed = group._groupActions.some(function (action) ***REMOVED***
        return seenNonDefaultActions.indexOf(action) !== -1;
      ***REMOVED***);

      // if no actions were used, report the error
      if (!actionUsed) ***REMOVED***
        var names = [];
        group._groupActions.forEach(function (action) ***REMOVED***
          if (action.help !== c.SUPPRESS) ***REMOVED***
            names.push(action.getName());
          ***REMOVED***
        ***REMOVED***);
        names = names.join(' ');
        var msg = 'one of the arguments ' + names + ' is required';
        self.error(msg);
      ***REMOVED***
    ***REMOVED***
  ***REMOVED***);

  // return the updated namespace and the extra arguments
  return [ namespace, extras ];
***REMOVED***;

ArgumentParser.prototype._readArgsFromFiles = function (argStrings) ***REMOVED***
  // expand arguments referencing files
  var self = this;
  var fs = require('fs');
  var newArgStrings = [];
  argStrings.forEach(function (argString) ***REMOVED***
    if (self.fromfilePrefixChars.indexOf(argString[0]) < 0) ***REMOVED***
      // for regular arguments, just add them back into the list
      newArgStrings.push(argString);
    ***REMOVED*** else ***REMOVED***
      // replace arguments referencing files with the file content
      try ***REMOVED***
        var argstrs = [];
        var filename = argString.slice(1);
        var content = fs.readFileSync(filename, 'utf8');
        content = content.trim().split('\n');
        content.forEach(function (argLine) ***REMOVED***
          self.convertArgLineToArgs(argLine).forEach(function (arg) ***REMOVED***
            argstrs.push(arg);
          ***REMOVED***);
          argstrs = self._readArgsFromFiles(argstrs);
        ***REMOVED***);
        newArgStrings.push.apply(newArgStrings, argstrs);
      ***REMOVED*** catch (error) ***REMOVED***
        return self.error(error.message);
      ***REMOVED***
    ***REMOVED***
  ***REMOVED***);
  return newArgStrings;
***REMOVED***;

ArgumentParser.prototype.convertArgLineToArgs = function (argLine) ***REMOVED***
  return [ argLine ];
***REMOVED***;

ArgumentParser.prototype._matchArgument = function (action, regexpArgStrings) ***REMOVED***

  // match the pattern for this action to the arg strings
  var regexpNargs = new RegExp('^' + this._getNargsPattern(action));
  var matches = regexpArgStrings.match(regexpNargs);
  var message;

  // throw an exception if we weren't able to find a match
  if (!matches) ***REMOVED***
    switch (action.nargs) ***REMOVED***
      /*eslint-disable no-undefined*/
      case undefined:
      case null:
        message = 'Expected one argument.';
        break;
      case c.OPTIONAL:
        message = 'Expected at most one argument.';
        break;
      case c.ONE_OR_MORE:
        message = 'Expected at least one argument.';
        break;
      default:
        message = 'Expected %s argument(s)';
    ***REMOVED***

    throw argumentErrorHelper(
      action,
      format(message, action.nargs)
    );
  ***REMOVED***
  // return the number of arguments matched
  return matches[1].length;
***REMOVED***;

ArgumentParser.prototype._matchArgumentsPartial = function (actions, regexpArgStrings) ***REMOVED***
  // progressively shorten the actions list by slicing off the
  // final actions until we find a match
  var self = this;
  var result = [];
  var actionSlice, pattern, matches;
  var i, j;

  function getLength(string) ***REMOVED***
    return string.length;
  ***REMOVED***

  for (i = actions.length; i > 0; i--) ***REMOVED***
    pattern = '';
    actionSlice = actions.slice(0, i);
    for (j = 0; j < actionSlice.length; j++) ***REMOVED***
      pattern += self._getNargsPattern(actionSlice[j]);
    ***REMOVED***

    pattern = new RegExp('^' + pattern);
    matches = regexpArgStrings.match(pattern);

    if (matches && matches.length > 0) ***REMOVED***
      // need only groups
      matches = matches.splice(1);
      result = result.concat(matches.map(getLength));
      break;
    ***REMOVED***
  ***REMOVED***

  // return the list of arg string counts
  return result;
***REMOVED***;

ArgumentParser.prototype._parseOptional = function (argString) ***REMOVED***
  var action, optionString, argExplicit, optionTuples;

  // if it's an empty string, it was meant to be a positional
  if (!argString) ***REMOVED***
    return null;
  ***REMOVED***

  // if it doesn't start with a prefix, it was meant to be positional
  if (this.prefixChars.indexOf(argString[0]) < 0) ***REMOVED***
    return null;
  ***REMOVED***

  // if the option string is present in the parser, return the action
  if (this._optionStringActions[argString]) ***REMOVED***
    return [ this._optionStringActions[argString], argString, null ];
  ***REMOVED***

  // if it's just a single character, it was meant to be positional
  if (argString.length === 1) ***REMOVED***
    return null;
  ***REMOVED***

  // if the option string before the "=" is present, return the action
  if (argString.indexOf('=') >= 0) ***REMOVED***
    optionString = argString.split('=', 1)[0];
    argExplicit = argString.slice(optionString.length + 1);

    if (this._optionStringActions[optionString]) ***REMOVED***
      action = this._optionStringActions[optionString];
      return [ action, optionString, argExplicit ];
    ***REMOVED***
  ***REMOVED***

  // search through all possible prefixes of the option string
  // and all actions in the parser for possible interpretations
  optionTuples = this._getOptionTuples(argString);

  // if multiple actions match, the option string was ambiguous
  if (optionTuples.length > 1) ***REMOVED***
    var optionStrings = optionTuples.map(function (optionTuple) ***REMOVED***
      return optionTuple[1];
    ***REMOVED***);
    this.error(format(
          'Ambiguous option: "%s" could match %s.',
          argString, optionStrings.join(', ')
    ));
  // if exactly one action matched, this segmentation is good,
  // so return the parsed action
  ***REMOVED*** else if (optionTuples.length === 1) ***REMOVED***
    return optionTuples[0];
  ***REMOVED***

  // if it was not found as an option, but it looks like a negative
  // number, it was meant to be positional
  // unless there are negative-number-like options
  if (argString.match(this._regexpNegativeNumber)) ***REMOVED***
    if (!this._hasNegativeNumberOptionals.some(Boolean)) ***REMOVED***
      return null;
    ***REMOVED***
  ***REMOVED***
  // if it contains a space, it was meant to be a positional
  if (argString.search(' ') >= 0) ***REMOVED***
    return null;
  ***REMOVED***

  // it was meant to be an optional but there is no such option
  // in this parser (though it might be a valid option in a subparser)
  return [ null, argString, null ];
***REMOVED***;

ArgumentParser.prototype._getOptionTuples = function (optionString) ***REMOVED***
  var result = [];
  var chars = this.prefixChars;
  var optionPrefix;
  var argExplicit;
  var action;
  var actionOptionString;

  // option strings starting with two prefix characters are only split at
  // the '='
  if (chars.indexOf(optionString[0]) >= 0 && chars.indexOf(optionString[1]) >= 0) ***REMOVED***
    if (optionString.indexOf('=') >= 0) ***REMOVED***
      var optionStringSplit = optionString.split('=', 1);

      optionPrefix = optionStringSplit[0];
      argExplicit = optionStringSplit[1];
    ***REMOVED*** else ***REMOVED***
      optionPrefix = optionString;
      argExplicit = null;
    ***REMOVED***

    for (actionOptionString in this._optionStringActions) ***REMOVED***
      if (actionOptionString.substr(0, optionPrefix.length) === optionPrefix) ***REMOVED***
        action = this._optionStringActions[actionOptionString];
        result.push([ action, actionOptionString, argExplicit ]);
      ***REMOVED***
    ***REMOVED***

  // single character options can be concatenated with their arguments
  // but multiple character options always have to have their argument
  // separate
  ***REMOVED*** else if (chars.indexOf(optionString[0]) >= 0 && chars.indexOf(optionString[1]) < 0) ***REMOVED***
    optionPrefix = optionString;
    argExplicit = null;
    var optionPrefixShort = optionString.substr(0, 2);
    var argExplicitShort = optionString.substr(2);

    for (actionOptionString in this._optionStringActions) ***REMOVED***
      if (!$$.has(this._optionStringActions, actionOptionString)) continue;

      action = this._optionStringActions[actionOptionString];
      if (actionOptionString === optionPrefixShort) ***REMOVED***
        result.push([ action, actionOptionString, argExplicitShort ]);
      ***REMOVED*** else if (actionOptionString.substr(0, optionPrefix.length) === optionPrefix) ***REMOVED***
        result.push([ action, actionOptionString, argExplicit ]);
      ***REMOVED***
    ***REMOVED***

  // shouldn't ever get here
  ***REMOVED*** else ***REMOVED***
    throw new Error(format('Unexpected option string: %s.', optionString));
  ***REMOVED***
  // return the collected option tuples
  return result;
***REMOVED***;

ArgumentParser.prototype._getNargsPattern = function (action) ***REMOVED***
  // in all examples below, we have to allow for '--' args
  // which are represented as '-' in the pattern
  var regexpNargs;

  switch (action.nargs) ***REMOVED***
    // the default (null) is assumed to be a single argument
    case undefined:
    case null:
      regexpNargs = '(-*A-*)';
      break;
    // allow zero or more arguments
    case c.OPTIONAL:
      regexpNargs = '(-*A?-*)';
      break;
    // allow zero or more arguments
    case c.ZERO_OR_MORE:
      regexpNargs = '(-*[A-]*)';
      break;
    // allow one or more arguments
    case c.ONE_OR_MORE:
      regexpNargs = '(-*A[A-]*)';
      break;
    // allow any number of options or arguments
    case c.REMAINDER:
      regexpNargs = '([-AO]*)';
      break;
    // allow one argument followed by any number of options or arguments
    case c.PARSER:
      regexpNargs = '(-*A[-AO]*)';
      break;
    // all others should be integers
    default:
      regexpNargs = '(-*' + $$.repeat('-*A', action.nargs) + '-*)';
  ***REMOVED***

  // if this is an optional action, -- is not allowed
  if (action.isOptional()) ***REMOVED***
    regexpNargs = regexpNargs.replace(/-\*/g, '');
    regexpNargs = regexpNargs.replace(/-/g, '');
  ***REMOVED***

  // return the pattern
  return regexpNargs;
***REMOVED***;

//
// Value conversion methods
//

ArgumentParser.prototype._getValues = function (action, argStrings) ***REMOVED***
  var self = this;

  // for everything but PARSER args, strip out '--'
  if (action.nargs !== c.PARSER && action.nargs !== c.REMAINDER) ***REMOVED***
    argStrings = argStrings.filter(function (arrayElement) ***REMOVED***
      return arrayElement !== '--';
    ***REMOVED***);
  ***REMOVED***

  var value, argString;

  // optional argument produces a default when not present
  if (argStrings.length === 0 && action.nargs === c.OPTIONAL) ***REMOVED***

    value = (action.isOptional()) ? action.constant : action.defaultValue;

    if (typeof (value) === 'string') ***REMOVED***
      value = this._getValue(action, value);
      this._checkValue(action, value);
    ***REMOVED***

  // when nargs='*' on a positional, if there were no command-line
  // args, use the default if it is anything other than None
  ***REMOVED*** else if (argStrings.length === 0 && action.nargs === c.ZERO_OR_MORE &&
    action.optionStrings.length === 0) ***REMOVED***

    value = (action.defaultValue || argStrings);
    this._checkValue(action, value);

  // single argument or optional argument produces a single value
  ***REMOVED*** else if (argStrings.length === 1 &&
        (!action.nargs || action.nargs === c.OPTIONAL)) ***REMOVED***

    argString = argStrings[0];
    value = this._getValue(action, argString);
    this._checkValue(action, value);

  // REMAINDER arguments convert all values, checking none
  ***REMOVED*** else if (action.nargs === c.REMAINDER) ***REMOVED***
    value = argStrings.map(function (v) ***REMOVED***
      return self._getValue(action, v);
    ***REMOVED***);

  // PARSER arguments convert all values, but check only the first
  ***REMOVED*** else if (action.nargs === c.PARSER) ***REMOVED***
    value = argStrings.map(function (v) ***REMOVED***
      return self._getValue(action, v);
    ***REMOVED***);
    this._checkValue(action, value[0]);

  // all other types of nargs produce a list
  ***REMOVED*** else ***REMOVED***
    value = argStrings.map(function (v) ***REMOVED***
      return self._getValue(action, v);
    ***REMOVED***);
    value.forEach(function (v) ***REMOVED***
      self._checkValue(action, v);
    ***REMOVED***);
  ***REMOVED***

  // return the converted value
  return value;
***REMOVED***;

ArgumentParser.prototype._getValue = function (action, argString) ***REMOVED***
  var result;

  var typeFunction = this._registryGet('type', action.type, action.type);
  if (typeof typeFunction !== 'function') ***REMOVED***
    var message = format('%s is not callable', typeFunction);
    throw argumentErrorHelper(action, message);
  ***REMOVED***

  // convert the value to the appropriate type
  try ***REMOVED***
    result = typeFunction(argString);

    // ArgumentTypeErrors indicate errors
    // If action.type is not a registered string, it is a function
    // Try to deduce its name for inclusion in the error message
    // Failing that, include the error message it raised.
  ***REMOVED*** catch (e) ***REMOVED***
    var name = null;
    if (typeof action.type === 'string') ***REMOVED***
      name = action.type;
    ***REMOVED*** else ***REMOVED***
      name = action.type.name || action.type.displayName || '<function>';
    ***REMOVED***
    var msg = format('Invalid %s value: %s', name, argString);
    if (name === '<function>') ***REMOVED*** msg += '\n' + e.message; ***REMOVED***
    throw argumentErrorHelper(action, msg);
  ***REMOVED***
  // return the converted value
  return result;
***REMOVED***;

ArgumentParser.prototype._checkValue = function (action, value) ***REMOVED***
  // converted value must be one of the choices (if specified)
  var choices = action.choices;
  if (choices) ***REMOVED***
    // choise for argument can by array or string
    if ((typeof choices === 'string' || Array.isArray(choices)) &&
        choices.indexOf(value) !== -1) ***REMOVED***
      return;
    ***REMOVED***
    // choise for subparsers can by only hash
    if (typeof choices === 'object' && !Array.isArray(choices) && choices[value]) ***REMOVED***
      return;
    ***REMOVED***

    if (typeof choices === 'string') ***REMOVED***
      choices = choices.split('').join(', ');
    ***REMOVED*** else if (Array.isArray(choices)) ***REMOVED***
      choices =  choices.join(', ');
    ***REMOVED*** else ***REMOVED***
      choices =  Object.keys(choices).join(', ');
    ***REMOVED***
    var message = format('Invalid choice: %s (choose from [%s])', value, choices);
    throw argumentErrorHelper(action, message);
  ***REMOVED***
***REMOVED***;

//
// Help formatting methods
//

/**
 * ArgumentParser#formatUsage -> string
 *
 * Return usage string
 *
 * See also [original guide][1]
 *
 * [1]:http://docs.python.org/dev/library/argparse.html#printing-help
 **/
ArgumentParser.prototype.formatUsage = function () ***REMOVED***
  var formatter = this._getFormatter();
  formatter.addUsage(this.usage, this._actions, this._mutuallyExclusiveGroups);
  return formatter.formatHelp();
***REMOVED***;

/**
 * ArgumentParser#formatHelp -> string
 *
 * Return help
 *
 * See also [original guide][1]
 *
 * [1]:http://docs.python.org/dev/library/argparse.html#printing-help
 **/
ArgumentParser.prototype.formatHelp = function () ***REMOVED***
  var formatter = this._getFormatter();

  // usage
  formatter.addUsage(this.usage, this._actions, this._mutuallyExclusiveGroups);

  // description
  formatter.addText(this.description);

  // positionals, optionals and user-defined groups
  this._actionGroups.forEach(function (actionGroup) ***REMOVED***
    formatter.startSection(actionGroup.title);
    formatter.addText(actionGroup.description);
    formatter.addArguments(actionGroup._groupActions);
    formatter.endSection();
  ***REMOVED***);

  // epilog
  formatter.addText(this.epilog);

  // determine help from format above
  return formatter.formatHelp();
***REMOVED***;

ArgumentParser.prototype._getFormatter = function () ***REMOVED***
  var FormatterClass = this.formatterClass;
  var formatter = new FormatterClass(***REMOVED*** prog: this.prog ***REMOVED***);
  return formatter;
***REMOVED***;

//
//  Print functions
//

/**
 * ArgumentParser#printUsage() -> Void
 *
 * Print usage
 *
 * See also [original guide][1]
 *
 * [1]:http://docs.python.org/dev/library/argparse.html#printing-help
 **/
ArgumentParser.prototype.printUsage = function () ***REMOVED***
  this._printMessage(this.formatUsage());
***REMOVED***;

/**
 * ArgumentParser#printHelp() -> Void
 *
 * Print help
 *
 * See also [original guide][1]
 *
 * [1]:http://docs.python.org/dev/library/argparse.html#printing-help
 **/
ArgumentParser.prototype.printHelp = function () ***REMOVED***
  this._printMessage(this.formatHelp());
***REMOVED***;

ArgumentParser.prototype._printMessage = function (message, stream) ***REMOVED***
  if (!stream) ***REMOVED***
    stream = process.stdout;
  ***REMOVED***
  if (message) ***REMOVED***
    stream.write('' + message);
  ***REMOVED***
***REMOVED***;

//
//  Exit functions
//

/**
 * ArgumentParser#exit(status=0, message) -> Void
 * - status (int): exit status
 * - message (string): message
 *
 * Print message in stderr/stdout and exit program
 **/
ArgumentParser.prototype.exit = function (status, message) ***REMOVED***
  if (message) ***REMOVED***
    if (status === 0) ***REMOVED***
      this._printMessage(message);
    ***REMOVED*** else ***REMOVED***
      this._printMessage(message, process.stderr);
    ***REMOVED***
  ***REMOVED***

  process.exit(status);
***REMOVED***;

/**
 * ArgumentParser#error(message) -> Void
 * - err (Error|string): message
 *
 * Error method Prints a usage message incorporating the message to stderr and
 * exits. If you override this in a subclass,
 * it should not return -- it should
 * either exit or throw an exception.
 *
 **/
ArgumentParser.prototype.error = function (err) ***REMOVED***
  var message;
  if (err instanceof Error) ***REMOVED***
    if (this.debug === true) ***REMOVED***
      throw err;
    ***REMOVED***
    message = err.message;
  ***REMOVED*** else ***REMOVED***
    message = err;
  ***REMOVED***
  var msg = format('%s: error: %s', this.prog, message) + c.EOL;

  if (this.debug === true) ***REMOVED***
    throw new Error(msg);
  ***REMOVED***

  this.printUsage(process.stderr);

  return this.exit(2, msg);
***REMOVED***;

module.exports = ArgumentParser;
