// eslint-utils/index.mjs
/*! @author Toru Nagashima <https://github.com/mysticatea> */
import evk from 'eslint-visitor-keys';

/**
 * Get the innermost scope which contains a given location.
 * @param ***REMOVED***Scope***REMOVED*** initialScope The initial scope to search.
 * @param ***REMOVED***Node***REMOVED*** node The location to search.
 * @returns ***REMOVED***Scope***REMOVED*** The innermost scope.
 */
function getInnermostScope(initialScope, node) ***REMOVED***
    const location = node.range[0];

    let scope = initialScope;
    let found = false;
    do ***REMOVED***
        found = false;
        for (const childScope of scope.childScopes) ***REMOVED***
            const range = childScope.block.range;

            if (range[0] <= location && location < range[1]) ***REMOVED***
                scope = childScope;
                found = true;
                break
            ***REMOVED***
        ***REMOVED***
    ***REMOVED*** while (found)

    return scope
***REMOVED***

/**
 * Find the variable of a given name.
 * @param ***REMOVED***Scope***REMOVED*** initialScope The scope to start finding.
 * @param ***REMOVED***string|Node***REMOVED*** nameOrNode The variable name to find. If this is a Node object then it should be an Identifier node.
 * @returns ***REMOVED***Variable|null***REMOVED*** The found variable or null.
 */
function findVariable(initialScope, nameOrNode) ***REMOVED***
    let name = "";
    let scope = initialScope;

    if (typeof nameOrNode === "string") ***REMOVED***
        name = nameOrNode;
    ***REMOVED*** else ***REMOVED***
        name = nameOrNode.name;
        scope = getInnermostScope(scope, nameOrNode);
    ***REMOVED***

    while (scope != null) ***REMOVED***
        const variable = scope.set.get(name);
        if (variable != null) ***REMOVED***
            return variable
        ***REMOVED***
        scope = scope.upper;
    ***REMOVED***

    return null
***REMOVED***

/**
 * Negate the result of `this` calling.
 * @param ***REMOVED***Token***REMOVED*** token The token to check.
 * @returns ***REMOVED***boolean***REMOVED*** `true` if the result of `this(token)` is `false`.
 */
function negate0(token) ***REMOVED***
    return !this(token) //eslint-disable-line no-invalid-this
***REMOVED***

/**
 * Creates the negate function of the given function.
 * @param ***REMOVED***function(Token):boolean***REMOVED*** f - The function to negate.
 * @returns ***REMOVED***function(Token):boolean***REMOVED*** Negated function.
 */
function negate(f) ***REMOVED***
    return negate0.bind(f)
***REMOVED***

/**
 * Checks if the given token is an arrow token or not.
 * @param ***REMOVED***Token***REMOVED*** token - The token to check.
 * @returns ***REMOVED***boolean***REMOVED*** `true` if the token is an arrow token.
 */
function isArrowToken(token) ***REMOVED***
    return token.value === "=>" && token.type === "Punctuator"
***REMOVED***

/**
 * Checks if the given token is a comma token or not.
 * @param ***REMOVED***Token***REMOVED*** token - The token to check.
 * @returns ***REMOVED***boolean***REMOVED*** `true` if the token is a comma token.
 */
function isCommaToken(token) ***REMOVED***
    return token.value === "," && token.type === "Punctuator"
***REMOVED***

/**
 * Checks if the given token is a semicolon token or not.
 * @param ***REMOVED***Token***REMOVED*** token - The token to check.
 * @returns ***REMOVED***boolean***REMOVED*** `true` if the token is a semicolon token.
 */
function isSemicolonToken(token) ***REMOVED***
    return token.value === ";" && token.type === "Punctuator"
***REMOVED***

/**
 * Checks if the given token is a colon token or not.
 * @param ***REMOVED***Token***REMOVED*** token - The token to check.
 * @returns ***REMOVED***boolean***REMOVED*** `true` if the token is a colon token.
 */
function isColonToken(token) ***REMOVED***
    return token.value === ":" && token.type === "Punctuator"
***REMOVED***

/**
 * Checks if the given token is an opening parenthesis token or not.
 * @param ***REMOVED***Token***REMOVED*** token - The token to check.
 * @returns ***REMOVED***boolean***REMOVED*** `true` if the token is an opening parenthesis token.
 */
function isOpeningParenToken(token) ***REMOVED***
    return token.value === "(" && token.type === "Punctuator"
***REMOVED***

/**
 * Checks if the given token is a closing parenthesis token or not.
 * @param ***REMOVED***Token***REMOVED*** token - The token to check.
 * @returns ***REMOVED***boolean***REMOVED*** `true` if the token is a closing parenthesis token.
 */
function isClosingParenToken(token) ***REMOVED***
    return token.value === ")" && token.type === "Punctuator"
***REMOVED***

/**
 * Checks if the given token is an opening square bracket token or not.
 * @param ***REMOVED***Token***REMOVED*** token - The token to check.
 * @returns ***REMOVED***boolean***REMOVED*** `true` if the token is an opening square bracket token.
 */
function isOpeningBracketToken(token) ***REMOVED***
    return token.value === "[" && token.type === "Punctuator"
***REMOVED***

/**
 * Checks if the given token is a closing square bracket token or not.
 * @param ***REMOVED***Token***REMOVED*** token - The token to check.
 * @returns ***REMOVED***boolean***REMOVED*** `true` if the token is a closing square bracket token.
 */
function isClosingBracketToken(token) ***REMOVED***
    return token.value === "]" && token.type === "Punctuator"
***REMOVED***

/**
 * Checks if the given token is an opening brace token or not.
 * @param ***REMOVED***Token***REMOVED*** token - The token to check.
 * @returns ***REMOVED***boolean***REMOVED*** `true` if the token is an opening brace token.
 */
function isOpeningBraceToken(token) ***REMOVED***
    return token.value === "***REMOVED***" && token.type === "Punctuator"
***REMOVED***

/**
 * Checks if the given token is a closing brace token or not.
 * @param ***REMOVED***Token***REMOVED*** token - The token to check.
 * @returns ***REMOVED***boolean***REMOVED*** `true` if the token is a closing brace token.
 */
function isClosingBraceToken(token) ***REMOVED***
    return token.value === "***REMOVED***" && token.type === "Punctuator"
***REMOVED***

/**
 * Checks if the given token is a comment token or not.
 * @param ***REMOVED***Token***REMOVED*** token - The token to check.
 * @returns ***REMOVED***boolean***REMOVED*** `true` if the token is a comment token.
 */
function isCommentToken(token) ***REMOVED***
    return (
        token.type === "Line" ||
        token.type === "Block" ||
        token.type === "Shebang"
    )
***REMOVED***

const isNotArrowToken = negate(isArrowToken);
const isNotCommaToken = negate(isCommaToken);
const isNotSemicolonToken = negate(isSemicolonToken);
const isNotColonToken = negate(isColonToken);
const isNotOpeningParenToken = negate(isOpeningParenToken);
const isNotClosingParenToken = negate(isClosingParenToken);
const isNotOpeningBracketToken = negate(isOpeningBracketToken);
const isNotClosingBracketToken = negate(isClosingBracketToken);
const isNotOpeningBraceToken = negate(isOpeningBraceToken);
const isNotClosingBraceToken = negate(isClosingBraceToken);
const isNotCommentToken = negate(isCommentToken);

/**
 * Get the `(` token of the given function node.
 * @param ***REMOVED***Node***REMOVED*** node - The function node to get.
 * @param ***REMOVED***SourceCode***REMOVED*** sourceCode - The source code object to get tokens.
 * @returns ***REMOVED***Token***REMOVED*** `(` token.
 */
function getOpeningParenOfParams(node, sourceCode) ***REMOVED***
    return node.id
        ? sourceCode.getTokenAfter(node.id, isOpeningParenToken)
        : sourceCode.getFirstToken(node, isOpeningParenToken)
***REMOVED***

/**
 * Get the location of the given function node for reporting.
 * @param ***REMOVED***Node***REMOVED*** node - The function node to get.
 * @param ***REMOVED***SourceCode***REMOVED*** sourceCode - The source code object to get tokens.
 * @returns ***REMOVED***string***REMOVED*** The location of the function node for reporting.
 */
function getFunctionHeadLocation(node, sourceCode) ***REMOVED***
    const parent = node.parent;
    let start = null;
    let end = null;

    if (node.type === "ArrowFunctionExpression") ***REMOVED***
        const arrowToken = sourceCode.getTokenBefore(node.body, isArrowToken);

        start = arrowToken.loc.start;
        end = arrowToken.loc.end;
    ***REMOVED*** else if (
        parent.type === "Property" ||
        parent.type === "MethodDefinition"
    ) ***REMOVED***
        start = parent.loc.start;
        end = getOpeningParenOfParams(node, sourceCode).loc.start;
    ***REMOVED*** else ***REMOVED***
        start = node.loc.start;
        end = getOpeningParenOfParams(node, sourceCode).loc.start;
    ***REMOVED***

    return ***REMOVED***
        start: Object.assign(***REMOVED******REMOVED***, start),
        end: Object.assign(***REMOVED******REMOVED***, end),
    ***REMOVED***
***REMOVED***

/* globals BigInt, globalThis, global, self, window */

const globalObject =
    typeof globalThis !== "undefined"
        ? globalThis
        : typeof self !== "undefined"
        ? self
        : typeof window !== "undefined"
        ? window
        : typeof global !== "undefined"
        ? global
        : ***REMOVED******REMOVED***;

const builtinNames = Object.freeze(
    new Set([
        "Array",
        "ArrayBuffer",
        "BigInt",
        "BigInt64Array",
        "BigUint64Array",
        "Boolean",
        "DataView",
        "Date",
        "decodeURI",
        "decodeURIComponent",
        "encodeURI",
        "encodeURIComponent",
        "escape",
        "Float32Array",
        "Float64Array",
        "Function",
        "Infinity",
        "Int16Array",
        "Int32Array",
        "Int8Array",
        "isFinite",
        "isNaN",
        "isPrototypeOf",
        "JSON",
        "Map",
        "Math",
        "NaN",
        "Number",
        "Object",
        "parseFloat",
        "parseInt",
        "Promise",
        "Proxy",
        "Reflect",
        "RegExp",
        "Set",
        "String",
        "Symbol",
        "Uint16Array",
        "Uint32Array",
        "Uint8Array",
        "Uint8ClampedArray",
        "undefined",
        "unescape",
        "WeakMap",
        "WeakSet",
    ])
);
const callAllowed = new Set(
    [
        Array.isArray,
        typeof BigInt === "function" ? BigInt : undefined,
        Boolean,
        Date,
        Date.parse,
        decodeURI,
        decodeURIComponent,
        encodeURI,
        encodeURIComponent,
        escape,
        isFinite,
        isNaN,
        isPrototypeOf,
        ...Object.getOwnPropertyNames(Math)
            .map(k => Math[k])
            .filter(f => typeof f === "function"),
        Number,
        Number.isFinite,
        Number.isNaN,
        Number.parseFloat,
        Number.parseInt,
        Object,
        Object.entries,
        Object.is,
        Object.isExtensible,
        Object.isFrozen,
        Object.isSealed,
        Object.keys,
        Object.values,
        parseFloat,
        parseInt,
        RegExp,
        String,
        String.fromCharCode,
        String.fromCodePoint,
        String.raw,
        Symbol,
        Symbol.for,
        Symbol.keyFor,
        unescape,
    ].filter(f => typeof f === "function")
);
const callPassThrough = new Set([
    Object.freeze,
    Object.preventExtensions,
    Object.seal,
]);

/**
 * Get the property descriptor.
 * @param ***REMOVED***object***REMOVED*** object The object to get.
 * @param ***REMOVED***string|number|symbol***REMOVED*** name The property name to get.
 */
function getPropertyDescriptor(object, name) ***REMOVED***
    let x = object;
    while ((typeof x === "object" || typeof x === "function") && x !== null) ***REMOVED***
        const d = Object.getOwnPropertyDescriptor(x, name);
        if (d) ***REMOVED***
            return d
        ***REMOVED***
        x = Object.getPrototypeOf(x);
    ***REMOVED***
    return null
***REMOVED***

/**
 * Check if a property is getter or not.
 * @param ***REMOVED***object***REMOVED*** object The object to check.
 * @param ***REMOVED***string|number|symbol***REMOVED*** name The property name to check.
 */
function isGetter(object, name) ***REMOVED***
    const d = getPropertyDescriptor(object, name);
    return d != null && d.get != null
***REMOVED***

/**
 * Get the element values of a given node list.
 * @param ***REMOVED***Node[]***REMOVED*** nodeList The node list to get values.
 * @param ***REMOVED***Scope|undefined***REMOVED*** initialScope The initial scope to find variables.
 * @returns ***REMOVED***any[]|null***REMOVED*** The value list if all nodes are constant. Otherwise, null.
 */
function getElementValues(nodeList, initialScope) ***REMOVED***
    const valueList = [];

    for (let i = 0; i < nodeList.length; ++i) ***REMOVED***
        const elementNode = nodeList[i];

        if (elementNode == null) ***REMOVED***
            valueList.length = i + 1;
        ***REMOVED*** else if (elementNode.type === "SpreadElement") ***REMOVED***
            const argument = getStaticValueR(elementNode.argument, initialScope);
            if (argument == null) ***REMOVED***
                return null
            ***REMOVED***
            valueList.push(...argument.value);
        ***REMOVED*** else ***REMOVED***
            const element = getStaticValueR(elementNode, initialScope);
            if (element == null) ***REMOVED***
                return null
            ***REMOVED***
            valueList.push(element.value);
        ***REMOVED***
    ***REMOVED***

    return valueList
***REMOVED***

const operations = Object.freeze(***REMOVED***
    ArrayExpression(node, initialScope) ***REMOVED***
        const elements = getElementValues(node.elements, initialScope);
        return elements != null ? ***REMOVED*** value: elements ***REMOVED*** : null
    ***REMOVED***,

    AssignmentExpression(node, initialScope) ***REMOVED***
        if (node.operator === "=") ***REMOVED***
            return getStaticValueR(node.right, initialScope)
        ***REMOVED***
        return null
    ***REMOVED***,

    //eslint-disable-next-line complexity
    BinaryExpression(node, initialScope) ***REMOVED***
        if (node.operator === "in" || node.operator === "instanceof") ***REMOVED***
            // Not supported.
            return null
        ***REMOVED***

        const left = getStaticValueR(node.left, initialScope);
        const right = getStaticValueR(node.right, initialScope);
        if (left != null && right != null) ***REMOVED***
            switch (node.operator) ***REMOVED***
                case "==":
                    return ***REMOVED*** value: left.value == right.value ***REMOVED*** //eslint-disable-line eqeqeq
                case "!=":
                    return ***REMOVED*** value: left.value != right.value ***REMOVED*** //eslint-disable-line eqeqeq
                case "===":
                    return ***REMOVED*** value: left.value === right.value ***REMOVED***
                case "!==":
                    return ***REMOVED*** value: left.value !== right.value ***REMOVED***
                case "<":
                    return ***REMOVED*** value: left.value < right.value ***REMOVED***
                case "<=":
                    return ***REMOVED*** value: left.value <= right.value ***REMOVED***
                case ">":
                    return ***REMOVED*** value: left.value > right.value ***REMOVED***
                case ">=":
                    return ***REMOVED*** value: left.value >= right.value ***REMOVED***
                case "<<":
                    return ***REMOVED*** value: left.value << right.value ***REMOVED***
                case ">>":
                    return ***REMOVED*** value: left.value >> right.value ***REMOVED***
                case ">>>":
                    return ***REMOVED*** value: left.value >>> right.value ***REMOVED***
                case "+":
                    return ***REMOVED*** value: left.value + right.value ***REMOVED***
                case "-":
                    return ***REMOVED*** value: left.value - right.value ***REMOVED***
                case "*":
                    return ***REMOVED*** value: left.value * right.value ***REMOVED***
                case "/":
                    return ***REMOVED*** value: left.value / right.value ***REMOVED***
                case "%":
                    return ***REMOVED*** value: left.value % right.value ***REMOVED***
                case "**":
                    return ***REMOVED*** value: Math.pow(left.value, right.value) ***REMOVED***
                case "|":
                    return ***REMOVED*** value: left.value | right.value ***REMOVED***
                case "^":
                    return ***REMOVED*** value: left.value ^ right.value ***REMOVED***
                case "&":
                    return ***REMOVED*** value: left.value & right.value ***REMOVED***

                // no default
            ***REMOVED***
        ***REMOVED***

        return null
    ***REMOVED***,

    CallExpression(node, initialScope) ***REMOVED***
        const calleeNode = node.callee;
        const args = getElementValues(node.arguments, initialScope);

        if (args != null) ***REMOVED***
            if (calleeNode.type === "MemberExpression") ***REMOVED***
                const object = getStaticValueR(calleeNode.object, initialScope);
                const property = calleeNode.computed
                    ? getStaticValueR(calleeNode.property, initialScope)
                    : ***REMOVED*** value: calleeNode.property.name ***REMOVED***;

                if (object != null && property != null) ***REMOVED***
                    const receiver = object.value;
                    const methodName = property.value;
                    if (callAllowed.has(receiver[methodName])) ***REMOVED***
                        return ***REMOVED*** value: receiver[methodName](...args) ***REMOVED***
                    ***REMOVED***
                    if (callPassThrough.has(receiver[methodName])) ***REMOVED***
                        return ***REMOVED*** value: args[0] ***REMOVED***
                    ***REMOVED***
                ***REMOVED***
            ***REMOVED*** else ***REMOVED***
                const callee = getStaticValueR(calleeNode, initialScope);
                if (callee != null) ***REMOVED***
                    const func = callee.value;
                    if (callAllowed.has(func)) ***REMOVED***
                        return ***REMOVED*** value: func(...args) ***REMOVED***
                    ***REMOVED***
                    if (callPassThrough.has(func)) ***REMOVED***
                        return ***REMOVED*** value: args[0] ***REMOVED***
                    ***REMOVED***
                ***REMOVED***
            ***REMOVED***
        ***REMOVED***

        return null
    ***REMOVED***,

    ConditionalExpression(node, initialScope) ***REMOVED***
        const test = getStaticValueR(node.test, initialScope);
        if (test != null) ***REMOVED***
            return test.value
                ? getStaticValueR(node.consequent, initialScope)
                : getStaticValueR(node.alternate, initialScope)
        ***REMOVED***
        return null
    ***REMOVED***,

    ExpressionStatement(node, initialScope) ***REMOVED***
        return getStaticValueR(node.expression, initialScope)
    ***REMOVED***,

    Identifier(node, initialScope) ***REMOVED***
        if (initialScope != null) ***REMOVED***
            const variable = findVariable(initialScope, node);

            // Built-in globals.
            if (
                variable != null &&
                variable.defs.length === 0 &&
                builtinNames.has(variable.name) &&
                variable.name in globalObject
            ) ***REMOVED***
                return ***REMOVED*** value: globalObject[variable.name] ***REMOVED***
            ***REMOVED***

            // Constants.
            if (variable != null && variable.defs.length === 1) ***REMOVED***
                const def = variable.defs[0];
                if (
                    def.parent &&
                    def.parent.kind === "const" &&
                    // TODO(mysticatea): don't support destructuring here.
                    def.node.id.type === "Identifier"
                ) ***REMOVED***
                    return getStaticValueR(def.node.init, initialScope)
                ***REMOVED***
            ***REMOVED***
        ***REMOVED***
        return null
    ***REMOVED***,

    Literal(node) ***REMOVED***
        //istanbul ignore if : this is implementation-specific behavior.
        if ((node.regex != null || node.bigint != null) && node.value == null) ***REMOVED***
            // It was a RegExp/BigInt literal, but Node.js didn't support it.
            return null
        ***REMOVED***
        return ***REMOVED*** value: node.value ***REMOVED***
    ***REMOVED***,

    LogicalExpression(node, initialScope) ***REMOVED***
        const left = getStaticValueR(node.left, initialScope);
        if (left != null) ***REMOVED***
            if (
                (node.operator === "||" && Boolean(left.value) === true) ||
                (node.operator === "&&" && Boolean(left.value) === false)
            ) ***REMOVED***
                return left
            ***REMOVED***

            const right = getStaticValueR(node.right, initialScope);
            if (right != null) ***REMOVED***
                return right
            ***REMOVED***
        ***REMOVED***

        return null
    ***REMOVED***,

    MemberExpression(node, initialScope) ***REMOVED***
        const object = getStaticValueR(node.object, initialScope);
        const property = node.computed
            ? getStaticValueR(node.property, initialScope)
            : ***REMOVED*** value: node.property.name ***REMOVED***;

        if (
            object != null &&
            property != null &&
            !isGetter(object.value, property.value)
        ) ***REMOVED***
            return ***REMOVED*** value: object.value[property.value] ***REMOVED***
        ***REMOVED***
        return null
    ***REMOVED***,

    NewExpression(node, initialScope) ***REMOVED***
        const callee = getStaticValueR(node.callee, initialScope);
        const args = getElementValues(node.arguments, initialScope);

        if (callee != null && args != null) ***REMOVED***
            const Func = callee.value;
            if (callAllowed.has(Func)) ***REMOVED***
                return ***REMOVED*** value: new Func(...args) ***REMOVED***
            ***REMOVED***
        ***REMOVED***

        return null
    ***REMOVED***,

    ObjectExpression(node, initialScope) ***REMOVED***
        const object = ***REMOVED******REMOVED***;

        for (const propertyNode of node.properties) ***REMOVED***
            if (propertyNode.type === "Property") ***REMOVED***
                if (propertyNode.kind !== "init") ***REMOVED***
                    return null
                ***REMOVED***
                const key = propertyNode.computed
                    ? getStaticValueR(propertyNode.key, initialScope)
                    : ***REMOVED*** value: propertyNode.key.name ***REMOVED***;
                const value = getStaticValueR(propertyNode.value, initialScope);
                if (key == null || value == null) ***REMOVED***
                    return null
                ***REMOVED***
                object[key.value] = value.value;
            ***REMOVED*** else if (
                propertyNode.type === "SpreadElement" ||
                propertyNode.type === "ExperimentalSpreadProperty"
            ) ***REMOVED***
                const argument = getStaticValueR(
                    propertyNode.argument,
                    initialScope
                );
                if (argument == null) ***REMOVED***
                    return null
                ***REMOVED***
                Object.assign(object, argument.value);
            ***REMOVED*** else ***REMOVED***
                return null
            ***REMOVED***
        ***REMOVED***

        return ***REMOVED*** value: object ***REMOVED***
    ***REMOVED***,

    SequenceExpression(node, initialScope) ***REMOVED***
        const last = node.expressions[node.expressions.length - 1];
        return getStaticValueR(last, initialScope)
    ***REMOVED***,

    TaggedTemplateExpression(node, initialScope) ***REMOVED***
        const tag = getStaticValueR(node.tag, initialScope);
        const expressions = getElementValues(
            node.quasi.expressions,
            initialScope
        );

        if (tag != null && expressions != null) ***REMOVED***
            const func = tag.value;
            const strings = node.quasi.quasis.map(q => q.value.cooked);
            strings.raw = node.quasi.quasis.map(q => q.value.raw);

            if (func === String.raw) ***REMOVED***
                return ***REMOVED*** value: func(strings, ...expressions) ***REMOVED***
            ***REMOVED***
        ***REMOVED***

        return null
    ***REMOVED***,

    TemplateLiteral(node, initialScope) ***REMOVED***
        const expressions = getElementValues(node.expressions, initialScope);
        if (expressions != null) ***REMOVED***
            let value = node.quasis[0].value.cooked;
            for (let i = 0; i < expressions.length; ++i) ***REMOVED***
                value += expressions[i];
                value += node.quasis[i + 1].value.cooked;
            ***REMOVED***
            return ***REMOVED*** value ***REMOVED***
        ***REMOVED***
        return null
    ***REMOVED***,

    UnaryExpression(node, initialScope) ***REMOVED***
        if (node.operator === "delete") ***REMOVED***
            // Not supported.
            return null
        ***REMOVED***
        if (node.operator === "void") ***REMOVED***
            return ***REMOVED*** value: undefined ***REMOVED***
        ***REMOVED***

        const arg = getStaticValueR(node.argument, initialScope);
        if (arg != null) ***REMOVED***
            switch (node.operator) ***REMOVED***
                case "-":
                    return ***REMOVED*** value: -arg.value ***REMOVED***
                case "+":
                    return ***REMOVED*** value: +arg.value ***REMOVED*** //eslint-disable-line no-implicit-coercion
                case "!":
                    return ***REMOVED*** value: !arg.value ***REMOVED***
                case "~":
                    return ***REMOVED*** value: ~arg.value ***REMOVED***
                case "typeof":
                    return ***REMOVED*** value: typeof arg.value ***REMOVED***

                // no default
            ***REMOVED***
        ***REMOVED***

        return null
    ***REMOVED***,
***REMOVED***);

/**
 * Get the value of a given node if it's a static value.
 * @param ***REMOVED***Node***REMOVED*** node The node to get.
 * @param ***REMOVED***Scope|undefined***REMOVED*** initialScope The scope to start finding variable.
 * @returns ***REMOVED******REMOVED***value:any***REMOVED***|null***REMOVED*** The static value of the node, or `null`.
 */
function getStaticValueR(node, initialScope) ***REMOVED***
    if (node != null && Object.hasOwnProperty.call(operations, node.type)) ***REMOVED***
        return operations[node.type](node, initialScope)
    ***REMOVED***
    return null
***REMOVED***

/**
 * Get the value of a given node if it's a static value.
 * @param ***REMOVED***Node***REMOVED*** node The node to get.
 * @param ***REMOVED***Scope***REMOVED*** [initialScope] The scope to start finding variable. Optional. If this scope was given, this tries to resolve identifier references which are in the given node as much as possible.
 * @returns ***REMOVED******REMOVED***value:any***REMOVED***|null***REMOVED*** The static value of the node, or `null`.
 */
function getStaticValue(node, initialScope = null) ***REMOVED***
    try ***REMOVED***
        return getStaticValueR(node, initialScope)
    ***REMOVED*** catch (_error) ***REMOVED***
        return null
    ***REMOVED***
***REMOVED***

/**
 * Get the value of a given node if it's a literal or a template literal.
 * @param ***REMOVED***Node***REMOVED*** node The node to get.
 * @param ***REMOVED***Scope***REMOVED*** [initialScope] The scope to start finding variable. Optional. If the node is an Identifier node and this scope was given, this checks the variable of the identifier, and returns the value of it if the variable is a constant.
 * @returns ***REMOVED***string|null***REMOVED*** The value of the node, or `null`.
 */
function getStringIfConstant(node, initialScope = null) ***REMOVED***
    // Handle the literals that the platform doesn't support natively.
    if (node && node.type === "Literal" && node.value === null) ***REMOVED***
        if (node.regex) ***REMOVED***
            return `/$***REMOVED***node.regex.pattern***REMOVED***/$***REMOVED***node.regex.flags***REMOVED***`
        ***REMOVED***
        if (node.bigint) ***REMOVED***
            return node.bigint
        ***REMOVED***
    ***REMOVED***

    const evaluated = getStaticValue(node, initialScope);
    return evaluated && String(evaluated.value)
***REMOVED***

/**
 * Get the property name from a MemberExpression node or a Property node.
 * @param ***REMOVED***Node***REMOVED*** node The node to get.
 * @param ***REMOVED***Scope***REMOVED*** [initialScope] The scope to start finding variable. Optional. If the node is a computed property node and this scope was given, this checks the computed property name by the `getStringIfConstant` function with the scope, and returns the value of it.
 * @returns ***REMOVED***string|null***REMOVED*** The property name of the node.
 */
function getPropertyName(node, initialScope) ***REMOVED***
    switch (node.type) ***REMOVED***
        case "MemberExpression":
            if (node.computed) ***REMOVED***
                return getStringIfConstant(node.property, initialScope)
            ***REMOVED***
            return node.property.name

        case "Property":
        case "MethodDefinition":
            if (node.computed) ***REMOVED***
                return getStringIfConstant(node.key, initialScope)
            ***REMOVED***
            if (node.key.type === "Literal") ***REMOVED***
                return String(node.key.value)
            ***REMOVED***
            return node.key.name

        // no default
    ***REMOVED***

    return null
***REMOVED***

/**
 * Get the name and kind of the given function node.
 * @param ***REMOVED***ASTNode***REMOVED*** node - The function node to get.
 * @returns ***REMOVED***string***REMOVED*** The name and kind of the function node.
 */
function getFunctionNameWithKind(node) ***REMOVED***
    const parent = node.parent;
    const tokens = [];

    if (parent.type === "MethodDefinition" && parent.static) ***REMOVED***
        tokens.push("static");
    ***REMOVED***
    if (node.async) ***REMOVED***
        tokens.push("async");
    ***REMOVED***
    if (node.generator) ***REMOVED***
        tokens.push("generator");
    ***REMOVED***

    if (node.type === "ArrowFunctionExpression") ***REMOVED***
        tokens.push("arrow", "function");
    ***REMOVED*** else if (
        parent.type === "Property" ||
        parent.type === "MethodDefinition"
    ) ***REMOVED***
        if (parent.kind === "constructor") ***REMOVED***
            return "constructor"
        ***REMOVED***
        if (parent.kind === "get") ***REMOVED***
            tokens.push("getter");
        ***REMOVED*** else if (parent.kind === "set") ***REMOVED***
            tokens.push("setter");
        ***REMOVED*** else ***REMOVED***
            tokens.push("method");
        ***REMOVED***
    ***REMOVED*** else ***REMOVED***
        tokens.push("function");
    ***REMOVED***

    if (node.id) ***REMOVED***
        tokens.push(`'$***REMOVED***node.id.name***REMOVED***'`);
    ***REMOVED*** else ***REMOVED***
        const name = getPropertyName(parent);

        if (name) ***REMOVED***
            tokens.push(`'$***REMOVED***name***REMOVED***'`);
        ***REMOVED***
    ***REMOVED***

    return tokens.join(" ")
***REMOVED***

const typeConversionBinaryOps = Object.freeze(
    new Set([
        "==",
        "!=",
        "<",
        "<=",
        ">",
        ">=",
        "<<",
        ">>",
        ">>>",
        "+",
        "-",
        "*",
        "/",
        "%",
        "|",
        "^",
        "&",
        "in",
    ])
);
const typeConversionUnaryOps = Object.freeze(new Set(["-", "+", "!", "~"]));
const visitor = Object.freeze(
    Object.assign(Object.create(null), ***REMOVED***
        $visit(node, options, visitorKeys) ***REMOVED***
            const ***REMOVED*** type ***REMOVED*** = node;

            if (typeof this[type] === "function") ***REMOVED***
                return this[type](node, options, visitorKeys)
            ***REMOVED***

            return this.$visitChildren(node, options, visitorKeys)
        ***REMOVED***,

        $visitChildren(node, options, visitorKeys) ***REMOVED***
            const ***REMOVED*** type ***REMOVED*** = node;

            for (const key of visitorKeys[type] || evk.getKeys(node)) ***REMOVED***
                const value = node[key];

                if (Array.isArray(value)) ***REMOVED***
                    for (const element of value) ***REMOVED***
                        if (
                            element &&
                            this.$visit(element, options, visitorKeys)
                        ) ***REMOVED***
                            return true
                        ***REMOVED***
                    ***REMOVED***
                ***REMOVED*** else if (value && this.$visit(value, options, visitorKeys)) ***REMOVED***
                    return true
                ***REMOVED***
            ***REMOVED***

            return false
        ***REMOVED***,

        ArrowFunctionExpression() ***REMOVED***
            return false
        ***REMOVED***,
        AssignmentExpression() ***REMOVED***
            return true
        ***REMOVED***,
        AwaitExpression() ***REMOVED***
            return true
        ***REMOVED***,
        BinaryExpression(node, options, visitorKeys) ***REMOVED***
            if (
                options.considerImplicitTypeConversion &&
                typeConversionBinaryOps.has(node.operator) &&
                (node.left.type !== "Literal" || node.right.type !== "Literal")
            ) ***REMOVED***
                return true
            ***REMOVED***
            return this.$visitChildren(node, options, visitorKeys)
        ***REMOVED***,
        CallExpression() ***REMOVED***
            return true
        ***REMOVED***,
        FunctionExpression() ***REMOVED***
            return false
        ***REMOVED***,
        ImportExpression() ***REMOVED***
            return true
        ***REMOVED***,
        MemberExpression(node, options, visitorKeys) ***REMOVED***
            if (options.considerGetters) ***REMOVED***
                return true
            ***REMOVED***
            if (
                options.considerImplicitTypeConversion &&
                node.computed &&
                node.property.type !== "Literal"
            ) ***REMOVED***
                return true
            ***REMOVED***
            return this.$visitChildren(node, options, visitorKeys)
        ***REMOVED***,
        MethodDefinition(node, options, visitorKeys) ***REMOVED***
            if (
                options.considerImplicitTypeConversion &&
                node.computed &&
                node.key.type !== "Literal"
            ) ***REMOVED***
                return true
            ***REMOVED***
            return this.$visitChildren(node, options, visitorKeys)
        ***REMOVED***,
        NewExpression() ***REMOVED***
            return true
        ***REMOVED***,
        Property(node, options, visitorKeys) ***REMOVED***
            if (
                options.considerImplicitTypeConversion &&
                node.computed &&
                node.key.type !== "Literal"
            ) ***REMOVED***
                return true
            ***REMOVED***
            return this.$visitChildren(node, options, visitorKeys)
        ***REMOVED***,
        UnaryExpression(node, options, visitorKeys) ***REMOVED***
            if (node.operator === "delete") ***REMOVED***
                return true
            ***REMOVED***
            if (
                options.considerImplicitTypeConversion &&
                typeConversionUnaryOps.has(node.operator) &&
                node.argument.type !== "Literal"
            ) ***REMOVED***
                return true
            ***REMOVED***
            return this.$visitChildren(node, options, visitorKeys)
        ***REMOVED***,
        UpdateExpression() ***REMOVED***
            return true
        ***REMOVED***,
        YieldExpression() ***REMOVED***
            return true
        ***REMOVED***,
    ***REMOVED***)
);

/**
 * Check whether a given node has any side effect or not.
 * @param ***REMOVED***Node***REMOVED*** node The node to get.
 * @param ***REMOVED***SourceCode***REMOVED*** sourceCode The source code object.
 * @param ***REMOVED***object***REMOVED*** [options] The option object.
 * @param ***REMOVED***boolean***REMOVED*** [options.considerGetters=false] If `true` then it considers member accesses as the node which has side effects.
 * @param ***REMOVED***boolean***REMOVED*** [options.considerImplicitTypeConversion=false] If `true` then it considers implicit type conversion as the node which has side effects.
 * @param ***REMOVED***object***REMOVED*** [options.visitorKeys=evk.KEYS] The keys to traverse nodes. Use `context.getSourceCode().visitorKeys`.
 * @returns ***REMOVED***boolean***REMOVED*** `true` if the node has a certain side effect.
 */
function hasSideEffect(
    node,
    sourceCode,
    ***REMOVED*** considerGetters = false, considerImplicitTypeConversion = false ***REMOVED*** = ***REMOVED******REMOVED***
) ***REMOVED***
    return visitor.$visit(
        node,
        ***REMOVED*** considerGetters, considerImplicitTypeConversion ***REMOVED***,
        sourceCode.visitorKeys || evk.KEYS
    )
***REMOVED***

/**
 * Get the left parenthesis of the parent node syntax if it exists.
 * E.g., `if (a) ***REMOVED******REMOVED***` then the `(`.
 * @param ***REMOVED***Node***REMOVED*** node The AST node to check.
 * @param ***REMOVED***SourceCode***REMOVED*** sourceCode The source code object to get tokens.
 * @returns ***REMOVED***Token|null***REMOVED*** The left parenthesis of the parent node syntax
 */
function getParentSyntaxParen(node, sourceCode) ***REMOVED***
    const parent = node.parent;

    switch (parent.type) ***REMOVED***
        case "CallExpression":
        case "NewExpression":
            if (parent.arguments.length === 1 && parent.arguments[0] === node) ***REMOVED***
                return sourceCode.getTokenAfter(
                    parent.callee,
                    isOpeningParenToken
                )
            ***REMOVED***
            return null

        case "DoWhileStatement":
            if (parent.test === node) ***REMOVED***
                return sourceCode.getTokenAfter(
                    parent.body,
                    isOpeningParenToken
                )
            ***REMOVED***
            return null

        case "IfStatement":
        case "WhileStatement":
            if (parent.test === node) ***REMOVED***
                return sourceCode.getFirstToken(parent, 1)
            ***REMOVED***
            return null

        case "ImportExpression":
            if (parent.source === node) ***REMOVED***
                return sourceCode.getFirstToken(parent, 1)
            ***REMOVED***
            return null

        case "SwitchStatement":
            if (parent.discriminant === node) ***REMOVED***
                return sourceCode.getFirstToken(parent, 1)
            ***REMOVED***
            return null

        case "WithStatement":
            if (parent.object === node) ***REMOVED***
                return sourceCode.getFirstToken(parent, 1)
            ***REMOVED***
            return null

        default:
            return null
    ***REMOVED***
***REMOVED***

/**
 * Check whether a given node is parenthesized or not.
 * @param ***REMOVED***number***REMOVED*** times The number of parantheses.
 * @param ***REMOVED***Node***REMOVED*** node The AST node to check.
 * @param ***REMOVED***SourceCode***REMOVED*** sourceCode The source code object to get tokens.
 * @returns ***REMOVED***boolean***REMOVED*** `true` if the node is parenthesized the given times.
 */
/**
 * Check whether a given node is parenthesized or not.
 * @param ***REMOVED***Node***REMOVED*** node The AST node to check.
 * @param ***REMOVED***SourceCode***REMOVED*** sourceCode The source code object to get tokens.
 * @returns ***REMOVED***boolean***REMOVED*** `true` if the node is parenthesized.
 */
function isParenthesized(
    timesOrNode,
    nodeOrSourceCode,
    optionalSourceCode
) ***REMOVED***
    let times, node, sourceCode, maybeLeftParen, maybeRightParen;
    if (typeof timesOrNode === "number") ***REMOVED***
        times = timesOrNode | 0;
        node = nodeOrSourceCode;
        sourceCode = optionalSourceCode;
        if (!(times >= 1)) ***REMOVED***
            throw new TypeError("'times' should be a positive integer.")
        ***REMOVED***
    ***REMOVED*** else ***REMOVED***
        times = 1;
        node = timesOrNode;
        sourceCode = nodeOrSourceCode;
    ***REMOVED***

    if (node == null) ***REMOVED***
        return false
    ***REMOVED***

    maybeLeftParen = maybeRightParen = node;
    do ***REMOVED***
        maybeLeftParen = sourceCode.getTokenBefore(maybeLeftParen);
        maybeRightParen = sourceCode.getTokenAfter(maybeRightParen);
    ***REMOVED*** while (
        maybeLeftParen != null &&
        maybeRightParen != null &&
        isOpeningParenToken(maybeLeftParen) &&
        isClosingParenToken(maybeRightParen) &&
        // Avoid false positive such as `if (a) ***REMOVED******REMOVED***`
        maybeLeftParen !== getParentSyntaxParen(node, sourceCode) &&
        --times > 0
    )

    return times === 0
***REMOVED***

/**
 * @author Toru Nagashima <https://github.com/mysticatea>
 * See LICENSE file in root directory for full license.
 */

const placeholder = /\$(?:[$&`']|[1-9][0-9]?)/gu;

/** @type ***REMOVED***WeakMap<PatternMatcher, ***REMOVED***pattern:RegExp,escaped:boolean***REMOVED***>***REMOVED*** */
const internal = new WeakMap();

/**
 * Check whether a given character is escaped or not.
 * @param ***REMOVED***string***REMOVED*** str The string to check.
 * @param ***REMOVED***number***REMOVED*** index The location of the character to check.
 * @returns ***REMOVED***boolean***REMOVED*** `true` if the character is escaped.
 */
function isEscaped(str, index) ***REMOVED***
    let escaped = false;
    for (let i = index - 1; i >= 0 && str.charCodeAt(i) === 0x5c; --i) ***REMOVED***
        escaped = !escaped;
    ***REMOVED***
    return escaped
***REMOVED***

/**
 * Replace a given string by a given matcher.
 * @param ***REMOVED***PatternMatcher***REMOVED*** matcher The pattern matcher.
 * @param ***REMOVED***string***REMOVED*** str The string to be replaced.
 * @param ***REMOVED***string***REMOVED*** replacement The new substring to replace each matched part.
 * @returns ***REMOVED***string***REMOVED*** The replaced string.
 */
function replaceS(matcher, str, replacement) ***REMOVED***
    const chunks = [];
    let index = 0;

    /** @type ***REMOVED***RegExpExecArray***REMOVED*** */
    let match = null;

    /**
     * @param ***REMOVED***string***REMOVED*** key The placeholder.
     * @returns ***REMOVED***string***REMOVED*** The replaced string.
     */
    function replacer(key) ***REMOVED***
        switch (key) ***REMOVED***
            case "$$":
                return "$"
            case "$&":
                return match[0]
            case "$`":
                return str.slice(0, match.index)
            case "$'":
                return str.slice(match.index + match[0].length)
            default: ***REMOVED***
                const i = key.slice(1);
                if (i in match) ***REMOVED***
                    return match[i]
                ***REMOVED***
                return key
            ***REMOVED***
        ***REMOVED***
    ***REMOVED***

    for (match of matcher.execAll(str)) ***REMOVED***
        chunks.push(str.slice(index, match.index));
        chunks.push(replacement.replace(placeholder, replacer));
        index = match.index + match[0].length;
    ***REMOVED***
    chunks.push(str.slice(index));

    return chunks.join("")
***REMOVED***

/**
 * Replace a given string by a given matcher.
 * @param ***REMOVED***PatternMatcher***REMOVED*** matcher The pattern matcher.
 * @param ***REMOVED***string***REMOVED*** str The string to be replaced.
 * @param ***REMOVED***(...strs[])=>string***REMOVED*** replace The function to replace each matched part.
 * @returns ***REMOVED***string***REMOVED*** The replaced string.
 */
function replaceF(matcher, str, replace) ***REMOVED***
    const chunks = [];
    let index = 0;

    for (const match of matcher.execAll(str)) ***REMOVED***
        chunks.push(str.slice(index, match.index));
        chunks.push(String(replace(...match, match.index, match.input)));
        index = match.index + match[0].length;
    ***REMOVED***
    chunks.push(str.slice(index));

    return chunks.join("")
***REMOVED***

/**
 * The class to find patterns as considering escape sequences.
 */
class PatternMatcher ***REMOVED***
    /**
     * Initialize this matcher.
     * @param ***REMOVED***RegExp***REMOVED*** pattern The pattern to match.
     * @param ***REMOVED******REMOVED***escaped:boolean***REMOVED******REMOVED*** options The options.
     */
    constructor(pattern, ***REMOVED*** escaped = false ***REMOVED*** = ***REMOVED******REMOVED***) ***REMOVED***
        if (!(pattern instanceof RegExp)) ***REMOVED***
            throw new TypeError("'pattern' should be a RegExp instance.")
        ***REMOVED***
        if (!pattern.flags.includes("g")) ***REMOVED***
            throw new Error("'pattern' should contains 'g' flag.")
        ***REMOVED***

        internal.set(this, ***REMOVED***
            pattern: new RegExp(pattern.source, pattern.flags),
            escaped: Boolean(escaped),
        ***REMOVED***);
    ***REMOVED***

    /**
     * Find the pattern in a given string.
     * @param ***REMOVED***string***REMOVED*** str The string to find.
     * @returns ***REMOVED***IterableIterator<RegExpExecArray>***REMOVED*** The iterator which iterate the matched information.
     */
    *execAll(str) ***REMOVED***
        const ***REMOVED*** pattern, escaped ***REMOVED*** = internal.get(this);
        let match = null;
        let lastIndex = 0;

        pattern.lastIndex = 0;
        while ((match = pattern.exec(str)) != null) ***REMOVED***
            if (escaped || !isEscaped(str, match.index)) ***REMOVED***
                lastIndex = pattern.lastIndex;
                yield match;
                pattern.lastIndex = lastIndex;
            ***REMOVED***
        ***REMOVED***
    ***REMOVED***

    /**
     * Check whether the pattern is found in a given string.
     * @param ***REMOVED***string***REMOVED*** str The string to check.
     * @returns ***REMOVED***boolean***REMOVED*** `true` if the pattern was found in the string.
     */
    test(str) ***REMOVED***
        const it = this.execAll(str);
        const ret = it.next();
        return !ret.done
    ***REMOVED***

    /**
     * Replace a given string.
     * @param ***REMOVED***string***REMOVED*** str The string to be replaced.
     * @param ***REMOVED***(string|((...strs:string[])=>string))***REMOVED*** replacer The string or function to replace. This is the same as the 2nd argument of `String.prototype.replace`.
     * @returns ***REMOVED***string***REMOVED*** The replaced string.
     */
    [Symbol.replace](str, replacer) ***REMOVED***
        return typeof replacer === "function"
            ? replaceF(this, String(str), replacer)
            : replaceS(this, String(str), String(replacer))
    ***REMOVED***
***REMOVED***

const IMPORT_TYPE = /^(?:Import|Export(?:All|Default|Named))Declaration$/u;
const has = Function.call.bind(Object.hasOwnProperty);

const READ = Symbol("read");
const CALL = Symbol("call");
const CONSTRUCT = Symbol("construct");
const ESM = Symbol("esm");

const requireCall = ***REMOVED*** require: ***REMOVED*** [CALL]: true ***REMOVED*** ***REMOVED***;

/**
 * Check whether a given variable is modified or not.
 * @param ***REMOVED***Variable***REMOVED*** variable The variable to check.
 * @returns ***REMOVED***boolean***REMOVED*** `true` if the variable is modified.
 */
function isModifiedGlobal(variable) ***REMOVED***
    return (
        variable == null ||
        variable.defs.length !== 0 ||
        variable.references.some(r => r.isWrite())
    )
***REMOVED***

/**
 * Check if the value of a given node is passed through to the parent syntax as-is.
 * For example, `a` and `b` in (`a || b` and `c ? a : b`) are passed through.
 * @param ***REMOVED***Node***REMOVED*** node A node to check.
 * @returns ***REMOVED***boolean***REMOVED*** `true` if the node is passed through.
 */
function isPassThrough(node) ***REMOVED***
    const parent = node.parent;

    switch (parent && parent.type) ***REMOVED***
        case "ConditionalExpression":
            return parent.consequent === node || parent.alternate === node
        case "LogicalExpression":
            return true
        case "SequenceExpression":
            return parent.expressions[parent.expressions.length - 1] === node

        default:
            return false
    ***REMOVED***
***REMOVED***

/**
 * The reference tracker.
 */
class ReferenceTracker ***REMOVED***
    /**
     * Initialize this tracker.
     * @param ***REMOVED***Scope***REMOVED*** globalScope The global scope.
     * @param ***REMOVED***object***REMOVED*** [options] The options.
     * @param ***REMOVED***"legacy"|"strict"***REMOVED*** [options.mode="strict"] The mode to determine the ImportDeclaration's behavior for CJS modules.
     * @param ***REMOVED***string[]***REMOVED*** [options.globalObjectNames=["global","self","window"]] The variable names for Global Object.
     */
    constructor(
        globalScope,
        ***REMOVED***
            mode = "strict",
            globalObjectNames = ["global", "self", "window"],
        ***REMOVED*** = ***REMOVED******REMOVED***
    ) ***REMOVED***
        this.variableStack = [];
        this.globalScope = globalScope;
        this.mode = mode;
        this.globalObjectNames = globalObjectNames.slice(0);
    ***REMOVED***

    /**
     * Iterate the references of global variables.
     * @param ***REMOVED***object***REMOVED*** traceMap The trace map.
     * @returns ***REMOVED***IterableIterator<***REMOVED***node:Node,path:string[],type:symbol,info:any***REMOVED***>***REMOVED*** The iterator to iterate references.
     */
    *iterateGlobalReferences(traceMap) ***REMOVED***
        for (const key of Object.keys(traceMap)) ***REMOVED***
            const nextTraceMap = traceMap[key];
            const path = [key];
            const variable = this.globalScope.set.get(key);

            if (isModifiedGlobal(variable)) ***REMOVED***
                continue
            ***REMOVED***

            yield* this._iterateVariableReferences(
                variable,
                path,
                nextTraceMap,
                true
            );
        ***REMOVED***

        for (const key of this.globalObjectNames) ***REMOVED***
            const path = [];
            const variable = this.globalScope.set.get(key);

            if (isModifiedGlobal(variable)) ***REMOVED***
                continue
            ***REMOVED***

            yield* this._iterateVariableReferences(
                variable,
                path,
                traceMap,
                false
            );
        ***REMOVED***
    ***REMOVED***

    /**
     * Iterate the references of CommonJS modules.
     * @param ***REMOVED***object***REMOVED*** traceMap The trace map.
     * @returns ***REMOVED***IterableIterator<***REMOVED***node:Node,path:string[],type:symbol,info:any***REMOVED***>***REMOVED*** The iterator to iterate references.
     */
    *iterateCjsReferences(traceMap) ***REMOVED***
        for (const ***REMOVED*** node ***REMOVED*** of this.iterateGlobalReferences(requireCall)) ***REMOVED***
            const key = getStringIfConstant(node.arguments[0]);
            if (key == null || !has(traceMap, key)) ***REMOVED***
                continue
            ***REMOVED***

            const nextTraceMap = traceMap[key];
            const path = [key];

            if (nextTraceMap[READ]) ***REMOVED***
                yield ***REMOVED***
                    node,
                    path,
                    type: READ,
                    info: nextTraceMap[READ],
                ***REMOVED***;
            ***REMOVED***
            yield* this._iteratePropertyReferences(node, path, nextTraceMap);
        ***REMOVED***
    ***REMOVED***

    /**
     * Iterate the references of ES modules.
     * @param ***REMOVED***object***REMOVED*** traceMap The trace map.
     * @returns ***REMOVED***IterableIterator<***REMOVED***node:Node,path:string[],type:symbol,info:any***REMOVED***>***REMOVED*** The iterator to iterate references.
     */
    *iterateEsmReferences(traceMap) ***REMOVED***
        const programNode = this.globalScope.block;

        for (const node of programNode.body) ***REMOVED***
            if (!IMPORT_TYPE.test(node.type) || node.source == null) ***REMOVED***
                continue
            ***REMOVED***
            const moduleId = node.source.value;

            if (!has(traceMap, moduleId)) ***REMOVED***
                continue
            ***REMOVED***
            const nextTraceMap = traceMap[moduleId];
            const path = [moduleId];

            if (nextTraceMap[READ]) ***REMOVED***
                yield ***REMOVED*** node, path, type: READ, info: nextTraceMap[READ] ***REMOVED***;
            ***REMOVED***

            if (node.type === "ExportAllDeclaration") ***REMOVED***
                for (const key of Object.keys(nextTraceMap)) ***REMOVED***
                    const exportTraceMap = nextTraceMap[key];
                    if (exportTraceMap[READ]) ***REMOVED***
                        yield ***REMOVED***
                            node,
                            path: path.concat(key),
                            type: READ,
                            info: exportTraceMap[READ],
                        ***REMOVED***;
                    ***REMOVED***
                ***REMOVED***
            ***REMOVED*** else ***REMOVED***
                for (const specifier of node.specifiers) ***REMOVED***
                    const esm = has(nextTraceMap, ESM);
                    const it = this._iterateImportReferences(
                        specifier,
                        path,
                        esm
                            ? nextTraceMap
                            : this.mode === "legacy"
                            ? Object.assign(
                                  ***REMOVED*** default: nextTraceMap ***REMOVED***,
                                  nextTraceMap
                              )
                            : ***REMOVED*** default: nextTraceMap ***REMOVED***
                    );

                    if (esm) ***REMOVED***
                        yield* it;
                    ***REMOVED*** else ***REMOVED***
                        for (const report of it) ***REMOVED***
                            report.path = report.path.filter(exceptDefault);
                            if (
                                report.path.length >= 2 ||
                                report.type !== READ
                            ) ***REMOVED***
                                yield report;
                            ***REMOVED***
                        ***REMOVED***
                    ***REMOVED***
                ***REMOVED***
            ***REMOVED***
        ***REMOVED***
    ***REMOVED***

    /**
     * Iterate the references for a given variable.
     * @param ***REMOVED***Variable***REMOVED*** variable The variable to iterate that references.
     * @param ***REMOVED***string[]***REMOVED*** path The current path.
     * @param ***REMOVED***object***REMOVED*** traceMap The trace map.
     * @param ***REMOVED***boolean***REMOVED*** shouldReport = The flag to report those references.
     * @returns ***REMOVED***IterableIterator<***REMOVED***node:Node,path:string[],type:symbol,info:any***REMOVED***>***REMOVED*** The iterator to iterate references.
     */
    *_iterateVariableReferences(variable, path, traceMap, shouldReport) ***REMOVED***
        if (this.variableStack.includes(variable)) ***REMOVED***
            return
        ***REMOVED***
        this.variableStack.push(variable);
        try ***REMOVED***
            for (const reference of variable.references) ***REMOVED***
                if (!reference.isRead()) ***REMOVED***
                    continue
                ***REMOVED***
                const node = reference.identifier;

                if (shouldReport && traceMap[READ]) ***REMOVED***
                    yield ***REMOVED*** node, path, type: READ, info: traceMap[READ] ***REMOVED***;
                ***REMOVED***
                yield* this._iteratePropertyReferences(node, path, traceMap);
            ***REMOVED***
        ***REMOVED*** finally ***REMOVED***
            this.variableStack.pop();
        ***REMOVED***
    ***REMOVED***

    /**
     * Iterate the references for a given AST node.
     * @param rootNode The AST node to iterate references.
     * @param ***REMOVED***string[]***REMOVED*** path The current path.
     * @param ***REMOVED***object***REMOVED*** traceMap The trace map.
     * @returns ***REMOVED***IterableIterator<***REMOVED***node:Node,path:string[],type:symbol,info:any***REMOVED***>***REMOVED*** The iterator to iterate references.
     */
    //eslint-disable-next-line complexity
    *_iteratePropertyReferences(rootNode, path, traceMap) ***REMOVED***
        let node = rootNode;
        while (isPassThrough(node)) ***REMOVED***
            node = node.parent;
        ***REMOVED***

        const parent = node.parent;
        if (parent.type === "MemberExpression") ***REMOVED***
            if (parent.object === node) ***REMOVED***
                const key = getPropertyName(parent);
                if (key == null || !has(traceMap, key)) ***REMOVED***
                    return
                ***REMOVED***

                path = path.concat(key); //eslint-disable-line no-param-reassign
                const nextTraceMap = traceMap[key];
                if (nextTraceMap[READ]) ***REMOVED***
                    yield ***REMOVED***
                        node: parent,
                        path,
                        type: READ,
                        info: nextTraceMap[READ],
                    ***REMOVED***;
                ***REMOVED***
                yield* this._iteratePropertyReferences(
                    parent,
                    path,
                    nextTraceMap
                );
            ***REMOVED***
            return
        ***REMOVED***
        if (parent.type === "CallExpression") ***REMOVED***
            if (parent.callee === node && traceMap[CALL]) ***REMOVED***
                yield ***REMOVED*** node: parent, path, type: CALL, info: traceMap[CALL] ***REMOVED***;
            ***REMOVED***
            return
        ***REMOVED***
        if (parent.type === "NewExpression") ***REMOVED***
            if (parent.callee === node && traceMap[CONSTRUCT]) ***REMOVED***
                yield ***REMOVED***
                    node: parent,
                    path,
                    type: CONSTRUCT,
                    info: traceMap[CONSTRUCT],
                ***REMOVED***;
            ***REMOVED***
            return
        ***REMOVED***
        if (parent.type === "AssignmentExpression") ***REMOVED***
            if (parent.right === node) ***REMOVED***
                yield* this._iterateLhsReferences(parent.left, path, traceMap);
                yield* this._iteratePropertyReferences(parent, path, traceMap);
            ***REMOVED***
            return
        ***REMOVED***
        if (parent.type === "AssignmentPattern") ***REMOVED***
            if (parent.right === node) ***REMOVED***
                yield* this._iterateLhsReferences(parent.left, path, traceMap);
            ***REMOVED***
            return
        ***REMOVED***
        if (parent.type === "VariableDeclarator") ***REMOVED***
            if (parent.init === node) ***REMOVED***
                yield* this._iterateLhsReferences(parent.id, path, traceMap);
            ***REMOVED***
        ***REMOVED***
    ***REMOVED***

    /**
     * Iterate the references for a given Pattern node.
     * @param ***REMOVED***Node***REMOVED*** patternNode The Pattern node to iterate references.
     * @param ***REMOVED***string[]***REMOVED*** path The current path.
     * @param ***REMOVED***object***REMOVED*** traceMap The trace map.
     * @returns ***REMOVED***IterableIterator<***REMOVED***node:Node,path:string[],type:symbol,info:any***REMOVED***>***REMOVED*** The iterator to iterate references.
     */
    *_iterateLhsReferences(patternNode, path, traceMap) ***REMOVED***
        if (patternNode.type === "Identifier") ***REMOVED***
            const variable = findVariable(this.globalScope, patternNode);
            if (variable != null) ***REMOVED***
                yield* this._iterateVariableReferences(
                    variable,
                    path,
                    traceMap,
                    false
                );
            ***REMOVED***
            return
        ***REMOVED***
        if (patternNode.type === "ObjectPattern") ***REMOVED***
            for (const property of patternNode.properties) ***REMOVED***
                const key = getPropertyName(property);

                if (key == null || !has(traceMap, key)) ***REMOVED***
                    continue
                ***REMOVED***

                const nextPath = path.concat(key);
                const nextTraceMap = traceMap[key];
                if (nextTraceMap[READ]) ***REMOVED***
                    yield ***REMOVED***
                        node: property,
                        path: nextPath,
                        type: READ,
                        info: nextTraceMap[READ],
                    ***REMOVED***;
                ***REMOVED***
                yield* this._iterateLhsReferences(
                    property.value,
                    nextPath,
                    nextTraceMap
                );
            ***REMOVED***
            return
        ***REMOVED***
        if (patternNode.type === "AssignmentPattern") ***REMOVED***
            yield* this._iterateLhsReferences(patternNode.left, path, traceMap);
        ***REMOVED***
    ***REMOVED***

    /**
     * Iterate the references for a given ModuleSpecifier node.
     * @param ***REMOVED***Node***REMOVED*** specifierNode The ModuleSpecifier node to iterate references.
     * @param ***REMOVED***string[]***REMOVED*** path The current path.
     * @param ***REMOVED***object***REMOVED*** traceMap The trace map.
     * @returns ***REMOVED***IterableIterator<***REMOVED***node:Node,path:string[],type:symbol,info:any***REMOVED***>***REMOVED*** The iterator to iterate references.
     */
    *_iterateImportReferences(specifierNode, path, traceMap) ***REMOVED***
        const type = specifierNode.type;

        if (type === "ImportSpecifier" || type === "ImportDefaultSpecifier") ***REMOVED***
            const key =
                type === "ImportDefaultSpecifier"
                    ? "default"
                    : specifierNode.imported.name;
            if (!has(traceMap, key)) ***REMOVED***
                return
            ***REMOVED***

            path = path.concat(key); //eslint-disable-line no-param-reassign
            const nextTraceMap = traceMap[key];
            if (nextTraceMap[READ]) ***REMOVED***
                yield ***REMOVED***
                    node: specifierNode,
                    path,
                    type: READ,
                    info: nextTraceMap[READ],
                ***REMOVED***;
            ***REMOVED***
            yield* this._iterateVariableReferences(
                findVariable(this.globalScope, specifierNode.local),
                path,
                nextTraceMap,
                false
            );

            return
        ***REMOVED***

        if (type === "ImportNamespaceSpecifier") ***REMOVED***
            yield* this._iterateVariableReferences(
                findVariable(this.globalScope, specifierNode.local),
                path,
                traceMap,
                false
            );
            return
        ***REMOVED***

        if (type === "ExportSpecifier") ***REMOVED***
            const key = specifierNode.local.name;
            if (!has(traceMap, key)) ***REMOVED***
                return
            ***REMOVED***

            path = path.concat(key); //eslint-disable-line no-param-reassign
            const nextTraceMap = traceMap[key];
            if (nextTraceMap[READ]) ***REMOVED***
                yield ***REMOVED***
                    node: specifierNode,
                    path,
                    type: READ,
                    info: nextTraceMap[READ],
                ***REMOVED***;
            ***REMOVED***
        ***REMOVED***
    ***REMOVED***
***REMOVED***

ReferenceTracker.READ = READ;
ReferenceTracker.CALL = CALL;
ReferenceTracker.CONSTRUCT = CONSTRUCT;
ReferenceTracker.ESM = ESM;

/**
 * This is a predicate function for Array#filter.
 * @param ***REMOVED***string***REMOVED*** name A name part.
 * @param ***REMOVED***number***REMOVED*** index The index of the name.
 * @returns ***REMOVED***boolean***REMOVED*** `false` if it's default.
 */
function exceptDefault(name, index) ***REMOVED***
    return !(index === 1 && name === "default")
***REMOVED***

var index = ***REMOVED***
    CALL,
    CONSTRUCT,
    ESM,
    findVariable,
    getFunctionHeadLocation,
    getFunctionNameWithKind,
    getInnermostScope,
    getPropertyName,
    getStaticValue,
    getStringIfConstant,
    hasSideEffect,
    isArrowToken,
    isClosingBraceToken,
    isClosingBracketToken,
    isClosingParenToken,
    isColonToken,
    isCommaToken,
    isCommentToken,
    isNotArrowToken,
    isNotClosingBraceToken,
    isNotClosingBracketToken,
    isNotClosingParenToken,
    isNotColonToken,
    isNotCommaToken,
    isNotCommentToken,
    isNotOpeningBraceToken,
    isNotOpeningBracketToken,
    isNotOpeningParenToken,
    isNotSemicolonToken,
    isOpeningBraceToken,
    isOpeningBracketToken,
    isOpeningParenToken,
    isParenthesized,
    isSemicolonToken,
    PatternMatcher,
    READ,
    ReferenceTracker,
***REMOVED***;

export default index;
export ***REMOVED*** CALL, CONSTRUCT, ESM, PatternMatcher, READ, ReferenceTracker, findVariable, getFunctionHeadLocation, getFunctionNameWithKind, getInnermostScope, getPropertyName, getStaticValue, getStringIfConstant, hasSideEffect, isArrowToken, isClosingBraceToken, isClosingBracketToken, isClosingParenToken, isColonToken, isCommaToken, isCommentToken, isNotArrowToken, isNotClosingBraceToken, isNotClosingBracketToken, isNotClosingParenToken, isNotColonToken, isNotCommaToken, isNotCommentToken, isNotOpeningBraceToken, isNotOpeningBracketToken, isNotOpeningParenToken, isNotSemicolonToken, isOpeningBraceToken, isOpeningBracketToken, isOpeningParenToken, isParenthesized, isSemicolonToken ***REMOVED***;
