// discord.js/src/client/voice/receiver/VoiceReceiver.js
const EventEmitter = require('events').EventEmitter;
const secretbox = require('../util/Secretbox');
const Readable = require('./VoiceReadable');
const OpusEncoders = require('../opus/OpusEngineList');

const nonce = Buffer.alloc(24);
nonce.fill(0);

/**
 * Receives voice data from a voice connection.
 * ```js
 * // Obtained using:
 * voiceChannel.join()
 *   .then(connection => ***REMOVED***
 *     const receiver = connection.createReceiver();
 *   ***REMOVED***);
 * ```
 * @extends ***REMOVED***EventEmitter***REMOVED***
 */
class VoiceReceiver extends EventEmitter ***REMOVED***
  constructor(connection) ***REMOVED***
    super();
    /*
      Need a queue because we don't get the ssrc of the user speaking until after the first few packets,
      so we queue up unknown SSRCs until they become known, then empty the queue
    */
    this.queues = new Map();
    this.pcmStreams = new Map();
    this.opusStreams = new Map();
    this.opusEncoders = new Map();

    /**
     * Whether or not this receiver has been destroyed
     * @type ***REMOVED***boolean***REMOVED***
     */
    this.destroyed = false;

    /**
     * The VoiceConnection that instantiated this
     * @type ***REMOVED***VoiceConnection***REMOVED***
     */
    this.voiceConnection = connection;

    this._listener = msg => ***REMOVED***
      const ssrc = +msg.readUInt32BE(8).toString(10);
      const user = connection.client.users.get(connection.ssrcMap.get(ssrc));
      if (!user) ***REMOVED***
        if (!this.queues.has(ssrc)) this.queues.set(ssrc, []);
        this.queues.get(ssrc).push(msg);
      ***REMOVED*** else ***REMOVED***
        if (this.queues.get(ssrc)) ***REMOVED***
          this.queues.get(ssrc).push(msg);
          this.queues.get(ssrc).map(m => this.handlePacket(m, user));
          this.queues.delete(ssrc);
          return;
        ***REMOVED***
        this.handlePacket(msg, user);
      ***REMOVED***
    ***REMOVED***;
    this.voiceConnection.sockets.udp.socket.on('message', this._listener);
  ***REMOVED***

  /**
   * If this VoiceReceiver has been destroyed, running `recreate()` will recreate the listener.
   * This avoids you having to create a new receiver.
   * <info>Any streams that you had prior to destroying the receiver will not be recreated.</info>
   */
  recreate() ***REMOVED***
    if (!this.destroyed) return;
    this.voiceConnection.sockets.udp.socket.on('message', this._listener);
    this.destroyed = false;
  ***REMOVED***

  /**
   * Destroy this VoiceReceiver, also ending any streams that it may be controlling.
   */
  destroy() ***REMOVED***
    this.voiceConnection.sockets.udp.socket.removeListener('message', this._listener);
    for (const [id, stream] of this.pcmStreams) ***REMOVED***
      stream._push(null);
      this.pcmStreams.delete(id);
    ***REMOVED***
    for (const [id, stream] of this.opusStreams) ***REMOVED***
      stream._push(null);
      this.opusStreams.delete(id);
    ***REMOVED***
    for (const [id, encoder] of this.opusEncoders) ***REMOVED***
      encoder.destroy();
      this.opusEncoders.delete(id);
    ***REMOVED***
    this.destroyed = true;
  ***REMOVED***

  /**
   * Invoked when a user stops speaking.
   * @param ***REMOVED***User***REMOVED*** user The user that stopped speaking
   * @private
   */
  stoppedSpeaking(user) ***REMOVED***
    const opusStream = this.opusStreams.get(user.id);
    const pcmStream = this.pcmStreams.get(user.id);
    const opusEncoder = this.opusEncoders.get(user.id);
    if (opusStream) ***REMOVED***
      opusStream.push(null);
      opusStream.open = false;
      this.opusStreams.delete(user.id);
    ***REMOVED***
    if (pcmStream) ***REMOVED***
      pcmStream.push(null);
      pcmStream.open = false;
      this.pcmStreams.delete(user.id);
    ***REMOVED***
    if (opusEncoder) ***REMOVED***
      opusEncoder.destroy();
      this.opusEncoders.delete(user.id);
    ***REMOVED***
  ***REMOVED***

  /**
   * Creates a readable stream for a user that provides opus data while the user is speaking. When the user
   * stops speaking, the stream is destroyed.
   * @param ***REMOVED***UserResolvable***REMOVED*** user The user to create the stream for
   * @returns ***REMOVED***ReadableStream***REMOVED***
   */
  createOpusStream(user) ***REMOVED***
    user = this.voiceConnection.voiceManager.client.resolver.resolveUser(user);
    if (!user) throw new Error('Couldn\'t resolve the user to create Opus stream.');
    if (this.opusStreams.get(user.id)) throw new Error('There is already an existing stream for that user.');
    const stream = new Readable();
    this.opusStreams.set(user.id, stream);
    return stream;
  ***REMOVED***

  /**
   * Creates a readable stream for a user that provides PCM data while the user is speaking. When the user
   * stops speaking, the stream is destroyed. The stream is 16-bit signed stereo PCM at 48KHz.
   * @param ***REMOVED***UserResolvable***REMOVED*** user The user to create the stream for
   * @returns ***REMOVED***ReadableStream***REMOVED***
   */
  createPCMStream(user) ***REMOVED***
    user = this.voiceConnection.voiceManager.client.resolver.resolveUser(user);
    if (!user) throw new Error('Couldn\'t resolve the user to create PCM stream.');
    if (this.pcmStreams.get(user.id)) throw new Error('There is already an existing stream for that user.');
    const stream = new Readable();
    this.pcmStreams.set(user.id, stream);
    return stream;
  ***REMOVED***

  handlePacket(msg, user) ***REMOVED***
    msg.copy(nonce, 0, 0, 12);
    let data = secretbox.methods.open(msg.slice(12), nonce, this.voiceConnection.authentication.secretKey.key);
    if (!data) ***REMOVED***
      /**
       * Emitted whenever a voice packet experiences a problem.
       * @event VoiceReceiver#warn
       * @param ***REMOVED***string***REMOVED*** reason The reason for the warning. If it happened because the voice packet could not be
       * decrypted, this would be `decrypt`. If it happened because the voice packet could not be decoded into
       * PCM, this would be `decode`
       * @param ***REMOVED***string***REMOVED*** message The warning message
       */
      this.emit('warn', 'decrypt', 'Failed to decrypt voice packet');
      return;
    ***REMOVED***
    data = Buffer.from(data);

    // Strip RTP Header Extensions (one-byte only)
    if (data[0] === 0xBE && data[1] === 0xDE && data.length > 4) ***REMOVED***
      const headerExtensionLength = data.readUInt16BE(2);
      let offset = 4;
      for (let i = 0; i < headerExtensionLength; i++) ***REMOVED***
        const byte = data[offset];
        offset++;
        if (byte === 0) ***REMOVED***
          continue;
        ***REMOVED***
        offset += 1 + (0b1111 & (byte >> 4));
      ***REMOVED***
      // Skip over undocumented Discord byte
      offset++;

      data = data.slice(offset);
    ***REMOVED***

    if (this.opusStreams.get(user.id)) this.opusStreams.get(user.id)._push(data);
    /**
     * Emitted whenever voice data is received from the voice connection. This is _always_ emitted (unlike PCM).
     * @event VoiceReceiver#opus
     * @param ***REMOVED***User***REMOVED*** user The user that is sending the buffer (is speaking)
     * @param ***REMOVED***Buffer***REMOVED*** buffer The opus buffer
     */
    this.emit('opus', user, data);
    if (this.listenerCount('pcm') > 0 || this.pcmStreams.size > 0) ***REMOVED***
      if (!this.opusEncoders.get(user.id)) this.opusEncoders.set(user.id, OpusEncoders.fetch());
      const ***REMOVED*** pcm, error ***REMOVED*** = VoiceReceiver._tryDecode(this.opusEncoders.get(user.id), data);
      if (error) ***REMOVED***
        this.emit('warn', 'decode', `Failed to decode packet voice to PCM because: $***REMOVED***error.message***REMOVED***`);
        return;
      ***REMOVED***
      if (this.pcmStreams.get(user.id)) this.pcmStreams.get(user.id)._push(pcm);
      /**
       * Emits decoded voice data when it's received. For performance reasons, the decoding will only
       * happen if there is at least one `pcm` listener on this receiver.
       * @event VoiceReceiver#pcm
       * @param ***REMOVED***User***REMOVED*** user The user that is sending the buffer (is speaking)
       * @param ***REMOVED***Buffer***REMOVED*** buffer The decoded buffer
       */
      this.emit('pcm', user, pcm);
    ***REMOVED***
  ***REMOVED***

  static _tryDecode(encoder, data) ***REMOVED***
    try ***REMOVED***
      return ***REMOVED*** pcm: encoder.decode(data) ***REMOVED***;
    ***REMOVED*** catch (error) ***REMOVED***
      return ***REMOVED*** error ***REMOVED***;
    ***REMOVED***
  ***REMOVED***
***REMOVED***

module.exports = VoiceReceiver;
