// discord.js/src/client/websocket/packets/handlers/Ready.js
const AbstractHandler = require('./AbstractHandler');

const ClientUser = require('../../../../structures/ClientUser');

class ReadyHandler extends AbstractHandler ***REMOVED***
  handle(packet) ***REMOVED***
    const client = this.packetManager.client;
    const data = packet.d;

    client.ws.heartbeat();

    data.user.user_settings = data.user_settings;
    data.user.user_guild_settings = data.user_guild_settings;

    const clientUser = new ClientUser(client, data.user);
    client.user = clientUser;
    client.readyAt = new Date();
    client.users.set(clientUser.id, clientUser);

    for (const guild of data.guilds) if (!client.guilds.has(guild.id)) client.dataManager.newGuild(guild);
    for (const privateDM of data.private_channels) client.dataManager.newChannel(privateDM);

    for (const relation of data.relationships) ***REMOVED***
      const user = client.dataManager.newUser(relation.user);
      if (relation.type === 1) ***REMOVED***
        client.user.friends.set(user.id, user);
      ***REMOVED*** else if (relation.type === 2) ***REMOVED***
        client.user.blocked.set(user.id, user);
      ***REMOVED***
    ***REMOVED***

    data.presences = data.presences || [];
    for (const presence of data.presences) ***REMOVED***
      client.dataManager.newUser(presence.user);
      client._setPresence(presence.user.id, presence);
    ***REMOVED***

    if (data.notes) ***REMOVED***
      for (const user of Object.keys(data.notes)) ***REMOVED***
        let note = data.notes[user];
        if (!note.length) note = null;

        client.user.notes.set(user, note);
      ***REMOVED***
    ***REMOVED***

    if (!client.user.bot && client.options.sync) client.setInterval(client.syncGuilds.bind(client), 30000);

    if (!client.users.has('1')) ***REMOVED***
      client.dataManager.newUser(***REMOVED***
        id: '1',
        username: 'Clyde',
        discriminator: '0000',
        avatar: 'https://discordapp.com/assets/f78426a064bc9dd24847519259bc42af.png',
        bot: true,
        status: 'online',
        game: null,
        verified: true,
      ***REMOVED***);
    ***REMOVED***

    const t = client.setTimeout(() => ***REMOVED***
      client.ws.connection.triggerReady();
    ***REMOVED***, 1200 * data.guilds.length);

    const guildCount = data.guilds.length;

    if (client.getMaxListeners() !== 0) client.setMaxListeners(client.getMaxListeners() + guildCount);

    client.once('ready', () => ***REMOVED***
      client.syncGuilds();
      if (client.getMaxListeners() !== 0) client.setMaxListeners(client.getMaxListeners() - guildCount);
      client.clearTimeout(t);
    ***REMOVED***);

    const ws = this.packetManager.ws;

    ws.sessionID = data.session_id;
    client.emit('debug', `READY $***REMOVED***ws.sessionID***REMOVED***`);
    ws.checkIfReady();
  ***REMOVED***
***REMOVED***

module.exports = ReadyHandler;
