// prop-types/factoryWithTypeCheckers.js
/**
 * Copyright (c) 2013-present, Facebook, Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

'use strict';

var ReactIs = require('react-is');
var assign = require('object-assign');

var ReactPropTypesSecret = require('./lib/ReactPropTypesSecret');
var checkPropTypes = require('./checkPropTypes');

var has = Function.call.bind(Object.prototype.hasOwnProperty);
var printWarning = function() ***REMOVED******REMOVED***;

if (process.env.NODE_ENV !== 'production') ***REMOVED***
  printWarning = function(text) ***REMOVED***
    var message = 'Warning: ' + text;
    if (typeof console !== 'undefined') ***REMOVED***
      console.error(message);
    ***REMOVED***
    try ***REMOVED***
      // --- Welcome to debugging React ---
      // This error was thrown as a convenience so that you can use this stack
      // to find the callsite that caused this warning to fire.
      throw new Error(message);
    ***REMOVED*** catch (x) ***REMOVED******REMOVED***
  ***REMOVED***;
***REMOVED***

function emptyFunctionThatReturnsNull() ***REMOVED***
  return null;
***REMOVED***

module.exports = function(isValidElement, throwOnDirectAccess) ***REMOVED***
  /* global Symbol */
  var ITERATOR_SYMBOL = typeof Symbol === 'function' && Symbol.iterator;
  var FAUX_ITERATOR_SYMBOL = '@@iterator'; // Before Symbol spec.

  /**
   * Returns the iterator method function contained on the iterable object.
   *
   * Be sure to invoke the function with the iterable as context:
   *
   *     var iteratorFn = getIteratorFn(myIterable);
   *     if (iteratorFn) ***REMOVED***
   *       var iterator = iteratorFn.call(myIterable);
   *       ...
   *     ***REMOVED***
   *
   * @param ***REMOVED***?object***REMOVED*** maybeIterable
   * @return ***REMOVED***?function***REMOVED***
   */
  function getIteratorFn(maybeIterable) ***REMOVED***
    var iteratorFn = maybeIterable && (ITERATOR_SYMBOL && maybeIterable[ITERATOR_SYMBOL] || maybeIterable[FAUX_ITERATOR_SYMBOL]);
    if (typeof iteratorFn === 'function') ***REMOVED***
      return iteratorFn;
    ***REMOVED***
  ***REMOVED***

  /**
   * Collection of methods that allow declaration and validation of props that are
   * supplied to React components. Example usage:
   *
   *   var Props = require('ReactPropTypes');
   *   var MyArticle = React.createClass(***REMOVED***
   *     propTypes: ***REMOVED***
   *       // An optional string prop named "description".
   *       description: Props.string,
   *
   *       // A required enum prop named "category".
   *       category: Props.oneOf(['News','Photos']).isRequired,
   *
   *       // A prop named "dialog" that requires an instance of Dialog.
   *       dialog: Props.instanceOf(Dialog).isRequired
   *     ***REMOVED***,
   *     render: function() ***REMOVED*** ... ***REMOVED***
   *   ***REMOVED***);
   *
   * A more formal specification of how these methods are used:
   *
   *   type := array|bool|func|object|number|string|oneOf([...])|instanceOf(...)
   *   decl := ReactPropTypes.***REMOVED***type***REMOVED***(.isRequired)?
   *
   * Each and every declaration produces a function with the same signature. This
   * allows the creation of custom validation functions. For example:
   *
   *  var MyLink = React.createClass(***REMOVED***
   *    propTypes: ***REMOVED***
   *      // An optional string or URI prop named "href".
   *      href: function(props, propName, componentName) ***REMOVED***
   *        var propValue = props[propName];
   *        if (propValue != null && typeof propValue !== 'string' &&
   *            !(propValue instanceof URI)) ***REMOVED***
   *          return new Error(
   *            'Expected a string or an URI for ' + propName + ' in ' +
   *            componentName
   *          );
   *        ***REMOVED***
   *      ***REMOVED***
   *    ***REMOVED***,
   *    render: function() ***REMOVED***...***REMOVED***
   *  ***REMOVED***);
   *
   * @internal
   */

  var ANONYMOUS = '<<anonymous>>';

  // Important!
  // Keep this list in sync with production version in `./factoryWithThrowingShims.js`.
  var ReactPropTypes = ***REMOVED***
    array: createPrimitiveTypeChecker('array'),
    bool: createPrimitiveTypeChecker('boolean'),
    func: createPrimitiveTypeChecker('function'),
    number: createPrimitiveTypeChecker('number'),
    object: createPrimitiveTypeChecker('object'),
    string: createPrimitiveTypeChecker('string'),
    symbol: createPrimitiveTypeChecker('symbol'),

    any: createAnyTypeChecker(),
    arrayOf: createArrayOfTypeChecker,
    element: createElementTypeChecker(),
    elementType: createElementTypeTypeChecker(),
    instanceOf: createInstanceTypeChecker,
    node: createNodeChecker(),
    objectOf: createObjectOfTypeChecker,
    oneOf: createEnumTypeChecker,
    oneOfType: createUnionTypeChecker,
    shape: createShapeTypeChecker,
    exact: createStrictShapeTypeChecker,
  ***REMOVED***;

  /**
   * inlined Object.is polyfill to avoid requiring consumers ship their own
   * https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/is
   */
  /*eslint-disable no-self-compare*/
  function is(x, y) ***REMOVED***
    // SameValue algorithm
    if (x === y) ***REMOVED***
      // Steps 1-5, 7-10
      // Steps 6.b-6.e: +0 != -0
      return x !== 0 || 1 / x === 1 / y;
    ***REMOVED*** else ***REMOVED***
      // Step 6.a: NaN == NaN
      return x !== x && y !== y;
    ***REMOVED***
  ***REMOVED***
  /*eslint-enable no-self-compare*/

  /**
   * We use an Error-like object for backward compatibility as people may call
   * PropTypes directly and inspect their output. However, we don't use real
   * Errors anymore. We don't inspect their stack anyway, and creating them
   * is prohibitively expensive if they are created too often, such as what
   * happens in oneOfType() for any type before the one that matched.
   */
  function PropTypeError(message) ***REMOVED***
    this.message = message;
    this.stack = '';
  ***REMOVED***
  // Make `instanceof Error` still work for returned errors.
  PropTypeError.prototype = Error.prototype;

  function createChainableTypeChecker(validate) ***REMOVED***
    if (process.env.NODE_ENV !== 'production') ***REMOVED***
      var manualPropTypeCallCache = ***REMOVED******REMOVED***;
      var manualPropTypeWarningCount = 0;
    ***REMOVED***
    function checkType(isRequired, props, propName, componentName, location, propFullName, secret) ***REMOVED***
      componentName = componentName || ANONYMOUS;
      propFullName = propFullName || propName;

      if (secret !== ReactPropTypesSecret) ***REMOVED***
        if (throwOnDirectAccess) ***REMOVED***
          // New behavior only for users of `prop-types` package
          var err = new Error(
            'Calling PropTypes validators directly is not supported by the `prop-types` package. ' +
            'Use `PropTypes.checkPropTypes()` to call them. ' +
            'Read more at http://fb.me/use-check-prop-types'
          );
          err.name = 'Invariant Violation';
          throw err;
        ***REMOVED*** else if (process.env.NODE_ENV !== 'production' && typeof console !== 'undefined') ***REMOVED***
          // Old behavior for people using React.PropTypes
          var cacheKey = componentName + ':' + propName;
          if (
            !manualPropTypeCallCache[cacheKey] &&
            // Avoid spamming the console because they are often not actionable except for lib authors
            manualPropTypeWarningCount < 3
          ) ***REMOVED***
            printWarning(
              'You are manually calling a React.PropTypes validation ' +
              'function for the `' + propFullName + '` prop on `' + componentName  + '`. This is deprecated ' +
              'and will throw in the standalone `prop-types` package. ' +
              'You may be seeing this warning due to a third-party PropTypes ' +
              'library. See https://fb.me/react-warning-dont-call-proptypes ' + 'for details.'
            );
            manualPropTypeCallCache[cacheKey] = true;
            manualPropTypeWarningCount++;
          ***REMOVED***
        ***REMOVED***
      ***REMOVED***
      if (props[propName] == null) ***REMOVED***
        if (isRequired) ***REMOVED***
          if (props[propName] === null) ***REMOVED***
            return new PropTypeError('The ' + location + ' `' + propFullName + '` is marked as required ' + ('in `' + componentName + '`, but its value is `null`.'));
          ***REMOVED***
          return new PropTypeError('The ' + location + ' `' + propFullName + '` is marked as required in ' + ('`' + componentName + '`, but its value is `undefined`.'));
        ***REMOVED***
        return null;
      ***REMOVED*** else ***REMOVED***
        return validate(props, propName, componentName, location, propFullName);
      ***REMOVED***
    ***REMOVED***

    var chainedCheckType = checkType.bind(null, false);
    chainedCheckType.isRequired = checkType.bind(null, true);

    return chainedCheckType;
  ***REMOVED***

  function createPrimitiveTypeChecker(expectedType) ***REMOVED***
    function validate(props, propName, componentName, location, propFullName, secret) ***REMOVED***
      var propValue = props[propName];
      var propType = getPropType(propValue);
      if (propType !== expectedType) ***REMOVED***
        // `propValue` being instance of, say, date/regexp, pass the 'object'
        // check, but we can offer a more precise error message here rather than
        // 'of type `object`'.
        var preciseType = getPreciseType(propValue);

        return new PropTypeError('Invalid ' + location + ' `' + propFullName + '` of type ' + ('`' + preciseType + '` supplied to `' + componentName + '`, expected ') + ('`' + expectedType + '`.'));
      ***REMOVED***
      return null;
    ***REMOVED***
    return createChainableTypeChecker(validate);
  ***REMOVED***

  function createAnyTypeChecker() ***REMOVED***
    return createChainableTypeChecker(emptyFunctionThatReturnsNull);
  ***REMOVED***

  function createArrayOfTypeChecker(typeChecker) ***REMOVED***
    function validate(props, propName, componentName, location, propFullName) ***REMOVED***
      if (typeof typeChecker !== 'function') ***REMOVED***
        return new PropTypeError('Property `' + propFullName + '` of component `' + componentName + '` has invalid PropType notation inside arrayOf.');
      ***REMOVED***
      var propValue = props[propName];
      if (!Array.isArray(propValue)) ***REMOVED***
        var propType = getPropType(propValue);
        return new PropTypeError('Invalid ' + location + ' `' + propFullName + '` of type ' + ('`' + propType + '` supplied to `' + componentName + '`, expected an array.'));
      ***REMOVED***
      for (var i = 0; i < propValue.length; i++) ***REMOVED***
        var error = typeChecker(propValue, i, componentName, location, propFullName + '[' + i + ']', ReactPropTypesSecret);
        if (error instanceof Error) ***REMOVED***
          return error;
        ***REMOVED***
      ***REMOVED***
      return null;
    ***REMOVED***
    return createChainableTypeChecker(validate);
  ***REMOVED***

  function createElementTypeChecker() ***REMOVED***
    function validate(props, propName, componentName, location, propFullName) ***REMOVED***
      var propValue = props[propName];
      if (!isValidElement(propValue)) ***REMOVED***
        var propType = getPropType(propValue);
        return new PropTypeError('Invalid ' + location + ' `' + propFullName + '` of type ' + ('`' + propType + '` supplied to `' + componentName + '`, expected a single ReactElement.'));
      ***REMOVED***
      return null;
    ***REMOVED***
    return createChainableTypeChecker(validate);
  ***REMOVED***

  function createElementTypeTypeChecker() ***REMOVED***
    function validate(props, propName, componentName, location, propFullName) ***REMOVED***
      var propValue = props[propName];
      if (!ReactIs.isValidElementType(propValue)) ***REMOVED***
        var propType = getPropType(propValue);
        return new PropTypeError('Invalid ' + location + ' `' + propFullName + '` of type ' + ('`' + propType + '` supplied to `' + componentName + '`, expected a single ReactElement type.'));
      ***REMOVED***
      return null;
    ***REMOVED***
    return createChainableTypeChecker(validate);
  ***REMOVED***

  function createInstanceTypeChecker(expectedClass) ***REMOVED***
    function validate(props, propName, componentName, location, propFullName) ***REMOVED***
      if (!(props[propName] instanceof expectedClass)) ***REMOVED***
        var expectedClassName = expectedClass.name || ANONYMOUS;
        var actualClassName = getClassName(props[propName]);
        return new PropTypeError('Invalid ' + location + ' `' + propFullName + '` of type ' + ('`' + actualClassName + '` supplied to `' + componentName + '`, expected ') + ('instance of `' + expectedClassName + '`.'));
      ***REMOVED***
      return null;
    ***REMOVED***
    return createChainableTypeChecker(validate);
  ***REMOVED***

  function createEnumTypeChecker(expectedValues) ***REMOVED***
    if (!Array.isArray(expectedValues)) ***REMOVED***
      if (process.env.NODE_ENV !== 'production') ***REMOVED***
        if (arguments.length > 1) ***REMOVED***
          printWarning(
            'Invalid arguments supplied to oneOf, expected an array, got ' + arguments.length + ' arguments. ' +
            'A common mistake is to write oneOf(x, y, z) instead of oneOf([x, y, z]).'
          );
        ***REMOVED*** else ***REMOVED***
          printWarning('Invalid argument supplied to oneOf, expected an array.');
        ***REMOVED***
      ***REMOVED***
      return emptyFunctionThatReturnsNull;
    ***REMOVED***

    function validate(props, propName, componentName, location, propFullName) ***REMOVED***
      var propValue = props[propName];
      for (var i = 0; i < expectedValues.length; i++) ***REMOVED***
        if (is(propValue, expectedValues[i])) ***REMOVED***
          return null;
        ***REMOVED***
      ***REMOVED***

      var valuesString = JSON.stringify(expectedValues, function replacer(key, value) ***REMOVED***
        var type = getPreciseType(value);
        if (type === 'symbol') ***REMOVED***
          return String(value);
        ***REMOVED***
        return value;
      ***REMOVED***);
      return new PropTypeError('Invalid ' + location + ' `' + propFullName + '` of value `' + String(propValue) + '` ' + ('supplied to `' + componentName + '`, expected one of ' + valuesString + '.'));
    ***REMOVED***
    return createChainableTypeChecker(validate);
  ***REMOVED***

  function createObjectOfTypeChecker(typeChecker) ***REMOVED***
    function validate(props, propName, componentName, location, propFullName) ***REMOVED***
      if (typeof typeChecker !== 'function') ***REMOVED***
        return new PropTypeError('Property `' + propFullName + '` of component `' + componentName + '` has invalid PropType notation inside objectOf.');
      ***REMOVED***
      var propValue = props[propName];
      var propType = getPropType(propValue);
      if (propType !== 'object') ***REMOVED***
        return new PropTypeError('Invalid ' + location + ' `' + propFullName + '` of type ' + ('`' + propType + '` supplied to `' + componentName + '`, expected an object.'));
      ***REMOVED***
      for (var key in propValue) ***REMOVED***
        if (has(propValue, key)) ***REMOVED***
          var error = typeChecker(propValue, key, componentName, location, propFullName + '.' + key, ReactPropTypesSecret);
          if (error instanceof Error) ***REMOVED***
            return error;
          ***REMOVED***
        ***REMOVED***
      ***REMOVED***
      return null;
    ***REMOVED***
    return createChainableTypeChecker(validate);
  ***REMOVED***

  function createUnionTypeChecker(arrayOfTypeCheckers) ***REMOVED***
    if (!Array.isArray(arrayOfTypeCheckers)) ***REMOVED***
      process.env.NODE_ENV !== 'production' ? printWarning('Invalid argument supplied to oneOfType, expected an instance of array.') : void 0;
      return emptyFunctionThatReturnsNull;
    ***REMOVED***

    for (var i = 0; i < arrayOfTypeCheckers.length; i++) ***REMOVED***
      var checker = arrayOfTypeCheckers[i];
      if (typeof checker !== 'function') ***REMOVED***
        printWarning(
          'Invalid argument supplied to oneOfType. Expected an array of check functions, but ' +
          'received ' + getPostfixForTypeWarning(checker) + ' at index ' + i + '.'
        );
        return emptyFunctionThatReturnsNull;
      ***REMOVED***
    ***REMOVED***

    function validate(props, propName, componentName, location, propFullName) ***REMOVED***
      for (var i = 0; i < arrayOfTypeCheckers.length; i++) ***REMOVED***
        var checker = arrayOfTypeCheckers[i];
        if (checker(props, propName, componentName, location, propFullName, ReactPropTypesSecret) == null) ***REMOVED***
          return null;
        ***REMOVED***
      ***REMOVED***

      return new PropTypeError('Invalid ' + location + ' `' + propFullName + '` supplied to ' + ('`' + componentName + '`.'));
    ***REMOVED***
    return createChainableTypeChecker(validate);
  ***REMOVED***

  function createNodeChecker() ***REMOVED***
    function validate(props, propName, componentName, location, propFullName) ***REMOVED***
      if (!isNode(props[propName])) ***REMOVED***
        return new PropTypeError('Invalid ' + location + ' `' + propFullName + '` supplied to ' + ('`' + componentName + '`, expected a ReactNode.'));
      ***REMOVED***
      return null;
    ***REMOVED***
    return createChainableTypeChecker(validate);
  ***REMOVED***

  function createShapeTypeChecker(shapeTypes) ***REMOVED***
    function validate(props, propName, componentName, location, propFullName) ***REMOVED***
      var propValue = props[propName];
      var propType = getPropType(propValue);
      if (propType !== 'object') ***REMOVED***
        return new PropTypeError('Invalid ' + location + ' `' + propFullName + '` of type `' + propType + '` ' + ('supplied to `' + componentName + '`, expected `object`.'));
      ***REMOVED***
      for (var key in shapeTypes) ***REMOVED***
        var checker = shapeTypes[key];
        if (!checker) ***REMOVED***
          continue;
        ***REMOVED***
        var error = checker(propValue, key, componentName, location, propFullName + '.' + key, ReactPropTypesSecret);
        if (error) ***REMOVED***
          return error;
        ***REMOVED***
      ***REMOVED***
      return null;
    ***REMOVED***
    return createChainableTypeChecker(validate);
  ***REMOVED***

  function createStrictShapeTypeChecker(shapeTypes) ***REMOVED***
    function validate(props, propName, componentName, location, propFullName) ***REMOVED***
      var propValue = props[propName];
      var propType = getPropType(propValue);
      if (propType !== 'object') ***REMOVED***
        return new PropTypeError('Invalid ' + location + ' `' + propFullName + '` of type `' + propType + '` ' + ('supplied to `' + componentName + '`, expected `object`.'));
      ***REMOVED***
      // We need to check all keys in case some are required but missing from
      // props.
      var allKeys = assign(***REMOVED******REMOVED***, props[propName], shapeTypes);
      for (var key in allKeys) ***REMOVED***
        var checker = shapeTypes[key];
        if (!checker) ***REMOVED***
          return new PropTypeError(
            'Invalid ' + location + ' `' + propFullName + '` key `' + key + '` supplied to `' + componentName + '`.' +
            '\nBad object: ' + JSON.stringify(props[propName], null, '  ') +
            '\nValid keys: ' +  JSON.stringify(Object.keys(shapeTypes), null, '  ')
          );
        ***REMOVED***
        var error = checker(propValue, key, componentName, location, propFullName + '.' + key, ReactPropTypesSecret);
        if (error) ***REMOVED***
          return error;
        ***REMOVED***
      ***REMOVED***
      return null;
    ***REMOVED***

    return createChainableTypeChecker(validate);
  ***REMOVED***

  function isNode(propValue) ***REMOVED***
    switch (typeof propValue) ***REMOVED***
      case 'number':
      case 'string':
      case 'undefined':
        return true;
      case 'boolean':
        return !propValue;
      case 'object':
        if (Array.isArray(propValue)) ***REMOVED***
          return propValue.every(isNode);
        ***REMOVED***
        if (propValue === null || isValidElement(propValue)) ***REMOVED***
          return true;
        ***REMOVED***

        var iteratorFn = getIteratorFn(propValue);
        if (iteratorFn) ***REMOVED***
          var iterator = iteratorFn.call(propValue);
          var step;
          if (iteratorFn !== propValue.entries) ***REMOVED***
            while (!(step = iterator.next()).done) ***REMOVED***
              if (!isNode(step.value)) ***REMOVED***
                return false;
              ***REMOVED***
            ***REMOVED***
          ***REMOVED*** else ***REMOVED***
            // Iterator will provide entry [k,v] tuples rather than values.
            while (!(step = iterator.next()).done) ***REMOVED***
              var entry = step.value;
              if (entry) ***REMOVED***
                if (!isNode(entry[1])) ***REMOVED***
                  return false;
                ***REMOVED***
              ***REMOVED***
            ***REMOVED***
          ***REMOVED***
        ***REMOVED*** else ***REMOVED***
          return false;
        ***REMOVED***

        return true;
      default:
        return false;
    ***REMOVED***
  ***REMOVED***

  function isSymbol(propType, propValue) ***REMOVED***
    // Native Symbol.
    if (propType === 'symbol') ***REMOVED***
      return true;
    ***REMOVED***

    // falsy value can't be a Symbol
    if (!propValue) ***REMOVED***
      return false;
    ***REMOVED***

    // 19.4.3.5 Symbol.prototype[@@toStringTag] === 'Symbol'
    if (propValue['@@toStringTag'] === 'Symbol') ***REMOVED***
      return true;
    ***REMOVED***

    // Fallback for non-spec compliant Symbols which are polyfilled.
    if (typeof Symbol === 'function' && propValue instanceof Symbol) ***REMOVED***
      return true;
    ***REMOVED***

    return false;
  ***REMOVED***

  // Equivalent of `typeof` but with special handling for array and regexp.
  function getPropType(propValue) ***REMOVED***
    var propType = typeof propValue;
    if (Array.isArray(propValue)) ***REMOVED***
      return 'array';
    ***REMOVED***
    if (propValue instanceof RegExp) ***REMOVED***
      // Old webkits (at least until Android 4.0) return 'function' rather than
      // 'object' for typeof a RegExp. We'll normalize this here so that /bla/
      // passes PropTypes.object.
      return 'object';
    ***REMOVED***
    if (isSymbol(propType, propValue)) ***REMOVED***
      return 'symbol';
    ***REMOVED***
    return propType;
  ***REMOVED***

  // This handles more types than `getPropType`. Only used for error messages.
  // See `createPrimitiveTypeChecker`.
  function getPreciseType(propValue) ***REMOVED***
    if (typeof propValue === 'undefined' || propValue === null) ***REMOVED***
      return '' + propValue;
    ***REMOVED***
    var propType = getPropType(propValue);
    if (propType === 'object') ***REMOVED***
      if (propValue instanceof Date) ***REMOVED***
        return 'date';
      ***REMOVED*** else if (propValue instanceof RegExp) ***REMOVED***
        return 'regexp';
      ***REMOVED***
    ***REMOVED***
    return propType;
  ***REMOVED***

  // Returns a string that is postfixed to a warning about an invalid type.
  // For example, "undefined" or "of type array"
  function getPostfixForTypeWarning(value) ***REMOVED***
    var type = getPreciseType(value);
    switch (type) ***REMOVED***
      case 'array':
      case 'object':
        return 'an ' + type;
      case 'boolean':
      case 'date':
      case 'regexp':
        return 'a ' + type;
      default:
        return type;
    ***REMOVED***
  ***REMOVED***

  // Returns class name of the object, if any.
  function getClassName(propValue) ***REMOVED***
    if (!propValue.constructor || !propValue.constructor.name) ***REMOVED***
      return ANONYMOUS;
    ***REMOVED***
    return propValue.constructor.name;
  ***REMOVED***

  ReactPropTypes.checkPropTypes = checkPropTypes;
  ReactPropTypes.resetWarningCache = checkPropTypes.resetWarningCache;
  ReactPropTypes.PropTypes = ReactPropTypes;

  return ReactPropTypes;
***REMOVED***;
