// prop-types/prop-types.js
(function(f)***REMOVED***if(typeof exports==="object"&&typeof module!=="undefined")***REMOVED***module.exports=f()***REMOVED***else if(typeof define==="function"&&define.amd)***REMOVED***define([],f)***REMOVED***else***REMOVED***var g;if(typeof window!=="undefined")***REMOVED***g=window***REMOVED***else if(typeof global!=="undefined")***REMOVED***g=global***REMOVED***else if(typeof self!=="undefined")***REMOVED***g=self***REMOVED***else***REMOVED***g=this***REMOVED***g.PropTypes = f()***REMOVED******REMOVED***)(function()***REMOVED***var define,module,exports;return (function e(t,n,r)***REMOVED***function s(o,u)***REMOVED***if(!n[o])***REMOVED***if(!t[o])***REMOVED***var a=typeof require=="function"&&require;if(!u&&a)return a(o,!0);if(i)return i(o,!0);var f=new Error("Cannot find module '"+o+"'");throw f.code="MODULE_NOT_FOUND",f***REMOVED***var l=n[o]=***REMOVED***exports:***REMOVED******REMOVED******REMOVED***;t[o][0].call(l.exports,function(e)***REMOVED***var n=t[o][1][e];return s(n?n:e)***REMOVED***,l,l.exports,e,t,n,r)***REMOVED***return n[o].exports***REMOVED***var i=typeof require=="function"&&require;for(var o=0;o<r.length;o++)s(r[o]);return s***REMOVED***)(***REMOVED***1:[function(require,module,exports)***REMOVED***
/**
 * Copyright (c) 2013-present, Facebook, Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

'use strict';

var printWarning = function() ***REMOVED******REMOVED***;

if ("development" !== 'production') ***REMOVED***
  var ReactPropTypesSecret = require('./lib/ReactPropTypesSecret');
  var loggedTypeFailures = ***REMOVED******REMOVED***;
  var has = Function.call.bind(Object.prototype.hasOwnProperty);

  printWarning = function(text) ***REMOVED***
    var message = 'Warning: ' + text;
    if (typeof console !== 'undefined') ***REMOVED***
      console.error(message);
    ***REMOVED***
    try ***REMOVED***
      // --- Welcome to debugging React ---
      // This error was thrown as a convenience so that you can use this stack
      // to find the callsite that caused this warning to fire.
      throw new Error(message);
    ***REMOVED*** catch (x) ***REMOVED******REMOVED***
  ***REMOVED***;
***REMOVED***

/**
 * Assert that the values match with the type specs.
 * Error messages are memorized and will only be shown once.
 *
 * @param ***REMOVED***object***REMOVED*** typeSpecs Map of name to a ReactPropType
 * @param ***REMOVED***object***REMOVED*** values Runtime values that need to be type-checked
 * @param ***REMOVED***string***REMOVED*** location e.g. "prop", "context", "child context"
 * @param ***REMOVED***string***REMOVED*** componentName Name of the component for error messages.
 * @param ***REMOVED***?Function***REMOVED*** getStack Returns the component stack.
 * @private
 */
function checkPropTypes(typeSpecs, values, location, componentName, getStack) ***REMOVED***
  if ("development" !== 'production') ***REMOVED***
    for (var typeSpecName in typeSpecs) ***REMOVED***
      if (has(typeSpecs, typeSpecName)) ***REMOVED***
        var error;
        // Prop type validation may throw. In case they do, we don't want to
        // fail the render phase where it didn't fail before. So we log it.
        // After these have been cleaned up, we'll let them throw.
        try ***REMOVED***
          // This is intentionally an invariant that gets caught. It's the same
          // behavior as without this statement except with a better message.
          if (typeof typeSpecs[typeSpecName] !== 'function') ***REMOVED***
            var err = Error(
              (componentName || 'React class') + ': ' + location + ' type `' + typeSpecName + '` is invalid; ' +
              'it must be a function, usually from the `prop-types` package, but received `' + typeof typeSpecs[typeSpecName] + '`.'
            );
            err.name = 'Invariant Violation';
            throw err;
          ***REMOVED***
          error = typeSpecs[typeSpecName](values, typeSpecName, componentName, location, null, ReactPropTypesSecret);
        ***REMOVED*** catch (ex) ***REMOVED***
          error = ex;
        ***REMOVED***
        if (error && !(error instanceof Error)) ***REMOVED***
          printWarning(
            (componentName || 'React class') + ': type specification of ' +
            location + ' `' + typeSpecName + '` is invalid; the type checker ' +
            'function must return `null` or an `Error` but returned a ' + typeof error + '. ' +
            'You may have forgotten to pass an argument to the type checker ' +
            'creator (arrayOf, instanceOf, objectOf, oneOf, oneOfType, and ' +
            'shape all require an argument).'
          );
        ***REMOVED***
        if (error instanceof Error && !(error.message in loggedTypeFailures)) ***REMOVED***
          // Only monitor this failure once because there tends to be a lot of the
          // same error.
          loggedTypeFailures[error.message] = true;

          var stack = getStack ? getStack() : '';

          printWarning(
            'Failed ' + location + ' type: ' + error.message + (stack != null ? stack : '')
          );
        ***REMOVED***
      ***REMOVED***
    ***REMOVED***
  ***REMOVED***
***REMOVED***

/**
 * Resets warning cache when testing.
 *
 * @private
 */
checkPropTypes.resetWarningCache = function() ***REMOVED***
  if ("development" !== 'production') ***REMOVED***
    loggedTypeFailures = ***REMOVED******REMOVED***;
  ***REMOVED***
***REMOVED***

module.exports = checkPropTypes;

***REMOVED***,***REMOVED***"./lib/ReactPropTypesSecret":5***REMOVED***],2:[function(require,module,exports)***REMOVED***
/**
 * Copyright (c) 2013-present, Facebook, Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

'use strict';

var ReactPropTypesSecret = require('./lib/ReactPropTypesSecret');

function emptyFunction() ***REMOVED******REMOVED***
function emptyFunctionWithReset() ***REMOVED******REMOVED***
emptyFunctionWithReset.resetWarningCache = emptyFunction;

module.exports = function() ***REMOVED***
  function shim(props, propName, componentName, location, propFullName, secret) ***REMOVED***
    if (secret === ReactPropTypesSecret) ***REMOVED***
      // It is still safe when called from React.
      return;
    ***REMOVED***
    var err = new Error(
      'Calling PropTypes validators directly is not supported by the `prop-types` package. ' +
      'Use PropTypes.checkPropTypes() to call them. ' +
      'Read more at http://fb.me/use-check-prop-types'
    );
    err.name = 'Invariant Violation';
    throw err;
  ***REMOVED***;
  shim.isRequired = shim;
  function getShim() ***REMOVED***
    return shim;
  ***REMOVED***;
  // Important!
  // Keep this list in sync with production version in `./factoryWithTypeCheckers.js`.
  var ReactPropTypes = ***REMOVED***
    array: shim,
    bool: shim,
    func: shim,
    number: shim,
    object: shim,
    string: shim,
    symbol: shim,

    any: shim,
    arrayOf: getShim,
    element: shim,
    elementType: shim,
    instanceOf: getShim,
    node: shim,
    objectOf: getShim,
    oneOf: getShim,
    oneOfType: getShim,
    shape: getShim,
    exact: getShim,

    checkPropTypes: emptyFunctionWithReset,
    resetWarningCache: emptyFunction
  ***REMOVED***;

  ReactPropTypes.PropTypes = ReactPropTypes;

  return ReactPropTypes;
***REMOVED***;

***REMOVED***,***REMOVED***"./lib/ReactPropTypesSecret":5***REMOVED***],3:[function(require,module,exports)***REMOVED***
/**
 * Copyright (c) 2013-present, Facebook, Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

'use strict';

var ReactIs = require('react-is');
var assign = require('object-assign');

var ReactPropTypesSecret = require('./lib/ReactPropTypesSecret');
var checkPropTypes = require('./checkPropTypes');

var has = Function.call.bind(Object.prototype.hasOwnProperty);
var printWarning = function() ***REMOVED******REMOVED***;

if ("development" !== 'production') ***REMOVED***
  printWarning = function(text) ***REMOVED***
    var message = 'Warning: ' + text;
    if (typeof console !== 'undefined') ***REMOVED***
      console.error(message);
    ***REMOVED***
    try ***REMOVED***
      // --- Welcome to debugging React ---
      // This error was thrown as a convenience so that you can use this stack
      // to find the callsite that caused this warning to fire.
      throw new Error(message);
    ***REMOVED*** catch (x) ***REMOVED******REMOVED***
  ***REMOVED***;
***REMOVED***

function emptyFunctionThatReturnsNull() ***REMOVED***
  return null;
***REMOVED***

module.exports = function(isValidElement, throwOnDirectAccess) ***REMOVED***
  /* global Symbol */
  var ITERATOR_SYMBOL = typeof Symbol === 'function' && Symbol.iterator;
  var FAUX_ITERATOR_SYMBOL = '@@iterator'; // Before Symbol spec.

  /**
   * Returns the iterator method function contained on the iterable object.
   *
   * Be sure to invoke the function with the iterable as context:
   *
   *     var iteratorFn = getIteratorFn(myIterable);
   *     if (iteratorFn) ***REMOVED***
   *       var iterator = iteratorFn.call(myIterable);
   *       ...
   *     ***REMOVED***
   *
   * @param ***REMOVED***?object***REMOVED*** maybeIterable
   * @return ***REMOVED***?function***REMOVED***
   */
  function getIteratorFn(maybeIterable) ***REMOVED***
    var iteratorFn = maybeIterable && (ITERATOR_SYMBOL && maybeIterable[ITERATOR_SYMBOL] || maybeIterable[FAUX_ITERATOR_SYMBOL]);
    if (typeof iteratorFn === 'function') ***REMOVED***
      return iteratorFn;
    ***REMOVED***
  ***REMOVED***

  /**
   * Collection of methods that allow declaration and validation of props that are
   * supplied to React components. Example usage:
   *
   *   var Props = require('ReactPropTypes');
   *   var MyArticle = React.createClass(***REMOVED***
   *     propTypes: ***REMOVED***
   *       // An optional string prop named "description".
   *       description: Props.string,
   *
   *       // A required enum prop named "category".
   *       category: Props.oneOf(['News','Photos']).isRequired,
   *
   *       // A prop named "dialog" that requires an instance of Dialog.
   *       dialog: Props.instanceOf(Dialog).isRequired
   *     ***REMOVED***,
   *     render: function() ***REMOVED*** ... ***REMOVED***
   *   ***REMOVED***);
   *
   * A more formal specification of how these methods are used:
   *
   *   type := array|bool|func|object|number|string|oneOf([...])|instanceOf(...)
   *   decl := ReactPropTypes.***REMOVED***type***REMOVED***(.isRequired)?
   *
   * Each and every declaration produces a function with the same signature. This
   * allows the creation of custom validation functions. For example:
   *
   *  var MyLink = React.createClass(***REMOVED***
   *    propTypes: ***REMOVED***
   *      // An optional string or URI prop named "href".
   *      href: function(props, propName, componentName) ***REMOVED***
   *        var propValue = props[propName];
   *        if (propValue != null && typeof propValue !== 'string' &&
   *            !(propValue instanceof URI)) ***REMOVED***
   *          return new Error(
   *            'Expected a string or an URI for ' + propName + ' in ' +
   *            componentName
   *          );
   *        ***REMOVED***
   *      ***REMOVED***
   *    ***REMOVED***,
   *    render: function() ***REMOVED***...***REMOVED***
   *  ***REMOVED***);
   *
   * @internal
   */

  var ANONYMOUS = '<<anonymous>>';

  // Important!
  // Keep this list in sync with production version in `./factoryWithThrowingShims.js`.
  var ReactPropTypes = ***REMOVED***
    array: createPrimitiveTypeChecker('array'),
    bool: createPrimitiveTypeChecker('boolean'),
    func: createPrimitiveTypeChecker('function'),
    number: createPrimitiveTypeChecker('number'),
    object: createPrimitiveTypeChecker('object'),
    string: createPrimitiveTypeChecker('string'),
    symbol: createPrimitiveTypeChecker('symbol'),

    any: createAnyTypeChecker(),
    arrayOf: createArrayOfTypeChecker,
    element: createElementTypeChecker(),
    elementType: createElementTypeTypeChecker(),
    instanceOf: createInstanceTypeChecker,
    node: createNodeChecker(),
    objectOf: createObjectOfTypeChecker,
    oneOf: createEnumTypeChecker,
    oneOfType: createUnionTypeChecker,
    shape: createShapeTypeChecker,
    exact: createStrictShapeTypeChecker,
  ***REMOVED***;

  /**
   * inlined Object.is polyfill to avoid requiring consumers ship their own
   * https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/is
   */
  /*eslint-disable no-self-compare*/
  function is(x, y) ***REMOVED***
    // SameValue algorithm
    if (x === y) ***REMOVED***
      // Steps 1-5, 7-10
      // Steps 6.b-6.e: +0 != -0
      return x !== 0 || 1 / x === 1 / y;
    ***REMOVED*** else ***REMOVED***
      // Step 6.a: NaN == NaN
      return x !== x && y !== y;
    ***REMOVED***
  ***REMOVED***
  /*eslint-enable no-self-compare*/

  /**
   * We use an Error-like object for backward compatibility as people may call
   * PropTypes directly and inspect their output. However, we don't use real
   * Errors anymore. We don't inspect their stack anyway, and creating them
   * is prohibitively expensive if they are created too often, such as what
   * happens in oneOfType() for any type before the one that matched.
   */
  function PropTypeError(message) ***REMOVED***
    this.message = message;
    this.stack = '';
  ***REMOVED***
  // Make `instanceof Error` still work for returned errors.
  PropTypeError.prototype = Error.prototype;

  function createChainableTypeChecker(validate) ***REMOVED***
    if ("development" !== 'production') ***REMOVED***
      var manualPropTypeCallCache = ***REMOVED******REMOVED***;
      var manualPropTypeWarningCount = 0;
    ***REMOVED***
    function checkType(isRequired, props, propName, componentName, location, propFullName, secret) ***REMOVED***
      componentName = componentName || ANONYMOUS;
      propFullName = propFullName || propName;

      if (secret !== ReactPropTypesSecret) ***REMOVED***
        if (throwOnDirectAccess) ***REMOVED***
          // New behavior only for users of `prop-types` package
          var err = new Error(
            'Calling PropTypes validators directly is not supported by the `prop-types` package. ' +
            'Use `PropTypes.checkPropTypes()` to call them. ' +
            'Read more at http://fb.me/use-check-prop-types'
          );
          err.name = 'Invariant Violation';
          throw err;
        ***REMOVED*** else if ("development" !== 'production' && typeof console !== 'undefined') ***REMOVED***
          // Old behavior for people using React.PropTypes
          var cacheKey = componentName + ':' + propName;
          if (
            !manualPropTypeCallCache[cacheKey] &&
            // Avoid spamming the console because they are often not actionable except for lib authors
            manualPropTypeWarningCount < 3
          ) ***REMOVED***
            printWarning(
              'You are manually calling a React.PropTypes validation ' +
              'function for the `' + propFullName + '` prop on `' + componentName  + '`. This is deprecated ' +
              'and will throw in the standalone `prop-types` package. ' +
              'You may be seeing this warning due to a third-party PropTypes ' +
              'library. See https://fb.me/react-warning-dont-call-proptypes ' + 'for details.'
            );
            manualPropTypeCallCache[cacheKey] = true;
            manualPropTypeWarningCount++;
          ***REMOVED***
        ***REMOVED***
      ***REMOVED***
      if (props[propName] == null) ***REMOVED***
        if (isRequired) ***REMOVED***
          if (props[propName] === null) ***REMOVED***
            return new PropTypeError('The ' + location + ' `' + propFullName + '` is marked as required ' + ('in `' + componentName + '`, but its value is `null`.'));
          ***REMOVED***
          return new PropTypeError('The ' + location + ' `' + propFullName + '` is marked as required in ' + ('`' + componentName + '`, but its value is `undefined`.'));
        ***REMOVED***
        return null;
      ***REMOVED*** else ***REMOVED***
        return validate(props, propName, componentName, location, propFullName);
      ***REMOVED***
    ***REMOVED***

    var chainedCheckType = checkType.bind(null, false);
    chainedCheckType.isRequired = checkType.bind(null, true);

    return chainedCheckType;
  ***REMOVED***

  function createPrimitiveTypeChecker(expectedType) ***REMOVED***
    function validate(props, propName, componentName, location, propFullName, secret) ***REMOVED***
      var propValue = props[propName];
      var propType = getPropType(propValue);
      if (propType !== expectedType) ***REMOVED***
        // `propValue` being instance of, say, date/regexp, pass the 'object'
        // check, but we can offer a more precise error message here rather than
        // 'of type `object`'.
        var preciseType = getPreciseType(propValue);

        return new PropTypeError('Invalid ' + location + ' `' + propFullName + '` of type ' + ('`' + preciseType + '` supplied to `' + componentName + '`, expected ') + ('`' + expectedType + '`.'));
      ***REMOVED***
      return null;
    ***REMOVED***
    return createChainableTypeChecker(validate);
  ***REMOVED***

  function createAnyTypeChecker() ***REMOVED***
    return createChainableTypeChecker(emptyFunctionThatReturnsNull);
  ***REMOVED***

  function createArrayOfTypeChecker(typeChecker) ***REMOVED***
    function validate(props, propName, componentName, location, propFullName) ***REMOVED***
      if (typeof typeChecker !== 'function') ***REMOVED***
        return new PropTypeError('Property `' + propFullName + '` of component `' + componentName + '` has invalid PropType notation inside arrayOf.');
      ***REMOVED***
      var propValue = props[propName];
      if (!Array.isArray(propValue)) ***REMOVED***
        var propType = getPropType(propValue);
        return new PropTypeError('Invalid ' + location + ' `' + propFullName + '` of type ' + ('`' + propType + '` supplied to `' + componentName + '`, expected an array.'));
      ***REMOVED***
      for (var i = 0; i < propValue.length; i++) ***REMOVED***
        var error = typeChecker(propValue, i, componentName, location, propFullName + '[' + i + ']', ReactPropTypesSecret);
        if (error instanceof Error) ***REMOVED***
          return error;
        ***REMOVED***
      ***REMOVED***
      return null;
    ***REMOVED***
    return createChainableTypeChecker(validate);
  ***REMOVED***

  function createElementTypeChecker() ***REMOVED***
    function validate(props, propName, componentName, location, propFullName) ***REMOVED***
      var propValue = props[propName];
      if (!isValidElement(propValue)) ***REMOVED***
        var propType = getPropType(propValue);
        return new PropTypeError('Invalid ' + location + ' `' + propFullName + '` of type ' + ('`' + propType + '` supplied to `' + componentName + '`, expected a single ReactElement.'));
      ***REMOVED***
      return null;
    ***REMOVED***
    return createChainableTypeChecker(validate);
  ***REMOVED***

  function createElementTypeTypeChecker() ***REMOVED***
    function validate(props, propName, componentName, location, propFullName) ***REMOVED***
      var propValue = props[propName];
      if (!ReactIs.isValidElementType(propValue)) ***REMOVED***
        var propType = getPropType(propValue);
        return new PropTypeError('Invalid ' + location + ' `' + propFullName + '` of type ' + ('`' + propType + '` supplied to `' + componentName + '`, expected a single ReactElement type.'));
      ***REMOVED***
      return null;
    ***REMOVED***
    return createChainableTypeChecker(validate);
  ***REMOVED***

  function createInstanceTypeChecker(expectedClass) ***REMOVED***
    function validate(props, propName, componentName, location, propFullName) ***REMOVED***
      if (!(props[propName] instanceof expectedClass)) ***REMOVED***
        var expectedClassName = expectedClass.name || ANONYMOUS;
        var actualClassName = getClassName(props[propName]);
        return new PropTypeError('Invalid ' + location + ' `' + propFullName + '` of type ' + ('`' + actualClassName + '` supplied to `' + componentName + '`, expected ') + ('instance of `' + expectedClassName + '`.'));
      ***REMOVED***
      return null;
    ***REMOVED***
    return createChainableTypeChecker(validate);
  ***REMOVED***

  function createEnumTypeChecker(expectedValues) ***REMOVED***
    if (!Array.isArray(expectedValues)) ***REMOVED***
      if ("development" !== 'production') ***REMOVED***
        if (arguments.length > 1) ***REMOVED***
          printWarning(
            'Invalid arguments supplied to oneOf, expected an array, got ' + arguments.length + ' arguments. ' +
            'A common mistake is to write oneOf(x, y, z) instead of oneOf([x, y, z]).'
          );
        ***REMOVED*** else ***REMOVED***
          printWarning('Invalid argument supplied to oneOf, expected an array.');
        ***REMOVED***
      ***REMOVED***
      return emptyFunctionThatReturnsNull;
    ***REMOVED***

    function validate(props, propName, componentName, location, propFullName) ***REMOVED***
      var propValue = props[propName];
      for (var i = 0; i < expectedValues.length; i++) ***REMOVED***
        if (is(propValue, expectedValues[i])) ***REMOVED***
          return null;
        ***REMOVED***
      ***REMOVED***

      var valuesString = JSON.stringify(expectedValues, function replacer(key, value) ***REMOVED***
        var type = getPreciseType(value);
        if (type === 'symbol') ***REMOVED***
          return String(value);
        ***REMOVED***
        return value;
      ***REMOVED***);
      return new PropTypeError('Invalid ' + location + ' `' + propFullName + '` of value `' + String(propValue) + '` ' + ('supplied to `' + componentName + '`, expected one of ' + valuesString + '.'));
    ***REMOVED***
    return createChainableTypeChecker(validate);
  ***REMOVED***

  function createObjectOfTypeChecker(typeChecker) ***REMOVED***
    function validate(props, propName, componentName, location, propFullName) ***REMOVED***
      if (typeof typeChecker !== 'function') ***REMOVED***
        return new PropTypeError('Property `' + propFullName + '` of component `' + componentName + '` has invalid PropType notation inside objectOf.');
      ***REMOVED***
      var propValue = props[propName];
      var propType = getPropType(propValue);
      if (propType !== 'object') ***REMOVED***
        return new PropTypeError('Invalid ' + location + ' `' + propFullName + '` of type ' + ('`' + propType + '` supplied to `' + componentName + '`, expected an object.'));
      ***REMOVED***
      for (var key in propValue) ***REMOVED***
        if (has(propValue, key)) ***REMOVED***
          var error = typeChecker(propValue, key, componentName, location, propFullName + '.' + key, ReactPropTypesSecret);
          if (error instanceof Error) ***REMOVED***
            return error;
          ***REMOVED***
        ***REMOVED***
      ***REMOVED***
      return null;
    ***REMOVED***
    return createChainableTypeChecker(validate);
  ***REMOVED***

  function createUnionTypeChecker(arrayOfTypeCheckers) ***REMOVED***
    if (!Array.isArray(arrayOfTypeCheckers)) ***REMOVED***
      "development" !== 'production' ? printWarning('Invalid argument supplied to oneOfType, expected an instance of array.') : void 0;
      return emptyFunctionThatReturnsNull;
    ***REMOVED***

    for (var i = 0; i < arrayOfTypeCheckers.length; i++) ***REMOVED***
      var checker = arrayOfTypeCheckers[i];
      if (typeof checker !== 'function') ***REMOVED***
        printWarning(
          'Invalid argument supplied to oneOfType. Expected an array of check functions, but ' +
          'received ' + getPostfixForTypeWarning(checker) + ' at index ' + i + '.'
        );
        return emptyFunctionThatReturnsNull;
      ***REMOVED***
    ***REMOVED***

    function validate(props, propName, componentName, location, propFullName) ***REMOVED***
      for (var i = 0; i < arrayOfTypeCheckers.length; i++) ***REMOVED***
        var checker = arrayOfTypeCheckers[i];
        if (checker(props, propName, componentName, location, propFullName, ReactPropTypesSecret) == null) ***REMOVED***
          return null;
        ***REMOVED***
      ***REMOVED***

      return new PropTypeError('Invalid ' + location + ' `' + propFullName + '` supplied to ' + ('`' + componentName + '`.'));
    ***REMOVED***
    return createChainableTypeChecker(validate);
  ***REMOVED***

  function createNodeChecker() ***REMOVED***
    function validate(props, propName, componentName, location, propFullName) ***REMOVED***
      if (!isNode(props[propName])) ***REMOVED***
        return new PropTypeError('Invalid ' + location + ' `' + propFullName + '` supplied to ' + ('`' + componentName + '`, expected a ReactNode.'));
      ***REMOVED***
      return null;
    ***REMOVED***
    return createChainableTypeChecker(validate);
  ***REMOVED***

  function createShapeTypeChecker(shapeTypes) ***REMOVED***
    function validate(props, propName, componentName, location, propFullName) ***REMOVED***
      var propValue = props[propName];
      var propType = getPropType(propValue);
      if (propType !== 'object') ***REMOVED***
        return new PropTypeError('Invalid ' + location + ' `' + propFullName + '` of type `' + propType + '` ' + ('supplied to `' + componentName + '`, expected `object`.'));
      ***REMOVED***
      for (var key in shapeTypes) ***REMOVED***
        var checker = shapeTypes[key];
        if (!checker) ***REMOVED***
          continue;
        ***REMOVED***
        var error = checker(propValue, key, componentName, location, propFullName + '.' + key, ReactPropTypesSecret);
        if (error) ***REMOVED***
          return error;
        ***REMOVED***
      ***REMOVED***
      return null;
    ***REMOVED***
    return createChainableTypeChecker(validate);
  ***REMOVED***

  function createStrictShapeTypeChecker(shapeTypes) ***REMOVED***
    function validate(props, propName, componentName, location, propFullName) ***REMOVED***
      var propValue = props[propName];
      var propType = getPropType(propValue);
      if (propType !== 'object') ***REMOVED***
        return new PropTypeError('Invalid ' + location + ' `' + propFullName + '` of type `' + propType + '` ' + ('supplied to `' + componentName + '`, expected `object`.'));
      ***REMOVED***
      // We need to check all keys in case some are required but missing from
      // props.
      var allKeys = assign(***REMOVED******REMOVED***, props[propName], shapeTypes);
      for (var key in allKeys) ***REMOVED***
        var checker = shapeTypes[key];
        if (!checker) ***REMOVED***
          return new PropTypeError(
            'Invalid ' + location + ' `' + propFullName + '` key `' + key + '` supplied to `' + componentName + '`.' +
            '\nBad object: ' + JSON.stringify(props[propName], null, '  ') +
            '\nValid keys: ' +  JSON.stringify(Object.keys(shapeTypes), null, '  ')
          );
        ***REMOVED***
        var error = checker(propValue, key, componentName, location, propFullName + '.' + key, ReactPropTypesSecret);
        if (error) ***REMOVED***
          return error;
        ***REMOVED***
      ***REMOVED***
      return null;
    ***REMOVED***

    return createChainableTypeChecker(validate);
  ***REMOVED***

  function isNode(propValue) ***REMOVED***
    switch (typeof propValue) ***REMOVED***
      case 'number':
      case 'string':
      case 'undefined':
        return true;
      case 'boolean':
        return !propValue;
      case 'object':
        if (Array.isArray(propValue)) ***REMOVED***
          return propValue.every(isNode);
        ***REMOVED***
        if (propValue === null || isValidElement(propValue)) ***REMOVED***
          return true;
        ***REMOVED***

        var iteratorFn = getIteratorFn(propValue);
        if (iteratorFn) ***REMOVED***
          var iterator = iteratorFn.call(propValue);
          var step;
          if (iteratorFn !== propValue.entries) ***REMOVED***
            while (!(step = iterator.next()).done) ***REMOVED***
              if (!isNode(step.value)) ***REMOVED***
                return false;
              ***REMOVED***
            ***REMOVED***
          ***REMOVED*** else ***REMOVED***
            // Iterator will provide entry [k,v] tuples rather than values.
            while (!(step = iterator.next()).done) ***REMOVED***
              var entry = step.value;
              if (entry) ***REMOVED***
                if (!isNode(entry[1])) ***REMOVED***
                  return false;
                ***REMOVED***
              ***REMOVED***
            ***REMOVED***
          ***REMOVED***
        ***REMOVED*** else ***REMOVED***
          return false;
        ***REMOVED***

        return true;
      default:
        return false;
    ***REMOVED***
  ***REMOVED***

  function isSymbol(propType, propValue) ***REMOVED***
    // Native Symbol.
    if (propType === 'symbol') ***REMOVED***
      return true;
    ***REMOVED***

    // falsy value can't be a Symbol
    if (!propValue) ***REMOVED***
      return false;
    ***REMOVED***

    // 19.4.3.5 Symbol.prototype[@@toStringTag] === 'Symbol'
    if (propValue['@@toStringTag'] === 'Symbol') ***REMOVED***
      return true;
    ***REMOVED***

    // Fallback for non-spec compliant Symbols which are polyfilled.
    if (typeof Symbol === 'function' && propValue instanceof Symbol) ***REMOVED***
      return true;
    ***REMOVED***

    return false;
  ***REMOVED***

  // Equivalent of `typeof` but with special handling for array and regexp.
  function getPropType(propValue) ***REMOVED***
    var propType = typeof propValue;
    if (Array.isArray(propValue)) ***REMOVED***
      return 'array';
    ***REMOVED***
    if (propValue instanceof RegExp) ***REMOVED***
      // Old webkits (at least until Android 4.0) return 'function' rather than
      // 'object' for typeof a RegExp. We'll normalize this here so that /bla/
      // passes PropTypes.object.
      return 'object';
    ***REMOVED***
    if (isSymbol(propType, propValue)) ***REMOVED***
      return 'symbol';
    ***REMOVED***
    return propType;
  ***REMOVED***

  // This handles more types than `getPropType`. Only used for error messages.
  // See `createPrimitiveTypeChecker`.
  function getPreciseType(propValue) ***REMOVED***
    if (typeof propValue === 'undefined' || propValue === null) ***REMOVED***
      return '' + propValue;
    ***REMOVED***
    var propType = getPropType(propValue);
    if (propType === 'object') ***REMOVED***
      if (propValue instanceof Date) ***REMOVED***
        return 'date';
      ***REMOVED*** else if (propValue instanceof RegExp) ***REMOVED***
        return 'regexp';
      ***REMOVED***
    ***REMOVED***
    return propType;
  ***REMOVED***

  // Returns a string that is postfixed to a warning about an invalid type.
  // For example, "undefined" or "of type array"
  function getPostfixForTypeWarning(value) ***REMOVED***
    var type = getPreciseType(value);
    switch (type) ***REMOVED***
      case 'array':
      case 'object':
        return 'an ' + type;
      case 'boolean':
      case 'date':
      case 'regexp':
        return 'a ' + type;
      default:
        return type;
    ***REMOVED***
  ***REMOVED***

  // Returns class name of the object, if any.
  function getClassName(propValue) ***REMOVED***
    if (!propValue.constructor || !propValue.constructor.name) ***REMOVED***
      return ANONYMOUS;
    ***REMOVED***
    return propValue.constructor.name;
  ***REMOVED***

  ReactPropTypes.checkPropTypes = checkPropTypes;
  ReactPropTypes.resetWarningCache = checkPropTypes.resetWarningCache;
  ReactPropTypes.PropTypes = ReactPropTypes;

  return ReactPropTypes;
***REMOVED***;

***REMOVED***,***REMOVED***"./checkPropTypes":1,"./lib/ReactPropTypesSecret":5,"object-assign":6,"react-is":10***REMOVED***],4:[function(require,module,exports)***REMOVED***
/**
 * Copyright (c) 2013-present, Facebook, Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

if ("development" !== 'production') ***REMOVED***
  var ReactIs = require('react-is');

  // By explicitly using `prop-types` you are opting into new development behavior.
  // http://fb.me/prop-types-in-prod
  var throwOnDirectAccess = true;
  module.exports = require('./factoryWithTypeCheckers')(ReactIs.isElement, throwOnDirectAccess);
***REMOVED*** else ***REMOVED***
  // By explicitly using `prop-types` you are opting into new production behavior.
  // http://fb.me/prop-types-in-prod
  module.exports = require('./factoryWithThrowingShims')();
***REMOVED***

***REMOVED***,***REMOVED***"./factoryWithThrowingShims":2,"./factoryWithTypeCheckers":3,"react-is":10***REMOVED***],5:[function(require,module,exports)***REMOVED***
/**
 * Copyright (c) 2013-present, Facebook, Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

'use strict';

var ReactPropTypesSecret = 'SECRET_DO_NOT_PASS_THIS_OR_YOU_WILL_BE_FIRED';

module.exports = ReactPropTypesSecret;

***REMOVED***,***REMOVED******REMOVED***],6:[function(require,module,exports)***REMOVED***
/*
object-assign
(c) Sindre Sorhus
@license MIT
*/

'use strict';
/* eslint-disable no-unused-vars */
var getOwnPropertySymbols = Object.getOwnPropertySymbols;
var hasOwnProperty = Object.prototype.hasOwnProperty;
var propIsEnumerable = Object.prototype.propertyIsEnumerable;

function toObject(val) ***REMOVED***
	if (val === null || val === undefined) ***REMOVED***
		throw new TypeError('Object.assign cannot be called with null or undefined');
	***REMOVED***

	return Object(val);
***REMOVED***

function shouldUseNative() ***REMOVED***
	try ***REMOVED***
		if (!Object.assign) ***REMOVED***
			return false;
		***REMOVED***

		// Detect buggy property enumeration order in older V8 versions.

		// https://bugs.chromium.org/p/v8/issues/detail?id=4118
		var test1 = new String('abc');  // eslint-disable-line no-new-wrappers
		test1[5] = 'de';
		if (Object.getOwnPropertyNames(test1)[0] === '5') ***REMOVED***
			return false;
		***REMOVED***

		// https://bugs.chromium.org/p/v8/issues/detail?id=3056
		var test2 = ***REMOVED******REMOVED***;
		for (var i = 0; i < 10; i++) ***REMOVED***
			test2['_' + String.fromCharCode(i)] = i;
		***REMOVED***
		var order2 = Object.getOwnPropertyNames(test2).map(function (n) ***REMOVED***
			return test2[n];
		***REMOVED***);
		if (order2.join('') !== '0123456789') ***REMOVED***
			return false;
		***REMOVED***

		// https://bugs.chromium.org/p/v8/issues/detail?id=3056
		var test3 = ***REMOVED******REMOVED***;
		'abcdefghijklmnopqrst'.split('').forEach(function (letter) ***REMOVED***
			test3[letter] = letter;
		***REMOVED***);
		if (Object.keys(Object.assign(***REMOVED******REMOVED***, test3)).join('') !==
				'abcdefghijklmnopqrst') ***REMOVED***
			return false;
		***REMOVED***

		return true;
	***REMOVED*** catch (err) ***REMOVED***
		// We don't expect any of the above to throw, but better to be safe.
		return false;
	***REMOVED***
***REMOVED***

module.exports = shouldUseNative() ? Object.assign : function (target, source) ***REMOVED***
	var from;
	var to = toObject(target);
	var symbols;

	for (var s = 1; s < arguments.length; s++) ***REMOVED***
		from = Object(arguments[s]);

		for (var key in from) ***REMOVED***
			if (hasOwnProperty.call(from, key)) ***REMOVED***
				to[key] = from[key];
			***REMOVED***
		***REMOVED***

		if (getOwnPropertySymbols) ***REMOVED***
			symbols = getOwnPropertySymbols(from);
			for (var i = 0; i < symbols.length; i++) ***REMOVED***
				if (propIsEnumerable.call(from, symbols[i])) ***REMOVED***
					to[symbols[i]] = from[symbols[i]];
				***REMOVED***
			***REMOVED***
		***REMOVED***
	***REMOVED***

	return to;
***REMOVED***;

***REMOVED***,***REMOVED******REMOVED***],7:[function(require,module,exports)***REMOVED***
// shim for using process in browser
var process = module.exports = ***REMOVED******REMOVED***;

// cached from whatever global is present so that test runners that stub it
// don't break things.  But we need to wrap it in a try catch in case it is
// wrapped in strict mode code which doesn't define any globals.  It's inside a
// function because try/catches deoptimize in certain engines.

var cachedSetTimeout;
var cachedClearTimeout;

function defaultSetTimout() ***REMOVED***
    throw new Error('setTimeout has not been defined');
***REMOVED***
function defaultClearTimeout () ***REMOVED***
    throw new Error('clearTimeout has not been defined');
***REMOVED***
(function () ***REMOVED***
    try ***REMOVED***
        if (typeof setTimeout === 'function') ***REMOVED***
            cachedSetTimeout = setTimeout;
        ***REMOVED*** else ***REMOVED***
            cachedSetTimeout = defaultSetTimout;
        ***REMOVED***
    ***REMOVED*** catch (e) ***REMOVED***
        cachedSetTimeout = defaultSetTimout;
    ***REMOVED***
    try ***REMOVED***
        if (typeof clearTimeout === 'function') ***REMOVED***
            cachedClearTimeout = clearTimeout;
        ***REMOVED*** else ***REMOVED***
            cachedClearTimeout = defaultClearTimeout;
        ***REMOVED***
    ***REMOVED*** catch (e) ***REMOVED***
        cachedClearTimeout = defaultClearTimeout;
    ***REMOVED***
***REMOVED*** ())
function runTimeout(fun) ***REMOVED***
    if (cachedSetTimeout === setTimeout) ***REMOVED***
        //normal enviroments in sane situations
        return setTimeout(fun, 0);
    ***REMOVED***
    // if setTimeout wasn't available but was latter defined
    if ((cachedSetTimeout === defaultSetTimout || !cachedSetTimeout) && setTimeout) ***REMOVED***
        cachedSetTimeout = setTimeout;
        return setTimeout(fun, 0);
    ***REMOVED***
    try ***REMOVED***
        // when when somebody has screwed with setTimeout but no I.E. maddness
        return cachedSetTimeout(fun, 0);
    ***REMOVED*** catch(e)***REMOVED***
        try ***REMOVED***
            // When we are in I.E. but the script has been evaled so I.E. doesn't trust the global object when called normally
            return cachedSetTimeout.call(null, fun, 0);
        ***REMOVED*** catch(e)***REMOVED***
            // same as above but when it's a version of I.E. that must have the global object for 'this', hopfully our context correct otherwise it will throw a global error
            return cachedSetTimeout.call(this, fun, 0);
        ***REMOVED***
    ***REMOVED***


***REMOVED***
function runClearTimeout(marker) ***REMOVED***
    if (cachedClearTimeout === clearTimeout) ***REMOVED***
        //normal enviroments in sane situations
        return clearTimeout(marker);
    ***REMOVED***
    // if clearTimeout wasn't available but was latter defined
    if ((cachedClearTimeout === defaultClearTimeout || !cachedClearTimeout) && clearTimeout) ***REMOVED***
        cachedClearTimeout = clearTimeout;
        return clearTimeout(marker);
    ***REMOVED***
    try ***REMOVED***
        // when when somebody has screwed with setTimeout but no I.E. maddness
        return cachedClearTimeout(marker);
    ***REMOVED*** catch (e)***REMOVED***
        try ***REMOVED***
            // When we are in I.E. but the script has been evaled so I.E. doesn't  trust the global object when called normally
            return cachedClearTimeout.call(null, marker);
        ***REMOVED*** catch (e)***REMOVED***
            // same as above but when it's a version of I.E. that must have the global object for 'this', hopfully our context correct otherwise it will throw a global error.
            // Some versions of I.E. have different rules for clearTimeout vs setTimeout
            return cachedClearTimeout.call(this, marker);
        ***REMOVED***
    ***REMOVED***



***REMOVED***
var queue = [];
var draining = false;
var currentQueue;
var queueIndex = -1;

function cleanUpNextTick() ***REMOVED***
    if (!draining || !currentQueue) ***REMOVED***
        return;
    ***REMOVED***
    draining = false;
    if (currentQueue.length) ***REMOVED***
        queue = currentQueue.concat(queue);
    ***REMOVED*** else ***REMOVED***
        queueIndex = -1;
    ***REMOVED***
    if (queue.length) ***REMOVED***
        drainQueue();
    ***REMOVED***
***REMOVED***

function drainQueue() ***REMOVED***
    if (draining) ***REMOVED***
        return;
    ***REMOVED***
    var timeout = runTimeout(cleanUpNextTick);
    draining = true;

    var len = queue.length;
    while(len) ***REMOVED***
        currentQueue = queue;
        queue = [];
        while (++queueIndex < len) ***REMOVED***
            if (currentQueue) ***REMOVED***
                currentQueue[queueIndex].run();
            ***REMOVED***
        ***REMOVED***
        queueIndex = -1;
        len = queue.length;
    ***REMOVED***
    currentQueue = null;
    draining = false;
    runClearTimeout(timeout);
***REMOVED***

process.nextTick = function (fun) ***REMOVED***
    var args = new Array(arguments.length - 1);
    if (arguments.length > 1) ***REMOVED***
        for (var i = 1; i < arguments.length; i++) ***REMOVED***
            args[i - 1] = arguments[i];
        ***REMOVED***
    ***REMOVED***
    queue.push(new Item(fun, args));
    if (queue.length === 1 && !draining) ***REMOVED***
        runTimeout(drainQueue);
    ***REMOVED***
***REMOVED***;

// v8 likes predictible objects
function Item(fun, array) ***REMOVED***
    this.fun = fun;
    this.array = array;
***REMOVED***
Item.prototype.run = function () ***REMOVED***
    this.fun.apply(null, this.array);
***REMOVED***;
process.title = 'browser';
process.browser = true;
process.env = ***REMOVED******REMOVED***;
process.argv = [];
process.version = ''; // empty string to avoid regexp issues
process.versions = ***REMOVED******REMOVED***;

function noop() ***REMOVED******REMOVED***

process.on = noop;
process.addListener = noop;
process.once = noop;
process.off = noop;
process.removeListener = noop;
process.removeAllListeners = noop;
process.emit = noop;
process.prependListener = noop;
process.prependOnceListener = noop;

process.listeners = function (name) ***REMOVED*** return [] ***REMOVED***

process.binding = function (name) ***REMOVED***
    throw new Error('process.binding is not supported');
***REMOVED***;

process.cwd = function () ***REMOVED*** return '/' ***REMOVED***;
process.chdir = function (dir) ***REMOVED***
    throw new Error('process.chdir is not supported');
***REMOVED***;
process.umask = function() ***REMOVED*** return 0; ***REMOVED***;

***REMOVED***,***REMOVED******REMOVED***],8:[function(require,module,exports)***REMOVED***
(function (process)***REMOVED***
/** @license React v16.8.1
 * react-is.development.js
 *
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

'use strict';



if (process.env.NODE_ENV !== "production") ***REMOVED***
  (function() ***REMOVED***
'use strict';

Object.defineProperty(exports, '__esModule', ***REMOVED*** value: true ***REMOVED***);

// The Symbol used to tag the ReactElement-like types. If there is no native Symbol
// nor polyfill, then a plain number is used for performance.
var hasSymbol = typeof Symbol === 'function' && Symbol.for;

var REACT_ELEMENT_TYPE = hasSymbol ? Symbol.for('react.element') : 0xeac7;
var REACT_PORTAL_TYPE = hasSymbol ? Symbol.for('react.portal') : 0xeaca;
var REACT_FRAGMENT_TYPE = hasSymbol ? Symbol.for('react.fragment') : 0xeacb;
var REACT_STRICT_MODE_TYPE = hasSymbol ? Symbol.for('react.strict_mode') : 0xeacc;
var REACT_PROFILER_TYPE = hasSymbol ? Symbol.for('react.profiler') : 0xead2;
var REACT_PROVIDER_TYPE = hasSymbol ? Symbol.for('react.provider') : 0xeacd;
var REACT_CONTEXT_TYPE = hasSymbol ? Symbol.for('react.context') : 0xeace;
var REACT_ASYNC_MODE_TYPE = hasSymbol ? Symbol.for('react.async_mode') : 0xeacf;
var REACT_CONCURRENT_MODE_TYPE = hasSymbol ? Symbol.for('react.concurrent_mode') : 0xeacf;
var REACT_FORWARD_REF_TYPE = hasSymbol ? Symbol.for('react.forward_ref') : 0xead0;
var REACT_SUSPENSE_TYPE = hasSymbol ? Symbol.for('react.suspense') : 0xead1;
var REACT_MEMO_TYPE = hasSymbol ? Symbol.for('react.memo') : 0xead3;
var REACT_LAZY_TYPE = hasSymbol ? Symbol.for('react.lazy') : 0xead4;

function isValidElementType(type) ***REMOVED***
  return typeof type === 'string' || typeof type === 'function' ||
  // Note: its typeof might be other than 'symbol' or 'number' if it's a polyfill.
  type === REACT_FRAGMENT_TYPE || type === REACT_CONCURRENT_MODE_TYPE || type === REACT_PROFILER_TYPE || type === REACT_STRICT_MODE_TYPE || type === REACT_SUSPENSE_TYPE || typeof type === 'object' && type !== null && (type.$$typeof === REACT_LAZY_TYPE || type.$$typeof === REACT_MEMO_TYPE || type.$$typeof === REACT_PROVIDER_TYPE || type.$$typeof === REACT_CONTEXT_TYPE || type.$$typeof === REACT_FORWARD_REF_TYPE);
***REMOVED***

/**
 * Forked from fbjs/warning:
 * https://github.com/facebook/fbjs/blob/e66ba20ad5be433eb54423f2b097d829324d9de6/packages/fbjs/src/__forks__/warning.js
 *
 * Only change is we use console.warn instead of console.error,
 * and do nothing when 'console' is not supported.
 * This really simplifies the code.
 * ---
 * Similar to invariant but only logs a warning if the condition is not met.
 * This can be used to log issues in development environments in critical
 * paths. Removing the logging code for production environments will keep the
 * same logic and follow the same code paths.
 */

var lowPriorityWarning = function () ***REMOVED******REMOVED***;

***REMOVED***
  var printWarning = function (format) ***REMOVED***
    for (var _len = arguments.length, args = Array(_len > 1 ? _len - 1 : 0), _key = 1; _key < _len; _key++) ***REMOVED***
      args[_key - 1] = arguments[_key];
    ***REMOVED***

    var argIndex = 0;
    var message = 'Warning: ' + format.replace(/%s/g, function () ***REMOVED***
      return args[argIndex++];
    ***REMOVED***);
    if (typeof console !== 'undefined') ***REMOVED***
      console.warn(message);
    ***REMOVED***
    try ***REMOVED***
      // --- Welcome to debugging React ---
      // This error was thrown as a convenience so that you can use this stack
      // to find the callsite that caused this warning to fire.
      throw new Error(message);
    ***REMOVED*** catch (x) ***REMOVED******REMOVED***
  ***REMOVED***;

  lowPriorityWarning = function (condition, format) ***REMOVED***
    if (format === undefined) ***REMOVED***
      throw new Error('`lowPriorityWarning(condition, format, ...args)` requires a warning ' + 'message argument');
    ***REMOVED***
    if (!condition) ***REMOVED***
      for (var _len2 = arguments.length, args = Array(_len2 > 2 ? _len2 - 2 : 0), _key2 = 2; _key2 < _len2; _key2++) ***REMOVED***
        args[_key2 - 2] = arguments[_key2];
      ***REMOVED***

      printWarning.apply(undefined, [format].concat(args));
    ***REMOVED***
  ***REMOVED***;
***REMOVED***

var lowPriorityWarning$1 = lowPriorityWarning;

function typeOf(object) ***REMOVED***
  if (typeof object === 'object' && object !== null) ***REMOVED***
    var $$typeof = object.$$typeof;
    switch ($$typeof) ***REMOVED***
      case REACT_ELEMENT_TYPE:
        var type = object.type;

        switch (type) ***REMOVED***
          case REACT_ASYNC_MODE_TYPE:
          case REACT_CONCURRENT_MODE_TYPE:
          case REACT_FRAGMENT_TYPE:
          case REACT_PROFILER_TYPE:
          case REACT_STRICT_MODE_TYPE:
          case REACT_SUSPENSE_TYPE:
            return type;
          default:
            var $$typeofType = type && type.$$typeof;

            switch ($$typeofType) ***REMOVED***
              case REACT_CONTEXT_TYPE:
              case REACT_FORWARD_REF_TYPE:
              case REACT_PROVIDER_TYPE:
                return $$typeofType;
              default:
                return $$typeof;
            ***REMOVED***
        ***REMOVED***
      case REACT_LAZY_TYPE:
      case REACT_MEMO_TYPE:
      case REACT_PORTAL_TYPE:
        return $$typeof;
    ***REMOVED***
  ***REMOVED***

  return undefined;
***REMOVED***

// AsyncMode is deprecated along with isAsyncMode
var AsyncMode = REACT_ASYNC_MODE_TYPE;
var ConcurrentMode = REACT_CONCURRENT_MODE_TYPE;
var ContextConsumer = REACT_CONTEXT_TYPE;
var ContextProvider = REACT_PROVIDER_TYPE;
var Element = REACT_ELEMENT_TYPE;
var ForwardRef = REACT_FORWARD_REF_TYPE;
var Fragment = REACT_FRAGMENT_TYPE;
var Lazy = REACT_LAZY_TYPE;
var Memo = REACT_MEMO_TYPE;
var Portal = REACT_PORTAL_TYPE;
var Profiler = REACT_PROFILER_TYPE;
var StrictMode = REACT_STRICT_MODE_TYPE;
var Suspense = REACT_SUSPENSE_TYPE;

var hasWarnedAboutDeprecatedIsAsyncMode = false;

// AsyncMode should be deprecated
function isAsyncMode(object) ***REMOVED***
  ***REMOVED***
    if (!hasWarnedAboutDeprecatedIsAsyncMode) ***REMOVED***
      hasWarnedAboutDeprecatedIsAsyncMode = true;
      lowPriorityWarning$1(false, 'The ReactIs.isAsyncMode() alias has been deprecated, ' + 'and will be removed in React 17+. Update your code to use ' + 'ReactIs.isConcurrentMode() instead. It has the exact same API.');
    ***REMOVED***
  ***REMOVED***
  return isConcurrentMode(object) || typeOf(object) === REACT_ASYNC_MODE_TYPE;
***REMOVED***
function isConcurrentMode(object) ***REMOVED***
  return typeOf(object) === REACT_CONCURRENT_MODE_TYPE;
***REMOVED***
function isContextConsumer(object) ***REMOVED***
  return typeOf(object) === REACT_CONTEXT_TYPE;
***REMOVED***
function isContextProvider(object) ***REMOVED***
  return typeOf(object) === REACT_PROVIDER_TYPE;
***REMOVED***
function isElement(object) ***REMOVED***
  return typeof object === 'object' && object !== null && object.$$typeof === REACT_ELEMENT_TYPE;
***REMOVED***
function isForwardRef(object) ***REMOVED***
  return typeOf(object) === REACT_FORWARD_REF_TYPE;
***REMOVED***
function isFragment(object) ***REMOVED***
  return typeOf(object) === REACT_FRAGMENT_TYPE;
***REMOVED***
function isLazy(object) ***REMOVED***
  return typeOf(object) === REACT_LAZY_TYPE;
***REMOVED***
function isMemo(object) ***REMOVED***
  return typeOf(object) === REACT_MEMO_TYPE;
***REMOVED***
function isPortal(object) ***REMOVED***
  return typeOf(object) === REACT_PORTAL_TYPE;
***REMOVED***
function isProfiler(object) ***REMOVED***
  return typeOf(object) === REACT_PROFILER_TYPE;
***REMOVED***
function isStrictMode(object) ***REMOVED***
  return typeOf(object) === REACT_STRICT_MODE_TYPE;
***REMOVED***
function isSuspense(object) ***REMOVED***
  return typeOf(object) === REACT_SUSPENSE_TYPE;
***REMOVED***

exports.typeOf = typeOf;
exports.AsyncMode = AsyncMode;
exports.ConcurrentMode = ConcurrentMode;
exports.ContextConsumer = ContextConsumer;
exports.ContextProvider = ContextProvider;
exports.Element = Element;
exports.ForwardRef = ForwardRef;
exports.Fragment = Fragment;
exports.Lazy = Lazy;
exports.Memo = Memo;
exports.Portal = Portal;
exports.Profiler = Profiler;
exports.StrictMode = StrictMode;
exports.Suspense = Suspense;
exports.isValidElementType = isValidElementType;
exports.isAsyncMode = isAsyncMode;
exports.isConcurrentMode = isConcurrentMode;
exports.isContextConsumer = isContextConsumer;
exports.isContextProvider = isContextProvider;
exports.isElement = isElement;
exports.isForwardRef = isForwardRef;
exports.isFragment = isFragment;
exports.isLazy = isLazy;
exports.isMemo = isMemo;
exports.isPortal = isPortal;
exports.isProfiler = isProfiler;
exports.isStrictMode = isStrictMode;
exports.isSuspense = isSuspense;
  ***REMOVED***)();
***REMOVED***

***REMOVED***).call(this,require('_process'))
***REMOVED***,***REMOVED***"_process":7***REMOVED***],9:[function(require,module,exports)***REMOVED***
/** @license React v16.8.1
 * react-is.production.min.js
 *
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

'use strict';Object.defineProperty(exports,"__esModule",***REMOVED***value:!0***REMOVED***);
var b="function"===typeof Symbol&&Symbol.for,c=b?Symbol.for("react.element"):60103,d=b?Symbol.for("react.portal"):60106,e=b?Symbol.for("react.fragment"):60107,f=b?Symbol.for("react.strict_mode"):60108,g=b?Symbol.for("react.profiler"):60114,h=b?Symbol.for("react.provider"):60109,k=b?Symbol.for("react.context"):60110,l=b?Symbol.for("react.async_mode"):60111,m=b?Symbol.for("react.concurrent_mode"):60111,n=b?Symbol.for("react.forward_ref"):60112,p=b?Symbol.for("react.suspense"):60113,q=b?Symbol.for("react.memo"):
60115,r=b?Symbol.for("react.lazy"):60116;function t(a)***REMOVED***if("object"===typeof a&&null!==a)***REMOVED***var u=a.$$typeof;switch(u)***REMOVED***case c:switch(a=a.type,a)***REMOVED***case l:case m:case e:case g:case f:case p:return a;default:switch(a=a&&a.$$typeof,a)***REMOVED***case k:case n:case h:return a;default:return u***REMOVED******REMOVED***case r:case q:case d:return u***REMOVED******REMOVED******REMOVED***function v(a)***REMOVED***return t(a)===m***REMOVED***exports.typeOf=t;exports.AsyncMode=l;exports.ConcurrentMode=m;exports.ContextConsumer=k;exports.ContextProvider=h;exports.Element=c;exports.ForwardRef=n;
exports.Fragment=e;exports.Lazy=r;exports.Memo=q;exports.Portal=d;exports.Profiler=g;exports.StrictMode=f;exports.Suspense=p;exports.isValidElementType=function(a)***REMOVED***return"string"===typeof a||"function"===typeof a||a===e||a===m||a===g||a===f||a===p||"object"===typeof a&&null!==a&&(a.$$typeof===r||a.$$typeof===q||a.$$typeof===h||a.$$typeof===k||a.$$typeof===n)***REMOVED***;exports.isAsyncMode=function(a)***REMOVED***return v(a)||t(a)===l***REMOVED***;exports.isConcurrentMode=v;exports.isContextConsumer=function(a)***REMOVED***return t(a)===k***REMOVED***;
exports.isContextProvider=function(a)***REMOVED***return t(a)===h***REMOVED***;exports.isElement=function(a)***REMOVED***return"object"===typeof a&&null!==a&&a.$$typeof===c***REMOVED***;exports.isForwardRef=function(a)***REMOVED***return t(a)===n***REMOVED***;exports.isFragment=function(a)***REMOVED***return t(a)===e***REMOVED***;exports.isLazy=function(a)***REMOVED***return t(a)===r***REMOVED***;exports.isMemo=function(a)***REMOVED***return t(a)===q***REMOVED***;exports.isPortal=function(a)***REMOVED***return t(a)===d***REMOVED***;exports.isProfiler=function(a)***REMOVED***return t(a)===g***REMOVED***;exports.isStrictMode=function(a)***REMOVED***return t(a)===f***REMOVED***;
exports.isSuspense=function(a)***REMOVED***return t(a)===p***REMOVED***;

***REMOVED***,***REMOVED******REMOVED***],10:[function(require,module,exports)***REMOVED***
(function (process)***REMOVED***
'use strict';

if (process.env.NODE_ENV === 'production') ***REMOVED***
  module.exports = require('./cjs/react-is.production.min.js');
***REMOVED*** else ***REMOVED***
  module.exports = require('./cjs/react-is.development.js');
***REMOVED***

***REMOVED***).call(this,require('_process'))
***REMOVED***,***REMOVED***"./cjs/react-is.development.js":8,"./cjs/react-is.production.min.js":9,"_process":7***REMOVED***]***REMOVED***,***REMOVED******REMOVED***,[4])(4)
***REMOVED***);
