// rxjs/src/internal/observable/interval.ts
import ***REMOVED*** Observable ***REMOVED*** from '../Observable';
import ***REMOVED*** async ***REMOVED*** from '../scheduler/async';
import ***REMOVED*** SchedulerAction, SchedulerLike ***REMOVED*** from '../types';
import ***REMOVED*** isNumeric ***REMOVED*** from '../util/isNumeric';
import ***REMOVED*** Subscriber ***REMOVED*** from '../Subscriber';

/**
 * Creates an Observable that emits sequential numbers every specified
 * interval of time, on a specified ***REMOVED***@link SchedulerLike***REMOVED***.
 *
 * <span class="informal">Emits incremental numbers periodically in time.
 * </span>
 *
 * ![](interval.png)
 *
 * `interval` returns an Observable that emits an infinite sequence of
 * ascending integers, with a constant interval of time of your choosing
 * between those emissions. The first emission is not sent immediately, but
 * only after the first period has passed. By default, this operator uses the
 * `async` ***REMOVED***@link SchedulerLike***REMOVED*** to provide a notion of time, but you may pass any
 * ***REMOVED***@link SchedulerLike***REMOVED*** to it.
 *
 * ## Example
 * Emits ascending numbers, one every second (1000ms) up to the number 3
 * ```ts
 * import ***REMOVED*** interval ***REMOVED*** from 'rxjs';
 * import ***REMOVED*** take ***REMOVED*** from 'rxjs/operators';
 *
 * const numbers = interval(1000);
 *
 * const takeFourNumbers = numbers.pipe(take(4));
 *
 * takeFourNumbers.subscribe(x => console.log('Next: ', x));
 *
 * // Logs:
 * // Next: 0
 * // Next: 1
 * // Next: 2
 * // Next: 3
 * ```
 *
 * @see ***REMOVED***@link timer***REMOVED***
 * @see ***REMOVED***@link delay***REMOVED***
 *
 * @param ***REMOVED***number***REMOVED*** [period=0] The interval size in milliseconds (by default)
 * or the time unit determined by the scheduler's clock.
 * @param ***REMOVED***SchedulerLike***REMOVED*** [scheduler=async] The ***REMOVED***@link SchedulerLike***REMOVED*** to use for scheduling
 * the emission of values, and providing a notion of "time".
 * @return ***REMOVED***Observable***REMOVED*** An Observable that emits a sequential number each time
 * interval.
 * @static true
 * @name interval
 * @owner Observable
 */
export function interval(period = 0,
                         scheduler: SchedulerLike = async): Observable<number> ***REMOVED***
  if (!isNumeric(period) || period < 0) ***REMOVED***
    period = 0;
  ***REMOVED***

  if (!scheduler || typeof scheduler.schedule !== 'function') ***REMOVED***
    scheduler = async;
  ***REMOVED***

  return new Observable<number>(subscriber => ***REMOVED***
    subscriber.add(
      scheduler.schedule(dispatch, period, ***REMOVED*** subscriber, counter: 0, period ***REMOVED***)
    );
    return subscriber;
  ***REMOVED***);
***REMOVED***

function dispatch(this: SchedulerAction<IntervalState>, state: IntervalState) ***REMOVED***
  const ***REMOVED*** subscriber, counter, period ***REMOVED*** = state;
  subscriber.next(counter);
  this.schedule(***REMOVED*** subscriber, counter: counter + 1, period ***REMOVED***, period);
***REMOVED***

interface IntervalState ***REMOVED***
  subscriber: Subscriber<number>;
  counter: number;
  period: number;
***REMOVED***
