// rxjs/src/internal/observable/timer.ts
import ***REMOVED*** Observable ***REMOVED*** from '../Observable';
import ***REMOVED*** SchedulerAction, SchedulerLike ***REMOVED*** from '../types';
import ***REMOVED*** async ***REMOVED*** from '../scheduler/async';
import ***REMOVED*** isNumeric ***REMOVED*** from '../util/isNumeric';
import ***REMOVED*** isScheduler ***REMOVED*** from '../util/isScheduler';
import ***REMOVED*** Subscriber ***REMOVED*** from '../Subscriber';

/**
 * Creates an Observable that starts emitting after an `dueTime` and
 * emits ever increasing numbers after each `period` of time thereafter.
 *
 * <span class="informal">Its like ***REMOVED***@link index/interval***REMOVED***, but you can specify when
 * should the emissions start.</span>
 *
 * ![](timer.png)
 *
 * `timer` returns an Observable that emits an infinite sequence of ascending
 * integers, with a constant interval of time, `period` of your choosing
 * between those emissions. The first emission happens after the specified
 * `dueTime`. The initial delay may be a `Date`. By default, this
 * operator uses the ***REMOVED***@link asyncScheduler***REMOVED*** ***REMOVED***@link SchedulerLike***REMOVED*** to provide a notion of time, but you
 * may pass any ***REMOVED***@link SchedulerLike***REMOVED*** to it. If `period` is not specified, the output
 * Observable emits only one value, `0`. Otherwise, it emits an infinite
 * sequence.
 *
 * ## Examples
 * ### Emits ascending numbers, one every second (1000ms), starting after 3 seconds
 * ```ts
 * import ***REMOVED*** timer ***REMOVED*** from 'rxjs';
 *
 * const numbers = timer(3000, 1000);
 * numbers.subscribe(x => console.log(x));
 * ```
 *
 * ### Emits one number after five seconds
 * ```ts
 * import ***REMOVED*** timer ***REMOVED*** from 'rxjs';
 *
 * const numbers = timer(5000);
 * numbers.subscribe(x => console.log(x));
 * ```
 * @see ***REMOVED***@link index/interval***REMOVED***
 * @see ***REMOVED***@link delay***REMOVED***
 *
 * @param ***REMOVED***number|Date***REMOVED*** [dueTime] The initial delay time specified as a Date object or as an integer denoting
 * milliseconds to wait before emitting the first value of 0`.
 * @param ***REMOVED***number|SchedulerLike***REMOVED*** [periodOrScheduler] The period of time between emissions of the
 * subsequent numbers.
 * @param ***REMOVED***SchedulerLike***REMOVED*** [scheduler=async] The ***REMOVED***@link SchedulerLike***REMOVED*** to use for scheduling
 * the emission of values, and providing a notion of "time".
 * @return ***REMOVED***Observable***REMOVED*** An Observable that emits a `0` after the
 * `dueTime` and ever increasing numbers after each `period` of time
 * thereafter.
 * @static true
 * @name timer
 * @owner Observable
 */
export function timer(dueTime: number | Date = 0,
                      periodOrScheduler?: number | SchedulerLike,
                      scheduler?: SchedulerLike): Observable<number> ***REMOVED***
  let period = -1;
  if (isNumeric(periodOrScheduler)) ***REMOVED***
    period = Number(periodOrScheduler) < 1 && 1 || Number(periodOrScheduler);
  ***REMOVED*** else if (isScheduler(periodOrScheduler)) ***REMOVED***
    scheduler = periodOrScheduler as any;
  ***REMOVED***

  if (!isScheduler(scheduler)) ***REMOVED***
    scheduler = async;
  ***REMOVED***

  return new Observable(subscriber => ***REMOVED***
    const due = isNumeric(dueTime)
      ? (dueTime as number)
      : (+dueTime - scheduler.now());

    return scheduler.schedule(dispatch, due, ***REMOVED***
      index: 0, period, subscriber
    ***REMOVED***);
  ***REMOVED***);
***REMOVED***

interface TimerState ***REMOVED***
  index: number;
  period: number;
  subscriber: Subscriber<number>;
***REMOVED***

function dispatch(this: SchedulerAction<TimerState>, state: TimerState) ***REMOVED***
  const ***REMOVED*** index, period, subscriber ***REMOVED*** = state;
  subscriber.next(index);

  if (subscriber.closed) ***REMOVED***
    return;
  ***REMOVED*** else if (period === -1) ***REMOVED***
    return subscriber.complete();
  ***REMOVED***

  state.index = index + 1;
  this.schedule(state, period);
***REMOVED***
