// rxjs/internal/util/Immediate.js
"use strict";
Object.defineProperty(exports, "__esModule", ***REMOVED*** value: true ***REMOVED***);
var nextHandle = 1;
var RESOLVED = (function () ***REMOVED*** return Promise.resolve(); ***REMOVED***)();
var activeHandles = ***REMOVED******REMOVED***;
function findAndClearHandle(handle) ***REMOVED***
    if (handle in activeHandles) ***REMOVED***
        delete activeHandles[handle];
        return true;
    ***REMOVED***
    return false;
***REMOVED***
exports.Immediate = ***REMOVED***
    setImmediate: function (cb) ***REMOVED***
        var handle = nextHandle++;
        activeHandles[handle] = true;
        RESOLVED.then(function () ***REMOVED*** return findAndClearHandle(handle) && cb(); ***REMOVED***);
        return handle;
    ***REMOVED***,
    clearImmediate: function (handle) ***REMOVED***
        findAndClearHandle(handle);
    ***REMOVED***,
***REMOVED***;
exports.TestTools = ***REMOVED***
    pending: function () ***REMOVED***
        return Object.keys(activeHandles).length;
    ***REMOVED***
***REMOVED***;
