// rxjs/_esm5/internal/Notification.js
/** PURE_IMPORTS_START _observable_empty,_observable_of,_observable_throwError PURE_IMPORTS_END */
import ***REMOVED*** empty ***REMOVED*** from './observable/empty';
import ***REMOVED*** of ***REMOVED*** from './observable/of';
import ***REMOVED*** throwError ***REMOVED*** from './observable/throwError';
export var NotificationKind;
/*@__PURE__*/ (function (NotificationKind) ***REMOVED***
    NotificationKind["NEXT"] = "N";
    NotificationKind["ERROR"] = "E";
    NotificationKind["COMPLETE"] = "C";
***REMOVED***)(NotificationKind || (NotificationKind = ***REMOVED******REMOVED***));
var Notification = /*@__PURE__*/ (function () ***REMOVED***
    function Notification(kind, value, error) ***REMOVED***
        this.kind = kind;
        this.value = value;
        this.error = error;
        this.hasValue = kind === 'N';
    ***REMOVED***
    Notification.prototype.observe = function (observer) ***REMOVED***
        switch (this.kind) ***REMOVED***
            case 'N':
                return observer.next && observer.next(this.value);
            case 'E':
                return observer.error && observer.error(this.error);
            case 'C':
                return observer.complete && observer.complete();
        ***REMOVED***
    ***REMOVED***;
    Notification.prototype.do = function (next, error, complete) ***REMOVED***
        var kind = this.kind;
        switch (kind) ***REMOVED***
            case 'N':
                return next && next(this.value);
            case 'E':
                return error && error(this.error);
            case 'C':
                return complete && complete();
        ***REMOVED***
    ***REMOVED***;
    Notification.prototype.accept = function (nextOrObserver, error, complete) ***REMOVED***
        if (nextOrObserver && typeof nextOrObserver.next === 'function') ***REMOVED***
            return this.observe(nextOrObserver);
        ***REMOVED***
        else ***REMOVED***
            return this.do(nextOrObserver, error, complete);
        ***REMOVED***
    ***REMOVED***;
    Notification.prototype.toObservable = function () ***REMOVED***
        var kind = this.kind;
        switch (kind) ***REMOVED***
            case 'N':
                return of(this.value);
            case 'E':
                return throwError(this.error);
            case 'C':
                return empty();
        ***REMOVED***
        throw new Error('unexpected notification kind value');
    ***REMOVED***;
    Notification.createNext = function (value) ***REMOVED***
        if (typeof value !== 'undefined') ***REMOVED***
            return new Notification('N', value);
        ***REMOVED***
        return Notification.undefinedValueNotification;
    ***REMOVED***;
    Notification.createError = function (err) ***REMOVED***
        return new Notification('E', undefined, err);
    ***REMOVED***;
    Notification.createComplete = function () ***REMOVED***
        return Notification.completeNotification;
    ***REMOVED***;
    Notification.completeNotification = new Notification('C');
    Notification.undefinedValueNotification = new Notification('N', undefined);
    return Notification;
***REMOVED***());
export ***REMOVED*** Notification ***REMOVED***;
