// rxjs/_esm2015/internal/operators/exhaustMap.js
import ***REMOVED*** OuterSubscriber ***REMOVED*** from '../OuterSubscriber';
import ***REMOVED*** InnerSubscriber ***REMOVED*** from '../InnerSubscriber';
import ***REMOVED*** subscribeToResult ***REMOVED*** from '../util/subscribeToResult';
import ***REMOVED*** map ***REMOVED*** from './map';
import ***REMOVED*** from ***REMOVED*** from '../observable/from';
export function exhaustMap(project, resultSelector) ***REMOVED***
    if (resultSelector) ***REMOVED***
        return (source) => source.pipe(exhaustMap((a, i) => from(project(a, i)).pipe(map((b, ii) => resultSelector(a, b, i, ii)))));
    ***REMOVED***
    return (source) => source.lift(new ExhaustMapOperator(project));
***REMOVED***
class ExhaustMapOperator ***REMOVED***
    constructor(project) ***REMOVED***
        this.project = project;
    ***REMOVED***
    call(subscriber, source) ***REMOVED***
        return source.subscribe(new ExhaustMapSubscriber(subscriber, this.project));
    ***REMOVED***
***REMOVED***
class ExhaustMapSubscriber extends OuterSubscriber ***REMOVED***
    constructor(destination, project) ***REMOVED***
        super(destination);
        this.project = project;
        this.hasSubscription = false;
        this.hasCompleted = false;
        this.index = 0;
    ***REMOVED***
    _next(value) ***REMOVED***
        if (!this.hasSubscription) ***REMOVED***
            this.tryNext(value);
        ***REMOVED***
    ***REMOVED***
    tryNext(value) ***REMOVED***
        let result;
        const index = this.index++;
        try ***REMOVED***
            result = this.project(value, index);
        ***REMOVED***
        catch (err) ***REMOVED***
            this.destination.error(err);
            return;
        ***REMOVED***
        this.hasSubscription = true;
        this._innerSub(result, value, index);
    ***REMOVED***
    _innerSub(result, value, index) ***REMOVED***
        const innerSubscriber = new InnerSubscriber(this, value, index);
        const destination = this.destination;
        destination.add(innerSubscriber);
        const innerSubscription = subscribeToResult(this, result, undefined, undefined, innerSubscriber);
        if (innerSubscription !== innerSubscriber) ***REMOVED***
            destination.add(innerSubscription);
        ***REMOVED***
    ***REMOVED***
    _complete() ***REMOVED***
        this.hasCompleted = true;
        if (!this.hasSubscription) ***REMOVED***
            this.destination.complete();
        ***REMOVED***
        this.unsubscribe();
    ***REMOVED***
    notifyNext(outerValue, innerValue, outerIndex, innerIndex, innerSub) ***REMOVED***
        this.destination.next(innerValue);
    ***REMOVED***
    notifyError(err) ***REMOVED***
        this.destination.error(err);
    ***REMOVED***
    notifyComplete(innerSub) ***REMOVED***
        const destination = this.destination;
        destination.remove(innerSub);
        this.hasSubscription = false;
        if (this.hasCompleted) ***REMOVED***
            this.destination.complete();
        ***REMOVED***
    ***REMOVED***
***REMOVED***
