// rxjs/_esm2015/internal/operators/find.js
import ***REMOVED*** Subscriber ***REMOVED*** from '../Subscriber';
export function find(predicate, thisArg) ***REMOVED***
    if (typeof predicate !== 'function') ***REMOVED***
        throw new TypeError('predicate is not a function');
    ***REMOVED***
    return (source) => source.lift(new FindValueOperator(predicate, source, false, thisArg));
***REMOVED***
export class FindValueOperator ***REMOVED***
    constructor(predicate, source, yieldIndex, thisArg) ***REMOVED***
        this.predicate = predicate;
        this.source = source;
        this.yieldIndex = yieldIndex;
        this.thisArg = thisArg;
    ***REMOVED***
    call(observer, source) ***REMOVED***
        return source.subscribe(new FindValueSubscriber(observer, this.predicate, this.source, this.yieldIndex, this.thisArg));
    ***REMOVED***
***REMOVED***
export class FindValueSubscriber extends Subscriber ***REMOVED***
    constructor(destination, predicate, source, yieldIndex, thisArg) ***REMOVED***
        super(destination);
        this.predicate = predicate;
        this.source = source;
        this.yieldIndex = yieldIndex;
        this.thisArg = thisArg;
        this.index = 0;
    ***REMOVED***
    notifyComplete(value) ***REMOVED***
        const destination = this.destination;
        destination.next(value);
        destination.complete();
        this.unsubscribe();
    ***REMOVED***
    _next(value) ***REMOVED***
        const ***REMOVED*** predicate, thisArg ***REMOVED*** = this;
        const index = this.index++;
        try ***REMOVED***
            const result = predicate.call(thisArg || this, value, index, this.source);
            if (result) ***REMOVED***
                this.notifyComplete(this.yieldIndex ? index : value);
            ***REMOVED***
        ***REMOVED***
        catch (err) ***REMOVED***
            this.destination.error(err);
        ***REMOVED***
    ***REMOVED***
    _complete() ***REMOVED***
        this.notifyComplete(this.yieldIndex ? -1 : undefined);
    ***REMOVED***
***REMOVED***
