// ajv/lib/definition_schema.js
'use strict';

var metaSchema = require('./refs/json-schema-draft-07.json');

module.exports = ***REMOVED***
  $id: 'https://github.com/epoberezkin/ajv/blob/master/lib/definition_schema.js',
  definitions: ***REMOVED***
    simpleTypes: metaSchema.definitions.simpleTypes
  ***REMOVED***,
  type: 'object',
  dependencies: ***REMOVED***
    schema: ['validate'],
    $data: ['validate'],
    statements: ['inline'],
    valid: ***REMOVED***not: ***REMOVED***required: ['macro']***REMOVED******REMOVED***
  ***REMOVED***,
  properties: ***REMOVED***
    type: metaSchema.properties.type,
    schema: ***REMOVED***type: 'boolean'***REMOVED***,
    statements: ***REMOVED***type: 'boolean'***REMOVED***,
    dependencies: ***REMOVED***
      type: 'array',
      items: ***REMOVED***type: 'string'***REMOVED***
    ***REMOVED***,
    metaSchema: ***REMOVED***type: 'object'***REMOVED***,
    modifying: ***REMOVED***type: 'boolean'***REMOVED***,
    valid: ***REMOVED***type: 'boolean'***REMOVED***,
    $data: ***REMOVED***type: 'boolean'***REMOVED***,
    async: ***REMOVED***type: 'boolean'***REMOVED***,
    errors: ***REMOVED***
      anyOf: [
        ***REMOVED***type: 'boolean'***REMOVED***,
        ***REMOVED***const: 'full'***REMOVED***
      ]
    ***REMOVED***
  ***REMOVED***
***REMOVED***;
