// axios/lib/core/mergeConfig.js
'use strict';

var utils = require('../utils');

/**
 * Config-specific merge-function which creates a new config-object
 * by merging two configuration objects together.
 *
 * @param ***REMOVED***Object***REMOVED*** config1
 * @param ***REMOVED***Object***REMOVED*** config2
 * @returns ***REMOVED***Object***REMOVED*** New object resulting from merging config2 to config1
 */
module.exports = function mergeConfig(config1, config2) ***REMOVED***
  // eslint-disable-next-line no-param-reassign
  config2 = config2 || ***REMOVED******REMOVED***;
  var config = ***REMOVED******REMOVED***;

  var valueFromConfig2Keys = ['url', 'method', 'params', 'data'];
  var mergeDeepPropertiesKeys = ['headers', 'auth', 'proxy'];
  var defaultToConfig2Keys = [
    'baseURL', 'url', 'transformRequest', 'transformResponse', 'paramsSerializer',
    'timeout', 'withCredentials', 'adapter', 'responseType', 'xsrfCookieName',
    'xsrfHeaderName', 'onUploadProgress', 'onDownloadProgress',
    'maxContentLength', 'validateStatus', 'maxRedirects', 'httpAgent',
    'httpsAgent', 'cancelToken', 'socketPath'
  ];

  utils.forEach(valueFromConfig2Keys, function valueFromConfig2(prop) ***REMOVED***
    if (typeof config2[prop] !== 'undefined') ***REMOVED***
      config[prop] = config2[prop];
    ***REMOVED***
  ***REMOVED***);

  utils.forEach(mergeDeepPropertiesKeys, function mergeDeepProperties(prop) ***REMOVED***
    if (utils.isObject(config2[prop])) ***REMOVED***
      config[prop] = utils.deepMerge(config1[prop], config2[prop]);
    ***REMOVED*** else if (typeof config2[prop] !== 'undefined') ***REMOVED***
      config[prop] = config2[prop];
    ***REMOVED*** else if (utils.isObject(config1[prop])) ***REMOVED***
      config[prop] = utils.deepMerge(config1[prop]);
    ***REMOVED*** else if (typeof config1[prop] !== 'undefined') ***REMOVED***
      config[prop] = config1[prop];
    ***REMOVED***
  ***REMOVED***);

  utils.forEach(defaultToConfig2Keys, function defaultToConfig2(prop) ***REMOVED***
    if (typeof config2[prop] !== 'undefined') ***REMOVED***
      config[prop] = config2[prop];
    ***REMOVED*** else if (typeof config1[prop] !== 'undefined') ***REMOVED***
      config[prop] = config1[prop];
    ***REMOVED***
  ***REMOVED***);

  var axiosKeys = valueFromConfig2Keys
    .concat(mergeDeepPropertiesKeys)
    .concat(defaultToConfig2Keys);

  var otherKeys = Object
    .keys(config2)
    .filter(function filterAxiosKeys(key) ***REMOVED***
      return axiosKeys.indexOf(key) === -1;
    ***REMOVED***);

  utils.forEach(otherKeys, function otherKeysDefaultToConfig2(prop) ***REMOVED***
    if (typeof config2[prop] !== 'undefined') ***REMOVED***
      config[prop] = config2[prop];
    ***REMOVED*** else if (typeof config1[prop] !== 'undefined') ***REMOVED***
      config[prop] = config1[prop];
    ***REMOVED***
  ***REMOVED***);

  return config;
***REMOVED***;
