// discord.js/src/client/voice/opus/BaseOpusEngine.js
/**
 * The base opus encoding engine.
 * @private
 */
class BaseOpus ***REMOVED***
  /**
   * @param ***REMOVED***Object***REMOVED*** [options] The options to apply to the Opus engine
   * @param ***REMOVED***number***REMOVED*** [options.bitrate=48] The desired bitrate (kbps)
   * @param ***REMOVED***boolean***REMOVED*** [options.fec=false] Whether to enable forward error correction
   * @param ***REMOVED***number***REMOVED*** [options.plp=0] The expected packet loss percentage
   */
  constructor(***REMOVED*** bitrate = 48, fec = false, plp = 0 ***REMOVED*** = ***REMOVED******REMOVED***) ***REMOVED***
    this.ctl = ***REMOVED***
      BITRATE: 4002,
      FEC: 4012,
      PLP: 4014,
    ***REMOVED***;

    this.samplingRate = 48000;
    this.channels = 2;

    /**
     * The desired bitrate (kbps)
     * @type ***REMOVED***number***REMOVED***
     */
    this.bitrate = bitrate;

    /**
     * Miscellaneous Opus options
     * @type ***REMOVED***Object***REMOVED***
     */
    this.options = ***REMOVED*** fec, plp ***REMOVED***;
  ***REMOVED***

  init() ***REMOVED***
    try ***REMOVED***
      this.setBitrate(this.bitrate);

      // Set FEC (forward error correction)
      if (this.options.fec) this.setFEC(this.options.fec);

      // Set PLP (expected packet loss percentage)
      if (this.options.plp) this.setPLP(this.options.plp);
    ***REMOVED*** catch (err) ***REMOVED***
      // Opus engine likely has no support for libopus CTL
    ***REMOVED***
  ***REMOVED***

  encode(buffer) ***REMOVED***
    return buffer;
  ***REMOVED***

  decode(buffer) ***REMOVED***
    return buffer;
  ***REMOVED***

  destroy() ***REMOVED******REMOVED*** // eslint-disable-line no-empty-function
***REMOVED***

module.exports = BaseOpus;
