// doctrine/lib/typed.js
/*
  Copyright (C) 2012-2014 Yusuke Suzuki <utatane.tea@gmail.com>
  Copyright (C) 2014 Dan Tao <daniel.tao@gmail.com>
  Copyright (C) 2013 Andrew Eisenberg <andrew@eisenberg.as>

  Redistribution and use in source and binary forms, with or without
  modification, are permitted provided that the following conditions are met:

    * Redistributions of source code must retain the above copyright
      notice, this list of conditions and the following disclaimer.
    * Redistributions in binary form must reproduce the above copyright
      notice, this list of conditions and the following disclaimer in the
      documentation and/or other materials provided with the distribution.

  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
  ARE DISCLAIMED. IN NO EVENT SHALL <COPYRIGHT HOLDER> BE LIABLE FOR ANY
  DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
  (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
  ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
  (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
  THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
*/

// "typed", the Type Expression Parser for doctrine.

(function () ***REMOVED***
    'use strict';

    var Syntax,
        Token,
        source,
        length,
        index,
        previous,
        token,
        value,
        esutils,
        utility;

    esutils = require('esutils');
    utility = require('./utility');

    Syntax = ***REMOVED***
        NullableLiteral: 'NullableLiteral',
        AllLiteral: 'AllLiteral',
        NullLiteral: 'NullLiteral',
        UndefinedLiteral: 'UndefinedLiteral',
        VoidLiteral: 'VoidLiteral',
        UnionType: 'UnionType',
        ArrayType: 'ArrayType',
        RecordType: 'RecordType',
        FieldType: 'FieldType',
        FunctionType: 'FunctionType',
        ParameterType: 'ParameterType',
        RestType: 'RestType',
        NonNullableType: 'NonNullableType',
        OptionalType: 'OptionalType',
        NullableType: 'NullableType',
        NameExpression: 'NameExpression',
        TypeApplication: 'TypeApplication'
    ***REMOVED***;

    Token = ***REMOVED***
        ILLEGAL: 0,    // ILLEGAL
        DOT_LT: 1,     // .<
        REST: 2,       // ...
        LT: 3,         // <
        GT: 4,         // >
        LPAREN: 5,     // (
        RPAREN: 6,     // )
        LBRACE: 7,     // ***REMOVED***
        RBRACE: 8,     // ***REMOVED***
        LBRACK: 9,    // [
        RBRACK: 10,    // ]
        COMMA: 11,     // ,
        COLON: 12,     // :
        STAR: 13,      // *
        PIPE: 14,      // |
        QUESTION: 15,  // ?
        BANG: 16,      // !
        EQUAL: 17,     // =
        NAME: 18,      // name token
        STRING: 19,    // string
        NUMBER: 20,    // number
        EOF: 21
    ***REMOVED***;

    function isTypeName(ch) ***REMOVED***
        return '><()***REMOVED******REMOVED***[],:*|?!='.indexOf(String.fromCharCode(ch)) === -1 && !esutils.code.isWhiteSpace(ch) && !esutils.code.isLineTerminator(ch);
    ***REMOVED***

    function Context(previous, index, token, value) ***REMOVED***
        this._previous = previous;
        this._index = index;
        this._token = token;
        this._value = value;
    ***REMOVED***

    Context.prototype.restore = function () ***REMOVED***
        previous = this._previous;
        index = this._index;
        token = this._token;
        value = this._value;
    ***REMOVED***;

    Context.save = function () ***REMOVED***
        return new Context(previous, index, token, value);
    ***REMOVED***;

    function advance() ***REMOVED***
        var ch = source.charAt(index);
        index += 1;
        return ch;
    ***REMOVED***

    function scanHexEscape(prefix) ***REMOVED***
        var i, len, ch, code = 0;

        len = (prefix === 'u') ? 4 : 2;
        for (i = 0; i < len; ++i) ***REMOVED***
            if (index < length && esutils.code.isHexDigit(source.charCodeAt(index))) ***REMOVED***
                ch = advance();
                code = code * 16 + '0123456789abcdef'.indexOf(ch.toLowerCase());
            ***REMOVED*** else ***REMOVED***
                return '';
            ***REMOVED***
        ***REMOVED***
        return String.fromCharCode(code);
    ***REMOVED***

    function scanString() ***REMOVED***
        var str = '', quote, ch, code, unescaped, restore; //TODO review removal octal = false
        quote = source.charAt(index);
        ++index;

        while (index < length) ***REMOVED***
            ch = advance();

            if (ch === quote) ***REMOVED***
                quote = '';
                break;
            ***REMOVED*** else if (ch === '\\') ***REMOVED***
                ch = advance();
                if (!esutils.code.isLineTerminator(ch.charCodeAt(0))) ***REMOVED***
                    switch (ch) ***REMOVED***
                    case 'n':
                        str += '\n';
                        break;
                    case 'r':
                        str += '\r';
                        break;
                    case 't':
                        str += '\t';
                        break;
                    case 'u':
                    case 'x':
                        restore = index;
                        unescaped = scanHexEscape(ch);
                        if (unescaped) ***REMOVED***
                            str += unescaped;
                        ***REMOVED*** else ***REMOVED***
                            index = restore;
                            str += ch;
                        ***REMOVED***
                        break;
                    case 'b':
                        str += '\b';
                        break;
                    case 'f':
                        str += '\f';
                        break;
                    case 'v':
                        str += '\v';
                        break;

                    default:
                        if (esutils.code.isOctalDigit(ch.charCodeAt(0))) ***REMOVED***
                            code = '01234567'.indexOf(ch);

                            // \0 is not octal escape sequence
                            // Deprecating unused code. TODO review removal
                            //if (code !== 0) ***REMOVED***
                            //    octal = true;
                            //***REMOVED***

                            if (index < length && esutils.code.isOctalDigit(source.charCodeAt(index))) ***REMOVED***
                                //TODO Review Removal octal = true;
                                code = code * 8 + '01234567'.indexOf(advance());

                                // 3 digits are only allowed when string starts
                                // with 0, 1, 2, 3
                                if ('0123'.indexOf(ch) >= 0 &&
                                        index < length &&
                                        esutils.code.isOctalDigit(source.charCodeAt(index))) ***REMOVED***
                                    code = code * 8 + '01234567'.indexOf(advance());
                                ***REMOVED***
                            ***REMOVED***
                            str += String.fromCharCode(code);
                        ***REMOVED*** else ***REMOVED***
                            str += ch;
                        ***REMOVED***
                        break;
                    ***REMOVED***
                ***REMOVED*** else ***REMOVED***
                    if (ch ===  '\r' && source.charCodeAt(index) === 0x0A  /* '\n' */) ***REMOVED***
                        ++index;
                    ***REMOVED***
                ***REMOVED***
            ***REMOVED*** else if (esutils.code.isLineTerminator(ch.charCodeAt(0))) ***REMOVED***
                break;
            ***REMOVED*** else ***REMOVED***
                str += ch;
            ***REMOVED***
        ***REMOVED***

        if (quote !== '') ***REMOVED***
            utility.throwError('unexpected quote');
        ***REMOVED***

        value = str;
        return Token.STRING;
    ***REMOVED***

    function scanNumber() ***REMOVED***
        var number, ch;

        number = '';
        ch = source.charCodeAt(index);

        if (ch !== 0x2E  /* '.' */) ***REMOVED***
            number = advance();
            ch = source.charCodeAt(index);

            if (number === '0') ***REMOVED***
                if (ch === 0x78  /* 'x' */ || ch === 0x58  /* 'X' */) ***REMOVED***
                    number += advance();
                    while (index < length) ***REMOVED***
                        ch = source.charCodeAt(index);
                        if (!esutils.code.isHexDigit(ch)) ***REMOVED***
                            break;
                        ***REMOVED***
                        number += advance();
                    ***REMOVED***

                    if (number.length <= 2) ***REMOVED***
                        // only 0x
                        utility.throwError('unexpected token');
                    ***REMOVED***

                    if (index < length) ***REMOVED***
                        ch = source.charCodeAt(index);
                        if (esutils.code.isIdentifierStart(ch)) ***REMOVED***
                            utility.throwError('unexpected token');
                        ***REMOVED***
                    ***REMOVED***
                    value = parseInt(number, 16);
                    return Token.NUMBER;
                ***REMOVED***

                if (esutils.code.isOctalDigit(ch)) ***REMOVED***
                    number += advance();
                    while (index < length) ***REMOVED***
                        ch = source.charCodeAt(index);
                        if (!esutils.code.isOctalDigit(ch)) ***REMOVED***
                            break;
                        ***REMOVED***
                        number += advance();
                    ***REMOVED***

                    if (index < length) ***REMOVED***
                        ch = source.charCodeAt(index);
                        if (esutils.code.isIdentifierStart(ch) || esutils.code.isDecimalDigit(ch)) ***REMOVED***
                            utility.throwError('unexpected token');
                        ***REMOVED***
                    ***REMOVED***
                    value = parseInt(number, 8);
                    return Token.NUMBER;
                ***REMOVED***

                if (esutils.code.isDecimalDigit(ch)) ***REMOVED***
                    utility.throwError('unexpected token');
                ***REMOVED***
            ***REMOVED***

            while (index < length) ***REMOVED***
                ch = source.charCodeAt(index);
                if (!esutils.code.isDecimalDigit(ch)) ***REMOVED***
                    break;
                ***REMOVED***
                number += advance();
            ***REMOVED***
        ***REMOVED***

        if (ch === 0x2E  /* '.' */) ***REMOVED***
            number += advance();
            while (index < length) ***REMOVED***
                ch = source.charCodeAt(index);
                if (!esutils.code.isDecimalDigit(ch)) ***REMOVED***
                    break;
                ***REMOVED***
                number += advance();
            ***REMOVED***
        ***REMOVED***

        if (ch === 0x65  /* 'e' */ || ch === 0x45  /* 'E' */) ***REMOVED***
            number += advance();

            ch = source.charCodeAt(index);
            if (ch === 0x2B  /* '+' */ || ch === 0x2D  /* '-' */) ***REMOVED***
                number += advance();
            ***REMOVED***

            ch = source.charCodeAt(index);
            if (esutils.code.isDecimalDigit(ch)) ***REMOVED***
                number += advance();
                while (index < length) ***REMOVED***
                    ch = source.charCodeAt(index);
                    if (!esutils.code.isDecimalDigit(ch)) ***REMOVED***
                        break;
                    ***REMOVED***
                    number += advance();
                ***REMOVED***
            ***REMOVED*** else ***REMOVED***
                utility.throwError('unexpected token');
            ***REMOVED***
        ***REMOVED***

        if (index < length) ***REMOVED***
            ch = source.charCodeAt(index);
            if (esutils.code.isIdentifierStart(ch)) ***REMOVED***
                utility.throwError('unexpected token');
            ***REMOVED***
        ***REMOVED***

        value = parseFloat(number);
        return Token.NUMBER;
    ***REMOVED***


    function scanTypeName() ***REMOVED***
        var ch, ch2;

        value = advance();
        while (index < length && isTypeName(source.charCodeAt(index))) ***REMOVED***
            ch = source.charCodeAt(index);
            if (ch === 0x2E  /* '.' */) ***REMOVED***
                if ((index + 1) >= length) ***REMOVED***
                    return Token.ILLEGAL;
                ***REMOVED***
                ch2 = source.charCodeAt(index + 1);
                if (ch2 === 0x3C  /* '<' */) ***REMOVED***
                    break;
                ***REMOVED***
            ***REMOVED***
            value += advance();
        ***REMOVED***
        return Token.NAME;
    ***REMOVED***

    function next() ***REMOVED***
        var ch;

        previous = index;

        while (index < length && esutils.code.isWhiteSpace(source.charCodeAt(index))) ***REMOVED***
            advance();
        ***REMOVED***
        if (index >= length) ***REMOVED***
            token = Token.EOF;
            return token;
        ***REMOVED***

        ch = source.charCodeAt(index);
        switch (ch) ***REMOVED***
        case 0x27:  /* ''' */
        case 0x22:  /* '"' */
            token = scanString();
            return token;

        case 0x3A:  /* ':' */
            advance();
            token = Token.COLON;
            return token;

        case 0x2C:  /* ',' */
            advance();
            token = Token.COMMA;
            return token;

        case 0x28:  /* '(' */
            advance();
            token = Token.LPAREN;
            return token;

        case 0x29:  /* ')' */
            advance();
            token = Token.RPAREN;
            return token;

        case 0x5B:  /* '[' */
            advance();
            token = Token.LBRACK;
            return token;

        case 0x5D:  /* ']' */
            advance();
            token = Token.RBRACK;
            return token;

        case 0x7B:  /* '***REMOVED***' */
            advance();
            token = Token.LBRACE;
            return token;

        case 0x7D:  /* '***REMOVED***' */
            advance();
            token = Token.RBRACE;
            return token;

        case 0x2E:  /* '.' */
            if (index + 1 < length) ***REMOVED***
                ch = source.charCodeAt(index + 1);
                if (ch === 0x3C  /* '<' */) ***REMOVED***
                    advance();  // '.'
                    advance();  // '<'
                    token = Token.DOT_LT;
                    return token;
                ***REMOVED***

                if (ch === 0x2E  /* '.' */ && index + 2 < length && source.charCodeAt(index + 2) === 0x2E  /* '.' */) ***REMOVED***
                    advance();  // '.'
                    advance();  // '.'
                    advance();  // '.'
                    token = Token.REST;
                    return token;
                ***REMOVED***

                if (esutils.code.isDecimalDigit(ch)) ***REMOVED***
                    token = scanNumber();
                    return token;
                ***REMOVED***
            ***REMOVED***
            token = Token.ILLEGAL;
            return token;

        case 0x3C:  /* '<' */
            advance();
            token = Token.LT;
            return token;

        case 0x3E:  /* '>' */
            advance();
            token = Token.GT;
            return token;

        case 0x2A:  /* '*' */
            advance();
            token = Token.STAR;
            return token;

        case 0x7C:  /* '|' */
            advance();
            token = Token.PIPE;
            return token;

        case 0x3F:  /* '?' */
            advance();
            token = Token.QUESTION;
            return token;

        case 0x21:  /* '!' */
            advance();
            token = Token.BANG;
            return token;

        case 0x3D:  /* '=' */
            advance();
            token = Token.EQUAL;
            return token;

        default:
            if (esutils.code.isDecimalDigit(ch)) ***REMOVED***
                token = scanNumber();
                return token;
            ***REMOVED***

            // type string permits following case,
            //
            // namespace.module.MyClass
            //
            // this reduced 1 token TK_NAME
            utility.assert(isTypeName(ch));
            token = scanTypeName();
            return token;
        ***REMOVED***
    ***REMOVED***

    function consume(target, text) ***REMOVED***
        utility.assert(token === target, text || 'consumed token not matched');
        next();
    ***REMOVED***

    function expect(target, message) ***REMOVED***
        if (token !== target) ***REMOVED***
            utility.throwError(message || 'unexpected token');
        ***REMOVED***
        next();
    ***REMOVED***

    // UnionType := '(' TypeUnionList ')'
    //
    // TypeUnionList :=
    //     <<empty>>
    //   | NonemptyTypeUnionList
    //
    // NonemptyTypeUnionList :=
    //     TypeExpression
    //   | TypeExpression '|' NonemptyTypeUnionList
    function parseUnionType() ***REMOVED***
        var elements;
        consume(Token.LPAREN, 'UnionType should start with (');
        elements = [];
        if (token !== Token.RPAREN) ***REMOVED***
            while (true) ***REMOVED***
                elements.push(parseTypeExpression());
                if (token === Token.RPAREN) ***REMOVED***
                    break;
                ***REMOVED***
                expect(Token.PIPE);
            ***REMOVED***
        ***REMOVED***
        consume(Token.RPAREN, 'UnionType should end with )');
        return ***REMOVED***
            type: Syntax.UnionType,
            elements: elements
        ***REMOVED***;
    ***REMOVED***

    // ArrayType := '[' ElementTypeList ']'
    //
    // ElementTypeList :=
    //     <<empty>>
    //  | TypeExpression
    //  | '...' TypeExpression
    //  | TypeExpression ',' ElementTypeList
    function parseArrayType() ***REMOVED***
        var elements;
        consume(Token.LBRACK, 'ArrayType should start with [');
        elements = [];
        while (token !== Token.RBRACK) ***REMOVED***
            if (token === Token.REST) ***REMOVED***
                consume(Token.REST);
                elements.push(***REMOVED***
                    type: Syntax.RestType,
                    expression: parseTypeExpression()
                ***REMOVED***);
                break;
            ***REMOVED*** else ***REMOVED***
                elements.push(parseTypeExpression());
            ***REMOVED***
            if (token !== Token.RBRACK) ***REMOVED***
                expect(Token.COMMA);
            ***REMOVED***
        ***REMOVED***
        expect(Token.RBRACK);
        return ***REMOVED***
            type: Syntax.ArrayType,
            elements: elements
        ***REMOVED***;
    ***REMOVED***

    function parseFieldName() ***REMOVED***
        var v = value;
        if (token === Token.NAME || token === Token.STRING) ***REMOVED***
            next();
            return v;
        ***REMOVED***

        if (token === Token.NUMBER) ***REMOVED***
            consume(Token.NUMBER);
            return String(v);
        ***REMOVED***

        utility.throwError('unexpected token');
    ***REMOVED***

    // FieldType :=
    //     FieldName
    //   | FieldName ':' TypeExpression
    //
    // FieldName :=
    //     NameExpression
    //   | StringLiteral
    //   | NumberLiteral
    //   | ReservedIdentifier
    function parseFieldType() ***REMOVED***
        var key;

        key = parseFieldName();
        if (token === Token.COLON) ***REMOVED***
            consume(Token.COLON);
            return ***REMOVED***
                type: Syntax.FieldType,
                key: key,
                value: parseTypeExpression()
            ***REMOVED***;
        ***REMOVED***
        return ***REMOVED***
            type: Syntax.FieldType,
            key: key,
            value: null
        ***REMOVED***;
    ***REMOVED***

    // RecordType := '***REMOVED***' FieldTypeList '***REMOVED***'
    //
    // FieldTypeList :=
    //     <<empty>>
    //   | FieldType
    //   | FieldType ',' FieldTypeList
    function parseRecordType() ***REMOVED***
        var fields;

        consume(Token.LBRACE, 'RecordType should start with ***REMOVED***');
        fields = [];
        if (token === Token.COMMA) ***REMOVED***
            consume(Token.COMMA);
        ***REMOVED*** else ***REMOVED***
            while (token !== Token.RBRACE) ***REMOVED***
                fields.push(parseFieldType());
                if (token !== Token.RBRACE) ***REMOVED***
                    expect(Token.COMMA);
                ***REMOVED***
            ***REMOVED***
        ***REMOVED***
        expect(Token.RBRACE);
        return ***REMOVED***
            type: Syntax.RecordType,
            fields: fields
        ***REMOVED***;
    ***REMOVED***

    // NameExpression :=
    //    Identifier
    //  | TagIdentifier ':' Identifier
    //
    // Tag identifier is one of "module", "external" or "event"
    // Identifier is the same as Token.NAME, including any dots, something like
    // namespace.module.MyClass
    function parseNameExpression() ***REMOVED***
        var name = value;
        expect(Token.NAME);

        if (token === Token.COLON && (
                name === 'module' ||
                name === 'external' ||
                name === 'event')) ***REMOVED***
            consume(Token.COLON);
            name += ':' + value;
            expect(Token.NAME);
        ***REMOVED***

        return ***REMOVED***
            type: Syntax.NameExpression,
            name: name
        ***REMOVED***;
    ***REMOVED***

    // TypeExpressionList :=
    //     TopLevelTypeExpression
    //   | TopLevelTypeExpression ',' TypeExpressionList
    function parseTypeExpressionList() ***REMOVED***
        var elements = [];

        elements.push(parseTop());
        while (token === Token.COMMA) ***REMOVED***
            consume(Token.COMMA);
            elements.push(parseTop());
        ***REMOVED***
        return elements;
    ***REMOVED***

    // TypeName :=
    //     NameExpression
    //   | NameExpression TypeApplication
    //
    // TypeApplication :=
    //     '.<' TypeExpressionList '>'
    //   | '<' TypeExpressionList '>'   // this is extension of doctrine
    function parseTypeName() ***REMOVED***
        var expr, applications;

        expr = parseNameExpression();
        if (token === Token.DOT_LT || token === Token.LT) ***REMOVED***
            next();
            applications = parseTypeExpressionList();
            expect(Token.GT);
            return ***REMOVED***
                type: Syntax.TypeApplication,
                expression: expr,
                applications: applications
            ***REMOVED***;
        ***REMOVED***
        return expr;
    ***REMOVED***

    // ResultType :=
    //     <<empty>>
    //   | ':' void
    //   | ':' TypeExpression
    //
    // BNF is above
    // but, we remove <<empty>> pattern, so token is always TypeToken::COLON
    function parseResultType() ***REMOVED***
        consume(Token.COLON, 'ResultType should start with :');
        if (token === Token.NAME && value === 'void') ***REMOVED***
            consume(Token.NAME);
            return ***REMOVED***
                type: Syntax.VoidLiteral
            ***REMOVED***;
        ***REMOVED***
        return parseTypeExpression();
    ***REMOVED***

    // ParametersType :=
    //     RestParameterType
    //   | NonRestParametersType
    //   | NonRestParametersType ',' RestParameterType
    //
    // RestParameterType :=
    //     '...'
    //     '...' Identifier
    //
    // NonRestParametersType :=
    //     ParameterType ',' NonRestParametersType
    //   | ParameterType
    //   | OptionalParametersType
    //
    // OptionalParametersType :=
    //     OptionalParameterType
    //   | OptionalParameterType, OptionalParametersType
    //
    // OptionalParameterType := ParameterType=
    //
    // ParameterType := TypeExpression | Identifier ':' TypeExpression
    //
    // Identifier is "new" or "this"
    function parseParametersType() ***REMOVED***
        var params = [], optionalSequence = false, expr, rest = false;

        while (token !== Token.RPAREN) ***REMOVED***
            if (token === Token.REST) ***REMOVED***
                // RestParameterType
                consume(Token.REST);
                rest = true;
            ***REMOVED***

            expr = parseTypeExpression();
            if (expr.type === Syntax.NameExpression && token === Token.COLON) ***REMOVED***
                // Identifier ':' TypeExpression
                consume(Token.COLON);
                expr = ***REMOVED***
                    type: Syntax.ParameterType,
                    name: expr.name,
                    expression: parseTypeExpression()
                ***REMOVED***;
            ***REMOVED***
            if (token === Token.EQUAL) ***REMOVED***
                consume(Token.EQUAL);
                expr = ***REMOVED***
                    type: Syntax.OptionalType,
                    expression: expr
                ***REMOVED***;
                optionalSequence = true;
            ***REMOVED*** else ***REMOVED***
                if (optionalSequence) ***REMOVED***
                    utility.throwError('unexpected token');
                ***REMOVED***
            ***REMOVED***
            if (rest) ***REMOVED***
                expr = ***REMOVED***
                    type: Syntax.RestType,
                    expression: expr
                ***REMOVED***;
            ***REMOVED***
            params.push(expr);
            if (token !== Token.RPAREN) ***REMOVED***
                expect(Token.COMMA);
            ***REMOVED***
        ***REMOVED***
        return params;
    ***REMOVED***

    // FunctionType := 'function' FunctionSignatureType
    //
    // FunctionSignatureType :=
    //   | TypeParameters '(' ')' ResultType
    //   | TypeParameters '(' ParametersType ')' ResultType
    //   | TypeParameters '(' 'this' ':' TypeName ')' ResultType
    //   | TypeParameters '(' 'this' ':' TypeName ',' ParametersType ')' ResultType
    function parseFunctionType() ***REMOVED***
        var isNew, thisBinding, params, result, fnType;
        utility.assert(token === Token.NAME && value === 'function', 'FunctionType should start with \'function\'');
        consume(Token.NAME);

        // Google Closure Compiler is not implementing TypeParameters.
        // So we do not. if we don't get '(', we see it as error.
        expect(Token.LPAREN);

        isNew = false;
        params = [];
        thisBinding = null;
        if (token !== Token.RPAREN) ***REMOVED***
            // ParametersType or 'this'
            if (token === Token.NAME &&
                    (value === 'this' || value === 'new')) ***REMOVED***
                // 'this' or 'new'
                // 'new' is Closure Compiler extension
                isNew = value === 'new';
                consume(Token.NAME);
                expect(Token.COLON);
                thisBinding = parseTypeName();
                if (token === Token.COMMA) ***REMOVED***
                    consume(Token.COMMA);
                    params = parseParametersType();
                ***REMOVED***
            ***REMOVED*** else ***REMOVED***
                params = parseParametersType();
            ***REMOVED***
        ***REMOVED***

        expect(Token.RPAREN);

        result = null;
        if (token === Token.COLON) ***REMOVED***
            result = parseResultType();
        ***REMOVED***

        fnType = ***REMOVED***
            type: Syntax.FunctionType,
            params: params,
            result: result
        ***REMOVED***;
        if (thisBinding) ***REMOVED***
            // avoid adding null 'new' and 'this' properties
            fnType['this'] = thisBinding;
            if (isNew) ***REMOVED***
                fnType['new'] = true;
            ***REMOVED***
        ***REMOVED***
        return fnType;
    ***REMOVED***

    // BasicTypeExpression :=
    //     '*'
    //   | 'null'
    //   | 'undefined'
    //   | TypeName
    //   | FunctionType
    //   | UnionType
    //   | RecordType
    //   | ArrayType
    function parseBasicTypeExpression() ***REMOVED***
        var context;
        switch (token) ***REMOVED***
        case Token.STAR:
            consume(Token.STAR);
            return ***REMOVED***
                type: Syntax.AllLiteral
            ***REMOVED***;

        case Token.LPAREN:
            return parseUnionType();

        case Token.LBRACK:
            return parseArrayType();

        case Token.LBRACE:
            return parseRecordType();

        case Token.NAME:
            if (value === 'null') ***REMOVED***
                consume(Token.NAME);
                return ***REMOVED***
                    type: Syntax.NullLiteral
                ***REMOVED***;
            ***REMOVED***

            if (value === 'undefined') ***REMOVED***
                consume(Token.NAME);
                return ***REMOVED***
                    type: Syntax.UndefinedLiteral
                ***REMOVED***;
            ***REMOVED***

            context = Context.save();
            if (value === 'function') ***REMOVED***
                try ***REMOVED***
                    return parseFunctionType();
                ***REMOVED*** catch (e) ***REMOVED***
                    context.restore();
                ***REMOVED***
            ***REMOVED***

            return parseTypeName();

        default:
            utility.throwError('unexpected token');
        ***REMOVED***
    ***REMOVED***

    // TypeExpression :=
    //     BasicTypeExpression
    //   | '?' BasicTypeExpression
    //   | '!' BasicTypeExpression
    //   | BasicTypeExpression '?'
    //   | BasicTypeExpression '!'
    //   | '?'
    //   | BasicTypeExpression '[]'
    function parseTypeExpression() ***REMOVED***
        var expr;

        if (token === Token.QUESTION) ***REMOVED***
            consume(Token.QUESTION);
            if (token === Token.COMMA || token === Token.EQUAL || token === Token.RBRACE ||
                    token === Token.RPAREN || token === Token.PIPE || token === Token.EOF ||
                    token === Token.RBRACK || token === Token.GT) ***REMOVED***
                return ***REMOVED***
                    type: Syntax.NullableLiteral
                ***REMOVED***;
            ***REMOVED***
            return ***REMOVED***
                type: Syntax.NullableType,
                expression: parseBasicTypeExpression(),
                prefix: true
            ***REMOVED***;
        ***REMOVED***

        if (token === Token.BANG) ***REMOVED***
            consume(Token.BANG);
            return ***REMOVED***
                type: Syntax.NonNullableType,
                expression: parseBasicTypeExpression(),
                prefix: true
            ***REMOVED***;
        ***REMOVED***

        expr = parseBasicTypeExpression();
        if (token === Token.BANG) ***REMOVED***
            consume(Token.BANG);
            return ***REMOVED***
                type: Syntax.NonNullableType,
                expression: expr,
                prefix: false
            ***REMOVED***;
        ***REMOVED***

        if (token === Token.QUESTION) ***REMOVED***
            consume(Token.QUESTION);
            return ***REMOVED***
                type: Syntax.NullableType,
                expression: expr,
                prefix: false
            ***REMOVED***;
        ***REMOVED***

        if (token === Token.LBRACK) ***REMOVED***
            consume(Token.LBRACK);
            expect(Token.RBRACK, 'expected an array-style type declaration (' + value + '[])');
            return ***REMOVED***
                type: Syntax.TypeApplication,
                expression: ***REMOVED***
                    type: Syntax.NameExpression,
                    name: 'Array'
                ***REMOVED***,
                applications: [expr]
            ***REMOVED***;
        ***REMOVED***

        return expr;
    ***REMOVED***

    // TopLevelTypeExpression :=
    //      TypeExpression
    //    | TypeUnionList
    //
    // This rule is Google Closure Compiler extension, not ES4
    // like,
    //   ***REMOVED*** number | string ***REMOVED***
    // If strict to ES4, we should write it as
    //   ***REMOVED*** (number|string) ***REMOVED***
    function parseTop() ***REMOVED***
        var expr, elements;

        expr = parseTypeExpression();
        if (token !== Token.PIPE) ***REMOVED***
            return expr;
        ***REMOVED***

        elements = [ expr ];
        consume(Token.PIPE);
        while (true) ***REMOVED***
            elements.push(parseTypeExpression());
            if (token !== Token.PIPE) ***REMOVED***
                break;
            ***REMOVED***
            consume(Token.PIPE);
        ***REMOVED***

        return ***REMOVED***
            type: Syntax.UnionType,
            elements: elements
        ***REMOVED***;
    ***REMOVED***

    function parseTopParamType() ***REMOVED***
        var expr;

        if (token === Token.REST) ***REMOVED***
            consume(Token.REST);
            return ***REMOVED***
                type: Syntax.RestType,
                expression: parseTop()
            ***REMOVED***;
        ***REMOVED***

        expr = parseTop();
        if (token === Token.EQUAL) ***REMOVED***
            consume(Token.EQUAL);
            return ***REMOVED***
                type: Syntax.OptionalType,
                expression: expr
            ***REMOVED***;
        ***REMOVED***

        return expr;
    ***REMOVED***

    function parseType(src, opt) ***REMOVED***
        var expr;

        source = src;
        length = source.length;
        index = 0;
        previous = 0;

        next();
        expr = parseTop();

        if (opt && opt.midstream) ***REMOVED***
            return ***REMOVED***
                expression: expr,
                index: previous
            ***REMOVED***;
        ***REMOVED***

        if (token !== Token.EOF) ***REMOVED***
            utility.throwError('not reach to EOF');
        ***REMOVED***

        return expr;
    ***REMOVED***

    function parseParamType(src, opt) ***REMOVED***
        var expr;

        source = src;
        length = source.length;
        index = 0;
        previous = 0;

        next();
        expr = parseTopParamType();

        if (opt && opt.midstream) ***REMOVED***
            return ***REMOVED***
                expression: expr,
                index: previous
            ***REMOVED***;
        ***REMOVED***

        if (token !== Token.EOF) ***REMOVED***
            utility.throwError('not reach to EOF');
        ***REMOVED***

        return expr;
    ***REMOVED***

    function stringifyImpl(node, compact, topLevel) ***REMOVED***
        var result, i, iz;

        switch (node.type) ***REMOVED***
        case Syntax.NullableLiteral:
            result = '?';
            break;

        case Syntax.AllLiteral:
            result = '*';
            break;

        case Syntax.NullLiteral:
            result = 'null';
            break;

        case Syntax.UndefinedLiteral:
            result = 'undefined';
            break;

        case Syntax.VoidLiteral:
            result = 'void';
            break;

        case Syntax.UnionType:
            if (!topLevel) ***REMOVED***
                result = '(';
            ***REMOVED*** else ***REMOVED***
                result = '';
            ***REMOVED***

            for (i = 0, iz = node.elements.length; i < iz; ++i) ***REMOVED***
                result += stringifyImpl(node.elements[i], compact);
                if ((i + 1) !== iz) ***REMOVED***
                    result += '|';
                ***REMOVED***
            ***REMOVED***

            if (!topLevel) ***REMOVED***
                result += ')';
            ***REMOVED***
            break;

        case Syntax.ArrayType:
            result = '[';
            for (i = 0, iz = node.elements.length; i < iz; ++i) ***REMOVED***
                result += stringifyImpl(node.elements[i], compact);
                if ((i + 1) !== iz) ***REMOVED***
                    result += compact ? ',' : ', ';
                ***REMOVED***
            ***REMOVED***
            result += ']';
            break;

        case Syntax.RecordType:
            result = '***REMOVED***';
            for (i = 0, iz = node.fields.length; i < iz; ++i) ***REMOVED***
                result += stringifyImpl(node.fields[i], compact);
                if ((i + 1) !== iz) ***REMOVED***
                    result += compact ? ',' : ', ';
                ***REMOVED***
            ***REMOVED***
            result += '***REMOVED***';
            break;

        case Syntax.FieldType:
            if (node.value) ***REMOVED***
                result = node.key + (compact ? ':' : ': ') + stringifyImpl(node.value, compact);
            ***REMOVED*** else ***REMOVED***
                result = node.key;
            ***REMOVED***
            break;

        case Syntax.FunctionType:
            result = compact ? 'function(' : 'function (';

            if (node['this']) ***REMOVED***
                if (node['new']) ***REMOVED***
                    result += (compact ? 'new:' : 'new: ');
                ***REMOVED*** else ***REMOVED***
                    result += (compact ? 'this:' : 'this: ');
                ***REMOVED***

                result += stringifyImpl(node['this'], compact);

                if (node.params.length !== 0) ***REMOVED***
                    result += compact ? ',' : ', ';
                ***REMOVED***
            ***REMOVED***

            for (i = 0, iz = node.params.length; i < iz; ++i) ***REMOVED***
                result += stringifyImpl(node.params[i], compact);
                if ((i + 1) !== iz) ***REMOVED***
                    result += compact ? ',' : ', ';
                ***REMOVED***
            ***REMOVED***

            result += ')';

            if (node.result) ***REMOVED***
                result += (compact ? ':' : ': ') + stringifyImpl(node.result, compact);
            ***REMOVED***
            break;

        case Syntax.ParameterType:
            result = node.name + (compact ? ':' : ': ') + stringifyImpl(node.expression, compact);
            break;

        case Syntax.RestType:
            result = '...';
            if (node.expression) ***REMOVED***
                result += stringifyImpl(node.expression, compact);
            ***REMOVED***
            break;

        case Syntax.NonNullableType:
            if (node.prefix) ***REMOVED***
                result = '!' + stringifyImpl(node.expression, compact);
            ***REMOVED*** else ***REMOVED***
                result = stringifyImpl(node.expression, compact) + '!';
            ***REMOVED***
            break;

        case Syntax.OptionalType:
            result = stringifyImpl(node.expression, compact) + '=';
            break;

        case Syntax.NullableType:
            if (node.prefix) ***REMOVED***
                result = '?' + stringifyImpl(node.expression, compact);
            ***REMOVED*** else ***REMOVED***
                result = stringifyImpl(node.expression, compact) + '?';
            ***REMOVED***
            break;

        case Syntax.NameExpression:
            result = node.name;
            break;

        case Syntax.TypeApplication:
            result = stringifyImpl(node.expression, compact) + '.<';
            for (i = 0, iz = node.applications.length; i < iz; ++i) ***REMOVED***
                result += stringifyImpl(node.applications[i], compact);
                if ((i + 1) !== iz) ***REMOVED***
                    result += compact ? ',' : ', ';
                ***REMOVED***
            ***REMOVED***
            result += '>';
            break;

        default:
            utility.throwError('Unknown type ' + node.type);
        ***REMOVED***

        return result;
    ***REMOVED***

    function stringify(node, options) ***REMOVED***
        if (options == null) ***REMOVED***
            options = ***REMOVED******REMOVED***;
        ***REMOVED***
        return stringifyImpl(node, options.compact, options.topLevel);
    ***REMOVED***

    exports.parseType = parseType;
    exports.parseParamType = parseParamType;
    exports.stringify = stringify;
    exports.Syntax = Syntax;
***REMOVED***());
/* vim: set sw=4 ts=4 et tw=80 : */
