// eslint/lib/cli-engine/config-array/config-array.js
/**
 * @fileoverview `ConfigArray` class.
 *
 * `ConfigArray` class expresses the full of a configuration. It has the entry
 * config file, base config files that were extended, loaded parsers, and loaded
 * plugins.
 *
 * `ConfigArray` class provies three properties and two methods.
 *
 * - `pluginEnvironments`
 * - `pluginProcessors`
 * - `pluginRules`
 *      The `Map` objects that contain the members of all plugins that this
 *      config array contains. Those map objects don't have mutation methods.
 *      Those keys are the member ID such as `pluginId/memberName`.
 * - `isRoot()`
 *      If `true` then this configuration has `root:true` property.
 * - `extractConfig(filePath)`
 *      Extract the final configuration for a given file. This means merging
 *      every config array element which that `criteria` property matched. The
 *      `filePath` argument must be an absolute path.
 *
 * `ConfigArrayFactory` provides the loading logic of config files.
 *
 * @author Toru Nagashima <https://github.com/mysticatea>
 */
"use strict";

//------------------------------------------------------------------------------
// Requirements
//------------------------------------------------------------------------------

const ***REMOVED*** ExtractedConfig ***REMOVED*** = require("./extracted-config");

//------------------------------------------------------------------------------
// Helpers
//------------------------------------------------------------------------------

// Define types for VSCode IntelliSense.
/** @typedef ***REMOVED***import("../../shared/types").Environment***REMOVED*** Environment */
/** @typedef ***REMOVED***import("../../shared/types").GlobalConf***REMOVED*** GlobalConf */
/** @typedef ***REMOVED***import("../../shared/types").RuleConf***REMOVED*** RuleConf */
/** @typedef ***REMOVED***import("../../shared/types").Rule***REMOVED*** Rule */
/** @typedef ***REMOVED***import("../../shared/types").Plugin***REMOVED*** Plugin */
/** @typedef ***REMOVED***import("../../shared/types").Processor***REMOVED*** Processor */
/** @typedef ***REMOVED***import("./config-dependency").DependentParser***REMOVED*** DependentParser */
/** @typedef ***REMOVED***import("./config-dependency").DependentPlugin***REMOVED*** DependentPlugin */
/** @typedef ***REMOVED***import("./override-tester")["OverrideTester"]***REMOVED*** OverrideTester */

/**
 * @typedef ***REMOVED***Object***REMOVED*** ConfigArrayElement
 * @property ***REMOVED***string***REMOVED*** name The name of this config element.
 * @property ***REMOVED***string***REMOVED*** filePath The path to the source file of this config element.
 * @property ***REMOVED***InstanceType<OverrideTester>|null***REMOVED*** criteria The tester for the `files` and `excludedFiles` of this config element.
 * @property ***REMOVED***Record<string, boolean>|undefined***REMOVED*** env The environment settings.
 * @property ***REMOVED***Record<string, GlobalConf>|undefined***REMOVED*** globals The global variable settings.
 * @property ***REMOVED***boolean|undefined***REMOVED*** noInlineConfig The flag that disables directive comments.
 * @property ***REMOVED***DependentParser|undefined***REMOVED*** parser The parser loader.
 * @property ***REMOVED***Object|undefined***REMOVED*** parserOptions The parser options.
 * @property ***REMOVED***Record<string, DependentPlugin>|undefined***REMOVED*** plugins The plugin loaders.
 * @property ***REMOVED***string|undefined***REMOVED*** processor The processor name to refer plugin's processor.
 * @property ***REMOVED***boolean|undefined***REMOVED*** reportUnusedDisableDirectives The flag to report unused `eslint-disable` comments.
 * @property ***REMOVED***boolean|undefined***REMOVED*** root The flag to express root.
 * @property ***REMOVED***Record<string, RuleConf>|undefined***REMOVED*** rules The rule settings
 * @property ***REMOVED***Object|undefined***REMOVED*** settings The shared settings.
 */

/**
 * @typedef ***REMOVED***Object***REMOVED*** ConfigArrayInternalSlots
 * @property ***REMOVED***Map<string, ExtractedConfig>***REMOVED*** cache The cache to extract configs.
 * @property ***REMOVED***ReadonlyMap<string, Environment>|null***REMOVED*** envMap The map from environment ID to environment definition.
 * @property ***REMOVED***ReadonlyMap<string, Processor>|null***REMOVED*** processorMap The map from processor ID to environment definition.
 * @property ***REMOVED***ReadonlyMap<string, Rule>|null***REMOVED*** ruleMap The map from rule ID to rule definition.
 */

/** @type ***REMOVED***WeakMap<ConfigArray, ConfigArrayInternalSlots>***REMOVED*** */
const internalSlotsMap = new class extends WeakMap ***REMOVED***
    get(key) ***REMOVED***
        let value = super.get(key);

        if (!value) ***REMOVED***
            value = ***REMOVED***
                cache: new Map(),
                envMap: null,
                processorMap: null,
                ruleMap: null
            ***REMOVED***;
            super.set(key, value);
        ***REMOVED***

        return value;
    ***REMOVED***
***REMOVED***();

/**
 * Get the indices which are matched to a given file.
 * @param ***REMOVED***ConfigArrayElement[]***REMOVED*** elements The elements.
 * @param ***REMOVED***string***REMOVED*** filePath The path to a target file.
 * @returns ***REMOVED***number[]***REMOVED*** The indices.
 */
function getMatchedIndices(elements, filePath) ***REMOVED***
    const indices = [];

    for (let i = elements.length - 1; i >= 0; --i) ***REMOVED***
        const element = elements[i];

        if (!element.criteria || element.criteria.test(filePath)) ***REMOVED***
            indices.push(i);
        ***REMOVED***
    ***REMOVED***

    return indices;
***REMOVED***

/**
 * Check if a value is a non-null object.
 * @param ***REMOVED***any***REMOVED*** x The value to check.
 * @returns ***REMOVED***boolean***REMOVED*** `true` if the value is a non-null object.
 */
function isNonNullObject(x) ***REMOVED***
    return typeof x === "object" && x !== null;
***REMOVED***

/**
 * Merge two objects.
 *
 * Assign every property values of `y` to `x` if `x` doesn't have the property.
 * If `x`'s property value is an object, it does recursive.
 *
 * @param ***REMOVED***Object***REMOVED*** target The destination to merge
 * @param ***REMOVED***Object|undefined***REMOVED*** source The source to merge.
 * @returns ***REMOVED***void***REMOVED***
 */
function mergeWithoutOverwrite(target, source) ***REMOVED***
    if (!isNonNullObject(source)) ***REMOVED***
        return;
    ***REMOVED***

    for (const key of Object.keys(source)) ***REMOVED***
        if (key === "__proto__") ***REMOVED***
            continue;
        ***REMOVED***

        if (isNonNullObject(target[key])) ***REMOVED***
            mergeWithoutOverwrite(target[key], source[key]);
        ***REMOVED*** else if (target[key] === void 0) ***REMOVED***
            if (isNonNullObject(source[key])) ***REMOVED***
                target[key] = Array.isArray(source[key]) ? [] : ***REMOVED******REMOVED***;
                mergeWithoutOverwrite(target[key], source[key]);
            ***REMOVED*** else if (source[key] !== void 0) ***REMOVED***
                target[key] = source[key];
            ***REMOVED***
        ***REMOVED***
    ***REMOVED***
***REMOVED***

/**
 * Merge plugins.
 * `target`'s definition is prior to `source`'s.
 *
 * @param ***REMOVED***Record<string, DependentPlugin>***REMOVED*** target The destination to merge
 * @param ***REMOVED***Record<string, DependentPlugin>|undefined***REMOVED*** source The source to merge.
 * @returns ***REMOVED***void***REMOVED***
 */
function mergePlugins(target, source) ***REMOVED***
    if (!isNonNullObject(source)) ***REMOVED***
        return;
    ***REMOVED***

    for (const key of Object.keys(source)) ***REMOVED***
        if (key === "__proto__") ***REMOVED***
            continue;
        ***REMOVED***
        const targetValue = target[key];
        const sourceValue = source[key];

        // Adopt the plugin which was found at first.
        if (targetValue === void 0) ***REMOVED***
            if (sourceValue.error) ***REMOVED***
                throw sourceValue.error;
            ***REMOVED***
            target[key] = sourceValue;
        ***REMOVED***
    ***REMOVED***
***REMOVED***

/**
 * Merge rule configs.
 * `target`'s definition is prior to `source`'s.
 *
 * @param ***REMOVED***Record<string, Array>***REMOVED*** target The destination to merge
 * @param ***REMOVED***Record<string, RuleConf>|undefined***REMOVED*** source The source to merge.
 * @returns ***REMOVED***void***REMOVED***
 */
function mergeRuleConfigs(target, source) ***REMOVED***
    if (!isNonNullObject(source)) ***REMOVED***
        return;
    ***REMOVED***

    for (const key of Object.keys(source)) ***REMOVED***
        if (key === "__proto__") ***REMOVED***
            continue;
        ***REMOVED***
        const targetDef = target[key];
        const sourceDef = source[key];

        // Adopt the rule config which was found at first.
        if (targetDef === void 0) ***REMOVED***
            if (Array.isArray(sourceDef)) ***REMOVED***
                target[key] = [...sourceDef];
            ***REMOVED*** else ***REMOVED***
                target[key] = [sourceDef];
            ***REMOVED***

        /*
         * If the first found rule config is severity only and the current rule
         * config has options, merge the severity and the options.
         */
        ***REMOVED*** else if (
            targetDef.length === 1 &&
            Array.isArray(sourceDef) &&
            sourceDef.length >= 2
        ) ***REMOVED***
            targetDef.push(...sourceDef.slice(1));
        ***REMOVED***
    ***REMOVED***
***REMOVED***

/**
 * Create the extracted config.
 * @param ***REMOVED***ConfigArray***REMOVED*** instance The config elements.
 * @param ***REMOVED***number[]***REMOVED*** indices The indices to use.
 * @returns ***REMOVED***ExtractedConfig***REMOVED*** The extracted config.
 */
function createConfig(instance, indices) ***REMOVED***
    const config = new ExtractedConfig();

    // Merge elements.
    for (const index of indices) ***REMOVED***
        const element = instance[index];

        // Adopt the parser which was found at first.
        if (!config.parser && element.parser) ***REMOVED***
            if (element.parser.error) ***REMOVED***
                throw element.parser.error;
            ***REMOVED***
            config.parser = element.parser;
        ***REMOVED***

        // Adopt the processor which was found at first.
        if (!config.processor && element.processor) ***REMOVED***
            config.processor = element.processor;
        ***REMOVED***

        // Adopt the noInlineConfig which was found at first.
        if (config.noInlineConfig === void 0 && element.noInlineConfig !== void 0) ***REMOVED***
            config.noInlineConfig = element.noInlineConfig;
            config.configNameOfNoInlineConfig = element.name;
        ***REMOVED***

        // Adopt the reportUnusedDisableDirectives which was found at first.
        if (config.reportUnusedDisableDirectives === void 0 && element.reportUnusedDisableDirectives !== void 0) ***REMOVED***
            config.reportUnusedDisableDirectives = element.reportUnusedDisableDirectives;
        ***REMOVED***

        // Merge others.
        mergeWithoutOverwrite(config.env, element.env);
        mergeWithoutOverwrite(config.globals, element.globals);
        mergeWithoutOverwrite(config.parserOptions, element.parserOptions);
        mergeWithoutOverwrite(config.settings, element.settings);
        mergePlugins(config.plugins, element.plugins);
        mergeRuleConfigs(config.rules, element.rules);
    ***REMOVED***

    return config;
***REMOVED***

/**
 * Collect definitions.
 * @template T, U
 * @param ***REMOVED***string***REMOVED*** pluginId The plugin ID for prefix.
 * @param ***REMOVED***Record<string,T>***REMOVED*** defs The definitions to collect.
 * @param ***REMOVED***Map<string, U>***REMOVED*** map The map to output.
 * @param ***REMOVED***function(T): U***REMOVED*** [normalize] The normalize function for each value.
 * @returns ***REMOVED***void***REMOVED***
 */
function collect(pluginId, defs, map, normalize) ***REMOVED***
    if (defs) ***REMOVED***
        const prefix = pluginId && `$***REMOVED***pluginId***REMOVED***/`;

        for (const [key, value] of Object.entries(defs)) ***REMOVED***
            map.set(
                `$***REMOVED***prefix***REMOVED***$***REMOVED***key***REMOVED***`,
                normalize ? normalize(value) : value
            );
        ***REMOVED***
    ***REMOVED***
***REMOVED***

/**
 * Normalize a rule definition.
 * @param ***REMOVED***Function|Rule***REMOVED*** rule The rule definition to normalize.
 * @returns ***REMOVED***Rule***REMOVED*** The normalized rule definition.
 */
function normalizePluginRule(rule) ***REMOVED***
    return typeof rule === "function" ? ***REMOVED*** create: rule ***REMOVED*** : rule;
***REMOVED***

/**
 * Delete the mutation methods from a given map.
 * @param ***REMOVED***Map<any, any>***REMOVED*** map The map object to delete.
 * @returns ***REMOVED***void***REMOVED***
 */
function deleteMutationMethods(map) ***REMOVED***
    Object.defineProperties(map, ***REMOVED***
        clear: ***REMOVED*** configurable: true, value: void 0 ***REMOVED***,
        delete: ***REMOVED*** configurable: true, value: void 0 ***REMOVED***,
        set: ***REMOVED*** configurable: true, value: void 0 ***REMOVED***
    ***REMOVED***);
***REMOVED***

/**
 * Create `envMap`, `processorMap`, `ruleMap` with the plugins in the config array.
 * @param ***REMOVED***ConfigArrayElement[]***REMOVED*** elements The config elements.
 * @param ***REMOVED***ConfigArrayInternalSlots***REMOVED*** slots The internal slots.
 * @returns ***REMOVED***void***REMOVED***
 */
function initPluginMemberMaps(elements, slots) ***REMOVED***
    const processed = new Set();

    slots.envMap = new Map();
    slots.processorMap = new Map();
    slots.ruleMap = new Map();

    for (const element of elements) ***REMOVED***
        if (!element.plugins) ***REMOVED***
            continue;
        ***REMOVED***

        for (const [pluginId, value] of Object.entries(element.plugins)) ***REMOVED***
            const plugin = value.definition;

            if (!plugin || processed.has(pluginId)) ***REMOVED***
                continue;
            ***REMOVED***
            processed.add(pluginId);

            collect(pluginId, plugin.environments, slots.envMap);
            collect(pluginId, plugin.processors, slots.processorMap);
            collect(pluginId, plugin.rules, slots.ruleMap, normalizePluginRule);
        ***REMOVED***
    ***REMOVED***

    deleteMutationMethods(slots.envMap);
    deleteMutationMethods(slots.processorMap);
    deleteMutationMethods(slots.ruleMap);
***REMOVED***

/**
 * Create `envMap`, `processorMap`, `ruleMap` with the plugins in the config array.
 * @param ***REMOVED***ConfigArray***REMOVED*** instance The config elements.
 * @returns ***REMOVED***ConfigArrayInternalSlots***REMOVED*** The extracted config.
 */
function ensurePluginMemberMaps(instance) ***REMOVED***
    const slots = internalSlotsMap.get(instance);

    if (!slots.ruleMap) ***REMOVED***
        initPluginMemberMaps(instance, slots);
    ***REMOVED***

    return slots;
***REMOVED***

//------------------------------------------------------------------------------
// Public Interface
//------------------------------------------------------------------------------

/**
 * The Config Array.
 *
 * `ConfigArray` instance contains all settings, parsers, and plugins.
 * You need to call `ConfigArray#extractConfig(filePath)` method in order to
 * extract, merge and get only the config data which is related to an arbitrary
 * file.
 *
 * @extends ***REMOVED***Array<ConfigArrayElement>***REMOVED***
 */
class ConfigArray extends Array ***REMOVED***

    /**
     * Get the plugin environments.
     * The returned map cannot be mutated.
     * @type ***REMOVED***ReadonlyMap<string, Environment>***REMOVED*** The plugin environments.
     */
    get pluginEnvironments() ***REMOVED***
        return ensurePluginMemberMaps(this).envMap;
    ***REMOVED***

    /**
     * Get the plugin processors.
     * The returned map cannot be mutated.
     * @type ***REMOVED***ReadonlyMap<string, Processor>***REMOVED*** The plugin processors.
     */
    get pluginProcessors() ***REMOVED***
        return ensurePluginMemberMaps(this).processorMap;
    ***REMOVED***

    /**
     * Get the plugin rules.
     * The returned map cannot be mutated.
     * @returns ***REMOVED***ReadonlyMap<string, Rule>***REMOVED*** The plugin rules.
     */
    get pluginRules() ***REMOVED***
        return ensurePluginMemberMaps(this).ruleMap;
    ***REMOVED***

    /**
     * Check if this config has `root` flag.
     * @returns ***REMOVED***boolean***REMOVED*** `true` if this config array is root.
     */
    isRoot() ***REMOVED***
        for (let i = this.length - 1; i >= 0; --i) ***REMOVED***
            const root = this[i].root;

            if (typeof root === "boolean") ***REMOVED***
                return root;
            ***REMOVED***
        ***REMOVED***
        return false;
    ***REMOVED***

    /**
     * Extract the config data which is related to a given file.
     * @param ***REMOVED***string***REMOVED*** filePath The absolute path to the target file.
     * @returns ***REMOVED***ExtractedConfig***REMOVED*** The extracted config data.
     */
    extractConfig(filePath) ***REMOVED***
        const ***REMOVED*** cache ***REMOVED*** = internalSlotsMap.get(this);
        const indices = getMatchedIndices(this, filePath);
        const cacheKey = indices.join(",");

        if (!cache.has(cacheKey)) ***REMOVED***
            cache.set(cacheKey, createConfig(this, indices));
        ***REMOVED***

        return cache.get(cacheKey);
    ***REMOVED***
***REMOVED***

const exportObject = ***REMOVED***
    ConfigArray,

    /**
     * Get the used extracted configs.
     * CLIEngine will use this method to collect used deprecated rules.
     * @param ***REMOVED***ConfigArray***REMOVED*** instance The config array object to get.
     * @returns ***REMOVED***ExtractedConfig[]***REMOVED*** The used extracted configs.
     * @private
     */
    getUsedExtractedConfigs(instance) ***REMOVED***
        const ***REMOVED*** cache ***REMOVED*** = internalSlotsMap.get(instance);

        return Array.from(cache.values());
    ***REMOVED***
***REMOVED***;

module.exports = exportObject;
