// eslint/lib/cli-engine/config-array/config-dependency.js
/**
 * @fileoverview `ConfigDependency` class.
 *
 * `ConfigDependency` class expresses a loaded parser or plugin.
 *
 * If the parser or plugin was loaded successfully, it has `definition` property
 * and `filePath` property. Otherwise, it has `error` property.
 *
 * When `JSON.stringify()` converted a `ConfigDependency` object to a JSON, it
 * omits `definition` property.
 *
 * `ConfigArrayFactory` creates `ConfigDependency` objects when it loads parsers
 * or plugins.
 *
 * @author Toru Nagashima <https://github.com/mysticatea>
 */
"use strict";

const util = require("util");

/**
 * The class is to store parsers or plugins.
 * This class hides the loaded object from `JSON.stringify()` and `console.log`.
 * @template T
 */
class ConfigDependency ***REMOVED***

    /**
     * Initialize this instance.
     * @param ***REMOVED***Object***REMOVED*** data The dependency data.
     * @param ***REMOVED***T***REMOVED*** [data.definition] The dependency if the loading succeeded.
     * @param ***REMOVED***Error***REMOVED*** [data.error] The error object if the loading failed.
     * @param ***REMOVED***string***REMOVED*** [data.filePath] The actual path to the dependency if the loading succeeded.
     * @param ***REMOVED***string***REMOVED*** data.id The ID of this dependency.
     * @param ***REMOVED***string***REMOVED*** data.importerName The name of the config file which loads this dependency.
     * @param ***REMOVED***string***REMOVED*** data.importerPath The path to the config file which loads this dependency.
     */
    constructor(***REMOVED***
        definition = null,
        error = null,
        filePath = null,
        id,
        importerName,
        importerPath
    ***REMOVED***) ***REMOVED***

        /**
         * The loaded dependency if the loading succeeded.
         * @type ***REMOVED***T|null***REMOVED***
         */
        this.definition = definition;

        /**
         * The error object if the loading failed.
         * @type ***REMOVED***Error|null***REMOVED***
         */
        this.error = error;

        /**
         * The loaded dependency if the loading succeeded.
         * @type ***REMOVED***string|null***REMOVED***
         */
        this.filePath = filePath;

        /**
         * The ID of this dependency.
         * @type ***REMOVED***string***REMOVED***
         */
        this.id = id;

        /**
         * The name of the config file which loads this dependency.
         * @type ***REMOVED***string***REMOVED***
         */
        this.importerName = importerName;

        /**
         * The path to the config file which loads this dependency.
         * @type ***REMOVED***string***REMOVED***
         */
        this.importerPath = importerPath;
    ***REMOVED***

    /**
     * @returns ***REMOVED***Object***REMOVED*** a JSON compatible object.
     */
    toJSON() ***REMOVED***
        const obj = this[util.inspect.custom]();

        // Display `error.message` (`Error#message` is unenumerable).
        if (obj.error instanceof Error) ***REMOVED***
            obj.error = ***REMOVED*** ...obj.error, message: obj.error.message ***REMOVED***;
        ***REMOVED***

        return obj;
    ***REMOVED***

    /**
     * @returns ***REMOVED***Object***REMOVED*** an object to display by `console.log()`.
     */
    [util.inspect.custom]() ***REMOVED***
        const ***REMOVED***
            definition: _ignore, // eslint-disable-line no-unused-vars
            ...obj
        ***REMOVED*** = this;

        return obj;
    ***REMOVED***
***REMOVED***

/** @typedef ***REMOVED***ConfigDependency<import("../../shared/types").Parser>***REMOVED*** DependentParser */
/** @typedef ***REMOVED***ConfigDependency<import("../../shared/types").Plugin>***REMOVED*** DependentPlugin */

module.exports = ***REMOVED*** ConfigDependency ***REMOVED***;
