// standard-engine/index.js
module.exports.cli = require('./bin/cmd')

module.exports.linter = Linter

var deglob = require('deglob')
var os = require('os')
var path = require('path')
var pkgConf = require('pkg-conf')

var HOME_OR_TMP = os.homedir() || os.tmpdir()

var DEFAULT_PATTERNS = [
  '**/*.js',
  '**/*.jsx',
  '**/*.mjs',
  '**/*.cjs'
]

var DEFAULT_IGNORE = [
  '**/*.min.js',
  'coverage/**',
  'node_modules/**',
  'vendor/**'
]

function Linter (opts) ***REMOVED***
  if (!(this instanceof Linter)) return new Linter(opts)
  if (!opts) opts = ***REMOVED******REMOVED***

  if (!opts.cmd) throw new Error('opts.cmd option is required')
  if (!opts.eslint) throw new Error('opts.eslint option is required')

  this.cmd = opts.cmd
  this.eslint = opts.eslint
  this.cwd = opts.cwd || process.cwd()
  this.customParseOpts = opts.parseOpts

  var m = opts.version && opts.version.match(/^(\d+)\./)
  var majorVersion = (m && m[1]) || '0'

  // Example cache location: .standard-v12-cache/
  var cacheLocation = path.join(HOME_OR_TMP, `.$***REMOVED***this.cmd***REMOVED***-v$***REMOVED***majorVersion***REMOVED***-cache/`)

  this.eslintConfig = Object.assign(***REMOVED***
    cache: true,
    cacheLocation: cacheLocation,
    envs: [],
    fix: false,
    globals: [],
    ignore: false,
    plugins: [],
    useEslintrc: false
  ***REMOVED***, opts.eslintConfig)

  if (this.eslintConfig.configFile != null) ***REMOVED***
    this.eslintConfig.resolvePluginsRelativeTo =
      path.dirname(this.eslintConfig.configFile)
  ***REMOVED***
***REMOVED***

/**
 * Lint text to enforce JavaScript Style.
 *
 * @param ***REMOVED***string***REMOVED*** text                   file text to lint
 * @param ***REMOVED***Object=***REMOVED*** opts                  options object
 * @param ***REMOVED***boolean=***REMOVED*** opts.fix             automatically fix problems
 * @param ***REMOVED***Array.<string>=***REMOVED*** opts.globals  custom global variables to declare
 * @param ***REMOVED***Array.<string>=***REMOVED*** opts.plugins  custom eslint plugins
 * @param ***REMOVED***Array.<string>=***REMOVED*** opts.envs     custom eslint environment
 * @param ***REMOVED***string=***REMOVED*** opts.parser           custom js parser (e.g. babel-eslint)
 * @param ***REMOVED***string=***REMOVED*** opts.filename         path of file containing the text being linted
 * @param ***REMOVED***boolean=***REMOVED*** opts.usePackageJson  use options from nearest package.json? (default: true)
 */
Linter.prototype.lintTextSync = function (text, opts) ***REMOVED***
  opts = this.parseOpts(opts)
  return new this.eslint.CLIEngine(opts.eslintConfig).executeOnText(text, opts.filename)
***REMOVED***

Linter.prototype.lintText = function (text, opts, cb) ***REMOVED***
  if (typeof opts === 'function') return this.lintText(text, null, opts)
  var result
  try ***REMOVED***
    result = this.lintTextSync(text, opts)
  ***REMOVED*** catch (err) ***REMOVED***
    return process.nextTick(cb, err)
  ***REMOVED***
  process.nextTick(cb, null, result)
***REMOVED***

/**
 * Lint files to enforce JavaScript Style.
 *
 * @param ***REMOVED***Array.<string>***REMOVED*** files          file globs to lint
 * @param ***REMOVED***Object=***REMOVED*** opts                  options object
 * @param ***REMOVED***Array.<string>=***REMOVED*** opts.ignore   file globs to ignore (has sane defaults)
 * @param ***REMOVED***string=***REMOVED*** opts.cwd              current working directory (default: process.cwd())
 * @param ***REMOVED***boolean=***REMOVED*** opts.fix             automatically fix problems
 * @param ***REMOVED***Array.<string>=***REMOVED*** opts.globals  custom global variables to declare
 * @param ***REMOVED***Array.<string>=***REMOVED*** opts.plugins  custom eslint plugins
 * @param ***REMOVED***Array.<string>=***REMOVED*** opts.envs     custom eslint environment
 * @param ***REMOVED***string=***REMOVED*** opts.parser           custom js parser (e.g. babel-eslint)
 * @param ***REMOVED***boolean=***REMOVED*** opts.usePackageJson  use options from nearest package.json? (default: true)
 * @param ***REMOVED***function(Error, Object)***REMOVED*** cb    callback
 */
Linter.prototype.lintFiles = function (files, opts, cb) ***REMOVED***
  var self = this
  if (typeof opts === 'function') return self.lintFiles(files, null, opts)
  opts = self.parseOpts(opts)

  if (typeof files === 'string') files = [files]
  if (files.length === 0) files = DEFAULT_PATTERNS

  var deglobOpts = ***REMOVED***
    ignore: opts.ignore,
    cwd: opts.cwd,
    useGitIgnore: true,
    usePackageJson: false
  ***REMOVED***

  deglob(files, deglobOpts, function (err, allFiles) ***REMOVED***
    if (err) return cb(err)

    var result
    try ***REMOVED***
      result = new self.eslint.CLIEngine(opts.eslintConfig).executeOnFiles(allFiles)
    ***REMOVED*** catch (err) ***REMOVED***
      return cb(err)
    ***REMOVED***

    if (opts.fix) ***REMOVED***
      self.eslint.CLIEngine.outputFixes(result)
    ***REMOVED***

    return cb(null, result)
  ***REMOVED***)
***REMOVED***

Linter.prototype.parseOpts = function (opts) ***REMOVED***
  var self = this

  if (!opts) opts = ***REMOVED******REMOVED***
  opts = Object.assign(***REMOVED******REMOVED***, opts)
  opts.eslintConfig = Object.assign(***REMOVED******REMOVED***, self.eslintConfig)
  opts.eslintConfig.fix = !!opts.fix

  if (!opts.cwd) opts.cwd = self.cwd
  opts.eslintConfig.cwd = opts.cwd

  // If no usePackageJson option is given, default to `true`
  var usePackageJson = opts.usePackageJson != null
    ? opts.usePackageJson
    : true

  var packageOpts = usePackageJson
    ? pkgConf.sync(self.cmd, ***REMOVED*** cwd: opts.cwd ***REMOVED***)
    : ***REMOVED******REMOVED***

  if (!opts.ignore) opts.ignore = []
  addIgnore(packageOpts.ignore)
  if (!packageOpts.noDefaultIgnore) ***REMOVED***
    addIgnore(DEFAULT_IGNORE)
  ***REMOVED***

  addGlobals(packageOpts.globals || packageOpts.global)
  addGlobals(opts.globals || opts.global)

  addPlugins(packageOpts.plugins || packageOpts.plugin)
  addPlugins(opts.plugins || opts.plugin)

  addEnvs(packageOpts.envs || packageOpts.env)
  addEnvs(opts.envs || opts.env)

  setParser(packageOpts.parser || opts.parser)

  if (self.customParseOpts) ***REMOVED***
    var rootDir
    if (usePackageJson) ***REMOVED***
      var filePath = pkgConf.filepath(packageOpts)
      rootDir = filePath ? path.dirname(filePath) : opts.cwd
    ***REMOVED*** else ***REMOVED***
      rootDir = opts.cwd
    ***REMOVED***
    opts = self.customParseOpts(opts, packageOpts, rootDir)
  ***REMOVED***

  function addIgnore (ignore) ***REMOVED***
    if (!ignore) return
    opts.ignore = opts.ignore.concat(ignore)
  ***REMOVED***

  function addGlobals (globals) ***REMOVED***
    if (!globals) return
    opts.eslintConfig.globals = self.eslintConfig.globals.concat(globals)
  ***REMOVED***

  function addPlugins (plugins) ***REMOVED***
    if (!plugins) return
    opts.eslintConfig.plugins = self.eslintConfig.plugins.concat(plugins)
  ***REMOVED***

  function addEnvs (envs) ***REMOVED***
    if (!envs) return
    if (!Array.isArray(envs) && typeof envs !== 'string') ***REMOVED***
      // envs can be an object in `package.json`
      envs = Object.keys(envs).filter(function (env) ***REMOVED*** return envs[env] ***REMOVED***)
    ***REMOVED***
    opts.eslintConfig.envs = self.eslintConfig.envs.concat(envs)
  ***REMOVED***

  function setParser (parser) ***REMOVED***
    if (!parser) return
    opts.eslintConfig.parser = parser
  ***REMOVED***

  return opts
***REMOVED***
