// eslint/lib/options.js
/**
 * @fileoverview Options configuration for optionator.
 * @author George Zahariev
 */

"use strict";

//------------------------------------------------------------------------------
// Requirements
//------------------------------------------------------------------------------

const optionator = require("optionator");

//------------------------------------------------------------------------------
// Initialization and Public Interface
//------------------------------------------------------------------------------

// exports "parse(args)", "generateHelp()", and "generateHelpForOption(optionName)"
module.exports = optionator(***REMOVED***
    prepend: "eslint [options] file.js [file.js] [dir]",
    defaults: ***REMOVED***
        concatRepeatedArrays: true,
        mergeRepeatedObjects: true
    ***REMOVED***,
    options: [
        ***REMOVED***
            heading: "Basic configuration"
        ***REMOVED***,
        ***REMOVED***
            option: "eslintrc",
            type: "Boolean",
            default: "true",
            description: "Disable use of configuration from .eslintrc.*"
        ***REMOVED***,
        ***REMOVED***
            option: "config",
            alias: "c",
            type: "path::String",
            description: "Use this configuration, overriding .eslintrc.* config options if present"
        ***REMOVED***,
        ***REMOVED***
            option: "env",
            type: "[String]",
            description: "Specify environments"
        ***REMOVED***,
        ***REMOVED***
            option: "ext",
            type: "[String]",
            default: ".js",
            description: "Specify JavaScript file extensions"
        ***REMOVED***,
        ***REMOVED***
            option: "global",
            type: "[String]",
            description: "Define global variables"
        ***REMOVED***,
        ***REMOVED***
            option: "parser",
            type: "String",
            description: "Specify the parser to be used"
        ***REMOVED***,
        ***REMOVED***
            option: "parser-options",
            type: "Object",
            description: "Specify parser options"
        ***REMOVED***,
        ***REMOVED***
            option: "resolve-plugins-relative-to",
            type: "path::String",
            description: "A folder where plugins should be resolved from, CWD by default"
        ***REMOVED***,
        ***REMOVED***
            heading: "Specifying rules and plugins"
        ***REMOVED***,
        ***REMOVED***
            option: "rulesdir",
            type: "[path::String]",
            description: "Use additional rules from this directory"
        ***REMOVED***,
        ***REMOVED***
            option: "plugin",
            type: "[String]",
            description: "Specify plugins"
        ***REMOVED***,
        ***REMOVED***
            option: "rule",
            type: "Object",
            description: "Specify rules"
        ***REMOVED***,
        ***REMOVED***
            heading: "Fixing problems"
        ***REMOVED***,
        ***REMOVED***
            option: "fix",
            type: "Boolean",
            default: false,
            description: "Automatically fix problems"
        ***REMOVED***,
        ***REMOVED***
            option: "fix-dry-run",
            type: "Boolean",
            default: false,
            description: "Automatically fix problems without saving the changes to the file system"
        ***REMOVED***,
        ***REMOVED***
            option: "fix-type",
            type: "Array",
            description: "Specify the types of fixes to apply (problem, suggestion, layout)"
        ***REMOVED***,
        ***REMOVED***
            heading: "Ignoring files"
        ***REMOVED***,
        ***REMOVED***
            option: "ignore-path",
            type: "path::String",
            description: "Specify path of ignore file"
        ***REMOVED***,
        ***REMOVED***
            option: "ignore",
            type: "Boolean",
            default: "true",
            description: "Disable use of ignore files and patterns"
        ***REMOVED***,
        ***REMOVED***
            option: "ignore-pattern",
            type: "[String]",
            description: "Pattern of files to ignore (in addition to those in .eslintignore)",
            concatRepeatedArrays: [true, ***REMOVED***
                oneValuePerFlag: true
            ***REMOVED***]
        ***REMOVED***,
        ***REMOVED***
            heading: "Using stdin"
        ***REMOVED***,
        ***REMOVED***
            option: "stdin",
            type: "Boolean",
            default: "false",
            description: "Lint code provided on <STDIN>"
        ***REMOVED***,
        ***REMOVED***
            option: "stdin-filename",
            type: "String",
            description: "Specify filename to process STDIN as"
        ***REMOVED***,
        ***REMOVED***
            heading: "Handling warnings"
        ***REMOVED***,
        ***REMOVED***
            option: "quiet",
            type: "Boolean",
            default: "false",
            description: "Report errors only"
        ***REMOVED***,
        ***REMOVED***
            option: "max-warnings",
            type: "Int",
            default: "-1",
            description: "Number of warnings to trigger nonzero exit code"
        ***REMOVED***,
        ***REMOVED***
            heading: "Output"
        ***REMOVED***,
        ***REMOVED***
            option: "output-file",
            alias: "o",
            type: "path::String",
            description: "Specify file to write report to"
        ***REMOVED***,
        ***REMOVED***
            option: "format",
            alias: "f",
            type: "String",
            default: "stylish",
            description: "Use a specific output format"
        ***REMOVED***,
        ***REMOVED***
            option: "color",
            type: "Boolean",
            alias: "no-color",
            description: "Force enabling/disabling of color"
        ***REMOVED***,
        ***REMOVED***
            heading: "Inline configuration comments"
        ***REMOVED***,
        ***REMOVED***
            option: "inline-config",
            type: "Boolean",
            default: "true",
            description: "Prevent comments from changing config or rules"
        ***REMOVED***,
        ***REMOVED***
            option: "report-unused-disable-directives",
            type: "Boolean",
            default: void 0,
            description: "Adds reported errors for unused eslint-disable directives"
        ***REMOVED***,
        ***REMOVED***
            heading: "Caching"
        ***REMOVED***,
        ***REMOVED***
            option: "cache",
            type: "Boolean",
            default: "false",
            description: "Only check changed files"
        ***REMOVED***,
        ***REMOVED***
            option: "cache-file",
            type: "path::String",
            default: ".eslintcache",
            description: "Path to the cache file. Deprecated: use --cache-location"
        ***REMOVED***,
        ***REMOVED***
            option: "cache-location",
            type: "path::String",
            description: "Path to the cache file or directory"
        ***REMOVED***,
        ***REMOVED***
            heading: "Miscellaneous"
        ***REMOVED***,
        ***REMOVED***
            option: "init",
            type: "Boolean",
            default: "false",
            description: "Run config initialization wizard"
        ***REMOVED***,
        ***REMOVED***
            option: "debug",
            type: "Boolean",
            default: false,
            description: "Output debugging information"
        ***REMOVED***,
        ***REMOVED***
            option: "help",
            alias: "h",
            type: "Boolean",
            description: "Show help"
        ***REMOVED***,
        ***REMOVED***
            option: "version",
            alias: "v",
            type: "Boolean",
            description: "Output the version number"
        ***REMOVED***,
        ***REMOVED***
            option: "print-config",
            type: "path::String",
            description: "Print the configuration for the given file"
        ***REMOVED***
    ]
***REMOVED***);
