// rxjs/internal/observable/combineLatest.js
"use strict";
var __extends = (this && this.__extends) || (function () ***REMOVED***
    var extendStatics = function (d, b) ***REMOVED***
        extendStatics = Object.setPrototypeOf ||
            (***REMOVED*** __proto__: [] ***REMOVED*** instanceof Array && function (d, b) ***REMOVED*** d.__proto__ = b; ***REMOVED***) ||
            function (d, b) ***REMOVED*** for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; ***REMOVED***;
        return extendStatics(d, b);
    ***REMOVED***
    return function (d, b) ***REMOVED***
        extendStatics(d, b);
        function __() ***REMOVED*** this.constructor = d; ***REMOVED***
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    ***REMOVED***;
***REMOVED***)();
Object.defineProperty(exports, "__esModule", ***REMOVED*** value: true ***REMOVED***);
var isScheduler_1 = require("../util/isScheduler");
var isArray_1 = require("../util/isArray");
var OuterSubscriber_1 = require("../OuterSubscriber");
var subscribeToResult_1 = require("../util/subscribeToResult");
var fromArray_1 = require("./fromArray");
var NONE = ***REMOVED******REMOVED***;
function combineLatest() ***REMOVED***
    var observables = [];
    for (var _i = 0; _i < arguments.length; _i++) ***REMOVED***
        observables[_i] = arguments[_i];
    ***REMOVED***
    var resultSelector = null;
    var scheduler = null;
    if (isScheduler_1.isScheduler(observables[observables.length - 1])) ***REMOVED***
        scheduler = observables.pop();
    ***REMOVED***
    if (typeof observables[observables.length - 1] === 'function') ***REMOVED***
        resultSelector = observables.pop();
    ***REMOVED***
    if (observables.length === 1 && isArray_1.isArray(observables[0])) ***REMOVED***
        observables = observables[0];
    ***REMOVED***
    return fromArray_1.fromArray(observables, scheduler).lift(new CombineLatestOperator(resultSelector));
***REMOVED***
exports.combineLatest = combineLatest;
var CombineLatestOperator = (function () ***REMOVED***
    function CombineLatestOperator(resultSelector) ***REMOVED***
        this.resultSelector = resultSelector;
    ***REMOVED***
    CombineLatestOperator.prototype.call = function (subscriber, source) ***REMOVED***
        return source.subscribe(new CombineLatestSubscriber(subscriber, this.resultSelector));
    ***REMOVED***;
    return CombineLatestOperator;
***REMOVED***());
exports.CombineLatestOperator = CombineLatestOperator;
var CombineLatestSubscriber = (function (_super) ***REMOVED***
    __extends(CombineLatestSubscriber, _super);
    function CombineLatestSubscriber(destination, resultSelector) ***REMOVED***
        var _this = _super.call(this, destination) || this;
        _this.resultSelector = resultSelector;
        _this.active = 0;
        _this.values = [];
        _this.observables = [];
        return _this;
    ***REMOVED***
    CombineLatestSubscriber.prototype._next = function (observable) ***REMOVED***
        this.values.push(NONE);
        this.observables.push(observable);
    ***REMOVED***;
    CombineLatestSubscriber.prototype._complete = function () ***REMOVED***
        var observables = this.observables;
        var len = observables.length;
        if (len === 0) ***REMOVED***
            this.destination.complete();
        ***REMOVED***
        else ***REMOVED***
            this.active = len;
            this.toRespond = len;
            for (var i = 0; i < len; i++) ***REMOVED***
                var observable = observables[i];
                this.add(subscribeToResult_1.subscribeToResult(this, observable, observable, i));
            ***REMOVED***
        ***REMOVED***
    ***REMOVED***;
    CombineLatestSubscriber.prototype.notifyComplete = function (unused) ***REMOVED***
        if ((this.active -= 1) === 0) ***REMOVED***
            this.destination.complete();
        ***REMOVED***
    ***REMOVED***;
    CombineLatestSubscriber.prototype.notifyNext = function (outerValue, innerValue, outerIndex, innerIndex, innerSub) ***REMOVED***
        var values = this.values;
        var oldVal = values[outerIndex];
        var toRespond = !this.toRespond
            ? 0
            : oldVal === NONE ? --this.toRespond : this.toRespond;
        values[outerIndex] = innerValue;
        if (toRespond === 0) ***REMOVED***
            if (this.resultSelector) ***REMOVED***
                this._tryResultSelector(values);
            ***REMOVED***
            else ***REMOVED***
                this.destination.next(values.slice());
            ***REMOVED***
        ***REMOVED***
    ***REMOVED***;
    CombineLatestSubscriber.prototype._tryResultSelector = function (values) ***REMOVED***
        var result;
        try ***REMOVED***
            result = this.resultSelector.apply(this, values);
        ***REMOVED***
        catch (err) ***REMOVED***
            this.destination.error(err);
            return;
        ***REMOVED***
        this.destination.next(result);
    ***REMOVED***;
    return CombineLatestSubscriber;
***REMOVED***(OuterSubscriber_1.OuterSubscriber));
exports.CombineLatestSubscriber = CombineLatestSubscriber;
