// rxjs/internal/observable/fromEvent.js
"use strict";
Object.defineProperty(exports, "__esModule", ***REMOVED*** value: true ***REMOVED***);
var Observable_1 = require("../Observable");
var isArray_1 = require("../util/isArray");
var isFunction_1 = require("../util/isFunction");
var map_1 = require("../operators/map");
var toString = (function () ***REMOVED*** return Object.prototype.toString; ***REMOVED***)();
function fromEvent(target, eventName, options, resultSelector) ***REMOVED***
    if (isFunction_1.isFunction(options)) ***REMOVED***
        resultSelector = options;
        options = undefined;
    ***REMOVED***
    if (resultSelector) ***REMOVED***
        return fromEvent(target, eventName, options).pipe(map_1.map(function (args) ***REMOVED*** return isArray_1.isArray(args) ? resultSelector.apply(void 0, args) : resultSelector(args); ***REMOVED***));
    ***REMOVED***
    return new Observable_1.Observable(function (subscriber) ***REMOVED***
        function handler(e) ***REMOVED***
            if (arguments.length > 1) ***REMOVED***
                subscriber.next(Array.prototype.slice.call(arguments));
            ***REMOVED***
            else ***REMOVED***
                subscriber.next(e);
            ***REMOVED***
        ***REMOVED***
        setupSubscription(target, eventName, handler, subscriber, options);
    ***REMOVED***);
***REMOVED***
exports.fromEvent = fromEvent;
function setupSubscription(sourceObj, eventName, handler, subscriber, options) ***REMOVED***
    var unsubscribe;
    if (isEventTarget(sourceObj)) ***REMOVED***
        var source_1 = sourceObj;
        sourceObj.addEventListener(eventName, handler, options);
        unsubscribe = function () ***REMOVED*** return source_1.removeEventListener(eventName, handler, options); ***REMOVED***;
    ***REMOVED***
    else if (isJQueryStyleEventEmitter(sourceObj)) ***REMOVED***
        var source_2 = sourceObj;
        sourceObj.on(eventName, handler);
        unsubscribe = function () ***REMOVED*** return source_2.off(eventName, handler); ***REMOVED***;
    ***REMOVED***
    else if (isNodeStyleEventEmitter(sourceObj)) ***REMOVED***
        var source_3 = sourceObj;
        sourceObj.addListener(eventName, handler);
        unsubscribe = function () ***REMOVED*** return source_3.removeListener(eventName, handler); ***REMOVED***;
    ***REMOVED***
    else if (sourceObj && sourceObj.length) ***REMOVED***
        for (var i = 0, len = sourceObj.length; i < len; i++) ***REMOVED***
            setupSubscription(sourceObj[i], eventName, handler, subscriber, options);
        ***REMOVED***
    ***REMOVED***
    else ***REMOVED***
        throw new TypeError('Invalid event target');
    ***REMOVED***
    subscriber.add(unsubscribe);
***REMOVED***
function isNodeStyleEventEmitter(sourceObj) ***REMOVED***
    return sourceObj && typeof sourceObj.addListener === 'function' && typeof sourceObj.removeListener === 'function';
***REMOVED***
function isJQueryStyleEventEmitter(sourceObj) ***REMOVED***
    return sourceObj && typeof sourceObj.on === 'function' && typeof sourceObj.off === 'function';
***REMOVED***
function isEventTarget(sourceObj) ***REMOVED***
    return sourceObj && typeof sourceObj.addEventListener === 'function' && typeof sourceObj.removeEventListener === 'function';
***REMOVED***
