// rxjs/internal/observable/race.js
"use strict";
var __extends = (this && this.__extends) || (function () ***REMOVED***
    var extendStatics = function (d, b) ***REMOVED***
        extendStatics = Object.setPrototypeOf ||
            (***REMOVED*** __proto__: [] ***REMOVED*** instanceof Array && function (d, b) ***REMOVED*** d.__proto__ = b; ***REMOVED***) ||
            function (d, b) ***REMOVED*** for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; ***REMOVED***;
        return extendStatics(d, b);
    ***REMOVED***
    return function (d, b) ***REMOVED***
        extendStatics(d, b);
        function __() ***REMOVED*** this.constructor = d; ***REMOVED***
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    ***REMOVED***;
***REMOVED***)();
Object.defineProperty(exports, "__esModule", ***REMOVED*** value: true ***REMOVED***);
var isArray_1 = require("../util/isArray");
var fromArray_1 = require("./fromArray");
var OuterSubscriber_1 = require("../OuterSubscriber");
var subscribeToResult_1 = require("../util/subscribeToResult");
function race() ***REMOVED***
    var observables = [];
    for (var _i = 0; _i < arguments.length; _i++) ***REMOVED***
        observables[_i] = arguments[_i];
    ***REMOVED***
    if (observables.length === 1) ***REMOVED***
        if (isArray_1.isArray(observables[0])) ***REMOVED***
            observables = observables[0];
        ***REMOVED***
        else ***REMOVED***
            return observables[0];
        ***REMOVED***
    ***REMOVED***
    return fromArray_1.fromArray(observables, undefined).lift(new RaceOperator());
***REMOVED***
exports.race = race;
var RaceOperator = (function () ***REMOVED***
    function RaceOperator() ***REMOVED***
    ***REMOVED***
    RaceOperator.prototype.call = function (subscriber, source) ***REMOVED***
        return source.subscribe(new RaceSubscriber(subscriber));
    ***REMOVED***;
    return RaceOperator;
***REMOVED***());
exports.RaceOperator = RaceOperator;
var RaceSubscriber = (function (_super) ***REMOVED***
    __extends(RaceSubscriber, _super);
    function RaceSubscriber(destination) ***REMOVED***
        var _this = _super.call(this, destination) || this;
        _this.hasFirst = false;
        _this.observables = [];
        _this.subscriptions = [];
        return _this;
    ***REMOVED***
    RaceSubscriber.prototype._next = function (observable) ***REMOVED***
        this.observables.push(observable);
    ***REMOVED***;
    RaceSubscriber.prototype._complete = function () ***REMOVED***
        var observables = this.observables;
        var len = observables.length;
        if (len === 0) ***REMOVED***
            this.destination.complete();
        ***REMOVED***
        else ***REMOVED***
            for (var i = 0; i < len && !this.hasFirst; i++) ***REMOVED***
                var observable = observables[i];
                var subscription = subscribeToResult_1.subscribeToResult(this, observable, observable, i);
                if (this.subscriptions) ***REMOVED***
                    this.subscriptions.push(subscription);
                ***REMOVED***
                this.add(subscription);
            ***REMOVED***
            this.observables = null;
        ***REMOVED***
    ***REMOVED***;
    RaceSubscriber.prototype.notifyNext = function (outerValue, innerValue, outerIndex, innerIndex, innerSub) ***REMOVED***
        if (!this.hasFirst) ***REMOVED***
            this.hasFirst = true;
            for (var i = 0; i < this.subscriptions.length; i++) ***REMOVED***
                if (i !== outerIndex) ***REMOVED***
                    var subscription = this.subscriptions[i];
                    subscription.unsubscribe();
                    this.remove(subscription);
                ***REMOVED***
            ***REMOVED***
            this.subscriptions = null;
        ***REMOVED***
        this.destination.next(innerValue);
    ***REMOVED***;
    return RaceSubscriber;
***REMOVED***(OuterSubscriber_1.OuterSubscriber));
exports.RaceSubscriber = RaceSubscriber;
