// eslint-plugin-node/lib/util/get-allow-modules.js
/**
 * @author Toru Nagashima
 * See LICENSE file in root directory for full license.
 */
"use strict"

const DEFAULT_VALUE = Object.freeze([])

/**
 * Gets `allowModules` property from a given option object.
 *
 * @param ***REMOVED***object|undefined***REMOVED*** option - An option object to get.
 * @returns ***REMOVED***string[]|null***REMOVED*** The `allowModules` value, or `null`.
 */
function get(option) ***REMOVED***
    if (option && option.allowModules && Array.isArray(option.allowModules)) ***REMOVED***
        return option.allowModules.map(String)
    ***REMOVED***
    return null
***REMOVED***

/**
 * Gets "allowModules" setting.
 *
 * 1. This checks `options` property, then returns it if exists.
 * 2. This checks `settings.node` property, then returns it if exists.
 * 3. This returns `[]`.
 *
 * @param ***REMOVED***RuleContext***REMOVED*** context - The rule context.
 * @returns ***REMOVED***string[]***REMOVED*** A list of extensions.
 */
module.exports = function getAllowModules(context) ***REMOVED***
    return (
        get(context.options && context.options[0]) ||
        get(context.settings && context.settings.node) ||
        DEFAULT_VALUE
    )
***REMOVED***

module.exports.schema = ***REMOVED***
    type: "array",
    items: ***REMOVED***
        type: "string",
        pattern: "^(?:@[a-zA-Z0-9_\\-.]+/)?[a-zA-Z0-9_\\-.]+$",
    ***REMOVED***,
    uniqueItems: true,
***REMOVED***
