// eslint/lib/cli-engine/cascading-config-array-factory.js
/**
 * @fileoverview `CascadingConfigArrayFactory` class.
 *
 * `CascadingConfigArrayFactory` class has a responsibility:
 *
 * 1. Handles cascading of config files.
 *
 * It provides two methods:
 *
 * - `getConfigArrayForFile(filePath)`
 *     Get the corresponded configuration of a given file. This method doesn't
 *     throw even if the given file didn't exist.
 * - `clearCache()`
 *     Clear the internal cache. You have to call this method when
 *     `additionalPluginPool` was updated if `baseConfig` or `cliConfig` depends
 *     on the additional plugins. (`CLIEngine#addPlugin()` method calls this.)
 *
 * @author Toru Nagashima <https://github.com/mysticatea>
 */
"use strict";

//------------------------------------------------------------------------------
// Requirements
//------------------------------------------------------------------------------

const os = require("os");
const path = require("path");
const ***REMOVED*** validateConfigArray ***REMOVED*** = require("../shared/config-validator");
const ***REMOVED*** ConfigArrayFactory ***REMOVED*** = require("./config-array-factory");
const ***REMOVED*** ConfigArray, ConfigDependency ***REMOVED*** = require("./config-array");
const loadRules = require("./load-rules");
const debug = require("debug")("eslint:cascading-config-array-factory");

//------------------------------------------------------------------------------
// Helpers
//------------------------------------------------------------------------------

// Define types for VSCode IntelliSense.
/** @typedef ***REMOVED***import("../shared/types").ConfigData***REMOVED*** ConfigData */
/** @typedef ***REMOVED***import("../shared/types").Parser***REMOVED*** Parser */
/** @typedef ***REMOVED***import("../shared/types").Plugin***REMOVED*** Plugin */
/** @typedef ***REMOVED***ReturnType<ConfigArrayFactory["create"]>***REMOVED*** ConfigArray */

/**
 * @typedef ***REMOVED***Object***REMOVED*** CascadingConfigArrayFactoryOptions
 * @property ***REMOVED***Map<string,Plugin>***REMOVED*** [additionalPluginPool] The map for additional plugins.
 * @property ***REMOVED***ConfigData***REMOVED*** [baseConfig] The config by `baseConfig` option.
 * @property ***REMOVED***ConfigData***REMOVED*** [cliConfig] The config by CLI options (`--env`, `--global`, `--parser`, `--parser-options`, `--plugin`, and `--rule`). CLI options overwrite the setting in config files.
 * @property ***REMOVED***string***REMOVED*** [cwd] The base directory to start lookup.
 * @property ***REMOVED***string[]***REMOVED*** [rulePaths] The value of `--rulesdir` option.
 * @property ***REMOVED***string***REMOVED*** [specificConfigPath] The value of `--config` option.
 * @property ***REMOVED***boolean***REMOVED*** [useEslintrc] if `false` then it doesn't load config files.
 */

/**
 * @typedef ***REMOVED***Object***REMOVED*** CascadingConfigArrayFactoryInternalSlots
 * @property ***REMOVED***ConfigArray***REMOVED*** baseConfigArray The config array of `baseConfig` option.
 * @property ***REMOVED***ConfigData***REMOVED*** baseConfigData The config data of `baseConfig` option. This is used to reset `baseConfigArray`.
 * @property ***REMOVED***ConfigArray***REMOVED*** cliConfigArray The config array of CLI options.
 * @property ***REMOVED***ConfigData***REMOVED*** cliConfigData The config data of CLI options. This is used to reset `cliConfigArray`.
 * @property ***REMOVED***ConfigArrayFactory***REMOVED*** configArrayFactory The factory for config arrays.
 * @property ***REMOVED***Map<string, ConfigArray>***REMOVED*** configCache The cache from directory paths to config arrays.
 * @property ***REMOVED***string***REMOVED*** cwd The base directory to start lookup.
 * @property ***REMOVED***WeakMap<ConfigArray, ConfigArray>***REMOVED*** finalizeCache The cache from config arrays to finalized config arrays.
 * @property ***REMOVED***string[]|null***REMOVED*** rulePaths The value of `--rulesdir` option. This is used to reset `baseConfigArray`.
 * @property ***REMOVED***string|null***REMOVED*** specificConfigPath The value of `--config` option. This is used to reset `cliConfigArray`.
 * @property ***REMOVED***boolean***REMOVED*** useEslintrc if `false` then it doesn't load config files.
 */

/** @type ***REMOVED***WeakMap<CascadingConfigArrayFactory, CascadingConfigArrayFactoryInternalSlots>***REMOVED*** */
const internalSlotsMap = new WeakMap();

/**
 * Create the config array from `baseConfig` and `rulePaths`.
 * @param ***REMOVED***CascadingConfigArrayFactoryInternalSlots***REMOVED*** slots The slots.
 * @returns ***REMOVED***ConfigArray***REMOVED*** The config array of the base configs.
 */
function createBaseConfigArray(***REMOVED***
    configArrayFactory,
    baseConfigData,
    rulePaths,
    cwd
***REMOVED***) ***REMOVED***
    const baseConfigArray = configArrayFactory.create(
        baseConfigData,
        ***REMOVED*** name: "BaseConfig" ***REMOVED***
    );

    if (rulePaths && rulePaths.length > 0) ***REMOVED***

        /*
         * Load rules `--rulesdir` option as a pseudo plugin.
         * Use a pseudo plugin to define rules of `--rulesdir`, so we can
         * validate the rule's options with only information in the config
         * array.
         */
        baseConfigArray.push(***REMOVED***
            name: "--rulesdir",
            filePath: "",
            plugins: ***REMOVED***
                "": new ConfigDependency(***REMOVED***
                    definition: ***REMOVED***
                        rules: rulePaths.reduce(
                            (map, rulesPath) => Object.assign(
                                map,
                                loadRules(rulesPath, cwd)
                            ),
                            ***REMOVED******REMOVED***
                        )
                    ***REMOVED***,
                    filePath: "",
                    id: "",
                    importerName: "--rulesdir",
                    importerPath: ""
                ***REMOVED***)
            ***REMOVED***
        ***REMOVED***);
    ***REMOVED***

    return baseConfigArray;
***REMOVED***

/**
 * Create the config array from CLI options.
 * @param ***REMOVED***CascadingConfigArrayFactoryInternalSlots***REMOVED*** slots The slots.
 * @returns ***REMOVED***ConfigArray***REMOVED*** The config array of the base configs.
 */
function createCLIConfigArray(***REMOVED***
    cliConfigData,
    configArrayFactory,
    specificConfigPath
***REMOVED***) ***REMOVED***
    const cliConfigArray = configArrayFactory.create(
        cliConfigData,
        ***REMOVED*** name: "CLIOptions" ***REMOVED***
    );

    if (specificConfigPath) ***REMOVED***
        cliConfigArray.unshift(
            ...configArrayFactory.loadFile(
                specificConfigPath,
                ***REMOVED*** name: "--config" ***REMOVED***
            )
        );
    ***REMOVED***

    return cliConfigArray;
***REMOVED***

/**
 * The error type when there are files matched by a glob, but all of them have been ignored.
 */
class ConfigurationNotFoundError extends Error ***REMOVED***

    /**
     * @param ***REMOVED***string***REMOVED*** directoryPath - The directory path.
     */
    constructor(directoryPath) ***REMOVED***
        super(`No ESLint configuration found in $***REMOVED***directoryPath***REMOVED***.`);
        this.messageTemplate = "no-config-found";
        this.messageData = ***REMOVED*** directoryPath ***REMOVED***;
    ***REMOVED***
***REMOVED***

/**
 * This class provides the functionality that enumerates every file which is
 * matched by given glob patterns and that configuration.
 */
class CascadingConfigArrayFactory ***REMOVED***

    /**
     * Initialize this enumerator.
     * @param ***REMOVED***CascadingConfigArrayFactoryOptions***REMOVED*** options The options.
     */
    constructor(***REMOVED***
        additionalPluginPool = new Map(),
        baseConfig: baseConfigData = null,
        cliConfig: cliConfigData = null,
        cwd = process.cwd(),
        resolvePluginsRelativeTo = cwd,
        rulePaths = [],
        specificConfigPath = null,
        useEslintrc = true
    ***REMOVED*** = ***REMOVED******REMOVED***) ***REMOVED***
        const configArrayFactory = new ConfigArrayFactory(***REMOVED***
            additionalPluginPool,
            cwd,
            resolvePluginsRelativeTo
        ***REMOVED***);

        internalSlotsMap.set(this, ***REMOVED***
            baseConfigArray: createBaseConfigArray(***REMOVED***
                baseConfigData,
                configArrayFactory,
                cwd,
                rulePaths
            ***REMOVED***),
            baseConfigData,
            cliConfigArray: createCLIConfigArray(***REMOVED***
                cliConfigData,
                configArrayFactory,
                specificConfigPath
            ***REMOVED***),
            cliConfigData,
            configArrayFactory,
            configCache: new Map(),
            cwd,
            finalizeCache: new WeakMap(),
            rulePaths,
            specificConfigPath,
            useEslintrc
        ***REMOVED***);
    ***REMOVED***

    /**
     * The path to the current working directory.
     * This is used by tests.
     * @type ***REMOVED***string***REMOVED***
     */
    get cwd() ***REMOVED***
        const ***REMOVED*** cwd ***REMOVED*** = internalSlotsMap.get(this);

        return cwd;
    ***REMOVED***

    /**
     * Get the config array of a given file.
     * If `filePath` was not given, it returns the config which contains only
     * `baseConfigData` and `cliConfigData`.
     * @param ***REMOVED***string***REMOVED*** [filePath] The file path to a file.
     * @returns ***REMOVED***ConfigArray***REMOVED*** The config array of the file.
     */
    getConfigArrayForFile(filePath) ***REMOVED***
        const ***REMOVED***
            baseConfigArray,
            cliConfigArray,
            cwd
        ***REMOVED*** = internalSlotsMap.get(this);

        if (!filePath) ***REMOVED***
            return new ConfigArray(...baseConfigArray, ...cliConfigArray);
        ***REMOVED***

        const directoryPath = path.dirname(path.resolve(cwd, filePath));

        debug(`Load config files for $***REMOVED***directoryPath***REMOVED***.`);

        return this._finalizeConfigArray(
            this._loadConfigInAncestors(directoryPath),
            directoryPath
        );
    ***REMOVED***

    /**
     * Clear config cache.
     * @returns ***REMOVED***void***REMOVED***
     */
    clearCache() ***REMOVED***
        const slots = internalSlotsMap.get(this);

        slots.baseConfigArray = createBaseConfigArray(slots);
        slots.cliConfigArray = createCLIConfigArray(slots);
        slots.configCache.clear();
    ***REMOVED***

    /**
     * Load and normalize config files from the ancestor directories.
     * @param ***REMOVED***string***REMOVED*** directoryPath The path to a leaf directory.
     * @returns ***REMOVED***ConfigArray***REMOVED*** The loaded config.
     * @private
     */
    _loadConfigInAncestors(directoryPath) ***REMOVED***
        const ***REMOVED***
            baseConfigArray,
            configArrayFactory,
            configCache,
            cwd,
            useEslintrc
        ***REMOVED*** = internalSlotsMap.get(this);

        if (!useEslintrc) ***REMOVED***
            return baseConfigArray;
        ***REMOVED***

        let configArray = configCache.get(directoryPath);

        // Hit cache.
        if (configArray) ***REMOVED***
            debug(`Cache hit: $***REMOVED***directoryPath***REMOVED***.`);
            return configArray;
        ***REMOVED***
        debug(`No cache found: $***REMOVED***directoryPath***REMOVED***.`);

        const homePath = os.homedir();

        // Consider this is root.
        if (directoryPath === homePath && cwd !== homePath) ***REMOVED***
            debug("Stop traversing because of considered root.");
            return this._cacheConfig(directoryPath, baseConfigArray);
        ***REMOVED***

        // Load the config on this directory.
        try ***REMOVED***
            configArray = configArrayFactory.loadInDirectory(directoryPath);
        ***REMOVED*** catch (error) ***REMOVED***
            /* istanbul ignore next */
            if (error.code === "EACCES") ***REMOVED***
                debug("Stop traversing because of 'EACCES' error.");
                return this._cacheConfig(directoryPath, baseConfigArray);
            ***REMOVED***
            throw error;
        ***REMOVED***

        if (configArray.length > 0 && configArray.isRoot()) ***REMOVED***
            debug("Stop traversing because of 'root:true'.");
            configArray.unshift(...baseConfigArray);
            return this._cacheConfig(directoryPath, configArray);
        ***REMOVED***

        // Load from the ancestors and merge it.
        const parentPath = path.dirname(directoryPath);
        const parentConfigArray = parentPath && parentPath !== directoryPath
            ? this._loadConfigInAncestors(parentPath)
            : baseConfigArray;

        if (configArray.length > 0) ***REMOVED***
            configArray.unshift(...parentConfigArray);
        ***REMOVED*** else ***REMOVED***
            configArray = parentConfigArray;
        ***REMOVED***

        // Cache and return.
        return this._cacheConfig(directoryPath, configArray);
    ***REMOVED***

    /**
     * Freeze and cache a given config.
     * @param ***REMOVED***string***REMOVED*** directoryPath The path to a directory as a cache key.
     * @param ***REMOVED***ConfigArray***REMOVED*** configArray The config array as a cache value.
     * @returns ***REMOVED***ConfigArray***REMOVED*** The `configArray` (frozen).
     */
    _cacheConfig(directoryPath, configArray) ***REMOVED***
        const ***REMOVED*** configCache ***REMOVED*** = internalSlotsMap.get(this);

        Object.freeze(configArray);
        configCache.set(directoryPath, configArray);

        return configArray;
    ***REMOVED***

    /**
     * Finalize a given config array.
     * Concatenate `--config` and other CLI options.
     * @param ***REMOVED***ConfigArray***REMOVED*** configArray The parent config array.
     * @param ***REMOVED***string***REMOVED*** directoryPath The path to the leaf directory to find config files.
     * @returns ***REMOVED***ConfigArray***REMOVED*** The loaded config.
     * @private
     */
    _finalizeConfigArray(configArray, directoryPath) ***REMOVED***
        const ***REMOVED***
            cliConfigArray,
            configArrayFactory,
            finalizeCache,
            useEslintrc
        ***REMOVED*** = internalSlotsMap.get(this);

        let finalConfigArray = finalizeCache.get(configArray);

        if (!finalConfigArray) ***REMOVED***
            finalConfigArray = configArray;

            // Load the personal config if there are no regular config files.
            if (
                useEslintrc &&
                configArray.every(c => !c.filePath) &&
                cliConfigArray.every(c => !c.filePath) // `--config` option can be a file.
            ) ***REMOVED***
                debug("Loading the config file of the home directory.");

                finalConfigArray = configArrayFactory.loadInDirectory(
                    os.homedir(),
                    ***REMOVED*** name: "PersonalConfig", parent: finalConfigArray ***REMOVED***
                );
            ***REMOVED***

            // Apply CLI options.
            if (cliConfigArray.length > 0) ***REMOVED***
                finalConfigArray = finalConfigArray.concat(cliConfigArray);
            ***REMOVED***

            // Validate rule settings and environments.
            validateConfigArray(finalConfigArray);

            // Cache it.
            Object.freeze(finalConfigArray);
            finalizeCache.set(configArray, finalConfigArray);

            debug(
                "Configuration was determined: %o on %s",
                finalConfigArray,
                directoryPath
            );
        ***REMOVED***

        if (useEslintrc && finalConfigArray.length === 0) ***REMOVED***
            throw new ConfigurationNotFoundError(directoryPath);
        ***REMOVED***

        return finalConfigArray;
    ***REMOVED***
***REMOVED***

//------------------------------------------------------------------------------
// Public Interface
//------------------------------------------------------------------------------

module.exports = ***REMOVED*** CascadingConfigArrayFactory ***REMOVED***;
