// eslint-plugin-node/lib/rules/no-unsupported-features/es-builtins.js
/**
 * @author Toru Nagashima
 * See LICENSE file in root directory for full license.
 */
"use strict"

const ***REMOVED*** READ ***REMOVED*** = require("eslint-utils")
const checkUnsupportedBuiltins = require("../../util/check-unsupported-builtins")
const enumeratePropertyNames = require("../../util/enumerate-property-names")

const trackMap = ***REMOVED***
    globals: ***REMOVED***
        Array: ***REMOVED***
            from: ***REMOVED*** [READ]: ***REMOVED*** supported: "4.0.0" ***REMOVED*** ***REMOVED***,
            of: ***REMOVED*** [READ]: ***REMOVED*** supported: "4.0.0" ***REMOVED*** ***REMOVED***,
        ***REMOVED***,
        BigInt: ***REMOVED***
            [READ]: ***REMOVED*** supported: "10.4.0" ***REMOVED***,
        ***REMOVED***,
        Map: ***REMOVED***
            [READ]: ***REMOVED*** supported: "0.12.0" ***REMOVED***,
        ***REMOVED***,
        Math: ***REMOVED***
            acosh: ***REMOVED*** [READ]: ***REMOVED*** supported: "0.12.0" ***REMOVED*** ***REMOVED***,
            asinh: ***REMOVED*** [READ]: ***REMOVED*** supported: "0.12.0" ***REMOVED*** ***REMOVED***,
            atanh: ***REMOVED*** [READ]: ***REMOVED*** supported: "0.12.0" ***REMOVED*** ***REMOVED***,
            cbrt: ***REMOVED*** [READ]: ***REMOVED*** supported: "0.12.0" ***REMOVED*** ***REMOVED***,
            clz32: ***REMOVED*** [READ]: ***REMOVED*** supported: "0.12.0" ***REMOVED*** ***REMOVED***,
            cosh: ***REMOVED*** [READ]: ***REMOVED*** supported: "0.12.0" ***REMOVED*** ***REMOVED***,
            expm1: ***REMOVED*** [READ]: ***REMOVED*** supported: "0.12.0" ***REMOVED*** ***REMOVED***,
            fround: ***REMOVED*** [READ]: ***REMOVED*** supported: "0.12.0" ***REMOVED*** ***REMOVED***,
            hypot: ***REMOVED*** [READ]: ***REMOVED*** supported: "0.12.0" ***REMOVED*** ***REMOVED***,
            imul: ***REMOVED*** [READ]: ***REMOVED*** supported: "0.12.0" ***REMOVED*** ***REMOVED***,
            log10: ***REMOVED*** [READ]: ***REMOVED*** supported: "0.12.0" ***REMOVED*** ***REMOVED***,
            log1p: ***REMOVED*** [READ]: ***REMOVED*** supported: "0.12.0" ***REMOVED*** ***REMOVED***,
            log2: ***REMOVED*** [READ]: ***REMOVED*** supported: "0.12.0" ***REMOVED*** ***REMOVED***,
            sign: ***REMOVED*** [READ]: ***REMOVED*** supported: "0.12.0" ***REMOVED*** ***REMOVED***,
            sinh: ***REMOVED*** [READ]: ***REMOVED*** supported: "0.12.0" ***REMOVED*** ***REMOVED***,
            tanh: ***REMOVED*** [READ]: ***REMOVED*** supported: "0.12.0" ***REMOVED*** ***REMOVED***,
            trunc: ***REMOVED*** [READ]: ***REMOVED*** supported: "0.12.0" ***REMOVED*** ***REMOVED***,
        ***REMOVED***,
        Number: ***REMOVED***
            EPSILON: ***REMOVED*** [READ]: ***REMOVED*** supported: "0.12.0" ***REMOVED*** ***REMOVED***,
            isFinite: ***REMOVED*** [READ]: ***REMOVED*** supported: "0.10.0" ***REMOVED*** ***REMOVED***,
            isInteger: ***REMOVED*** [READ]: ***REMOVED*** supported: "0.12.0" ***REMOVED*** ***REMOVED***,
            isNaN: ***REMOVED*** [READ]: ***REMOVED*** supported: "0.10.0" ***REMOVED*** ***REMOVED***,
            isSafeInteger: ***REMOVED*** [READ]: ***REMOVED*** supported: "0.12.0" ***REMOVED*** ***REMOVED***,
            MAX_SAFE_INTEGER: ***REMOVED*** [READ]: ***REMOVED*** supported: "0.12.0" ***REMOVED*** ***REMOVED***,
            MIN_SAFE_INTEGER: ***REMOVED*** [READ]: ***REMOVED*** supported: "0.12.0" ***REMOVED*** ***REMOVED***,
            parseFloat: ***REMOVED*** [READ]: ***REMOVED*** supported: "0.12.0" ***REMOVED*** ***REMOVED***,
            parseInt: ***REMOVED*** [READ]: ***REMOVED*** supported: "0.12.0" ***REMOVED*** ***REMOVED***,
        ***REMOVED***,
        Object: ***REMOVED***
            assign: ***REMOVED*** [READ]: ***REMOVED*** supported: "4.0.0" ***REMOVED*** ***REMOVED***,
            fromEntries: ***REMOVED*** [READ]: ***REMOVED*** supported: "12.0.0" ***REMOVED*** ***REMOVED***,
            getOwnPropertySymbols: ***REMOVED*** [READ]: ***REMOVED*** supported: "0.12.0" ***REMOVED*** ***REMOVED***,
            is: ***REMOVED*** [READ]: ***REMOVED*** supported: "0.10.0" ***REMOVED*** ***REMOVED***,
            setPrototypeOf: ***REMOVED*** [READ]: ***REMOVED*** supported: "0.12.0" ***REMOVED*** ***REMOVED***,
            values: ***REMOVED*** [READ]: ***REMOVED*** supported: "7.0.0" ***REMOVED*** ***REMOVED***,
            entries: ***REMOVED*** [READ]: ***REMOVED*** supported: "7.0.0" ***REMOVED*** ***REMOVED***,
            getOwnPropertyDescriptors: ***REMOVED*** [READ]: ***REMOVED*** supported: "7.0.0" ***REMOVED*** ***REMOVED***,
        ***REMOVED***,
        Promise: ***REMOVED***
            [READ]: ***REMOVED*** supported: "0.12.0" ***REMOVED***,
            allSettled: ***REMOVED*** [READ]: ***REMOVED*** supported: "12.9.0" ***REMOVED*** ***REMOVED***,
        ***REMOVED***,
        Proxy: ***REMOVED***
            [READ]: ***REMOVED*** supported: "6.0.0" ***REMOVED***,
        ***REMOVED***,
        Reflect: ***REMOVED***
            [READ]: ***REMOVED*** supported: "6.0.0" ***REMOVED***,
        ***REMOVED***,
        Set: ***REMOVED***
            [READ]: ***REMOVED*** supported: "0.12.0" ***REMOVED***,
        ***REMOVED***,
        String: ***REMOVED***
            fromCodePoint: ***REMOVED*** [READ]: ***REMOVED*** supported: "4.0.0" ***REMOVED*** ***REMOVED***,
            raw: ***REMOVED*** [READ]: ***REMOVED*** supported: "4.0.0" ***REMOVED*** ***REMOVED***,
        ***REMOVED***,
        Symbol: ***REMOVED***
            [READ]: ***REMOVED*** supported: "0.12.0" ***REMOVED***,
        ***REMOVED***,
        Int8Array: ***REMOVED***
            [READ]: ***REMOVED*** supported: "0.10.0" ***REMOVED***,
        ***REMOVED***,
        Uint8Array: ***REMOVED***
            [READ]: ***REMOVED*** supported: "0.10.0" ***REMOVED***,
        ***REMOVED***,
        Uint8ClampedArray: ***REMOVED***
            [READ]: ***REMOVED*** supported: "0.10.0" ***REMOVED***,
        ***REMOVED***,
        Int16Array: ***REMOVED***
            [READ]: ***REMOVED*** supported: "0.10.0" ***REMOVED***,
        ***REMOVED***,
        Uint16Array: ***REMOVED***
            [READ]: ***REMOVED*** supported: "0.10.0" ***REMOVED***,
        ***REMOVED***,
        Int32Array: ***REMOVED***
            [READ]: ***REMOVED*** supported: "0.10.0" ***REMOVED***,
        ***REMOVED***,
        Uint32Array: ***REMOVED***
            [READ]: ***REMOVED*** supported: "0.10.0" ***REMOVED***,
        ***REMOVED***,
        BigInt64Array: ***REMOVED***
            [READ]: ***REMOVED*** supported: "10.4.0" ***REMOVED***,
        ***REMOVED***,
        BigUint64Array: ***REMOVED***
            [READ]: ***REMOVED*** supported: "10.4.0" ***REMOVED***,
        ***REMOVED***,
        Float32Array: ***REMOVED***
            [READ]: ***REMOVED*** supported: "0.10.0" ***REMOVED***,
        ***REMOVED***,
        Float64Array: ***REMOVED***
            [READ]: ***REMOVED*** supported: "0.10.0" ***REMOVED***,
        ***REMOVED***,
        DataView: ***REMOVED***
            [READ]: ***REMOVED*** supported: "0.10.0" ***REMOVED***,
        ***REMOVED***,
        WeakMap: ***REMOVED***
            [READ]: ***REMOVED*** supported: "0.12.0" ***REMOVED***,
        ***REMOVED***,
        WeakSet: ***REMOVED***
            [READ]: ***REMOVED*** supported: "0.12.0" ***REMOVED***,
        ***REMOVED***,
        Atomics: ***REMOVED***
            [READ]: ***REMOVED*** supported: "8.10.0" ***REMOVED***,
        ***REMOVED***,
        SharedArrayBuffer: ***REMOVED***
            [READ]: ***REMOVED*** supported: "8.10.0" ***REMOVED***,
        ***REMOVED***,
    ***REMOVED***,
***REMOVED***

module.exports = ***REMOVED***
    meta: ***REMOVED***
        docs: ***REMOVED***
            description:
                "disallow unsupported ECMAScript built-ins on the specified version",
            category: "Possible Errors",
            recommended: true,
            url:
                "https://github.com/mysticatea/eslint-plugin-node/blob/v10.0.0/docs/rules/no-unsupported-features/es-builtins.md",
        ***REMOVED***,
        type: "problem",
        fixable: null,
        schema: [
            ***REMOVED***
                type: "object",
                properties: ***REMOVED***
                    version: ***REMOVED***
                        type: "string",
                    ***REMOVED***,
                    ignores: ***REMOVED***
                        type: "array",
                        items: ***REMOVED***
                            enum: Array.from(
                                enumeratePropertyNames(trackMap.globals)
                            ),
                        ***REMOVED***,
                        uniqueItems: true,
                    ***REMOVED***,
                ***REMOVED***,
                additionalProperties: false,
            ***REMOVED***,
        ],
        messages: ***REMOVED***
            unsupported:
                "The '***REMOVED******REMOVED***name***REMOVED******REMOVED***' is not supported until Node.js ***REMOVED******REMOVED***supported***REMOVED******REMOVED***. The configured version range is '***REMOVED******REMOVED***version***REMOVED******REMOVED***'.",
        ***REMOVED***,
    ***REMOVED***,
    create(context) ***REMOVED***
        return ***REMOVED***
            "Program:exit"() ***REMOVED***
                checkUnsupportedBuiltins(context, trackMap)
            ***REMOVED***,
        ***REMOVED***
    ***REMOVED***,
***REMOVED***
