// eslint-plugin-node/lib/rules/no-unsupported-features/es-syntax.js
/**
 * @author Toru Nagashima
 * See LICENSE file in root directory for full license.
 */
"use strict"

const ***REMOVED*** rules: esRules ***REMOVED*** = require("eslint-plugin-es")
const ***REMOVED*** getInnermostScope ***REMOVED*** = require("eslint-utils")
const ***REMOVED*** Range ***REMOVED*** = require("semver") //eslint-disable-line no-unused-vars
const getConfiguredNodeVersion = require("../../util/get-configured-node-version")
const getSemverRange = require("../../util/get-semver-range")
const mergeVisitorsInPlace = require("../../util/merge-visitors-in-place")

const getOrSet = /^(?:g|s)et$/u
const features = ***REMOVED***
    //--------------------------------------------------------------------------
    // ES2015
    //--------------------------------------------------------------------------
    arrowFunctions: ***REMOVED***
        ruleId: "no-arrow-functions",
        cases: [
            ***REMOVED***
                supported: "4.0.0",
                messageId: "no-arrow-functions",
            ***REMOVED***,
        ],
    ***REMOVED***,
    binaryNumericLiterals: ***REMOVED***
        ruleId: "no-binary-numeric-literals",
        cases: [
            ***REMOVED***
                supported: "4.0.0",
                messageId: "no-binary-numeric-literals",
            ***REMOVED***,
        ],
    ***REMOVED***,
    blockScopedFunctions: ***REMOVED***
        ruleId: "no-block-scoped-functions",
        cases: [
            ***REMOVED***
                supported: "6.0.0",
                test: info => !info.isStrict,
                messageId: "no-block-scoped-functions-sloppy",
            ***REMOVED***,
            ***REMOVED***
                supported: "4.0.0",
                messageId: "no-block-scoped-functions-strict",
            ***REMOVED***,
        ],
    ***REMOVED***,
    blockScopedVariables: ***REMOVED***
        ruleId: "no-block-scoped-variables",
        cases: [
            ***REMOVED***
                supported: "6.0.0",
                test: info => !info.isStrict,
                messageId: "no-block-scoped-variables-sloppy",
            ***REMOVED***,
            ***REMOVED***
                supported: "4.0.0",
                messageId: "no-block-scoped-variables-strict",
            ***REMOVED***,
        ],
    ***REMOVED***,
    classes: ***REMOVED***
        ruleId: "no-classes",
        cases: [
            ***REMOVED***
                supported: "6.0.0",
                test: info => !info.isStrict,
                messageId: "no-classes-sloppy",
            ***REMOVED***,
            ***REMOVED***
                supported: "4.0.0",
                messageId: "no-classes-strict",
            ***REMOVED***,
        ],
    ***REMOVED***,
    computedProperties: ***REMOVED***
        ruleId: "no-computed-properties",
        cases: [
            ***REMOVED***
                supported: "4.0.0",
                messageId: "no-computed-properties",
            ***REMOVED***,
        ],
    ***REMOVED***,
    defaultParameters: ***REMOVED***
        ruleId: "no-default-parameters",
        cases: [
            ***REMOVED***
                supported: "6.0.0",
                messageId: "no-default-parameters",
            ***REMOVED***,
        ],
    ***REMOVED***,
    destructuring: ***REMOVED***
        ruleId: "no-destructuring",
        cases: [
            ***REMOVED***
                supported: "6.0.0",
                messageId: "no-destructuring",
            ***REMOVED***,
        ],
    ***REMOVED***,
    forOfLoops: ***REMOVED***
        ruleId: "no-for-of-loops",
        cases: [
            ***REMOVED***
                supported: "0.12.0",
                messageId: "no-for-of-loops",
            ***REMOVED***,
        ],
    ***REMOVED***,
    generators: ***REMOVED***
        ruleId: "no-generators",
        cases: [
            ***REMOVED***
                supported: "4.0.0",
                messageId: "no-generators",
            ***REMOVED***,
        ],
    ***REMOVED***,
    modules: ***REMOVED***
        ruleId: "no-modules",
        cases: [
            ***REMOVED***
                supported: null,
                messageId: "no-modules",
            ***REMOVED***,
        ],
    ***REMOVED***,
    "new.target": ***REMOVED***
        ruleId: "no-new-target",
        cases: [
            ***REMOVED***
                supported: "5.0.0",
                messageId: "no-new-target",
            ***REMOVED***,
        ],
    ***REMOVED***,
    objectSuperProperties: ***REMOVED***
        ruleId: "no-object-super-properties",
        cases: [
            ***REMOVED***
                supported: "4.0.0",
                messageId: "no-object-super-properties",
            ***REMOVED***,
        ],
    ***REMOVED***,
    octalNumericLiterals: ***REMOVED***
        ruleId: "no-octal-numeric-literals",
        cases: [
            ***REMOVED***
                supported: "4.0.0",
                messageId: "no-octal-numeric-literals",
            ***REMOVED***,
        ],
    ***REMOVED***,
    propertyShorthands: ***REMOVED***
        ruleId: "no-property-shorthands",
        cases: [
            ***REMOVED***
                supported: "6.0.0",
                test: info =>
                    info.node.shorthand && getOrSet.test(info.node.key.name),
                messageId: "no-property-shorthands-getset",
            ***REMOVED***,
            ***REMOVED***
                supported: "4.0.0",
                messageId: "no-property-shorthands",
            ***REMOVED***,
        ],
    ***REMOVED***,
    regexpU: ***REMOVED***
        ruleId: "no-regexp-u-flag",
        cases: [
            ***REMOVED***
                supported: "6.0.0",
                messageId: "no-regexp-u-flag",
            ***REMOVED***,
        ],
    ***REMOVED***,
    regexpY: ***REMOVED***
        ruleId: "no-regexp-y-flag",
        cases: [
            ***REMOVED***
                supported: "6.0.0",
                messageId: "no-regexp-y-flag",
            ***REMOVED***,
        ],
    ***REMOVED***,
    restParameters: ***REMOVED***
        ruleId: "no-rest-parameters",
        cases: [
            ***REMOVED***
                supported: "6.0.0",
                messageId: "no-rest-parameters",
            ***REMOVED***,
        ],
    ***REMOVED***,
    spreadElements: ***REMOVED***
        ruleId: "no-spread-elements",
        cases: [
            ***REMOVED***
                supported: "5.0.0",
                messageId: "no-spread-elements",
            ***REMOVED***,
        ],
    ***REMOVED***,
    templateLiterals: ***REMOVED***
        ruleId: "no-template-literals",
        cases: [
            ***REMOVED***
                supported: "4.0.0",
                messageId: "no-template-literals",
            ***REMOVED***,
        ],
    ***REMOVED***,
    unicodeCodePointEscapes: ***REMOVED***
        ruleId: "no-unicode-codepoint-escapes",
        cases: [
            ***REMOVED***
                supported: "4.0.0",
                messageId: "no-unicode-codepoint-escapes",
            ***REMOVED***,
        ],
    ***REMOVED***,

    //--------------------------------------------------------------------------
    // ES2016
    //--------------------------------------------------------------------------
    exponentialOperators: ***REMOVED***
        ruleId: "no-exponential-operators",
        cases: [
            ***REMOVED***
                supported: "7.0.0",
                messageId: "no-exponential-operators",
            ***REMOVED***,
        ],
    ***REMOVED***,

    //--------------------------------------------------------------------------
    // ES2017
    //--------------------------------------------------------------------------
    asyncFunctions: ***REMOVED***
        ruleId: "no-async-functions",
        cases: [
            ***REMOVED***
                supported: "7.6.0",
                messageId: "no-async-functions",
            ***REMOVED***,
        ],
    ***REMOVED***,
    trailingCommasInFunctions: ***REMOVED***
        ruleId: "no-trailing-function-commas",
        cases: [
            ***REMOVED***
                supported: "8.0.0",
                messageId: "no-trailing-function-commas",
            ***REMOVED***,
        ],
    ***REMOVED***,

    //--------------------------------------------------------------------------
    // ES2018
    //--------------------------------------------------------------------------
    asyncIteration: ***REMOVED***
        ruleId: "no-async-iteration",
        cases: [
            ***REMOVED***
                supported: "10.0.0",
                messageId: "no-async-iteration",
            ***REMOVED***,
        ],
    ***REMOVED***,
    malformedTemplateLiterals: ***REMOVED***
        ruleId: "no-malformed-template-literals",
        cases: [
            ***REMOVED***
                supported: "8.10.0",
                messageId: "no-malformed-template-literals",
            ***REMOVED***,
        ],
    ***REMOVED***,
    regexpLookbehind: ***REMOVED***
        ruleId: "no-regexp-lookbehind-assertions",
        cases: [
            ***REMOVED***
                supported: "8.10.0",
                messageId: "no-regexp-lookbehind-assertions",
            ***REMOVED***,
        ],
    ***REMOVED***,
    regexpNamedCaptureGroups: ***REMOVED***
        ruleId: "no-regexp-named-capture-groups",
        cases: [
            ***REMOVED***
                supported: "10.0.0",
                messageId: "no-regexp-named-capture-groups",
            ***REMOVED***,
        ],
    ***REMOVED***,
    regexpS: ***REMOVED***
        ruleId: "no-regexp-s-flag",
        cases: [
            ***REMOVED***
                supported: "8.10.0",
                messageId: "no-regexp-s-flag",
            ***REMOVED***,
        ],
    ***REMOVED***,
    regexpUnicodeProperties: ***REMOVED***
        ruleId: "no-regexp-unicode-property-escapes",
        cases: [
            ***REMOVED***
                supported: "10.0.0",
                messageId: "no-regexp-unicode-property-escapes",
            ***REMOVED***,
        ],
    ***REMOVED***,
    restSpreadProperties: ***REMOVED***
        ruleId: "no-rest-spread-properties",
        cases: [
            ***REMOVED***
                supported: "8.3.0",
                messageId: "no-rest-spread-properties",
            ***REMOVED***,
        ],
    ***REMOVED***,

    //--------------------------------------------------------------------------
    // ES2019
    //--------------------------------------------------------------------------
    jsonSuperset: ***REMOVED***
        ruleId: "no-json-superset",
        cases: [
            ***REMOVED***
                supported: "10.0.0",
                messageId: "no-json-superset",
            ***REMOVED***,
        ],
    ***REMOVED***,
    optionalCatchBinding: ***REMOVED***
        ruleId: "no-optional-catch-binding",
        cases: [
            ***REMOVED***
                supported: "10.0.0",
                messageId: "no-optional-catch-binding",
            ***REMOVED***,
        ],
    ***REMOVED***,

    //--------------------------------------------------------------------------
    // ES2020
    //--------------------------------------------------------------------------
    bigint: ***REMOVED***
        ruleId: "no-bigint",
        cases: [
            ***REMOVED***
                supported: "10.4.0",
                test: info => info.node.type === "Literal",
                messageId: "no-bigint",
            ***REMOVED***,
        ],
    ***REMOVED***,
    dynamicImport: ***REMOVED***
        ruleId: "no-dynamic-import",
        cases: [
            ***REMOVED***
                supported: null,
                messageId: "no-dynamic-import",
            ***REMOVED***,
        ],
    ***REMOVED***,
***REMOVED***
const keywords = Object.keys(features)

/**
 * Parses the options.
 * @param ***REMOVED***RuleContext***REMOVED*** context The rule context.
 * @returns ***REMOVED******REMOVED***version:Range,ignores:Set<string>***REMOVED******REMOVED*** Parsed value.
 */
function parseOptions(context) ***REMOVED***
    const raw = context.options[0] || ***REMOVED******REMOVED***
    const filePath = context.getFilename()
    const version = getConfiguredNodeVersion(raw.version, filePath)
    const ignores = new Set(raw.ignores || [])

    return Object.freeze(***REMOVED*** version, ignores ***REMOVED***)
***REMOVED***

/**
 * Find the scope that a given node belongs to.
 * @param ***REMOVED***Scope***REMOVED*** initialScope The initial scope to find.
 * @param ***REMOVED***Node***REMOVED*** node The AST node.
 * @returns ***REMOVED***Scope***REMOVED*** The scope that the node belongs to.
 */
function normalizeScope(initialScope, node) ***REMOVED***
    let scope = getInnermostScope(initialScope, node)

    while (scope && scope.block === node) ***REMOVED***
        scope = scope.upper
    ***REMOVED***

    return scope
***REMOVED***

/**
 * Define the visitor object as merging the rules of eslint-plugin-es.
 * @param ***REMOVED***RuleContext***REMOVED*** context The rule context.
 * @param ***REMOVED******REMOVED***version:Range,ignores:Set<string>***REMOVED******REMOVED*** options The options.
 * @returns ***REMOVED***object***REMOVED*** The defined visitor.
 */
function defineVisitor(context, options) ***REMOVED***
    const testInfoPrototype = ***REMOVED***
        get isStrict() ***REMOVED***
            return normalizeScope(context.getScope(), this.node).isStrict
        ***REMOVED***,
    ***REMOVED***

    /**
     * Check whether a given case object is full-supported on the configured node version.
     * @param ***REMOVED******REMOVED***supported:string***REMOVED******REMOVED*** aCase The case object to check.
     * @returns ***REMOVED***boolean***REMOVED*** `true` if it's supporting.
     */
    function isNotSupportingVersion(aCase) ***REMOVED***
        return (
            !aCase.supported ||
            options.version.intersects(getSemverRange(`<$***REMOVED***aCase.supported***REMOVED***`))
        )
    ***REMOVED***

    /**
     * Define the predicate function to check whether a given case object is supported on the configured node version.
     * @param ***REMOVED***Node***REMOVED*** node The node which is reported.
     * @returns ***REMOVED***function(aCase:***REMOVED***supported:string***REMOVED***):boolean***REMOVED*** The predicate function.
     */
    function isNotSupportingOn(node) ***REMOVED***
        return aCase =>
            isNotSupportingVersion(aCase) &&
            (!aCase.test || aCase.test(***REMOVED*** node, __proto__: testInfoPrototype ***REMOVED***))
    ***REMOVED***

    return (
        keywords
            // Omit full-supported features and ignored features by options
            // because this rule never reports those.
            .filter(
                keyword =>
                    !options.ignores.has(keyword) &&
                    features[keyword].cases.some(isNotSupportingVersion)
            )
            // Merge remaining features with overriding `context.report()`.
            .reduce((visitor, keyword) => ***REMOVED***
                const ***REMOVED*** ruleId, cases ***REMOVED*** = features[keyword]
                const rule = esRules[ruleId]
                const thisContext = ***REMOVED***
                    __proto__: context,

                    // Override `context.report()` then:
                    // - ignore if it's supported.
                    // - override reporting messages.
                    report(descriptor) ***REMOVED***
                        // Set additional information.
                        if (descriptor.data) ***REMOVED***
                            descriptor.data.version = options.version.raw
                        ***REMOVED*** else ***REMOVED***
                            descriptor.data = ***REMOVED*** version: options.version.raw ***REMOVED***
                        ***REMOVED***
                        descriptor.fix = undefined

                        // Test and report.
                        const node = descriptor.node
                        const hitCase = cases.find(isNotSupportingOn(node))
                        if (hitCase) ***REMOVED***
                            descriptor.messageId = hitCase.messageId
                            descriptor.data.supported = hitCase.supported
                            super.report(descriptor)
                        ***REMOVED***
                    ***REMOVED***,
                ***REMOVED***
                return mergeVisitorsInPlace(visitor, rule.create(thisContext))
            ***REMOVED***, ***REMOVED******REMOVED***)
    )
***REMOVED***

module.exports = ***REMOVED***
    meta: ***REMOVED***
        docs: ***REMOVED***
            description:
                "disallow unsupported ECMAScript syntax on the specified version",
            category: "Possible Errors",
            recommended: true,
            url:
                "https://github.com/mysticatea/eslint-plugin-node/blob/v10.0.0/docs/rules/no-unsupported-features/es-syntax.md",
        ***REMOVED***,
        type: "problem",
        fixable: null,
        schema: [
            ***REMOVED***
                type: "object",
                properties: ***REMOVED***
                    version: ***REMOVED***
                        type: "string",
                    ***REMOVED***,
                    ignores: ***REMOVED***
                        type: "array",
                        items: ***REMOVED***
                            enum: Object.keys(features),
                        ***REMOVED***,
                        uniqueItems: true,
                    ***REMOVED***,
                ***REMOVED***,
                additionalProperties: false,
            ***REMOVED***,
        ],
        messages: ***REMOVED***
            //------------------------------------------------------------------
            // ES2015
            //------------------------------------------------------------------
            "no-arrow-functions":
                "Arrow functions are not supported until Node.js ***REMOVED******REMOVED***supported***REMOVED******REMOVED***. The configured version range is '***REMOVED******REMOVED***version***REMOVED******REMOVED***'.",
            "no-binary-numeric-literals":
                "Binary numeric literals are not supported until Node.js ***REMOVED******REMOVED***supported***REMOVED******REMOVED***. The configured version range is '***REMOVED******REMOVED***version***REMOVED******REMOVED***'.",
            "no-block-scoped-functions-strict":
                "Block-scoped functions in strict mode are not supported until Node.js ***REMOVED******REMOVED***supported***REMOVED******REMOVED***. The configured version range is '***REMOVED******REMOVED***version***REMOVED******REMOVED***'.",
            "no-block-scoped-functions-sloppy":
                "Block-scoped functions in non-strict mode are not supported until Node.js ***REMOVED******REMOVED***supported***REMOVED******REMOVED***. The configured version range is '***REMOVED******REMOVED***version***REMOVED******REMOVED***'.",
            "no-block-scoped-variables-strict":
                "Block-scoped variables in strict mode are not supported until Node.js ***REMOVED******REMOVED***supported***REMOVED******REMOVED***. The configured version range is '***REMOVED******REMOVED***version***REMOVED******REMOVED***'.",
            "no-block-scoped-variables-sloppy":
                "Block-scoped variables in non-strict mode are not supported until Node.js ***REMOVED******REMOVED***supported***REMOVED******REMOVED***. The configured version range is '***REMOVED******REMOVED***version***REMOVED******REMOVED***'.",
            "no-classes-strict":
                "Classes in strict mode are not supported until Node.js ***REMOVED******REMOVED***supported***REMOVED******REMOVED***. The configured version range is '***REMOVED******REMOVED***version***REMOVED******REMOVED***'.",
            "no-classes-sloppy":
                "Classes in non-strict mode are not supported until Node.js ***REMOVED******REMOVED***supported***REMOVED******REMOVED***. The configured version range is '***REMOVED******REMOVED***version***REMOVED******REMOVED***'.",
            "no-computed-properties":
                "Computed properties are not supported until Node.js ***REMOVED******REMOVED***supported***REMOVED******REMOVED***. The configured version range is '***REMOVED******REMOVED***version***REMOVED******REMOVED***'.",
            "no-default-parameters":
                "Default parameters are not supported until Node.js ***REMOVED******REMOVED***supported***REMOVED******REMOVED***. The configured version range is '***REMOVED******REMOVED***version***REMOVED******REMOVED***'.",
            "no-destructuring":
                "Destructuring is not supported until Node.js ***REMOVED******REMOVED***supported***REMOVED******REMOVED***. The configured version range is '***REMOVED******REMOVED***version***REMOVED******REMOVED***'.",
            "no-for-of-loops":
                "'for-of' loops are not supported until Node.js ***REMOVED******REMOVED***supported***REMOVED******REMOVED***. The configured version range is '***REMOVED******REMOVED***version***REMOVED******REMOVED***'.",
            "no-generators":
                "Generator functions are not supported until Node.js ***REMOVED******REMOVED***supported***REMOVED******REMOVED***. The configured version range is '***REMOVED******REMOVED***version***REMOVED******REMOVED***'.",
            "no-modules":
                "Import and export declarations are not supported yet.",
            "no-new-target":
                "'new.target' is not supported until Node.js ***REMOVED******REMOVED***supported***REMOVED******REMOVED***. The configured version range is '***REMOVED******REMOVED***version***REMOVED******REMOVED***'.",
            "no-object-super-properties":
                "'super' in object literals is not supported until Node.js ***REMOVED******REMOVED***supported***REMOVED******REMOVED***. The configured version range is '***REMOVED******REMOVED***version***REMOVED******REMOVED***'.",
            "no-octal-numeric-literals":
                "Octal numeric literals are not supported until Node.js ***REMOVED******REMOVED***supported***REMOVED******REMOVED***. The configured version range is '***REMOVED******REMOVED***version***REMOVED******REMOVED***'.",
            "no-property-shorthands":
                "Property shorthands are not supported until Node.js ***REMOVED******REMOVED***supported***REMOVED******REMOVED***. The configured version range is '***REMOVED******REMOVED***version***REMOVED******REMOVED***'.",
            "no-property-shorthands-getset":
                "Property shorthands of 'get' and 'set' are not supported until Node.js ***REMOVED******REMOVED***supported***REMOVED******REMOVED***. The configured version range is '***REMOVED******REMOVED***version***REMOVED******REMOVED***'.",
            "no-regexp-u-flag":
                "RegExp 'u' flag is not supported until Node.js ***REMOVED******REMOVED***supported***REMOVED******REMOVED***. The configured version range is '***REMOVED******REMOVED***version***REMOVED******REMOVED***'.",
            "no-regexp-y-flag":
                "RegExp 'y' flag is not supported until Node.js ***REMOVED******REMOVED***supported***REMOVED******REMOVED***. The configured version range is '***REMOVED******REMOVED***version***REMOVED******REMOVED***'.",
            "no-rest-parameters":
                "Rest parameters are not supported until Node.js ***REMOVED******REMOVED***supported***REMOVED******REMOVED***. The configured version range is '***REMOVED******REMOVED***version***REMOVED******REMOVED***'.",
            "no-spread-elements":
                "Spread elements are not supported until Node.js ***REMOVED******REMOVED***supported***REMOVED******REMOVED***. The configured version range is '***REMOVED******REMOVED***version***REMOVED******REMOVED***'.",
            "no-template-literals":
                "Template literals are not supported until Node.js ***REMOVED******REMOVED***supported***REMOVED******REMOVED***. The configured version range is '***REMOVED******REMOVED***version***REMOVED******REMOVED***'.",
            "no-unicode-codepoint-escapes":
                "Unicode code point escapes are not supported until Node.js ***REMOVED******REMOVED***supported***REMOVED******REMOVED***. The configured version range is '***REMOVED******REMOVED***version***REMOVED******REMOVED***'.",

            //------------------------------------------------------------------
            // ES2016
            //------------------------------------------------------------------
            "no-exponential-operators":
                "Exponential operators are not supported until Node.js ***REMOVED******REMOVED***supported***REMOVED******REMOVED***. The configured version range is '***REMOVED******REMOVED***version***REMOVED******REMOVED***'.",

            //------------------------------------------------------------------
            // ES2017
            //------------------------------------------------------------------
            "no-async-functions":
                "Async functions are not supported until Node.js ***REMOVED******REMOVED***supported***REMOVED******REMOVED***. The configured version range is '***REMOVED******REMOVED***version***REMOVED******REMOVED***'.",
            "no-trailing-function-commas":
                "Trailing commas in function syntax are not supported until Node.js ***REMOVED******REMOVED***supported***REMOVED******REMOVED***. The configured version range is '***REMOVED******REMOVED***version***REMOVED******REMOVED***'.",

            //------------------------------------------------------------------
            // ES2018
            //------------------------------------------------------------------
            "no-async-iteration":
                "Async iteration is not supported until Node.js ***REMOVED******REMOVED***supported***REMOVED******REMOVED***. The configured version range is '***REMOVED******REMOVED***version***REMOVED******REMOVED***'.",
            "no-malformed-template-literals":
                "Malformed template literals are not supported until Node.js ***REMOVED******REMOVED***supported***REMOVED******REMOVED***. The configured version range is '***REMOVED******REMOVED***version***REMOVED******REMOVED***'.",
            "no-regexp-lookbehind-assertions":
                "RegExp lookbehind assertions are not supported until Node.js ***REMOVED******REMOVED***supported***REMOVED******REMOVED***. The configured version range is '***REMOVED******REMOVED***version***REMOVED******REMOVED***'.",
            "no-regexp-named-capture-groups":
                "RegExp named capture groups are not supported until Node.js ***REMOVED******REMOVED***supported***REMOVED******REMOVED***. The configured version range is '***REMOVED******REMOVED***version***REMOVED******REMOVED***'.",
            "no-regexp-s-flag":
                "RegExp 's' flag is not supported until Node.js ***REMOVED******REMOVED***supported***REMOVED******REMOVED***. The configured version range is '***REMOVED******REMOVED***version***REMOVED******REMOVED***'.",
            "no-regexp-unicode-property-escapes":
                "RegExp Unicode property escapes are not supported until Node.js ***REMOVED******REMOVED***supported***REMOVED******REMOVED***. The configured version range is '***REMOVED******REMOVED***version***REMOVED******REMOVED***'.",
            "no-rest-spread-properties":
                "Rest/spread properties are not supported until Node.js ***REMOVED******REMOVED***supported***REMOVED******REMOVED***. The configured version range is '***REMOVED******REMOVED***version***REMOVED******REMOVED***'.",

            //------------------------------------------------------------------
            // ES2019
            //------------------------------------------------------------------
            "no-json-superset":
                "'\\u***REMOVED******REMOVED***code***REMOVED******REMOVED***' in string literals is not supported until Node.js ***REMOVED******REMOVED***supported***REMOVED******REMOVED***. The configured version range is '***REMOVED******REMOVED***version***REMOVED******REMOVED***'.",
            "no-optional-catch-binding":
                "The omission of 'catch' binding is not supported until Node.js ***REMOVED******REMOVED***supported***REMOVED******REMOVED***. The configured version range is '***REMOVED******REMOVED***version***REMOVED******REMOVED***'.",

            //------------------------------------------------------------------
            // ES2020
            //------------------------------------------------------------------
            "no-bigint":
                "Bigint literals are not supported until Node.js ***REMOVED******REMOVED***supported***REMOVED******REMOVED***. The configured version range is '***REMOVED******REMOVED***version***REMOVED******REMOVED***'.",
            "no-dynamic-import":
                "'import()' expressions are not supported until Node.js ***REMOVED******REMOVED***supported***REMOVED******REMOVED***. The configured version range is '***REMOVED******REMOVED***version***REMOVED******REMOVED***'.",
        ***REMOVED***,
    ***REMOVED***,
    create(context) ***REMOVED***
        return defineVisitor(context, parseOptions(context))
    ***REMOVED***,
***REMOVED***
