// eslint-plugin-node/lib/rules/no-unsupported-features/node-builtins.js
/**
 * @author Toru Nagashima
 * See LICENSE file in root directory for full license.
 */
"use strict"

const ***REMOVED*** READ ***REMOVED*** = require("eslint-utils")
const checkUnsupportedBuiltins = require("../../util/check-unsupported-builtins")
const enumeratePropertyNames = require("../../util/enumerate-property-names")

/*eslint-disable camelcase */
const trackMap = ***REMOVED***
    globals: ***REMOVED***
        queueMicrotask: ***REMOVED***
            [READ]: ***REMOVED*** supported: "12.0.0", experimental: "11.0.0" ***REMOVED***,
        ***REMOVED***,
        require: ***REMOVED***
            resolve: ***REMOVED***
                paths: ***REMOVED*** [READ]: ***REMOVED*** supported: "8.9.0" ***REMOVED*** ***REMOVED***,
            ***REMOVED***,
        ***REMOVED***,
    ***REMOVED***,
    modules: ***REMOVED***
        assert: ***REMOVED***
            strict: ***REMOVED***
                [READ]: ***REMOVED*** supported: "9.9.0" ***REMOVED***,
                doesNotReject: ***REMOVED*** [READ]: ***REMOVED*** supported: "10.0.0" ***REMOVED*** ***REMOVED***,
                rejects: ***REMOVED*** [READ]: ***REMOVED*** supported: "10.0.0" ***REMOVED*** ***REMOVED***,
            ***REMOVED***,
            deepStrictEqual: ***REMOVED*** [READ]: ***REMOVED*** supported: "4.0.0" ***REMOVED*** ***REMOVED***,
            doesNotReject: ***REMOVED*** [READ]: ***REMOVED*** supported: "10.0.0" ***REMOVED*** ***REMOVED***,
            notDeepStrictEqual: ***REMOVED*** [READ]: ***REMOVED*** supported: "4.0.0" ***REMOVED*** ***REMOVED***,
            rejects: ***REMOVED*** [READ]: ***REMOVED*** supported: "10.0.0" ***REMOVED*** ***REMOVED***,
        ***REMOVED***,
        async_hooks: ***REMOVED***
            [READ]: ***REMOVED*** supported: "8.0.0" ***REMOVED***,
            createHook: ***REMOVED*** [READ]: ***REMOVED*** supported: "8.1.0" ***REMOVED*** ***REMOVED***,
        ***REMOVED***,
        buffer: ***REMOVED***
            Buffer: ***REMOVED***
                alloc: ***REMOVED*** [READ]: ***REMOVED*** supported: "4.5.0" ***REMOVED*** ***REMOVED***,
                allocUnsafe: ***REMOVED*** [READ]: ***REMOVED*** supported: "4.5.0" ***REMOVED*** ***REMOVED***,
                allocUnsafeSlow: ***REMOVED*** [READ]: ***REMOVED*** supported: "4.5.0" ***REMOVED*** ***REMOVED***,
                from: ***REMOVED*** [READ]: ***REMOVED*** supported: "4.5.0" ***REMOVED*** ***REMOVED***,
            ***REMOVED***,
            kMaxLength: ***REMOVED*** [READ]: ***REMOVED*** supported: "3.0.0" ***REMOVED*** ***REMOVED***,
            transcode: ***REMOVED*** [READ]: ***REMOVED*** supported: "7.1.0" ***REMOVED*** ***REMOVED***,
            constants: ***REMOVED*** [READ]: ***REMOVED*** supported: "8.2.0" ***REMOVED*** ***REMOVED***,
        ***REMOVED***,
        child_process: ***REMOVED***
            ChildProcess: ***REMOVED*** [READ]: ***REMOVED*** supported: "2.2.0" ***REMOVED*** ***REMOVED***,
        ***REMOVED***,
        console: ***REMOVED***
            clear: ***REMOVED*** [READ]: ***REMOVED*** supported: "8.3.0" ***REMOVED*** ***REMOVED***,
            count: ***REMOVED*** [READ]: ***REMOVED*** supported: "8.3.0" ***REMOVED*** ***REMOVED***,
            countReset: ***REMOVED*** [READ]: ***REMOVED*** supported: "8.3.0" ***REMOVED*** ***REMOVED***,
            debug: ***REMOVED*** [READ]: ***REMOVED*** supported: "8.0.0" ***REMOVED*** ***REMOVED***,
            dirxml: ***REMOVED*** [READ]: ***REMOVED*** supported: "8.0.0" ***REMOVED*** ***REMOVED***,
            group: ***REMOVED*** [READ]: ***REMOVED*** supported: "8.5.0" ***REMOVED*** ***REMOVED***,
            groupCollapsed: ***REMOVED*** [READ]: ***REMOVED*** supported: "8.5.0" ***REMOVED*** ***REMOVED***,
            groupEnd: ***REMOVED*** [READ]: ***REMOVED*** supported: "8.5.0" ***REMOVED*** ***REMOVED***,
            table: ***REMOVED*** [READ]: ***REMOVED*** supported: "10.0.0" ***REMOVED*** ***REMOVED***,
            markTimeline: ***REMOVED*** [READ]: ***REMOVED*** supported: "8.0.0" ***REMOVED*** ***REMOVED***,
            profile: ***REMOVED*** [READ]: ***REMOVED*** supported: "8.0.0" ***REMOVED*** ***REMOVED***,
            profileEnd: ***REMOVED*** [READ]: ***REMOVED*** supported: "8.0.0" ***REMOVED*** ***REMOVED***,
            timeLog: ***REMOVED*** [READ]: ***REMOVED*** supported: "10.7.0" ***REMOVED*** ***REMOVED***,
            timeStamp: ***REMOVED*** [READ]: ***REMOVED*** supported: "8.0.0" ***REMOVED*** ***REMOVED***,
            timeline: ***REMOVED*** [READ]: ***REMOVED*** supported: "8.0.0" ***REMOVED*** ***REMOVED***,
            timelineEnd: ***REMOVED*** [READ]: ***REMOVED*** supported: "8.0.0" ***REMOVED*** ***REMOVED***,
        ***REMOVED***,
        crypto: ***REMOVED***
            Certificate: ***REMOVED***
                exportChallenge: ***REMOVED*** [READ]: ***REMOVED*** supported: "9.0.0" ***REMOVED*** ***REMOVED***,
                exportPublicKey: ***REMOVED*** [READ]: ***REMOVED*** supported: "9.0.0" ***REMOVED*** ***REMOVED***,
                verifySpkac: ***REMOVED*** [READ]: ***REMOVED*** supported: "9.0.0" ***REMOVED*** ***REMOVED***,
            ***REMOVED***,
            KeyObject: ***REMOVED*** [READ]: ***REMOVED*** supported: "11.6.0" ***REMOVED*** ***REMOVED***,
            createPrivateKey: ***REMOVED*** [READ]: ***REMOVED*** supported: "11.6.0" ***REMOVED*** ***REMOVED***,
            createPublicKey: ***REMOVED*** [READ]: ***REMOVED*** supported: "11.6.0" ***REMOVED*** ***REMOVED***,
            createSecretKey: ***REMOVED*** [READ]: ***REMOVED*** supported: "11.6.0" ***REMOVED*** ***REMOVED***,
            constants: ***REMOVED*** [READ]: ***REMOVED*** supported: "6.3.0" ***REMOVED*** ***REMOVED***,
            fips: ***REMOVED*** [READ]: ***REMOVED*** supported: "6.0.0" ***REMOVED*** ***REMOVED***,
            generateKeyPair: ***REMOVED*** [READ]: ***REMOVED*** supported: "10.12.0" ***REMOVED*** ***REMOVED***,
            generateKeyPairSync: ***REMOVED*** [READ]: ***REMOVED*** supported: "10.12.0" ***REMOVED*** ***REMOVED***,
            getCurves: ***REMOVED*** [READ]: ***REMOVED*** supported: "2.3.0" ***REMOVED*** ***REMOVED***,
            getFips: ***REMOVED*** [READ]: ***REMOVED*** supported: "10.0.0" ***REMOVED*** ***REMOVED***,
            privateEncrypt: ***REMOVED*** [READ]: ***REMOVED*** supported: "1.1.0" ***REMOVED*** ***REMOVED***,
            publicDecrypt: ***REMOVED*** [READ]: ***REMOVED*** supported: "1.1.0" ***REMOVED*** ***REMOVED***,
            randomFillSync: ***REMOVED*** [READ]: ***REMOVED*** supported: "7.10.0" ***REMOVED*** ***REMOVED***,
            randomFill: ***REMOVED*** [READ]: ***REMOVED*** supported: "7.10.0" ***REMOVED*** ***REMOVED***,
            scrypt: ***REMOVED*** [READ]: ***REMOVED*** supported: "10.5.0" ***REMOVED*** ***REMOVED***,
            scryptSync: ***REMOVED*** [READ]: ***REMOVED*** supported: "10.5.0" ***REMOVED*** ***REMOVED***,
            setFips: ***REMOVED*** [READ]: ***REMOVED*** supported: "10.0.0" ***REMOVED*** ***REMOVED***,
            sign: ***REMOVED*** [READ]: ***REMOVED*** supported: "12.0.0" ***REMOVED*** ***REMOVED***,
            timingSafeEqual: ***REMOVED*** [READ]: ***REMOVED*** supported: "6.6.0" ***REMOVED*** ***REMOVED***,
            verify: ***REMOVED*** [READ]: ***REMOVED*** supported: "12.0.0" ***REMOVED*** ***REMOVED***,
        ***REMOVED***,
        dns: ***REMOVED***
            Resolver: ***REMOVED*** [READ]: ***REMOVED*** supported: "8.3.0" ***REMOVED*** ***REMOVED***,
            resolvePtr: ***REMOVED*** [READ]: ***REMOVED*** supported: "6.0.0" ***REMOVED*** ***REMOVED***,
            promises: ***REMOVED***
                [READ]: ***REMOVED*** supported: "11.14.0", experimental: "10.6.0" ***REMOVED***,
            ***REMOVED***,
        ***REMOVED***,
        events: ***REMOVED***
            EventEmitter: ***REMOVED***
                once: ***REMOVED*** [READ]: ***REMOVED*** supported: "11.13.0" ***REMOVED*** ***REMOVED***,
            ***REMOVED***,
            once: ***REMOVED*** [READ]: ***REMOVED*** supported: "11.13.0" ***REMOVED*** ***REMOVED***,
        ***REMOVED***,
        fs: ***REMOVED***
            Dirent: ***REMOVED*** [READ]: ***REMOVED*** supported: "10.10.0" ***REMOVED*** ***REMOVED***,
            copyFile: ***REMOVED*** [READ]: ***REMOVED*** supported: "8.5.0" ***REMOVED*** ***REMOVED***,
            copyFileSync: ***REMOVED*** [READ]: ***REMOVED*** supported: "8.5.0" ***REMOVED*** ***REMOVED***,
            mkdtemp: ***REMOVED*** [READ]: ***REMOVED*** supported: "5.10.0" ***REMOVED*** ***REMOVED***,
            mkdtempSync: ***REMOVED*** [READ]: ***REMOVED*** supported: "5.10.0" ***REMOVED*** ***REMOVED***,
            realpath: ***REMOVED***
                native: ***REMOVED*** [READ]: ***REMOVED*** supported: "9.2.0" ***REMOVED*** ***REMOVED***,
            ***REMOVED***,
            realpathSync: ***REMOVED***
                native: ***REMOVED*** [READ]: ***REMOVED*** supported: "9.2.0" ***REMOVED*** ***REMOVED***,
            ***REMOVED***,
            promises: ***REMOVED***
                [READ]: ***REMOVED*** supported: "11.14.0", experimental: "10.1.0" ***REMOVED***,
            ***REMOVED***,
        ***REMOVED***,
        http2: ***REMOVED***
            [READ]: ***REMOVED*** supported: "8.4.0" ***REMOVED***,
        ***REMOVED***,
        inspector: ***REMOVED***
            [READ]: ***REMOVED*** supported: "8.0.0" ***REMOVED***,
        ***REMOVED***,
        module: ***REMOVED***
            builtinModules: ***REMOVED*** [READ]: ***REMOVED*** supported: "9.3.0" ***REMOVED*** ***REMOVED***,
        ***REMOVED***,
        os: ***REMOVED***
            constants: ***REMOVED***
                [READ]: ***REMOVED*** supported: "6.3.0" ***REMOVED***,
                priority: ***REMOVED*** [READ]: ***REMOVED*** supported: "10.10.0" ***REMOVED*** ***REMOVED***,
            ***REMOVED***,
            getPriority: ***REMOVED*** [READ]: ***REMOVED*** supported: "10.10.0" ***REMOVED*** ***REMOVED***,
            homedir: ***REMOVED*** [READ]: ***REMOVED*** supported: "2.3.0" ***REMOVED*** ***REMOVED***,
            setPriority: ***REMOVED*** [READ]: ***REMOVED*** supported: "10.10.0" ***REMOVED*** ***REMOVED***,
            userInfo: ***REMOVED*** [READ]: ***REMOVED*** supported: "6.0.0" ***REMOVED*** ***REMOVED***,
        ***REMOVED***,
        path: ***REMOVED***
            toNamespacedPath: ***REMOVED*** [READ]: ***REMOVED*** supported: "9.0.0" ***REMOVED*** ***REMOVED***,
        ***REMOVED***,
        perf_hooks: ***REMOVED***
            [READ]: ***REMOVED*** supported: "8.5.0" ***REMOVED***,
            monitorEventLoopDelay: ***REMOVED*** [READ]: ***REMOVED*** supported: "11.10.0" ***REMOVED*** ***REMOVED***,
        ***REMOVED***,
        process: ***REMOVED***
            allowedNodeEnvironmentFlags: ***REMOVED*** [READ]: ***REMOVED*** supported: "10.10.0" ***REMOVED*** ***REMOVED***,
            argv0: ***REMOVED*** [READ]: ***REMOVED*** supported: "6.4.0" ***REMOVED*** ***REMOVED***,
            channel: ***REMOVED*** [READ]: ***REMOVED*** supported: "7.1.0" ***REMOVED*** ***REMOVED***,
            cpuUsage: ***REMOVED*** [READ]: ***REMOVED*** supported: "6.1.0" ***REMOVED*** ***REMOVED***,
            emitWarning: ***REMOVED*** [READ]: ***REMOVED*** supported: "6.0.0" ***REMOVED*** ***REMOVED***,
            getegid: ***REMOVED*** [READ]: ***REMOVED*** supported: "2.0.0" ***REMOVED*** ***REMOVED***,
            geteuid: ***REMOVED*** [READ]: ***REMOVED*** supported: "2.0.0" ***REMOVED*** ***REMOVED***,
            hasUncaughtExceptionCaptureCallback: ***REMOVED***
                [READ]: ***REMOVED*** supported: "9.3.0" ***REMOVED***,
            ***REMOVED***,
            hrtime: ***REMOVED***
                bigint: ***REMOVED*** [READ]: ***REMOVED*** supported: "10.7.0" ***REMOVED*** ***REMOVED***,
            ***REMOVED***,
            ppid: ***REMOVED*** [READ]: ***REMOVED*** supported: "9.2.0" ***REMOVED*** ***REMOVED***,
            release: ***REMOVED*** [READ]: ***REMOVED*** supported: "3.0.0" ***REMOVED*** ***REMOVED***,
            report: ***REMOVED*** [READ]: ***REMOVED*** supported: null, experimental: "11.8.0" ***REMOVED*** ***REMOVED***,
            setegid: ***REMOVED*** [READ]: ***REMOVED*** supported: "2.0.0" ***REMOVED*** ***REMOVED***,
            seteuid: ***REMOVED*** [READ]: ***REMOVED*** supported: "2.0.0" ***REMOVED*** ***REMOVED***,
            setUncaughtExceptionCaptureCallback: ***REMOVED***
                [READ]: ***REMOVED*** supported: "9.3.0" ***REMOVED***,
            ***REMOVED***,
            stdout: ***REMOVED***
                getColorDepth: ***REMOVED*** [READ]: ***REMOVED*** supported: "9.9.0" ***REMOVED*** ***REMOVED***,
                hasColor: ***REMOVED*** [READ]: ***REMOVED*** supported: "11.13.0" ***REMOVED*** ***REMOVED***,
            ***REMOVED***,
            stderr: ***REMOVED***
                getColorDepth: ***REMOVED*** [READ]: ***REMOVED*** supported: "9.9.0" ***REMOVED*** ***REMOVED***,
                hasColor: ***REMOVED*** [READ]: ***REMOVED*** supported: "11.13.0" ***REMOVED*** ***REMOVED***,
            ***REMOVED***,
        ***REMOVED***,
        stream: ***REMOVED***
            finished: ***REMOVED*** [READ]: ***REMOVED*** supported: "10.0.0" ***REMOVED*** ***REMOVED***,
            pipeline: ***REMOVED*** [READ]: ***REMOVED*** supported: "10.0.0" ***REMOVED*** ***REMOVED***,
        ***REMOVED***,
        trace_events: ***REMOVED***
            [READ]: ***REMOVED*** supported: "10.0.0" ***REMOVED***,
        ***REMOVED***,
        url: ***REMOVED***
            URL: ***REMOVED*** [READ]: ***REMOVED*** supported: "7.0.0" ***REMOVED*** ***REMOVED***,
            URLSearchParams: ***REMOVED*** [READ]: ***REMOVED*** supported: "7.5.0" ***REMOVED*** ***REMOVED***,
            domainToASCII: ***REMOVED*** [READ]: ***REMOVED*** supported: "7.4.0" ***REMOVED*** ***REMOVED***,
            domainToUnicode: ***REMOVED*** [READ]: ***REMOVED*** supported: "7.4.0" ***REMOVED*** ***REMOVED***,
        ***REMOVED***,
        util: ***REMOVED***
            callbackify: ***REMOVED*** [READ]: ***REMOVED*** supported: "8.2.0" ***REMOVED*** ***REMOVED***,
            formatWithOptions: ***REMOVED*** [READ]: ***REMOVED*** supported: "10.0.0" ***REMOVED*** ***REMOVED***,
            getSystemErrorName: ***REMOVED*** [READ]: ***REMOVED*** supported: "9.7.0" ***REMOVED*** ***REMOVED***,
            inspect: ***REMOVED***
                custom: ***REMOVED*** [READ]: ***REMOVED*** supported: "6.6.0" ***REMOVED*** ***REMOVED***,
                defaultOptions: ***REMOVED*** [READ]: ***REMOVED*** supported: "6.4.0" ***REMOVED*** ***REMOVED***,
                replDefaults: ***REMOVED*** [READ]: ***REMOVED*** supported: "11.12.0" ***REMOVED*** ***REMOVED***,
            ***REMOVED***,
            isDeepStrictEqual: ***REMOVED*** [READ]: ***REMOVED*** supported: "9.0.0" ***REMOVED*** ***REMOVED***,
            promisify: ***REMOVED*** [READ]: ***REMOVED*** supported: "8.0.0" ***REMOVED*** ***REMOVED***,
            TextDecoder: ***REMOVED*** [READ]: ***REMOVED*** supported: "8.3.0" ***REMOVED*** ***REMOVED***,
            TextEncoder: ***REMOVED*** [READ]: ***REMOVED*** supported: "8.3.0" ***REMOVED*** ***REMOVED***,
            types: ***REMOVED***
                [READ]: ***REMOVED*** supported: "10.0.0" ***REMOVED***,
                isBoxedPrimitive: ***REMOVED*** [READ]: ***REMOVED*** supported: "10.11.0" ***REMOVED*** ***REMOVED***,
            ***REMOVED***,
        ***REMOVED***,
        v8: ***REMOVED***
            [READ]: ***REMOVED*** supported: "1.0.0" ***REMOVED***,
            DefaultDeserializer: ***REMOVED*** [READ]: ***REMOVED*** supported: "8.0.0" ***REMOVED*** ***REMOVED***,
            DefaultSerializer: ***REMOVED*** [READ]: ***REMOVED*** supported: "8.0.0" ***REMOVED*** ***REMOVED***,
            Deserializer: ***REMOVED*** [READ]: ***REMOVED*** supported: "8.0.0" ***REMOVED*** ***REMOVED***,
            Serializer: ***REMOVED*** [READ]: ***REMOVED*** supported: "8.0.0" ***REMOVED*** ***REMOVED***,
            cachedDataVersionTag: ***REMOVED*** [READ]: ***REMOVED*** supported: "8.0.0" ***REMOVED*** ***REMOVED***,
            deserialize: ***REMOVED*** [READ]: ***REMOVED*** supported: "8.0.0" ***REMOVED*** ***REMOVED***,
            getHeapSnapshot: ***REMOVED*** [READ]: ***REMOVED*** supported: "11.13.0" ***REMOVED*** ***REMOVED***,
            getHeapSpaceStatistics: ***REMOVED*** [READ]: ***REMOVED*** supported: "6.0.0" ***REMOVED*** ***REMOVED***,
            serialize: ***REMOVED*** [READ]: ***REMOVED*** supported: "8.0.0" ***REMOVED*** ***REMOVED***,
            writeHeapSnapshot: ***REMOVED*** [READ]: ***REMOVED*** supported: "11.13.0" ***REMOVED*** ***REMOVED***,
        ***REMOVED***,
        vm: ***REMOVED***
            Module: ***REMOVED*** [READ]: ***REMOVED*** supported: "9.6.0" ***REMOVED*** ***REMOVED***,
            compileFunction: ***REMOVED*** [READ]: ***REMOVED*** supported: "10.10.0" ***REMOVED*** ***REMOVED***,
        ***REMOVED***,
        worker_threads: ***REMOVED***
            [READ]: ***REMOVED*** supported: null, experimental: "10.5.0" ***REMOVED***,
        ***REMOVED***,
    ***REMOVED***,
***REMOVED***
Object.assign(trackMap.globals, ***REMOVED***
    Buffer: trackMap.modules.buffer.Buffer,
    TextDecoder: ***REMOVED***
        ...trackMap.modules.util.TextDecoder,
        [READ]: ***REMOVED*** supported: "11.0.0" ***REMOVED***,
    ***REMOVED***,
    TextEncoder: ***REMOVED***
        ...trackMap.modules.util.TextEncoder,
        [READ]: ***REMOVED*** supported: "11.0.0" ***REMOVED***,
    ***REMOVED***,
    URL: ***REMOVED***
        ...trackMap.modules.url.URL,
        [READ]: ***REMOVED*** supported: "10.0.0" ***REMOVED***,
    ***REMOVED***,
    URLSearchParams: ***REMOVED***
        ...trackMap.modules.url.URLSearchParams,
        [READ]: ***REMOVED*** supported: "10.0.0" ***REMOVED***,
    ***REMOVED***,
    console: trackMap.modules.console,
    process: trackMap.modules.process,
***REMOVED***)
/*eslint-enable camelcase */

module.exports = ***REMOVED***
    meta: ***REMOVED***
        docs: ***REMOVED***
            description:
                "disallow unsupported Node.js built-in APIs on the specified version",
            category: "Possible Errors",
            recommended: true,
            url:
                "https://github.com/mysticatea/eslint-plugin-node/blob/v10.0.0/docs/rules/no-unsupported-features/node-builtins.md",
        ***REMOVED***,
        type: "problem",
        fixable: null,
        schema: [
            ***REMOVED***
                type: "object",
                properties: ***REMOVED***
                    version: ***REMOVED***
                        type: "string",
                    ***REMOVED***,
                    ignores: ***REMOVED***
                        type: "array",
                        items: ***REMOVED***
                            enum: Array.from(
                                new Set([
                                    ...enumeratePropertyNames(trackMap.globals),
                                    ...enumeratePropertyNames(trackMap.modules),
                                ])
                            ),
                        ***REMOVED***,
                        uniqueItems: true,
                    ***REMOVED***,
                ***REMOVED***,
                additionalProperties: false,
            ***REMOVED***,
        ],
        messages: ***REMOVED***
            unsupported:
                "The '***REMOVED******REMOVED***name***REMOVED******REMOVED***' is not supported until Node.js ***REMOVED******REMOVED***supported***REMOVED******REMOVED***. The configured version range is '***REMOVED******REMOVED***version***REMOVED******REMOVED***'.",
        ***REMOVED***,
    ***REMOVED***,
    create(context) ***REMOVED***
        return ***REMOVED***
            "Program:exit"() ***REMOVED***
                checkUnsupportedBuiltins(context, trackMap)
            ***REMOVED***,
        ***REMOVED***
    ***REMOVED***,
***REMOVED***
