// eslint/lib/cli-engine/formatters/html.js
/**
 * @fileoverview HTML reporter
 * @author Julian Laval
 */
"use strict";

const lodash = require("lodash");
const fs = require("fs");
const path = require("path");

//------------------------------------------------------------------------------
// Helpers
//------------------------------------------------------------------------------

const pageTemplate = lodash.template(fs.readFileSync(path.join(__dirname, "html-template-page.html"), "utf-8"));
const messageTemplate = lodash.template(fs.readFileSync(path.join(__dirname, "html-template-message.html"), "utf-8"));
const resultTemplate = lodash.template(fs.readFileSync(path.join(__dirname, "html-template-result.html"), "utf-8"));

/**
 * Given a word and a count, append an s if count is not one.
 * @param ***REMOVED***string***REMOVED*** word A word in its singular form.
 * @param ***REMOVED***int***REMOVED*** count A number controlling whether word should be pluralized.
 * @returns ***REMOVED***string***REMOVED*** The original word with an s on the end if count is not one.
 */
function pluralize(word, count) ***REMOVED***
    return (count === 1 ? word : `$***REMOVED***word***REMOVED***s`);
***REMOVED***

/**
 * Renders text along the template of x problems (x errors, x warnings)
 * @param ***REMOVED***string***REMOVED*** totalErrors Total errors
 * @param ***REMOVED***string***REMOVED*** totalWarnings Total warnings
 * @returns ***REMOVED***string***REMOVED*** The formatted string, pluralized where necessary
 */
function renderSummary(totalErrors, totalWarnings) ***REMOVED***
    const totalProblems = totalErrors + totalWarnings;
    let renderedText = `$***REMOVED***totalProblems***REMOVED*** $***REMOVED***pluralize("problem", totalProblems)***REMOVED***`;

    if (totalProblems !== 0) ***REMOVED***
        renderedText += ` ($***REMOVED***totalErrors***REMOVED*** $***REMOVED***pluralize("error", totalErrors)***REMOVED***, $***REMOVED***totalWarnings***REMOVED*** $***REMOVED***pluralize("warning", totalWarnings)***REMOVED***)`;
    ***REMOVED***
    return renderedText;
***REMOVED***

/**
 * Get the color based on whether there are errors/warnings...
 * @param ***REMOVED***string***REMOVED*** totalErrors Total errors
 * @param ***REMOVED***string***REMOVED*** totalWarnings Total warnings
 * @returns ***REMOVED***int***REMOVED*** The color code (0 = green, 1 = yellow, 2 = red)
 */
function renderColor(totalErrors, totalWarnings) ***REMOVED***
    if (totalErrors !== 0) ***REMOVED***
        return 2;
    ***REMOVED***
    if (totalWarnings !== 0) ***REMOVED***
        return 1;
    ***REMOVED***
    return 0;
***REMOVED***

/**
 * Get HTML (table rows) describing the messages.
 * @param ***REMOVED***Array***REMOVED*** messages Messages.
 * @param ***REMOVED***int***REMOVED*** parentIndex Index of the parent HTML row.
 * @param ***REMOVED***Object***REMOVED*** rulesMeta Dictionary containing metadata for each rule executed by the analysis.
 * @returns ***REMOVED***string***REMOVED*** HTML (table rows) describing the messages.
 */
function renderMessages(messages, parentIndex, rulesMeta) ***REMOVED***

    /**
     * Get HTML (table row) describing a message.
     * @param ***REMOVED***Object***REMOVED*** message Message.
     * @returns ***REMOVED***string***REMOVED*** HTML (table row) describing a message.
     */
    return lodash.map(messages, message => ***REMOVED***
        const lineNumber = message.line || 0;
        const columnNumber = message.column || 0;
        let ruleUrl;

        if (rulesMeta) ***REMOVED***
            const meta = rulesMeta[message.ruleId];

            ruleUrl = lodash.get(meta, "docs.url", null);
        ***REMOVED***

        return messageTemplate(***REMOVED***
            parentIndex,
            lineNumber,
            columnNumber,
            severityNumber: message.severity,
            severityName: message.severity === 1 ? "Warning" : "Error",
            message: message.message,
            ruleId: message.ruleId,
            ruleUrl
        ***REMOVED***);
    ***REMOVED***).join("\n");
***REMOVED***

/**
 * @param ***REMOVED***Array***REMOVED*** results Test results.
 * @param ***REMOVED***Object***REMOVED*** rulesMeta Dictionary containing metadata for each rule executed by the analysis.
 * @returns ***REMOVED***string***REMOVED*** HTML string describing the results.
 */
function renderResults(results, rulesMeta) ***REMOVED***
    return lodash.map(results, (result, index) => resultTemplate(***REMOVED***
        index,
        color: renderColor(result.errorCount, result.warningCount),
        filePath: result.filePath,
        summary: renderSummary(result.errorCount, result.warningCount)

    ***REMOVED***) + renderMessages(result.messages, index, rulesMeta)).join("\n");
***REMOVED***

//------------------------------------------------------------------------------
// Public Interface
//------------------------------------------------------------------------------

module.exports = function(results, data) ***REMOVED***
    let totalErrors,
        totalWarnings;

    const metaData = data ? data.rulesMeta : ***REMOVED******REMOVED***;

    totalErrors = 0;
    totalWarnings = 0;

    // Iterate over results to get totals
    results.forEach(result => ***REMOVED***
        totalErrors += result.errorCount;
        totalWarnings += result.warningCount;
    ***REMOVED***);

    return pageTemplate(***REMOVED***
        date: new Date(),
        reportColor: renderColor(totalErrors, totalWarnings),
        reportSummary: renderSummary(totalErrors, totalWarnings),
        results: renderResults(results, metaData)
    ***REMOVED***);
***REMOVED***;
