// eslint/lib/cli-engine/formatters/table.js
/**
 * @fileoverview "table reporter.
 * @author Gajus Kuizinas <gajus@gajus.com>
 */
"use strict";

//------------------------------------------------------------------------------
// Requirements
//------------------------------------------------------------------------------

const chalk = require("chalk"),
    table = require("table").table;

//------------------------------------------------------------------------------
// Helpers
//------------------------------------------------------------------------------

/**
 * Given a word and a count, append an "s" if count is not one.
 * @param ***REMOVED***string***REMOVED*** word A word.
 * @param ***REMOVED***number***REMOVED*** count Quantity.
 * @returns ***REMOVED***string***REMOVED*** The original word with an s on the end if count is not one.
 */
function pluralize(word, count) ***REMOVED***
    return (count === 1 ? word : `$***REMOVED***word***REMOVED***s`);
***REMOVED***

/**
 * Draws text table.
 * @param ***REMOVED***Array<Object>***REMOVED*** messages Error messages relating to a specific file.
 * @returns ***REMOVED***string***REMOVED*** A text table.
 */
function drawTable(messages) ***REMOVED***
    const rows = [];

    if (messages.length === 0) ***REMOVED***
        return "";
    ***REMOVED***

    rows.push([
        chalk.bold("Line"),
        chalk.bold("Column"),
        chalk.bold("Type"),
        chalk.bold("Message"),
        chalk.bold("Rule ID")
    ]);

    messages.forEach(message => ***REMOVED***
        let messageType;

        if (message.fatal || message.severity === 2) ***REMOVED***
            messageType = chalk.red("error");
        ***REMOVED*** else ***REMOVED***
            messageType = chalk.yellow("warning");
        ***REMOVED***

        rows.push([
            message.line || 0,
            message.column || 0,
            messageType,
            message.message,
            message.ruleId || ""
        ]);
    ***REMOVED***);

    return table(rows, ***REMOVED***
        columns: ***REMOVED***
            0: ***REMOVED***
                width: 8,
                wrapWord: true
            ***REMOVED***,
            1: ***REMOVED***
                width: 8,
                wrapWord: true
            ***REMOVED***,
            2: ***REMOVED***
                width: 8,
                wrapWord: true
            ***REMOVED***,
            3: ***REMOVED***
                paddingRight: 5,
                width: 50,
                wrapWord: true
            ***REMOVED***,
            4: ***REMOVED***
                width: 20,
                wrapWord: true
            ***REMOVED***
        ***REMOVED***,
        drawHorizontalLine(index) ***REMOVED***
            return index === 1;
        ***REMOVED***
    ***REMOVED***);
***REMOVED***

/**
 * Draws a report (multiple tables).
 * @param ***REMOVED***Array***REMOVED*** results Report results for every file.
 * @returns ***REMOVED***string***REMOVED*** A column of text tables.
 */
function drawReport(results) ***REMOVED***
    let files;

    files = results.map(result => ***REMOVED***
        if (!result.messages.length) ***REMOVED***
            return "";
        ***REMOVED***

        return `\n$***REMOVED***result.filePath***REMOVED***\n\n$***REMOVED***drawTable(result.messages)***REMOVED***`;
    ***REMOVED***);

    files = files.filter(content => content.trim());

    return files.join("");
***REMOVED***

//------------------------------------------------------------------------------
// Public Interface
//------------------------------------------------------------------------------

module.exports = function(report) ***REMOVED***
    let result,
        errorCount,
        warningCount;

    result = "";
    errorCount = 0;
    warningCount = 0;

    report.forEach(fileReport => ***REMOVED***
        errorCount += fileReport.errorCount;
        warningCount += fileReport.warningCount;
    ***REMOVED***);

    if (errorCount || warningCount) ***REMOVED***
        result = drawReport(report);
    ***REMOVED***

    result += `\n$***REMOVED***table([
        [
            chalk.red(pluralize(`$***REMOVED***errorCount***REMOVED*** Error`, errorCount))
        ],
        [
            chalk.yellow(pluralize(`$***REMOVED***warningCount***REMOVED*** Warning`, warningCount))
        ]
    ], ***REMOVED***
        columns: ***REMOVED***
            0: ***REMOVED***
                width: 110,
                wrapWord: true
            ***REMOVED***
        ***REMOVED***,
        drawHorizontalLine() ***REMOVED***
            return true;
        ***REMOVED***
    ***REMOVED***)***REMOVED***`;

    return result;
***REMOVED***;
