// eslint/lib/cli-engine/formatters/tap.js
/**
 * @fileoverview TAP reporter
 * @author Jonathan Kingston
 */
"use strict";

const yaml = require("js-yaml");

//------------------------------------------------------------------------------
// Helper Functions
//------------------------------------------------------------------------------

/**
 * Returns a canonical error level string based upon the error message passed in.
 * @param ***REMOVED***Object***REMOVED*** message Individual error message provided by eslint
 * @returns ***REMOVED***string***REMOVED*** Error level string
 */
function getMessageType(message) ***REMOVED***
    if (message.fatal || message.severity === 2) ***REMOVED***
        return "error";
    ***REMOVED***
    return "warning";
***REMOVED***

/**
 * Takes in a JavaScript object and outputs a TAP diagnostics string
 * @param ***REMOVED***Object***REMOVED*** diagnostic JavaScript object to be embedded as YAML into output.
 * @returns ***REMOVED***string***REMOVED*** diagnostics string with YAML embedded - TAP version 13 compliant
 */
function outputDiagnostics(diagnostic) ***REMOVED***
    const prefix = "  ";
    let output = `$***REMOVED***prefix***REMOVED***---\n`;

    output += prefix + yaml.safeDump(diagnostic).split("\n").join(`\n$***REMOVED***prefix***REMOVED***`);
    output += "...\n";
    return output;
***REMOVED***

//------------------------------------------------------------------------------
// Public Interface
//------------------------------------------------------------------------------

module.exports = function(results) ***REMOVED***
    let output = `TAP version 13\n1..$***REMOVED***results.length***REMOVED***\n`;

    results.forEach((result, id) => ***REMOVED***
        const messages = result.messages;
        let testResult = "ok";
        let diagnostics = ***REMOVED******REMOVED***;

        if (messages.length > 0) ***REMOVED***
            messages.forEach(message => ***REMOVED***
                const severity = getMessageType(message);
                const diagnostic = ***REMOVED***
                    message: message.message,
                    severity,
                    data: ***REMOVED***
                        line: message.line || 0,
                        column: message.column || 0,
                        ruleId: message.ruleId || ""
                    ***REMOVED***
                ***REMOVED***;

                // This ensures a warning message is not flagged as error
                if (severity === "error") ***REMOVED***
                    testResult = "not ok";
                ***REMOVED***

                /*
                 * If we have multiple messages place them under a messages key
                 * The first error will be logged as message key
                 * This is to adhere to TAP 13 loosely defined specification of having a message key
                 */
                if ("message" in diagnostics) ***REMOVED***
                    if (typeof diagnostics.messages === "undefined") ***REMOVED***
                        diagnostics.messages = [];
                    ***REMOVED***
                    diagnostics.messages.push(diagnostic);
                ***REMOVED*** else ***REMOVED***
                    diagnostics = diagnostic;
                ***REMOVED***
            ***REMOVED***);
        ***REMOVED***

        output += `$***REMOVED***testResult***REMOVED*** $***REMOVED***id + 1***REMOVED*** - $***REMOVED***result.filePath***REMOVED***\n`;

        // If we have an error include diagnostics
        if (messages.length > 0) ***REMOVED***
            output += outputDiagnostics(diagnostics);
        ***REMOVED***

    ***REMOVED***);

    return output;
***REMOVED***;
