// table/dist/drawBorder.js
"use strict";

Object.defineProperty(exports, "__esModule", ***REMOVED***
  value: true
***REMOVED***);
exports.drawBorderTop = exports.drawBorderJoin = exports.drawBorderBottom = exports.drawBorder = void 0;

/**
 * @typedef drawBorder~parts
 * @property ***REMOVED***string***REMOVED*** left
 * @property ***REMOVED***string***REMOVED*** right
 * @property ***REMOVED***string***REMOVED*** body
 * @property ***REMOVED***string***REMOVED*** join
 */

/**
 * @param ***REMOVED***number[]***REMOVED*** columnSizeIndex
 * @param ***REMOVED***drawBorder~parts***REMOVED*** parts
 * @returns ***REMOVED***string***REMOVED***
 */
const drawBorder = (columnSizeIndex, parts) => ***REMOVED***
  const columns = columnSizeIndex.map(size => ***REMOVED***
    return parts.body.repeat(size);
  ***REMOVED***).join(parts.join);
  return parts.left + columns + parts.right + '\n';
***REMOVED***;
/**
 * @typedef drawBorderTop~parts
 * @property ***REMOVED***string***REMOVED*** topLeft
 * @property ***REMOVED***string***REMOVED*** topRight
 * @property ***REMOVED***string***REMOVED*** topBody
 * @property ***REMOVED***string***REMOVED*** topJoin
 */

/**
 * @param ***REMOVED***number[]***REMOVED*** columnSizeIndex
 * @param ***REMOVED***drawBorderTop~parts***REMOVED*** parts
 * @returns ***REMOVED***string***REMOVED***
 */


exports.drawBorder = drawBorder;

const drawBorderTop = (columnSizeIndex, parts) => ***REMOVED***
  const border = drawBorder(columnSizeIndex, ***REMOVED***
    body: parts.topBody,
    join: parts.topJoin,
    left: parts.topLeft,
    right: parts.topRight
  ***REMOVED***);

  if (border === '\n') ***REMOVED***
    return '';
  ***REMOVED***

  return border;
***REMOVED***;
/**
 * @typedef drawBorderJoin~parts
 * @property ***REMOVED***string***REMOVED*** joinLeft
 * @property ***REMOVED***string***REMOVED*** joinRight
 * @property ***REMOVED***string***REMOVED*** joinBody
 * @property ***REMOVED***string***REMOVED*** joinJoin
 */

/**
 * @param ***REMOVED***number[]***REMOVED*** columnSizeIndex
 * @param ***REMOVED***drawBorderJoin~parts***REMOVED*** parts
 * @returns ***REMOVED***string***REMOVED***
 */


exports.drawBorderTop = drawBorderTop;

const drawBorderJoin = (columnSizeIndex, parts) => ***REMOVED***
  return drawBorder(columnSizeIndex, ***REMOVED***
    body: parts.joinBody,
    join: parts.joinJoin,
    left: parts.joinLeft,
    right: parts.joinRight
  ***REMOVED***);
***REMOVED***;
/**
 * @typedef drawBorderBottom~parts
 * @property ***REMOVED***string***REMOVED*** topLeft
 * @property ***REMOVED***string***REMOVED*** topRight
 * @property ***REMOVED***string***REMOVED*** topBody
 * @property ***REMOVED***string***REMOVED*** topJoin
 */

/**
 * @param ***REMOVED***number[]***REMOVED*** columnSizeIndex
 * @param ***REMOVED***drawBorderBottom~parts***REMOVED*** parts
 * @returns ***REMOVED***string***REMOVED***
 */


exports.drawBorderJoin = drawBorderJoin;

const drawBorderBottom = (columnSizeIndex, parts) => ***REMOVED***
  return drawBorder(columnSizeIndex, ***REMOVED***
    body: parts.bottomBody,
    join: parts.bottomJoin,
    left: parts.bottomLeft,
    right: parts.bottomRight
  ***REMOVED***);
***REMOVED***;

exports.drawBorderBottom = drawBorderBottom;
