// table/dist/makeConfig.js
"use strict";

Object.defineProperty(exports, "__esModule", ***REMOVED***
  value: true
***REMOVED***);
exports.default = void 0;

var _cloneDeep2 = _interopRequireDefault(require("lodash/cloneDeep"));

var _isUndefined2 = _interopRequireDefault(require("lodash/isUndefined"));

var _times2 = _interopRequireDefault(require("lodash/times"));

var _getBorderCharacters = _interopRequireDefault(require("./getBorderCharacters"));

var _validateConfig = _interopRequireDefault(require("./validateConfig"));

var _calculateMaximumColumnWidthIndex = _interopRequireDefault(require("./calculateMaximumColumnWidthIndex"));

function _interopRequireDefault(obj) ***REMOVED*** return obj && obj.__esModule ? obj : ***REMOVED*** default: obj ***REMOVED***; ***REMOVED***

/**
 * Merges user provided border characters with the default border ("honeywell") characters.
 *
 * @param ***REMOVED***Object***REMOVED*** border
 * @returns ***REMOVED***Object***REMOVED***
 */
const makeBorder = (border = ***REMOVED******REMOVED***) => ***REMOVED***
  return Object.assign(***REMOVED******REMOVED***, (0, _getBorderCharacters.default)('honeywell'), border);
***REMOVED***;
/**
 * Creates a configuration for every column using default
 * values for the missing configuration properties.
 *
 * @param ***REMOVED***Array[]***REMOVED*** rows
 * @param ***REMOVED***Object***REMOVED*** columns
 * @param ***REMOVED***Object***REMOVED*** columnDefault
 * @returns ***REMOVED***Object***REMOVED***
 */


const makeColumns = (rows, columns = ***REMOVED******REMOVED***, columnDefault = ***REMOVED******REMOVED***) => ***REMOVED***
  const maximumColumnWidthIndex = (0, _calculateMaximumColumnWidthIndex.default)(rows);
  (0, _times2.default)(rows[0].length, index => ***REMOVED***
    if ((0, _isUndefined2.default)(columns[index])) ***REMOVED***
      columns[index] = ***REMOVED******REMOVED***;
    ***REMOVED***

    columns[index] = Object.assign(***REMOVED***
      alignment: 'left',
      paddingLeft: 1,
      paddingRight: 1,
      truncate: Infinity,
      width: maximumColumnWidthIndex[index],
      wrapWord: false
    ***REMOVED***, columnDefault, columns[index]);
  ***REMOVED***);
  return columns;
***REMOVED***;
/**
 * Makes a new configuration object out of the userConfig object
 * using default values for the missing configuration properties.
 *
 * @param ***REMOVED***Array[]***REMOVED*** rows
 * @param ***REMOVED***Object***REMOVED*** userConfig
 * @returns ***REMOVED***Object***REMOVED***
 */


const makeConfig = (rows, userConfig = ***REMOVED******REMOVED***) => ***REMOVED***
  (0, _validateConfig.default)('config.json', userConfig);
  const config = (0, _cloneDeep2.default)(userConfig);
  config.border = makeBorder(config.border);
  config.columns = makeColumns(rows, config.columns, config.columnDefault);

  if (!config.drawHorizontalLine) ***REMOVED***
    /**
         * @returns ***REMOVED***boolean***REMOVED***
         */
    config.drawHorizontalLine = () => ***REMOVED***
      return true;
    ***REMOVED***;
  ***REMOVED***

  if (config.singleLine === undefined) ***REMOVED***
    config.singleLine = false;
  ***REMOVED***

  return config;
***REMOVED***;

var _default = makeConfig;
exports.default = _default;
