// tsutils/typeguard/2.8/type.js
"use strict";
Object.defineProperty(exports, "__esModule", ***REMOVED*** value: true ***REMOVED***);
const ts = require("typescript");
function isConditionalType(type) ***REMOVED***
    return (type.flags & ts.TypeFlags.Conditional) !== 0;
***REMOVED***
exports.isConditionalType = isConditionalType;
function isEnumType(type) ***REMOVED***
    return (type.flags & ts.TypeFlags.Enum) !== 0;
***REMOVED***
exports.isEnumType = isEnumType;
function isGenericType(type) ***REMOVED***
    return (type.flags & ts.TypeFlags.Object) !== 0 &&
        (type.objectFlags & ts.ObjectFlags.ClassOrInterface) !== 0 &&
        (type.objectFlags & ts.ObjectFlags.Reference) !== 0;
***REMOVED***
exports.isGenericType = isGenericType;
function isIndexedAccessType(type) ***REMOVED***
    return (type.flags & ts.TypeFlags.IndexedAccess) !== 0;
***REMOVED***
exports.isIndexedAccessType = isIndexedAccessType;
function isIndexedAccessype(type) ***REMOVED***
    return (type.flags & ts.TypeFlags.Index) !== 0;
***REMOVED***
exports.isIndexedAccessype = isIndexedAccessype;
function isInstantiableType(type) ***REMOVED***
    return (type.flags & ts.TypeFlags.Instantiable) !== 0;
***REMOVED***
exports.isInstantiableType = isInstantiableType;
function isInterfaceType(type) ***REMOVED***
    return (type.flags & ts.TypeFlags.Object) !== 0 &&
        (type.objectFlags & ts.ObjectFlags.ClassOrInterface) !== 0;
***REMOVED***
exports.isInterfaceType = isInterfaceType;
function isIntersectionType(type) ***REMOVED***
    return (type.flags & ts.TypeFlags.Intersection) !== 0;
***REMOVED***
exports.isIntersectionType = isIntersectionType;
function isLiteralType(type) ***REMOVED***
    return (type.flags & (ts.TypeFlags.StringOrNumberLiteral | ts.TypeFlags.BigIntLiteral)) !== 0;
***REMOVED***
exports.isLiteralType = isLiteralType;
function isObjectType(type) ***REMOVED***
    return (type.flags & ts.TypeFlags.Object) !== 0;
***REMOVED***
exports.isObjectType = isObjectType;
function isSubstitutionType(type) ***REMOVED***
    return (type.flags & ts.TypeFlags.Substitution) !== 0;
***REMOVED***
exports.isSubstitutionType = isSubstitutionType;
function isTypeParameter(type) ***REMOVED***
    return (type.flags & ts.TypeFlags.TypeParameter) !== 0;
***REMOVED***
exports.isTypeParameter = isTypeParameter;
function isTypeReference(type) ***REMOVED***
    return (type.flags & ts.TypeFlags.Object) !== 0 &&
        (type.objectFlags & ts.ObjectFlags.Reference) !== 0;
***REMOVED***
exports.isTypeReference = isTypeReference;
function isTypeVariable(type) ***REMOVED***
    return (type.flags & ts.TypeFlags.TypeVariable) !== 0;
***REMOVED***
exports.isTypeVariable = isTypeVariable;
function isUnionOrIntersectionType(type) ***REMOVED***
    return (type.flags & ts.TypeFlags.UnionOrIntersection) !== 0;
***REMOVED***
exports.isUnionOrIntersectionType = isUnionOrIntersectionType;
function isUnionType(type) ***REMOVED***
    return (type.flags & ts.TypeFlags.Union) !== 0;
***REMOVED***
exports.isUnionType = isUnionType;
function isUniqueESSymbolType(type) ***REMOVED***
    return (type.flags & ts.TypeFlags.UniqueESSymbol) !== 0;
***REMOVED***
exports.isUniqueESSymbolType = isUniqueESSymbolType;
