// spdx-correct/index.js
/*
Copyright spdx-correct.js contributors

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/
var parse = require('spdx-expression-parse')
var spdxLicenseIds = require('spdx-license-ids')

function valid (string) ***REMOVED***
  try ***REMOVED***
    parse(string)
    return true
  ***REMOVED*** catch (error) ***REMOVED***
    return false
  ***REMOVED***
***REMOVED***

// Common transpositions of license identifier acronyms
var transpositions = [
  ['APGL', 'AGPL'],
  ['Gpl', 'GPL'],
  ['GLP', 'GPL'],
  ['APL', 'Apache'],
  ['ISD', 'ISC'],
  ['GLP', 'GPL'],
  ['IST', 'ISC'],
  ['Claude', 'Clause'],
  [' or later', '+'],
  [' International', ''],
  ['GNU', 'GPL'],
  ['GUN', 'GPL'],
  ['+', ''],
  ['GNU GPL', 'GPL'],
  ['GNU/GPL', 'GPL'],
  ['GNU GLP', 'GPL'],
  ['GNU General Public License', 'GPL'],
  ['Gnu public license', 'GPL'],
  ['GNU Public License', 'GPL'],
  ['GNU GENERAL PUBLIC LICENSE', 'GPL'],
  ['MTI', 'MIT'],
  ['Mozilla Public License', 'MPL'],
  ['WTH', 'WTF'],
  ['-License', '']
]

var TRANSPOSED = 0
var CORRECT = 1

// Simple corrections to nearly valid identifiers.
var transforms = [
  // e.g. 'mit'
  function (argument) ***REMOVED***
    return argument.toUpperCase()
  ***REMOVED***,
  // e.g. 'MIT '
  function (argument) ***REMOVED***
    return argument.trim()
  ***REMOVED***,
  // e.g. 'M.I.T.'
  function (argument) ***REMOVED***
    return argument.replace(/\./g, '')
  ***REMOVED***,
  // e.g. 'Apache- 2.0'
  function (argument) ***REMOVED***
    return argument.replace(/\s+/g, '')
  ***REMOVED***,
  // e.g. 'CC BY 4.0''
  function (argument) ***REMOVED***
    return argument.replace(/\s+/g, '-')
  ***REMOVED***,
  // e.g. 'LGPLv2.1'
  function (argument) ***REMOVED***
    return argument.replace('v', '-')
  ***REMOVED***,
  // e.g. 'Apache 2.0'
  function (argument) ***REMOVED***
    return argument.replace(/,?\s*(\d)/, '-$1')
  ***REMOVED***,
  // e.g. 'GPL 2'
  function (argument) ***REMOVED***
    return argument.replace(/,?\s*(\d)/, '-$1.0')
  ***REMOVED***,
  // e.g. 'Apache Version 2.0'
  function (argument) ***REMOVED***
    return argument
      .replace(/,?\s*(V\.|v\.|V|v|Version|version)\s*(\d)/, '-$2')
  ***REMOVED***,
  // e.g. 'Apache Version 2'
  function (argument) ***REMOVED***
    return argument
      .replace(/,?\s*(V\.|v\.|V|v|Version|version)\s*(\d)/, '-$2.0')
  ***REMOVED***,
  // e.g. 'ZLIB'
  function (argument) ***REMOVED***
    return argument[0].toUpperCase() + argument.slice(1)
  ***REMOVED***,
  // e.g. 'MPL/2.0'
  function (argument) ***REMOVED***
    return argument.replace('/', '-')
  ***REMOVED***,
  // e.g. 'Apache 2'
  function (argument) ***REMOVED***
    return argument
      .replace(/\s*V\s*(\d)/, '-$1')
      .replace(/(\d)$/, '$1.0')
  ***REMOVED***,
  // e.g. 'GPL-2.0', 'GPL-3.0'
  function (argument) ***REMOVED***
    if (argument.indexOf('3.0') !== -1) ***REMOVED***
      return argument + '-or-later'
    ***REMOVED*** else ***REMOVED***
      return argument + '-only'
    ***REMOVED***
  ***REMOVED***,
  // e.g. 'GPL-2.0-'
  function (argument) ***REMOVED***
    return argument + 'only'
  ***REMOVED***,
  // e.g. 'GPL2'
  function (argument) ***REMOVED***
    return argument.replace(/(\d)$/, '-$1.0')
  ***REMOVED***,
  // e.g. 'BSD 3'
  function (argument) ***REMOVED***
    return argument.replace(/(-| )?(\d)$/, '-$2-Clause')
  ***REMOVED***,
  // e.g. 'BSD clause 3'
  function (argument) ***REMOVED***
    return argument.replace(/(-| )clause(-| )(\d)/, '-$3-Clause')
  ***REMOVED***,
  // e.g. 'BY-NC-4.0'
  function (argument) ***REMOVED***
    return 'CC-' + argument
  ***REMOVED***,
  // e.g. 'BY-NC'
  function (argument) ***REMOVED***
    return 'CC-' + argument + '-4.0'
  ***REMOVED***,
  // e.g. 'Attribution-NonCommercial'
  function (argument) ***REMOVED***
    return argument
      .replace('Attribution', 'BY')
      .replace('NonCommercial', 'NC')
      .replace('NoDerivatives', 'ND')
      .replace(/ (\d)/, '-$1')
      .replace(/ ?International/, '')
  ***REMOVED***,
  // e.g. 'Attribution-NonCommercial'
  function (argument) ***REMOVED***
    return 'CC-' +
      argument
        .replace('Attribution', 'BY')
        .replace('NonCommercial', 'NC')
        .replace('NoDerivatives', 'ND')
        .replace(/ (\d)/, '-$1')
        .replace(/ ?International/, '') +
      '-4.0'
  ***REMOVED***
]

var licensesWithVersions = spdxLicenseIds
  .map(function (id) ***REMOVED***
    var match = /^(.*)-\d+\.\d+$/.exec(id)
    return match
      ? [match[0], match[1]]
      : [id, null]
  ***REMOVED***)
  .reduce(function (objectMap, item) ***REMOVED***
    var key = item[1]
    objectMap[key] = objectMap[key] || []
    objectMap[key].push(item[0])
    return objectMap
  ***REMOVED***, ***REMOVED******REMOVED***)

var licensesWithOneVersion = Object.keys(licensesWithVersions)
  .map(function makeEntries (key) ***REMOVED***
    return [key, licensesWithVersions[key]]
  ***REMOVED***)
  .filter(function identifySoleVersions (item) ***REMOVED***
    return (
      // Licenses has just one valid version suffix.
      item[1].length === 1 &&
      item[0] !== null &&
      // APL will be considered Apache, rather than APL-1.0
      item[0] !== 'APL'
    )
  ***REMOVED***)
  .map(function createLastResorts (item) ***REMOVED***
    return [item[0], item[1][0]]
  ***REMOVED***)

licensesWithVersions = undefined

// If all else fails, guess that strings containing certain substrings
// meant to identify certain licenses.
var lastResorts = [
  ['UNLI', 'Unlicense'],
  ['WTF', 'WTFPL'],
  ['2 CLAUSE', 'BSD-2-Clause'],
  ['2-CLAUSE', 'BSD-2-Clause'],
  ['3 CLAUSE', 'BSD-3-Clause'],
  ['3-CLAUSE', 'BSD-3-Clause'],
  ['AFFERO', 'AGPL-3.0-or-later'],
  ['AGPL', 'AGPL-3.0-or-later'],
  ['APACHE', 'Apache-2.0'],
  ['ARTISTIC', 'Artistic-2.0'],
  ['Affero', 'AGPL-3.0-or-later'],
  ['BEER', 'Beerware'],
  ['BOOST', 'BSL-1.0'],
  ['BSD', 'BSD-2-Clause'],
  ['CDDL', 'CDDL-1.1'],
  ['ECLIPSE', 'EPL-1.0'],
  ['FUCK', 'WTFPL'],
  ['GNU', 'GPL-3.0-or-later'],
  ['LGPL', 'LGPL-3.0-or-later'],
  ['GPLV1', 'GPL-1.0-only'],
  ['GPL-1', 'GPL-1.0-only'],
  ['GPLV2', 'GPL-2.0-only'],
  ['GPL-2', 'GPL-2.0-only'],
  ['GPL', 'GPL-3.0-or-later'],
  ['MIT +NO-FALSE-ATTRIBS', 'MITNFA'],
  ['MIT', 'MIT'],
  ['MPL', 'MPL-2.0'],
  ['X11', 'X11'],
  ['ZLIB', 'Zlib']
].concat(licensesWithOneVersion)

var SUBSTRING = 0
var IDENTIFIER = 1

var validTransformation = function (identifier) ***REMOVED***
  for (var i = 0; i < transforms.length; i++) ***REMOVED***
    var transformed = transforms[i](identifier).trim()
    if (transformed !== identifier && valid(transformed)) ***REMOVED***
      return transformed
    ***REMOVED***
  ***REMOVED***
  return null
***REMOVED***

var validLastResort = function (identifier) ***REMOVED***
  var upperCased = identifier.toUpperCase()
  for (var i = 0; i < lastResorts.length; i++) ***REMOVED***
    var lastResort = lastResorts[i]
    if (upperCased.indexOf(lastResort[SUBSTRING]) > -1) ***REMOVED***
      return lastResort[IDENTIFIER]
    ***REMOVED***
  ***REMOVED***
  return null
***REMOVED***

var anyCorrection = function (identifier, check) ***REMOVED***
  for (var i = 0; i < transpositions.length; i++) ***REMOVED***
    var transposition = transpositions[i]
    var transposed = transposition[TRANSPOSED]
    if (identifier.indexOf(transposed) > -1) ***REMOVED***
      var corrected = identifier.replace(
        transposed,
        transposition[CORRECT]
      )
      var checked = check(corrected)
      if (checked !== null) ***REMOVED***
        return checked
      ***REMOVED***
    ***REMOVED***
  ***REMOVED***
  return null
***REMOVED***

module.exports = function (identifier, options) ***REMOVED***
  options = options || ***REMOVED******REMOVED***
  var upgrade = options.upgrade === undefined ? true : !!options.upgrade
  function postprocess (value) ***REMOVED***
    return upgrade ? upgradeGPLs(value) : value
  ***REMOVED***
  var validArugment = (
    typeof identifier === 'string' &&
    identifier.trim().length !== 0
  )
  if (!validArugment) ***REMOVED***
    throw Error('Invalid argument. Expected non-empty string.')
  ***REMOVED***
  identifier = identifier.trim()
  if (valid(identifier)) ***REMOVED***
    return postprocess(identifier)
  ***REMOVED***
  var noPlus = identifier.replace(/\+$/, '').trim()
  if (valid(noPlus)) ***REMOVED***
    return postprocess(noPlus)
  ***REMOVED***
  var transformed = validTransformation(identifier)
  if (transformed !== null) ***REMOVED***
    return postprocess(transformed)
  ***REMOVED***
  transformed = anyCorrection(identifier, function (argument) ***REMOVED***
    if (valid(argument)) ***REMOVED***
      return argument
    ***REMOVED***
    return validTransformation(argument)
  ***REMOVED***)
  if (transformed !== null) ***REMOVED***
    return postprocess(transformed)
  ***REMOVED***
  transformed = validLastResort(identifier)
  if (transformed !== null) ***REMOVED***
    return postprocess(transformed)
  ***REMOVED***
  transformed = anyCorrection(identifier, validLastResort)
  if (transformed !== null) ***REMOVED***
    return postprocess(transformed)
  ***REMOVED***
  return null
***REMOVED***

function upgradeGPLs (value) ***REMOVED***
  if ([
    'GPL-1.0', 'LGPL-1.0', 'AGPL-1.0',
    'GPL-2.0', 'LGPL-2.0', 'AGPL-2.0',
    'LGPL-2.1'
  ].indexOf(value) !== -1) ***REMOVED***
    return value + '-only'
  ***REMOVED*** else if ([
    'GPL-1.0+', 'GPL-2.0+', 'GPL-3.0+',
    'LGPL-2.0+', 'LGPL-2.1+', 'LGPL-3.0+',
    'AGPL-1.0+', 'AGPL-3.0+'
  ].indexOf(value) !== -1) ***REMOVED***
    return value.replace(/\+$/, '-or-later')
  ***REMOVED*** else if (['GPL-3.0', 'LGPL-3.0', 'AGPL-3.0'].indexOf(value) !== -1) ***REMOVED***
    return value + '-or-later'
  ***REMOVED*** else ***REMOVED***
    return value
  ***REMOVED***
***REMOVED***
