// prelude-ls/lib/Str.js
// Generated by LiveScript 1.4.0
var split, join, lines, unlines, words, unwords, chars, unchars, reverse, repeat, capitalize, camelize, dasherize;
split = curry$(function(sep, str)***REMOVED***
  return str.split(sep);
***REMOVED***);
join = curry$(function(sep, xs)***REMOVED***
  return xs.join(sep);
***REMOVED***);
lines = function(str)***REMOVED***
  if (!str.length) ***REMOVED***
    return [];
  ***REMOVED***
  return str.split('\n');
***REMOVED***;
unlines = function(it)***REMOVED***
  return it.join('\n');
***REMOVED***;
words = function(str)***REMOVED***
  if (!str.length) ***REMOVED***
    return [];
  ***REMOVED***
  return str.split(/[ ]+/);
***REMOVED***;
unwords = function(it)***REMOVED***
  return it.join(' ');
***REMOVED***;
chars = function(it)***REMOVED***
  return it.split('');
***REMOVED***;
unchars = function(it)***REMOVED***
  return it.join('');
***REMOVED***;
reverse = function(str)***REMOVED***
  return str.split('').reverse().join('');
***REMOVED***;
repeat = curry$(function(n, str)***REMOVED***
  var result, i$;
  result = '';
  for (i$ = 0; i$ < n; ++i$) ***REMOVED***
    result += str;
  ***REMOVED***
  return result;
***REMOVED***);
capitalize = function(str)***REMOVED***
  return str.charAt(0).toUpperCase() + str.slice(1);
***REMOVED***;
camelize = function(it)***REMOVED***
  return it.replace(/[-_]+(.)?/g, function(arg$, c)***REMOVED***
    return (c != null ? c : '').toUpperCase();
  ***REMOVED***);
***REMOVED***;
dasherize = function(str)***REMOVED***
  return str.replace(/([^-A-Z])([A-Z]+)/g, function(arg$, lower, upper)***REMOVED***
    return lower + "-" + (upper.length > 1
      ? upper
      : upper.toLowerCase());
  ***REMOVED***).replace(/^([A-Z]+)/, function(arg$, upper)***REMOVED***
    if (upper.length > 1) ***REMOVED***
      return upper + "-";
    ***REMOVED*** else ***REMOVED***
      return upper.toLowerCase();
    ***REMOVED***
  ***REMOVED***);
***REMOVED***;
module.exports = ***REMOVED***
  split: split,
  join: join,
  lines: lines,
  unlines: unlines,
  words: words,
  unwords: unwords,
  chars: chars,
  unchars: unchars,
  reverse: reverse,
  repeat: repeat,
  capitalize: capitalize,
  camelize: camelize,
  dasherize: dasherize
***REMOVED***;
function curry$(f, bound)***REMOVED***
  var context,
  _curry = function(args) ***REMOVED***
    return f.length > 1 ? function()***REMOVED***
      var params = args ? args.concat() : [];
      context = bound ? context || this : this;
      return params.push.apply(params, arguments) <
          f.length && arguments.length ?
        _curry.call(context, params) : f.apply(context, params);
    ***REMOVED*** : f;
  ***REMOVED***;
  return _curry();
***REMOVED***
