// tsutils/util/convert-ast.js
"use strict";
Object.defineProperty(exports, "__esModule", ***REMOVED*** value: true ***REMOVED***);
const ts = require("typescript");
const util_1 = require("./util");
function convertAst(sourceFile) ***REMOVED***
    const wrapped = ***REMOVED***
        node: sourceFile,
        parent: undefined,
        kind: ts.SyntaxKind.SourceFile,
        children: [],
        next: undefined,
        skip: undefined,
    ***REMOVED***;
    const flat = [];
    let current = wrapped;
    let previous = current;
    ts.forEachChild(sourceFile, function wrap(node) ***REMOVED***
        flat.push(node);
        const parent = current;
        previous.next = current = ***REMOVED***
            node,
            parent,
            kind: node.kind,
            children: [],
            next: undefined,
            skip: undefined,
        ***REMOVED***;
        if (previous !== parent)
            setSkip(previous, current);
        previous = current;
        parent.children.push(current);
        if (util_1.isNodeKind(node.kind))
            ts.forEachChild(node, wrap);
        current = parent;
    ***REMOVED***);
    return ***REMOVED***
        wrapped,
        flat,
    ***REMOVED***;
***REMOVED***
exports.convertAst = convertAst;
function setSkip(node, skip) ***REMOVED***
    do ***REMOVED***
        node.skip = skip;
        node = node.parent;
    ***REMOVED*** while (node !== skip.parent);
***REMOVED***
