// tsutils/util/type.js
"use strict";
Object.defineProperty(exports, "__esModule", ***REMOVED*** value: true ***REMOVED***);
const ts = require("typescript");
const type_1 = require("../typeguard/type");
const util_1 = require("./util");
const node_1 = require("../typeguard/node");
function isEmptyObjectType(type) ***REMOVED***
    if (type_1.isObjectType(type) &&
        type.objectFlags & ts.ObjectFlags.Anonymous &&
        type.getProperties().length === 0 &&
        type.getCallSignatures().length === 0 &&
        type.getConstructSignatures().length === 0 &&
        type.getStringIndexType() === undefined &&
        type.getNumberIndexType() === undefined) ***REMOVED***
        const baseTypes = type.getBaseTypes();
        return baseTypes === undefined || baseTypes.every(isEmptyObjectType);
    ***REMOVED***
    return false;
***REMOVED***
exports.isEmptyObjectType = isEmptyObjectType;
function removeOptionalityFromType(checker, type) ***REMOVED***
    if (!containsTypeWithFlag(type, ts.TypeFlags.Undefined))
        return type;
    const allowsNull = containsTypeWithFlag(type, ts.TypeFlags.Null);
    type = checker.getNonNullableType(type);
    return allowsNull ? checker.getNullableType(type, ts.TypeFlags.Null) : type;
***REMOVED***
exports.removeOptionalityFromType = removeOptionalityFromType;
function containsTypeWithFlag(type, flag) ***REMOVED***
    for (const t of unionTypeParts(type))
        if (util_1.isTypeFlagSet(t, flag))
            return true;
    return false;
***REMOVED***
function isTypeAssignableToNumber(checker, type) ***REMOVED***
    return isTypeAssignableTo(checker, type, ts.TypeFlags.NumberLike);
***REMOVED***
exports.isTypeAssignableToNumber = isTypeAssignableToNumber;
function isTypeAssignableToString(checker, type) ***REMOVED***
    return isTypeAssignableTo(checker, type, ts.TypeFlags.StringLike);
***REMOVED***
exports.isTypeAssignableToString = isTypeAssignableToString;
function isTypeAssignableTo(checker, type, flags) ***REMOVED***
    flags |= ts.TypeFlags.Any;
    let typeParametersSeen;
    return (function check(t) ***REMOVED***
        if (type_1.isTypeParameter(t) && t.symbol !== undefined && t.symbol.declarations !== undefined) ***REMOVED***
            if (typeParametersSeen === undefined) ***REMOVED***
                typeParametersSeen = new Set([t]);
            ***REMOVED***
            else if (!typeParametersSeen.has(t)) ***REMOVED***
                typeParametersSeen.add(t);
            ***REMOVED***
            else ***REMOVED***
                return false;
            ***REMOVED***
            const declaration = t.symbol.declarations[0];
            if (declaration.constraint === undefined)
                return true;
            return check(checker.getTypeFromTypeNode(declaration.constraint));
        ***REMOVED***
        if (type_1.isUnionType(t))
            return t.types.every(check);
        if (type_1.isIntersectionType(t))
            return t.types.some(check);
        return util_1.isTypeFlagSet(t, flags);
    ***REMOVED***)(type);
***REMOVED***
function getCallSignaturesOfType(type) ***REMOVED***
    if (type_1.isUnionType(type)) ***REMOVED***
        const signatures = [];
        for (const t of type.types)
            signatures.push(...getCallSignaturesOfType(t));
        return signatures;
    ***REMOVED***
    if (type_1.isIntersectionType(type)) ***REMOVED***
        let signatures;
        for (const t of type.types) ***REMOVED***
            const sig = getCallSignaturesOfType(t);
            if (sig.length !== 0) ***REMOVED***
                if (signatures !== undefined)
                    return [];
                signatures = sig;
            ***REMOVED***
        ***REMOVED***
        return signatures === undefined ? [] : signatures;
    ***REMOVED***
    return type.getCallSignatures();
***REMOVED***
exports.getCallSignaturesOfType = getCallSignaturesOfType;
function unionTypeParts(type) ***REMOVED***
    return type_1.isUnionType(type) ? type.types : [type];
***REMOVED***
exports.unionTypeParts = unionTypeParts;
function intersectionTypeParts(type) ***REMOVED***
    return type_1.isIntersectionType(type) ? type.types : [type];
***REMOVED***
exports.intersectionTypeParts = intersectionTypeParts;
function someTypePart(type, predicate, cb) ***REMOVED***
    return predicate(type) ? type.types.some(cb) : cb(type);
***REMOVED***
exports.someTypePart = someTypePart;
function isThenableType(checker, node, type = checker.getTypeAtLocation(node)) ***REMOVED***
    for (const ty of unionTypeParts(checker.getApparentType(type))) ***REMOVED***
        const then = ty.getProperty('then');
        if (then === undefined)
            continue;
        const thenType = checker.getTypeOfSymbolAtLocation(then, node);
        for (const t of unionTypeParts(thenType))
            for (const signature of t.getCallSignatures())
                if (signature.parameters.length !== 0 && isCallback(checker, signature.parameters[0], node))
                    return true;
    ***REMOVED***
    return false;
***REMOVED***
exports.isThenableType = isThenableType;
function isCallback(checker, param, node) ***REMOVED***
    let type = checker.getApparentType(checker.getTypeOfSymbolAtLocation(param, node));
    if (param.valueDeclaration.dotDotDotToken) ***REMOVED***
        type = type.getNumberIndexType();
        if (type === undefined)
            return false;
    ***REMOVED***
    for (const t of unionTypeParts(type))
        if (t.getCallSignatures().length !== 0)
            return true;
    return false;
***REMOVED***
function isFalsyType(type) ***REMOVED***
    if (type.flags & (ts.TypeFlags.Undefined | ts.TypeFlags.Null | ts.TypeFlags.Void))
        return true;
    if (type_1.isLiteralType(type))
        return !type.value;
    return isBooleanLiteralType(type, false);
***REMOVED***
exports.isFalsyType = isFalsyType;
function isBooleanLiteralType(type, literal) ***REMOVED***
    return util_1.isTypeFlagSet(type, ts.TypeFlags.BooleanLiteral) &&
        type.intrinsicName === (literal ? 'true' : 'false');
***REMOVED***
exports.isBooleanLiteralType = isBooleanLiteralType;
function getPropertyOfType(type, name) ***REMOVED***
    if (!name.startsWith('__'))
        return type.getProperty(name);
    return type.getProperties().find((s) => s.escapedName === name);
***REMOVED***
exports.getPropertyOfType = getPropertyOfType;
function isPropertyReadonlyInType(type, name, checker) ***REMOVED***
    let seenProperty = false;
    let seenReadonlySignature = false;
    for (const t of unionTypeParts(type)) ***REMOVED***
        if (getPropertyOfType(t, name) === undefined) ***REMOVED***
            const index = (util_1.isNumericPropertyName(name) ? checker.getIndexInfoOfType(t, ts.IndexKind.Number) : undefined) ||
                checker.getIndexInfoOfType(t, ts.IndexKind.String);
            if (index !== undefined && index.isReadonly) ***REMOVED***
                if (seenProperty)
                    return true;
                seenReadonlySignature = true;
            ***REMOVED***
        ***REMOVED***
        else if (seenReadonlySignature || isReadonlyPropertyIntersection(t, name, checker)) ***REMOVED***
            return true;
        ***REMOVED***
        else ***REMOVED***
            seenProperty = true;
        ***REMOVED***
    ***REMOVED***
    return false;
***REMOVED***
exports.isPropertyReadonlyInType = isPropertyReadonlyInType;
function isReadonlyPropertyIntersection(type, name, checker) ***REMOVED***
    return someTypePart(type, type_1.isIntersectionType, (t) => ***REMOVED***
        const prop = getPropertyOfType(t, name);
        if (prop === undefined)
            return false;
        if (prop.flags & ts.SymbolFlags.Transient) ***REMOVED***
            if (/^(?:[1-9]\d*|0)$/.test(name) && type_1.isTupleTypeReference(t))
                return t.target.readonly;
            switch (isReadonlyPropertyFromMappedType(t, name, checker)) ***REMOVED***
                case true:
                    return true;
                case false:
                    return false;
                default:
            ***REMOVED***
        ***REMOVED***
        return (util_1.isSymbolFlagSet(prop, ts.SymbolFlags.ValueModule) ||
            symbolHasReadonlyDeclaration(prop, checker));
    ***REMOVED***);
***REMOVED***
function isReadonlyPropertyFromMappedType(type, name, checker) ***REMOVED***
    if (!type_1.isObjectType(type) || !util_1.isObjectFlagSet(type, ts.ObjectFlags.Mapped))
        return;
    const declaration = type.symbol.declarations[0];
    if (declaration.readonlyToken !== undefined && !/^__@[^@]+$/.test(name))
        return declaration.readonlyToken.kind !== ts.SyntaxKind.MinusToken;
    return isPropertyReadonlyInType(type.modifiersType, name, checker);
***REMOVED***
function symbolHasReadonlyDeclaration(symbol, checker) ***REMOVED***
    return (symbol.flags & ts.SymbolFlags.Accessor) === ts.SymbolFlags.GetAccessor ||
        symbol.declarations !== undefined &&
            symbol.declarations.some((node) => util_1.isModifierFlagSet(node, ts.ModifierFlags.Readonly) ||
                node_1.isVariableDeclaration(node) && util_1.isNodeFlagSet(node.parent, ts.NodeFlags.Const) ||
                node_1.isCallExpression(node) && util_1.isReadonlyAssignmentDeclaration(node, checker) ||
                node_1.isEnumMember(node) ||
                (node_1.isPropertyAssignment(node) || node_1.isShorthandPropertyAssignment(node)) && util_1.isInConstContext(node.parent));
***REMOVED***
exports.symbolHasReadonlyDeclaration = symbolHasReadonlyDeclaration;
function getPropertyNameFromType(type) ***REMOVED***
    if (type.flags & (ts.TypeFlags.StringLiteral | ts.TypeFlags.NumberLiteral)) ***REMOVED***
        const value = String(type.value);
        return ***REMOVED*** displayName: value, symbolName: ts.escapeLeadingUnderscores(value) ***REMOVED***;
    ***REMOVED***
    if (type_1.isUniqueESSymbolType(type))
        return ***REMOVED***
            displayName: `[$***REMOVED***type.symbol ? type.symbol.name : type.escapedName.replace(/^__@|@\d+$/g, '')***REMOVED***]`,
            symbolName: type.escapedName,
        ***REMOVED***;
***REMOVED***
exports.getPropertyNameFromType = getPropertyNameFromType;
function getConstructorTypeOfClassLikeDeclaration(node, checker) ***REMOVED***
    return checker.getDeclaredTypeOfSymbol(node.name !== undefined ? checker.getSymbolAtLocation(node.name) : checker.getTypeAtLocation(node).symbol);
***REMOVED***
exports.getConstructorTypeOfClassLikeDeclaration = getConstructorTypeOfClassLikeDeclaration;
function getInstanceTypeOfClassLikeDeclaration(node, checker) ***REMOVED***
    return node.kind === ts.SyntaxKind.ClassDeclaration
        ? checker.getTypeAtLocation(node)
        : checker.getTypeOfSymbolAtLocation(checker.getTypeAtLocation(node).getProperty('prototype'), node);
***REMOVED***
exports.getInstanceTypeOfClassLikeDeclaration = getInstanceTypeOfClassLikeDeclaration;
function getIteratorYieldResultFromIteratorResult(type, node, checker) ***REMOVED***
    return type_1.isUnionType(type) && type.types.find((t) => ***REMOVED***
        const done = t.getProperty('done');
        return done !== undefined &&
            isBooleanLiteralType(removeOptionalityFromType(checker, checker.getTypeOfSymbolAtLocation(done, node)), false);
    ***REMOVED***) || type;
***REMOVED***
exports.getIteratorYieldResultFromIteratorResult = getIteratorYieldResultFromIteratorResult;
