// tsutils/util/util.js
"use strict";
Object.defineProperty(exports, "__esModule", ***REMOVED*** value: true ***REMOVED***);
const ts = require("typescript");
const node_1 = require("../typeguard/node");
const _3_2_1 = require("../typeguard/3.2");
const type_1 = require("./type");
function getChildOfKind(node, kind, sourceFile) ***REMOVED***
    for (const child of node.getChildren(sourceFile))
        if (child.kind === kind)
            return child;
***REMOVED***
exports.getChildOfKind = getChildOfKind;
function isTokenKind(kind) ***REMOVED***
    return kind >= ts.SyntaxKind.FirstToken && kind <= ts.SyntaxKind.LastToken;
***REMOVED***
exports.isTokenKind = isTokenKind;
function isNodeKind(kind) ***REMOVED***
    return kind >= ts.SyntaxKind.FirstNode;
***REMOVED***
exports.isNodeKind = isNodeKind;
function isAssignmentKind(kind) ***REMOVED***
    return kind >= ts.SyntaxKind.FirstAssignment && kind <= ts.SyntaxKind.LastAssignment;
***REMOVED***
exports.isAssignmentKind = isAssignmentKind;
function isTypeNodeKind(kind) ***REMOVED***
    return kind >= ts.SyntaxKind.FirstTypeNode && kind <= ts.SyntaxKind.LastTypeNode;
***REMOVED***
exports.isTypeNodeKind = isTypeNodeKind;
function isJsDocKind(kind) ***REMOVED***
    return kind >= ts.SyntaxKind.FirstJSDocNode && kind <= ts.SyntaxKind.LastJSDocNode;
***REMOVED***
exports.isJsDocKind = isJsDocKind;
function isKeywordKind(kind) ***REMOVED***
    return kind >= ts.SyntaxKind.FirstKeyword && kind <= ts.SyntaxKind.LastKeyword;
***REMOVED***
exports.isKeywordKind = isKeywordKind;
function isThisParameter(parameter) ***REMOVED***
    return parameter.name.kind === ts.SyntaxKind.Identifier && parameter.name.originalKeywordKind === ts.SyntaxKind.ThisKeyword;
***REMOVED***
exports.isThisParameter = isThisParameter;
function getModifier(node, kind) ***REMOVED***
    if (node.modifiers !== undefined)
        for (const modifier of node.modifiers)
            if (modifier.kind === kind)
                return modifier;
***REMOVED***
exports.getModifier = getModifier;
function hasModifier(modifiers, ...kinds) ***REMOVED***
    if (modifiers === undefined)
        return false;
    for (const modifier of modifiers)
        if (kinds.includes(modifier.kind))
            return true;
    return false;
***REMOVED***
exports.hasModifier = hasModifier;
function isParameterProperty(node) ***REMOVED***
    return hasModifier(node.modifiers, ts.SyntaxKind.PublicKeyword, ts.SyntaxKind.ProtectedKeyword, ts.SyntaxKind.PrivateKeyword, ts.SyntaxKind.ReadonlyKeyword);
***REMOVED***
exports.isParameterProperty = isParameterProperty;
function hasAccessModifier(node) ***REMOVED***
    return hasModifier(node.modifiers, ts.SyntaxKind.PublicKeyword, ts.SyntaxKind.ProtectedKeyword, ts.SyntaxKind.PrivateKeyword);
***REMOVED***
exports.hasAccessModifier = hasAccessModifier;
function isFlagSet(obj, flag) ***REMOVED***
    return (obj.flags & flag) !== 0;
***REMOVED***
exports.isNodeFlagSet = isFlagSet;
exports.isTypeFlagSet = isFlagSet;
exports.isSymbolFlagSet = isFlagSet;
function isObjectFlagSet(objectType, flag) ***REMOVED***
    return (objectType.objectFlags & flag) !== 0;
***REMOVED***
exports.isObjectFlagSet = isObjectFlagSet;
function isModifierFlagSet(node, flag) ***REMOVED***
    return (ts.getCombinedModifierFlags(node) & flag) !== 0;
***REMOVED***
exports.isModifierFlagSet = isModifierFlagSet;
function getPreviousStatement(statement) ***REMOVED***
    const parent = statement.parent;
    if (node_1.isBlockLike(parent)) ***REMOVED***
        const index = parent.statements.indexOf(statement);
        if (index > 0)
            return parent.statements[index - 1];
    ***REMOVED***
***REMOVED***
exports.getPreviousStatement = getPreviousStatement;
function getNextStatement(statement) ***REMOVED***
    const parent = statement.parent;
    if (node_1.isBlockLike(parent)) ***REMOVED***
        const index = parent.statements.indexOf(statement);
        if (index < parent.statements.length)
            return parent.statements[index + 1];
    ***REMOVED***
***REMOVED***
exports.getNextStatement = getNextStatement;
function getPreviousToken(node, sourceFile) ***REMOVED***
    let parent = node.parent;
    while (parent !== undefined && parent.pos === node.pos)
        parent = parent.parent;
    if (parent === undefined)
        return;
    outer: while (true) ***REMOVED***
        const children = parent.getChildren(sourceFile);
        for (let i = children.length - 1; i >= 0; --i) ***REMOVED***
            const child = children[i];
            if (child.pos < node.pos && child.kind !== ts.SyntaxKind.JSDocComment) ***REMOVED***
                if (isTokenKind(child.kind))
                    return child;
                parent = child;
                continue outer;
            ***REMOVED***
        ***REMOVED***
        return;
    ***REMOVED***
***REMOVED***
exports.getPreviousToken = getPreviousToken;
function getNextToken(node, sourceFile = node.getSourceFile()) ***REMOVED***
    if (node.kind === ts.SyntaxKind.SourceFile || node.kind === ts.SyntaxKind.EndOfFileToken)
        return;
    const end = node.end;
    node = node.parent;
    while (node.end === end) ***REMOVED***
        if (node.parent === undefined)
            return node.endOfFileToken;
        node = node.parent;
    ***REMOVED***
    return getTokenAtPositionWorker(node, end, sourceFile, false);
***REMOVED***
exports.getNextToken = getNextToken;
function getTokenAtPosition(parent, pos, sourceFile, allowJsDoc) ***REMOVED***
    if (pos < parent.pos || pos >= parent.end)
        return;
    if (isTokenKind(parent.kind))
        return parent;
    if (sourceFile === undefined)
        sourceFile = parent.getSourceFile();
    return getTokenAtPositionWorker(parent, pos, sourceFile, allowJsDoc === true);
***REMOVED***
exports.getTokenAtPosition = getTokenAtPosition;
function getTokenAtPositionWorker(node, pos, sourceFile, allowJsDoc) ***REMOVED***
    outer: while (true) ***REMOVED***
        for (const child of node.getChildren(sourceFile)) ***REMOVED***
            if (child.end > pos && (allowJsDoc || child.kind !== ts.SyntaxKind.JSDocComment)) ***REMOVED***
                if (isTokenKind(child.kind))
                    return child;
                node = child;
                continue outer;
            ***REMOVED***
        ***REMOVED***
        return;
    ***REMOVED***
***REMOVED***
function getCommentAtPosition(sourceFile, pos, parent = sourceFile) ***REMOVED***
    const token = getTokenAtPosition(parent, pos, sourceFile);
    if (token === undefined || token.kind === ts.SyntaxKind.JsxText || pos >= token.end - (ts.tokenToString(token.kind) || '').length)
        return;
    const startPos = token.pos === 0
        ? (ts.getShebang(sourceFile.text) || '').length
        : token.pos;
    return startPos !== 0 && ts.forEachTrailingCommentRange(sourceFile.text, startPos, commentAtPositionCallback, pos) ||
        ts.forEachLeadingCommentRange(sourceFile.text, startPos, commentAtPositionCallback, pos);
***REMOVED***
exports.getCommentAtPosition = getCommentAtPosition;
function commentAtPositionCallback(pos, end, kind, _nl, at) ***REMOVED***
    return at >= pos && at < end ? ***REMOVED*** pos, end, kind ***REMOVED*** : undefined;
***REMOVED***
function isPositionInComment(sourceFile, pos, parent) ***REMOVED***
    return getCommentAtPosition(sourceFile, pos, parent) !== undefined;
***REMOVED***
exports.isPositionInComment = isPositionInComment;
function commentText(sourceText, comment) ***REMOVED***
    return sourceText.substring(comment.pos + 2, comment.kind === ts.SyntaxKind.SingleLineCommentTrivia ? comment.end : comment.end - 2);
***REMOVED***
exports.commentText = commentText;
function getWrappedNodeAtPosition(wrap, pos) ***REMOVED***
    if (wrap.node.pos > pos || wrap.node.end <= pos)
        return;
    outer: while (true) ***REMOVED***
        for (const child of wrap.children) ***REMOVED***
            if (child.node.pos > pos)
                return wrap;
            if (child.node.end > pos) ***REMOVED***
                wrap = child;
                continue outer;
            ***REMOVED***
        ***REMOVED***
        return wrap;
    ***REMOVED***
***REMOVED***
exports.getWrappedNodeAtPosition = getWrappedNodeAtPosition;
function getPropertyName(propertyName) ***REMOVED***
    if (propertyName.kind === ts.SyntaxKind.ComputedPropertyName) ***REMOVED***
        if (!node_1.isLiteralExpression(propertyName.expression))
            return;
        if (_3_2_1.isBigIntLiteral(propertyName.expression))
            return propertyName.expression.text.slice(0, -1);
        return propertyName.expression.text;
    ***REMOVED***
    return propertyName.text;
***REMOVED***
exports.getPropertyName = getPropertyName;
function forEachDestructuringIdentifier(pattern, fn) ***REMOVED***
    for (const element of pattern.elements) ***REMOVED***
        if (element.kind !== ts.SyntaxKind.BindingElement)
            continue;
        let result;
        if (element.name.kind === ts.SyntaxKind.Identifier) ***REMOVED***
            result = fn(element);
        ***REMOVED***
        else ***REMOVED***
            result = forEachDestructuringIdentifier(element.name, fn);
        ***REMOVED***
        if (result)
            return result;
    ***REMOVED***
***REMOVED***
exports.forEachDestructuringIdentifier = forEachDestructuringIdentifier;
function forEachDeclaredVariable(declarationList, cb) ***REMOVED***
    for (const declaration of declarationList.declarations) ***REMOVED***
        let result;
        if (declaration.name.kind === ts.SyntaxKind.Identifier) ***REMOVED***
            result = cb(declaration);
        ***REMOVED***
        else ***REMOVED***
            result = forEachDestructuringIdentifier(declaration.name, cb);
        ***REMOVED***
        if (result)
            return result;
    ***REMOVED***
***REMOVED***
exports.forEachDeclaredVariable = forEachDeclaredVariable;
var VariableDeclarationKind;
(function (VariableDeclarationKind) ***REMOVED***
    VariableDeclarationKind[VariableDeclarationKind["Var"] = 0] = "Var";
    VariableDeclarationKind[VariableDeclarationKind["Let"] = 1] = "Let";
    VariableDeclarationKind[VariableDeclarationKind["Const"] = 2] = "Const";
***REMOVED***)(VariableDeclarationKind = exports.VariableDeclarationKind || (exports.VariableDeclarationKind = ***REMOVED******REMOVED***));
function getVariableDeclarationKind(declarationList) ***REMOVED***
    if (declarationList.flags & ts.NodeFlags.Let)
        return 1;
    if (declarationList.flags & ts.NodeFlags.Const)
        return 2;
    return 0;
***REMOVED***
exports.getVariableDeclarationKind = getVariableDeclarationKind;
function isBlockScopedVariableDeclarationList(declarationList) ***REMOVED***
    return (declarationList.flags & ts.NodeFlags.BlockScoped) !== 0;
***REMOVED***
exports.isBlockScopedVariableDeclarationList = isBlockScopedVariableDeclarationList;
function isBlockScopedVariableDeclaration(declaration) ***REMOVED***
    const parent = declaration.parent;
    return parent.kind === ts.SyntaxKind.CatchClause ||
        isBlockScopedVariableDeclarationList(parent);
***REMOVED***
exports.isBlockScopedVariableDeclaration = isBlockScopedVariableDeclaration;
function isBlockScopedDeclarationStatement(statement) ***REMOVED***
    switch (statement.kind) ***REMOVED***
        case ts.SyntaxKind.VariableStatement:
            return isBlockScopedVariableDeclarationList(statement.declarationList);
        case ts.SyntaxKind.ClassDeclaration:
        case ts.SyntaxKind.EnumDeclaration:
        case ts.SyntaxKind.InterfaceDeclaration:
        case ts.SyntaxKind.TypeAliasDeclaration:
            return true;
        default:
            return false;
    ***REMOVED***
***REMOVED***
exports.isBlockScopedDeclarationStatement = isBlockScopedDeclarationStatement;
function isInSingleStatementContext(statement) ***REMOVED***
    switch (statement.parent.kind) ***REMOVED***
        case ts.SyntaxKind.ForStatement:
        case ts.SyntaxKind.ForInStatement:
        case ts.SyntaxKind.ForOfStatement:
        case ts.SyntaxKind.WhileStatement:
        case ts.SyntaxKind.DoStatement:
        case ts.SyntaxKind.IfStatement:
        case ts.SyntaxKind.WithStatement:
        case ts.SyntaxKind.LabeledStatement:
            return true;
        default:
            return false;
    ***REMOVED***
***REMOVED***
exports.isInSingleStatementContext = isInSingleStatementContext;
var ScopeBoundary;
(function (ScopeBoundary) ***REMOVED***
    ScopeBoundary[ScopeBoundary["None"] = 0] = "None";
    ScopeBoundary[ScopeBoundary["Function"] = 1] = "Function";
    ScopeBoundary[ScopeBoundary["Block"] = 2] = "Block";
    ScopeBoundary[ScopeBoundary["Type"] = 4] = "Type";
    ScopeBoundary[ScopeBoundary["ConditionalType"] = 8] = "ConditionalType";
***REMOVED***)(ScopeBoundary = exports.ScopeBoundary || (exports.ScopeBoundary = ***REMOVED******REMOVED***));
var ScopeBoundarySelector;
(function (ScopeBoundarySelector) ***REMOVED***
    ScopeBoundarySelector[ScopeBoundarySelector["Function"] = 1] = "Function";
    ScopeBoundarySelector[ScopeBoundarySelector["Block"] = 3] = "Block";
    ScopeBoundarySelector[ScopeBoundarySelector["Type"] = 7] = "Type";
    ScopeBoundarySelector[ScopeBoundarySelector["InferType"] = 8] = "InferType";
***REMOVED***)(ScopeBoundarySelector = exports.ScopeBoundarySelector || (exports.ScopeBoundarySelector = ***REMOVED******REMOVED***));
function isScopeBoundary(node) ***REMOVED***
    return isFunctionScopeBoundary(node) || isBlockScopeBoundary(node) || isTypeScopeBoundary(node);
***REMOVED***
exports.isScopeBoundary = isScopeBoundary;
function isTypeScopeBoundary(node) ***REMOVED***
    switch (node.kind) ***REMOVED***
        case ts.SyntaxKind.InterfaceDeclaration:
        case ts.SyntaxKind.TypeAliasDeclaration:
        case ts.SyntaxKind.MappedType:
            return 4;
        case ts.SyntaxKind.ConditionalType:
            return 8;
        default:
            return 0;
    ***REMOVED***
***REMOVED***
exports.isTypeScopeBoundary = isTypeScopeBoundary;
function isFunctionScopeBoundary(node) ***REMOVED***
    switch (node.kind) ***REMOVED***
        case ts.SyntaxKind.FunctionExpression:
        case ts.SyntaxKind.ArrowFunction:
        case ts.SyntaxKind.Constructor:
        case ts.SyntaxKind.ModuleDeclaration:
        case ts.SyntaxKind.ClassDeclaration:
        case ts.SyntaxKind.ClassExpression:
        case ts.SyntaxKind.EnumDeclaration:
        case ts.SyntaxKind.MethodDeclaration:
        case ts.SyntaxKind.FunctionDeclaration:
        case ts.SyntaxKind.GetAccessor:
        case ts.SyntaxKind.SetAccessor:
        case ts.SyntaxKind.MethodSignature:
        case ts.SyntaxKind.CallSignature:
        case ts.SyntaxKind.ConstructSignature:
        case ts.SyntaxKind.ConstructorType:
        case ts.SyntaxKind.FunctionType:
            return 1;
        case ts.SyntaxKind.SourceFile:
            return ts.isExternalModule(node) ? 1 : 0;
        default:
            return 0;
    ***REMOVED***
***REMOVED***
exports.isFunctionScopeBoundary = isFunctionScopeBoundary;
function isBlockScopeBoundary(node) ***REMOVED***
    switch (node.kind) ***REMOVED***
        case ts.SyntaxKind.Block:
            const parent = node.parent;
            return parent.kind !== ts.SyntaxKind.CatchClause &&
                (parent.kind === ts.SyntaxKind.SourceFile ||
                    !isFunctionScopeBoundary(parent))
                ? 2
                : 0;
        case ts.SyntaxKind.ForStatement:
        case ts.SyntaxKind.ForInStatement:
        case ts.SyntaxKind.ForOfStatement:
        case ts.SyntaxKind.CaseBlock:
        case ts.SyntaxKind.CatchClause:
        case ts.SyntaxKind.WithStatement:
            return 2;
        default:
            return 0;
    ***REMOVED***
***REMOVED***
exports.isBlockScopeBoundary = isBlockScopeBoundary;
function hasOwnThisReference(node) ***REMOVED***
    switch (node.kind) ***REMOVED***
        case ts.SyntaxKind.ClassDeclaration:
        case ts.SyntaxKind.ClassExpression:
        case ts.SyntaxKind.FunctionExpression:
            return true;
        case ts.SyntaxKind.FunctionDeclaration:
            return node.body !== undefined;
        case ts.SyntaxKind.MethodDeclaration:
        case ts.SyntaxKind.GetAccessor:
        case ts.SyntaxKind.SetAccessor:
            return node.parent.kind === ts.SyntaxKind.ObjectLiteralExpression;
        default:
            return false;
    ***REMOVED***
***REMOVED***
exports.hasOwnThisReference = hasOwnThisReference;
function isFunctionWithBody(node) ***REMOVED***
    switch (node.kind) ***REMOVED***
        case ts.SyntaxKind.GetAccessor:
        case ts.SyntaxKind.SetAccessor:
        case ts.SyntaxKind.FunctionDeclaration:
        case ts.SyntaxKind.MethodDeclaration:
        case ts.SyntaxKind.Constructor:
            return node.body !== undefined;
        case ts.SyntaxKind.FunctionExpression:
        case ts.SyntaxKind.ArrowFunction:
            return true;
        default:
            return false;
    ***REMOVED***
***REMOVED***
exports.isFunctionWithBody = isFunctionWithBody;
function forEachToken(node, cb, sourceFile = node.getSourceFile()) ***REMOVED***
    return (function iterate(child) ***REMOVED***
        if (isTokenKind(child.kind))
            return cb(child);
        if (child.kind !== ts.SyntaxKind.JSDocComment)
            return child.getChildren(sourceFile).forEach(iterate);
    ***REMOVED***)(node);
***REMOVED***
exports.forEachToken = forEachToken;
function forEachTokenWithTrivia(node, cb, sourceFile = node.getSourceFile()) ***REMOVED***
    const fullText = sourceFile.text;
    const scanner = ts.createScanner(sourceFile.languageVersion, false, sourceFile.languageVariant, fullText);
    return forEachToken(node, (token) => ***REMOVED***
        const tokenStart = token.kind === ts.SyntaxKind.JsxText || token.pos === token.end ? token.pos : token.getStart(sourceFile);
        if (tokenStart !== token.pos) ***REMOVED***
            scanner.setTextPos(token.pos);
            let kind = scanner.scan();
            let pos = scanner.getTokenPos();
            while (pos < tokenStart) ***REMOVED***
                const textPos = scanner.getTextPos();
                cb(fullText, kind, ***REMOVED*** pos, end: textPos ***REMOVED***, token.parent);
                if (textPos === tokenStart)
                    break;
                kind = scanner.scan();
                pos = scanner.getTokenPos();
            ***REMOVED***
        ***REMOVED***
        return cb(fullText, token.kind, ***REMOVED*** end: token.end, pos: tokenStart ***REMOVED***, token.parent);
    ***REMOVED***, sourceFile);
***REMOVED***
exports.forEachTokenWithTrivia = forEachTokenWithTrivia;
function forEachComment(node, cb, sourceFile = node.getSourceFile()) ***REMOVED***
    const fullText = sourceFile.text;
    const notJsx = sourceFile.languageVariant !== ts.LanguageVariant.JSX;
    return forEachToken(node, (token) => ***REMOVED***
        if (token.pos === token.end)
            return;
        if (token.kind !== ts.SyntaxKind.JsxText)
            ts.forEachLeadingCommentRange(fullText, token.pos === 0 ? (ts.getShebang(fullText) || '').length : token.pos, commentCallback);
        if (notJsx || canHaveTrailingTrivia(token))
            return ts.forEachTrailingCommentRange(fullText, token.end, commentCallback);
    ***REMOVED***, sourceFile);
    function commentCallback(pos, end, kind) ***REMOVED***
        cb(fullText, ***REMOVED*** pos, end, kind ***REMOVED***);
    ***REMOVED***
***REMOVED***
exports.forEachComment = forEachComment;
function canHaveTrailingTrivia(token) ***REMOVED***
    switch (token.kind) ***REMOVED***
        case ts.SyntaxKind.CloseBraceToken:
            return token.parent.kind !== ts.SyntaxKind.JsxExpression || !isJsxElementOrFragment(token.parent.parent);
        case ts.SyntaxKind.GreaterThanToken:
            switch (token.parent.kind) ***REMOVED***
                case ts.SyntaxKind.JsxOpeningElement:
                    return token.end !== token.parent.end;
                case ts.SyntaxKind.JsxOpeningFragment:
                    return false;
                case ts.SyntaxKind.JsxSelfClosingElement:
                    return token.end !== token.parent.end ||
                        !isJsxElementOrFragment(token.parent.parent);
                case ts.SyntaxKind.JsxClosingElement:
                case ts.SyntaxKind.JsxClosingFragment:
                    return !isJsxElementOrFragment(token.parent.parent.parent);
            ***REMOVED***
    ***REMOVED***
    return true;
***REMOVED***
function isJsxElementOrFragment(node) ***REMOVED***
    return node.kind === ts.SyntaxKind.JsxElement || node.kind === ts.SyntaxKind.JsxFragment;
***REMOVED***
function getLineRanges(sourceFile) ***REMOVED***
    const lineStarts = sourceFile.getLineStarts();
    const result = [];
    const length = lineStarts.length;
    const sourceText = sourceFile.text;
    let pos = 0;
    for (let i = 1; i < length; ++i) ***REMOVED***
        const end = lineStarts[i];
        let lineEnd = end;
        for (; lineEnd > pos; --lineEnd)
            if (!ts.isLineBreak(sourceText.charCodeAt(lineEnd - 1)))
                break;
        result.push(***REMOVED***
            pos,
            end,
            contentLength: lineEnd - pos,
        ***REMOVED***);
        pos = end;
    ***REMOVED***
    result.push(***REMOVED***
        pos,
        end: sourceFile.end,
        contentLength: sourceFile.end - pos,
    ***REMOVED***);
    return result;
***REMOVED***
exports.getLineRanges = getLineRanges;
function getLineBreakStyle(sourceFile) ***REMOVED***
    const lineStarts = sourceFile.getLineStarts();
    return lineStarts.length === 1 || lineStarts[1] < 2 || sourceFile.text[lineStarts[1] - 2] !== '\r'
        ? '\n'
        : '\r\n';
***REMOVED***
exports.getLineBreakStyle = getLineBreakStyle;
let cachedScanner;
function scanToken(text, languageVersion) ***REMOVED***
    if (cachedScanner === undefined) ***REMOVED***
        cachedScanner = ts.createScanner(languageVersion, false, undefined, text);
    ***REMOVED***
    else ***REMOVED***
        cachedScanner.setScriptTarget(languageVersion);
        cachedScanner.setText(text);
    ***REMOVED***
    cachedScanner.scan();
    return cachedScanner;
***REMOVED***
function isValidIdentifier(text, languageVersion = ts.ScriptTarget.Latest) ***REMOVED***
    const scan = scanToken(text, languageVersion);
    return scan.isIdentifier() && scan.getTextPos() === text.length && scan.getTokenPos() === 0;
***REMOVED***
exports.isValidIdentifier = isValidIdentifier;
function charSize(ch) ***REMOVED***
    return ch >= 0x10000 ? 2 : 1;
***REMOVED***
function isValidPropertyAccess(text, languageVersion = ts.ScriptTarget.Latest) ***REMOVED***
    if (text.length === 0)
        return false;
    let ch = text.codePointAt(0);
    if (!ts.isIdentifierStart(ch, languageVersion))
        return false;
    for (let i = charSize(ch); i < text.length; i += charSize(ch)) ***REMOVED***
        ch = text.codePointAt(i);
        if (!ts.isIdentifierPart(ch, languageVersion))
            return false;
    ***REMOVED***
    return true;
***REMOVED***
exports.isValidPropertyAccess = isValidPropertyAccess;
function isValidPropertyName(text, languageVersion = ts.ScriptTarget.Latest) ***REMOVED***
    if (isValidPropertyAccess(text, languageVersion))
        return true;
    const scan = scanToken(text, languageVersion);
    return scan.getTextPos() === text.length &&
        scan.getToken() === ts.SyntaxKind.NumericLiteral && scan.getTokenValue() === text;
***REMOVED***
exports.isValidPropertyName = isValidPropertyName;
function isValidNumericLiteral(text, languageVersion = ts.ScriptTarget.Latest) ***REMOVED***
    const scan = scanToken(text, languageVersion);
    return scan.getToken() === ts.SyntaxKind.NumericLiteral && scan.getTextPos() === text.length && scan.getTokenPos() === 0;
***REMOVED***
exports.isValidNumericLiteral = isValidNumericLiteral;
function isValidJsxIdentifier(text, languageVersion = ts.ScriptTarget.Latest) ***REMOVED***
    if (text.length === 0)
        return false;
    let ch = text.codePointAt(0);
    if (!ts.isIdentifierStart(ch, languageVersion))
        return false;
    for (let i = charSize(ch); i < text.length; i += charSize(ch)) ***REMOVED***
        ch = text.codePointAt(i);
        if (!ts.isIdentifierPart(ch, languageVersion) && ch !== 45)
            return false;
    ***REMOVED***
    return true;
***REMOVED***
exports.isValidJsxIdentifier = isValidJsxIdentifier;
function isNumericPropertyName(name) ***REMOVED***
    return String(+name) === name;
***REMOVED***
exports.isNumericPropertyName = isNumericPropertyName;
function isSameLine(sourceFile, pos1, pos2) ***REMOVED***
    return ts.getLineAndCharacterOfPosition(sourceFile, pos1).line === ts.getLineAndCharacterOfPosition(sourceFile, pos2).line;
***REMOVED***
exports.isSameLine = isSameLine;
var SideEffectOptions;
(function (SideEffectOptions) ***REMOVED***
    SideEffectOptions[SideEffectOptions["None"] = 0] = "None";
    SideEffectOptions[SideEffectOptions["TaggedTemplate"] = 1] = "TaggedTemplate";
    SideEffectOptions[SideEffectOptions["Constructor"] = 2] = "Constructor";
    SideEffectOptions[SideEffectOptions["JsxElement"] = 4] = "JsxElement";
***REMOVED***)(SideEffectOptions = exports.SideEffectOptions || (exports.SideEffectOptions = ***REMOVED******REMOVED***));
function hasSideEffects(node, options) ***REMOVED***
    switch (node.kind) ***REMOVED***
        case ts.SyntaxKind.CallExpression:
        case ts.SyntaxKind.PostfixUnaryExpression:
        case ts.SyntaxKind.AwaitExpression:
        case ts.SyntaxKind.YieldExpression:
        case ts.SyntaxKind.DeleteExpression:
            return true;
        case ts.SyntaxKind.TypeAssertionExpression:
        case ts.SyntaxKind.AsExpression:
        case ts.SyntaxKind.ParenthesizedExpression:
        case ts.SyntaxKind.NonNullExpression:
        case ts.SyntaxKind.VoidExpression:
        case ts.SyntaxKind.TypeOfExpression:
        case ts.SyntaxKind.PropertyAccessExpression:
        case ts.SyntaxKind.SpreadElement:
        case ts.SyntaxKind.PartiallyEmittedExpression:
            return hasSideEffects(node.expression, options);
        case ts.SyntaxKind.BinaryExpression:
            return isAssignmentKind(node.operatorToken.kind) ||
                hasSideEffects(node.left, options) ||
                hasSideEffects(node.right, options);
        case ts.SyntaxKind.PrefixUnaryExpression:
            switch (node.operator) ***REMOVED***
                case ts.SyntaxKind.PlusPlusToken:
                case ts.SyntaxKind.MinusMinusToken:
                    return true;
                default:
                    return hasSideEffects(node.operand, options);
            ***REMOVED***
        case ts.SyntaxKind.ElementAccessExpression:
            return hasSideEffects(node.expression, options) ||
                node.argumentExpression !== undefined &&
                    hasSideEffects(node.argumentExpression, options);
        case ts.SyntaxKind.ConditionalExpression:
            return hasSideEffects(node.condition, options) ||
                hasSideEffects(node.whenTrue, options) ||
                hasSideEffects(node.whenFalse, options);
        case ts.SyntaxKind.NewExpression:
            if (options & 2 || hasSideEffects(node.expression, options))
                return true;
            if (node.arguments !== undefined)
                for (const child of node.arguments)
                    if (hasSideEffects(child, options))
                        return true;
            return false;
        case ts.SyntaxKind.TaggedTemplateExpression:
            if (options & 1 || hasSideEffects(node.tag, options))
                return true;
            if (node.template.kind === ts.SyntaxKind.NoSubstitutionTemplateLiteral)
                return false;
            node = node.template;
        case ts.SyntaxKind.TemplateExpression:
            for (const child of node.templateSpans)
                if (hasSideEffects(child.expression, options))
                    return true;
            return false;
        case ts.SyntaxKind.ClassExpression:
            return classExpressionHasSideEffects(node, options);
        case ts.SyntaxKind.ArrayLiteralExpression:
            for (const child of node.elements)
                if (hasSideEffects(child, options))
                    return true;
            return false;
        case ts.SyntaxKind.ObjectLiteralExpression:
            for (const child of node.properties) ***REMOVED***
                if (child.name !== undefined && child.name.kind === ts.SyntaxKind.ComputedPropertyName &&
                    hasSideEffects(child.name.expression, options))
                    return true;
                switch (child.kind) ***REMOVED***
                    case ts.SyntaxKind.PropertyAssignment:
                        if (hasSideEffects(child.initializer, options))
                            return true;
                        break;
                    case ts.SyntaxKind.SpreadAssignment:
                        if (hasSideEffects(child.expression, options))
                            return true;
                ***REMOVED***
            ***REMOVED***
            return false;
        case ts.SyntaxKind.JsxExpression:
            return node.expression !== undefined && hasSideEffects(node.expression, options);
        case ts.SyntaxKind.JsxElement:
        case ts.SyntaxKind.JsxFragment:
            for (const child of node.children)
                if (child.kind !== ts.SyntaxKind.JsxText && hasSideEffects(child, options))
                    return true;
            if (node.kind === ts.SyntaxKind.JsxFragment)
                return false;
            node = node.openingElement;
        case ts.SyntaxKind.JsxSelfClosingElement:
        case ts.SyntaxKind.JsxOpeningElement:
            if (options & 4)
                return true;
            for (const child of node.attributes.properties) ***REMOVED***
                if (child.kind === ts.SyntaxKind.JsxSpreadAttribute) ***REMOVED***
                    if (hasSideEffects(child.expression, options))
                        return true;
                ***REMOVED***
                else if (child.initializer !== undefined && hasSideEffects(child.initializer, options)) ***REMOVED***
                    return true;
                ***REMOVED***
            ***REMOVED***
            return false;
        case ts.SyntaxKind.CommaListExpression:
            for (const child of node.elements)
                if (hasSideEffects(child, options))
                    return true;
            return false;
        default:
            return false;
    ***REMOVED***
***REMOVED***
exports.hasSideEffects = hasSideEffects;
function classExpressionHasSideEffects(node, options) ***REMOVED***
    if (node.heritageClauses !== undefined && node.heritageClauses[0].token === ts.SyntaxKind.ExtendsKeyword)
        for (const base of node.heritageClauses[0].types)
            if (hasSideEffects(base.expression, options))
                return true;
    for (const child of node.members)
        if (child.name !== undefined && child.name.kind === ts.SyntaxKind.ComputedPropertyName &&
            hasSideEffects(child.name.expression, options) ||
            node_1.isPropertyDeclaration(child) && child.initializer !== undefined &&
                hasSideEffects(child.initializer, options))
            return true;
    return false;
***REMOVED***
function getDeclarationOfBindingElement(node) ***REMOVED***
    let parent = node.parent.parent;
    while (parent.kind === ts.SyntaxKind.BindingElement)
        parent = parent.parent.parent;
    return parent;
***REMOVED***
exports.getDeclarationOfBindingElement = getDeclarationOfBindingElement;
function isExpressionValueUsed(node) ***REMOVED***
    while (true) ***REMOVED***
        const parent = node.parent;
        switch (parent.kind) ***REMOVED***
            case ts.SyntaxKind.CallExpression:
            case ts.SyntaxKind.NewExpression:
            case ts.SyntaxKind.ElementAccessExpression:
            case ts.SyntaxKind.WhileStatement:
            case ts.SyntaxKind.DoStatement:
            case ts.SyntaxKind.WithStatement:
            case ts.SyntaxKind.ThrowStatement:
            case ts.SyntaxKind.ReturnStatement:
            case ts.SyntaxKind.JsxExpression:
            case ts.SyntaxKind.JsxSpreadAttribute:
            case ts.SyntaxKind.JsxElement:
            case ts.SyntaxKind.JsxFragment:
            case ts.SyntaxKind.JsxSelfClosingElement:
            case ts.SyntaxKind.ComputedPropertyName:
            case ts.SyntaxKind.ArrowFunction:
            case ts.SyntaxKind.ExportSpecifier:
            case ts.SyntaxKind.ExportAssignment:
            case ts.SyntaxKind.ImportDeclaration:
            case ts.SyntaxKind.ExternalModuleReference:
            case ts.SyntaxKind.Decorator:
            case ts.SyntaxKind.TaggedTemplateExpression:
            case ts.SyntaxKind.TemplateSpan:
            case ts.SyntaxKind.ExpressionWithTypeArguments:
            case ts.SyntaxKind.TypeOfExpression:
            case ts.SyntaxKind.AwaitExpression:
            case ts.SyntaxKind.YieldExpression:
            case ts.SyntaxKind.LiteralType:
            case ts.SyntaxKind.JsxAttributes:
            case ts.SyntaxKind.JsxOpeningElement:
            case ts.SyntaxKind.JsxClosingElement:
            case ts.SyntaxKind.IfStatement:
            case ts.SyntaxKind.CaseClause:
            case ts.SyntaxKind.SwitchStatement:
                return true;
            case ts.SyntaxKind.PropertyAccessExpression:
                return parent.expression === node;
            case ts.SyntaxKind.QualifiedName:
                return parent.left === node;
            case ts.SyntaxKind.ShorthandPropertyAssignment:
                return parent.objectAssignmentInitializer === node ||
                    !isInDestructuringAssignment(parent);
            case ts.SyntaxKind.PropertyAssignment:
                return parent.initializer === node && !isInDestructuringAssignment(parent);
            case ts.SyntaxKind.SpreadAssignment:
            case ts.SyntaxKind.SpreadElement:
            case ts.SyntaxKind.ArrayLiteralExpression:
                return !isInDestructuringAssignment(parent);
            case ts.SyntaxKind.ParenthesizedExpression:
            case ts.SyntaxKind.AsExpression:
            case ts.SyntaxKind.TypeAssertionExpression:
            case ts.SyntaxKind.PostfixUnaryExpression:
            case ts.SyntaxKind.PrefixUnaryExpression:
            case ts.SyntaxKind.NonNullExpression:
                node = parent;
                break;
            case ts.SyntaxKind.ForStatement:
                return parent.condition === node;
            case ts.SyntaxKind.ForInStatement:
            case ts.SyntaxKind.ForOfStatement:
                return parent.expression === node;
            case ts.SyntaxKind.ConditionalExpression:
                if (parent.condition === node)
                    return true;
                node = parent;
                break;
            case ts.SyntaxKind.PropertyDeclaration:
            case ts.SyntaxKind.BindingElement:
            case ts.SyntaxKind.VariableDeclaration:
            case ts.SyntaxKind.Parameter:
            case ts.SyntaxKind.EnumMember:
                return parent.initializer === node;
            case ts.SyntaxKind.ImportEqualsDeclaration:
                return parent.moduleReference === node;
            case ts.SyntaxKind.CommaListExpression:
                if (parent.elements[parent.elements.length - 1] !== node)
                    return false;
                node = parent;
                break;
            case ts.SyntaxKind.BinaryExpression:
                if (parent.right === node) ***REMOVED***
                    if (parent.operatorToken.kind === ts.SyntaxKind.CommaToken) ***REMOVED***
                        node = parent;
                        break;
                    ***REMOVED***
                    return true;
                ***REMOVED***
                switch (parent.operatorToken.kind) ***REMOVED***
                    case ts.SyntaxKind.CommaToken:
                    case ts.SyntaxKind.EqualsToken:
                        return false;
                    case ts.SyntaxKind.EqualsEqualsEqualsToken:
                    case ts.SyntaxKind.EqualsEqualsToken:
                    case ts.SyntaxKind.ExclamationEqualsEqualsToken:
                    case ts.SyntaxKind.ExclamationEqualsToken:
                    case ts.SyntaxKind.InstanceOfKeyword:
                    case ts.SyntaxKind.PlusToken:
                    case ts.SyntaxKind.MinusToken:
                    case ts.SyntaxKind.AsteriskToken:
                    case ts.SyntaxKind.SlashToken:
                    case ts.SyntaxKind.PercentToken:
                    case ts.SyntaxKind.AsteriskAsteriskToken:
                    case ts.SyntaxKind.GreaterThanToken:
                    case ts.SyntaxKind.GreaterThanGreaterThanToken:
                    case ts.SyntaxKind.GreaterThanGreaterThanGreaterThanToken:
                    case ts.SyntaxKind.GreaterThanEqualsToken:
                    case ts.SyntaxKind.LessThanToken:
                    case ts.SyntaxKind.LessThanLessThanToken:
                    case ts.SyntaxKind.LessThanEqualsToken:
                    case ts.SyntaxKind.AmpersandToken:
                    case ts.SyntaxKind.BarToken:
                    case ts.SyntaxKind.CaretToken:
                    case ts.SyntaxKind.BarBarToken:
                    case ts.SyntaxKind.AmpersandAmpersandToken:
                    case ts.SyntaxKind.InKeyword:
                        return true;
                    default:
                        node = parent;
                ***REMOVED***
                break;
            default:
                return false;
        ***REMOVED***
    ***REMOVED***
***REMOVED***
exports.isExpressionValueUsed = isExpressionValueUsed;
function isInDestructuringAssignment(node) ***REMOVED***
    switch (node.kind) ***REMOVED***
        case ts.SyntaxKind.ShorthandPropertyAssignment:
            if (node.objectAssignmentInitializer !== undefined)
                return true;
        case ts.SyntaxKind.PropertyAssignment:
        case ts.SyntaxKind.SpreadAssignment:
            node = node.parent;
            break;
        case ts.SyntaxKind.SpreadElement:
            if (node.parent.kind !== ts.SyntaxKind.ArrayLiteralExpression)
                return false;
            node = node.parent;
    ***REMOVED***
    while (true) ***REMOVED***
        switch (node.parent.kind) ***REMOVED***
            case ts.SyntaxKind.BinaryExpression:
                return node.parent.left === node &&
                    node.parent.operatorToken.kind === ts.SyntaxKind.EqualsToken;
            case ts.SyntaxKind.ForOfStatement:
                return node.parent.initializer === node;
            case ts.SyntaxKind.ArrayLiteralExpression:
            case ts.SyntaxKind.ObjectLiteralExpression:
                node = node.parent;
                break;
            case ts.SyntaxKind.SpreadAssignment:
            case ts.SyntaxKind.PropertyAssignment:
                node = node.parent.parent;
                break;
            case ts.SyntaxKind.SpreadElement:
                if (node.parent.parent.kind !== ts.SyntaxKind.ArrayLiteralExpression)
                    return false;
                node = node.parent.parent;
                break;
            default:
                return false;
        ***REMOVED***
    ***REMOVED***
***REMOVED***
var AccessKind;
(function (AccessKind) ***REMOVED***
    AccessKind[AccessKind["None"] = 0] = "None";
    AccessKind[AccessKind["Read"] = 1] = "Read";
    AccessKind[AccessKind["Write"] = 2] = "Write";
    AccessKind[AccessKind["Delete"] = 4] = "Delete";
    AccessKind[AccessKind["ReadWrite"] = 3] = "ReadWrite";
    AccessKind[AccessKind["Modification"] = 6] = "Modification";
***REMOVED***)(AccessKind = exports.AccessKind || (exports.AccessKind = ***REMOVED******REMOVED***));
function getAccessKind(node) ***REMOVED***
    const parent = node.parent;
    switch (parent.kind) ***REMOVED***
        case ts.SyntaxKind.DeleteExpression:
            return 4;
        case ts.SyntaxKind.PostfixUnaryExpression:
            return 3;
        case ts.SyntaxKind.PrefixUnaryExpression:
            return parent.operator === ts.SyntaxKind.PlusPlusToken ||
                parent.operator === ts.SyntaxKind.MinusMinusToken
                ? 3
                : 1;
        case ts.SyntaxKind.BinaryExpression:
            return parent.right === node
                ? 1
                : !isAssignmentKind(parent.operatorToken.kind)
                    ? 1
                    : parent.operatorToken.kind === ts.SyntaxKind.EqualsToken
                        ? 2
                        : 3;
        case ts.SyntaxKind.ShorthandPropertyAssignment:
            return parent.objectAssignmentInitializer === node
                ? 1
                : isInDestructuringAssignment(parent)
                    ? 2
                    : 1;
        case ts.SyntaxKind.PropertyAssignment:
            return parent.name === node
                ? 0
                : isInDestructuringAssignment(parent)
                    ? 2
                    : 1;
        case ts.SyntaxKind.ArrayLiteralExpression:
        case ts.SyntaxKind.SpreadElement:
        case ts.SyntaxKind.SpreadAssignment:
            return isInDestructuringAssignment(parent)
                ? 2
                : 1;
        case ts.SyntaxKind.ParenthesizedExpression:
        case ts.SyntaxKind.NonNullExpression:
        case ts.SyntaxKind.TypeAssertionExpression:
        case ts.SyntaxKind.AsExpression:
            return getAccessKind(parent);
        case ts.SyntaxKind.ForOfStatement:
        case ts.SyntaxKind.ForInStatement:
            return parent.initializer === node
                ? 2
                : 1;
        case ts.SyntaxKind.ExpressionWithTypeArguments:
            return parent.parent.token === ts.SyntaxKind.ExtendsKeyword &&
                parent.parent.parent.kind !== ts.SyntaxKind.InterfaceDeclaration
                ? 1
                : 0;
        case ts.SyntaxKind.ComputedPropertyName:
        case ts.SyntaxKind.ExpressionStatement:
        case ts.SyntaxKind.TypeOfExpression:
        case ts.SyntaxKind.ElementAccessExpression:
        case ts.SyntaxKind.ForStatement:
        case ts.SyntaxKind.IfStatement:
        case ts.SyntaxKind.DoStatement:
        case ts.SyntaxKind.WhileStatement:
        case ts.SyntaxKind.SwitchStatement:
        case ts.SyntaxKind.WithStatement:
        case ts.SyntaxKind.ThrowStatement:
        case ts.SyntaxKind.CallExpression:
        case ts.SyntaxKind.NewExpression:
        case ts.SyntaxKind.TaggedTemplateExpression:
        case ts.SyntaxKind.JsxExpression:
        case ts.SyntaxKind.Decorator:
        case ts.SyntaxKind.TemplateSpan:
        case ts.SyntaxKind.JsxOpeningElement:
        case ts.SyntaxKind.JsxSelfClosingElement:
        case ts.SyntaxKind.JsxSpreadAttribute:
        case ts.SyntaxKind.VoidExpression:
        case ts.SyntaxKind.ReturnStatement:
        case ts.SyntaxKind.AwaitExpression:
        case ts.SyntaxKind.YieldExpression:
        case ts.SyntaxKind.ConditionalExpression:
        case ts.SyntaxKind.CaseClause:
        case ts.SyntaxKind.JsxElement:
            return 1;
        case ts.SyntaxKind.ArrowFunction:
            return parent.body === node
                ? 1
                : 2;
        case ts.SyntaxKind.PropertyDeclaration:
        case ts.SyntaxKind.VariableDeclaration:
        case ts.SyntaxKind.Parameter:
        case ts.SyntaxKind.EnumMember:
        case ts.SyntaxKind.BindingElement:
        case ts.SyntaxKind.JsxAttribute:
            return parent.initializer === node
                ? 1
                : 0;
        case ts.SyntaxKind.PropertyAccessExpression:
            return parent.expression === node
                ? 1
                : 0;
        case ts.SyntaxKind.ExportAssignment:
            return parent.isExportEquals
                ? 1
                : 0;
    ***REMOVED***
    return 0;
***REMOVED***
exports.getAccessKind = getAccessKind;
function isReassignmentTarget(node) ***REMOVED***
    return (getAccessKind(node) & 2) !== 0;
***REMOVED***
exports.isReassignmentTarget = isReassignmentTarget;
function canHaveJsDoc(node) ***REMOVED***
    const kind = node.kind;
    switch (kind) ***REMOVED***
        case ts.SyntaxKind.Parameter:
        case ts.SyntaxKind.CallSignature:
        case ts.SyntaxKind.ConstructSignature:
        case ts.SyntaxKind.MethodSignature:
        case ts.SyntaxKind.PropertySignature:
        case ts.SyntaxKind.ArrowFunction:
        case ts.SyntaxKind.ParenthesizedExpression:
        case ts.SyntaxKind.SpreadAssignment:
        case ts.SyntaxKind.ShorthandPropertyAssignment:
        case ts.SyntaxKind.PropertyAssignment:
        case ts.SyntaxKind.FunctionExpression:
        case ts.SyntaxKind.FunctionDeclaration:
        case ts.SyntaxKind.LabeledStatement:
        case ts.SyntaxKind.ExpressionStatement:
        case ts.SyntaxKind.VariableStatement:
        case ts.SyntaxKind.Constructor:
        case ts.SyntaxKind.MethodDeclaration:
        case ts.SyntaxKind.PropertyDeclaration:
        case ts.SyntaxKind.GetAccessor:
        case ts.SyntaxKind.SetAccessor:
        case ts.SyntaxKind.ClassDeclaration:
        case ts.SyntaxKind.ClassExpression:
        case ts.SyntaxKind.InterfaceDeclaration:
        case ts.SyntaxKind.TypeAliasDeclaration:
        case ts.SyntaxKind.EnumMember:
        case ts.SyntaxKind.EnumDeclaration:
        case ts.SyntaxKind.ModuleDeclaration:
        case ts.SyntaxKind.ImportEqualsDeclaration:
        case ts.SyntaxKind.IndexSignature:
        case ts.SyntaxKind.FunctionType:
        case ts.SyntaxKind.ConstructorType:
        case ts.SyntaxKind.JSDocFunctionType:
        case ts.SyntaxKind.EndOfFileToken:
        case ts.SyntaxKind.ExportDeclaration:
            return true;
        default:
            return false;
    ***REMOVED***
***REMOVED***
exports.canHaveJsDoc = canHaveJsDoc;
function getJsDoc(node, sourceFile) ***REMOVED***
    if (node.kind === ts.SyntaxKind.EndOfFileToken)
        return parseJsDocWorker(node, sourceFile || node.parent);
    const result = [];
    for (const child of node.getChildren(sourceFile)) ***REMOVED***
        if (!node_1.isJsDoc(child))
            break;
        result.push(child);
    ***REMOVED***
    return result;
***REMOVED***
exports.getJsDoc = getJsDoc;
function parseJsDocOfNode(node, considerTrailingComments, sourceFile = node.getSourceFile()) ***REMOVED***
    if (canHaveJsDoc(node) && node.kind !== ts.SyntaxKind.EndOfFileToken) ***REMOVED***
        const result = getJsDoc(node, sourceFile);
        if (result.length !== 0 || !considerTrailingComments)
            return result;
    ***REMOVED***
    return parseJsDocWorker(node, sourceFile, considerTrailingComments);
***REMOVED***
exports.parseJsDocOfNode = parseJsDocOfNode;
function parseJsDocWorker(node, sourceFile, considerTrailingComments) ***REMOVED***
    const nodeStart = node.getStart(sourceFile);
    const start = ts[considerTrailingComments && isSameLine(sourceFile, node.pos, nodeStart)
        ? 'forEachTrailingCommentRange'
        : 'forEachLeadingCommentRange'](sourceFile.text, node.pos, (pos, _end, kind) => kind === ts.SyntaxKind.MultiLineCommentTrivia && sourceFile.text[pos + 2] === '*' ? ***REMOVED*** pos ***REMOVED*** : undefined);
    if (start === undefined)
        return [];
    const startPos = start.pos;
    const text = sourceFile.text.slice(startPos, nodeStart);
    const newSourceFile = ts.createSourceFile('jsdoc.ts', `$***REMOVED***text***REMOVED***var a;`, sourceFile.languageVersion);
    const result = getJsDoc(newSourceFile.statements[0], newSourceFile);
    for (const doc of result)
        updateNode(doc, node);
    return result;
    function updateNode(n, parent) ***REMOVED***
        n.pos += startPos;
        n.end += startPos;
        n.parent = parent;
        return ts.forEachChild(n, (child) => updateNode(child, n), (children) => ***REMOVED***
            children.pos += startPos;
            children.end += startPos;
            for (const child of children)
                updateNode(child, n);
        ***REMOVED***);
    ***REMOVED***
***REMOVED***
var ImportKind;
(function (ImportKind) ***REMOVED***
    ImportKind[ImportKind["ImportDeclaration"] = 1] = "ImportDeclaration";
    ImportKind[ImportKind["ImportEquals"] = 2] = "ImportEquals";
    ImportKind[ImportKind["ExportFrom"] = 4] = "ExportFrom";
    ImportKind[ImportKind["DynamicImport"] = 8] = "DynamicImport";
    ImportKind[ImportKind["Require"] = 16] = "Require";
    ImportKind[ImportKind["ImportType"] = 32] = "ImportType";
    ImportKind[ImportKind["All"] = 63] = "All";
    ImportKind[ImportKind["AllImports"] = 59] = "AllImports";
    ImportKind[ImportKind["AllStaticImports"] = 3] = "AllStaticImports";
    ImportKind[ImportKind["AllImportExpressions"] = 24] = "AllImportExpressions";
    ImportKind[ImportKind["AllRequireLike"] = 18] = "AllRequireLike";
    ImportKind[ImportKind["AllNestedImports"] = 56] = "AllNestedImports";
    ImportKind[ImportKind["AllTopLevelImports"] = 7] = "AllTopLevelImports";
***REMOVED***)(ImportKind = exports.ImportKind || (exports.ImportKind = ***REMOVED******REMOVED***));
function findImports(sourceFile, kinds) ***REMOVED***
    const result = [];
    for (const node of findImportLikeNodes(sourceFile, kinds)) ***REMOVED***
        switch (node.kind) ***REMOVED***
            case ts.SyntaxKind.ImportDeclaration:
                addIfTextualLiteral(node.moduleSpecifier);
                break;
            case ts.SyntaxKind.ImportEqualsDeclaration:
                addIfTextualLiteral(node.moduleReference.expression);
                break;
            case ts.SyntaxKind.ExportDeclaration:
                addIfTextualLiteral(node.moduleSpecifier);
                break;
            case ts.SyntaxKind.CallExpression:
                addIfTextualLiteral(node.arguments[0]);
                break;
            case ts.SyntaxKind.ImportType:
                if (node_1.isLiteralTypeNode(node.argument))
                    addIfTextualLiteral(node.argument.literal);
                break;
            default:
                throw new Error('unexpected node');
        ***REMOVED***
    ***REMOVED***
    return result;
    function addIfTextualLiteral(node) ***REMOVED***
        if (node_1.isTextualLiteral(node))
            result.push(node);
    ***REMOVED***
***REMOVED***
exports.findImports = findImports;
function findImportLikeNodes(sourceFile, kinds) ***REMOVED***
    return new ImportFinder(sourceFile, kinds).find();
***REMOVED***
exports.findImportLikeNodes = findImportLikeNodes;
class ImportFinder ***REMOVED***
    constructor(_sourceFile, _options) ***REMOVED***
        this._sourceFile = _sourceFile;
        this._options = _options;
        this._result = [];
    ***REMOVED***
    find() ***REMOVED***
        if (this._sourceFile.isDeclarationFile)
            this._options &= ~24;
        if (this._options & 7)
            this._findImports(this._sourceFile.statements);
        if (this._options & 56)
            this._findNestedImports();
        return this._result;
    ***REMOVED***
    _findImports(statements) ***REMOVED***
        for (const statement of statements) ***REMOVED***
            if (node_1.isImportDeclaration(statement)) ***REMOVED***
                if (this._options & 1)
                    this._result.push(statement);
            ***REMOVED***
            else if (node_1.isImportEqualsDeclaration(statement)) ***REMOVED***
                if (this._options & 2 &&
                    statement.moduleReference.kind === ts.SyntaxKind.ExternalModuleReference)
                    this._result.push(statement);
            ***REMOVED***
            else if (node_1.isExportDeclaration(statement)) ***REMOVED***
                if (statement.moduleSpecifier !== undefined && this._options & 4)
                    this._result.push(statement);
            ***REMOVED***
            else if (node_1.isModuleDeclaration(statement)) ***REMOVED***
                this._findImportsInModule(statement);
            ***REMOVED***
        ***REMOVED***
    ***REMOVED***
    _findImportsInModule(declaration) ***REMOVED***
        if (declaration.body === undefined)
            return;
        if (declaration.body.kind === ts.SyntaxKind.ModuleDeclaration)
            return this._findImportsInModule(declaration.body);
        this._findImports(declaration.body.statements);
    ***REMOVED***
    _findNestedImports() ***REMOVED***
        let re;
        if ((this._options & 56) === 16) ***REMOVED***
            re = /\brequire\s*[</(]/g;
        ***REMOVED***
        else if (this._options & 16) ***REMOVED***
            re = /\b(?:import|require)\s*[</(]/g;
        ***REMOVED***
        else ***REMOVED***
            re = /\bimport\s*[</(]/g;
        ***REMOVED***
        const isJavaScriptFile = (this._sourceFile.flags & ts.NodeFlags.JavaScriptFile) !== 0;
        for (let match = re.exec(this._sourceFile.text); match !== null; match = re.exec(this._sourceFile.text)) ***REMOVED***
            const token = getTokenAtPositionWorker(this._sourceFile, match.index, this._sourceFile, match[0][0] === 'i' && isJavaScriptFile);
            if (token.kind === ts.SyntaxKind.ImportKeyword) ***REMOVED***
                if (token.end - 'import'.length !== match.index)
                    continue;
                switch (token.parent.kind) ***REMOVED***
                    case ts.SyntaxKind.ImportType:
                        this._result.push(token.parent);
                        break;
                    case ts.SyntaxKind.CallExpression:
                        if (token.parent.arguments.length === 1)
                            this._result.push(token.parent);
                ***REMOVED***
            ***REMOVED***
            else if (token.kind === ts.SyntaxKind.Identifier &&
                token.end - 'require'.length === match.index &&
                token.parent.kind === ts.SyntaxKind.CallExpression &&
                token.parent.expression === token &&
                token.parent.arguments.length === 1) ***REMOVED***
                this._result.push(token.parent);
            ***REMOVED***
        ***REMOVED***
    ***REMOVED***
***REMOVED***
function isStatementInAmbientContext(node) ***REMOVED***
    while (node.flags & ts.NodeFlags.NestedNamespace)
        node = node.parent;
    return hasModifier(node.modifiers, ts.SyntaxKind.DeclareKeyword) || isAmbientModuleBlock(node.parent);
***REMOVED***
exports.isStatementInAmbientContext = isStatementInAmbientContext;
function isAmbientModuleBlock(node) ***REMOVED***
    while (node.kind === ts.SyntaxKind.ModuleBlock) ***REMOVED***
        do
            node = node.parent;
        while (node.flags & ts.NodeFlags.NestedNamespace);
        if (hasModifier(node.modifiers, ts.SyntaxKind.DeclareKeyword))
            return true;
        node = node.parent;
    ***REMOVED***
    return false;
***REMOVED***
exports.isAmbientModuleBlock = isAmbientModuleBlock;
function getIIFE(func) ***REMOVED***
    let node = func.parent;
    while (node.kind === ts.SyntaxKind.ParenthesizedExpression)
        node = node.parent;
    return node_1.isCallExpression(node) && func.end <= node.expression.end ? node : undefined;
***REMOVED***
exports.getIIFE = getIIFE;
function isStrictCompilerOptionEnabled(options, option) ***REMOVED***
    return (options.strict ? options[option] !== false : options[option] === true) &&
        (option !== 'strictPropertyInitialization' || isStrictCompilerOptionEnabled(options, 'strictNullChecks'));
***REMOVED***
exports.isStrictCompilerOptionEnabled = isStrictCompilerOptionEnabled;
function isCompilerOptionEnabled(options, option) ***REMOVED***
    switch (option) ***REMOVED***
        case 'stripInternal':
            return options.stripInternal === true && isCompilerOptionEnabled(options, 'declaration');
        case 'declaration':
            return options.declaration || isCompilerOptionEnabled(options, 'composite');
        case 'incremental':
            return options.incremental === undefined ? isCompilerOptionEnabled(options, 'composite') : options.incremental;
        case 'skipDefaultLibCheck':
            return options.skipDefaultLibCheck || isCompilerOptionEnabled(options, 'skipLibCheck');
        case 'suppressImplicitAnyIndexErrors':
            return options.suppressImplicitAnyIndexErrors === true && isCompilerOptionEnabled(options, 'noImplicitAny');
        case 'allowSyntheticDefaultImports':
            return options.allowSyntheticDefaultImports !== undefined
                ? options.allowSyntheticDefaultImports
                : isCompilerOptionEnabled(options, 'esModuleInterop') || options.module === ts.ModuleKind.System;
        case 'noImplicitAny':
        case 'noImplicitThis':
        case 'strictNullChecks':
        case 'strictFunctionTypes':
        case 'strictPropertyInitialization':
        case 'alwaysStrict':
        case 'strictBindCallApply':
            return isStrictCompilerOptionEnabled(options, option);
    ***REMOVED***
    return options[option] === true;
***REMOVED***
exports.isCompilerOptionEnabled = isCompilerOptionEnabled;
function isAmbientModule(node) ***REMOVED***
    return node.name.kind === ts.SyntaxKind.StringLiteral || (node.flags & ts.NodeFlags.GlobalAugmentation) !== 0;
***REMOVED***
exports.isAmbientModule = isAmbientModule;
function getCheckJsDirective(source) ***REMOVED***
    let directive;
    ts.forEachLeadingCommentRange(source, (ts.getShebang(source) || '').length, (pos, end, kind) => ***REMOVED***
        if (kind === ts.SyntaxKind.SingleLineCommentTrivia) ***REMOVED***
            const text = source.slice(pos, end);
            const match = /^\/***REMOVED***2,3***REMOVED***\s*@ts-(no)?check(?:\s|$)/i.exec(text);
            if (match !== null)
                directive = ***REMOVED*** pos, end, enabled: match[1] === undefined ***REMOVED***;
        ***REMOVED***
    ***REMOVED***);
    return directive;
***REMOVED***
exports.getCheckJsDirective = getCheckJsDirective;
function isConstAssertion(node) ***REMOVED***
    return node_1.isTypeReferenceNode(node.type) &&
        node.type.typeName.kind === ts.SyntaxKind.Identifier &&
        node.type.typeName.escapedText === 'const';
***REMOVED***
exports.isConstAssertion = isConstAssertion;
function isInConstContext(node) ***REMOVED***
    let current = node;
    while (true) ***REMOVED***
        const parent = current.parent;
        outer: switch (parent.kind) ***REMOVED***
            case ts.SyntaxKind.TypeAssertionExpression:
            case ts.SyntaxKind.AsExpression:
                return isConstAssertion(parent);
            case ts.SyntaxKind.PrefixUnaryExpression:
                if (current.kind !== ts.SyntaxKind.NumericLiteral)
                    return false;
                switch (parent.operator) ***REMOVED***
                    case ts.SyntaxKind.PlusToken:
                    case ts.SyntaxKind.MinusToken:
                        current = parent;
                        break outer;
                    default:
                        return false;
                ***REMOVED***
            case ts.SyntaxKind.PropertyAssignment:
                if (parent.initializer !== current)
                    return false;
                current = parent.parent;
                break;
            case ts.SyntaxKind.ShorthandPropertyAssignment:
                current = parent.parent;
                break;
            case ts.SyntaxKind.ParenthesizedExpression:
            case ts.SyntaxKind.ArrayLiteralExpression:
            case ts.SyntaxKind.ObjectLiteralExpression:
                current = parent;
                break;
            default:
                return false;
        ***REMOVED***
    ***REMOVED***
***REMOVED***
exports.isInConstContext = isInConstContext;
function isReadonlyAssignmentDeclaration(node, checker) ***REMOVED***
    if (!isBindableObjectDefinePropertyCall(node))
        return false;
    const descriptorType = checker.getTypeAtLocation(node.arguments[2]);
    if (descriptorType.getProperty('value') === undefined)
        return descriptorType.getProperty('set') === undefined;
    const writableProp = descriptorType.getProperty('writable');
    if (writableProp === undefined)
        return false;
    const writableType = writableProp.valueDeclaration !== undefined && node_1.isPropertyAssignment(writableProp.valueDeclaration)
        ? checker.getTypeAtLocation(writableProp.valueDeclaration.initializer)
        : checker.getTypeOfSymbolAtLocation(writableProp, node.arguments[2]);
    return type_1.isBooleanLiteralType(writableType, false);
***REMOVED***
exports.isReadonlyAssignmentDeclaration = isReadonlyAssignmentDeclaration;
function isBindableObjectDefinePropertyCall(node) ***REMOVED***
    return node.arguments.length === 3 &&
        node_1.isEntityNameExpression(node.arguments[0]) &&
        node_1.isNumericOrStringLikeLiteral(node.arguments[1]) &&
        node_1.isPropertyAccessExpression(node.expression) &&
        node.expression.name.escapedText === 'defineProperty' &&
        node_1.isIdentifier(node.expression.expression) &&
        node.expression.expression.escapedText === 'Object';
***REMOVED***
exports.isBindableObjectDefinePropertyCall = isBindableObjectDefinePropertyCall;
function isWellKnownSymbolLiterally(node) ***REMOVED***
    return ts.isPropertyAccessExpression(node) &&
        ts.isIdentifier(node.expression) &&
        node.expression.escapedText === 'Symbol';
***REMOVED***
exports.isWellKnownSymbolLiterally = isWellKnownSymbolLiterally;
function getPropertyNameOfWellKnownSymbol(node) ***REMOVED***
    return ***REMOVED***
        displayName: `[Symbol.$***REMOVED***node.name.text***REMOVED***]`,
        symbolName: ('__@' + node.name.text),
    ***REMOVED***;
***REMOVED***
exports.getPropertyNameOfWellKnownSymbol = getPropertyNameOfWellKnownSymbol;
function getLateBoundPropertyNames(node, checker) ***REMOVED***
    const result = ***REMOVED***
        known: true,
        names: [],
    ***REMOVED***;
    node = unwrapParentheses(node);
    if (isWellKnownSymbolLiterally(node)) ***REMOVED***
        result.names.push(getPropertyNameOfWellKnownSymbol(node));
    ***REMOVED***
    else ***REMOVED***
        const type = checker.getTypeAtLocation(node);
        for (const key of type_1.unionTypeParts(checker.getBaseConstraintOfType(type) || type)) ***REMOVED***
            const propertyName = type_1.getPropertyNameFromType(key);
            if (propertyName) ***REMOVED***
                result.names.push(propertyName);
            ***REMOVED***
            else ***REMOVED***
                result.known = false;
            ***REMOVED***
        ***REMOVED***
    ***REMOVED***
    return result;
***REMOVED***
exports.getLateBoundPropertyNames = getLateBoundPropertyNames;
function getLateBoundPropertyNamesOfPropertyName(node, checker) ***REMOVED***
    const staticName = getPropertyName(node);
    return staticName !== undefined
        ? ***REMOVED*** known: true, names: [***REMOVED*** displayName: staticName, symbolName: ts.escapeLeadingUnderscores(staticName) ***REMOVED***] ***REMOVED***
        : getLateBoundPropertyNames(node.expression, checker);
***REMOVED***
exports.getLateBoundPropertyNamesOfPropertyName = getLateBoundPropertyNamesOfPropertyName;
function getSingleLateBoundPropertyNameOfPropertyName(node, checker) ***REMOVED***
    const staticName = getPropertyName(node);
    if (staticName !== undefined)
        return ***REMOVED*** displayName: staticName, symbolName: ts.escapeLeadingUnderscores(staticName) ***REMOVED***;
    const ***REMOVED*** expression ***REMOVED*** = node;
    return isWellKnownSymbolLiterally(expression)
        ? getPropertyNameOfWellKnownSymbol(expression)
        : type_1.getPropertyNameFromType(checker.getTypeAtLocation(expression));
***REMOVED***
exports.getSingleLateBoundPropertyNameOfPropertyName = getSingleLateBoundPropertyNameOfPropertyName;
function unwrapParentheses(node) ***REMOVED***
    while (node.kind === ts.SyntaxKind.ParenthesizedExpression)
        node = node.expression;
    return node;
***REMOVED***
exports.unwrapParentheses = unwrapParentheses;
