// eslint/lib/source-code/token-store/backward-token-comment-cursor.js
/**
 * @fileoverview Define the cursor which iterates tokens and comments in reverse.
 * @author Toru Nagashima
 */
"use strict";

//------------------------------------------------------------------------------
// Requirements
//------------------------------------------------------------------------------

const Cursor = require("./cursor");
const utils = require("./utils");

//------------------------------------------------------------------------------
// Exports
//------------------------------------------------------------------------------

/**
 * The cursor which iterates tokens and comments in reverse.
 */
module.exports = class BackwardTokenCommentCursor extends Cursor ***REMOVED***

    /**
     * Initializes this cursor.
     * @param ***REMOVED***Token[]***REMOVED*** tokens - The array of tokens.
     * @param ***REMOVED***Comment[]***REMOVED*** comments - The array of comments.
     * @param ***REMOVED***Object***REMOVED*** indexMap - The map from locations to indices in `tokens`.
     * @param ***REMOVED***number***REMOVED*** startLoc - The start location of the iteration range.
     * @param ***REMOVED***number***REMOVED*** endLoc - The end location of the iteration range.
     */
    constructor(tokens, comments, indexMap, startLoc, endLoc) ***REMOVED***
        super();
        this.tokens = tokens;
        this.comments = comments;
        this.tokenIndex = utils.getLastIndex(tokens, indexMap, endLoc);
        this.commentIndex = utils.search(comments, endLoc) - 1;
        this.border = startLoc;
    ***REMOVED***

    /** @inheritdoc */
    moveNext() ***REMOVED***
        const token = (this.tokenIndex >= 0) ? this.tokens[this.tokenIndex] : null;
        const comment = (this.commentIndex >= 0) ? this.comments[this.commentIndex] : null;

        if (token && (!comment || token.range[1] > comment.range[1])) ***REMOVED***
            this.current = token;
            this.tokenIndex -= 1;
        ***REMOVED*** else if (comment) ***REMOVED***
            this.current = comment;
            this.commentIndex -= 1;
        ***REMOVED*** else ***REMOVED***
            this.current = null;
        ***REMOVED***

        return Boolean(this.current) && (this.border === -1 || this.current.range[0] >= this.border);
    ***REMOVED***
***REMOVED***;
