// eslint/lib/source-code/token-store/forward-token-comment-cursor.js
/**
 * @fileoverview Define the cursor which iterates tokens and comments.
 * @author Toru Nagashima
 */
"use strict";

//------------------------------------------------------------------------------
// Requirements
//------------------------------------------------------------------------------

const Cursor = require("./cursor");
const utils = require("./utils");

//------------------------------------------------------------------------------
// Exports
//------------------------------------------------------------------------------

/**
 * The cursor which iterates tokens and comments.
 */
module.exports = class ForwardTokenCommentCursor extends Cursor ***REMOVED***

    /**
     * Initializes this cursor.
     * @param ***REMOVED***Token[]***REMOVED*** tokens - The array of tokens.
     * @param ***REMOVED***Comment[]***REMOVED*** comments - The array of comments.
     * @param ***REMOVED***Object***REMOVED*** indexMap - The map from locations to indices in `tokens`.
     * @param ***REMOVED***number***REMOVED*** startLoc - The start location of the iteration range.
     * @param ***REMOVED***number***REMOVED*** endLoc - The end location of the iteration range.
     */
    constructor(tokens, comments, indexMap, startLoc, endLoc) ***REMOVED***
        super();
        this.tokens = tokens;
        this.comments = comments;
        this.tokenIndex = utils.getFirstIndex(tokens, indexMap, startLoc);
        this.commentIndex = utils.search(comments, startLoc);
        this.border = endLoc;
    ***REMOVED***

    /** @inheritdoc */
    moveNext() ***REMOVED***
        const token = (this.tokenIndex < this.tokens.length) ? this.tokens[this.tokenIndex] : null;
        const comment = (this.commentIndex < this.comments.length) ? this.comments[this.commentIndex] : null;

        if (token && (!comment || token.range[0] < comment.range[0])) ***REMOVED***
            this.current = token;
            this.tokenIndex += 1;
        ***REMOVED*** else if (comment) ***REMOVED***
            this.current = comment;
            this.commentIndex += 1;
        ***REMOVED*** else ***REMOVED***
            this.current = null;
        ***REMOVED***

        return Boolean(this.current) && (this.border === -1 || this.current.range[1] <= this.border);
    ***REMOVED***
***REMOVED***;
