// eslint-plugin-es/lib/rules/no-async-iteration.js
/**
 * @author Toru Nagashima <https://github.com/mysticatea>
 * See LICENSE file in root directory for full license.
 */
"use strict"

module.exports = ***REMOVED***
    meta: ***REMOVED***
        docs: ***REMOVED***
            description: "disallow async iteration.",
            category: "ES2018",
            recommended: false,
            url:
                "http://mysticatea.github.io/eslint-plugin-es/rules/no-async-iteration.html",
        ***REMOVED***,
        fixable: null,
        messages: ***REMOVED***
            forbidden: "ES2018 async iteration is forbidden.",
        ***REMOVED***,
        schema: [],
        type: "problem",
    ***REMOVED***,
    create(context) ***REMOVED***
        return ***REMOVED***
            ":function[async=true][generator=true]"(node) ***REMOVED***
                context.report(***REMOVED*** node, messageId: "forbidden" ***REMOVED***)
            ***REMOVED***,
            "ForOfStatement[await=true]"(node) ***REMOVED***
                context.report(***REMOVED*** node, messageId: "forbidden" ***REMOVED***)
            ***REMOVED***,
        ***REMOVED***
    ***REMOVED***,
***REMOVED***
