// eslint-plugin-es/lib/rules/no-bigint.js
/**
 * @author Toru Nagashima <https://github.com/mysticatea>
 * See LICENSE file in root directory for full license.
 */
"use strict"

const ***REMOVED*** ReferenceTracker ***REMOVED*** = require("eslint-utils")

module.exports = ***REMOVED***
    meta: ***REMOVED***
        docs: ***REMOVED***
            description: "disallow `bigint` syntax and built-ins",
            category: "ES2020",
            recommended: false,
            url:
                "http://mysticatea.github.io/eslint-plugin-es/rules/no-bigint.html",
        ***REMOVED***,
        fixable: null,
        messages: ***REMOVED***
            forbidden: "ES2020 BigInt is forbidden.",
        ***REMOVED***,
        schema: [],
        type: "problem",
    ***REMOVED***,
    create(context) ***REMOVED***
        return ***REMOVED***
            Literal(node) ***REMOVED***
                if (node.bigint != null) ***REMOVED***
                    context.report(***REMOVED*** messageId: "forbidden", node ***REMOVED***)
                ***REMOVED***
            ***REMOVED***,

            "Program:exit"() ***REMOVED***
                const tracker = new ReferenceTracker(context.getScope())
                const references = tracker.iterateGlobalReferences(***REMOVED***
                    BigInt: ***REMOVED*** [ReferenceTracker.READ]: true ***REMOVED***,
                    BigInt64Array: ***REMOVED*** [ReferenceTracker.READ]: true ***REMOVED***,
                    BigUint64Array: ***REMOVED*** [ReferenceTracker.READ]: true ***REMOVED***,
                ***REMOVED***)

                for (const ***REMOVED*** node ***REMOVED*** of references) ***REMOVED***
                    context.report(***REMOVED*** messageId: "forbidden", node ***REMOVED***)
                ***REMOVED***
            ***REMOVED***,
        ***REMOVED***
    ***REMOVED***,
***REMOVED***
