// eslint-plugin-es/lib/rules/no-computed-properties.js
/**
 * @author Toru Nagashima <https://github.com/mysticatea>
 * See LICENSE file in root directory for full license.
 */
"use strict"

module.exports = ***REMOVED***
    meta: ***REMOVED***
        docs: ***REMOVED***
            description: "disallow computed properties.",
            category: "ES2015",
            recommended: false,
            url:
                "http://mysticatea.github.io/eslint-plugin-es/rules/no-computed-properties.html",
        ***REMOVED***,
        fixable: null,
        messages: ***REMOVED***
            forbidden: "ES2015 computed properties are forbidden.",
        ***REMOVED***,
        schema: [],
        type: "problem",
    ***REMOVED***,
    create(context) ***REMOVED***
        return ***REMOVED***
            ":matches(Property, MethodDefinition)[computed=true]"(node) ***REMOVED***
                context.report(***REMOVED*** node, messageId: "forbidden" ***REMOVED***)
            ***REMOVED***,
        ***REMOVED***
    ***REMOVED***,
***REMOVED***
