// eslint-plugin-es/lib/rules/no-exponential-operators.js
/**
 * @author Toru Nagashima <https://github.com/mysticatea>
 * See LICENSE file in root directory for full license.
 */
"use strict"

module.exports = ***REMOVED***
    meta: ***REMOVED***
        docs: ***REMOVED***
            description: "disallow exponential operators.",
            category: "ES2016",
            recommended: false,
            url:
                "http://mysticatea.github.io/eslint-plugin-es/rules/no-exponential-operators.html",
        ***REMOVED***,
        fixable: null,
        messages: ***REMOVED***
            forbidden: "ES2016 exponential operators are forbidden.",
        ***REMOVED***,
        schema: [],
        type: "problem",
    ***REMOVED***,
    create(context) ***REMOVED***
        return ***REMOVED***
            "AssignmentExpression[operator='**='], BinaryExpression[operator='**']"(
                node
            ) ***REMOVED***
                context.report(***REMOVED*** node, messageId: "forbidden" ***REMOVED***)
            ***REMOVED***,
        ***REMOVED***
    ***REMOVED***,
***REMOVED***
