// eslint-plugin-es/lib/rules/no-json-superset.js
/**
 * @author Toru Nagashima <https://github.com/mysticatea>
 * See LICENSE file in root directory for full license.
 */
"use strict"

const ***REMOVED*** definePatternSearchGenerator ***REMOVED*** = require("../utils")
const iterateTargetChars = definePatternSearchGenerator(/[\u2028\u2029]/gu)

module.exports = ***REMOVED***
    meta: ***REMOVED***
        docs: ***REMOVED***
            description: "disallow `\\u2028` and `\\u2029` in string literals.",
            category: "ES2019",
            recommended: false,
            url:
                "http://mysticatea.github.io/eslint-plugin-es/rules/no-json-superset.html",
        ***REMOVED***,
        fixable: "code",
        messages: ***REMOVED***
            forbidden: "ES2019 '\\u***REMOVED******REMOVED***code***REMOVED******REMOVED***' in string literals is forbidden.",
        ***REMOVED***,
        schema: [],
        type: "problem",
    ***REMOVED***,
    create(context) ***REMOVED***
        const sourceCode = context.getSourceCode()
        return ***REMOVED***
            Literal(node) ***REMOVED***
                if (typeof node.value !== "string") ***REMOVED***
                    return
                ***REMOVED***

                const offset = node.range[0]
                for (const ***REMOVED*** index ***REMOVED*** of iterateTargetChars(node.raw)) ***REMOVED***
                    const code = node.raw.codePointAt(index).toString(16)
                    const loc = sourceCode.getLocFromIndex(offset + index)

                    context.report(***REMOVED***
                        node,
                        loc,
                        messageId: "forbidden",
                        data: ***REMOVED*** code ***REMOVED***,
                        fix(fixer) ***REMOVED***
                            return fixer.replaceTextRange(
                                [offset + index, offset + index + 1],
                                `\\u$***REMOVED***code***REMOVED***`
                            )
                        ***REMOVED***,
                    ***REMOVED***)
                ***REMOVED***
            ***REMOVED***,
        ***REMOVED***
    ***REMOVED***,
***REMOVED***
