// eslint-plugin-es/lib/rules/no-malformed-template-literals.js
/**
 * @author Toru Nagashima <https://github.com/mysticatea>
 * See LICENSE file in root directory for full license.
 */
"use strict"

module.exports = ***REMOVED***
    meta: ***REMOVED***
        docs: ***REMOVED***
            description:
                "disallow template literals with invalid escape sequences.",
            category: "ES2018",
            recommended: false,
            url:
                "http://mysticatea.github.io/eslint-plugin-es/rules/no-malformed-template-literals.html",
        ***REMOVED***,
        fixable: null,
        messages: ***REMOVED***
            forbidden:
                "ES2018 template literals with invalid escape sequences are forbidden.",
        ***REMOVED***,
        schema: [],
        type: "problem",
    ***REMOVED***,
    create(context) ***REMOVED***
        const reported = new Set()
        return ***REMOVED***
            "TemplateElement[value.cooked=null]"(elementNode) ***REMOVED***
                const node = elementNode.parent
                if (!reported.has(node)) ***REMOVED***
                    reported.add(node)
                    context.report(***REMOVED*** node, messageId: "forbidden" ***REMOVED***)
                ***REMOVED***
            ***REMOVED***,
        ***REMOVED***
    ***REMOVED***,
***REMOVED***
