// eslint-plugin-es/lib/rules/no-number-isinteger.js
/**
 * @author Toru Nagashima <https://github.com/mysticatea>
 * See LICENSE file in root directory for full license.
 */
"use strict"

const ***REMOVED*** READ, ReferenceTracker ***REMOVED*** = require("eslint-utils")

module.exports = ***REMOVED***
    meta: ***REMOVED***
        docs: ***REMOVED***
            description: "disallow the `Number.isInteger` method.",
            category: "ES2015",
            recommended: false,
            url:
                "http://mysticatea.github.io/eslint-plugin-es/rules/no-number-isinteger.html",
        ***REMOVED***,
        fixable: null,
        messages: ***REMOVED***
            forbidden: "ES2015 '***REMOVED******REMOVED***name***REMOVED******REMOVED***' method is forbidden.",
        ***REMOVED***,
        schema: [],
        type: "problem",
    ***REMOVED***,
    create(context) ***REMOVED***
        return ***REMOVED***
            "Program:exit"() ***REMOVED***
                const tracker = new ReferenceTracker(context.getScope())
                for (const ***REMOVED*** node, path ***REMOVED*** of tracker.iterateGlobalReferences(***REMOVED***
                    Number: ***REMOVED***
                        isInteger: ***REMOVED*** [READ]: true ***REMOVED***,
                    ***REMOVED***,
                ***REMOVED***)) ***REMOVED***
                    context.report(***REMOVED***
                        node,
                        messageId: "forbidden",
                        data: ***REMOVED*** name: path.join(".") ***REMOVED***,
                    ***REMOVED***)
                ***REMOVED***
            ***REMOVED***,
        ***REMOVED***
    ***REMOVED***,
***REMOVED***
