// eslint-plugin-es/lib/rules/no-object-super-properties.js
/**
 * @author Toru Nagashima <https://github.com/mysticatea>
 * See LICENSE file in root directory for full license.
 */
"use strict"

module.exports = ***REMOVED***
    meta: ***REMOVED***
        docs: ***REMOVED***
            description:
                "disallow `super` property accesses in object literals.",
            category: "ES2015",
            recommended: false,
            url:
                "http://mysticatea.github.io/eslint-plugin-es/rules/no-object-super-properties.html",
        ***REMOVED***,
        fixable: null,
        messages: ***REMOVED***
            forbidden:
                "ES2015 'super' property accesses in object literals are forbidden.",
        ***REMOVED***,
        schema: [],
        type: "problem",
    ***REMOVED***,
    create(context) ***REMOVED***
        let stack = null

        return ***REMOVED***
            Super(node) ***REMOVED***
                if (stack && stack.inObjectMethod) ***REMOVED***
                    context.report(***REMOVED*** node, messageId: "forbidden" ***REMOVED***)
                ***REMOVED***
            ***REMOVED***,

            ":matches(FunctionExpression, FunctionDeclaration)"(node) ***REMOVED***
                const ***REMOVED*** type, method ***REMOVED*** = node.parent
                stack = ***REMOVED***
                    inObjectMethod: type === "Property" && method === true,
                    upper: stack,
                ***REMOVED***
            ***REMOVED***,
            ":matches(FunctionExpression, FunctionDeclaration):exit"() ***REMOVED***
                stack = stack.upper
            ***REMOVED***,
        ***REMOVED***
    ***REMOVED***,
***REMOVED***
