// eslint-plugin-es/lib/rules/no-promise-all-settled.js
/**
 * @author Toru Nagashima <https://github.com/mysticatea>
 * See LICENSE file in root directory for full license.
 */
"use strict"

const ***REMOVED*** READ, ReferenceTracker ***REMOVED*** = require("eslint-utils")

module.exports = ***REMOVED***
    meta: ***REMOVED***
        docs: ***REMOVED***
            description: "disallow `Promise.allSettled` function",
            category: "ES2020",
            recommended: false,
            url:
                "http://mysticatea.github.io/eslint-plugin-es/rules/no-promise-all-settled.html",
        ***REMOVED***,
        fixable: null,
        messages: ***REMOVED***
            forbidden: "ES2020 'Promise.allSettled' function is forbidden.",
        ***REMOVED***,
        schema: [],
        type: "problem",
    ***REMOVED***,
    create(context) ***REMOVED***
        return ***REMOVED***
            "Program:exit"() ***REMOVED***
                const tracker = new ReferenceTracker(context.getScope())
                for (const ***REMOVED*** node ***REMOVED*** of tracker.iterateGlobalReferences(***REMOVED***
                    Promise: ***REMOVED*** allSettled: ***REMOVED*** [READ]: true ***REMOVED*** ***REMOVED***,
                ***REMOVED***)) ***REMOVED***
                    context.report(***REMOVED*** node, messageId: "forbidden" ***REMOVED***)
                ***REMOVED***
            ***REMOVED***,
        ***REMOVED***
    ***REMOVED***,
***REMOVED***
