// eslint-plugin-es/lib/rules/no-rest-spread-properties.js
/**
 * @author Toru Nagashima <https://github.com/mysticatea>
 * See LICENSE file in root directory for full license.
 */
"use strict"

module.exports = ***REMOVED***
    meta: ***REMOVED***
        docs: ***REMOVED***
            description: "disallow rest/spread properties.",
            category: "ES2018",
            recommended: false,
            url:
                "http://mysticatea.github.io/eslint-plugin-es/rules/no-rest-spread-properties.html",
        ***REMOVED***,
        fixable: null,
        messages: ***REMOVED***
            forbidden: "ES2018 rest/spread properties are forbidden.",
        ***REMOVED***,
        schema: [],
        type: "problem",
    ***REMOVED***,
    create(context) ***REMOVED***
        return ***REMOVED***
            "ObjectPattern > RestElement"(node) ***REMOVED***
                context.report(***REMOVED*** node, messageId: "forbidden" ***REMOVED***)
            ***REMOVED***,
            "ObjectExpression > SpreadElement"(node) ***REMOVED***
                context.report(***REMOVED*** node, messageId: "forbidden" ***REMOVED***)
            ***REMOVED***,
        ***REMOVED***
    ***REMOVED***,
***REMOVED***
