// eslint-plugin-es/lib/rules/no-regexp-u-flag.js
/**
 * @author Toru Nagashima <https://github.com/mysticatea>
 * See LICENSE file in root directory for full license.
 */
"use strict"

const ***REMOVED*** getRegExpCalls ***REMOVED*** = require("../utils")

module.exports = ***REMOVED***
    meta: ***REMOVED***
        docs: ***REMOVED***
            description: "disallow RegExp `u` flag.",
            category: "ES2015",
            recommended: false,
            url:
                "http://mysticatea.github.io/eslint-plugin-es/rules/no-regexp-u-flag.html",
        ***REMOVED***,
        fixable: null,
        messages: ***REMOVED***
            forbidden: "ES2015 RegExp 'u' flag is forbidden.",
        ***REMOVED***,
        schema: [],
        type: "problem",
    ***REMOVED***,
    create(context) ***REMOVED***
        return ***REMOVED***
            "Literal[regex]"(node) ***REMOVED***
                if (node.regex.flags.includes("u")) ***REMOVED***
                    context.report(***REMOVED*** node, messageId: "forbidden" ***REMOVED***)
                ***REMOVED***
            ***REMOVED***,

            "Program:exit"() ***REMOVED***
                const scope = context.getScope()

                for (const ***REMOVED*** node, flags ***REMOVED*** of getRegExpCalls(scope)) ***REMOVED***
                    if (flags && flags.includes("u")) ***REMOVED***
                        context.report(***REMOVED*** node, messageId: "forbidden" ***REMOVED***)
                    ***REMOVED***
                ***REMOVED***
            ***REMOVED***,
        ***REMOVED***
    ***REMOVED***,
***REMOVED***
