// eslint-plugin-es/lib/rules/no-regexp-named-capture-groups.js
/**
 * @author Toru Nagashima <https://github.com/mysticatea>
 * See LICENSE file in root directory for full license.
 */
"use strict"

const ***REMOVED*** RegExpValidator ***REMOVED*** = require("regexpp")
const ***REMOVED*** getRegExpCalls ***REMOVED*** = require("../utils")

/**
 * Verify a given regular expression.
 * @param ***REMOVED***RuleContext***REMOVED*** context The rule context to report.
 * @param ***REMOVED***Node***REMOVED*** node The AST node to report.
 * @param ***REMOVED***string***REMOVED*** pattern The pattern part of a RegExp.
 * @param ***REMOVED***string***REMOVED*** flags The flags part of a RegExp.
 * @returns ***REMOVED***void***REMOVED***
 */
function verify(context, node, pattern, flags) ***REMOVED***
    try ***REMOVED***
        let found = false

        new RegExpValidator(***REMOVED***
            onCapturingGroupEnter(_start, name) ***REMOVED***
                if (name) ***REMOVED***
                    found = true
                ***REMOVED***
            ***REMOVED***,
            onBackreference(_start, _end, ref) ***REMOVED***
                if (typeof ref === "string") ***REMOVED***
                    found = true
                ***REMOVED***
            ***REMOVED***,
        ***REMOVED***).validatePattern(pattern, 0, pattern.length, flags.includes("u"))

        if (found) ***REMOVED***
            context.report(***REMOVED*** node, messageId: "forbidden" ***REMOVED***)
        ***REMOVED***
    ***REMOVED*** catch (error) ***REMOVED***
        //istanbul ignore else
        if (error.message.startsWith("Invalid regular expression:")) ***REMOVED***
            return
        ***REMOVED***
        //istanbul ignore next
        throw error
    ***REMOVED***
***REMOVED***

module.exports = ***REMOVED***
    meta: ***REMOVED***
        docs: ***REMOVED***
            description: "disallow RegExp named capture groups.",
            category: "ES2018",
            recommended: false,
            url:
                "http://mysticatea.github.io/eslint-plugin-es/rules/no-regexp-named-capture-groups.html",
        ***REMOVED***,
        fixable: null,
        messages: ***REMOVED***
            forbidden: "ES2018 RegExp named capture groups are forbidden.",
        ***REMOVED***,
        schema: [],
        type: "problem",
    ***REMOVED***,
    create(context) ***REMOVED***
        return ***REMOVED***
            "Literal[regex]"(node) ***REMOVED***
                const ***REMOVED*** pattern, flags ***REMOVED*** = node.regex
                verify(context, node, pattern || "", flags || "")
            ***REMOVED***,

            "Program:exit"() ***REMOVED***
                const scope = context.getScope()
                for (const ***REMOVED*** node, pattern, flags ***REMOVED*** of getRegExpCalls(scope)) ***REMOVED***
                    verify(context, node, pattern || "", flags || "")
                ***REMOVED***
            ***REMOVED***,
        ***REMOVED***
    ***REMOVED***,
***REMOVED***
