// eslint-plugin-es/lib/rules/no-trailing-commas.js
/**
 * @author Toru Nagashima <https://github.com/mysticatea>
 * See LICENSE file in root directory for full license.
 */
"use strict"

const ***REMOVED*** isCommaToken ***REMOVED*** = require("../utils")

module.exports = ***REMOVED***
    meta: ***REMOVED***
        docs: ***REMOVED***
            description: "disallow trailing commas in array/object literals.",
            category: "ES5",
            recommended: false,
            url:
                "http://mysticatea.github.io/eslint-plugin-es/rules/no-trailing-commas.html",
        ***REMOVED***,
        fixable: null,
        messages: ***REMOVED***
            forbidden:
                "ES5 trailing commas in array/object literals are forbidden.",
        ***REMOVED***,
        schema: [],
        type: "problem",
    ***REMOVED***,
    create(context) ***REMOVED***
        const sourceCode = context.getSourceCode()
        return ***REMOVED***
            "ArrayExpression, ArrayPattern, ObjectExpression, ObjectPattern"(
                node
            ) ***REMOVED***
                const token = sourceCode.getLastToken(node, 1)
                if (isCommaToken(token)) ***REMOVED***
                    context.report(***REMOVED*** node, messageId: "forbidden" ***REMOVED***)
                ***REMOVED***
            ***REMOVED***,
        ***REMOVED***
    ***REMOVED***,
***REMOVED***
