// eslint-plugin-react/lib/rules/forbid-dom-props.js
/**
 * @fileoverview Forbid certain props on DOM Nodes
 * @author David Vázquez
 */

'use strict';

const docsUrl = require('../util/docsUrl');

// ------------------------------------------------------------------------------
// Constants
// ------------------------------------------------------------------------------

const DEFAULTS = [];

// ------------------------------------------------------------------------------
// Rule Definition
// ------------------------------------------------------------------------------

module.exports = ***REMOVED***
  meta: ***REMOVED***
    docs: ***REMOVED***
      description: 'Forbid certain props on DOM Nodes',
      category: 'Best Practices',
      recommended: false,
      url: docsUrl('forbid-dom-props')
    ***REMOVED***,

    schema: [***REMOVED***
      type: 'object',
      properties: ***REMOVED***
        forbid: ***REMOVED***
          type: 'array',
          items: ***REMOVED***
            type: 'string',
            minLength: 1
          ***REMOVED***,
          uniqueItems: true
        ***REMOVED***
      ***REMOVED***,
      additionalProperties: false
    ***REMOVED***]
  ***REMOVED***,

  create(context) ***REMOVED***
    function isForbidden(prop) ***REMOVED***
      const configuration = context.options[0] || ***REMOVED******REMOVED***;

      const forbid = configuration.forbid || DEFAULTS;
      return forbid.indexOf(prop) >= 0;
    ***REMOVED***

    return ***REMOVED***
      JSXAttribute(node) ***REMOVED***
        const tag = node.parent.name.name;
        if (!(tag && tag[0] !== tag[0].toUpperCase())) ***REMOVED***
          // This is a Component, not  a DOM node, so exit.
          return;
        ***REMOVED***

        const prop = node.name.name;

        if (!isForbidden(prop)) ***REMOVED***
          return;
        ***REMOVED***

        context.report(***REMOVED***
          node,
          message: `Prop \`$***REMOVED***prop***REMOVED***\` is forbidden on DOM Nodes`
        ***REMOVED***);
      ***REMOVED***
    ***REMOVED***;
  ***REMOVED***
***REMOVED***;
