// eslint-plugin-react/lib/rules/jsx-closing-tag-location.js
/**
 * @fileoverview Validate closing tag location in JSX
 * @author Ross Solomon
 */

'use strict';

const astUtil = require('../util/ast');
const docsUrl = require('../util/docsUrl');

// ------------------------------------------------------------------------------
// Rule Definition
// ------------------------------------------------------------------------------
module.exports = ***REMOVED***
  meta: ***REMOVED***
    docs: ***REMOVED***
      description: 'Validate closing tag location for multiline JSX',
      category: 'Stylistic Issues',
      recommended: false,
      url: docsUrl('jsx-closing-tag-location')
    ***REMOVED***,
    fixable: 'whitespace'
  ***REMOVED***,

  create(context) ***REMOVED***
    function handleClosingElement(node) ***REMOVED***
      if (!node.parent) ***REMOVED***
        return;
      ***REMOVED***

      const opening = node.parent.openingElement || node.parent.openingFragment;
      if (opening.loc.start.line === node.loc.start.line) ***REMOVED***
        return;
      ***REMOVED***

      if (opening.loc.start.column === node.loc.start.column) ***REMOVED***
        return;
      ***REMOVED***

      let message;
      if (!astUtil.isNodeFirstInLine(context, node)) ***REMOVED***
        message = 'Closing tag of a multiline JSX expression must be on its own line.';
      ***REMOVED*** else ***REMOVED***
        message = 'Expected closing tag to match indentation of opening.';
      ***REMOVED***

      context.report(***REMOVED***
        node,
        loc: node.loc,
        message,
        fix(fixer) ***REMOVED***
          const indent = Array(opening.loc.start.column + 1).join(' ');
          if (astUtil.isNodeFirstInLine(context, node)) ***REMOVED***
            return fixer.replaceTextRange(
              [node.range[0] - node.loc.start.column, node.range[0]],
              indent
            );
          ***REMOVED***

          return fixer.insertTextBefore(node, `\n$***REMOVED***indent***REMOVED***`);
        ***REMOVED***
      ***REMOVED***);
    ***REMOVED***

    return ***REMOVED***
      JSXClosingElement: handleClosingElement,
      JSXClosingFragment: handleClosingElement
    ***REMOVED***;
  ***REMOVED***
***REMOVED***;
