// eslint-plugin-react/lib/rules/jsx-indent.js
/**
 * @fileoverview Validate JSX indentation
 * @author Yannick Croissant

 * This rule has been ported and modified from eslint and nodeca.
 * @author Vitaly Puzrin
 * @author Gyandeep Singh
 * @copyright 2015 Vitaly Puzrin. All rights reserved.
 * @copyright 2015 Gyandeep Singh. All rights reserved.
 Copyright (C) 2014 by Vitaly Puzrin

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the 'Software'), to deal
 in the Software without restriction, including without limitation the rights
 to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 copies of the Software, and to permit persons to whom the Software is
 furnished to do so, subject to the following conditions:

 The above copyright notice and this permission notice shall be included in
 all copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED 'AS IS', WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 THE SOFTWARE.
 */

'use strict';

const astUtil = require('../util/ast');
const docsUrl = require('../util/docsUrl');

// ------------------------------------------------------------------------------
// Rule Definition
// ------------------------------------------------------------------------------
module.exports = ***REMOVED***
  meta: ***REMOVED***
    docs: ***REMOVED***
      description: 'Validate JSX indentation',
      category: 'Stylistic Issues',
      recommended: false,
      url: docsUrl('jsx-indent')
    ***REMOVED***,
    fixable: 'whitespace',
    schema: [***REMOVED***
      oneOf: [***REMOVED***
        enum: ['tab']
      ***REMOVED***, ***REMOVED***
        type: 'integer'
      ***REMOVED***]
    ***REMOVED***, ***REMOVED***
      type: 'object',
      properties: ***REMOVED***
        checkAttributes: ***REMOVED***
          type: 'boolean'
        ***REMOVED***,
        indentLogicalExpressions: ***REMOVED***
          type: 'boolean'
        ***REMOVED***
      ***REMOVED***,
      additionalProperties: false
    ***REMOVED***]
  ***REMOVED***,

  create(context) ***REMOVED***
    const MESSAGE = 'Expected indentation of ***REMOVED******REMOVED***needed***REMOVED******REMOVED*** ***REMOVED******REMOVED***type***REMOVED******REMOVED*** ***REMOVED******REMOVED***characters***REMOVED******REMOVED*** but found ***REMOVED******REMOVED***gotten***REMOVED******REMOVED***.';

    const extraColumnStart = 0;
    let indentType = 'space';
    let indentSize = 4;

    if (context.options.length) ***REMOVED***
      if (context.options[0] === 'tab') ***REMOVED***
        indentSize = 1;
        indentType = 'tab';
      ***REMOVED*** else if (typeof context.options[0] === 'number') ***REMOVED***
        indentSize = context.options[0];
        indentType = 'space';
      ***REMOVED***
    ***REMOVED***

    const indentChar = indentType === 'space' ? ' ' : '\t';
    const options = context.options[1] || ***REMOVED******REMOVED***;
    const checkAttributes = options.checkAttributes || false;
    const indentLogicalExpressions = options.indentLogicalExpressions || false;

    /**
     * Responsible for fixing the indentation issue fix
     * @param ***REMOVED***ASTNode***REMOVED*** node Node violating the indent rule
     * @param ***REMOVED***Number***REMOVED*** needed Expected indentation character count
     * @returns ***REMOVED***Function***REMOVED*** function to be executed by the fixer
     * @private
     */
    function getFixerFunction(node, needed) ***REMOVED***
      return function (fixer) ***REMOVED***
        const indent = Array(needed + 1).join(indentChar);
        return fixer.replaceTextRange(
          [node.range[0] - node.loc.start.column, node.range[0]],
          indent
        );
      ***REMOVED***;
    ***REMOVED***

    /**
     * Reports a given indent violation and properly pluralizes the message
     * @param ***REMOVED***ASTNode***REMOVED*** node Node violating the indent rule
     * @param ***REMOVED***Number***REMOVED*** needed Expected indentation character count
     * @param ***REMOVED***Number***REMOVED*** gotten Indentation character count in the actual node/code
     * @param ***REMOVED***Object***REMOVED*** [loc] Error line and column location
     */
    function report(node, needed, gotten, loc) ***REMOVED***
      const msgContext = ***REMOVED***
        needed,
        type: indentType,
        characters: needed === 1 ? 'character' : 'characters',
        gotten
      ***REMOVED***;

      if (loc) ***REMOVED***
        context.report(***REMOVED***
          node,
          loc,
          message: MESSAGE,
          data: msgContext,
          fix: getFixerFunction(node, needed)
        ***REMOVED***);
      ***REMOVED*** else ***REMOVED***
        context.report(***REMOVED***
          node,
          message: MESSAGE,
          data: msgContext,
          fix: getFixerFunction(node, needed)
        ***REMOVED***);
      ***REMOVED***
    ***REMOVED***

    /**
     * Get node indent
     * @param ***REMOVED***ASTNode***REMOVED*** node Node to examine
     * @param ***REMOVED***Boolean***REMOVED*** [byLastLine] get indent of node's last line
     * @param ***REMOVED***Boolean***REMOVED*** [excludeCommas] skip comma on start of line
     * @return ***REMOVED***Number***REMOVED*** Indent
     */
    function getNodeIndent(node, byLastLine, excludeCommas) ***REMOVED***
      byLastLine = byLastLine || false;
      excludeCommas = excludeCommas || false;

      let src = context.getSourceCode().getText(node, node.loc.start.column + extraColumnStart);
      const lines = src.split('\n');
      if (byLastLine) ***REMOVED***
        src = lines[lines.length - 1];
      ***REMOVED*** else ***REMOVED***
        src = lines[0];
      ***REMOVED***

      const skip = excludeCommas ? ',' : '';

      let regExp;
      if (indentType === 'space') ***REMOVED***
        regExp = new RegExp(`^[ $***REMOVED***skip***REMOVED***]+`);
      ***REMOVED*** else ***REMOVED***
        regExp = new RegExp(`^[\t$***REMOVED***skip***REMOVED***]+`);
      ***REMOVED***

      const indent = regExp.exec(src);
      return indent ? indent[0].length : 0;
    ***REMOVED***

    /**
     * Check if the node is the right member of a logical expression
     * @param ***REMOVED***ASTNode***REMOVED*** node The node to check
     * @return ***REMOVED***Boolean***REMOVED*** true if its the case, false if not
     */
    function isRightInLogicalExp(node) ***REMOVED***
      return (
        node.parent &&
        node.parent.parent &&
        node.parent.parent.type === 'LogicalExpression' &&
        node.parent.parent.right === node.parent &&
        !indentLogicalExpressions
      );
    ***REMOVED***

    /**
     * Check if the node is the alternate member of a conditional expression
     * @param ***REMOVED***ASTNode***REMOVED*** node The node to check
     * @return ***REMOVED***Boolean***REMOVED*** true if its the case, false if not
     */
    function isAlternateInConditionalExp(node) ***REMOVED***
      return (
        node.parent &&
        node.parent.parent &&
        node.parent.parent.type === 'ConditionalExpression' &&
        node.parent.parent.alternate === node.parent &&
        context.getSourceCode().getTokenBefore(node).value !== '('
      );
    ***REMOVED***

    /**
     * Check indent for nodes list
     * @param ***REMOVED***ASTNode***REMOVED*** node The node to check
     * @param ***REMOVED***Number***REMOVED*** indent needed indent
     * @param ***REMOVED***Boolean***REMOVED*** [excludeCommas] skip comma on start of line
     */
    function checkNodesIndent(node, indent, excludeCommas) ***REMOVED***
      const nodeIndent = getNodeIndent(node, false, excludeCommas);
      const isCorrectRightInLogicalExp = isRightInLogicalExp(node) && (nodeIndent - indent) === indentSize;
      const isCorrectAlternateInCondExp = isAlternateInConditionalExp(node) && (nodeIndent - indent) === 0;
      if (
        nodeIndent !== indent &&
        astUtil.isNodeFirstInLine(context, node) &&
        !isCorrectRightInLogicalExp &&
        !isCorrectAlternateInCondExp
      ) ***REMOVED***
        report(node, indent, nodeIndent);
      ***REMOVED***
    ***REMOVED***

    function handleOpeningElement(node) ***REMOVED***
      const sourceCode = context.getSourceCode();
      let prevToken = sourceCode.getTokenBefore(node);
      if (!prevToken) ***REMOVED***
        return;
      ***REMOVED***
      // Use the parent in a list or an array
      if (prevToken.type === 'JSXText' || prevToken.type === 'Punctuator' && prevToken.value === ',') ***REMOVED***
        prevToken = sourceCode.getNodeByRangeIndex(prevToken.range[0]);
        prevToken = prevToken.type === 'Literal' || prevToken.type === 'JSXText' ? prevToken.parent : prevToken;
      // Use the first non-punctuator token in a conditional expression
      ***REMOVED*** else if (prevToken.type === 'Punctuator' && prevToken.value === ':') ***REMOVED***
        do ***REMOVED***
          prevToken = sourceCode.getTokenBefore(prevToken);
        ***REMOVED*** while (prevToken.type === 'Punctuator' && prevToken.value !== '/');
        prevToken = sourceCode.getNodeByRangeIndex(prevToken.range[0]);
        while (prevToken.parent && prevToken.parent.type !== 'ConditionalExpression') ***REMOVED***
          prevToken = prevToken.parent;
        ***REMOVED***
      ***REMOVED***
      prevToken = prevToken.type === 'JSXExpressionContainer' ? prevToken.expression : prevToken;
      const parentElementIndent = getNodeIndent(prevToken);
      const indent = (
        prevToken.loc.start.line === node.loc.start.line ||
        isRightInLogicalExp(node) ||
        isAlternateInConditionalExp(node)
      ) ? 0 : indentSize;
      checkNodesIndent(node, parentElementIndent + indent);
    ***REMOVED***

    function handleClosingElement(node) ***REMOVED***
      if (!node.parent) ***REMOVED***
        return;
      ***REMOVED***
      const peerElementIndent = getNodeIndent(node.parent.openingElement || node.parent.openingFragment);
      checkNodesIndent(node, peerElementIndent);
    ***REMOVED***

    function handleAttribute(node) ***REMOVED***
      if (!checkAttributes || (!node.value || node.value.type !== 'JSXExpressionContainer')) ***REMOVED***
        return;
      ***REMOVED***
      const nameIndent = getNodeIndent(node.name);
      const lastToken = context.getSourceCode().getLastToken(node.value);
      const firstInLine = astUtil.getFirstNodeInLine(context, lastToken);
      const indent = node.name.loc.start.line === firstInLine.loc.start.line ? 0 : nameIndent;
      checkNodesIndent(firstInLine, indent);
    ***REMOVED***

    return ***REMOVED***
      JSXOpeningElement: handleOpeningElement,
      JSXOpeningFragment: handleOpeningElement,
      JSXClosingElement: handleClosingElement,
      JSXClosingFragment: handleClosingElement,
      JSXAttribute: handleAttribute,
      JSXExpressionContainer(node) ***REMOVED***
        if (!node.parent) ***REMOVED***
          return;
        ***REMOVED***
        const parentNodeIndent = getNodeIndent(node.parent);
        checkNodesIndent(node, parentNodeIndent + indentSize);
      ***REMOVED***
    ***REMOVED***;
  ***REMOVED***
***REMOVED***;
