// eslint-plugin-react/lib/rules/jsx-uses-vars.js
/**
 * @fileoverview Prevent variables used in JSX to be marked as unused
 * @author Yannick Croissant
 */

'use strict';

const docsUrl = require('../util/docsUrl');

// ------------------------------------------------------------------------------
// Rule Definition
// ------------------------------------------------------------------------------

module.exports = ***REMOVED***
  meta: ***REMOVED***
    docs: ***REMOVED***
      description: 'Prevent variables used in JSX to be marked as unused',
      category: 'Best Practices',
      recommended: true,
      url: docsUrl('jsx-uses-vars')
    ***REMOVED***,
    schema: []
  ***REMOVED***,

  create(context) ***REMOVED***
    return ***REMOVED***
      JSXOpeningElement(node) ***REMOVED***
        let name;
        if (node.name.namespace && node.name.namespace.name) ***REMOVED***
          // <Foo:Bar>
          name = node.name.namespace.name;
        ***REMOVED*** else if (node.name.name) ***REMOVED***
          // <Foo>
          name = node.name.name;
        ***REMOVED*** else if (node.name.object) ***REMOVED***
          // <Foo...Bar>
          let parent = node.name.object;
          while (parent.object) ***REMOVED***
            parent = parent.object;
          ***REMOVED***
          name = parent.name;
        ***REMOVED*** else ***REMOVED***
          return;
        ***REMOVED***

        context.markVariableAsUsed(name);
      ***REMOVED***

    ***REMOVED***;
  ***REMOVED***
***REMOVED***;
