// eslint-plugin-react/lib/rules/jsx-uses-react.js
/**
 * @fileoverview Prevent React to be marked as unused
 * @author Glen Mailer
 */

'use strict';

const pragmaUtil = require('../util/pragma');
const docsUrl = require('../util/docsUrl');

// ------------------------------------------------------------------------------
// Rule Definition
// ------------------------------------------------------------------------------

module.exports = ***REMOVED***
  meta: ***REMOVED***
    docs: ***REMOVED***
      description: 'Prevent React to be marked as unused',
      category: 'Best Practices',
      recommended: true,
      url: docsUrl('jsx-uses-react')
    ***REMOVED***,
    schema: []
  ***REMOVED***,

  create(context) ***REMOVED***
    const pragma = pragmaUtil.getFromContext(context);

    function handleOpeningElement() ***REMOVED***
      context.markVariableAsUsed(pragma);
    ***REMOVED***
    // --------------------------------------------------------------------------
    // Public
    // --------------------------------------------------------------------------

    return ***REMOVED***
      JSXOpeningElement: handleOpeningElement,
      JSXOpeningFragment: handleOpeningElement
    ***REMOVED***;
  ***REMOVED***
***REMOVED***;
