// eslint-plugin-react/lib/rules/jsx-space-before-closing.js
/**
 * @fileoverview Validate spacing before closing bracket in JSX.
 * @author ryym
 * @deprecated
 */

'use strict';

const getTokenBeforeClosingBracket = require('../util/getTokenBeforeClosingBracket');
const docsUrl = require('../util/docsUrl');
const log = require('../util/log');

let isWarnedForDeprecation = false;

// ------------------------------------------------------------------------------
// Rule Definition
// ------------------------------------------------------------------------------

module.exports = ***REMOVED***
  meta: ***REMOVED***
    deprecated: true,
    docs: ***REMOVED***
      description: 'Validate spacing before closing bracket in JSX',
      category: 'Stylistic Issues',
      recommended: false,
      url: docsUrl('jsx-space-before-closing')
    ***REMOVED***,
    fixable: 'code',

    schema: [***REMOVED***
      enum: ['always', 'never']
    ***REMOVED***]
  ***REMOVED***,

  create(context) ***REMOVED***
    const configuration = context.options[0] || 'always';

    const NEVER_MESSAGE = 'A space is forbidden before closing bracket';
    const ALWAYS_MESSAGE = 'A space is required before closing bracket';

    // --------------------------------------------------------------------------
    // Public
    // --------------------------------------------------------------------------

    return ***REMOVED***
      JSXOpeningElement(node) ***REMOVED***
        if (!node.selfClosing) ***REMOVED***
          return;
        ***REMOVED***

        const sourceCode = context.getSourceCode();

        const leftToken = getTokenBeforeClosingBracket(node);
        const closingSlash = sourceCode.getTokenAfter(leftToken);

        if (leftToken.loc.end.line !== closingSlash.loc.start.line) ***REMOVED***
          return;
        ***REMOVED***

        if (configuration === 'always' && !sourceCode.isSpaceBetweenTokens(leftToken, closingSlash)) ***REMOVED***
          context.report(***REMOVED***
            loc: closingSlash.loc.start,
            message: ALWAYS_MESSAGE,
            fix(fixer) ***REMOVED***
              return fixer.insertTextBefore(closingSlash, ' ');
            ***REMOVED***
          ***REMOVED***);
        ***REMOVED*** else if (configuration === 'never' && sourceCode.isSpaceBetweenTokens(leftToken, closingSlash)) ***REMOVED***
          context.report(***REMOVED***
            loc: closingSlash.loc.start,
            message: NEVER_MESSAGE,
            fix(fixer) ***REMOVED***
              const previousToken = sourceCode.getTokenBefore(closingSlash);
              return fixer.removeRange([previousToken.range[1], closingSlash.range[0]]);
            ***REMOVED***
          ***REMOVED***);
        ***REMOVED***
      ***REMOVED***,

      Program() ***REMOVED***
        if (isWarnedForDeprecation) ***REMOVED***
          return;
        ***REMOVED***

        log('The react/jsx-space-before-closing rule is deprecated. ' +
            'Please use the react/jsx-tag-spacing rule with the ' +
            '"beforeSelfClosing" option instead.');
        isWarnedForDeprecation = true;
      ***REMOVED***
    ***REMOVED***;
  ***REMOVED***
***REMOVED***;
