// eslint-plugin-react/lib/rules/no-find-dom-node.js
/**
 * @fileoverview Prevent usage of findDOMNode
 * @author Yannick Croissant
 */

'use strict';

const docsUrl = require('../util/docsUrl');

// ------------------------------------------------------------------------------
// Rule Definition
// ------------------------------------------------------------------------------

module.exports = ***REMOVED***
  meta: ***REMOVED***
    docs: ***REMOVED***
      description: 'Prevent usage of findDOMNode',
      category: 'Best Practices',
      recommended: true,
      url: docsUrl('no-find-dom-node')
    ***REMOVED***,
    schema: []
  ***REMOVED***,

  create(context) ***REMOVED***
    // --------------------------------------------------------------------------
    // Public
    // --------------------------------------------------------------------------

    return ***REMOVED***

      CallExpression(node) ***REMOVED***
        const callee = node.callee;

        const isfindDOMNode = (callee.name === 'findDOMNode') ||
          (callee.property && callee.property.name === 'findDOMNode');
        if (!isfindDOMNode) ***REMOVED***
          return;
        ***REMOVED***

        context.report(***REMOVED***
          node: callee,
          message: 'Do not use findDOMNode'
        ***REMOVED***);
      ***REMOVED***
    ***REMOVED***;
  ***REMOVED***
***REMOVED***;
