// eslint-plugin-react/lib/rules/no-is-mounted.js
/**
 * @fileoverview Prevent usage of isMounted
 * @author Joe Lencioni
 */

'use strict';

const docsUrl = require('../util/docsUrl');

// ------------------------------------------------------------------------------
// Rule Definition
// ------------------------------------------------------------------------------

module.exports = ***REMOVED***
  meta: ***REMOVED***
    docs: ***REMOVED***
      description: 'Prevent usage of isMounted',
      category: 'Best Practices',
      recommended: true,
      url: docsUrl('no-is-mounted')
    ***REMOVED***,
    schema: []
  ***REMOVED***,

  create(context) ***REMOVED***
    // --------------------------------------------------------------------------
    // Public
    // --------------------------------------------------------------------------

    return ***REMOVED***

      CallExpression(node) ***REMOVED***
        const callee = node.callee;
        if (callee.type !== 'MemberExpression') ***REMOVED***
          return;
        ***REMOVED***
        if (callee.object.type !== 'ThisExpression' || callee.property.name !== 'isMounted') ***REMOVED***
          return;
        ***REMOVED***
        const ancestors = context.getAncestors(callee);
        for (let i = 0, j = ancestors.length; i < j; i++) ***REMOVED***
          if (ancestors[i].type === 'Property' || ancestors[i].type === 'MethodDefinition') ***REMOVED***
            context.report(***REMOVED***
              node: callee,
              message: 'Do not use isMounted'
            ***REMOVED***);
            break;
          ***REMOVED***
        ***REMOVED***
      ***REMOVED***
    ***REMOVED***;
  ***REMOVED***
***REMOVED***;
