// eslint-plugin-react/lib/rules/no-this-in-sfc.js
/**
 * @fileoverview Report "this" being used in stateless functional components.
 */

'use strict';

const Components = require('../util/Components');
const docsUrl = require('../util/docsUrl');

// ------------------------------------------------------------------------------
// Constants
// ------------------------------------------------------------------------------

const ERROR_MESSAGE = 'Stateless functional components should not use this';

// ------------------------------------------------------------------------------
// Rule Definition
// ------------------------------------------------------------------------------

module.exports = ***REMOVED***
  meta: ***REMOVED***
    docs: ***REMOVED***
      description: 'Report "this" being used in stateless components',
      category: 'Possible Errors',
      recommended: false,
      url: docsUrl('no-this-in-sfc')
    ***REMOVED***,
    schema: []
  ***REMOVED***,

  create: Components.detect((context, components, utils) => (***REMOVED***
    MemberExpression(node) ***REMOVED***
      if (node.object.type === 'ThisExpression') ***REMOVED***
        const component = components.get(utils.getParentStatelessComponent());
        if (!component || component.node && component.node.parent && component.node.parent.type === 'Property') ***REMOVED***
          return;
        ***REMOVED***
        context.report(***REMOVED***
          node,
          message: ERROR_MESSAGE
        ***REMOVED***);
      ***REMOVED***
    ***REMOVED***
  ***REMOVED***))
***REMOVED***;
