// eslint-plugin-react/lib/rules/no-unknown-property.js
/**
 * @fileoverview Prevent usage of unknown DOM property
 * @author Yannick Croissant
 */

'use strict';

const docsUrl = require('../util/docsUrl');

// ------------------------------------------------------------------------------
// Constants
// ------------------------------------------------------------------------------

const DEFAULTS = ***REMOVED***
  ignore: []
***REMOVED***;

const UNKNOWN_MESSAGE = 'Unknown property \'***REMOVED******REMOVED***name***REMOVED******REMOVED***\' found, use \'***REMOVED******REMOVED***standardName***REMOVED******REMOVED***\' instead';
const WRONG_TAG_MESSAGE = 'Invalid property \'***REMOVED******REMOVED***name***REMOVED******REMOVED***\' found on tag \'***REMOVED******REMOVED***tagName***REMOVED******REMOVED***\', but it is only allowed on: ***REMOVED******REMOVED***allowedTags***REMOVED******REMOVED***';

const DOM_ATTRIBUTE_NAMES = ***REMOVED***
  'accept-charset': 'acceptCharset',
  class: 'className',
  for: 'htmlFor',
  'http-equiv': 'httpEquiv',
  crossorigin: 'crossOrigin'
***REMOVED***;

const ATTRIBUTE_TAGS_MAP = ***REMOVED***
  crossOrigin: ['script', 'img', 'video', 'audio', 'link']
***REMOVED***;

const SVGDOM_ATTRIBUTE_NAMES = ***REMOVED***
  'accent-height': 'accentHeight',
  'alignment-baseline': 'alignmentBaseline',
  'arabic-form': 'arabicForm',
  'baseline-shift': 'baselineShift',
  'cap-height': 'capHeight',
  'clip-path': 'clipPath',
  'clip-rule': 'clipRule',
  'color-interpolation': 'colorInterpolation',
  'color-interpolation-filters': 'colorInterpolationFilters',
  'color-profile': 'colorProfile',
  'color-rendering': 'colorRendering',
  'dominant-baseline': 'dominantBaseline',
  'enable-background': 'enableBackground',
  'fill-opacity': 'fillOpacity',
  'fill-rule': 'fillRule',
  'flood-color': 'floodColor',
  'flood-opacity': 'floodOpacity',
  'font-family': 'fontFamily',
  'font-size': 'fontSize',
  'font-size-adjust': 'fontSizeAdjust',
  'font-stretch': 'fontStretch',
  'font-style': 'fontStyle',
  'font-variant': 'fontVariant',
  'font-weight': 'fontWeight',
  'glyph-name': 'glyphName',
  'glyph-orientation-horizontal': 'glyphOrientationHorizontal',
  'glyph-orientation-vertical': 'glyphOrientationVertical',
  'horiz-adv-x': 'horizAdvX',
  'horiz-origin-x': 'horizOriginX',
  'image-rendering': 'imageRendering',
  'letter-spacing': 'letterSpacing',
  'lighting-color': 'lightingColor',
  'marker-end': 'markerEnd',
  'marker-mid': 'markerMid',
  'marker-start': 'markerStart',
  'overline-position': 'overlinePosition',
  'overline-thickness': 'overlineThickness',
  'paint-order': 'paintOrder',
  'panose-1': 'panose1',
  'pointer-events': 'pointerEvents',
  'rendering-intent': 'renderingIntent',
  'shape-rendering': 'shapeRendering',
  'stop-color': 'stopColor',
  'stop-opacity': 'stopOpacity',
  'strikethrough-position': 'strikethroughPosition',
  'strikethrough-thickness': 'strikethroughThickness',
  'stroke-dasharray': 'strokeDasharray',
  'stroke-dashoffset': 'strokeDashoffset',
  'stroke-linecap': 'strokeLinecap',
  'stroke-linejoin': 'strokeLinejoin',
  'stroke-miterlimit': 'strokeMiterlimit',
  'stroke-opacity': 'strokeOpacity',
  'stroke-width': 'strokeWidth',
  'text-anchor': 'textAnchor',
  'text-decoration': 'textDecoration',
  'text-rendering': 'textRendering',
  'underline-position': 'underlinePosition',
  'underline-thickness': 'underlineThickness',
  'unicode-bidi': 'unicodeBidi',
  'unicode-range': 'unicodeRange',
  'units-per-em': 'unitsPerEm',
  'v-alphabetic': 'vAlphabetic',
  'v-hanging': 'vHanging',
  'v-ideographic': 'vIdeographic',
  'v-mathematical': 'vMathematical',
  'vector-effect': 'vectorEffect',
  'vert-adv-y': 'vertAdvY',
  'vert-origin-x': 'vertOriginX',
  'vert-origin-y': 'vertOriginY',
  'word-spacing': 'wordSpacing',
  'writing-mode': 'writingMode',
  'x-height': 'xHeight',
  'xlink:actuate': 'xlinkActuate',
  'xlink:arcrole': 'xlinkArcrole',
  'xlink:href': 'xlinkHref',
  'xlink:role': 'xlinkRole',
  'xlink:show': 'xlinkShow',
  'xlink:title': 'xlinkTitle',
  'xlink:type': 'xlinkType',
  'xml:base': 'xmlBase',
  'xml:lang': 'xmlLang',
  'xml:space': 'xmlSpace'
***REMOVED***;

const DOM_PROPERTY_NAMES = [
  // Standard
  'acceptCharset', 'accessKey', 'allowFullScreen', 'allowTransparency', 'autoComplete', 'autoFocus', 'autoPlay',
  'cellPadding', 'cellSpacing', 'classID', 'className', 'colSpan', 'contentEditable', 'contextMenu',
  'dateTime', 'encType', 'formAction', 'formEncType', 'formMethod', 'formNoValidate', 'formTarget',
  'frameBorder', 'hrefLang', 'htmlFor', 'httpEquiv', 'inputMode', 'keyParams', 'keyType', 'marginHeight', 'marginWidth',
  'maxLength', 'mediaGroup', 'minLength', 'noValidate', 'onAnimationEnd', 'onAnimationIteration', 'onAnimationStart',
  'onBlur', 'onChange', 'onClick', 'onContextMenu', 'onCopy', 'onCompositionEnd', 'onCompositionStart',
  'onCompositionUpdate', 'onCut', 'onDoubleClick', 'onDrag', 'onDragEnd', 'onDragEnter', 'onDragExit', 'onDragLeave',
  'onError', 'onFocus', 'onInput', 'onKeyDown', 'onKeyPress', 'onKeyUp', 'onLoad', 'onWheel', 'onDragOver',
  'onDragStart', 'onDrop', 'onMouseDown', 'onMouseEnter', 'onMouseLeave', 'onMouseMove', 'onMouseOut', 'onMouseOver',
  'onMouseUp', 'onPaste', 'onScroll', 'onSelect', 'onSubmit', 'onTransitionEnd', 'radioGroup', 'readOnly', 'rowSpan',
  'spellCheck', 'srcDoc', 'srcLang', 'srcSet', 'tabIndex', 'useMap',
  // Non standard
  'autoCapitalize', 'autoCorrect',
  'autoSave',
  'itemProp', 'itemScope', 'itemType', 'itemRef', 'itemID'
];

// ------------------------------------------------------------------------------
// Helpers
// ------------------------------------------------------------------------------

/**
 * Checks if a node matches the JSX tag convention.
 * @param ***REMOVED***Object***REMOVED*** node - JSX element being tested.
 * @returns ***REMOVED***boolean***REMOVED*** Whether or not the node name match the JSX tag convention.
 */
const tagConvention = /^[a-z][^-]*$/;
function isTagName(node) ***REMOVED***
  if (tagConvention.test(node.parent.name.name)) ***REMOVED***
    // http://www.w3.org/TR/custom-elements/#type-extension-semantics
    return !node.parent.attributes.some(attrNode => (
      attrNode.type === 'JSXAttribute' &&
        attrNode.name.type === 'JSXIdentifier' &&
        attrNode.name.name === 'is'
    ));
  ***REMOVED***
  return false;
***REMOVED***

/**
 * Extracts the tag name for the JSXAttribute
 * @param ***REMOVED***JSXAttribute***REMOVED*** node - JSXAttribute being tested.
 * @returns ***REMOVED***String|null***REMOVED*** tag name
 */
function getTagName(node) ***REMOVED***
  if (node && node.parent && node.parent.name && node.parent.name) ***REMOVED***
    return node.parent.name.name;
  ***REMOVED***
  return null;
***REMOVED***

/**
 * Test wether the tag name for the JSXAttribute is
 * something like <Foo.bar />
 * @param ***REMOVED***JSXAttribute***REMOVED*** node - JSXAttribute being tested.
 * @returns ***REMOVED***Boolean***REMOVED*** result
 */
function tagNameHasDot(node) ***REMOVED***
  return !!(
    node.parent &&
    node.parent.name &&
    node.parent.name.type === 'JSXMemberExpression'
  );
***REMOVED***

/**
 * Get the standard name of the attribute.
 * @param ***REMOVED***String***REMOVED*** name - Name of the attribute.
 * @returns ***REMOVED***String***REMOVED*** The standard name of the attribute.
 */
function getStandardName(name) ***REMOVED***
  if (DOM_ATTRIBUTE_NAMES[name]) ***REMOVED***
    return DOM_ATTRIBUTE_NAMES[name];
  ***REMOVED***
  if (SVGDOM_ATTRIBUTE_NAMES[name]) ***REMOVED***
    return SVGDOM_ATTRIBUTE_NAMES[name];
  ***REMOVED***
  let i = -1;
  const found = DOM_PROPERTY_NAMES.some((element, index) => ***REMOVED***
    i = index;
    return element.toLowerCase() === name;
  ***REMOVED***);
  return found ? DOM_PROPERTY_NAMES[i] : null;
***REMOVED***

// ------------------------------------------------------------------------------
// Rule Definition
// ------------------------------------------------------------------------------

module.exports = ***REMOVED***
  meta: ***REMOVED***
    docs: ***REMOVED***
      description: 'Prevent usage of unknown DOM property',
      category: 'Possible Errors',
      recommended: true,
      url: docsUrl('no-unknown-property')
    ***REMOVED***,
    fixable: 'code',

    schema: [***REMOVED***
      type: 'object',
      properties: ***REMOVED***
        ignore: ***REMOVED***
          type: 'array',
          items: ***REMOVED***
            type: 'string'
          ***REMOVED***
        ***REMOVED***
      ***REMOVED***,
      additionalProperties: false
    ***REMOVED***]
  ***REMOVED***,

  create(context) ***REMOVED***
    function getIgnoreConfig() ***REMOVED***
      return context.options[0] && context.options[0].ignore || DEFAULTS.ignore;
    ***REMOVED***

    return ***REMOVED***
      JSXAttribute(node) ***REMOVED***
        const ignoreNames = getIgnoreConfig();
        const name = context.getSourceCode().getText(node.name);
        if (ignoreNames.indexOf(name) >= 0) ***REMOVED***
          return;
        ***REMOVED***

        // Ignore tags like <Foo.bar />
        if (tagNameHasDot(node)) ***REMOVED***
          return;
        ***REMOVED***

        const tagName = getTagName(node);
        const allowedTags = ATTRIBUTE_TAGS_MAP[name];
        if (tagName && allowedTags && /[^A-Z]/.test(tagName.charAt(0)) && allowedTags.indexOf(tagName) === -1) ***REMOVED***
          context.report(***REMOVED***
            node,
            message: WRONG_TAG_MESSAGE,
            data: ***REMOVED***
              name,
              tagName,
              allowedTags: allowedTags.join(', ')
            ***REMOVED***
          ***REMOVED***);
        ***REMOVED***

        const standardName = getStandardName(name);
        if (!isTagName(node) || !standardName) ***REMOVED***
          return;
        ***REMOVED***
        context.report(***REMOVED***
          node,
          message: UNKNOWN_MESSAGE,
          data: ***REMOVED***
            name,
            standardName
          ***REMOVED***,
          fix(fixer) ***REMOVED***
            return fixer.replaceText(node.name, standardName);
          ***REMOVED***
        ***REMOVED***);
      ***REMOVED***
    ***REMOVED***;
  ***REMOVED***
***REMOVED***;
