// eslint-plugin-react/lib/rules/prefer-es6-class.js
/**
 * @fileoverview Enforce ES5 or ES6 class for React Components
 * @author Dan Hamilton
 */

'use strict';

const Components = require('../util/Components');
const docsUrl = require('../util/docsUrl');

// ------------------------------------------------------------------------------
// Rule Definition
// ------------------------------------------------------------------------------

module.exports = ***REMOVED***
  meta: ***REMOVED***
    docs: ***REMOVED***
      description: 'Enforce ES5 or ES6 class for React Components',
      category: 'Stylistic Issues',
      recommended: false,
      url: docsUrl('prefer-es6-class')
    ***REMOVED***,

    schema: [***REMOVED***
      enum: ['always', 'never']
    ***REMOVED***]
  ***REMOVED***,

  create: Components.detect((context, components, utils) => ***REMOVED***
    const configuration = context.options[0] || 'always';

    return ***REMOVED***
      ObjectExpression(node) ***REMOVED***
        if (utils.isES5Component(node) && configuration === 'always') ***REMOVED***
          context.report(***REMOVED***
            node,
            message: 'Component should use es6 class instead of createClass'
          ***REMOVED***);
        ***REMOVED***
      ***REMOVED***,
      ClassDeclaration(node) ***REMOVED***
        if (utils.isES6Component(node) && configuration === 'never') ***REMOVED***
          context.report(***REMOVED***
            node,
            message: 'Component should use createClass instead of es6 class'
          ***REMOVED***);
        ***REMOVED***
      ***REMOVED***
    ***REMOVED***;
  ***REMOVED***)
***REMOVED***;
