// eslint-plugin-react/lib/rules/state-in-constructor.js
/**
 * @fileoverview Enforce the state initialization style to be either in a constructor or with a class property
 * @author Kanitkorn Sujautra
 */

'use strict';

const Components = require('../util/Components');
const docsUrl = require('../util/docsUrl');

// ------------------------------------------------------------------------------
// Rule Definition
// ------------------------------------------------------------------------------

module.exports = ***REMOVED***
  meta: ***REMOVED***
    docs: ***REMOVED***
      description: 'State initialization in an ES6 class component should be in a constructor',
      category: 'Stylistic Issues',
      recommended: false,
      url: docsUrl('state-in-constructor')
    ***REMOVED***,
    schema: [***REMOVED***
      enum: ['always', 'never']
    ***REMOVED***]
  ***REMOVED***,

  create: Components.detect((context, components, utils) => ***REMOVED***
    const option = context.options[0] || 'always';
    return ***REMOVED***
      ClassProperty(node) ***REMOVED***
        if (
          option === 'always' &&
          !node.static &&
          node.key.name === 'state' &&
          utils.getParentES6Component()
        ) ***REMOVED***
          context.report(***REMOVED***
            node,
            message: 'State initialization should be in a constructor'
          ***REMOVED***);
        ***REMOVED***
      ***REMOVED***,
      AssignmentExpression(node) ***REMOVED***
        if (
          option === 'never' &&
          utils.isStateMemberExpression(node.left) &&
          utils.inConstructor() &&
          utils.getParentES6Component()
        ) ***REMOVED***
          context.report(***REMOVED***
            node,
            message: 'State initialization should be in a class property'
          ***REMOVED***);
        ***REMOVED***
      ***REMOVED***
    ***REMOVED***;
  ***REMOVED***)
***REMOVED***;
