// snekfetch/src/browser.js
function buildRequest(method, url) ***REMOVED***
  return ***REMOVED***
    method,
    path: url,
    redirect: this.options.followRedirects ? 'follow' : 'manual',
    headers: ***REMOVED******REMOVED***,
    setHeader(name, value) ***REMOVED***
      this.headers[name.toLowerCase()] = value;
    ***REMOVED***,
    getHeader(name) ***REMOVED***
      return this.headers[name.toLowerCase()];
    ***REMOVED***,
  ***REMOVED***;
***REMOVED***

function finalizeRequest() ***REMOVED***
  this._finalizeRequest();
  if (this.data)
    this.request.body = this.data;
  return window.fetch(this.request.path, this.request)
    .then((r) => r.text().then((t) => ***REMOVED***
      const headers = ***REMOVED******REMOVED***;
      for (const [k, v] of r.headers.entries())
        headers[k.toLowerCase()] = v;
      return ***REMOVED*** response: r, raw: t, headers ***REMOVED***;
    ***REMOVED***));
***REMOVED***

module.exports = ***REMOVED***
  buildRequest, finalizeRequest,
  shouldSendRaw: () => false,
  METHODS: ['GET', 'HEAD', 'POST', 'PUT', 'DELETE', 'CONNECT', 'OPTIONS', 'PATCH'],
  STATUS_CODES: ***REMOVED******REMOVED***,
  Extension: Object,
  FormData: window.FormData,
***REMOVED***;
