// snekfetch/src/index.js
const browser = typeof window !== 'undefined';
const querystring = require('querystring');
const transport = browser ? require('./browser') : require('./node');

/**
 * Snekfetch
 * @extends Stream.Readable
 * @extends Promise
 */
class Snekfetch extends transport.Extension ***REMOVED***
  /**
   * Options to pass to the Snekfetch constructor
   * @typedef ***REMOVED***object***REMOVED*** SnekfetchOptions
   * @memberof Snekfetch
   * @property ***REMOVED***object***REMOVED*** [headers] Headers to initialize the request with
   * @property ***REMOVED***object|string|Buffer***REMOVED*** [data] Data to initialize the request with
   * @property ***REMOVED***string|Object***REMOVED*** [query] Query to intialize the request with
   * @property ***REMOVED***boolean***REMOVED*** [followRedirects=true] If the request should follow redirects
   * @property ***REMOVED***object***REMOVED*** [qs=querystring] Querystring module to use, any object providing
   * `stringify` and `parse` for querystrings
   * @property ***REMOVED***number***REMOVED*** [version = 1] The http version to use [1 or 2]
   * @property ***REMOVED***external:Agent***REMOVED*** [agent] Whether to use an http agent
   */

  /**
   * Create a request.
   * Usually you'll want to do `Snekfetch#method(url [, options])` instead of
   * `new Snekfetch(method, url [, options])`
   * @param ***REMOVED***string***REMOVED*** method HTTP method
   * @param ***REMOVED***string***REMOVED*** url URL
   * @param ***REMOVED***SnekfetchOptions***REMOVED*** [opts] Options
   */
  constructor(method, url, opts = ***REMOVED******REMOVED***) ***REMOVED***
    super();
    this.options = Object.assign(***REMOVED*** version: 1, qs: querystring, followRedirects: true ***REMOVED***, opts);
    this.request = transport.buildRequest.call(this, method, url, opts);
    if (opts.headers)
      this.set(opts.headers);
    if (opts.query)
      this.query(opts.query);
    if (opts.data)
      this.send(opts.data);
  ***REMOVED***

  /**
   * Add a query param to the request
   * @param ***REMOVED***string|Object***REMOVED*** name Name of query param or object to add to query
   * @param ***REMOVED***string***REMOVED*** [value] If name is a string value, this will be the value of the query param
   * @returns ***REMOVED***Snekfetch***REMOVED*** This request
   */
  query(name, value) ***REMOVED***
    if (!this.request.query)
      this.request.query = ***REMOVED******REMOVED***;
    if (name !== null && typeof name === 'object') ***REMOVED***
      for (const [k, v] of Object.entries(name))
        this.query(k, v);
    ***REMOVED*** else ***REMOVED***
      this.request.query[name] = value;
    ***REMOVED***

    return this;
  ***REMOVED***

  /**
   * Add a header to the request
   * @param ***REMOVED***string|Object***REMOVED*** name Name of query param or object to add to headers
   * @param ***REMOVED***string***REMOVED*** [value] If name is a string value, this will be the value of the header
   * @returns ***REMOVED***Snekfetch***REMOVED*** This request
   */
  set(name, value) ***REMOVED***
    if (name !== null && typeof name === 'object') ***REMOVED***
      for (const key of Object.keys(name))
        this.set(key, name[key]);
    ***REMOVED*** else ***REMOVED***
      this.request.setHeader(name, value);
    ***REMOVED***

    return this;
  ***REMOVED***

  /**
   * Attach a form data object
   * @param ***REMOVED***string***REMOVED*** name Name of the form attachment
   * @param ***REMOVED***string|Object|Buffer***REMOVED*** data Data for the attachment
   * @param ***REMOVED***string***REMOVED*** [filename] Optional filename if form attachment name needs to be overridden
   * @returns ***REMOVED***Snekfetch***REMOVED*** This request
   */
  attach(...args) ***REMOVED***
    const form = this.data instanceof transport.FormData ? this.data : this.data = new transport.FormData();
    if (typeof args[0] === 'object') ***REMOVED***
      for (const [k, v] of Object.entries(args[0]))
        this.attach(k, v);
    ***REMOVED*** else ***REMOVED***
      form.append(...args);
    ***REMOVED***

    return this;
  ***REMOVED***

  /**
   * Send data with the request
   * @param ***REMOVED***string|Buffer|Object***REMOVED*** data Data to send
   * @returns ***REMOVED***Snekfetch***REMOVED*** This request
   */
  send(data) ***REMOVED***
    if (data instanceof transport.FormData || transport.shouldSendRaw(data)) ***REMOVED***
      this.data = data;
    ***REMOVED*** else if (data !== null && typeof data === 'object') ***REMOVED***
      const header = this.request.getHeader('content-type');
      let serialize;
      if (header) ***REMOVED***
        if (header.includes('json'))
          serialize = JSON.stringify;
        else if (header.includes('urlencoded'))
          serialize = this.options.qs.stringify;
      ***REMOVED*** else ***REMOVED***
        this.set('Content-Type', 'application/json');
        serialize = JSON.stringify;
      ***REMOVED***
      this.data = serialize(data);
    ***REMOVED*** else ***REMOVED***
      this.data = data;
    ***REMOVED***
    return this;
  ***REMOVED***

  then(resolver, rejector) ***REMOVED***
    if (this._response)
      return this._response.then(resolver, rejector);
    // eslint-disable-next-line no-return-assign
    return this._response = transport.finalizeRequest.call(this)
      .then((***REMOVED*** response, raw, redirect, headers ***REMOVED***) => ***REMOVED***
        if (redirect) ***REMOVED***
          let method = this.request.method;
          if ([301, 302].includes(response.statusCode)) ***REMOVED***
            if (method !== 'HEAD')
              method = 'GET';
            this.data = null;
          ***REMOVED*** else if (response.statusCode === 303) ***REMOVED***
            method = 'GET';
          ***REMOVED***

          const redirectHeaders = this.request.getHeaders();
          delete redirectHeaders.host;
          return new Snekfetch(method, redirect, ***REMOVED***
            data: this.data,
            headers: redirectHeaders,
            version: this.options.version,
          ***REMOVED***);
        ***REMOVED***

        const statusCode = response.statusCode || response.status;
        // forgive me :(
        const self = this; // eslint-disable-line consistent-this
        /**
         * Response from Snekfetch
         * @typedef ***REMOVED***Object***REMOVED*** SnekfetchResponse
         * @memberof Snekfetch
         * @prop ***REMOVED***HTTP.Request***REMOVED*** request
         * @prop ***REMOVED***?string|object|Buffer***REMOVED*** body Processed response body
         * @prop ***REMOVED***string***REMOVED*** text Raw response body
         * @prop ***REMOVED***boolean***REMOVED*** ok If the response code is >= 200 and < 300
         * @prop ***REMOVED***number***REMOVED*** status HTTP status code
         * @prop ***REMOVED***string***REMOVED*** statusText Human readable HTTP status
         */
        const res = ***REMOVED***
          request: this.request,
          get body() ***REMOVED***
            delete res.body;
            const type = this.headers['content-type'];
            if (type && type.includes('application/json')) ***REMOVED***
              try ***REMOVED***
                res.body = JSON.parse(res.text);
              ***REMOVED*** catch (err) ***REMOVED***
                res.body = res.text;
              ***REMOVED***
            ***REMOVED*** else if (type && type.includes('application/x-www-form-urlencoded')) ***REMOVED***
              res.body = self.options.qs.parse(res.text);
            ***REMOVED*** else ***REMOVED***
              res.body = raw;
            ***REMOVED***

            return res.body;
          ***REMOVED***,
          text: raw.toString(),
          ok: statusCode >= 200 && statusCode < 400,
          headers: headers || response.headers,
          status: statusCode,
          statusText: response.statusText || transport.STATUS_CODES[response.statusCode],
        ***REMOVED***;

        if (res.ok) ***REMOVED***
          return res;
        ***REMOVED*** else ***REMOVED***
          const err = new Error(`$***REMOVED***res.status***REMOVED*** $***REMOVED***res.statusText***REMOVED***`.trim());
          Object.assign(err, res);
          return Promise.reject(err);
        ***REMOVED***
      ***REMOVED***)
      .then(resolver, rejector);
  ***REMOVED***

  catch(rejector) ***REMOVED***
    return this.then(null, rejector);
  ***REMOVED***

  /**
   * End the request
   * @param ***REMOVED***Function***REMOVED*** [cb] Optional callback to handle the response
   * @returns ***REMOVED***Promise***REMOVED*** This request
   */
  end(cb) ***REMOVED***
    return this.then(
      (res) => cb ? cb(null, res) : res,
      (err) => cb ? cb(err, err.status ? err : null) : Promise.reject(err)
    );
  ***REMOVED***

  _finalizeRequest() ***REMOVED***
    if (!this.request)
      return;

    if (this.request.method !== 'HEAD')
      this.set('Accept-Encoding', 'gzip, deflate');
    if (this.data && this.data.getBoundary)
      this.set('Content-Type', `multipart/form-data; boundary=$***REMOVED***this.data.getBoundary()***REMOVED***`);

    if (this.request.query) ***REMOVED***
      const [path, query] = this.request.path.split('?');
      this.request.path = `$***REMOVED***path***REMOVED***?$***REMOVED***this.options.qs.stringify(this.request.query)***REMOVED***$***REMOVED***query ? `&$***REMOVED***query***REMOVED***` : ''***REMOVED***`;
    ***REMOVED***
  ***REMOVED***
***REMOVED***

/**
 * Create a ((THIS)) request
 * @dynamic this.METHODS
 * @method Snekfetch.((THIS)lowerCase)
 * @param ***REMOVED***string***REMOVED*** url The url to request
 * @param ***REMOVED***Snekfetch.snekfetchOptions***REMOVED*** [opts] Options
 * @returns ***REMOVED***Snekfetch***REMOVED***
 */
Snekfetch.METHODS = transport.METHODS.concat('BREW').filter((m) => m !== 'M-SEARCH');
for (const method of Snekfetch.METHODS) ***REMOVED***
  Snekfetch[method.toLowerCase()] = function runMethod(url, opts) ***REMOVED***
    const Constructor = this.prototype instanceof Snekfetch ? this : Snekfetch;
    return new Constructor(method, url, opts);
  ***REMOVED***;
***REMOVED***

module.exports = Snekfetch;

/**
 * @external Agent
 * @see ***REMOVED***@link https://nodejs.org/api/http.html#http_class_http_agent***REMOVED***
 */
