// eslint-plugin-standard/rules/array-bracket-even-spacing.js
'use strict'

/**
 * @fileoverview Disallows or enforces spaces inside of array brackets.
 * @author Jamund Ferguson
 * @copyright 2015 Jamund Ferguson. All rights reserved.
 * @copyright 2014 Brandyn Bennett. All rights reserved.
 * @copyright 2014 Michael Ficarra. No rights reserved.
 * @copyright 2014 Vignesh Anand. All rights reserved.
 */
// ------------------------------------------------------------------------------
// Rule Definition
// ------------------------------------------------------------------------------

module.exports = ***REMOVED***
  meta: ***REMOVED***
    docs: ***REMOVED***
      url: 'https://github.com/standard/eslint-plugin-standard#rules-explanations'
    ***REMOVED***
  ***REMOVED***,

  create: function (context) ***REMOVED***
    var spaced = context.options[0] === 'always'
    var either = context.options[0] === 'either'

    /**
     * Determines whether an option is set, relative to the spacing option.
     * If spaced is "always", then check whether option is set to false.
     * If spaced is "never", then check whether option is set to true.
     * @param ***REMOVED***Object***REMOVED*** option - The option to exclude.
     * @returns ***REMOVED***boolean***REMOVED*** Whether or not the property is excluded.
     */
    function isOptionSet (option) ***REMOVED***
      return context.options[1] != null ? context.options[1][option] === !spaced : false
    ***REMOVED***

    var options = ***REMOVED***
      either: either,
      spaced: spaced,
      singleElementException: isOptionSet('singleValue'),
      objectsInArraysException: isOptionSet('objectsInArrays'),
      arraysInArraysException: isOptionSet('arraysInArrays')
    ***REMOVED***

    // --------------------------------------------------------------------------
    // Helpers
    // --------------------------------------------------------------------------

    /**
     * Determines whether two adjacent tokens are have whitespace between them.
     * @param ***REMOVED***Object***REMOVED*** left - The left token object.
     * @param ***REMOVED***Object***REMOVED*** right - The right token object.
     * @returns ***REMOVED***boolean***REMOVED*** Whether or not there is space between the tokens.
     */
    function isSpaced (left, right) ***REMOVED***
      return left.range[1] < right.range[0]
    ***REMOVED***

    /**
     * Determines whether two adjacent tokens are on the same line.
     * @param ***REMOVED***Object***REMOVED*** left - The left token object.
     * @param ***REMOVED***Object***REMOVED*** right - The right token object.
     * @returns ***REMOVED***boolean***REMOVED*** Whether or not the tokens are on the same line.
     */
    function isSameLine (left, right) ***REMOVED***
      return left.loc.start.line === right.loc.start.line
    ***REMOVED***

    /**
     * Reports that there shouldn't be a space after the first token
     * @param ***REMOVED***ASTNode***REMOVED*** node - The node to report in the event of an error.
     * @param ***REMOVED***Token***REMOVED*** token - The token to use for the report.
     * @returns ***REMOVED***void***REMOVED***
     */
    function reportNoBeginningSpace (node, token) ***REMOVED***
      context.report(node, token.loc.start,
        "There should be no space after '" + token.value + "'")
    ***REMOVED***

    /**
     * Reports that there shouldn't be a space before the last token
     * @param ***REMOVED***ASTNode***REMOVED*** node - The node to report in the event of an error.
     * @param ***REMOVED***Token***REMOVED*** token - The token to use for the report.
     * @returns ***REMOVED***void***REMOVED***
     */
    function reportNoEndingSpace (node, token) ***REMOVED***
      context.report(node, token.loc.start,
        "There should be no space before '" + token.value + "'")
    ***REMOVED***

    /**
     * Reports that there should be a space after the first token
     * @param ***REMOVED***ASTNode***REMOVED*** node - The node to report in the event of an error.
     * @param ***REMOVED***Token***REMOVED*** token - The token to use for the report.
     * @returns ***REMOVED***void***REMOVED***
     */
    function reportRequiredBeginningSpace (node, token) ***REMOVED***
      context.report(node, token.loc.start,
        "A space is required after '" + token.value + "'")
    ***REMOVED***

    /**
     * Reports that there should be a space before the last token
     * @param ***REMOVED***ASTNode***REMOVED*** node - The node to report in the event of an error.
     * @param ***REMOVED***Token***REMOVED*** token - The token to use for the report.
     * @returns ***REMOVED***void***REMOVED***
     */
    function reportRequiredEndingSpace (node, token) ***REMOVED***
      context.report(node, token.loc.start,
        "A space is required before '" + token.value + "'")
    ***REMOVED***

    /**
     * Checks if a start and end brace in a node are spaced evenly
     * and not too long (>1 space)
     * @param node
     * @param start
     * @param end
     * @returns ***REMOVED***boolean***REMOVED***
     */
    function isEvenlySpacedAndNotTooLong (node, start, end) ***REMOVED***
      var expectedSpace = start[1].range[0] - start[0].range[1]
      var endSpace = end[1].range[0] - end[0].range[1]
      return endSpace === expectedSpace && endSpace <= 1
    ***REMOVED***

    /**
     * Validates the spacing around array brackets
     * @param ***REMOVED***ASTNode***REMOVED*** node - The node we're checking for spacing
     * @returns ***REMOVED***void***REMOVED***
     */
    function validateArraySpacing (node) ***REMOVED***
      if (node.elements.length === 0) ***REMOVED***
        return
      ***REMOVED***

      var first = context.getFirstToken(node)
      var second = context.getFirstToken(node, 1)
      var penultimate = context.getLastToken(node, 1)
      var last = context.getLastToken(node)

      var openingBracketMustBeSpaced =
      (options.objectsInArraysException && second.value === '***REMOVED***') ||
      (options.arraysInArraysException && second.value === '[') ||
      (options.singleElementException && node.elements.length === 1)
        ? !options.spaced : options.spaced

      var closingBracketMustBeSpaced =
      (options.objectsInArraysException && penultimate.value === '***REMOVED***') ||
      (options.arraysInArraysException && penultimate.value === ']') ||
      (options.singleElementException && node.elements.length === 1)
        ? !options.spaced : options.spaced

      // we only care about evenly spaced things
      if (options.either) ***REMOVED***
        // newlines at any point means return
        if (!isSameLine(first, last)) ***REMOVED***
          return
        ***REMOVED***

        // confirm that the object expression/literal is spaced evenly
        if (!isEvenlySpacedAndNotTooLong(node, [first, second], [penultimate, last])) ***REMOVED***
          context.report(node, 'Expected consistent spacing')
        ***REMOVED***

        return
      ***REMOVED***

      if (isSameLine(first, second)) ***REMOVED***
        if (openingBracketMustBeSpaced && !isSpaced(first, second)) ***REMOVED***
          reportRequiredBeginningSpace(node, first)
        ***REMOVED***
        if (!openingBracketMustBeSpaced && isSpaced(first, second)) ***REMOVED***
          reportNoBeginningSpace(node, first)
        ***REMOVED***
      ***REMOVED***

      if (isSameLine(penultimate, last)) ***REMOVED***
        if (closingBracketMustBeSpaced && !isSpaced(penultimate, last)) ***REMOVED***
          reportRequiredEndingSpace(node, last)
        ***REMOVED***
        if (!closingBracketMustBeSpaced && isSpaced(penultimate, last)) ***REMOVED***
          reportNoEndingSpace(node, last)
        ***REMOVED***
      ***REMOVED***
    ***REMOVED***

    // --------------------------------------------------------------------------
    // Public
    // --------------------------------------------------------------------------

    return ***REMOVED***
      ArrayPattern: validateArraySpacing,
      ArrayExpression: validateArraySpacing
    ***REMOVED***
  ***REMOVED***
***REMOVED***
