// eslint-plugin-standard/rules/computed-property-even-spacing.js
'use strict'

/**
 * @fileoverview Disallows or enforces spaces inside computed properties.
 * @author Jamund Ferguson
 * @copyright 2015 Jamund Ferguson. All rights reserved.
 */
// ------------------------------------------------------------------------------
// Rule Definition
// ------------------------------------------------------------------------------

module.exports = ***REMOVED***
  meta: ***REMOVED***
    docs: ***REMOVED***
      url: 'https://github.com/standard/eslint-plugin-standard#rules-explanations'
    ***REMOVED***
  ***REMOVED***,

  create: function (context) ***REMOVED***
    var propertyNameMustBeSpaced = context.options[0] === 'always' // default is "never"
    var propertyNameMustBeEven = context.options[0] === 'even' // default is "never"

    // --------------------------------------------------------------------------
    // Helpers
    // --------------------------------------------------------------------------

    /**
     * Determines whether two adjacent tokens are have whitespace between them.
     * @param ***REMOVED***Object***REMOVED*** left - The left token object.
     * @param ***REMOVED***Object***REMOVED*** right - The right token object.
     * @returns ***REMOVED***boolean***REMOVED*** Whether or not there is space between the tokens.
     */
    function isSpaced (left, right) ***REMOVED***
      return left.range[1] < right.range[0]
    ***REMOVED***

    /**
     * Determines whether two adjacent tokens are on the same line.
     * @param ***REMOVED***Object***REMOVED*** left - The left token object.
     * @param ***REMOVED***Object***REMOVED*** right - The right token object.
     * @returns ***REMOVED***boolean***REMOVED*** Whether or not the tokens are on the same line.
     */
    function isSameLine (left, right) ***REMOVED***
      return left.loc.start.line === right.loc.start.line
    ***REMOVED***

    /**
     * Reports that there shouldn't be a space after the first token
     * @param ***REMOVED***ASTNode***REMOVED*** node - The node to report in the event of an error.
     * @param ***REMOVED***Token***REMOVED*** token - The token to use for the report.
     * @returns ***REMOVED***void***REMOVED***
     */
    function reportNoBeginningSpace (node, token) ***REMOVED***
      context.report(node, token.loc.start,
        "There should be no space after '" + token.value + "'")
    ***REMOVED***

    /**
     * Reports that there shouldn't be a space before the last token
     * @param ***REMOVED***ASTNode***REMOVED*** node - The node to report in the event of an error.
     * @param ***REMOVED***Token***REMOVED*** token - The token to use for the report.
     * @returns ***REMOVED***void***REMOVED***
     */
    function reportNoEndingSpace (node, token) ***REMOVED***
      context.report(node, token.loc.start,
        "There should be no space before '" + token.value + "'")
    ***REMOVED***

    /**
     * Reports that there should be a space after the first token
     * @param ***REMOVED***ASTNode***REMOVED*** node - The node to report in the event of an error.
     * @param ***REMOVED***Token***REMOVED*** token - The token to use for the report.
     * @returns ***REMOVED***void***REMOVED***
     */
    function reportRequiredBeginningSpace (node, token) ***REMOVED***
      context.report(node, token.loc.start,
        "A space is required after '" + token.value + "'")
    ***REMOVED***

    /**
     * Reports that there should be a space before the last token
     * @param ***REMOVED***ASTNode***REMOVED*** node - The node to report in the event of an error.
     * @param ***REMOVED***Token***REMOVED*** token - The token to use for the report.
     * @returns ***REMOVED***void***REMOVED***
     */
    function reportRequiredEndingSpace (node, token) ***REMOVED***
      context.report(node, token.loc.start,
        "A space is required before '" + token.value + "'")
    ***REMOVED***

    /**
     * Returns a function that checks the spacing of a node on the property name
     * that was passed in.
     * @param ***REMOVED***String***REMOVED*** propertyName The property on the node to check for spacing
     * @returns ***REMOVED***Function***REMOVED*** A function that will check spacing on a node
     */
    function checkSpacing (propertyName) ***REMOVED***
      return function (node) ***REMOVED***
        if (!node.computed) ***REMOVED***
          return
        ***REMOVED***

        var property = node[propertyName]

        var before = context.getTokenBefore(property)
        var first = context.getFirstToken(property)
        var last = context.getLastToken(property)
        var after = context.getTokenAfter(property)
        var startSpace, endSpace

        if (propertyNameMustBeEven) ***REMOVED***
          if (!isSameLine(before, after)) ***REMOVED***
            context.report(node, 'Expected "[" and "]" to be on the same line')
            return
          ***REMOVED***
          startSpace = first.loc.start.column - before.loc.end.column
          endSpace = after.loc.start.column - last.loc.end.column

          if (startSpace !== endSpace || startSpace > 1) ***REMOVED***
            context.report(node, 'Expected 1 or 0 spaces around "[" and "]"')
          ***REMOVED***

          return
        ***REMOVED***

        if (isSameLine(before, first)) ***REMOVED***
          if (propertyNameMustBeSpaced) ***REMOVED***
            if (!isSpaced(before, first) && isSameLine(before, first)) ***REMOVED***
              reportRequiredBeginningSpace(node, before)
            ***REMOVED***
          ***REMOVED*** else ***REMOVED***
            if (isSpaced(before, first)) ***REMOVED***
              reportNoBeginningSpace(node, before)
            ***REMOVED***
          ***REMOVED***
        ***REMOVED***

        if (isSameLine(last, after)) ***REMOVED***
          if (propertyNameMustBeSpaced) ***REMOVED***
            if (!isSpaced(last, after) && isSameLine(last, after)) ***REMOVED***
              reportRequiredEndingSpace(node, after)
            ***REMOVED***
          ***REMOVED*** else ***REMOVED***
            if (isSpaced(last, after)) ***REMOVED***
              reportNoEndingSpace(node, after)
            ***REMOVED***
          ***REMOVED***
        ***REMOVED***
      ***REMOVED***
    ***REMOVED***

    // --------------------------------------------------------------------------
    // Public
    // --------------------------------------------------------------------------

    return ***REMOVED***
      Property: checkSpacing('key'),
      MemberExpression: checkSpacing('property')
    ***REMOVED***
  ***REMOVED***
***REMOVED***
