// eslint-plugin-standard/rules/no-callback-literal.js
/**
 * Ensures that the callback pattern is followed properly
 * with an Error object (or undefined or null) in the first position.
 */

'use strict'

// ------------------------------------------------------------------------------
// Helpers
// ------------------------------------------------------------------------------

/**
 * Determine if a node has a possiblity to be an Error object
 * @param  ***REMOVED***ASTNode***REMOVED***  node  ASTNode to check
 * @returns ***REMOVED***boolean***REMOVED***       True if there is a chance it contains an Error obj
 */
function couldBeError (node) ***REMOVED***
  switch (node.type) ***REMOVED***
    case 'Identifier':
    case 'CallExpression':
    case 'NewExpression':
    case 'MemberExpression':
    case 'TaggedTemplateExpression':
    case 'YieldExpression':
      return true // possibly an error object.

    case 'AssignmentExpression':
      return couldBeError(node.right)

    case 'SequenceExpression':
      var exprs = node.expressions
      return exprs.length !== 0 && couldBeError(exprs[exprs.length - 1])

    case 'LogicalExpression':
      return couldBeError(node.left) || couldBeError(node.right)

    case 'ConditionalExpression':
      return couldBeError(node.consequent) || couldBeError(node.alternate)

    default:
      return node.value === null
  ***REMOVED***
***REMOVED***

// ------------------------------------------------------------------------------
// Rule Definition
// ------------------------------------------------------------------------------

module.exports = ***REMOVED***
  meta: ***REMOVED***
    docs: ***REMOVED***
      url: 'https://github.com/standard/eslint-plugin-standard#rules-explanations'
    ***REMOVED***
  ***REMOVED***,

  create: function (context) ***REMOVED***
    var callbackNames = context.options[0] || ['callback', 'cb']

    function isCallback (name) ***REMOVED***
      return callbackNames.indexOf(name) > -1
    ***REMOVED***

    return ***REMOVED***

      CallExpression: function (node) ***REMOVED***
        var errorArg = node.arguments[0]
        var calleeName = node.callee.name

        if (errorArg && !couldBeError(errorArg) && isCallback(calleeName)) ***REMOVED***
          context.report(node, 'Unexpected literal in error position of callback.')
        ***REMOVED***
      ***REMOVED***
    ***REMOVED***
  ***REMOVED***
***REMOVED***
