// eslint-plugin-standard/rules/object-curly-even-spacing.js
'use strict'

/**
 * @fileoverview Disallows or enforces spaces inside of object literals.
 * @author Jamund Ferguson
 * @copyright 2014 Brandyn Bennett. All rights reserved.
 * @copyright 2014 Michael Ficarra. No rights reserved.
 * @copyright 2014 Vignesh Anand. All rights reserved.
 * @copyright 2015 Jamund Ferguson. All rights reserved.
 */
// ------------------------------------------------------------------------------
// Rule Definition
// ------------------------------------------------------------------------------

module.exports = ***REMOVED***
  meta: ***REMOVED***
    docs: ***REMOVED***
      url: 'https://github.com/standard/eslint-plugin-standard#rules-explanations'
    ***REMOVED***
  ***REMOVED***,

  create: function (context) ***REMOVED***
    var spaced = context.options[0] === 'always'
    var either = context.options[0] === 'either'

    /**
     * Determines whether an option is set, relative to the spacing option.
     * If spaced is "always", then check whether option is set to false.
     * If spaced is "never", then check whether option is set to true.
     * @param ***REMOVED***Object***REMOVED*** option - The option to exclude.
     * @returns ***REMOVED***boolean***REMOVED*** Whether or not the property is excluded.
     */
    function isOptionSet (option) ***REMOVED***
      return context.options[1] != null ? context.options[1][option] === !spaced : false
    ***REMOVED***

    var options = ***REMOVED***
      spaced: spaced,
      either: either,
      arraysInObjectsException: isOptionSet('arraysInObjects'),
      objectsInObjectsException: isOptionSet('objectsInObjects')
    ***REMOVED***

    // --------------------------------------------------------------------------
    // Helpers
    // --------------------------------------------------------------------------

    /**
     * Determines whether two adjacent tokens are have whitespace between them.
     * @param ***REMOVED***Object***REMOVED*** left - The left token object.
     * @param ***REMOVED***Object***REMOVED*** right - The right token object.
     * @returns ***REMOVED***boolean***REMOVED*** Whether or not there is space between the tokens.
     */
    function isSpaced (left, right) ***REMOVED***
      return left.range[1] < right.range[0]
    ***REMOVED***

    /**
     * Determines whether two adjacent tokens are on the same line.
     * @param ***REMOVED***Object***REMOVED*** left - The left token object.
     * @param ***REMOVED***Object***REMOVED*** right - The right token object.
     * @returns ***REMOVED***boolean***REMOVED*** Whether or not the tokens are on the same line.
     */
    function isSameLine (left, right) ***REMOVED***
      return left.loc.start.line === right.loc.start.line
    ***REMOVED***

    /**
     * Reports that there shouldn't be a space after the first token
     * @param ***REMOVED***ASTNode***REMOVED*** node - The node to report in the event of an error.
     * @param ***REMOVED***Token***REMOVED*** token - The token to use for the report.
     * @returns ***REMOVED***void***REMOVED***
     */
    function reportNoBeginningSpace (node, token) ***REMOVED***
      context.report(node, token.loc.start,
        "There should be no space after '" + token.value + "'")
    ***REMOVED***

    /**
     * Reports that there shouldn't be a space before the last token
     * @param ***REMOVED***ASTNode***REMOVED*** node - The node to report in the event of an error.
     * @param ***REMOVED***Token***REMOVED*** token - The token to use for the report.
     * @returns ***REMOVED***void***REMOVED***
     */
    function reportNoEndingSpace (node, token) ***REMOVED***
      context.report(node, token.loc.start,
        "There should be no space before '" + token.value + "'")
    ***REMOVED***

    /**
     * Reports that there should be a space after the first token
     * @param ***REMOVED***ASTNode***REMOVED*** node - The node to report in the event of an error.
     * @param ***REMOVED***Token***REMOVED*** token - The token to use for the report.
     * @returns ***REMOVED***void***REMOVED***
     */
    function reportRequiredBeginningSpace (node, token) ***REMOVED***
      context.report(node, token.loc.start,
        "A space is required after '" + token.value + "'")
    ***REMOVED***

    /**
     * Reports that there should be a space before the last token
     * @param ***REMOVED***ASTNode***REMOVED*** node - The node to report in the event of an error.
     * @param ***REMOVED***Token***REMOVED*** token - The token to use for the report.
     * @returns ***REMOVED***void***REMOVED***
     */
    function reportRequiredEndingSpace (node, token) ***REMOVED***
      context.report(node, token.loc.start,
        "A space is required before '" + token.value + "'")
    ***REMOVED***

    /**
     * Checks if a start and end brace in a node are spaced evenly
     * and not too long (>1 space)
     * @param node
     * @param start
     * @param end
     * @returns ***REMOVED***boolean***REMOVED***
     */
    function isEvenlySpacedAndNotTooLong (node, start, end) ***REMOVED***
      var expectedSpace = start[1].range[0] - start[0].range[1]
      var endSpace = end[1].range[0] - end[0].range[1]
      return endSpace === expectedSpace && endSpace <= 1
    ***REMOVED***

    /**
     * Determines if spacing in curly braces is valid.
     * @param ***REMOVED***ASTNode***REMOVED*** node The AST node to check.
     * @param ***REMOVED***Token***REMOVED*** first The first token to check (should be the opening brace)
     * @param ***REMOVED***Token***REMOVED*** second The second token to check (should be first after the opening brace)
     * @param ***REMOVED***Token***REMOVED*** penultimate The penultimate token to check (should be last before closing brace)
     * @param ***REMOVED***Token***REMOVED*** last The last token to check (should be closing brace)
     * @returns ***REMOVED***void***REMOVED***
     */
    function validateBraceSpacing (node, first, second, penultimate, last) ***REMOVED***
      var closingCurlyBraceMustBeSpaced =
      (options.arraysInObjectsException && penultimate.value === ']') ||
      (options.objectsInObjectsException && penultimate.value === '***REMOVED***')
        ? !options.spaced : options.spaced

      // we only care about evenly spaced things
      if (options.either) ***REMOVED***
        // newlines at any point means return
        if (!isSameLine(first, last)) ***REMOVED***
          return
        ***REMOVED***

        // confirm that the object expression/literal is spaced evenly
        if (!isEvenlySpacedAndNotTooLong(node, [first, second], [penultimate, last])) ***REMOVED***
          context.report(node, 'Expected consistent spacing')
        ***REMOVED***

        return
      ***REMOVED***

      // ***REMOVED*** and key are on same line
      if (isSameLine(first, second)) ***REMOVED***
        if (options.spaced && !isSpaced(first, second)) ***REMOVED***
          reportRequiredBeginningSpace(node, first)
        ***REMOVED***
        if (!options.spaced && isSpaced(first, second)) ***REMOVED***
          reportNoBeginningSpace(node, first)
        ***REMOVED***
      ***REMOVED***

      // final key and ***REMOVED*** ore on the same line
      if (isSameLine(penultimate, last)) ***REMOVED***
        if (closingCurlyBraceMustBeSpaced && !isSpaced(penultimate, last)) ***REMOVED***
          reportRequiredEndingSpace(node, last)
        ***REMOVED***
        if (!closingCurlyBraceMustBeSpaced && isSpaced(penultimate, last)) ***REMOVED***
          reportNoEndingSpace(node, last)
        ***REMOVED***
      ***REMOVED***
    ***REMOVED***

    // --------------------------------------------------------------------------
    // Public
    // --------------------------------------------------------------------------

    return ***REMOVED***
      // var ***REMOVED***x***REMOVED*** = y
      ObjectPattern: function (node) ***REMOVED***
        if (node.properties.length === 0) ***REMOVED***
          return
        ***REMOVED***

        var firstSpecifier = node.properties[0]
        var lastSpecifier = node.properties[node.properties.length - 1]

        var first = context.getTokenBefore(firstSpecifier)
        var second = context.getFirstToken(firstSpecifier)
        var penultimate = context.getLastToken(lastSpecifier)
        var last = context.getTokenAfter(lastSpecifier)

        // support trailing commas
        if (last.value === ',') ***REMOVED***
          penultimate = last
          last = context.getTokenAfter(last)
        ***REMOVED***

        validateBraceSpacing(node, first, second, penultimate, last)
      ***REMOVED***,

      // import ***REMOVED***y***REMOVED*** from 'x'
      ImportDeclaration: function (node) ***REMOVED***
        var firstSpecifier = node.specifiers[0]
        var lastSpecifier = node.specifiers[node.specifiers.length - 1]

        // don't do anything for namespace or default imports
        if (firstSpecifier && lastSpecifier && firstSpecifier.type === 'ImportSpecifier' && lastSpecifier.type === 'ImportSpecifier') ***REMOVED***
          var first = context.getTokenBefore(firstSpecifier)
          var second = context.getFirstToken(firstSpecifier)
          var penultimate = context.getLastToken(lastSpecifier)
          var last = context.getTokenAfter(lastSpecifier)

          validateBraceSpacing(node, first, second, penultimate, last)
        ***REMOVED***
      ***REMOVED***,

      // export ***REMOVED***name***REMOVED*** from 'yo'
      ExportNamedDeclaration: function (node) ***REMOVED***
        if (!node.specifiers.length) ***REMOVED***
          return
        ***REMOVED***

        var firstSpecifier = node.specifiers[0]
        var lastSpecifier = node.specifiers[node.specifiers.length - 1]
        var first = context.getTokenBefore(firstSpecifier)
        var second = context.getFirstToken(firstSpecifier)
        var penultimate = context.getLastToken(lastSpecifier)
        var last = context.getTokenAfter(lastSpecifier)

        validateBraceSpacing(node, first, second, penultimate, last)
      ***REMOVED***,

      // var y = ***REMOVED***x: 'y'***REMOVED***
      ObjectExpression: function (node) ***REMOVED***
        if (node.properties.length === 0) ***REMOVED***
          return
        ***REMOVED***

        var first = context.getFirstToken(node)
        var second = context.getFirstToken(node, 1)
        var penultimate = context.getLastToken(node, 1)
        var last = context.getLastToken(node)

        validateBraceSpacing(node, first, second, penultimate, last)
      ***REMOVED***
    ***REMOVED***
  ***REMOVED***
***REMOVED***
