// v8-compile-cache/v8-compile-cache.js
'use strict';

const Module = require('module');
const crypto = require('crypto');
const fs = require('fs');
const path = require('path');
const vm = require('vm');
const os = require('os');

const hasOwnProperty = Object.prototype.hasOwnProperty;

//------------------------------------------------------------------------------
// FileSystemBlobStore
//------------------------------------------------------------------------------

class FileSystemBlobStore ***REMOVED***
  constructor(directory, prefix) ***REMOVED***
    const name = prefix ? slashEscape(prefix + '.') : '';
    this._blobFilename = path.join(directory, name + 'BLOB');
    this._mapFilename = path.join(directory, name + 'MAP');
    this._lockFilename = path.join(directory, name + 'LOCK');
    this._directory = directory;
    this._load();
  ***REMOVED***

  has(key, invalidationKey) ***REMOVED***
    if (hasOwnProperty.call(this._memoryBlobs, key)) ***REMOVED***
      return this._invalidationKeys[key] === invalidationKey;
    ***REMOVED*** else if (hasOwnProperty.call(this._storedMap, key)) ***REMOVED***
      return this._storedMap[key][0] === invalidationKey;
    ***REMOVED***
    return false;
  ***REMOVED***

  get(key, invalidationKey) ***REMOVED***
    if (hasOwnProperty.call(this._memoryBlobs, key)) ***REMOVED***
      if (this._invalidationKeys[key] === invalidationKey) ***REMOVED***
        return this._memoryBlobs[key];
      ***REMOVED***
    ***REMOVED*** else if (hasOwnProperty.call(this._storedMap, key)) ***REMOVED***
      const mapping = this._storedMap[key];
      if (mapping[0] === invalidationKey) ***REMOVED***
        return this._storedBlob.slice(mapping[1], mapping[2]);
      ***REMOVED***
    ***REMOVED***
  ***REMOVED***

  set(key, invalidationKey, buffer) ***REMOVED***
    this._invalidationKeys[key] = invalidationKey;
    this._memoryBlobs[key] = buffer;
    this._dirty = true;
  ***REMOVED***

  delete(key) ***REMOVED***
    if (hasOwnProperty.call(this._memoryBlobs, key)) ***REMOVED***
      this._dirty = true;
      delete this._memoryBlobs[key];
    ***REMOVED***
    if (hasOwnProperty.call(this._invalidationKeys, key)) ***REMOVED***
      this._dirty = true;
      delete this._invalidationKeys[key];
    ***REMOVED***
    if (hasOwnProperty.call(this._storedMap, key)) ***REMOVED***
      this._dirty = true;
      delete this._storedMap[key];
    ***REMOVED***
  ***REMOVED***

  isDirty() ***REMOVED***
    return this._dirty;
  ***REMOVED***

  save() ***REMOVED***
    const dump = this._getDump();
    const blobToStore = Buffer.concat(dump[0]);
    const mapToStore = JSON.stringify(dump[1]);

    try ***REMOVED***
      mkdirpSync(this._directory);
      fs.writeFileSync(this._lockFilename, 'LOCK', ***REMOVED***flag: 'wx'***REMOVED***);
    ***REMOVED*** catch (error) ***REMOVED***
      // Swallow the exception if we fail to acquire the lock.
      return false;
    ***REMOVED***

    try ***REMOVED***
      fs.writeFileSync(this._blobFilename, blobToStore);
      fs.writeFileSync(this._mapFilename, mapToStore);
    ***REMOVED*** catch (error) ***REMOVED***
      throw error;
    ***REMOVED*** finally ***REMOVED***
      fs.unlinkSync(this._lockFilename);
    ***REMOVED***

    return true;
  ***REMOVED***

  _load() ***REMOVED***
    try ***REMOVED***
      this._storedBlob = fs.readFileSync(this._blobFilename);
      this._storedMap = JSON.parse(fs.readFileSync(this._mapFilename));
    ***REMOVED*** catch (e) ***REMOVED***
      this._storedBlob = Buffer.alloc(0);
      this._storedMap = ***REMOVED******REMOVED***;
    ***REMOVED***
    this._dirty = false;
    this._memoryBlobs = ***REMOVED******REMOVED***;
    this._invalidationKeys = ***REMOVED******REMOVED***;
  ***REMOVED***

  _getDump() ***REMOVED***
    const buffers = [];
    const newMap = ***REMOVED******REMOVED***;
    let offset = 0;

    function push(key, invalidationKey, buffer) ***REMOVED***
      buffers.push(buffer);
      newMap[key] = [invalidationKey, offset, offset + buffer.length];
      offset += buffer.length;
    ***REMOVED***

    for (const key of Object.keys(this._memoryBlobs)) ***REMOVED***
      const buffer = this._memoryBlobs[key];
      const invalidationKey = this._invalidationKeys[key];
      push(key, invalidationKey, buffer);
    ***REMOVED***

    for (const key of Object.keys(this._storedMap)) ***REMOVED***
      if (hasOwnProperty.call(newMap, key)) continue;
      const mapping = this._storedMap[key];
      const buffer = this._storedBlob.slice(mapping[1], mapping[2]);
      push(key, mapping[0], buffer);
    ***REMOVED***

    return [buffers, newMap];
  ***REMOVED***
***REMOVED***

//------------------------------------------------------------------------------
// NativeCompileCache
//------------------------------------------------------------------------------

class NativeCompileCache ***REMOVED***
  constructor() ***REMOVED***
    this._cacheStore = null;
    this._previousModuleCompile = null;
  ***REMOVED***

  setCacheStore(cacheStore) ***REMOVED***
    this._cacheStore = cacheStore;
  ***REMOVED***

  install() ***REMOVED***
    const self = this;
    const hasRequireResolvePaths = typeof require.resolve.paths === 'function';
    this._previousModuleCompile = Module.prototype._compile;
    Module.prototype._compile = function(content, filename) ***REMOVED***
      const mod = this;

      function require(id) ***REMOVED***
        return mod.require(id);
      ***REMOVED***

      // https://github.com/nodejs/node/blob/v10.15.3/lib/internal/modules/cjs/helpers.js#L28
      function resolve(request, options) ***REMOVED***
        return Module._resolveFilename(request, mod, false, options);
      ***REMOVED***
      require.resolve = resolve;

      // https://github.com/nodejs/node/blob/v10.15.3/lib/internal/modules/cjs/helpers.js#L37
      // resolve.resolve.paths was added in v8.9.0
      if (hasRequireResolvePaths) ***REMOVED***
        resolve.paths = function paths(request) ***REMOVED***
          return Module._resolveLookupPaths(request, mod, true);
        ***REMOVED***;
      ***REMOVED***

      require.main = process.mainModule;

      // Enable support to add extra extension types
      require.extensions = Module._extensions;
      require.cache = Module._cache;

      const dirname = path.dirname(filename);

      const compiledWrapper = self._moduleCompile(filename, content);

      // We skip the debugger setup because by the time we run, node has already
      // done that itself.

      // `Buffer` is included for Electron.
      // See https://github.com/zertosh/v8-compile-cache/pull/10#issuecomment-518042543
      const args = [mod.exports, require, mod, filename, dirname, process, global, Buffer];
      return compiledWrapper.apply(mod.exports, args);
    ***REMOVED***;
  ***REMOVED***

  uninstall() ***REMOVED***
    Module.prototype._compile = this._previousModuleCompile;
  ***REMOVED***

  _moduleCompile(filename, content) ***REMOVED***
    // https://github.com/nodejs/node/blob/v7.5.0/lib/module.js#L511

    // Remove shebang
    var contLen = content.length;
    if (contLen >= 2) ***REMOVED***
      if (content.charCodeAt(0) === 35/*#*/ &&
          content.charCodeAt(1) === 33/*!*/) ***REMOVED***
        if (contLen === 2) ***REMOVED***
          // Exact match
          content = '';
        ***REMOVED*** else ***REMOVED***
          // Find end of shebang line and slice it off
          var i = 2;
          for (; i < contLen; ++i) ***REMOVED***
            var code = content.charCodeAt(i);
            if (code === 10/*\n*/ || code === 13/*\r*/) break;
          ***REMOVED***
          if (i === contLen) ***REMOVED***
            content = '';
          ***REMOVED*** else ***REMOVED***
            // Note that this actually includes the newline character(s) in the
            // new output. This duplicates the behavior of the regular
            // expression that was previously used to replace the shebang line
            content = content.slice(i);
          ***REMOVED***
        ***REMOVED***
      ***REMOVED***
    ***REMOVED***

    // create wrapper function
    var wrapper = Module.wrap(content);

    var invalidationKey = crypto
      .createHash('sha1')
      .update(content, 'utf8')
      .digest('hex');

    var buffer = this._cacheStore.get(filename, invalidationKey);

    var script = new vm.Script(wrapper, ***REMOVED***
      filename: filename,
      lineOffset: 0,
      displayErrors: true,
      cachedData: buffer,
      produceCachedData: true,
    ***REMOVED***);

    if (script.cachedDataProduced) ***REMOVED***
      this._cacheStore.set(filename, invalidationKey, script.cachedData);
    ***REMOVED*** else if (script.cachedDataRejected) ***REMOVED***
      this._cacheStore.delete(filename);
    ***REMOVED***

    var compiledWrapper = script.runInThisContext(***REMOVED***
      filename: filename,
      lineOffset: 0,
      columnOffset: 0,
      displayErrors: true,
    ***REMOVED***);

    return compiledWrapper;
  ***REMOVED***
***REMOVED***

//------------------------------------------------------------------------------
// utilities
//
// https://github.com/substack/node-mkdirp/blob/f2003bb/index.js#L55-L98
// https://github.com/zertosh/slash-escape/blob/e7ebb99/slash-escape.js
//------------------------------------------------------------------------------

function mkdirpSync(p_) ***REMOVED***
  _mkdirpSync(path.resolve(p_), parseInt('0777', 8) & ~process.umask());
***REMOVED***

function _mkdirpSync(p, mode) ***REMOVED***
  try ***REMOVED***
    fs.mkdirSync(p, mode);
  ***REMOVED*** catch (err0) ***REMOVED***
    if (err0.code === 'ENOENT') ***REMOVED***
      _mkdirpSync(path.dirname(p));
      _mkdirpSync(p);
    ***REMOVED*** else ***REMOVED***
      try ***REMOVED***
        const stat = fs.statSync(p);
        if (!stat.isDirectory()) ***REMOVED*** throw err0; ***REMOVED***
      ***REMOVED*** catch (err1) ***REMOVED***
        throw err0;
      ***REMOVED***
    ***REMOVED***
  ***REMOVED***
***REMOVED***

function slashEscape(str) ***REMOVED***
  const ESCAPE_LOOKUP = ***REMOVED***
    '\\': 'zB',
    ':': 'zC',
    '/': 'zS',
    '\x00': 'z0',
    'z': 'zZ',
  ***REMOVED***;
  return str.replace(/[\\:\/\x00z]/g, match => (ESCAPE_LOOKUP[match]));
***REMOVED***

function supportsCachedData() ***REMOVED***
  const script = new vm.Script('""', ***REMOVED***produceCachedData: true***REMOVED***);
  // chakracore, as of v1.7.1.0, returns `false`.
  return script.cachedDataProduced === true;
***REMOVED***

function getCacheDir() ***REMOVED***
  // Avoid cache ownership issues on POSIX systems.
  const dirname = typeof process.getuid === 'function'
    ? 'v8-compile-cache-' + process.getuid()
    : 'v8-compile-cache';
  const version = typeof process.versions.v8 === 'string'
    ? process.versions.v8
    : typeof process.versions.chakracore === 'string'
      ? 'chakracore-' + process.versions.chakracore
      : 'node-' + process.version;
  const cacheDir = path.join(os.tmpdir(), dirname, version);
  return cacheDir;
***REMOVED***

function getParentName() ***REMOVED***
  // `module.parent.filename` is undefined or null when:
  //    * node -e 'require("v8-compile-cache")'
  //    * node -r 'v8-compile-cache'
  //    * Or, requiring from the REPL.
  const parentName = module.parent && typeof module.parent.filename === 'string'
    ? module.parent.filename
    : process.cwd();
  return parentName;
***REMOVED***

//------------------------------------------------------------------------------
// main
//------------------------------------------------------------------------------

if (!process.env.DISABLE_V8_COMPILE_CACHE && supportsCachedData()) ***REMOVED***
  const cacheDir = getCacheDir();
  const prefix = getParentName();
  const blobStore = new FileSystemBlobStore(cacheDir, prefix);

  const nativeCompileCache = new NativeCompileCache();
  nativeCompileCache.setCacheStore(blobStore);
  nativeCompileCache.install();

  process.once('exit', code => ***REMOVED***
    if (blobStore.isDirty()) ***REMOVED***
      blobStore.save();
    ***REMOVED***
    nativeCompileCache.uninstall();
  ***REMOVED***);
***REMOVED***

module.exports.__TEST__ = ***REMOVED***
  FileSystemBlobStore,
  NativeCompileCache,
  mkdirpSync,
  slashEscape,
  supportsCachedData,
  getCacheDir,
  getParentName,
***REMOVED***;
