// rxjs/_esm5/internal/scheduler/QueueAction.js
/** PURE_IMPORTS_START tslib,_AsyncAction PURE_IMPORTS_END */
import * as tslib_1 from "tslib";
import ***REMOVED*** AsyncAction ***REMOVED*** from './AsyncAction';
var QueueAction = /*@__PURE__*/ (function (_super) ***REMOVED***
    tslib_1.__extends(QueueAction, _super);
    function QueueAction(scheduler, work) ***REMOVED***
        var _this = _super.call(this, scheduler, work) || this;
        _this.scheduler = scheduler;
        _this.work = work;
        return _this;
    ***REMOVED***
    QueueAction.prototype.schedule = function (state, delay) ***REMOVED***
        if (delay === void 0) ***REMOVED***
            delay = 0;
        ***REMOVED***
        if (delay > 0) ***REMOVED***
            return _super.prototype.schedule.call(this, state, delay);
        ***REMOVED***
        this.delay = delay;
        this.state = state;
        this.scheduler.flush(this);
        return this;
    ***REMOVED***;
    QueueAction.prototype.execute = function (state, delay) ***REMOVED***
        return (delay > 0 || this.closed) ?
            _super.prototype.execute.call(this, state, delay) :
            this._execute(state, delay);
    ***REMOVED***;
    QueueAction.prototype.requestAsyncId = function (scheduler, id, delay) ***REMOVED***
        if (delay === void 0) ***REMOVED***
            delay = 0;
        ***REMOVED***
        if ((delay !== null && delay > 0) || (delay === null && this.delay > 0)) ***REMOVED***
            return _super.prototype.requestAsyncId.call(this, scheduler, id, delay);
        ***REMOVED***
        return scheduler.flush(this);
    ***REMOVED***;
    return QueueAction;
***REMOVED***(AsyncAction));
export ***REMOVED*** QueueAction ***REMOVED***;
