// regexpp/index.mjs
/*! @author Toru Nagashima <https://github.com/mysticatea> */


var ast = /*#__PURE__*/Object.freeze(***REMOVED***

***REMOVED***);

let largeIdStartPattern = null;
let largeIdContinuePattern = null;
function isIdStart(cp) ***REMOVED***
    if (cp < 0x41)
        return false;
    if (cp < 0x5b)
        return true;
    if (cp < 0x61)
        return false;
    if (cp < 0x7b)
        return true;
    return isLargeIdStart(cp);
***REMOVED***
function isIdContinue(cp) ***REMOVED***
    if (cp < 0x30)
        return false;
    if (cp < 0x3a)
        return true;
    if (cp < 0x41)
        return false;
    if (cp < 0x5b)
        return true;
    if (cp === 0x5f)
        return true;
    if (cp < 0x61)
        return false;
    if (cp < 0x7b)
        return true;
    return isLargeIdStart(cp) || isLargeIdContinue(cp);
***REMOVED***
function isLargeIdStart(cp) ***REMOVED***
    if (!largeIdStartPattern) ***REMOVED***
        largeIdStartPattern = new RegExp("^[\xaa\xb5\xba\xc0-\xd6\xd8-\xf6\xf8-\u02c1\u02c6-\u02d1" +
            "\u02e0-\u02e4\u02ec\u02ee\u0370-\u0374\u0376\u0377" +
            "\u037a-\u037d\u037f\u0386\u0388-\u038a\u038c\u038e-\u03a1" +
            "\u03a3-\u03f5\u03f7-\u0481\u048a-\u052f\u0531-\u0556\u0559" +
            "\u0560-\u0588\u05d0-\u05ea\u05ef-\u05f2\u0620-\u064a" +
            "\u066e\u066f\u0671-\u06d3\u06d5\u06e5\u06e6\u06ee\u06ef" +
            "\u06fa-\u06fc\u06ff\u0710\u0712-\u072f\u074d-\u07a5\u07b1" +
            "\u07ca-\u07ea\u07f4\u07f5\u07fa\u0800-\u0815\u081a\u0824" +
            "\u0828\u0840-\u0858\u0860-\u086a\u08a0-\u08b4\u08b6-\u08bd" +
            "\u0904-\u0939\u093d\u0950\u0958-\u0961\u0971-\u0980" +
            "\u0985-\u098c\u098f\u0990\u0993-\u09a8\u09aa-\u09b0\u09b2" +
            "\u09b6-\u09b9\u09bd\u09ce\u09dc\u09dd\u09df-\u09e1" +
            "\u09f0\u09f1\u09fc\u0a05-\u0a0a\u0a0f\u0a10\u0a13-\u0a28" +
            "\u0a2a-\u0a30\u0a32\u0a33\u0a35\u0a36\u0a38\u0a39" +
            "\u0a59-\u0a5c\u0a5e\u0a72-\u0a74\u0a85-\u0a8d\u0a8f-\u0a91" +
            "\u0a93-\u0aa8\u0aaa-\u0ab0\u0ab2\u0ab3\u0ab5-\u0ab9\u0abd" +
            "\u0ad0\u0ae0\u0ae1\u0af9\u0b05-\u0b0c\u0b0f\u0b10" +
            "\u0b13-\u0b28\u0b2a-\u0b30\u0b32\u0b33\u0b35-\u0b39\u0b3d" +
            "\u0b5c\u0b5d\u0b5f-\u0b61\u0b71\u0b83\u0b85-\u0b8a" +
            "\u0b8e-\u0b90\u0b92-\u0b95\u0b99\u0b9a\u0b9c\u0b9e\u0b9f" +
            "\u0ba3\u0ba4\u0ba8-\u0baa\u0bae-\u0bb9\u0bd0\u0c05-\u0c0c" +
            "\u0c0e-\u0c10\u0c12-\u0c28\u0c2a-\u0c39\u0c3d\u0c58-\u0c5a" +
            "\u0c60\u0c61\u0c80\u0c85-\u0c8c\u0c8e-\u0c90\u0c92-\u0ca8" +
            "\u0caa-\u0cb3\u0cb5-\u0cb9\u0cbd\u0cde\u0ce0\u0ce1" +
            "\u0cf1\u0cf2\u0d05-\u0d0c\u0d0e-\u0d10\u0d12-\u0d3a\u0d3d" +
            "\u0d4e\u0d54-\u0d56\u0d5f-\u0d61\u0d7a-\u0d7f\u0d85-\u0d96" +
            "\u0d9a-\u0db1\u0db3-\u0dbb\u0dbd\u0dc0-\u0dc6\u0e01-\u0e30" +
            "\u0e32\u0e33\u0e40-\u0e46\u0e81\u0e82\u0e84\u0e86-\u0e8a" +
            "\u0e8c-\u0ea3\u0ea5\u0ea7-\u0eb0\u0eb2\u0eb3\u0ebd" +
            "\u0ec0-\u0ec4\u0ec6\u0edc-\u0edf\u0f00\u0f40-\u0f47" +
            "\u0f49-\u0f6c\u0f88-\u0f8c\u1000-\u102a\u103f\u1050-\u1055" +
            "\u105a-\u105d\u1061\u1065\u1066\u106e-\u1070\u1075-\u1081" +
            "\u108e\u10a0-\u10c5\u10c7\u10cd\u10d0-\u10fa\u10fc-\u1248" +
            "\u124a-\u124d\u1250-\u1256\u1258\u125a-\u125d\u1260-\u1288" +
            "\u128a-\u128d\u1290-\u12b0\u12b2-\u12b5\u12b8-\u12be\u12c0" +
            "\u12c2-\u12c5\u12c8-\u12d6\u12d8-\u1310\u1312-\u1315" +
            "\u1318-\u135a\u1380-\u138f\u13a0-\u13f5\u13f8-\u13fd" +
            "\u1401-\u166c\u166f-\u167f\u1681-\u169a\u16a0-\u16ea" +
            "\u16ee-\u16f8\u1700-\u170c\u170e-\u1711\u1720-\u1731" +
            "\u1740-\u1751\u1760-\u176c\u176e-\u1770\u1780-\u17b3\u17d7" +
            "\u17dc\u1820-\u1878\u1880-\u18a8\u18aa\u18b0-\u18f5" +
            "\u1900-\u191e\u1950-\u196d\u1970-\u1974\u1980-\u19ab" +
            "\u19b0-\u19c9\u1a00-\u1a16\u1a20-\u1a54\u1aa7\u1b05-\u1b33" +
            "\u1b45-\u1b4b\u1b83-\u1ba0\u1bae\u1baf\u1bba-\u1be5" +
            "\u1c00-\u1c23\u1c4d-\u1c4f\u1c5a-\u1c7d\u1c80-\u1c88" +
            "\u1c90-\u1cba\u1cbd-\u1cbf\u1ce9-\u1cec\u1cee-\u1cf3" +
            "\u1cf5\u1cf6\u1cfa\u1d00-\u1dbf\u1e00-\u1f15\u1f18-\u1f1d" +
            "\u1f20-\u1f45\u1f48-\u1f4d\u1f50-\u1f57\u1f59\u1f5b\u1f5d" +
            "\u1f5f-\u1f7d\u1f80-\u1fb4\u1fb6-\u1fbc\u1fbe\u1fc2-\u1fc4" +
            "\u1fc6-\u1fcc\u1fd0-\u1fd3\u1fd6-\u1fdb\u1fe0-\u1fec" +
            "\u1ff2-\u1ff4\u1ff6-\u1ffc\u2071\u207f\u2090-\u209c\u2102" +
            "\u2107\u210a-\u2113\u2115\u2118-\u211d\u2124\u2126\u2128" +
            "\u212a-\u2139\u213c-\u213f\u2145-\u2149\u214e\u2160-\u2188" +
            "\u2c00-\u2c2e\u2c30-\u2c5e\u2c60-\u2ce4\u2ceb-\u2cee" +
            "\u2cf2\u2cf3\u2d00-\u2d25\u2d27\u2d2d\u2d30-\u2d67\u2d6f" +
            "\u2d80-\u2d96\u2da0-\u2da6\u2da8-\u2dae\u2db0-\u2db6" +
            "\u2db8-\u2dbe\u2dc0-\u2dc6\u2dc8-\u2dce\u2dd0-\u2dd6" +
            "\u2dd8-\u2dde\u3005-\u3007\u3021-\u3029\u3031-\u3035" +
            "\u3038-\u303c\u3041-\u3096\u309b-\u309f\u30a1-\u30fa" +
            "\u30fc-\u30ff\u3105-\u312f\u3131-\u318e\u31a0-\u31ba" +
            "\u31f0-\u31ff\u3400-\u4db5\u4e00-\u9fef\ua000-\ua48c" +
            "\ua4d0-\ua4fd\ua500-\ua60c\ua610-\ua61f\ua62a\ua62b" +
            "\ua640-\ua66e\ua67f-\ua69d\ua6a0-\ua6ef\ua717-\ua71f" +
            "\ua722-\ua788\ua78b-\ua7bf\ua7c2-\ua7c6\ua7f7-\ua801" +
            "\ua803-\ua805\ua807-\ua80a\ua80c-\ua822\ua840-\ua873" +
            "\ua882-\ua8b3\ua8f2-\ua8f7\ua8fb\ua8fd\ua8fe\ua90a-\ua925" +
            "\ua930-\ua946\ua960-\ua97c\ua984-\ua9b2\ua9cf\ua9e0-\ua9e4" +
            "\ua9e6-\ua9ef\ua9fa-\ua9fe\uaa00-\uaa28\uaa40-\uaa42" +
            "\uaa44-\uaa4b\uaa60-\uaa76\uaa7a\uaa7e-\uaaaf\uaab1" +
            "\uaab5\uaab6\uaab9-\uaabd\uaac0\uaac2\uaadb-\uaadd" +
            "\uaae0-\uaaea\uaaf2-\uaaf4\uab01-\uab06\uab09-\uab0e" +
            "\uab11-\uab16\uab20-\uab26\uab28-\uab2e\uab30-\uab5a" +
            "\uab5c-\uab67\uab70-\uabe2\uac00-\ud7a3\ud7b0-\ud7c6" +
            "\ud7cb-\ud7fb\uf900-\ufa6d\ufa70-\ufad9\ufb00-\ufb06" +
            "\ufb13-\ufb17\ufb1d\ufb1f-\ufb28\ufb2a-\ufb36\ufb38-\ufb3c" +
            "\ufb3e\ufb40\ufb41\ufb43\ufb44\ufb46-\ufbb1\ufbd3-\ufd3d" +
            "\ufd50-\ufd8f\ufd92-\ufdc7\ufdf0-\ufdfb\ufe70-\ufe74" +
            "\ufe76-\ufefc\uff21-\uff3a\uff41-\uff5a\uff66-\uffbe" +
            "\uffc2-\uffc7\uffca-\uffcf\uffd2-\uffd7\uffda-\uffdc" +
            "\u***REMOVED***10000***REMOVED***-\u***REMOVED***1000b***REMOVED***\u***REMOVED***1000d***REMOVED***-\u***REMOVED***10026***REMOVED***\u***REMOVED***10028***REMOVED***-\u***REMOVED***1003a***REMOVED***" +
            "\u***REMOVED***1003c***REMOVED***\u***REMOVED***1003d***REMOVED***\u***REMOVED***1003f***REMOVED***-\u***REMOVED***1004d***REMOVED***\u***REMOVED***10050***REMOVED***-\u***REMOVED***1005d***REMOVED***" +
            "\u***REMOVED***10080***REMOVED***-\u***REMOVED***100fa***REMOVED***\u***REMOVED***10140***REMOVED***-\u***REMOVED***10174***REMOVED***\u***REMOVED***10280***REMOVED***-\u***REMOVED***1029c***REMOVED***" +
            "\u***REMOVED***102a0***REMOVED***-\u***REMOVED***102d0***REMOVED***\u***REMOVED***10300***REMOVED***-\u***REMOVED***1031f***REMOVED***\u***REMOVED***1032d***REMOVED***-\u***REMOVED***1034a***REMOVED***" +
            "\u***REMOVED***10350***REMOVED***-\u***REMOVED***10375***REMOVED***\u***REMOVED***10380***REMOVED***-\u***REMOVED***1039d***REMOVED***\u***REMOVED***103a0***REMOVED***-\u***REMOVED***103c3***REMOVED***" +
            "\u***REMOVED***103c8***REMOVED***-\u***REMOVED***103cf***REMOVED***\u***REMOVED***103d1***REMOVED***-\u***REMOVED***103d5***REMOVED***\u***REMOVED***10400***REMOVED***-\u***REMOVED***1049d***REMOVED***" +
            "\u***REMOVED***104b0***REMOVED***-\u***REMOVED***104d3***REMOVED***\u***REMOVED***104d8***REMOVED***-\u***REMOVED***104fb***REMOVED***\u***REMOVED***10500***REMOVED***-\u***REMOVED***10527***REMOVED***" +
            "\u***REMOVED***10530***REMOVED***-\u***REMOVED***10563***REMOVED***\u***REMOVED***10600***REMOVED***-\u***REMOVED***10736***REMOVED***\u***REMOVED***10740***REMOVED***-\u***REMOVED***10755***REMOVED***" +
            "\u***REMOVED***10760***REMOVED***-\u***REMOVED***10767***REMOVED***\u***REMOVED***10800***REMOVED***-\u***REMOVED***10805***REMOVED***\u***REMOVED***10808***REMOVED***" +
            "\u***REMOVED***1080a***REMOVED***-\u***REMOVED***10835***REMOVED***\u***REMOVED***10837***REMOVED***\u***REMOVED***10838***REMOVED***\u***REMOVED***1083c***REMOVED***" +
            "\u***REMOVED***1083f***REMOVED***-\u***REMOVED***10855***REMOVED***\u***REMOVED***10860***REMOVED***-\u***REMOVED***10876***REMOVED***\u***REMOVED***10880***REMOVED***-\u***REMOVED***1089e***REMOVED***" +
            "\u***REMOVED***108e0***REMOVED***-\u***REMOVED***108f2***REMOVED***\u***REMOVED***108f4***REMOVED***\u***REMOVED***108f5***REMOVED***\u***REMOVED***10900***REMOVED***-\u***REMOVED***10915***REMOVED***" +
            "\u***REMOVED***10920***REMOVED***-\u***REMOVED***10939***REMOVED***\u***REMOVED***10980***REMOVED***-\u***REMOVED***109b7***REMOVED***\u***REMOVED***109be***REMOVED***\u***REMOVED***109bf***REMOVED***" +
            "\u***REMOVED***10a00***REMOVED***\u***REMOVED***10a10***REMOVED***-\u***REMOVED***10a13***REMOVED***\u***REMOVED***10a15***REMOVED***-\u***REMOVED***10a17***REMOVED***" +
            "\u***REMOVED***10a19***REMOVED***-\u***REMOVED***10a35***REMOVED***\u***REMOVED***10a60***REMOVED***-\u***REMOVED***10a7c***REMOVED***\u***REMOVED***10a80***REMOVED***-\u***REMOVED***10a9c***REMOVED***" +
            "\u***REMOVED***10ac0***REMOVED***-\u***REMOVED***10ac7***REMOVED***\u***REMOVED***10ac9***REMOVED***-\u***REMOVED***10ae4***REMOVED***\u***REMOVED***10b00***REMOVED***-\u***REMOVED***10b35***REMOVED***" +
            "\u***REMOVED***10b40***REMOVED***-\u***REMOVED***10b55***REMOVED***\u***REMOVED***10b60***REMOVED***-\u***REMOVED***10b72***REMOVED***\u***REMOVED***10b80***REMOVED***-\u***REMOVED***10b91***REMOVED***" +
            "\u***REMOVED***10c00***REMOVED***-\u***REMOVED***10c48***REMOVED***\u***REMOVED***10c80***REMOVED***-\u***REMOVED***10cb2***REMOVED***\u***REMOVED***10cc0***REMOVED***-\u***REMOVED***10cf2***REMOVED***" +
            "\u***REMOVED***10d00***REMOVED***-\u***REMOVED***10d23***REMOVED***\u***REMOVED***10f00***REMOVED***-\u***REMOVED***10f1c***REMOVED***\u***REMOVED***10f27***REMOVED***" +
            "\u***REMOVED***10f30***REMOVED***-\u***REMOVED***10f45***REMOVED***\u***REMOVED***10fe0***REMOVED***-\u***REMOVED***10ff6***REMOVED***\u***REMOVED***11003***REMOVED***-\u***REMOVED***11037***REMOVED***" +
            "\u***REMOVED***11083***REMOVED***-\u***REMOVED***110af***REMOVED***\u***REMOVED***110d0***REMOVED***-\u***REMOVED***110e8***REMOVED***\u***REMOVED***11103***REMOVED***-\u***REMOVED***11126***REMOVED***" +
            "\u***REMOVED***11144***REMOVED***\u***REMOVED***11150***REMOVED***-\u***REMOVED***11172***REMOVED***\u***REMOVED***11176***REMOVED***\u***REMOVED***11183***REMOVED***-\u***REMOVED***111b2***REMOVED***" +
            "\u***REMOVED***111c1***REMOVED***-\u***REMOVED***111c4***REMOVED***\u***REMOVED***111da***REMOVED***\u***REMOVED***111dc***REMOVED***\u***REMOVED***11200***REMOVED***-\u***REMOVED***11211***REMOVED***" +
            "\u***REMOVED***11213***REMOVED***-\u***REMOVED***1122b***REMOVED***\u***REMOVED***11280***REMOVED***-\u***REMOVED***11286***REMOVED***\u***REMOVED***11288***REMOVED***" +
            "\u***REMOVED***1128a***REMOVED***-\u***REMOVED***1128d***REMOVED***\u***REMOVED***1128f***REMOVED***-\u***REMOVED***1129d***REMOVED***\u***REMOVED***1129f***REMOVED***-\u***REMOVED***112a8***REMOVED***" +
            "\u***REMOVED***112b0***REMOVED***-\u***REMOVED***112de***REMOVED***\u***REMOVED***11305***REMOVED***-\u***REMOVED***1130c***REMOVED***\u***REMOVED***1130f***REMOVED***\u***REMOVED***11310***REMOVED***" +
            "\u***REMOVED***11313***REMOVED***-\u***REMOVED***11328***REMOVED***\u***REMOVED***1132a***REMOVED***-\u***REMOVED***11330***REMOVED***\u***REMOVED***11332***REMOVED***\u***REMOVED***11333***REMOVED***" +
            "\u***REMOVED***11335***REMOVED***-\u***REMOVED***11339***REMOVED***\u***REMOVED***1133d***REMOVED***\u***REMOVED***11350***REMOVED***\u***REMOVED***1135d***REMOVED***-\u***REMOVED***11361***REMOVED***" +
            "\u***REMOVED***11400***REMOVED***-\u***REMOVED***11434***REMOVED***\u***REMOVED***11447***REMOVED***-\u***REMOVED***1144a***REMOVED***\u***REMOVED***1145f***REMOVED***" +
            "\u***REMOVED***11480***REMOVED***-\u***REMOVED***114af***REMOVED***\u***REMOVED***114c4***REMOVED***\u***REMOVED***114c5***REMOVED***\u***REMOVED***114c7***REMOVED***" +
            "\u***REMOVED***11580***REMOVED***-\u***REMOVED***115ae***REMOVED***\u***REMOVED***115d8***REMOVED***-\u***REMOVED***115db***REMOVED***\u***REMOVED***11600***REMOVED***-\u***REMOVED***1162f***REMOVED***" +
            "\u***REMOVED***11644***REMOVED***\u***REMOVED***11680***REMOVED***-\u***REMOVED***116aa***REMOVED***\u***REMOVED***116b8***REMOVED***\u***REMOVED***11700***REMOVED***-\u***REMOVED***1171a***REMOVED***" +
            "\u***REMOVED***11800***REMOVED***-\u***REMOVED***1182b***REMOVED***\u***REMOVED***118a0***REMOVED***-\u***REMOVED***118df***REMOVED***\u***REMOVED***118ff***REMOVED***" +
            "\u***REMOVED***119a0***REMOVED***-\u***REMOVED***119a7***REMOVED***\u***REMOVED***119aa***REMOVED***-\u***REMOVED***119d0***REMOVED***\u***REMOVED***119e1***REMOVED***\u***REMOVED***119e3***REMOVED***" +
            "\u***REMOVED***11a00***REMOVED***\u***REMOVED***11a0b***REMOVED***-\u***REMOVED***11a32***REMOVED***\u***REMOVED***11a3a***REMOVED***\u***REMOVED***11a50***REMOVED***" +
            "\u***REMOVED***11a5c***REMOVED***-\u***REMOVED***11a89***REMOVED***\u***REMOVED***11a9d***REMOVED***\u***REMOVED***11ac0***REMOVED***-\u***REMOVED***11af8***REMOVED***" +
            "\u***REMOVED***11c00***REMOVED***-\u***REMOVED***11c08***REMOVED***\u***REMOVED***11c0a***REMOVED***-\u***REMOVED***11c2e***REMOVED***\u***REMOVED***11c40***REMOVED***" +
            "\u***REMOVED***11c72***REMOVED***-\u***REMOVED***11c8f***REMOVED***\u***REMOVED***11d00***REMOVED***-\u***REMOVED***11d06***REMOVED***\u***REMOVED***11d08***REMOVED***\u***REMOVED***11d09***REMOVED***" +
            "\u***REMOVED***11d0b***REMOVED***-\u***REMOVED***11d30***REMOVED***\u***REMOVED***11d46***REMOVED***\u***REMOVED***11d60***REMOVED***-\u***REMOVED***11d65***REMOVED***" +
            "\u***REMOVED***11d67***REMOVED***\u***REMOVED***11d68***REMOVED***\u***REMOVED***11d6a***REMOVED***-\u***REMOVED***11d89***REMOVED***\u***REMOVED***11d98***REMOVED***" +
            "\u***REMOVED***11ee0***REMOVED***-\u***REMOVED***11ef2***REMOVED***\u***REMOVED***12000***REMOVED***-\u***REMOVED***12399***REMOVED***\u***REMOVED***12400***REMOVED***-\u***REMOVED***1246e***REMOVED***" +
            "\u***REMOVED***12480***REMOVED***-\u***REMOVED***12543***REMOVED***\u***REMOVED***13000***REMOVED***-\u***REMOVED***1342e***REMOVED***\u***REMOVED***14400***REMOVED***-\u***REMOVED***14646***REMOVED***" +
            "\u***REMOVED***16800***REMOVED***-\u***REMOVED***16a38***REMOVED***\u***REMOVED***16a40***REMOVED***-\u***REMOVED***16a5e***REMOVED***\u***REMOVED***16ad0***REMOVED***-\u***REMOVED***16aed***REMOVED***" +
            "\u***REMOVED***16b00***REMOVED***-\u***REMOVED***16b2f***REMOVED***\u***REMOVED***16b40***REMOVED***-\u***REMOVED***16b43***REMOVED***\u***REMOVED***16b63***REMOVED***-\u***REMOVED***16b77***REMOVED***" +
            "\u***REMOVED***16b7d***REMOVED***-\u***REMOVED***16b8f***REMOVED***\u***REMOVED***16e40***REMOVED***-\u***REMOVED***16e7f***REMOVED***\u***REMOVED***16f00***REMOVED***-\u***REMOVED***16f4a***REMOVED***" +
            "\u***REMOVED***16f50***REMOVED***\u***REMOVED***16f93***REMOVED***-\u***REMOVED***16f9f***REMOVED***\u***REMOVED***16fe0***REMOVED***\u***REMOVED***16fe1***REMOVED***\u***REMOVED***16fe3***REMOVED***" +
            "\u***REMOVED***17000***REMOVED***-\u***REMOVED***187f7***REMOVED***\u***REMOVED***18800***REMOVED***-\u***REMOVED***18af2***REMOVED***\u***REMOVED***1b000***REMOVED***-\u***REMOVED***1b11e***REMOVED***" +
            "\u***REMOVED***1b150***REMOVED***-\u***REMOVED***1b152***REMOVED***\u***REMOVED***1b164***REMOVED***-\u***REMOVED***1b167***REMOVED***\u***REMOVED***1b170***REMOVED***-\u***REMOVED***1b2fb***REMOVED***" +
            "\u***REMOVED***1bc00***REMOVED***-\u***REMOVED***1bc6a***REMOVED***\u***REMOVED***1bc70***REMOVED***-\u***REMOVED***1bc7c***REMOVED***\u***REMOVED***1bc80***REMOVED***-\u***REMOVED***1bc88***REMOVED***" +
            "\u***REMOVED***1bc90***REMOVED***-\u***REMOVED***1bc99***REMOVED***\u***REMOVED***1d400***REMOVED***-\u***REMOVED***1d454***REMOVED***\u***REMOVED***1d456***REMOVED***-\u***REMOVED***1d49c***REMOVED***" +
            "\u***REMOVED***1d49e***REMOVED***\u***REMOVED***1d49f***REMOVED***\u***REMOVED***1d4a2***REMOVED***\u***REMOVED***1d4a5***REMOVED***\u***REMOVED***1d4a6***REMOVED***" +
            "\u***REMOVED***1d4a9***REMOVED***-\u***REMOVED***1d4ac***REMOVED***\u***REMOVED***1d4ae***REMOVED***-\u***REMOVED***1d4b9***REMOVED***\u***REMOVED***1d4bb***REMOVED***" +
            "\u***REMOVED***1d4bd***REMOVED***-\u***REMOVED***1d4c3***REMOVED***\u***REMOVED***1d4c5***REMOVED***-\u***REMOVED***1d505***REMOVED***\u***REMOVED***1d507***REMOVED***-\u***REMOVED***1d50a***REMOVED***" +
            "\u***REMOVED***1d50d***REMOVED***-\u***REMOVED***1d514***REMOVED***\u***REMOVED***1d516***REMOVED***-\u***REMOVED***1d51c***REMOVED***\u***REMOVED***1d51e***REMOVED***-\u***REMOVED***1d539***REMOVED***" +
            "\u***REMOVED***1d53b***REMOVED***-\u***REMOVED***1d53e***REMOVED***\u***REMOVED***1d540***REMOVED***-\u***REMOVED***1d544***REMOVED***\u***REMOVED***1d546***REMOVED***" +
            "\u***REMOVED***1d54a***REMOVED***-\u***REMOVED***1d550***REMOVED***\u***REMOVED***1d552***REMOVED***-\u***REMOVED***1d6a5***REMOVED***\u***REMOVED***1d6a8***REMOVED***-\u***REMOVED***1d6c0***REMOVED***" +
            "\u***REMOVED***1d6c2***REMOVED***-\u***REMOVED***1d6da***REMOVED***\u***REMOVED***1d6dc***REMOVED***-\u***REMOVED***1d6fa***REMOVED***\u***REMOVED***1d6fc***REMOVED***-\u***REMOVED***1d714***REMOVED***" +
            "\u***REMOVED***1d716***REMOVED***-\u***REMOVED***1d734***REMOVED***\u***REMOVED***1d736***REMOVED***-\u***REMOVED***1d74e***REMOVED***\u***REMOVED***1d750***REMOVED***-\u***REMOVED***1d76e***REMOVED***" +
            "\u***REMOVED***1d770***REMOVED***-\u***REMOVED***1d788***REMOVED***\u***REMOVED***1d78a***REMOVED***-\u***REMOVED***1d7a8***REMOVED***\u***REMOVED***1d7aa***REMOVED***-\u***REMOVED***1d7c2***REMOVED***" +
            "\u***REMOVED***1d7c4***REMOVED***-\u***REMOVED***1d7cb***REMOVED***\u***REMOVED***1e100***REMOVED***-\u***REMOVED***1e12c***REMOVED***\u***REMOVED***1e137***REMOVED***-\u***REMOVED***1e13d***REMOVED***" +
            "\u***REMOVED***1e14e***REMOVED***\u***REMOVED***1e2c0***REMOVED***-\u***REMOVED***1e2eb***REMOVED***\u***REMOVED***1e800***REMOVED***-\u***REMOVED***1e8c4***REMOVED***" +
            "\u***REMOVED***1e900***REMOVED***-\u***REMOVED***1e943***REMOVED***\u***REMOVED***1e94b***REMOVED***\u***REMOVED***1ee00***REMOVED***-\u***REMOVED***1ee03***REMOVED***" +
            "\u***REMOVED***1ee05***REMOVED***-\u***REMOVED***1ee1f***REMOVED***\u***REMOVED***1ee21***REMOVED***\u***REMOVED***1ee22***REMOVED***\u***REMOVED***1ee24***REMOVED***\u***REMOVED***1ee27***REMOVED***" +
            "\u***REMOVED***1ee29***REMOVED***-\u***REMOVED***1ee32***REMOVED***\u***REMOVED***1ee34***REMOVED***-\u***REMOVED***1ee37***REMOVED***\u***REMOVED***1ee39***REMOVED***\u***REMOVED***1ee3b***REMOVED***" +
            "\u***REMOVED***1ee42***REMOVED***\u***REMOVED***1ee47***REMOVED***\u***REMOVED***1ee49***REMOVED***\u***REMOVED***1ee4b***REMOVED***\u***REMOVED***1ee4d***REMOVED***-\u***REMOVED***1ee4f***REMOVED***" +
            "\u***REMOVED***1ee51***REMOVED***\u***REMOVED***1ee52***REMOVED***\u***REMOVED***1ee54***REMOVED***\u***REMOVED***1ee57***REMOVED***\u***REMOVED***1ee59***REMOVED***\u***REMOVED***1ee5b***REMOVED***" +
            "\u***REMOVED***1ee5d***REMOVED***\u***REMOVED***1ee5f***REMOVED***\u***REMOVED***1ee61***REMOVED***\u***REMOVED***1ee62***REMOVED***\u***REMOVED***1ee64***REMOVED***" +
            "\u***REMOVED***1ee67***REMOVED***-\u***REMOVED***1ee6a***REMOVED***\u***REMOVED***1ee6c***REMOVED***-\u***REMOVED***1ee72***REMOVED***\u***REMOVED***1ee74***REMOVED***-\u***REMOVED***1ee77***REMOVED***" +
            "\u***REMOVED***1ee79***REMOVED***-\u***REMOVED***1ee7c***REMOVED***\u***REMOVED***1ee7e***REMOVED***\u***REMOVED***1ee80***REMOVED***-\u***REMOVED***1ee89***REMOVED***" +
            "\u***REMOVED***1ee8b***REMOVED***-\u***REMOVED***1ee9b***REMOVED***\u***REMOVED***1eea1***REMOVED***-\u***REMOVED***1eea3***REMOVED***\u***REMOVED***1eea5***REMOVED***-\u***REMOVED***1eea9***REMOVED***" +
            "\u***REMOVED***1eeab***REMOVED***-\u***REMOVED***1eebb***REMOVED***\u***REMOVED***20000***REMOVED***-\u***REMOVED***2a6d6***REMOVED***\u***REMOVED***2a700***REMOVED***-\u***REMOVED***2b734***REMOVED***" +
            "\u***REMOVED***2b740***REMOVED***-\u***REMOVED***2b81d***REMOVED***\u***REMOVED***2b820***REMOVED***-\u***REMOVED***2cea1***REMOVED***\u***REMOVED***2ceb0***REMOVED***-\u***REMOVED***2ebe0***REMOVED***" +
            "\u***REMOVED***2f800***REMOVED***-\u***REMOVED***2fa1d***REMOVED***]$", "u");
    ***REMOVED***
    return largeIdStartPattern.test(String.fromCodePoint(cp));
***REMOVED***
function isLargeIdContinue(cp) ***REMOVED***
    if (!largeIdContinuePattern) ***REMOVED***
        largeIdContinuePattern = new RegExp("^[\xb7\u0300-\u036f\u0387\u0483-\u0487\u0591-\u05bd\u05bf" +
            "\u05c1\u05c2\u05c4\u05c5\u05c7\u0610-\u061a\u064b-\u0669" +
            "\u0670\u06d6-\u06dc\u06df-\u06e4\u06e7\u06e8\u06ea-\u06ed" +
            "\u06f0-\u06f9\u0711\u0730-\u074a\u07a6-\u07b0\u07c0-\u07c9" +
            "\u07eb-\u07f3\u07fd\u0816-\u0819\u081b-\u0823\u0825-\u0827" +
            "\u0829-\u082d\u0859-\u085b\u08d3-\u08e1\u08e3-\u0903" +
            "\u093a-\u093c\u093e-\u094f\u0951-\u0957\u0962\u0963" +
            "\u0966-\u096f\u0981-\u0983\u09bc\u09be-\u09c4\u09c7\u09c8" +
            "\u09cb-\u09cd\u09d7\u09e2\u09e3\u09e6-\u09ef\u09fe" +
            "\u0a01-\u0a03\u0a3c\u0a3e-\u0a42\u0a47\u0a48\u0a4b-\u0a4d" +
            "\u0a51\u0a66-\u0a71\u0a75\u0a81-\u0a83\u0abc\u0abe-\u0ac5" +
            "\u0ac7-\u0ac9\u0acb-\u0acd\u0ae2\u0ae3\u0ae6-\u0aef" +
            "\u0afa-\u0aff\u0b01-\u0b03\u0b3c\u0b3e-\u0b44\u0b47\u0b48" +
            "\u0b4b-\u0b4d\u0b56\u0b57\u0b62\u0b63\u0b66-\u0b6f\u0b82" +
            "\u0bbe-\u0bc2\u0bc6-\u0bc8\u0bca-\u0bcd\u0bd7\u0be6-\u0bef" +
            "\u0c00-\u0c04\u0c3e-\u0c44\u0c46-\u0c48\u0c4a-\u0c4d" +
            "\u0c55\u0c56\u0c62\u0c63\u0c66-\u0c6f\u0c81-\u0c83\u0cbc" +
            "\u0cbe-\u0cc4\u0cc6-\u0cc8\u0cca-\u0ccd\u0cd5\u0cd6" +
            "\u0ce2\u0ce3\u0ce6-\u0cef\u0d00-\u0d03\u0d3b\u0d3c" +
            "\u0d3e-\u0d44\u0d46-\u0d48\u0d4a-\u0d4d\u0d57\u0d62\u0d63" +
            "\u0d66-\u0d6f\u0d82\u0d83\u0dca\u0dcf-\u0dd4\u0dd6" +
            "\u0dd8-\u0ddf\u0de6-\u0def\u0df2\u0df3\u0e31\u0e34-\u0e3a" +
            "\u0e47-\u0e4e\u0e50-\u0e59\u0eb1\u0eb4-\u0ebc\u0ec8-\u0ecd" +
            "\u0ed0-\u0ed9\u0f18\u0f19\u0f20-\u0f29\u0f35\u0f37\u0f39" +
            "\u0f3e\u0f3f\u0f71-\u0f84\u0f86\u0f87\u0f8d-\u0f97" +
            "\u0f99-\u0fbc\u0fc6\u102b-\u103e\u1040-\u1049\u1056-\u1059" +
            "\u105e-\u1060\u1062-\u1064\u1067-\u106d\u1071-\u1074" +
            "\u1082-\u108d\u108f-\u109d\u135d-\u135f\u1369-\u1371" +
            "\u1712-\u1714\u1732-\u1734\u1752\u1753\u1772\u1773" +
            "\u17b4-\u17d3\u17dd\u17e0-\u17e9\u180b-\u180d\u1810-\u1819" +
            "\u18a9\u1920-\u192b\u1930-\u193b\u1946-\u194f\u19d0-\u19da" +
            "\u1a17-\u1a1b\u1a55-\u1a5e\u1a60-\u1a7c\u1a7f-\u1a89" +
            "\u1a90-\u1a99\u1ab0-\u1abd\u1b00-\u1b04\u1b34-\u1b44" +
            "\u1b50-\u1b59\u1b6b-\u1b73\u1b80-\u1b82\u1ba1-\u1bad" +
            "\u1bb0-\u1bb9\u1be6-\u1bf3\u1c24-\u1c37\u1c40-\u1c49" +
            "\u1c50-\u1c59\u1cd0-\u1cd2\u1cd4-\u1ce8\u1ced\u1cf4" +
            "\u1cf7-\u1cf9\u1dc0-\u1df9\u1dfb-\u1dff\u203f\u2040\u2054" +
            "\u20d0-\u20dc\u20e1\u20e5-\u20f0\u2cef-\u2cf1\u2d7f" +
            "\u2de0-\u2dff\u302a-\u302f\u3099\u309a\ua620-\ua629\ua66f" +
            "\ua674-\ua67d\ua69e\ua69f\ua6f0\ua6f1\ua802\ua806\ua80b" +
            "\ua823-\ua827\ua880\ua881\ua8b4-\ua8c5\ua8d0-\ua8d9" +
            "\ua8e0-\ua8f1\ua8ff-\ua909\ua926-\ua92d\ua947-\ua953" +
            "\ua980-\ua983\ua9b3-\ua9c0\ua9d0-\ua9d9\ua9e5\ua9f0-\ua9f9" +
            "\uaa29-\uaa36\uaa43\uaa4c\uaa4d\uaa50-\uaa59\uaa7b-\uaa7d" +
            "\uaab0\uaab2-\uaab4\uaab7\uaab8\uaabe\uaabf\uaac1" +
            "\uaaeb-\uaaef\uaaf5\uaaf6\uabe3-\uabea\uabec\uabed" +
            "\uabf0-\uabf9\ufb1e\ufe00-\ufe0f\ufe20-\ufe2f\ufe33\ufe34" +
            "\ufe4d-\ufe4f\uff10-\uff19\uff3f\u***REMOVED***101fd***REMOVED***\u***REMOVED***102e0***REMOVED***" +
            "\u***REMOVED***10376***REMOVED***-\u***REMOVED***1037a***REMOVED***\u***REMOVED***104a0***REMOVED***-\u***REMOVED***104a9***REMOVED***\u***REMOVED***10a01***REMOVED***-\u***REMOVED***10a03***REMOVED***" +
            "\u***REMOVED***10a05***REMOVED***\u***REMOVED***10a06***REMOVED***\u***REMOVED***10a0c***REMOVED***-\u***REMOVED***10a0f***REMOVED***\u***REMOVED***10a38***REMOVED***-\u***REMOVED***10a3a***REMOVED***" +
            "\u***REMOVED***10a3f***REMOVED***\u***REMOVED***10ae5***REMOVED***\u***REMOVED***10ae6***REMOVED***\u***REMOVED***10d24***REMOVED***-\u***REMOVED***10d27***REMOVED***" +
            "\u***REMOVED***10d30***REMOVED***-\u***REMOVED***10d39***REMOVED***\u***REMOVED***10f46***REMOVED***-\u***REMOVED***10f50***REMOVED***\u***REMOVED***11000***REMOVED***-\u***REMOVED***11002***REMOVED***" +
            "\u***REMOVED***11038***REMOVED***-\u***REMOVED***11046***REMOVED***\u***REMOVED***11066***REMOVED***-\u***REMOVED***1106f***REMOVED***\u***REMOVED***1107f***REMOVED***-\u***REMOVED***11082***REMOVED***" +
            "\u***REMOVED***110b0***REMOVED***-\u***REMOVED***110ba***REMOVED***\u***REMOVED***110f0***REMOVED***-\u***REMOVED***110f9***REMOVED***\u***REMOVED***11100***REMOVED***-\u***REMOVED***11102***REMOVED***" +
            "\u***REMOVED***11127***REMOVED***-\u***REMOVED***11134***REMOVED***\u***REMOVED***11136***REMOVED***-\u***REMOVED***1113f***REMOVED***\u***REMOVED***11145***REMOVED***\u***REMOVED***11146***REMOVED***" +
            "\u***REMOVED***11173***REMOVED***\u***REMOVED***11180***REMOVED***-\u***REMOVED***11182***REMOVED***\u***REMOVED***111b3***REMOVED***-\u***REMOVED***111c0***REMOVED***" +
            "\u***REMOVED***111c9***REMOVED***-\u***REMOVED***111cc***REMOVED***\u***REMOVED***111d0***REMOVED***-\u***REMOVED***111d9***REMOVED***\u***REMOVED***1122c***REMOVED***-\u***REMOVED***11237***REMOVED***" +
            "\u***REMOVED***1123e***REMOVED***\u***REMOVED***112df***REMOVED***-\u***REMOVED***112ea***REMOVED***\u***REMOVED***112f0***REMOVED***-\u***REMOVED***112f9***REMOVED***" +
            "\u***REMOVED***11300***REMOVED***-\u***REMOVED***11303***REMOVED***\u***REMOVED***1133b***REMOVED***\u***REMOVED***1133c***REMOVED***\u***REMOVED***1133e***REMOVED***-\u***REMOVED***11344***REMOVED***" +
            "\u***REMOVED***11347***REMOVED***\u***REMOVED***11348***REMOVED***\u***REMOVED***1134b***REMOVED***-\u***REMOVED***1134d***REMOVED***\u***REMOVED***11357***REMOVED***" +
            "\u***REMOVED***11362***REMOVED***\u***REMOVED***11363***REMOVED***\u***REMOVED***11366***REMOVED***-\u***REMOVED***1136c***REMOVED***\u***REMOVED***11370***REMOVED***-\u***REMOVED***11374***REMOVED***" +
            "\u***REMOVED***11435***REMOVED***-\u***REMOVED***11446***REMOVED***\u***REMOVED***11450***REMOVED***-\u***REMOVED***11459***REMOVED***\u***REMOVED***1145e***REMOVED***" +
            "\u***REMOVED***114b0***REMOVED***-\u***REMOVED***114c3***REMOVED***\u***REMOVED***114d0***REMOVED***-\u***REMOVED***114d9***REMOVED***\u***REMOVED***115af***REMOVED***-\u***REMOVED***115b5***REMOVED***" +
            "\u***REMOVED***115b8***REMOVED***-\u***REMOVED***115c0***REMOVED***\u***REMOVED***115dc***REMOVED***\u***REMOVED***115dd***REMOVED***\u***REMOVED***11630***REMOVED***-\u***REMOVED***11640***REMOVED***" +
            "\u***REMOVED***11650***REMOVED***-\u***REMOVED***11659***REMOVED***\u***REMOVED***116ab***REMOVED***-\u***REMOVED***116b7***REMOVED***\u***REMOVED***116c0***REMOVED***-\u***REMOVED***116c9***REMOVED***" +
            "\u***REMOVED***1171d***REMOVED***-\u***REMOVED***1172b***REMOVED***\u***REMOVED***11730***REMOVED***-\u***REMOVED***11739***REMOVED***\u***REMOVED***1182c***REMOVED***-\u***REMOVED***1183a***REMOVED***" +
            "\u***REMOVED***118e0***REMOVED***-\u***REMOVED***118e9***REMOVED***\u***REMOVED***119d1***REMOVED***-\u***REMOVED***119d7***REMOVED***\u***REMOVED***119da***REMOVED***-\u***REMOVED***119e0***REMOVED***" +
            "\u***REMOVED***119e4***REMOVED***\u***REMOVED***11a01***REMOVED***-\u***REMOVED***11a0a***REMOVED***\u***REMOVED***11a33***REMOVED***-\u***REMOVED***11a39***REMOVED***" +
            "\u***REMOVED***11a3b***REMOVED***-\u***REMOVED***11a3e***REMOVED***\u***REMOVED***11a47***REMOVED***\u***REMOVED***11a51***REMOVED***-\u***REMOVED***11a5b***REMOVED***" +
            "\u***REMOVED***11a8a***REMOVED***-\u***REMOVED***11a99***REMOVED***\u***REMOVED***11c2f***REMOVED***-\u***REMOVED***11c36***REMOVED***\u***REMOVED***11c38***REMOVED***-\u***REMOVED***11c3f***REMOVED***" +
            "\u***REMOVED***11c50***REMOVED***-\u***REMOVED***11c59***REMOVED***\u***REMOVED***11c92***REMOVED***-\u***REMOVED***11ca7***REMOVED***\u***REMOVED***11ca9***REMOVED***-\u***REMOVED***11cb6***REMOVED***" +
            "\u***REMOVED***11d31***REMOVED***-\u***REMOVED***11d36***REMOVED***\u***REMOVED***11d3a***REMOVED***\u***REMOVED***11d3c***REMOVED***\u***REMOVED***11d3d***REMOVED***" +
            "\u***REMOVED***11d3f***REMOVED***-\u***REMOVED***11d45***REMOVED***\u***REMOVED***11d47***REMOVED***\u***REMOVED***11d50***REMOVED***-\u***REMOVED***11d59***REMOVED***" +
            "\u***REMOVED***11d8a***REMOVED***-\u***REMOVED***11d8e***REMOVED***\u***REMOVED***11d90***REMOVED***\u***REMOVED***11d91***REMOVED***\u***REMOVED***11d93***REMOVED***-\u***REMOVED***11d97***REMOVED***" +
            "\u***REMOVED***11da0***REMOVED***-\u***REMOVED***11da9***REMOVED***\u***REMOVED***11ef3***REMOVED***-\u***REMOVED***11ef6***REMOVED***\u***REMOVED***16a60***REMOVED***-\u***REMOVED***16a69***REMOVED***" +
            "\u***REMOVED***16af0***REMOVED***-\u***REMOVED***16af4***REMOVED***\u***REMOVED***16b30***REMOVED***-\u***REMOVED***16b36***REMOVED***\u***REMOVED***16b50***REMOVED***-\u***REMOVED***16b59***REMOVED***" +
            "\u***REMOVED***16f4f***REMOVED***\u***REMOVED***16f51***REMOVED***-\u***REMOVED***16f87***REMOVED***\u***REMOVED***16f8f***REMOVED***-\u***REMOVED***16f92***REMOVED***" +
            "\u***REMOVED***1bc9d***REMOVED***\u***REMOVED***1bc9e***REMOVED***\u***REMOVED***1d165***REMOVED***-\u***REMOVED***1d169***REMOVED***\u***REMOVED***1d16d***REMOVED***-\u***REMOVED***1d172***REMOVED***" +
            "\u***REMOVED***1d17b***REMOVED***-\u***REMOVED***1d182***REMOVED***\u***REMOVED***1d185***REMOVED***-\u***REMOVED***1d18b***REMOVED***\u***REMOVED***1d1aa***REMOVED***-\u***REMOVED***1d1ad***REMOVED***" +
            "\u***REMOVED***1d242***REMOVED***-\u***REMOVED***1d244***REMOVED***\u***REMOVED***1d7ce***REMOVED***-\u***REMOVED***1d7ff***REMOVED***\u***REMOVED***1da00***REMOVED***-\u***REMOVED***1da36***REMOVED***" +
            "\u***REMOVED***1da3b***REMOVED***-\u***REMOVED***1da6c***REMOVED***\u***REMOVED***1da75***REMOVED***\u***REMOVED***1da84***REMOVED***\u***REMOVED***1da9b***REMOVED***-\u***REMOVED***1da9f***REMOVED***" +
            "\u***REMOVED***1daa1***REMOVED***-\u***REMOVED***1daaf***REMOVED***\u***REMOVED***1e000***REMOVED***-\u***REMOVED***1e006***REMOVED***\u***REMOVED***1e008***REMOVED***-\u***REMOVED***1e018***REMOVED***" +
            "\u***REMOVED***1e01b***REMOVED***-\u***REMOVED***1e021***REMOVED***\u***REMOVED***1e023***REMOVED***\u***REMOVED***1e024***REMOVED***\u***REMOVED***1e026***REMOVED***-\u***REMOVED***1e02a***REMOVED***" +
            "\u***REMOVED***1e130***REMOVED***-\u***REMOVED***1e136***REMOVED***\u***REMOVED***1e140***REMOVED***-\u***REMOVED***1e149***REMOVED***\u***REMOVED***1e2ec***REMOVED***-\u***REMOVED***1e2f9***REMOVED***" +
            "\u***REMOVED***1e8d0***REMOVED***-\u***REMOVED***1e8d6***REMOVED***\u***REMOVED***1e944***REMOVED***-\u***REMOVED***1e94a***REMOVED***\u***REMOVED***1e950***REMOVED***-\u***REMOVED***1e959***REMOVED***" +
            "\u***REMOVED***e0100***REMOVED***-\u***REMOVED***e01ef***REMOVED***]$", "u");
    ***REMOVED***
    return largeIdContinuePattern.test(String.fromCodePoint(cp));
***REMOVED***

const gcNamePattern = /^(?:General_Category|gc)$/u;
const scNamePattern = /^(?:Script(?:_Extensions)?|scx?)$/u;
const gcValuePatterns = ***REMOVED***
    es2018: null,
    es2019: null,
    es2020: null,
***REMOVED***;
const scValuePatterns = ***REMOVED***
    es2018: null,
    es2019: null,
    es2020: null,
***REMOVED***;
const binPropertyPatterns = ***REMOVED***
    es2018: null,
    es2019: null,
    es2020: null,
***REMOVED***;
function isValidUnicodeProperty(version, name, value) ***REMOVED***
    if (gcNamePattern.test(name)) ***REMOVED***
        if (version >= 2018) ***REMOVED***
            if (!gcValuePatterns.es2018) ***REMOVED***
                gcValuePatterns.es2018 = new RegExp("^(?:C|Cased_Letter|Cc|Cf|Close_Punctuation|Cn|Co|" +
                    "Combining_Mark|Connector_Punctuation|Control|Cs|" +
                    "Currency_Symbol|Dash_Punctuation|Decimal_Number|" +
                    "Enclosing_Mark|Final_Punctuation|Format|" +
                    "Initial_Punctuation|L|LC|Letter|Letter_Number|" +
                    "Line_Separator|Ll|Lm|Lo|Lowercase_Letter|Lt|Lu|M|" +
                    "Mark|Math_Symbol|Mc|Me|Mn|Modifier_Letter|" +
                    "Modifier_Symbol|N|Nd|Nl|No|Nonspacing_Mark|Number|" +
                    "Open_Punctuation|Other|Other_Letter|Other_Number|" +
                    "Other_Punctuation|Other_Symbol|P|" +
                    "Paragraph_Separator|Pc|Pd|Pe|Pf|Pi|Po|Private_Use|" +
                    "Ps|Punctuation|S|Sc|Separator|Sk|Sm|So|" +
                    "Space_Separator|Spacing_Mark|Surrogate|Symbol|" +
                    "Titlecase_Letter|Unassigned|Uppercase_Letter|Z|Zl|" +
                    "Zp|Zs|cntrl|digit|punct)$", "u");
            ***REMOVED***
            if (gcValuePatterns.es2018.test(value)) ***REMOVED***
                return true;
            ***REMOVED***
        ***REMOVED***
    ***REMOVED***
    if (scNamePattern.test(name)) ***REMOVED***
        if (version >= 2018) ***REMOVED***
            if (!scValuePatterns.es2018) ***REMOVED***
                scValuePatterns.es2018 = new RegExp("^(?:Adlam|Adlm|Aghb|Ahom|Anatolian_Hieroglyphs|Arab|" +
                    "Arabic|Armenian|Armi|Armn|Avestan|Avst|Bali|" +
                    "Balinese|Bamu|Bamum|Bass|Bassa_Vah|Batak|Batk|Beng|" +
                    "Bengali|Bhaiksuki|Bhks|Bopo|Bopomofo|Brah|Brahmi|" +
                    "Brai|Braille|Bugi|Buginese|Buhd|Buhid|Cakm|" +
                    "Canadian_Aboriginal|Cans|Cari|Carian|" +
                    "Caucasian_Albanian|Chakma|Cham|Cher|Cherokee|Common|" +
                    "Copt|Coptic|Cprt|Cuneiform|Cypriot|Cyrillic|Cyrl|" +
                    "Deseret|Deva|Devanagari|Dsrt|Dupl|Duployan|Egyp|" +
                    "Egyptian_Hieroglyphs|Elba|Elbasan|Ethi|Ethiopic|" +
                    "Geor|Georgian|Glag|Glagolitic|Gonm|Goth|Gothic|Gran|" +
                    "Grantha|Greek|Grek|Gujarati|Gujr|Gurmukhi|Guru|Han|" +
                    "Hang|Hangul|Hani|Hano|Hanunoo|Hatr|Hatran|Hebr|" +
                    "Hebrew|Hira|Hiragana|Hluw|Hmng|Hung|" +
                    "Imperial_Aramaic|Inherited|Inscriptional_Pahlavi|" +
                    "Inscriptional_Parthian|Ital|Java|Javanese|Kaithi|" +
                    "Kali|Kana|Kannada|Katakana|Kayah_Li|Khar|Kharoshthi|" +
                    "Khmer|Khmr|Khoj|Khojki|Khudawadi|Knda|Kthi|Lana|Lao|" +
                    "Laoo|Latin|Latn|Lepc|Lepcha|Limb|Limbu|Lina|Linb|" +
                    "Linear_A|Linear_B|Lisu|Lyci|Lycian|Lydi|Lydian|" +
                    "Mahajani|Mahj|Malayalam|Mand|Mandaic|Mani|" +
                    "Manichaean|Marc|Marchen|Masaram_Gondi|Meetei_Mayek|" +
                    "Mend|Mende_Kikakui|Merc|Mero|Meroitic_Cursive|" +
                    "Meroitic_Hieroglyphs|Miao|Mlym|Modi|Mong|Mongolian|" +
                    "Mro|Mroo|Mtei|Mult|Multani|Myanmar|Mymr|Nabataean|" +
                    "Narb|Nbat|New_Tai_Lue|Newa|Nko|Nkoo|Nshu|Nushu|Ogam|" +
                    "Ogham|Ol_Chiki|Olck|Old_Hungarian|Old_Italic|" +
                    "Old_North_Arabian|Old_Permic|Old_Persian|" +
                    "Old_South_Arabian|Old_Turkic|Oriya|Orkh|Orya|Osage|" +
                    "Osge|Osma|Osmanya|Pahawh_Hmong|Palm|Palmyrene|" +
                    "Pau_Cin_Hau|Pauc|Perm|Phag|Phags_Pa|Phli|Phlp|Phnx|" +
                    "Phoenician|Plrd|Prti|Psalter_Pahlavi|Qaac|Qaai|" +
                    "Rejang|Rjng|Runic|Runr|Samaritan|Samr|Sarb|Saur|" +
                    "Saurashtra|Sgnw|Sharada|Shavian|Shaw|Shrd|Sidd|" +
                    "Siddham|SignWriting|Sind|Sinh|Sinhala|Sora|" +
                    "Sora_Sompeng|Soyo|Soyombo|Sund|Sundanese|Sylo|" +
                    "Syloti_Nagri|Syrc|Syriac|Tagalog|Tagb|Tagbanwa|" +
                    "Tai_Le|Tai_Tham|Tai_Viet|Takr|Takri|Tale|Talu|Tamil|" +
                    "Taml|Tang|Tangut|Tavt|Telu|Telugu|Tfng|Tglg|Thaa|" +
                    "Thaana|Thai|Tibetan|Tibt|Tifinagh|Tirh|Tirhuta|Ugar|" +
                    "Ugaritic|Vai|Vaii|Wara|Warang_Citi|Xpeo|Xsux|Yi|" +
                    "Yiii|Zanabazar_Square|Zanb|Zinh|Zyyy)$", "u");
            ***REMOVED***
            if (scValuePatterns.es2018.test(value)) ***REMOVED***
                return true;
            ***REMOVED***
        ***REMOVED***
        if (version >= 2019) ***REMOVED***
            if (!scValuePatterns.es2019) ***REMOVED***
                scValuePatterns.es2019 = new RegExp("^(?:Dogr|Dogra|Gong|Gunjala_Gondi|Hanifi_Rohingya|" +
                    "Maka|Makasar|Medefaidrin|Medf|Old_Sogdian|Rohg|Sogd|" +
                    "Sogdian|Sogo)$", "u");
            ***REMOVED***
            if (scValuePatterns.es2019.test(value)) ***REMOVED***
                return true;
            ***REMOVED***
        ***REMOVED***
        if (version >= 2020) ***REMOVED***
            if (!scValuePatterns.es2020) ***REMOVED***
                scValuePatterns.es2020 = new RegExp("^(?:Hmnp|Nand|Nandinagari|Nyiakeng_Puachue_Hmong|" +
                    "Wancho|Wcho)$", "u");
            ***REMOVED***
            if (scValuePatterns.es2020.test(value)) ***REMOVED***
                return true;
            ***REMOVED***
        ***REMOVED***
    ***REMOVED***
    return false;
***REMOVED***
function isValidLoneUnicodeProperty(version, value) ***REMOVED***
    if (version >= 2018) ***REMOVED***
        if (!binPropertyPatterns.es2018) ***REMOVED***
            binPropertyPatterns.es2018 = new RegExp("^(?:AHex|ASCII|ASCII_Hex_Digit|Alpha|Alphabetic|Any|" +
                "Assigned|Bidi_C|Bidi_Control|Bidi_M|Bidi_Mirrored|CI|" +
                "CWCF|CWCM|CWKCF|CWL|CWT|CWU|Case_Ignorable|Cased|" +
                "Changes_When_Casefolded|Changes_When_Casemapped|" +
                "Changes_When_Lowercased|Changes_When_NFKC_Casefolded|" +
                "Changes_When_Titlecased|Changes_When_Uppercased|DI|Dash|" +
                "Default_Ignorable_Code_Point|Dep|Deprecated|Dia|" +
                "Diacritic|Emoji|Emoji_Component|Emoji_Modifier|" +
                "Emoji_Modifier_Base|Emoji_Presentation|Ext|Extender|" +
                "Gr_Base|Gr_Ext|Grapheme_Base|Grapheme_Extend|Hex|" +
                "Hex_Digit|IDC|IDS|IDSB|IDST|IDS_Binary_Operator|" +
                "IDS_Trinary_Operator|ID_Continue|ID_Start|Ideo|" +
                "Ideographic|Join_C|Join_Control|LOE|" +
                "Logical_Order_Exception|Lower|Lowercase|Math|NChar|" +
                "Noncharacter_Code_Point|Pat_Syn|Pat_WS|Pattern_Syntax|" +
                "Pattern_White_Space|QMark|Quotation_Mark|RI|Radical|" +
                "Regional_Indicator|SD|STerm|Sentence_Terminal|" +
                "Soft_Dotted|Term|Terminal_Punctuation|UIdeo|" +
                "Unified_Ideograph|Upper|Uppercase|VS|Variation_Selector|" +
                "White_Space|XIDC|XIDS|XID_Continue|XID_Start|space)$", "u");
        ***REMOVED***
        if (binPropertyPatterns.es2018.test(value)) ***REMOVED***
            return true;
        ***REMOVED***
    ***REMOVED***
    if (version >= 2019) ***REMOVED***
        if (!binPropertyPatterns.es2019) ***REMOVED***
            binPropertyPatterns.es2019 = new RegExp("^(?:Extended_Pictographic)$", "u");
        ***REMOVED***
        if (binPropertyPatterns.es2019.test(value)) ***REMOVED***
            return true;
        ***REMOVED***
    ***REMOVED***
    return false;
***REMOVED***

const Backspace = 0x08;
const CharacterTabulation = 0x09;
const LineFeed = 0x0a;
const LineTabulation = 0x0b;
const FormFeed = 0x0c;
const CarriageReturn = 0x0d;
const ExclamationMark = 0x21;
const DollarSign = 0x24;
const LeftParenthesis = 0x28;
const RightParenthesis = 0x29;
const Asterisk = 0x2a;
const PlusSign = 0x2b;
const Comma = 0x2c;
const HyphenMinus = 0x2d;
const FullStop = 0x2e;
const Solidus = 0x2f;
const DigitZero = 0x30;
const DigitOne = 0x31;
const DigitSeven = 0x37;
const DigitNine = 0x39;
const Colon = 0x3a;
const LessThanSign = 0x3c;
const EqualsSign = 0x3d;
const GreaterThanSign = 0x3e;
const QuestionMark = 0x3f;
const LatinCapitalLetterA = 0x41;
const LatinCapitalLetterB = 0x42;
const LatinCapitalLetterD = 0x44;
const LatinCapitalLetterF = 0x46;
const LatinCapitalLetterP = 0x50;
const LatinCapitalLetterS = 0x53;
const LatinCapitalLetterW = 0x57;
const LatinCapitalLetterZ = 0x5a;
const LowLine = 0x5f;
const LatinSmallLetterA = 0x61;
const LatinSmallLetterB = 0x62;
const LatinSmallLetterC = 0x63;
const LatinSmallLetterD = 0x64;
const LatinSmallLetterF = 0x66;
const LatinSmallLetterG = 0x67;
const LatinSmallLetterI = 0x69;
const LatinSmallLetterK = 0x6b;
const LatinSmallLetterM = 0x6d;
const LatinSmallLetterN = 0x6e;
const LatinSmallLetterP = 0x70;
const LatinSmallLetterR = 0x72;
const LatinSmallLetterS = 0x73;
const LatinSmallLetterT = 0x74;
const LatinSmallLetterU = 0x75;
const LatinSmallLetterV = 0x76;
const LatinSmallLetterW = 0x77;
const LatinSmallLetterX = 0x78;
const LatinSmallLetterY = 0x79;
const LatinSmallLetterZ = 0x7a;
const LeftSquareBracket = 0x5b;
const ReverseSolidus = 0x5c;
const RightSquareBracket = 0x5d;
const CircumflexAccent = 0x5e;
const LeftCurlyBracket = 0x7b;
const VerticalLine = 0x7c;
const RightCurlyBracket = 0x7d;
const ZeroWidthNonJoiner = 0x200c;
const ZeroWidthJoiner = 0x200d;
const LineSeparator = 0x2028;
const ParagraphSeparator = 0x2029;
const MinCodePoint = 0x00;
const MaxCodePoint = 0x10ffff;
function isLatinLetter(code) ***REMOVED***
    return ((code >= LatinCapitalLetterA && code <= LatinCapitalLetterZ) ||
        (code >= LatinSmallLetterA && code <= LatinSmallLetterZ));
***REMOVED***
function isDecimalDigit(code) ***REMOVED***
    return code >= DigitZero && code <= DigitNine;
***REMOVED***
function isOctalDigit(code) ***REMOVED***
    return code >= DigitZero && code <= DigitSeven;
***REMOVED***
function isHexDigit(code) ***REMOVED***
    return ((code >= DigitZero && code <= DigitNine) ||
        (code >= LatinCapitalLetterA && code <= LatinCapitalLetterF) ||
        (code >= LatinSmallLetterA && code <= LatinSmallLetterF));
***REMOVED***
function isLineTerminator(code) ***REMOVED***
    return (code === LineFeed ||
        code === CarriageReturn ||
        code === LineSeparator ||
        code === ParagraphSeparator);
***REMOVED***
function isValidUnicode(code) ***REMOVED***
    return code >= MinCodePoint && code <= MaxCodePoint;
***REMOVED***
function digitToInt(code) ***REMOVED***
    if (code >= LatinSmallLetterA && code <= LatinSmallLetterF) ***REMOVED***
        return code - LatinSmallLetterA + 10;
    ***REMOVED***
    if (code >= LatinCapitalLetterA && code <= LatinCapitalLetterF) ***REMOVED***
        return code - LatinCapitalLetterA + 10;
    ***REMOVED***
    return code - DigitZero;
***REMOVED***

const legacyImpl = ***REMOVED***
    at(s, end, i) ***REMOVED***
        return i < end ? s.charCodeAt(i) : -1;
    ***REMOVED***,
    width(c) ***REMOVED***
        return 1;
    ***REMOVED***,
***REMOVED***;
const unicodeImpl = ***REMOVED***
    at(s, end, i) ***REMOVED***
        return i < end ? s.codePointAt(i) : -1;
    ***REMOVED***,
    width(c) ***REMOVED***
        return c > 0xffff ? 2 : 1;
    ***REMOVED***,
***REMOVED***;
class Reader ***REMOVED***
    constructor() ***REMOVED***
        this._impl = legacyImpl;
        this._s = "";
        this._i = 0;
        this._end = 0;
        this._cp1 = -1;
        this._w1 = 1;
        this._cp2 = -1;
        this._w2 = 1;
        this._cp3 = -1;
        this._w3 = 1;
        this._cp4 = -1;
    ***REMOVED***
    get source() ***REMOVED***
        return this._s;
    ***REMOVED***
    get index() ***REMOVED***
        return this._i;
    ***REMOVED***
    get currentCodePoint() ***REMOVED***
        return this._cp1;
    ***REMOVED***
    get nextCodePoint() ***REMOVED***
        return this._cp2;
    ***REMOVED***
    get nextCodePoint2() ***REMOVED***
        return this._cp3;
    ***REMOVED***
    get nextCodePoint3() ***REMOVED***
        return this._cp4;
    ***REMOVED***
    reset(source, start, end, uFlag) ***REMOVED***
        this._impl = uFlag ? unicodeImpl : legacyImpl;
        this._s = source;
        this._end = end;
        this.rewind(start);
    ***REMOVED***
    rewind(index) ***REMOVED***
        const impl = this._impl;
        this._i = index;
        this._cp1 = impl.at(this._s, this._end, index);
        this._w1 = impl.width(this._cp1);
        this._cp2 = impl.at(this._s, this._end, index + this._w1);
        this._w2 = impl.width(this._cp2);
        this._cp3 = impl.at(this._s, this._end, index + this._w1 + this._w2);
        this._w3 = impl.width(this._cp3);
        this._cp4 = impl.at(this._s, this._end, index + this._w1 + this._w2 + this._w3);
    ***REMOVED***
    advance() ***REMOVED***
        if (this._cp1 !== -1) ***REMOVED***
            const impl = this._impl;
            this._i += this._w1;
            this._cp1 = this._cp2;
            this._w1 = this._w2;
            this._cp2 = this._cp3;
            this._w2 = impl.width(this._cp2);
            this._cp3 = this._cp4;
            this._w3 = impl.width(this._cp3);
            this._cp4 = impl.at(this._s, this._end, this._i + this._w1 + this._w2 + this._w3);
        ***REMOVED***
    ***REMOVED***
    eat(cp) ***REMOVED***
        if (this._cp1 === cp) ***REMOVED***
            this.advance();
            return true;
        ***REMOVED***
        return false;
    ***REMOVED***
    eat2(cp1, cp2) ***REMOVED***
        if (this._cp1 === cp1 && this._cp2 === cp2) ***REMOVED***
            this.advance();
            this.advance();
            return true;
        ***REMOVED***
        return false;
    ***REMOVED***
    eat3(cp1, cp2, cp3) ***REMOVED***
        if (this._cp1 === cp1 && this._cp2 === cp2 && this._cp3 === cp3) ***REMOVED***
            this.advance();
            this.advance();
            this.advance();
            return true;
        ***REMOVED***
        return false;
    ***REMOVED***
***REMOVED***

class RegExpSyntaxError extends SyntaxError ***REMOVED***
    constructor(source, uFlag, index, message) ***REMOVED***
        if (source) ***REMOVED***
            if (!source.startsWith("/")) ***REMOVED***
                source = `/$***REMOVED***source***REMOVED***/$***REMOVED***uFlag ? "u" : ""***REMOVED***`;
            ***REMOVED***
            source = `: $***REMOVED***source***REMOVED***`;
        ***REMOVED***
        super(`Invalid regular expression$***REMOVED***source***REMOVED***: $***REMOVED***message***REMOVED***`);
        this.index = index;
    ***REMOVED***
***REMOVED***

function isSyntaxCharacter(cp) ***REMOVED***
    return (cp === CircumflexAccent ||
        cp === DollarSign ||
        cp === ReverseSolidus ||
        cp === FullStop ||
        cp === Asterisk ||
        cp === PlusSign ||
        cp === QuestionMark ||
        cp === LeftParenthesis ||
        cp === RightParenthesis ||
        cp === LeftSquareBracket ||
        cp === RightSquareBracket ||
        cp === LeftCurlyBracket ||
        cp === RightCurlyBracket ||
        cp === VerticalLine);
***REMOVED***
function isRegExpIdentifierStart(cp) ***REMOVED***
    return isIdStart(cp) || cp === DollarSign || cp === LowLine;
***REMOVED***
function isRegExpIdentifierPart(cp) ***REMOVED***
    return (isIdContinue(cp) ||
        cp === DollarSign ||
        cp === LowLine ||
        cp === ZeroWidthNonJoiner ||
        cp === ZeroWidthJoiner);
***REMOVED***
function isUnicodePropertyNameCharacter(cp) ***REMOVED***
    return isLatinLetter(cp) || cp === LowLine;
***REMOVED***
function isUnicodePropertyValueCharacter(cp) ***REMOVED***
    return isUnicodePropertyNameCharacter(cp) || isDecimalDigit(cp);
***REMOVED***
class RegExpValidator ***REMOVED***
    constructor(options) ***REMOVED***
        this._reader = new Reader();
        this._uFlag = false;
        this._nFlag = false;
        this._lastIntValue = 0;
        this._lastMinValue = 0;
        this._lastMaxValue = 0;
        this._lastStrValue = "";
        this._lastKeyValue = "";
        this._lastValValue = "";
        this._lastAssertionIsQuantifiable = false;
        this._numCapturingParens = 0;
        this._groupNames = new Set();
        this._backreferenceNames = new Set();
        this._options = options || ***REMOVED******REMOVED***;
    ***REMOVED***
    validateLiteral(source, start = 0, end = source.length) ***REMOVED***
        this._uFlag = this._nFlag = false;
        this.reset(source, start, end);
        this.onLiteralEnter(start);
        if (this.eat(Solidus) && this.eatRegExpBody() && this.eat(Solidus)) ***REMOVED***
            const flagStart = this.index;
            const uFlag = source.includes("u", flagStart);
            this.validateFlags(source, flagStart, end);
            this.validatePattern(source, start + 1, flagStart - 1, uFlag);
        ***REMOVED***
        else if (start >= end) ***REMOVED***
            this.raise("Empty");
        ***REMOVED***
        else ***REMOVED***
            const c = String.fromCodePoint(this.currentCodePoint);
            this.raise(`Unexpected character '$***REMOVED***c***REMOVED***'`);
        ***REMOVED***
        this.onLiteralLeave(start, end);
    ***REMOVED***
    validateFlags(source, start = 0, end = source.length) ***REMOVED***
        const existingFlags = new Set();
        let global = false;
        let ignoreCase = false;
        let multiline = false;
        let sticky = false;
        let unicode = false;
        let dotAll = false;
        for (let i = start; i < end; ++i) ***REMOVED***
            const flag = source.charCodeAt(i);
            if (existingFlags.has(flag)) ***REMOVED***
                this.raise(`Duplicated flag '$***REMOVED***source[i]***REMOVED***'`);
            ***REMOVED***
            existingFlags.add(flag);
            if (flag === LatinSmallLetterG) ***REMOVED***
                global = true;
            ***REMOVED***
            else if (flag === LatinSmallLetterI) ***REMOVED***
                ignoreCase = true;
            ***REMOVED***
            else if (flag === LatinSmallLetterM) ***REMOVED***
                multiline = true;
            ***REMOVED***
            else if (flag === LatinSmallLetterU && this.ecmaVersion >= 2015) ***REMOVED***
                unicode = true;
            ***REMOVED***
            else if (flag === LatinSmallLetterY && this.ecmaVersion >= 2015) ***REMOVED***
                sticky = true;
            ***REMOVED***
            else if (flag === LatinSmallLetterS && this.ecmaVersion >= 2018) ***REMOVED***
                dotAll = true;
            ***REMOVED***
            else ***REMOVED***
                this.raise(`Invalid flag '$***REMOVED***source[i]***REMOVED***'`);
            ***REMOVED***
        ***REMOVED***
        this.onFlags(start, end, global, ignoreCase, multiline, unicode, sticky, dotAll);
    ***REMOVED***
    validatePattern(source, start = 0, end = source.length, uFlag = false) ***REMOVED***
        this._uFlag = uFlag && this.ecmaVersion >= 2015;
        this._nFlag = uFlag && this.ecmaVersion >= 2018;
        this.reset(source, start, end);
        this.pattern();
        if (!this._nFlag &&
            this.ecmaVersion >= 2018 &&
            this._groupNames.size > 0) ***REMOVED***
            this._nFlag = true;
            this.rewind(start);
            this.pattern();
        ***REMOVED***
    ***REMOVED***
    get strict() ***REMOVED***
        return Boolean(this._options.strict || this._uFlag);
    ***REMOVED***
    get ecmaVersion() ***REMOVED***
        return this._options.ecmaVersion || 2020;
    ***REMOVED***
    onLiteralEnter(start) ***REMOVED***
        if (this._options.onLiteralEnter) ***REMOVED***
            this._options.onLiteralEnter(start);
        ***REMOVED***
    ***REMOVED***
    onLiteralLeave(start, end) ***REMOVED***
        if (this._options.onLiteralLeave) ***REMOVED***
            this._options.onLiteralLeave(start, end);
        ***REMOVED***
    ***REMOVED***
    onFlags(start, end, global, ignoreCase, multiline, unicode, sticky, dotAll) ***REMOVED***
        if (this._options.onFlags) ***REMOVED***
            this._options.onFlags(start, end, global, ignoreCase, multiline, unicode, sticky, dotAll);
        ***REMOVED***
    ***REMOVED***
    onPatternEnter(start) ***REMOVED***
        if (this._options.onPatternEnter) ***REMOVED***
            this._options.onPatternEnter(start);
        ***REMOVED***
    ***REMOVED***
    onPatternLeave(start, end) ***REMOVED***
        if (this._options.onPatternLeave) ***REMOVED***
            this._options.onPatternLeave(start, end);
        ***REMOVED***
    ***REMOVED***
    onDisjunctionEnter(start) ***REMOVED***
        if (this._options.onDisjunctionEnter) ***REMOVED***
            this._options.onDisjunctionEnter(start);
        ***REMOVED***
    ***REMOVED***
    onDisjunctionLeave(start, end) ***REMOVED***
        if (this._options.onDisjunctionLeave) ***REMOVED***
            this._options.onDisjunctionLeave(start, end);
        ***REMOVED***
    ***REMOVED***
    onAlternativeEnter(start, index) ***REMOVED***
        if (this._options.onAlternativeEnter) ***REMOVED***
            this._options.onAlternativeEnter(start, index);
        ***REMOVED***
    ***REMOVED***
    onAlternativeLeave(start, end, index) ***REMOVED***
        if (this._options.onAlternativeLeave) ***REMOVED***
            this._options.onAlternativeLeave(start, end, index);
        ***REMOVED***
    ***REMOVED***
    onGroupEnter(start) ***REMOVED***
        if (this._options.onGroupEnter) ***REMOVED***
            this._options.onGroupEnter(start);
        ***REMOVED***
    ***REMOVED***
    onGroupLeave(start, end) ***REMOVED***
        if (this._options.onGroupLeave) ***REMOVED***
            this._options.onGroupLeave(start, end);
        ***REMOVED***
    ***REMOVED***
    onCapturingGroupEnter(start, name) ***REMOVED***
        if (this._options.onCapturingGroupEnter) ***REMOVED***
            this._options.onCapturingGroupEnter(start, name);
        ***REMOVED***
    ***REMOVED***
    onCapturingGroupLeave(start, end, name) ***REMOVED***
        if (this._options.onCapturingGroupLeave) ***REMOVED***
            this._options.onCapturingGroupLeave(start, end, name);
        ***REMOVED***
    ***REMOVED***
    onQuantifier(start, end, min, max, greedy) ***REMOVED***
        if (this._options.onQuantifier) ***REMOVED***
            this._options.onQuantifier(start, end, min, max, greedy);
        ***REMOVED***
    ***REMOVED***
    onLookaroundAssertionEnter(start, kind, negate) ***REMOVED***
        if (this._options.onLookaroundAssertionEnter) ***REMOVED***
            this._options.onLookaroundAssertionEnter(start, kind, negate);
        ***REMOVED***
    ***REMOVED***
    onLookaroundAssertionLeave(start, end, kind, negate) ***REMOVED***
        if (this._options.onLookaroundAssertionLeave) ***REMOVED***
            this._options.onLookaroundAssertionLeave(start, end, kind, negate);
        ***REMOVED***
    ***REMOVED***
    onEdgeAssertion(start, end, kind) ***REMOVED***
        if (this._options.onEdgeAssertion) ***REMOVED***
            this._options.onEdgeAssertion(start, end, kind);
        ***REMOVED***
    ***REMOVED***
    onWordBoundaryAssertion(start, end, kind, negate) ***REMOVED***
        if (this._options.onWordBoundaryAssertion) ***REMOVED***
            this._options.onWordBoundaryAssertion(start, end, kind, negate);
        ***REMOVED***
    ***REMOVED***
    onAnyCharacterSet(start, end, kind) ***REMOVED***
        if (this._options.onAnyCharacterSet) ***REMOVED***
            this._options.onAnyCharacterSet(start, end, kind);
        ***REMOVED***
    ***REMOVED***
    onEscapeCharacterSet(start, end, kind, negate) ***REMOVED***
        if (this._options.onEscapeCharacterSet) ***REMOVED***
            this._options.onEscapeCharacterSet(start, end, kind, negate);
        ***REMOVED***
    ***REMOVED***
    onUnicodePropertyCharacterSet(start, end, kind, key, value, negate) ***REMOVED***
        if (this._options.onUnicodePropertyCharacterSet) ***REMOVED***
            this._options.onUnicodePropertyCharacterSet(start, end, kind, key, value, negate);
        ***REMOVED***
    ***REMOVED***
    onCharacter(start, end, value) ***REMOVED***
        if (this._options.onCharacter) ***REMOVED***
            this._options.onCharacter(start, end, value);
        ***REMOVED***
    ***REMOVED***
    onBackreference(start, end, ref) ***REMOVED***
        if (this._options.onBackreference) ***REMOVED***
            this._options.onBackreference(start, end, ref);
        ***REMOVED***
    ***REMOVED***
    onCharacterClassEnter(start, negate) ***REMOVED***
        if (this._options.onCharacterClassEnter) ***REMOVED***
            this._options.onCharacterClassEnter(start, negate);
        ***REMOVED***
    ***REMOVED***
    onCharacterClassLeave(start, end, negate) ***REMOVED***
        if (this._options.onCharacterClassLeave) ***REMOVED***
            this._options.onCharacterClassLeave(start, end, negate);
        ***REMOVED***
    ***REMOVED***
    onCharacterClassRange(start, end, min, max) ***REMOVED***
        if (this._options.onCharacterClassRange) ***REMOVED***
            this._options.onCharacterClassRange(start, end, min, max);
        ***REMOVED***
    ***REMOVED***
    get source() ***REMOVED***
        return this._reader.source;
    ***REMOVED***
    get index() ***REMOVED***
        return this._reader.index;
    ***REMOVED***
    get currentCodePoint() ***REMOVED***
        return this._reader.currentCodePoint;
    ***REMOVED***
    get nextCodePoint() ***REMOVED***
        return this._reader.nextCodePoint;
    ***REMOVED***
    get nextCodePoint2() ***REMOVED***
        return this._reader.nextCodePoint2;
    ***REMOVED***
    get nextCodePoint3() ***REMOVED***
        return this._reader.nextCodePoint3;
    ***REMOVED***
    reset(source, start, end) ***REMOVED***
        this._reader.reset(source, start, end, this._uFlag);
    ***REMOVED***
    rewind(index) ***REMOVED***
        this._reader.rewind(index);
    ***REMOVED***
    advance() ***REMOVED***
        this._reader.advance();
    ***REMOVED***
    eat(cp) ***REMOVED***
        return this._reader.eat(cp);
    ***REMOVED***
    eat2(cp1, cp2) ***REMOVED***
        return this._reader.eat2(cp1, cp2);
    ***REMOVED***
    eat3(cp1, cp2, cp3) ***REMOVED***
        return this._reader.eat3(cp1, cp2, cp3);
    ***REMOVED***
    raise(message) ***REMOVED***
        throw new RegExpSyntaxError(this.source, this._uFlag, this.index, message);
    ***REMOVED***
    eatRegExpBody() ***REMOVED***
        const start = this.index;
        let inClass = false;
        let escaped = false;
        for (;;) ***REMOVED***
            const cp = this.currentCodePoint;
            if (cp === -1 || isLineTerminator(cp)) ***REMOVED***
                const kind = inClass ? "character class" : "regular expression";
                this.raise(`Unterminated $***REMOVED***kind***REMOVED***`);
            ***REMOVED***
            if (escaped) ***REMOVED***
                escaped = false;
            ***REMOVED***
            else if (cp === ReverseSolidus) ***REMOVED***
                escaped = true;
            ***REMOVED***
            else if (cp === LeftSquareBracket) ***REMOVED***
                inClass = true;
            ***REMOVED***
            else if (cp === RightSquareBracket) ***REMOVED***
                inClass = false;
            ***REMOVED***
            else if ((cp === Solidus && !inClass) ||
                (cp === Asterisk && this.index === start)) ***REMOVED***
                break;
            ***REMOVED***
            this.advance();
        ***REMOVED***
        return this.index !== start;
    ***REMOVED***
    pattern() ***REMOVED***
        const start = this.index;
        this._numCapturingParens = this.countCapturingParens();
        this._groupNames.clear();
        this._backreferenceNames.clear();
        this.onPatternEnter(start);
        this.disjunction();
        const cp = this.currentCodePoint;
        if (this.currentCodePoint !== -1) ***REMOVED***
            if (cp === RightParenthesis) ***REMOVED***
                this.raise("Unmatched ')'");
            ***REMOVED***
            if (cp === ReverseSolidus) ***REMOVED***
                this.raise("\\ at end of pattern");
            ***REMOVED***
            if (cp === RightSquareBracket || cp === RightCurlyBracket) ***REMOVED***
                this.raise("Lone quantifier brackets");
            ***REMOVED***
            const c = String.fromCodePoint(cp);
            this.raise(`Unexpected character '$***REMOVED***c***REMOVED***'`);
        ***REMOVED***
        for (const name of this._backreferenceNames) ***REMOVED***
            if (!this._groupNames.has(name)) ***REMOVED***
                this.raise("Invalid named capture referenced");
            ***REMOVED***
        ***REMOVED***
        this.onPatternLeave(start, this.index);
    ***REMOVED***
    countCapturingParens() ***REMOVED***
        const start = this.index;
        let inClass = false;
        let escaped = false;
        let count = 0;
        let cp = 0;
        while ((cp = this.currentCodePoint) !== -1) ***REMOVED***
            if (escaped) ***REMOVED***
                escaped = false;
            ***REMOVED***
            else if (cp === ReverseSolidus) ***REMOVED***
                escaped = true;
            ***REMOVED***
            else if (cp === LeftSquareBracket) ***REMOVED***
                inClass = true;
            ***REMOVED***
            else if (cp === RightSquareBracket) ***REMOVED***
                inClass = false;
            ***REMOVED***
            else if (cp === LeftParenthesis &&
                !inClass &&
                (this.nextCodePoint !== QuestionMark ||
                    (this.nextCodePoint2 === LessThanSign &&
                        this.nextCodePoint3 !== EqualsSign &&
                        this.nextCodePoint3 !== ExclamationMark))) ***REMOVED***
                count += 1;
            ***REMOVED***
            this.advance();
        ***REMOVED***
        this.rewind(start);
        return count;
    ***REMOVED***
    disjunction() ***REMOVED***
        const start = this.index;
        let i = 0;
        this.onDisjunctionEnter(start);
        this.alternative(i++);
        while (this.eat(VerticalLine)) ***REMOVED***
            this.alternative(i++);
        ***REMOVED***
        if (this.eatQuantifier(true)) ***REMOVED***
            this.raise("Nothing to repeat");
        ***REMOVED***
        if (this.eat(LeftCurlyBracket)) ***REMOVED***
            this.raise("Lone quantifier brackets");
        ***REMOVED***
        this.onDisjunctionLeave(start, this.index);
    ***REMOVED***
    alternative(i) ***REMOVED***
        const start = this.index;
        this.onAlternativeEnter(start, i);
        while (this.currentCodePoint !== -1 && this.eatTerm()) ***REMOVED***
        ***REMOVED***
        this.onAlternativeLeave(start, this.index, i);
    ***REMOVED***
    eatTerm() ***REMOVED***
        if (this.eatAssertion()) ***REMOVED***
            if (this._lastAssertionIsQuantifiable) ***REMOVED***
                this.eatQuantifier();
            ***REMOVED***
            return true;
        ***REMOVED***
        if (this.strict ? this.eatAtom() : this.eatExtendedAtom()) ***REMOVED***
            this.eatQuantifier();
            return true;
        ***REMOVED***
        return false;
    ***REMOVED***
    eatAssertion() ***REMOVED***
        const start = this.index;
        this._lastAssertionIsQuantifiable = false;
        if (this.eat(CircumflexAccent)) ***REMOVED***
            this.onEdgeAssertion(start, this.index, "start");
            return true;
        ***REMOVED***
        if (this.eat(DollarSign)) ***REMOVED***
            this.onEdgeAssertion(start, this.index, "end");
            return true;
        ***REMOVED***
        if (this.eat2(ReverseSolidus, LatinCapitalLetterB)) ***REMOVED***
            this.onWordBoundaryAssertion(start, this.index, "word", true);
            return true;
        ***REMOVED***
        if (this.eat2(ReverseSolidus, LatinSmallLetterB)) ***REMOVED***
            this.onWordBoundaryAssertion(start, this.index, "word", false);
            return true;
        ***REMOVED***
        if (this.eat2(LeftParenthesis, QuestionMark)) ***REMOVED***
            const lookbehind = this.ecmaVersion >= 2018 && this.eat(LessThanSign);
            let negate = false;
            if (this.eat(EqualsSign) || (negate = this.eat(ExclamationMark))) ***REMOVED***
                const kind = lookbehind ? "lookbehind" : "lookahead";
                this.onLookaroundAssertionEnter(start, kind, negate);
                this.disjunction();
                if (!this.eat(RightParenthesis)) ***REMOVED***
                    this.raise("Unterminated group");
                ***REMOVED***
                this._lastAssertionIsQuantifiable = !lookbehind && !this.strict;
                this.onLookaroundAssertionLeave(start, this.index, kind, negate);
                return true;
            ***REMOVED***
            this.rewind(start);
        ***REMOVED***
        return false;
    ***REMOVED***
    eatQuantifier(noError = false) ***REMOVED***
        const start = this.index;
        let min = 0;
        let max = 0;
        let greedy = false;
        if (this.eat(Asterisk)) ***REMOVED***
            min = 0;
            max = Number.POSITIVE_INFINITY;
        ***REMOVED***
        else if (this.eat(PlusSign)) ***REMOVED***
            min = 1;
            max = Number.POSITIVE_INFINITY;
        ***REMOVED***
        else if (this.eat(QuestionMark)) ***REMOVED***
            min = 0;
            max = 1;
        ***REMOVED***
        else if (this.eatBracedQuantifier(noError)) ***REMOVED***
            min = this._lastMinValue;
            max = this._lastMaxValue;
        ***REMOVED***
        else ***REMOVED***
            return false;
        ***REMOVED***
        greedy = !this.eat(QuestionMark);
        if (!noError) ***REMOVED***
            this.onQuantifier(start, this.index, min, max, greedy);
        ***REMOVED***
        return true;
    ***REMOVED***
    eatBracedQuantifier(noError) ***REMOVED***
        const start = this.index;
        if (this.eat(LeftCurlyBracket)) ***REMOVED***
            this._lastMinValue = 0;
            this._lastMaxValue = Number.POSITIVE_INFINITY;
            if (this.eatDecimalDigits()) ***REMOVED***
                this._lastMinValue = this._lastMaxValue = this._lastIntValue;
                if (this.eat(Comma)) ***REMOVED***
                    this._lastMaxValue = this.eatDecimalDigits()
                        ? this._lastIntValue
                        : Number.POSITIVE_INFINITY;
                ***REMOVED***
                if (this.eat(RightCurlyBracket)) ***REMOVED***
                    if (!noError && this._lastMaxValue < this._lastMinValue) ***REMOVED***
                        this.raise("numbers out of order in ***REMOVED******REMOVED*** quantifier");
                    ***REMOVED***
                    return true;
                ***REMOVED***
            ***REMOVED***
            if (!noError && this.strict) ***REMOVED***
                this.raise("Incomplete quantifier");
            ***REMOVED***
            this.rewind(start);
        ***REMOVED***
        return false;
    ***REMOVED***
    eatAtom() ***REMOVED***
        return (this.eatPatternCharacter() ||
            this.eatDot() ||
            this.eatReverseSolidusAtomEscape() ||
            this.eatCharacterClass() ||
            this.eatUncapturingGroup() ||
            this.eatCapturingGroup());
    ***REMOVED***
    eatDot() ***REMOVED***
        if (this.eat(FullStop)) ***REMOVED***
            this.onAnyCharacterSet(this.index - 1, this.index, "any");
            return true;
        ***REMOVED***
        return false;
    ***REMOVED***
    eatReverseSolidusAtomEscape() ***REMOVED***
        const start = this.index;
        if (this.eat(ReverseSolidus)) ***REMOVED***
            if (this.eatAtomEscape()) ***REMOVED***
                return true;
            ***REMOVED***
            this.rewind(start);
        ***REMOVED***
        return false;
    ***REMOVED***
    eatUncapturingGroup() ***REMOVED***
        const start = this.index;
        if (this.eat3(LeftParenthesis, QuestionMark, Colon)) ***REMOVED***
            this.onGroupEnter(start);
            this.disjunction();
            if (!this.eat(RightParenthesis)) ***REMOVED***
                this.raise("Unterminated group");
            ***REMOVED***
            this.onGroupLeave(start, this.index);
            return true;
        ***REMOVED***
        return false;
    ***REMOVED***
    eatCapturingGroup() ***REMOVED***
        const start = this.index;
        if (this.eat(LeftParenthesis)) ***REMOVED***
            this._lastStrValue = "";
            if (this.ecmaVersion >= 2018) ***REMOVED***
                this.groupSpecifier();
            ***REMOVED***
            else if (this.currentCodePoint === QuestionMark) ***REMOVED***
                this.raise("Invalid group");
            ***REMOVED***
            const name = this._lastStrValue || null;
            this.onCapturingGroupEnter(start, name);
            this.disjunction();
            if (!this.eat(RightParenthesis)) ***REMOVED***
                this.raise("Unterminated group");
            ***REMOVED***
            this.onCapturingGroupLeave(start, this.index, name);
            return true;
        ***REMOVED***
        return false;
    ***REMOVED***
    eatExtendedAtom() ***REMOVED***
        return (this.eatDot() ||
            this.eatReverseSolidusAtomEscape() ||
            this.eatReverseSolidusFollowedByC() ||
            this.eatCharacterClass() ||
            this.eatUncapturingGroup() ||
            this.eatCapturingGroup() ||
            this.eatInvalidBracedQuantifier() ||
            this.eatExtendedPatternCharacter());
    ***REMOVED***
    eatReverseSolidusFollowedByC() ***REMOVED***
        if (this.currentCodePoint === ReverseSolidus &&
            this.nextCodePoint === LatinSmallLetterC) ***REMOVED***
            this._lastIntValue = this.currentCodePoint;
            this.advance();
            this.onCharacter(this.index - 1, this.index, ReverseSolidus);
            return true;
        ***REMOVED***
        return false;
    ***REMOVED***
    eatInvalidBracedQuantifier() ***REMOVED***
        if (this.eatBracedQuantifier(true)) ***REMOVED***
            this.raise("Nothing to repeat");
        ***REMOVED***
        return false;
    ***REMOVED***
    eatSyntaxCharacter() ***REMOVED***
        if (isSyntaxCharacter(this.currentCodePoint)) ***REMOVED***
            this._lastIntValue = this.currentCodePoint;
            this.advance();
            return true;
        ***REMOVED***
        return false;
    ***REMOVED***
    eatPatternCharacter() ***REMOVED***
        const start = this.index;
        const cp = this.currentCodePoint;
        if (cp !== -1 && !isSyntaxCharacter(cp)) ***REMOVED***
            this.advance();
            this.onCharacter(start, this.index, cp);
            return true;
        ***REMOVED***
        return false;
    ***REMOVED***
    eatExtendedPatternCharacter() ***REMOVED***
        const start = this.index;
        const cp = this.currentCodePoint;
        if (cp !== -1 &&
            cp !== CircumflexAccent &&
            cp !== DollarSign &&
            cp !== ReverseSolidus &&
            cp !== FullStop &&
            cp !== Asterisk &&
            cp !== PlusSign &&
            cp !== QuestionMark &&
            cp !== LeftParenthesis &&
            cp !== RightParenthesis &&
            cp !== LeftSquareBracket &&
            cp !== VerticalLine) ***REMOVED***
            this.advance();
            this.onCharacter(start, this.index, cp);
            return true;
        ***REMOVED***
        return false;
    ***REMOVED***
    groupSpecifier() ***REMOVED***
        this._lastStrValue = "";
        if (this.eat(QuestionMark)) ***REMOVED***
            if (this.eatGroupName()) ***REMOVED***
                if (!this._groupNames.has(this._lastStrValue)) ***REMOVED***
                    this._groupNames.add(this._lastStrValue);
                    return;
                ***REMOVED***
                this.raise("Duplicate capture group name");
            ***REMOVED***
            this.raise("Invalid group");
        ***REMOVED***
    ***REMOVED***
    eatGroupName() ***REMOVED***
        this._lastStrValue = "";
        if (this.eat(LessThanSign)) ***REMOVED***
            if (this.eatRegExpIdentifierName() && this.eat(GreaterThanSign)) ***REMOVED***
                return true;
            ***REMOVED***
            this.raise("Invalid capture group name");
        ***REMOVED***
        return false;
    ***REMOVED***
    eatRegExpIdentifierName() ***REMOVED***
        this._lastStrValue = "";
        if (this.eatRegExpIdentifierStart()) ***REMOVED***
            this._lastStrValue += String.fromCodePoint(this._lastIntValue);
            while (this.eatRegExpIdentifierPart()) ***REMOVED***
                this._lastStrValue += String.fromCodePoint(this._lastIntValue);
            ***REMOVED***
            return true;
        ***REMOVED***
        return false;
    ***REMOVED***
    eatRegExpIdentifierStart() ***REMOVED***
        const start = this.index;
        let cp = this.currentCodePoint;
        this.advance();
        if (cp === ReverseSolidus && this.eatRegExpUnicodeEscapeSequence()) ***REMOVED***
            cp = this._lastIntValue;
        ***REMOVED***
        if (isRegExpIdentifierStart(cp)) ***REMOVED***
            this._lastIntValue = cp;
            return true;
        ***REMOVED***
        if (this.index !== start) ***REMOVED***
            this.rewind(start);
        ***REMOVED***
        return false;
    ***REMOVED***
    eatRegExpIdentifierPart() ***REMOVED***
        const start = this.index;
        let cp = this.currentCodePoint;
        this.advance();
        if (cp === ReverseSolidus && this.eatRegExpUnicodeEscapeSequence()) ***REMOVED***
            cp = this._lastIntValue;
        ***REMOVED***
        if (isRegExpIdentifierPart(cp)) ***REMOVED***
            this._lastIntValue = cp;
            return true;
        ***REMOVED***
        if (this.index !== start) ***REMOVED***
            this.rewind(start);
        ***REMOVED***
        return false;
    ***REMOVED***
    eatAtomEscape() ***REMOVED***
        if (this.eatBackreference() ||
            this.eatCharacterClassEscape() ||
            this.eatCharacterEscape() ||
            (this._nFlag && this.eatKGroupName())) ***REMOVED***
            return true;
        ***REMOVED***
        if (this.strict || this._uFlag) ***REMOVED***
            this.raise("Invalid escape");
        ***REMOVED***
        return false;
    ***REMOVED***
    eatBackreference() ***REMOVED***
        const start = this.index;
        if (this.eatDecimalEscape()) ***REMOVED***
            const n = this._lastIntValue;
            if (n <= this._numCapturingParens) ***REMOVED***
                this.onBackreference(start - 1, this.index, n);
                return true;
            ***REMOVED***
            if (this.strict) ***REMOVED***
                this.raise("Invalid escape");
            ***REMOVED***
            this.rewind(start);
        ***REMOVED***
        return false;
    ***REMOVED***
    eatKGroupName() ***REMOVED***
        const start = this.index;
        if (this.eat(LatinSmallLetterK)) ***REMOVED***
            if (this.eatGroupName()) ***REMOVED***
                const groupName = this._lastStrValue;
                this._backreferenceNames.add(groupName);
                this.onBackreference(start - 1, this.index, groupName);
                return true;
            ***REMOVED***
            this.raise("Invalid named reference");
        ***REMOVED***
        return false;
    ***REMOVED***
    eatCharacterEscape() ***REMOVED***
        const start = this.index;
        if (this.eatControlEscape() ||
            this.eatCControlLetter() ||
            this.eatZero() ||
            this.eatHexEscapeSequence() ||
            this.eatRegExpUnicodeEscapeSequence() ||
            (!this.strict && this.eatLegacyOctalEscapeSequence()) ||
            this.eatIdentityEscape()) ***REMOVED***
            this.onCharacter(start - 1, this.index, this._lastIntValue);
            return true;
        ***REMOVED***
        return false;
    ***REMOVED***
    eatCControlLetter() ***REMOVED***
        const start = this.index;
        if (this.eat(LatinSmallLetterC)) ***REMOVED***
            if (this.eatControlLetter()) ***REMOVED***
                return true;
            ***REMOVED***
            this.rewind(start);
        ***REMOVED***
        return false;
    ***REMOVED***
    eatZero() ***REMOVED***
        if (this.currentCodePoint === DigitZero &&
            !isDecimalDigit(this.nextCodePoint)) ***REMOVED***
            this._lastIntValue = 0;
            this.advance();
            return true;
        ***REMOVED***
        return false;
    ***REMOVED***
    eatControlEscape() ***REMOVED***
        if (this.eat(LatinSmallLetterT)) ***REMOVED***
            this._lastIntValue = CharacterTabulation;
            return true;
        ***REMOVED***
        if (this.eat(LatinSmallLetterN)) ***REMOVED***
            this._lastIntValue = LineFeed;
            return true;
        ***REMOVED***
        if (this.eat(LatinSmallLetterV)) ***REMOVED***
            this._lastIntValue = LineTabulation;
            return true;
        ***REMOVED***
        if (this.eat(LatinSmallLetterF)) ***REMOVED***
            this._lastIntValue = FormFeed;
            return true;
        ***REMOVED***
        if (this.eat(LatinSmallLetterR)) ***REMOVED***
            this._lastIntValue = CarriageReturn;
            return true;
        ***REMOVED***
        return false;
    ***REMOVED***
    eatControlLetter() ***REMOVED***
        const cp = this.currentCodePoint;
        if (isLatinLetter(cp)) ***REMOVED***
            this.advance();
            this._lastIntValue = cp % 0x20;
            return true;
        ***REMOVED***
        return false;
    ***REMOVED***
    eatRegExpUnicodeEscapeSequence() ***REMOVED***
        const start = this.index;
        if (this.eat(LatinSmallLetterU)) ***REMOVED***
            if (this.eatFixedHexDigits(4)) ***REMOVED***
                const lead = this._lastIntValue;
                if (this._uFlag && lead >= 0xd800 && lead <= 0xdbff) ***REMOVED***
                    const leadSurrogateEnd = this.index;
                    if (this.eat(ReverseSolidus) &&
                        this.eat(LatinSmallLetterU) &&
                        this.eatFixedHexDigits(4)) ***REMOVED***
                        const trail = this._lastIntValue;
                        if (trail >= 0xdc00 && trail <= 0xdfff) ***REMOVED***
                            this._lastIntValue =
                                (lead - 0xd800) * 0x400 +
                                    (trail - 0xdc00) +
                                    0x10000;
                            return true;
                        ***REMOVED***
                    ***REMOVED***
                    this.rewind(leadSurrogateEnd);
                    this._lastIntValue = lead;
                ***REMOVED***
                return true;
            ***REMOVED***
            if (this._uFlag &&
                this.eat(LeftCurlyBracket) &&
                this.eatHexDigits() &&
                this.eat(RightCurlyBracket) &&
                isValidUnicode(this._lastIntValue)) ***REMOVED***
                return true;
            ***REMOVED***
            if (this.strict || this._uFlag) ***REMOVED***
                this.raise("Invalid unicode escape");
            ***REMOVED***
            this.rewind(start);
        ***REMOVED***
        return false;
    ***REMOVED***
    eatIdentityEscape() ***REMOVED***
        if (this._uFlag) ***REMOVED***
            if (this.eatSyntaxCharacter()) ***REMOVED***
                return true;
            ***REMOVED***
            if (this.eat(Solidus)) ***REMOVED***
                this._lastIntValue = Solidus;
                return true;
            ***REMOVED***
            return false;
        ***REMOVED***
        if (this.isValidIdentityEscape(this.currentCodePoint)) ***REMOVED***
            this._lastIntValue = this.currentCodePoint;
            this.advance();
            return true;
        ***REMOVED***
        return false;
    ***REMOVED***
    isValidIdentityEscape(cp) ***REMOVED***
        if (cp === -1) ***REMOVED***
            return false;
        ***REMOVED***
        if (this.strict) ***REMOVED***
            return !isIdContinue(cp);
        ***REMOVED***
        return (cp !== LatinSmallLetterC &&
            (!this._nFlag || cp !== LatinSmallLetterK));
    ***REMOVED***
    eatDecimalEscape() ***REMOVED***
        this._lastIntValue = 0;
        let cp = this.currentCodePoint;
        if (cp >= DigitOne && cp <= DigitNine) ***REMOVED***
            do ***REMOVED***
                this._lastIntValue = 10 * this._lastIntValue + (cp - DigitZero);
                this.advance();
            ***REMOVED*** while ((cp = this.currentCodePoint) >= DigitZero &&
                cp <= DigitNine);
            return true;
        ***REMOVED***
        return false;
    ***REMOVED***
    eatCharacterClassEscape() ***REMOVED***
        const start = this.index;
        if (this.eat(LatinSmallLetterD)) ***REMOVED***
            this._lastIntValue = -1;
            this.onEscapeCharacterSet(start - 1, this.index, "digit", false);
            return true;
        ***REMOVED***
        if (this.eat(LatinCapitalLetterD)) ***REMOVED***
            this._lastIntValue = -1;
            this.onEscapeCharacterSet(start - 1, this.index, "digit", true);
            return true;
        ***REMOVED***
        if (this.eat(LatinSmallLetterS)) ***REMOVED***
            this._lastIntValue = -1;
            this.onEscapeCharacterSet(start - 1, this.index, "space", false);
            return true;
        ***REMOVED***
        if (this.eat(LatinCapitalLetterS)) ***REMOVED***
            this._lastIntValue = -1;
            this.onEscapeCharacterSet(start - 1, this.index, "space", true);
            return true;
        ***REMOVED***
        if (this.eat(LatinSmallLetterW)) ***REMOVED***
            this._lastIntValue = -1;
            this.onEscapeCharacterSet(start - 1, this.index, "word", false);
            return true;
        ***REMOVED***
        if (this.eat(LatinCapitalLetterW)) ***REMOVED***
            this._lastIntValue = -1;
            this.onEscapeCharacterSet(start - 1, this.index, "word", true);
            return true;
        ***REMOVED***
        let negate = false;
        if (this._uFlag &&
            this.ecmaVersion >= 2018 &&
            (this.eat(LatinSmallLetterP) ||
                (negate = this.eat(LatinCapitalLetterP)))) ***REMOVED***
            this._lastIntValue = -1;
            if (this.eat(LeftCurlyBracket) &&
                this.eatUnicodePropertyValueExpression() &&
                this.eat(RightCurlyBracket)) ***REMOVED***
                this.onUnicodePropertyCharacterSet(start - 1, this.index, "property", this._lastKeyValue, this._lastValValue || null, negate);
                return true;
            ***REMOVED***
            this.raise("Invalid property name");
        ***REMOVED***
        return false;
    ***REMOVED***
    eatUnicodePropertyValueExpression() ***REMOVED***
        const start = this.index;
        if (this.eatUnicodePropertyName() && this.eat(EqualsSign)) ***REMOVED***
            this._lastKeyValue = this._lastStrValue;
            if (this.eatUnicodePropertyValue()) ***REMOVED***
                this._lastValValue = this._lastStrValue;
                if (isValidUnicodeProperty(this.ecmaVersion, this._lastKeyValue, this._lastValValue)) ***REMOVED***
                    return true;
                ***REMOVED***
                this.raise("Invalid property name");
            ***REMOVED***
        ***REMOVED***
        this.rewind(start);
        if (this.eatLoneUnicodePropertyNameOrValue()) ***REMOVED***
            const nameOrValue = this._lastStrValue;
            if (isValidUnicodeProperty(this.ecmaVersion, "General_Category", nameOrValue)) ***REMOVED***
                this._lastKeyValue = "General_Category";
                this._lastValValue = nameOrValue;
                return true;
            ***REMOVED***
            if (isValidLoneUnicodeProperty(this.ecmaVersion, nameOrValue)) ***REMOVED***
                this._lastKeyValue = nameOrValue;
                this._lastValValue = "";
                return true;
            ***REMOVED***
            this.raise("Invalid property name");
        ***REMOVED***
        return false;
    ***REMOVED***
    eatUnicodePropertyName() ***REMOVED***
        this._lastStrValue = "";
        while (isUnicodePropertyNameCharacter(this.currentCodePoint)) ***REMOVED***
            this._lastStrValue += String.fromCodePoint(this.currentCodePoint);
            this.advance();
        ***REMOVED***
        return this._lastStrValue !== "";
    ***REMOVED***
    eatUnicodePropertyValue() ***REMOVED***
        this._lastStrValue = "";
        while (isUnicodePropertyValueCharacter(this.currentCodePoint)) ***REMOVED***
            this._lastStrValue += String.fromCodePoint(this.currentCodePoint);
            this.advance();
        ***REMOVED***
        return this._lastStrValue !== "";
    ***REMOVED***
    eatLoneUnicodePropertyNameOrValue() ***REMOVED***
        return this.eatUnicodePropertyValue();
    ***REMOVED***
    eatCharacterClass() ***REMOVED***
        const start = this.index;
        if (this.eat(LeftSquareBracket)) ***REMOVED***
            const negate = this.eat(CircumflexAccent);
            this.onCharacterClassEnter(start, negate);
            this.classRanges();
            if (!this.eat(RightSquareBracket)) ***REMOVED***
                this.raise("Unterminated character class");
            ***REMOVED***
            this.onCharacterClassLeave(start, this.index, negate);
            return true;
        ***REMOVED***
        return false;
    ***REMOVED***
    classRanges() ***REMOVED***
        let start = this.index;
        while (this.eatClassAtom()) ***REMOVED***
            const left = this._lastIntValue;
            const hyphenStart = this.index;
            if (this.eat(HyphenMinus)) ***REMOVED***
                this.onCharacter(hyphenStart, this.index, HyphenMinus);
                if (this.eatClassAtom()) ***REMOVED***
                    const right = this._lastIntValue;
                    if (left === -1 || right === -1) ***REMOVED***
                        if (this.strict) ***REMOVED***
                            this.raise("Invalid character class");
                        ***REMOVED***
                    ***REMOVED***
                    else if (left > right) ***REMOVED***
                        this.raise("Range out of order in character class");
                    ***REMOVED***
                    else ***REMOVED***
                        this.onCharacterClassRange(start, this.index, left, right);
                    ***REMOVED***
                ***REMOVED***
            ***REMOVED***
            start = this.index;
        ***REMOVED***
    ***REMOVED***
    eatClassAtom() ***REMOVED***
        const start = this.index;
        if (this.eat(ReverseSolidus)) ***REMOVED***
            if (this.eatClassEscape()) ***REMOVED***
                return true;
            ***REMOVED***
            if (this._uFlag) ***REMOVED***
                this.raise("Invalid escape");
            ***REMOVED***
            this.rewind(start);
        ***REMOVED***
        const cp = this.currentCodePoint;
        if (cp !== -1 && cp !== RightSquareBracket) ***REMOVED***
            this.advance();
            this._lastIntValue = cp;
            this.onCharacter(start, this.index, cp);
            return true;
        ***REMOVED***
        return false;
    ***REMOVED***
    eatClassEscape() ***REMOVED***
        const start = this.index;
        if (this.eat(LatinSmallLetterB)) ***REMOVED***
            this._lastIntValue = Backspace;
            this.onCharacter(start - 1, this.index, Backspace);
            return true;
        ***REMOVED***
        if (this._uFlag && this.eat(HyphenMinus)) ***REMOVED***
            this._lastIntValue = HyphenMinus;
            this.onCharacter(start - 1, this.index, HyphenMinus);
            return true;
        ***REMOVED***
        if (!this._uFlag && this.eat(LatinSmallLetterC)) ***REMOVED***
            if (this.eatClassControlLetter()) ***REMOVED***
                this.onCharacter(start - 1, this.index, this._lastIntValue);
                return true;
            ***REMOVED***
            this.rewind(start);
        ***REMOVED***
        return this.eatCharacterClassEscape() || this.eatCharacterEscape();
    ***REMOVED***
    eatClassControlLetter() ***REMOVED***
        const cp = this.currentCodePoint;
        if (isDecimalDigit(cp) || cp === LowLine) ***REMOVED***
            this.advance();
            this._lastIntValue = cp % 0x20;
            return true;
        ***REMOVED***
        return false;
    ***REMOVED***
    eatHexEscapeSequence() ***REMOVED***
        const start = this.index;
        if (this.eat(LatinSmallLetterX)) ***REMOVED***
            if (this.eatFixedHexDigits(2)) ***REMOVED***
                return true;
            ***REMOVED***
            if (this._uFlag) ***REMOVED***
                this.raise("Invalid escape");
            ***REMOVED***
            this.rewind(start);
        ***REMOVED***
        return false;
    ***REMOVED***
    eatDecimalDigits() ***REMOVED***
        const start = this.index;
        this._lastIntValue = 0;
        while (isDecimalDigit(this.currentCodePoint)) ***REMOVED***
            this._lastIntValue =
                10 * this._lastIntValue + digitToInt(this.currentCodePoint);
            this.advance();
        ***REMOVED***
        return this.index !== start;
    ***REMOVED***
    eatHexDigits() ***REMOVED***
        const start = this.index;
        this._lastIntValue = 0;
        while (isHexDigit(this.currentCodePoint)) ***REMOVED***
            this._lastIntValue =
                16 * this._lastIntValue + digitToInt(this.currentCodePoint);
            this.advance();
        ***REMOVED***
        return this.index !== start;
    ***REMOVED***
    eatLegacyOctalEscapeSequence() ***REMOVED***
        if (this.eatOctalDigit()) ***REMOVED***
            const n1 = this._lastIntValue;
            if (this.eatOctalDigit()) ***REMOVED***
                const n2 = this._lastIntValue;
                if (n1 <= 3 && this.eatOctalDigit()) ***REMOVED***
                    this._lastIntValue = n1 * 64 + n2 * 8 + this._lastIntValue;
                ***REMOVED***
                else ***REMOVED***
                    this._lastIntValue = n1 * 8 + n2;
                ***REMOVED***
            ***REMOVED***
            else ***REMOVED***
                this._lastIntValue = n1;
            ***REMOVED***
            return true;
        ***REMOVED***
        return false;
    ***REMOVED***
    eatOctalDigit() ***REMOVED***
        const cp = this.currentCodePoint;
        if (isOctalDigit(cp)) ***REMOVED***
            this.advance();
            this._lastIntValue = cp - DigitZero;
            return true;
        ***REMOVED***
        this._lastIntValue = 0;
        return false;
    ***REMOVED***
    eatFixedHexDigits(length) ***REMOVED***
        const start = this.index;
        this._lastIntValue = 0;
        for (let i = 0; i < length; ++i) ***REMOVED***
            const cp = this.currentCodePoint;
            if (!isHexDigit(cp)) ***REMOVED***
                this.rewind(start);
                return false;
            ***REMOVED***
            this._lastIntValue = 16 * this._lastIntValue + digitToInt(cp);
            this.advance();
        ***REMOVED***
        return true;
    ***REMOVED***
***REMOVED***

const DummyPattern = ***REMOVED******REMOVED***;
const DummyFlags = ***REMOVED******REMOVED***;
const DummyCapturingGroup = ***REMOVED******REMOVED***;
class RegExpParserState ***REMOVED***
    constructor(options) ***REMOVED***
        this._node = DummyPattern;
        this._flags = DummyFlags;
        this._backreferences = [];
        this._capturingGroups = [];
        this.source = "";
        this.strict = Boolean(options && options.strict);
        this.ecmaVersion = (options && options.ecmaVersion) || 2020;
    ***REMOVED***
    get pattern() ***REMOVED***
        if (this._node.type !== "Pattern") ***REMOVED***
            throw new Error("UnknownError");
        ***REMOVED***
        return this._node;
    ***REMOVED***
    get flags() ***REMOVED***
        if (this._flags.type !== "Flags") ***REMOVED***
            throw new Error("UnknownError");
        ***REMOVED***
        return this._flags;
    ***REMOVED***
    onFlags(start, end, global, ignoreCase, multiline, unicode, sticky, dotAll) ***REMOVED***
        this._flags = ***REMOVED***
            type: "Flags",
            parent: null,
            start,
            end,
            raw: this.source.slice(start, end),
            global,
            ignoreCase,
            multiline,
            unicode,
            sticky,
            dotAll,
        ***REMOVED***;
    ***REMOVED***
    onPatternEnter(start) ***REMOVED***
        this._node = ***REMOVED***
            type: "Pattern",
            parent: null,
            start,
            end: start,
            raw: "",
            alternatives: [],
        ***REMOVED***;
        this._backreferences.length = 0;
        this._capturingGroups.length = 0;
    ***REMOVED***
    onPatternLeave(start, end) ***REMOVED***
        this._node.end = end;
        this._node.raw = this.source.slice(start, end);
        for (const reference of this._backreferences) ***REMOVED***
            const ref = reference.ref;
            const group = typeof ref === "number"
                ? this._capturingGroups[ref - 1]
                : this._capturingGroups.find(g => g.name === ref);
            reference.resolved = group;
            group.references.push(reference);
        ***REMOVED***
    ***REMOVED***
    onAlternativeEnter(start) ***REMOVED***
        const parent = this._node;
        if (parent.type !== "Assertion" &&
            parent.type !== "CapturingGroup" &&
            parent.type !== "Group" &&
            parent.type !== "Pattern") ***REMOVED***
            throw new Error("UnknownError");
        ***REMOVED***
        this._node = ***REMOVED***
            type: "Alternative",
            parent,
            start,
            end: start,
            raw: "",
            elements: [],
        ***REMOVED***;
        parent.alternatives.push(this._node);
    ***REMOVED***
    onAlternativeLeave(start, end) ***REMOVED***
        const node = this._node;
        if (node.type !== "Alternative") ***REMOVED***
            throw new Error("UnknownError");
        ***REMOVED***
        node.end = end;
        node.raw = this.source.slice(start, end);
        this._node = node.parent;
    ***REMOVED***
    onGroupEnter(start) ***REMOVED***
        const parent = this._node;
        if (parent.type !== "Alternative") ***REMOVED***
            throw new Error("UnknownError");
        ***REMOVED***
        this._node = ***REMOVED***
            type: "Group",
            parent,
            start,
            end: start,
            raw: "",
            alternatives: [],
        ***REMOVED***;
        parent.elements.push(this._node);
    ***REMOVED***
    onGroupLeave(start, end) ***REMOVED***
        const node = this._node;
        if (node.type !== "Group" || node.parent.type !== "Alternative") ***REMOVED***
            throw new Error("UnknownError");
        ***REMOVED***
        node.end = end;
        node.raw = this.source.slice(start, end);
        this._node = node.parent;
    ***REMOVED***
    onCapturingGroupEnter(start, name) ***REMOVED***
        const parent = this._node;
        if (parent.type !== "Alternative") ***REMOVED***
            throw new Error("UnknownError");
        ***REMOVED***
        this._node = ***REMOVED***
            type: "CapturingGroup",
            parent,
            start,
            end: start,
            raw: "",
            name,
            alternatives: [],
            references: [],
        ***REMOVED***;
        parent.elements.push(this._node);
        this._capturingGroups.push(this._node);
    ***REMOVED***
    onCapturingGroupLeave(start, end) ***REMOVED***
        const node = this._node;
        if (node.type !== "CapturingGroup" ||
            node.parent.type !== "Alternative") ***REMOVED***
            throw new Error("UnknownError");
        ***REMOVED***
        node.end = end;
        node.raw = this.source.slice(start, end);
        this._node = node.parent;
    ***REMOVED***
    onQuantifier(start, end, min, max, greedy) ***REMOVED***
        const parent = this._node;
        if (parent.type !== "Alternative") ***REMOVED***
            throw new Error("UnknownError");
        ***REMOVED***
        const element = parent.elements.pop();
        if (element == null ||
            element.type === "Quantifier" ||
            (element.type === "Assertion" && element.kind !== "lookahead")) ***REMOVED***
            throw new Error("UnknownError");
        ***REMOVED***
        const node = ***REMOVED***
            type: "Quantifier",
            parent,
            start: element.start,
            end,
            raw: this.source.slice(element.start, end),
            min,
            max,
            greedy,
            element,
        ***REMOVED***;
        parent.elements.push(node);
        element.parent = node;
    ***REMOVED***
    onLookaroundAssertionEnter(start, kind, negate) ***REMOVED***
        const parent = this._node;
        if (parent.type !== "Alternative") ***REMOVED***
            throw new Error("UnknownError");
        ***REMOVED***
        const node = (this._node = ***REMOVED***
            type: "Assertion",
            parent,
            start,
            end: start,
            raw: "",
            kind,
            negate,
            alternatives: [],
        ***REMOVED***);
        parent.elements.push(node);
    ***REMOVED***
    onLookaroundAssertionLeave(start, end) ***REMOVED***
        const node = this._node;
        if (node.type !== "Assertion" || node.parent.type !== "Alternative") ***REMOVED***
            throw new Error("UnknownError");
        ***REMOVED***
        node.end = end;
        node.raw = this.source.slice(start, end);
        this._node = node.parent;
    ***REMOVED***
    onEdgeAssertion(start, end, kind) ***REMOVED***
        const parent = this._node;
        if (parent.type !== "Alternative") ***REMOVED***
            throw new Error("UnknownError");
        ***REMOVED***
        parent.elements.push(***REMOVED***
            type: "Assertion",
            parent,
            start,
            end,
            raw: this.source.slice(start, end),
            kind,
        ***REMOVED***);
    ***REMOVED***
    onWordBoundaryAssertion(start, end, kind, negate) ***REMOVED***
        const parent = this._node;
        if (parent.type !== "Alternative") ***REMOVED***
            throw new Error("UnknownError");
        ***REMOVED***
        parent.elements.push(***REMOVED***
            type: "Assertion",
            parent,
            start,
            end,
            raw: this.source.slice(start, end),
            kind,
            negate,
        ***REMOVED***);
    ***REMOVED***
    onAnyCharacterSet(start, end, kind) ***REMOVED***
        const parent = this._node;
        if (parent.type !== "Alternative") ***REMOVED***
            throw new Error("UnknownError");
        ***REMOVED***
        parent.elements.push(***REMOVED***
            type: "CharacterSet",
            parent,
            start,
            end,
            raw: this.source.slice(start, end),
            kind,
        ***REMOVED***);
    ***REMOVED***
    onEscapeCharacterSet(start, end, kind, negate) ***REMOVED***
        const parent = this._node;
        if (parent.type !== "Alternative" && parent.type !== "CharacterClass") ***REMOVED***
            throw new Error("UnknownError");
        ***REMOVED***
        parent.elements.push(***REMOVED***
            type: "CharacterSet",
            parent,
            start,
            end,
            raw: this.source.slice(start, end),
            kind,
            negate,
        ***REMOVED***);
    ***REMOVED***
    onUnicodePropertyCharacterSet(start, end, kind, key, value, negate) ***REMOVED***
        const parent = this._node;
        if (parent.type !== "Alternative" && parent.type !== "CharacterClass") ***REMOVED***
            throw new Error("UnknownError");
        ***REMOVED***
        parent.elements.push(***REMOVED***
            type: "CharacterSet",
            parent,
            start,
            end,
            raw: this.source.slice(start, end),
            kind,
            key,
            value,
            negate,
        ***REMOVED***);
    ***REMOVED***
    onCharacter(start, end, value) ***REMOVED***
        const parent = this._node;
        if (parent.type !== "Alternative" && parent.type !== "CharacterClass") ***REMOVED***
            throw new Error("UnknownError");
        ***REMOVED***
        parent.elements.push(***REMOVED***
            type: "Character",
            parent,
            start,
            end,
            raw: this.source.slice(start, end),
            value,
        ***REMOVED***);
    ***REMOVED***
    onBackreference(start, end, ref) ***REMOVED***
        const parent = this._node;
        if (parent.type !== "Alternative") ***REMOVED***
            throw new Error("UnknownError");
        ***REMOVED***
        const node = ***REMOVED***
            type: "Backreference",
            parent,
            start,
            end,
            raw: this.source.slice(start, end),
            ref,
            resolved: DummyCapturingGroup,
        ***REMOVED***;
        parent.elements.push(node);
        this._backreferences.push(node);
    ***REMOVED***
    onCharacterClassEnter(start, negate) ***REMOVED***
        const parent = this._node;
        if (parent.type !== "Alternative") ***REMOVED***
            throw new Error("UnknownError");
        ***REMOVED***
        this._node = ***REMOVED***
            type: "CharacterClass",
            parent,
            start,
            end: start,
            raw: "",
            negate,
            elements: [],
        ***REMOVED***;
        parent.elements.push(this._node);
    ***REMOVED***
    onCharacterClassLeave(start, end) ***REMOVED***
        const node = this._node;
        if (node.type !== "CharacterClass" ||
            node.parent.type !== "Alternative") ***REMOVED***
            throw new Error("UnknownError");
        ***REMOVED***
        node.end = end;
        node.raw = this.source.slice(start, end);
        this._node = node.parent;
    ***REMOVED***
    onCharacterClassRange(start, end) ***REMOVED***
        const parent = this._node;
        if (parent.type !== "CharacterClass") ***REMOVED***
            throw new Error("UnknownError");
        ***REMOVED***
        const elements = parent.elements;
        const max = elements.pop();
        const hyphen = elements.pop();
        const min = elements.pop();
        if (!min ||
            !max ||
            !hyphen ||
            min.type !== "Character" ||
            max.type !== "Character" ||
            hyphen.type !== "Character" ||
            hyphen.value !== HyphenMinus) ***REMOVED***
            throw new Error("UnknownError");
        ***REMOVED***
        const node = ***REMOVED***
            type: "CharacterClassRange",
            parent,
            start,
            end,
            raw: this.source.slice(start, end),
            min,
            max,
        ***REMOVED***;
        min.parent = node;
        max.parent = node;
        elements.push(node);
    ***REMOVED***
***REMOVED***
class RegExpParser ***REMOVED***
    constructor(options) ***REMOVED***
        this._state = new RegExpParserState(options);
        this._validator = new RegExpValidator(this._state);
    ***REMOVED***
    parseLiteral(source, start = 0, end = source.length) ***REMOVED***
        this._state.source = source;
        this._validator.validateLiteral(source, start, end);
        const pattern = this._state.pattern;
        const flags = this._state.flags;
        const literal = ***REMOVED***
            type: "RegExpLiteral",
            parent: null,
            start,
            end,
            raw: source,
            pattern,
            flags,
        ***REMOVED***;
        pattern.parent = literal;
        flags.parent = literal;
        return literal;
    ***REMOVED***
    parseFlags(source, start = 0, end = source.length) ***REMOVED***
        this._state.source = source;
        this._validator.validateFlags(source, start, end);
        return this._state.flags;
    ***REMOVED***
    parsePattern(source, start = 0, end = source.length, uFlag = false) ***REMOVED***
        this._state.source = source;
        this._validator.validatePattern(source, start, end, uFlag);
        return this._state.pattern;
    ***REMOVED***
***REMOVED***

class RegExpVisitor ***REMOVED***
    constructor(handlers) ***REMOVED***
        this._handlers = handlers;
    ***REMOVED***
    visit(node) ***REMOVED***
        switch (node.type) ***REMOVED***
            case "Alternative":
                this.visitAlternative(node);
                break;
            case "Assertion":
                this.visitAssertion(node);
                break;
            case "Backreference":
                this.visitBackreference(node);
                break;
            case "CapturingGroup":
                this.visitCapturingGroup(node);
                break;
            case "Character":
                this.visitCharacter(node);
                break;
            case "CharacterClass":
                this.visitCharacterClass(node);
                break;
            case "CharacterClassRange":
                this.visitCharacterClassRange(node);
                break;
            case "CharacterSet":
                this.visitCharacterSet(node);
                break;
            case "Flags":
                this.visitFlags(node);
                break;
            case "Group":
                this.visitGroup(node);
                break;
            case "Pattern":
                this.visitPattern(node);
                break;
            case "Quantifier":
                this.visitQuantifier(node);
                break;
            case "RegExpLiteral":
                this.visitRegExpLiteral(node);
                break;
            default:
                throw new Error(`Unknown type: $***REMOVED***node.type***REMOVED***`);
        ***REMOVED***
    ***REMOVED***
    visitAlternative(node) ***REMOVED***
        if (this._handlers.onAlternativeEnter) ***REMOVED***
            this._handlers.onAlternativeEnter(node);
        ***REMOVED***
        node.elements.forEach(this.visit, this);
        if (this._handlers.onAlternativeLeave) ***REMOVED***
            this._handlers.onAlternativeLeave(node);
        ***REMOVED***
    ***REMOVED***
    visitAssertion(node) ***REMOVED***
        if (this._handlers.onAssertionEnter) ***REMOVED***
            this._handlers.onAssertionEnter(node);
        ***REMOVED***
        if (node.kind === "lookahead" || node.kind === "lookbehind") ***REMOVED***
            node.alternatives.forEach(this.visit, this);
        ***REMOVED***
        if (this._handlers.onAssertionLeave) ***REMOVED***
            this._handlers.onAssertionLeave(node);
        ***REMOVED***
    ***REMOVED***
    visitBackreference(node) ***REMOVED***
        if (this._handlers.onBackreferenceEnter) ***REMOVED***
            this._handlers.onBackreferenceEnter(node);
        ***REMOVED***
        if (this._handlers.onBackreferenceLeave) ***REMOVED***
            this._handlers.onBackreferenceLeave(node);
        ***REMOVED***
    ***REMOVED***
    visitCapturingGroup(node) ***REMOVED***
        if (this._handlers.onCapturingGroupEnter) ***REMOVED***
            this._handlers.onCapturingGroupEnter(node);
        ***REMOVED***
        node.alternatives.forEach(this.visit, this);
        if (this._handlers.onCapturingGroupLeave) ***REMOVED***
            this._handlers.onCapturingGroupLeave(node);
        ***REMOVED***
    ***REMOVED***
    visitCharacter(node) ***REMOVED***
        if (this._handlers.onCharacterEnter) ***REMOVED***
            this._handlers.onCharacterEnter(node);
        ***REMOVED***
        if (this._handlers.onCharacterLeave) ***REMOVED***
            this._handlers.onCharacterLeave(node);
        ***REMOVED***
    ***REMOVED***
    visitCharacterClass(node) ***REMOVED***
        if (this._handlers.onCharacterClassEnter) ***REMOVED***
            this._handlers.onCharacterClassEnter(node);
        ***REMOVED***
        node.elements.forEach(this.visit, this);
        if (this._handlers.onCharacterClassLeave) ***REMOVED***
            this._handlers.onCharacterClassLeave(node);
        ***REMOVED***
    ***REMOVED***
    visitCharacterClassRange(node) ***REMOVED***
        if (this._handlers.onCharacterClassRangeEnter) ***REMOVED***
            this._handlers.onCharacterClassRangeEnter(node);
        ***REMOVED***
        this.visitCharacter(node.min);
        this.visitCharacter(node.max);
        if (this._handlers.onCharacterClassRangeLeave) ***REMOVED***
            this._handlers.onCharacterClassRangeLeave(node);
        ***REMOVED***
    ***REMOVED***
    visitCharacterSet(node) ***REMOVED***
        if (this._handlers.onCharacterSetEnter) ***REMOVED***
            this._handlers.onCharacterSetEnter(node);
        ***REMOVED***
        if (this._handlers.onCharacterSetLeave) ***REMOVED***
            this._handlers.onCharacterSetLeave(node);
        ***REMOVED***
    ***REMOVED***
    visitFlags(node) ***REMOVED***
        if (this._handlers.onFlagsEnter) ***REMOVED***
            this._handlers.onFlagsEnter(node);
        ***REMOVED***
        if (this._handlers.onFlagsLeave) ***REMOVED***
            this._handlers.onFlagsLeave(node);
        ***REMOVED***
    ***REMOVED***
    visitGroup(node) ***REMOVED***
        if (this._handlers.onGroupEnter) ***REMOVED***
            this._handlers.onGroupEnter(node);
        ***REMOVED***
        node.alternatives.forEach(this.visit, this);
        if (this._handlers.onGroupLeave) ***REMOVED***
            this._handlers.onGroupLeave(node);
        ***REMOVED***
    ***REMOVED***
    visitPattern(node) ***REMOVED***
        if (this._handlers.onPatternEnter) ***REMOVED***
            this._handlers.onPatternEnter(node);
        ***REMOVED***
        node.alternatives.forEach(this.visit, this);
        if (this._handlers.onPatternLeave) ***REMOVED***
            this._handlers.onPatternLeave(node);
        ***REMOVED***
    ***REMOVED***
    visitQuantifier(node) ***REMOVED***
        if (this._handlers.onQuantifierEnter) ***REMOVED***
            this._handlers.onQuantifierEnter(node);
        ***REMOVED***
        this.visit(node.element);
        if (this._handlers.onQuantifierLeave) ***REMOVED***
            this._handlers.onQuantifierLeave(node);
        ***REMOVED***
    ***REMOVED***
    visitRegExpLiteral(node) ***REMOVED***
        if (this._handlers.onRegExpLiteralEnter) ***REMOVED***
            this._handlers.onRegExpLiteralEnter(node);
        ***REMOVED***
        this.visitPattern(node.pattern);
        this.visitFlags(node.flags);
        if (this._handlers.onRegExpLiteralLeave) ***REMOVED***
            this._handlers.onRegExpLiteralLeave(node);
        ***REMOVED***
    ***REMOVED***
***REMOVED***

function parseRegExpLiteral(source, options) ***REMOVED***
    return new RegExpParser(options).parseLiteral(String(source));
***REMOVED***
function validateRegExpLiteral(source, options) ***REMOVED***
    return new RegExpValidator(options).validateLiteral(source);
***REMOVED***
function visitRegExpAST(node, handlers) ***REMOVED***
    new RegExpVisitor(handlers).visit(node);
***REMOVED***

export ***REMOVED*** ast as AST, RegExpParser, RegExpValidator, parseRegExpLiteral, validateRegExpLiteral, visitRegExpAST ***REMOVED***;
