// eslint/lib/init/config-file.js
/**
 * @fileoverview Helper to locate and load configuration files.
 * @author Nicholas C. Zakas
 */

"use strict";

//------------------------------------------------------------------------------
// Requirements
//------------------------------------------------------------------------------

const fs = require("fs"),
    path = require("path"),
    stringify = require("json-stable-stringify-without-jsonify");

const debug = require("debug")("eslint:config-file");

//------------------------------------------------------------------------------
// Helpers
//------------------------------------------------------------------------------

/**
 * Determines sort order for object keys for json-stable-stringify
 *
 * see: https://github.com/samn/json-stable-stringify#cmp
 *
 * @param   ***REMOVED***Object***REMOVED*** a The first comparison object (***REMOVED***key: akey, value: avalue***REMOVED***)
 * @param   ***REMOVED***Object***REMOVED*** b The second comparison object (***REMOVED***key: bkey, value: bvalue***REMOVED***)
 * @returns ***REMOVED***number***REMOVED***   1 or -1, used in stringify cmp method
 */
function sortByKey(a, b) ***REMOVED***
    return a.key > b.key ? 1 : -1;
***REMOVED***

//------------------------------------------------------------------------------
// Private
//------------------------------------------------------------------------------

/**
 * Writes a configuration file in JSON format.
 * @param ***REMOVED***Object***REMOVED*** config The configuration object to write.
 * @param ***REMOVED***string***REMOVED*** filePath The filename to write to.
 * @returns ***REMOVED***void***REMOVED***
 * @private
 */
function writeJSONConfigFile(config, filePath) ***REMOVED***
    debug(`Writing JSON config file: $***REMOVED***filePath***REMOVED***`);

    const content = stringify(config, ***REMOVED*** cmp: sortByKey, space: 4 ***REMOVED***);

    fs.writeFileSync(filePath, content, "utf8");
***REMOVED***

/**
 * Writes a configuration file in YAML format.
 * @param ***REMOVED***Object***REMOVED*** config The configuration object to write.
 * @param ***REMOVED***string***REMOVED*** filePath The filename to write to.
 * @returns ***REMOVED***void***REMOVED***
 * @private
 */
function writeYAMLConfigFile(config, filePath) ***REMOVED***
    debug(`Writing YAML config file: $***REMOVED***filePath***REMOVED***`);

    // lazy load YAML to improve performance when not used
    const yaml = require("js-yaml");

    const content = yaml.safeDump(config, ***REMOVED*** sortKeys: true ***REMOVED***);

    fs.writeFileSync(filePath, content, "utf8");
***REMOVED***

/**
 * Writes a configuration file in JavaScript format.
 * @param ***REMOVED***Object***REMOVED*** config The configuration object to write.
 * @param ***REMOVED***string***REMOVED*** filePath The filename to write to.
 * @throws ***REMOVED***Error***REMOVED*** If an error occurs linting the config file contents.
 * @returns ***REMOVED***void***REMOVED***
 * @private
 */
function writeJSConfigFile(config, filePath) ***REMOVED***
    debug(`Writing JS config file: $***REMOVED***filePath***REMOVED***`);

    let contentToWrite;
    const stringifiedContent = `module.exports = $***REMOVED***stringify(config, ***REMOVED*** cmp: sortByKey, space: 4 ***REMOVED***)***REMOVED***;`;

    try ***REMOVED***
        const ***REMOVED*** CLIEngine ***REMOVED*** = require("../cli-engine");
        const linter = new CLIEngine(***REMOVED***
            baseConfig: config,
            fix: true,
            useEslintrc: false
        ***REMOVED***);
        const report = linter.executeOnText(stringifiedContent);

        contentToWrite = report.results[0].output || stringifiedContent;
    ***REMOVED*** catch (e) ***REMOVED***
        debug("Error linting JavaScript config file, writing unlinted version");
        const errorMessage = e.message;

        contentToWrite = stringifiedContent;
        e.message = "An error occurred while generating your JavaScript config file. ";
        e.message += "A config file was still generated, but the config file itself may not follow your linting rules.";
        e.message += `\nError: $***REMOVED***errorMessage***REMOVED***`;
        throw e;
    ***REMOVED*** finally ***REMOVED***
        fs.writeFileSync(filePath, contentToWrite, "utf8");
    ***REMOVED***
***REMOVED***

/**
 * Writes a configuration file.
 * @param ***REMOVED***Object***REMOVED*** config The configuration object to write.
 * @param ***REMOVED***string***REMOVED*** filePath The filename to write to.
 * @returns ***REMOVED***void***REMOVED***
 * @throws ***REMOVED***Error***REMOVED*** When an unknown file type is specified.
 * @private
 */
function write(config, filePath) ***REMOVED***
    switch (path.extname(filePath)) ***REMOVED***
        case ".js":
            writeJSConfigFile(config, filePath);
            break;

        case ".json":
            writeJSONConfigFile(config, filePath);
            break;

        case ".yaml":
        case ".yml":
            writeYAMLConfigFile(config, filePath);
            break;

        default:
            throw new Error("Can't write to unknown file type.");
    ***REMOVED***
***REMOVED***

//------------------------------------------------------------------------------
// Public Interface
//------------------------------------------------------------------------------

module.exports = ***REMOVED***
    write
***REMOVED***;
