// eslint-plugin-import/lib/rules/max-dependencies.js
'use strict';

var _slicedToArray = function () ***REMOVED*** function sliceIterator(arr, i) ***REMOVED*** var _arr = []; var _n = true; var _d = false; var _e = undefined; try ***REMOVED*** for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) ***REMOVED*** _arr.push(_s.value); if (i && _arr.length === i) break; ***REMOVED*** ***REMOVED*** catch (err) ***REMOVED*** _d = true; _e = err; ***REMOVED*** finally ***REMOVED*** try ***REMOVED*** if (!_n && _i["return"]) _i["return"](); ***REMOVED*** finally ***REMOVED*** if (_d) throw _e; ***REMOVED*** ***REMOVED*** return _arr; ***REMOVED*** return function (arr, i) ***REMOVED*** if (Array.isArray(arr)) ***REMOVED*** return arr; ***REMOVED*** else if (Symbol.iterator in Object(arr)) ***REMOVED*** return sliceIterator(arr, i); ***REMOVED*** else ***REMOVED*** throw new TypeError("Invalid attempt to destructure non-iterable instance"); ***REMOVED*** ***REMOVED***; ***REMOVED***();

var _staticRequire = require('../core/staticRequire');

var _staticRequire2 = _interopRequireDefault(_staticRequire);

var _docsUrl = require('../docsUrl');

var _docsUrl2 = _interopRequireDefault(_docsUrl);

function _interopRequireDefault(obj) ***REMOVED*** return obj && obj.__esModule ? obj : ***REMOVED*** default: obj ***REMOVED***; ***REMOVED***

const DEFAULT_MAX = 10;

const countDependencies = (dependencies, lastNode, context) => ***REMOVED***
  var _ref = context.options[0] || ***REMOVED*** max: DEFAULT_MAX ***REMOVED***;

  const max = _ref.max;


  if (dependencies.size > max) ***REMOVED***
    context.report(lastNode, `Maximum number of dependencies ($***REMOVED***max***REMOVED***) exceeded.`);
  ***REMOVED***
***REMOVED***;

module.exports = ***REMOVED***
  meta: ***REMOVED***
    type: 'suggestion',
    docs: ***REMOVED***
      url: (0, _docsUrl2.default)('max-dependencies')
    ***REMOVED***,

    schema: [***REMOVED***
      'type': 'object',
      'properties': ***REMOVED***
        'max': ***REMOVED*** 'type': 'number' ***REMOVED***
      ***REMOVED***,
      'additionalProperties': false
    ***REMOVED***]
  ***REMOVED***,

  create: context => ***REMOVED***
    const dependencies = new Set(); // keep track of dependencies
    let lastNode; // keep track of the last node to report on

    return ***REMOVED***
      ImportDeclaration(node) ***REMOVED***
        dependencies.add(node.source.value);
        lastNode = node.source;
      ***REMOVED***,

      CallExpression(node) ***REMOVED***
        if ((0, _staticRequire2.default)(node)) ***REMOVED***
          var _node$arguments = _slicedToArray(node.arguments, 1);

          const requirePath = _node$arguments[0];

          dependencies.add(requirePath.value);
          lastNode = node;
        ***REMOVED***
      ***REMOVED***,

      'Program:exit': function () ***REMOVED***
        countDependencies(dependencies, lastNode, context);
      ***REMOVED***
    ***REMOVED***;
  ***REMOVED***
***REMOVED***;
