// js-yaml/lib/js-yaml/common.js
'use strict';


function isNothing(subject) ***REMOVED***
  return (typeof subject === 'undefined') || (subject === null);
***REMOVED***


function isObject(subject) ***REMOVED***
  return (typeof subject === 'object') && (subject !== null);
***REMOVED***


function toArray(sequence) ***REMOVED***
  if (Array.isArray(sequence)) return sequence;
  else if (isNothing(sequence)) return [];

  return [ sequence ];
***REMOVED***


function extend(target, source) ***REMOVED***
  var index, length, key, sourceKeys;

  if (source) ***REMOVED***
    sourceKeys = Object.keys(source);

    for (index = 0, length = sourceKeys.length; index < length; index += 1) ***REMOVED***
      key = sourceKeys[index];
      target[key] = source[key];
    ***REMOVED***
  ***REMOVED***

  return target;
***REMOVED***


function repeat(string, count) ***REMOVED***
  var result = '', cycle;

  for (cycle = 0; cycle < count; cycle += 1) ***REMOVED***
    result += string;
  ***REMOVED***

  return result;
***REMOVED***


function isNegativeZero(number) ***REMOVED***
  return (number === 0) && (Number.NEGATIVE_INFINITY === 1 / number);
***REMOVED***


module.exports.isNothing      = isNothing;
module.exports.isObject       = isObject;
module.exports.toArray        = toArray;
module.exports.repeat         = repeat;
module.exports.isNegativeZero = isNegativeZero;
module.exports.extend         = extend;
