// js-yaml/lib/js-yaml/dumper.js
'use strict';

/*eslint-disable no-use-before-define*/

var common              = require('./common');
var YAMLException       = require('./exception');
var DEFAULT_FULL_SCHEMA = require('./schema/default_full');
var DEFAULT_SAFE_SCHEMA = require('./schema/default_safe');

var _toString       = Object.prototype.toString;
var _hasOwnProperty = Object.prototype.hasOwnProperty;

var CHAR_TAB                  = 0x09; /* Tab */
var CHAR_LINE_FEED            = 0x0A; /* LF */
var CHAR_SPACE                = 0x20; /* Space */
var CHAR_EXCLAMATION          = 0x21; /* ! */
var CHAR_DOUBLE_QUOTE         = 0x22; /* " */
var CHAR_SHARP                = 0x23; /* # */
var CHAR_PERCENT              = 0x25; /* % */
var CHAR_AMPERSAND            = 0x26; /* & */
var CHAR_SINGLE_QUOTE         = 0x27; /* ' */
var CHAR_ASTERISK             = 0x2A; /* * */
var CHAR_COMMA                = 0x2C; /* , */
var CHAR_MINUS                = 0x2D; /* - */
var CHAR_COLON                = 0x3A; /* : */
var CHAR_GREATER_THAN         = 0x3E; /* > */
var CHAR_QUESTION             = 0x3F; /* ? */
var CHAR_COMMERCIAL_AT        = 0x40; /* @ */
var CHAR_LEFT_SQUARE_BRACKET  = 0x5B; /* [ */
var CHAR_RIGHT_SQUARE_BRACKET = 0x5D; /* ] */
var CHAR_GRAVE_ACCENT         = 0x60; /* ` */
var CHAR_LEFT_CURLY_BRACKET   = 0x7B; /* ***REMOVED*** */
var CHAR_VERTICAL_LINE        = 0x7C; /* | */
var CHAR_RIGHT_CURLY_BRACKET  = 0x7D; /* ***REMOVED*** */

var ESCAPE_SEQUENCES = ***REMOVED******REMOVED***;

ESCAPE_SEQUENCES[0x00]   = '\\0';
ESCAPE_SEQUENCES[0x07]   = '\\a';
ESCAPE_SEQUENCES[0x08]   = '\\b';
ESCAPE_SEQUENCES[0x09]   = '\\t';
ESCAPE_SEQUENCES[0x0A]   = '\\n';
ESCAPE_SEQUENCES[0x0B]   = '\\v';
ESCAPE_SEQUENCES[0x0C]   = '\\f';
ESCAPE_SEQUENCES[0x0D]   = '\\r';
ESCAPE_SEQUENCES[0x1B]   = '\\e';
ESCAPE_SEQUENCES[0x22]   = '\\"';
ESCAPE_SEQUENCES[0x5C]   = '\\\\';
ESCAPE_SEQUENCES[0x85]   = '\\N';
ESCAPE_SEQUENCES[0xA0]   = '\\_';
ESCAPE_SEQUENCES[0x2028] = '\\L';
ESCAPE_SEQUENCES[0x2029] = '\\P';

var DEPRECATED_BOOLEANS_SYNTAX = [
  'y', 'Y', 'yes', 'Yes', 'YES', 'on', 'On', 'ON',
  'n', 'N', 'no', 'No', 'NO', 'off', 'Off', 'OFF'
];

function compileStyleMap(schema, map) ***REMOVED***
  var result, keys, index, length, tag, style, type;

  if (map === null) return ***REMOVED******REMOVED***;

  result = ***REMOVED******REMOVED***;
  keys = Object.keys(map);

  for (index = 0, length = keys.length; index < length; index += 1) ***REMOVED***
    tag = keys[index];
    style = String(map[tag]);

    if (tag.slice(0, 2) === '!!') ***REMOVED***
      tag = 'tag:yaml.org,2002:' + tag.slice(2);
    ***REMOVED***
    type = schema.compiledTypeMap['fallback'][tag];

    if (type && _hasOwnProperty.call(type.styleAliases, style)) ***REMOVED***
      style = type.styleAliases[style];
    ***REMOVED***

    result[tag] = style;
  ***REMOVED***

  return result;
***REMOVED***

function encodeHex(character) ***REMOVED***
  var string, handle, length;

  string = character.toString(16).toUpperCase();

  if (character <= 0xFF) ***REMOVED***
    handle = 'x';
    length = 2;
  ***REMOVED*** else if (character <= 0xFFFF) ***REMOVED***
    handle = 'u';
    length = 4;
  ***REMOVED*** else if (character <= 0xFFFFFFFF) ***REMOVED***
    handle = 'U';
    length = 8;
  ***REMOVED*** else ***REMOVED***
    throw new YAMLException('code point within a string may not be greater than 0xFFFFFFFF');
  ***REMOVED***

  return '\\' + handle + common.repeat('0', length - string.length) + string;
***REMOVED***

function State(options) ***REMOVED***
  this.schema        = options['schema'] || DEFAULT_FULL_SCHEMA;
  this.indent        = Math.max(1, (options['indent'] || 2));
  this.noArrayIndent = options['noArrayIndent'] || false;
  this.skipInvalid   = options['skipInvalid'] || false;
  this.flowLevel     = (common.isNothing(options['flowLevel']) ? -1 : options['flowLevel']);
  this.styleMap      = compileStyleMap(this.schema, options['styles'] || null);
  this.sortKeys      = options['sortKeys'] || false;
  this.lineWidth     = options['lineWidth'] || 80;
  this.noRefs        = options['noRefs'] || false;
  this.noCompatMode  = options['noCompatMode'] || false;
  this.condenseFlow  = options['condenseFlow'] || false;

  this.implicitTypes = this.schema.compiledImplicit;
  this.explicitTypes = this.schema.compiledExplicit;

  this.tag = null;
  this.result = '';

  this.duplicates = [];
  this.usedDuplicates = null;
***REMOVED***

// Indents every line in a string. Empty lines (\n only) are not indented.
function indentString(string, spaces) ***REMOVED***
  var ind = common.repeat(' ', spaces),
      position = 0,
      next = -1,
      result = '',
      line,
      length = string.length;

  while (position < length) ***REMOVED***
    next = string.indexOf('\n', position);
    if (next === -1) ***REMOVED***
      line = string.slice(position);
      position = length;
    ***REMOVED*** else ***REMOVED***
      line = string.slice(position, next + 1);
      position = next + 1;
    ***REMOVED***

    if (line.length && line !== '\n') result += ind;

    result += line;
  ***REMOVED***

  return result;
***REMOVED***

function generateNextLine(state, level) ***REMOVED***
  return '\n' + common.repeat(' ', state.indent * level);
***REMOVED***

function testImplicitResolving(state, str) ***REMOVED***
  var index, length, type;

  for (index = 0, length = state.implicitTypes.length; index < length; index += 1) ***REMOVED***
    type = state.implicitTypes[index];

    if (type.resolve(str)) ***REMOVED***
      return true;
    ***REMOVED***
  ***REMOVED***

  return false;
***REMOVED***

// [33] s-white ::= s-space | s-tab
function isWhitespace(c) ***REMOVED***
  return c === CHAR_SPACE || c === CHAR_TAB;
***REMOVED***

// Returns true if the character can be printed without escaping.
// From YAML 1.2: "any allowed characters known to be non-printable
// should also be escaped. [However,] This isn’t mandatory"
// Derived from nb-char - \t - #x85 - #xA0 - #x2028 - #x2029.
function isPrintable(c) ***REMOVED***
  return  (0x00020 <= c && c <= 0x00007E)
      || ((0x000A1 <= c && c <= 0x00D7FF) && c !== 0x2028 && c !== 0x2029)
      || ((0x0E000 <= c && c <= 0x00FFFD) && c !== 0xFEFF /* BOM */)
      ||  (0x10000 <= c && c <= 0x10FFFF);
***REMOVED***

// Simplified test for values allowed after the first character in plain style.
function isPlainSafe(c) ***REMOVED***
  // Uses a subset of nb-char - c-flow-indicator - ":" - "#"
  // where nb-char ::= c-printable - b-char - c-byte-order-mark.
  return isPrintable(c) && c !== 0xFEFF
    // - c-flow-indicator
    && c !== CHAR_COMMA
    && c !== CHAR_LEFT_SQUARE_BRACKET
    && c !== CHAR_RIGHT_SQUARE_BRACKET
    && c !== CHAR_LEFT_CURLY_BRACKET
    && c !== CHAR_RIGHT_CURLY_BRACKET
    // - ":" - "#"
    && c !== CHAR_COLON
    && c !== CHAR_SHARP;
***REMOVED***

// Simplified test for values allowed as the first character in plain style.
function isPlainSafeFirst(c) ***REMOVED***
  // Uses a subset of ns-char - c-indicator
  // where ns-char = nb-char - s-white.
  return isPrintable(c) && c !== 0xFEFF
    && !isWhitespace(c) // - s-white
    // - (c-indicator ::=
    // “-” | “?” | “:” | “,” | “[” | “]” | “***REMOVED***” | “***REMOVED***”
    && c !== CHAR_MINUS
    && c !== CHAR_QUESTION
    && c !== CHAR_COLON
    && c !== CHAR_COMMA
    && c !== CHAR_LEFT_SQUARE_BRACKET
    && c !== CHAR_RIGHT_SQUARE_BRACKET
    && c !== CHAR_LEFT_CURLY_BRACKET
    && c !== CHAR_RIGHT_CURLY_BRACKET
    // | “#” | “&” | “*” | “!” | “|” | “>” | “'” | “"”
    && c !== CHAR_SHARP
    && c !== CHAR_AMPERSAND
    && c !== CHAR_ASTERISK
    && c !== CHAR_EXCLAMATION
    && c !== CHAR_VERTICAL_LINE
    && c !== CHAR_GREATER_THAN
    && c !== CHAR_SINGLE_QUOTE
    && c !== CHAR_DOUBLE_QUOTE
    // | “%” | “@” | “`”)
    && c !== CHAR_PERCENT
    && c !== CHAR_COMMERCIAL_AT
    && c !== CHAR_GRAVE_ACCENT;
***REMOVED***

// Determines whether block indentation indicator is required.
function needIndentIndicator(string) ***REMOVED***
  var leadingSpaceRe = /^\n* /;
  return leadingSpaceRe.test(string);
***REMOVED***

var STYLE_PLAIN   = 1,
    STYLE_SINGLE  = 2,
    STYLE_LITERAL = 3,
    STYLE_FOLDED  = 4,
    STYLE_DOUBLE  = 5;

// Determines which scalar styles are possible and returns the preferred style.
// lineWidth = -1 => no limit.
// Pre-conditions: str.length > 0.
// Post-conditions:
//    STYLE_PLAIN or STYLE_SINGLE => no \n are in the string.
//    STYLE_LITERAL => no lines are suitable for folding (or lineWidth is -1).
//    STYLE_FOLDED => a line > lineWidth and can be folded (and lineWidth != -1).
function chooseScalarStyle(string, singleLineOnly, indentPerLevel, lineWidth, testAmbiguousType) ***REMOVED***
  var i;
  var char;
  var hasLineBreak = false;
  var hasFoldableLine = false; // only checked if shouldTrackWidth
  var shouldTrackWidth = lineWidth !== -1;
  var previousLineBreak = -1; // count the first line correctly
  var plain = isPlainSafeFirst(string.charCodeAt(0))
          && !isWhitespace(string.charCodeAt(string.length - 1));

  if (singleLineOnly) ***REMOVED***
    // Case: no block styles.
    // Check for disallowed characters to rule out plain and single.
    for (i = 0; i < string.length; i++) ***REMOVED***
      char = string.charCodeAt(i);
      if (!isPrintable(char)) ***REMOVED***
        return STYLE_DOUBLE;
      ***REMOVED***
      plain = plain && isPlainSafe(char);
    ***REMOVED***
  ***REMOVED*** else ***REMOVED***
    // Case: block styles permitted.
    for (i = 0; i < string.length; i++) ***REMOVED***
      char = string.charCodeAt(i);
      if (char === CHAR_LINE_FEED) ***REMOVED***
        hasLineBreak = true;
        // Check if any line can be folded.
        if (shouldTrackWidth) ***REMOVED***
          hasFoldableLine = hasFoldableLine ||
            // Foldable line = too long, and not more-indented.
            (i - previousLineBreak - 1 > lineWidth &&
             string[previousLineBreak + 1] !== ' ');
          previousLineBreak = i;
        ***REMOVED***
      ***REMOVED*** else if (!isPrintable(char)) ***REMOVED***
        return STYLE_DOUBLE;
      ***REMOVED***
      plain = plain && isPlainSafe(char);
    ***REMOVED***
    // in case the end is missing a \n
    hasFoldableLine = hasFoldableLine || (shouldTrackWidth &&
      (i - previousLineBreak - 1 > lineWidth &&
       string[previousLineBreak + 1] !== ' '));
  ***REMOVED***
  // Although every style can represent \n without escaping, prefer block styles
  // for multiline, since they're more readable and they don't add empty lines.
  // Also prefer folding a super-long line.
  if (!hasLineBreak && !hasFoldableLine) ***REMOVED***
    // Strings interpretable as another type have to be quoted;
    // e.g. the string 'true' vs. the boolean true.
    return plain && !testAmbiguousType(string)
      ? STYLE_PLAIN : STYLE_SINGLE;
  ***REMOVED***
  // Edge case: block indentation indicator can only have one digit.
  if (indentPerLevel > 9 && needIndentIndicator(string)) ***REMOVED***
    return STYLE_DOUBLE;
  ***REMOVED***
  // At this point we know block styles are valid.
  // Prefer literal style unless we want to fold.
  return hasFoldableLine ? STYLE_FOLDED : STYLE_LITERAL;
***REMOVED***

// Note: line breaking/folding is implemented for only the folded style.
// NB. We drop the last trailing newline (if any) of a returned block scalar
//  since the dumper adds its own newline. This always works:
//    • No ending newline => unaffected; already using strip "-" chomping.
//    • Ending newline    => removed then restored.
//  Importantly, this keeps the "+" chomp indicator from gaining an extra line.
function writeScalar(state, string, level, iskey) ***REMOVED***
  state.dump = (function () ***REMOVED***
    if (string.length === 0) ***REMOVED***
      return "''";
    ***REMOVED***
    if (!state.noCompatMode &&
        DEPRECATED_BOOLEANS_SYNTAX.indexOf(string) !== -1) ***REMOVED***
      return "'" + string + "'";
    ***REMOVED***

    var indent = state.indent * Math.max(1, level); // no 0-indent scalars
    // As indentation gets deeper, let the width decrease monotonically
    // to the lower bound min(state.lineWidth, 40).
    // Note that this implies
    //  state.lineWidth ≤ 40 + state.indent: width is fixed at the lower bound.
    //  state.lineWidth > 40 + state.indent: width decreases until the lower bound.
    // This behaves better than a constant minimum width which disallows narrower options,
    // or an indent threshold which causes the width to suddenly increase.
    var lineWidth = state.lineWidth === -1
      ? -1 : Math.max(Math.min(state.lineWidth, 40), state.lineWidth - indent);

    // Without knowing if keys are implicit/explicit, assume implicit for safety.
    var singleLineOnly = iskey
      // No block styles in flow mode.
      || (state.flowLevel > -1 && level >= state.flowLevel);
    function testAmbiguity(string) ***REMOVED***
      return testImplicitResolving(state, string);
    ***REMOVED***

    switch (chooseScalarStyle(string, singleLineOnly, state.indent, lineWidth, testAmbiguity)) ***REMOVED***
      case STYLE_PLAIN:
        return string;
      case STYLE_SINGLE:
        return "'" + string.replace(/'/g, "''") + "'";
      case STYLE_LITERAL:
        return '|' + blockHeader(string, state.indent)
          + dropEndingNewline(indentString(string, indent));
      case STYLE_FOLDED:
        return '>' + blockHeader(string, state.indent)
          + dropEndingNewline(indentString(foldString(string, lineWidth), indent));
      case STYLE_DOUBLE:
        return '"' + escapeString(string, lineWidth) + '"';
      default:
        throw new YAMLException('impossible error: invalid scalar style');
    ***REMOVED***
  ***REMOVED***());
***REMOVED***

// Pre-conditions: string is valid for a block scalar, 1 <= indentPerLevel <= 9.
function blockHeader(string, indentPerLevel) ***REMOVED***
  var indentIndicator = needIndentIndicator(string) ? String(indentPerLevel) : '';

  // note the special case: the string '\n' counts as a "trailing" empty line.
  var clip =          string[string.length - 1] === '\n';
  var keep = clip && (string[string.length - 2] === '\n' || string === '\n');
  var chomp = keep ? '+' : (clip ? '' : '-');

  return indentIndicator + chomp + '\n';
***REMOVED***

// (See the note for writeScalar.)
function dropEndingNewline(string) ***REMOVED***
  return string[string.length - 1] === '\n' ? string.slice(0, -1) : string;
***REMOVED***

// Note: a long line without a suitable break point will exceed the width limit.
// Pre-conditions: every char in str isPrintable, str.length > 0, width > 0.
function foldString(string, width) ***REMOVED***
  // In folded style, $k$ consecutive newlines output as $k+1$ newlines—
  // unless they're before or after a more-indented line, or at the very
  // beginning or end, in which case $k$ maps to $k$.
  // Therefore, parse each chunk as newline(s) followed by a content line.
  var lineRe = /(\n+)([^\n]*)/g;

  // first line (possibly an empty line)
  var result = (function () ***REMOVED***
    var nextLF = string.indexOf('\n');
    nextLF = nextLF !== -1 ? nextLF : string.length;
    lineRe.lastIndex = nextLF;
    return foldLine(string.slice(0, nextLF), width);
  ***REMOVED***());
  // If we haven't reached the first content line yet, don't add an extra \n.
  var prevMoreIndented = string[0] === '\n' || string[0] === ' ';
  var moreIndented;

  // rest of the lines
  var match;
  while ((match = lineRe.exec(string))) ***REMOVED***
    var prefix = match[1], line = match[2];
    moreIndented = (line[0] === ' ');
    result += prefix
      + (!prevMoreIndented && !moreIndented && line !== ''
        ? '\n' : '')
      + foldLine(line, width);
    prevMoreIndented = moreIndented;
  ***REMOVED***

  return result;
***REMOVED***

// Greedy line breaking.
// Picks the longest line under the limit each time,
// otherwise settles for the shortest line over the limit.
// NB. More-indented lines *cannot* be folded, as that would add an extra \n.
function foldLine(line, width) ***REMOVED***
  if (line === '' || line[0] === ' ') return line;

  // Since a more-indented line adds a \n, breaks can't be followed by a space.
  var breakRe = / [^ ]/g; // note: the match index will always be <= length-2.
  var match;
  // start is an inclusive index. end, curr, and next are exclusive.
  var start = 0, end, curr = 0, next = 0;
  var result = '';

  // Invariants: 0 <= start <= length-1.
  //   0 <= curr <= next <= max(0, length-2). curr - start <= width.
  // Inside the loop:
  //   A match implies length >= 2, so curr and next are <= length-2.
  while ((match = breakRe.exec(line))) ***REMOVED***
    next = match.index;
    // maintain invariant: curr - start <= width
    if (next - start > width) ***REMOVED***
      end = (curr > start) ? curr : next; // derive end <= length-2
      result += '\n' + line.slice(start, end);
      // skip the space that was output as \n
      start = end + 1;                    // derive start <= length-1
    ***REMOVED***
    curr = next;
  ***REMOVED***

  // By the invariants, start <= length-1, so there is something left over.
  // It is either the whole string or a part starting from non-whitespace.
  result += '\n';
  // Insert a break if the remainder is too long and there is a break available.
  if (line.length - start > width && curr > start) ***REMOVED***
    result += line.slice(start, curr) + '\n' + line.slice(curr + 1);
  ***REMOVED*** else ***REMOVED***
    result += line.slice(start);
  ***REMOVED***

  return result.slice(1); // drop extra \n joiner
***REMOVED***

// Escapes a double-quoted string.
function escapeString(string) ***REMOVED***
  var result = '';
  var char, nextChar;
  var escapeSeq;

  for (var i = 0; i < string.length; i++) ***REMOVED***
    char = string.charCodeAt(i);
    // Check for surrogate pairs (reference Unicode 3.0 section "3.7 Surrogates").
    if (char >= 0xD800 && char <= 0xDBFF/* high surrogate */) ***REMOVED***
      nextChar = string.charCodeAt(i + 1);
      if (nextChar >= 0xDC00 && nextChar <= 0xDFFF/* low surrogate */) ***REMOVED***
        // Combine the surrogate pair and store it escaped.
        result += encodeHex((char - 0xD800) * 0x400 + nextChar - 0xDC00 + 0x10000);
        // Advance index one extra since we already used that char here.
        i++; continue;
      ***REMOVED***
    ***REMOVED***
    escapeSeq = ESCAPE_SEQUENCES[char];
    result += !escapeSeq && isPrintable(char)
      ? string[i]
      : escapeSeq || encodeHex(char);
  ***REMOVED***

  return result;
***REMOVED***

function writeFlowSequence(state, level, object) ***REMOVED***
  var _result = '',
      _tag    = state.tag,
      index,
      length;

  for (index = 0, length = object.length; index < length; index += 1) ***REMOVED***
    // Write only valid elements.
    if (writeNode(state, level, object[index], false, false)) ***REMOVED***
      if (index !== 0) _result += ',' + (!state.condenseFlow ? ' ' : '');
      _result += state.dump;
    ***REMOVED***
  ***REMOVED***

  state.tag = _tag;
  state.dump = '[' + _result + ']';
***REMOVED***

function writeBlockSequence(state, level, object, compact) ***REMOVED***
  var _result = '',
      _tag    = state.tag,
      index,
      length;

  for (index = 0, length = object.length; index < length; index += 1) ***REMOVED***
    // Write only valid elements.
    if (writeNode(state, level + 1, object[index], true, true)) ***REMOVED***
      if (!compact || index !== 0) ***REMOVED***
        _result += generateNextLine(state, level);
      ***REMOVED***

      if (state.dump && CHAR_LINE_FEED === state.dump.charCodeAt(0)) ***REMOVED***
        _result += '-';
      ***REMOVED*** else ***REMOVED***
        _result += '- ';
      ***REMOVED***

      _result += state.dump;
    ***REMOVED***
  ***REMOVED***

  state.tag = _tag;
  state.dump = _result || '[]'; // Empty sequence if no valid values.
***REMOVED***

function writeFlowMapping(state, level, object) ***REMOVED***
  var _result       = '',
      _tag          = state.tag,
      objectKeyList = Object.keys(object),
      index,
      length,
      objectKey,
      objectValue,
      pairBuffer;

  for (index = 0, length = objectKeyList.length; index < length; index += 1) ***REMOVED***
    pairBuffer = state.condenseFlow ? '"' : '';

    if (index !== 0) pairBuffer += ', ';

    objectKey = objectKeyList[index];
    objectValue = object[objectKey];

    if (!writeNode(state, level, objectKey, false, false)) ***REMOVED***
      continue; // Skip this pair because of invalid key;
    ***REMOVED***

    if (state.dump.length > 1024) pairBuffer += '? ';

    pairBuffer += state.dump + (state.condenseFlow ? '"' : '') + ':' + (state.condenseFlow ? '' : ' ');

    if (!writeNode(state, level, objectValue, false, false)) ***REMOVED***
      continue; // Skip this pair because of invalid value.
    ***REMOVED***

    pairBuffer += state.dump;

    // Both key and value are valid.
    _result += pairBuffer;
  ***REMOVED***

  state.tag = _tag;
  state.dump = '***REMOVED***' + _result + '***REMOVED***';
***REMOVED***

function writeBlockMapping(state, level, object, compact) ***REMOVED***
  var _result       = '',
      _tag          = state.tag,
      objectKeyList = Object.keys(object),
      index,
      length,
      objectKey,
      objectValue,
      explicitPair,
      pairBuffer;

  // Allow sorting keys so that the output file is deterministic
  if (state.sortKeys === true) ***REMOVED***
    // Default sorting
    objectKeyList.sort();
  ***REMOVED*** else if (typeof state.sortKeys === 'function') ***REMOVED***
    // Custom sort function
    objectKeyList.sort(state.sortKeys);
  ***REMOVED*** else if (state.sortKeys) ***REMOVED***
    // Something is wrong
    throw new YAMLException('sortKeys must be a boolean or a function');
  ***REMOVED***

  for (index = 0, length = objectKeyList.length; index < length; index += 1) ***REMOVED***
    pairBuffer = '';

    if (!compact || index !== 0) ***REMOVED***
      pairBuffer += generateNextLine(state, level);
    ***REMOVED***

    objectKey = objectKeyList[index];
    objectValue = object[objectKey];

    if (!writeNode(state, level + 1, objectKey, true, true, true)) ***REMOVED***
      continue; // Skip this pair because of invalid key.
    ***REMOVED***

    explicitPair = (state.tag !== null && state.tag !== '?') ||
                   (state.dump && state.dump.length > 1024);

    if (explicitPair) ***REMOVED***
      if (state.dump && CHAR_LINE_FEED === state.dump.charCodeAt(0)) ***REMOVED***
        pairBuffer += '?';
      ***REMOVED*** else ***REMOVED***
        pairBuffer += '? ';
      ***REMOVED***
    ***REMOVED***

    pairBuffer += state.dump;

    if (explicitPair) ***REMOVED***
      pairBuffer += generateNextLine(state, level);
    ***REMOVED***

    if (!writeNode(state, level + 1, objectValue, true, explicitPair)) ***REMOVED***
      continue; // Skip this pair because of invalid value.
    ***REMOVED***

    if (state.dump && CHAR_LINE_FEED === state.dump.charCodeAt(0)) ***REMOVED***
      pairBuffer += ':';
    ***REMOVED*** else ***REMOVED***
      pairBuffer += ': ';
    ***REMOVED***

    pairBuffer += state.dump;

    // Both key and value are valid.
    _result += pairBuffer;
  ***REMOVED***

  state.tag = _tag;
  state.dump = _result || '***REMOVED******REMOVED***'; // Empty mapping if no valid pairs.
***REMOVED***

function detectType(state, object, explicit) ***REMOVED***
  var _result, typeList, index, length, type, style;

  typeList = explicit ? state.explicitTypes : state.implicitTypes;

  for (index = 0, length = typeList.length; index < length; index += 1) ***REMOVED***
    type = typeList[index];

    if ((type.instanceOf  || type.predicate) &&
        (!type.instanceOf || ((typeof object === 'object') && (object instanceof type.instanceOf))) &&
        (!type.predicate  || type.predicate(object))) ***REMOVED***

      state.tag = explicit ? type.tag : '?';

      if (type.represent) ***REMOVED***
        style = state.styleMap[type.tag] || type.defaultStyle;

        if (_toString.call(type.represent) === '[object Function]') ***REMOVED***
          _result = type.represent(object, style);
        ***REMOVED*** else if (_hasOwnProperty.call(type.represent, style)) ***REMOVED***
          _result = type.represent[style](object, style);
        ***REMOVED*** else ***REMOVED***
          throw new YAMLException('!<' + type.tag + '> tag resolver accepts not "' + style + '" style');
        ***REMOVED***

        state.dump = _result;
      ***REMOVED***

      return true;
    ***REMOVED***
  ***REMOVED***

  return false;
***REMOVED***

// Serializes `object` and writes it to global `result`.
// Returns true on success, or false on invalid object.
//
function writeNode(state, level, object, block, compact, iskey) ***REMOVED***
  state.tag = null;
  state.dump = object;

  if (!detectType(state, object, false)) ***REMOVED***
    detectType(state, object, true);
  ***REMOVED***

  var type = _toString.call(state.dump);

  if (block) ***REMOVED***
    block = (state.flowLevel < 0 || state.flowLevel > level);
  ***REMOVED***

  var objectOrArray = type === '[object Object]' || type === '[object Array]',
      duplicateIndex,
      duplicate;

  if (objectOrArray) ***REMOVED***
    duplicateIndex = state.duplicates.indexOf(object);
    duplicate = duplicateIndex !== -1;
  ***REMOVED***

  if ((state.tag !== null && state.tag !== '?') || duplicate || (state.indent !== 2 && level > 0)) ***REMOVED***
    compact = false;
  ***REMOVED***

  if (duplicate && state.usedDuplicates[duplicateIndex]) ***REMOVED***
    state.dump = '*ref_' + duplicateIndex;
  ***REMOVED*** else ***REMOVED***
    if (objectOrArray && duplicate && !state.usedDuplicates[duplicateIndex]) ***REMOVED***
      state.usedDuplicates[duplicateIndex] = true;
    ***REMOVED***
    if (type === '[object Object]') ***REMOVED***
      if (block && (Object.keys(state.dump).length !== 0)) ***REMOVED***
        writeBlockMapping(state, level, state.dump, compact);
        if (duplicate) ***REMOVED***
          state.dump = '&ref_' + duplicateIndex + state.dump;
        ***REMOVED***
      ***REMOVED*** else ***REMOVED***
        writeFlowMapping(state, level, state.dump);
        if (duplicate) ***REMOVED***
          state.dump = '&ref_' + duplicateIndex + ' ' + state.dump;
        ***REMOVED***
      ***REMOVED***
    ***REMOVED*** else if (type === '[object Array]') ***REMOVED***
      var arrayLevel = (state.noArrayIndent && (level > 0)) ? level - 1 : level;
      if (block && (state.dump.length !== 0)) ***REMOVED***
        writeBlockSequence(state, arrayLevel, state.dump, compact);
        if (duplicate) ***REMOVED***
          state.dump = '&ref_' + duplicateIndex + state.dump;
        ***REMOVED***
      ***REMOVED*** else ***REMOVED***
        writeFlowSequence(state, arrayLevel, state.dump);
        if (duplicate) ***REMOVED***
          state.dump = '&ref_' + duplicateIndex + ' ' + state.dump;
        ***REMOVED***
      ***REMOVED***
    ***REMOVED*** else if (type === '[object String]') ***REMOVED***
      if (state.tag !== '?') ***REMOVED***
        writeScalar(state, state.dump, level, iskey);
      ***REMOVED***
    ***REMOVED*** else ***REMOVED***
      if (state.skipInvalid) return false;
      throw new YAMLException('unacceptable kind of an object to dump ' + type);
    ***REMOVED***

    if (state.tag !== null && state.tag !== '?') ***REMOVED***
      state.dump = '!<' + state.tag + '> ' + state.dump;
    ***REMOVED***
  ***REMOVED***

  return true;
***REMOVED***

function getDuplicateReferences(object, state) ***REMOVED***
  var objects = [],
      duplicatesIndexes = [],
      index,
      length;

  inspectNode(object, objects, duplicatesIndexes);

  for (index = 0, length = duplicatesIndexes.length; index < length; index += 1) ***REMOVED***
    state.duplicates.push(objects[duplicatesIndexes[index]]);
  ***REMOVED***
  state.usedDuplicates = new Array(length);
***REMOVED***

function inspectNode(object, objects, duplicatesIndexes) ***REMOVED***
  var objectKeyList,
      index,
      length;

  if (object !== null && typeof object === 'object') ***REMOVED***
    index = objects.indexOf(object);
    if (index !== -1) ***REMOVED***
      if (duplicatesIndexes.indexOf(index) === -1) ***REMOVED***
        duplicatesIndexes.push(index);
      ***REMOVED***
    ***REMOVED*** else ***REMOVED***
      objects.push(object);

      if (Array.isArray(object)) ***REMOVED***
        for (index = 0, length = object.length; index < length; index += 1) ***REMOVED***
          inspectNode(object[index], objects, duplicatesIndexes);
        ***REMOVED***
      ***REMOVED*** else ***REMOVED***
        objectKeyList = Object.keys(object);

        for (index = 0, length = objectKeyList.length; index < length; index += 1) ***REMOVED***
          inspectNode(object[objectKeyList[index]], objects, duplicatesIndexes);
        ***REMOVED***
      ***REMOVED***
    ***REMOVED***
  ***REMOVED***
***REMOVED***

function dump(input, options) ***REMOVED***
  options = options || ***REMOVED******REMOVED***;

  var state = new State(options);

  if (!state.noRefs) getDuplicateReferences(input, state);

  if (writeNode(state, 0, input, true, true)) return state.dump + '\n';

  return '';
***REMOVED***

function safeDump(input, options) ***REMOVED***
  return dump(input, common.extend(***REMOVED*** schema: DEFAULT_SAFE_SCHEMA ***REMOVED***, options));
***REMOVED***

module.exports.dump     = dump;
module.exports.safeDump = safeDump;
