// js-yaml/lib/js-yaml/exception.js
// YAML error class. http://stackoverflow.com/questions/8458984
//
'use strict';

function YAMLException(reason, mark) ***REMOVED***
  // Super constructor
  Error.call(this);

  this.name = 'YAMLException';
  this.reason = reason;
  this.mark = mark;
  this.message = (this.reason || '(unknown reason)') + (this.mark ? ' ' + this.mark.toString() : '');

  // Include stack trace in error object
  if (Error.captureStackTrace) ***REMOVED***
    // Chrome and NodeJS
    Error.captureStackTrace(this, this.constructor);
  ***REMOVED*** else ***REMOVED***
    // FF, IE 10+ and Safari 6+. Fallback for others
    this.stack = (new Error()).stack || '';
  ***REMOVED***
***REMOVED***


// Inherit from Error
YAMLException.prototype = Object.create(Error.prototype);
YAMLException.prototype.constructor = YAMLException;


YAMLException.prototype.toString = function toString(compact) ***REMOVED***
  var result = this.name + ': ';

  result += this.reason || '(unknown reason)';

  if (!compact && this.mark) ***REMOVED***
    result += ' ' + this.mark.toString();
  ***REMOVED***

  return result;
***REMOVED***;


module.exports = YAMLException;
