// js-yaml/lib/js-yaml/loader.js
'use strict';

/*eslint-disable max-len,no-use-before-define*/

var common              = require('./common');
var YAMLException       = require('./exception');
var Mark                = require('./mark');
var DEFAULT_SAFE_SCHEMA = require('./schema/default_safe');
var DEFAULT_FULL_SCHEMA = require('./schema/default_full');


var _hasOwnProperty = Object.prototype.hasOwnProperty;


var CONTEXT_FLOW_IN   = 1;
var CONTEXT_FLOW_OUT  = 2;
var CONTEXT_BLOCK_IN  = 3;
var CONTEXT_BLOCK_OUT = 4;


var CHOMPING_CLIP  = 1;
var CHOMPING_STRIP = 2;
var CHOMPING_KEEP  = 3;


var PATTERN_NON_PRINTABLE         = /[\x00-\x08\x0B\x0C\x0E-\x1F\x7F-\x84\x86-\x9F\uFFFE\uFFFF]|[\uD800-\uDBFF](?![\uDC00-\uDFFF])|(?:[^\uD800-\uDBFF]|^)[\uDC00-\uDFFF]/;
var PATTERN_NON_ASCII_LINE_BREAKS = /[\x85\u2028\u2029]/;
var PATTERN_FLOW_INDICATORS       = /[,\[\]\***REMOVED***\***REMOVED***]/;
var PATTERN_TAG_HANDLE            = /^(?:!|!!|![a-z\-]+!)$/i;
var PATTERN_TAG_URI               = /^(?:!|[^,\[\]\***REMOVED***\***REMOVED***])(?:%[0-9a-f]***REMOVED***2***REMOVED***|[0-9a-z\-#;\/\?:@&=\+\$,_\.!~\*'\(\)\[\]])*$/i;


function _class(obj) ***REMOVED*** return Object.prototype.toString.call(obj); ***REMOVED***

function is_EOL(c) ***REMOVED***
  return (c === 0x0A/* LF */) || (c === 0x0D/* CR */);
***REMOVED***

function is_WHITE_SPACE(c) ***REMOVED***
  return (c === 0x09/* Tab */) || (c === 0x20/* Space */);
***REMOVED***

function is_WS_OR_EOL(c) ***REMOVED***
  return (c === 0x09/* Tab */) ||
         (c === 0x20/* Space */) ||
         (c === 0x0A/* LF */) ||
         (c === 0x0D/* CR */);
***REMOVED***

function is_FLOW_INDICATOR(c) ***REMOVED***
  return c === 0x2C/* , */ ||
         c === 0x5B/* [ */ ||
         c === 0x5D/* ] */ ||
         c === 0x7B/* ***REMOVED*** */ ||
         c === 0x7D/* ***REMOVED*** */;
***REMOVED***

function fromHexCode(c) ***REMOVED***
  var lc;

  if ((0x30/* 0 */ <= c) && (c <= 0x39/* 9 */)) ***REMOVED***
    return c - 0x30;
  ***REMOVED***

  /*eslint-disable no-bitwise*/
  lc = c | 0x20;

  if ((0x61/* a */ <= lc) && (lc <= 0x66/* f */)) ***REMOVED***
    return lc - 0x61 + 10;
  ***REMOVED***

  return -1;
***REMOVED***

function escapedHexLen(c) ***REMOVED***
  if (c === 0x78/* x */) ***REMOVED*** return 2; ***REMOVED***
  if (c === 0x75/* u */) ***REMOVED*** return 4; ***REMOVED***
  if (c === 0x55/* U */) ***REMOVED*** return 8; ***REMOVED***
  return 0;
***REMOVED***

function fromDecimalCode(c) ***REMOVED***
  if ((0x30/* 0 */ <= c) && (c <= 0x39/* 9 */)) ***REMOVED***
    return c - 0x30;
  ***REMOVED***

  return -1;
***REMOVED***

function simpleEscapeSequence(c) ***REMOVED***
  /* eslint-disable indent */
  return (c === 0x30/* 0 */) ? '\x00' :
        (c === 0x61/* a */) ? '\x07' :
        (c === 0x62/* b */) ? '\x08' :
        (c === 0x74/* t */) ? '\x09' :
        (c === 0x09/* Tab */) ? '\x09' :
        (c === 0x6E/* n */) ? '\x0A' :
        (c === 0x76/* v */) ? '\x0B' :
        (c === 0x66/* f */) ? '\x0C' :
        (c === 0x72/* r */) ? '\x0D' :
        (c === 0x65/* e */) ? '\x1B' :
        (c === 0x20/* Space */) ? ' ' :
        (c === 0x22/* " */) ? '\x22' :
        (c === 0x2F/* / */) ? '/' :
        (c === 0x5C/* \ */) ? '\x5C' :
        (c === 0x4E/* N */) ? '\x85' :
        (c === 0x5F/* _ */) ? '\xA0' :
        (c === 0x4C/* L */) ? '\u2028' :
        (c === 0x50/* P */) ? '\u2029' : '';
***REMOVED***

function charFromCodepoint(c) ***REMOVED***
  if (c <= 0xFFFF) ***REMOVED***
    return String.fromCharCode(c);
  ***REMOVED***
  // Encode UTF-16 surrogate pair
  // https://en.wikipedia.org/wiki/UTF-16#Code_points_U.2B010000_to_U.2B10FFFF
  return String.fromCharCode(
    ((c - 0x010000) >> 10) + 0xD800,
    ((c - 0x010000) & 0x03FF) + 0xDC00
  );
***REMOVED***

var simpleEscapeCheck = new Array(256); // integer, for fast access
var simpleEscapeMap = new Array(256);
for (var i = 0; i < 256; i++) ***REMOVED***
  simpleEscapeCheck[i] = simpleEscapeSequence(i) ? 1 : 0;
  simpleEscapeMap[i] = simpleEscapeSequence(i);
***REMOVED***


function State(input, options) ***REMOVED***
  this.input = input;

  this.filename  = options['filename']  || null;
  this.schema    = options['schema']    || DEFAULT_FULL_SCHEMA;
  this.onWarning = options['onWarning'] || null;
  this.legacy    = options['legacy']    || false;
  this.json      = options['json']      || false;
  this.listener  = options['listener']  || null;

  this.implicitTypes = this.schema.compiledImplicit;
  this.typeMap       = this.schema.compiledTypeMap;

  this.length     = input.length;
  this.position   = 0;
  this.line       = 0;
  this.lineStart  = 0;
  this.lineIndent = 0;

  this.documents = [];

  /*
  this.version;
  this.checkLineBreaks;
  this.tagMap;
  this.anchorMap;
  this.tag;
  this.anchor;
  this.kind;
  this.result;*/

***REMOVED***


function generateError(state, message) ***REMOVED***
  return new YAMLException(
    message,
    new Mark(state.filename, state.input, state.position, state.line, (state.position - state.lineStart)));
***REMOVED***

function throwError(state, message) ***REMOVED***
  throw generateError(state, message);
***REMOVED***

function throwWarning(state, message) ***REMOVED***
  if (state.onWarning) ***REMOVED***
    state.onWarning.call(null, generateError(state, message));
  ***REMOVED***
***REMOVED***


var directiveHandlers = ***REMOVED***

  YAML: function handleYamlDirective(state, name, args) ***REMOVED***

    var match, major, minor;

    if (state.version !== null) ***REMOVED***
      throwError(state, 'duplication of %YAML directive');
    ***REMOVED***

    if (args.length !== 1) ***REMOVED***
      throwError(state, 'YAML directive accepts exactly one argument');
    ***REMOVED***

    match = /^([0-9]+)\.([0-9]+)$/.exec(args[0]);

    if (match === null) ***REMOVED***
      throwError(state, 'ill-formed argument of the YAML directive');
    ***REMOVED***

    major = parseInt(match[1], 10);
    minor = parseInt(match[2], 10);

    if (major !== 1) ***REMOVED***
      throwError(state, 'unacceptable YAML version of the document');
    ***REMOVED***

    state.version = args[0];
    state.checkLineBreaks = (minor < 2);

    if (minor !== 1 && minor !== 2) ***REMOVED***
      throwWarning(state, 'unsupported YAML version of the document');
    ***REMOVED***
  ***REMOVED***,

  TAG: function handleTagDirective(state, name, args) ***REMOVED***

    var handle, prefix;

    if (args.length !== 2) ***REMOVED***
      throwError(state, 'TAG directive accepts exactly two arguments');
    ***REMOVED***

    handle = args[0];
    prefix = args[1];

    if (!PATTERN_TAG_HANDLE.test(handle)) ***REMOVED***
      throwError(state, 'ill-formed tag handle (first argument) of the TAG directive');
    ***REMOVED***

    if (_hasOwnProperty.call(state.tagMap, handle)) ***REMOVED***
      throwError(state, 'there is a previously declared suffix for "' + handle + '" tag handle');
    ***REMOVED***

    if (!PATTERN_TAG_URI.test(prefix)) ***REMOVED***
      throwError(state, 'ill-formed tag prefix (second argument) of the TAG directive');
    ***REMOVED***

    state.tagMap[handle] = prefix;
  ***REMOVED***
***REMOVED***;


function captureSegment(state, start, end, checkJson) ***REMOVED***
  var _position, _length, _character, _result;

  if (start < end) ***REMOVED***
    _result = state.input.slice(start, end);

    if (checkJson) ***REMOVED***
      for (_position = 0, _length = _result.length; _position < _length; _position += 1) ***REMOVED***
        _character = _result.charCodeAt(_position);
        if (!(_character === 0x09 ||
              (0x20 <= _character && _character <= 0x10FFFF))) ***REMOVED***
          throwError(state, 'expected valid JSON character');
        ***REMOVED***
      ***REMOVED***
    ***REMOVED*** else if (PATTERN_NON_PRINTABLE.test(_result)) ***REMOVED***
      throwError(state, 'the stream contains non-printable characters');
    ***REMOVED***

    state.result += _result;
  ***REMOVED***
***REMOVED***

function mergeMappings(state, destination, source, overridableKeys) ***REMOVED***
  var sourceKeys, key, index, quantity;

  if (!common.isObject(source)) ***REMOVED***
    throwError(state, 'cannot merge mappings; the provided source object is unacceptable');
  ***REMOVED***

  sourceKeys = Object.keys(source);

  for (index = 0, quantity = sourceKeys.length; index < quantity; index += 1) ***REMOVED***
    key = sourceKeys[index];

    if (!_hasOwnProperty.call(destination, key)) ***REMOVED***
      destination[key] = source[key];
      overridableKeys[key] = true;
    ***REMOVED***
  ***REMOVED***
***REMOVED***

function storeMappingPair(state, _result, overridableKeys, keyTag, keyNode, valueNode, startLine, startPos) ***REMOVED***
  var index, quantity;

  // The output is a plain object here, so keys can only be strings.
  // We need to convert keyNode to a string, but doing so can hang the process
  // (deeply nested arrays that explode exponentially using aliases).
  if (Array.isArray(keyNode)) ***REMOVED***
    keyNode = Array.prototype.slice.call(keyNode);

    for (index = 0, quantity = keyNode.length; index < quantity; index += 1) ***REMOVED***
      if (Array.isArray(keyNode[index])) ***REMOVED***
        throwError(state, 'nested arrays are not supported inside keys');
      ***REMOVED***

      if (typeof keyNode === 'object' && _class(keyNode[index]) === '[object Object]') ***REMOVED***
        keyNode[index] = '[object Object]';
      ***REMOVED***
    ***REMOVED***
  ***REMOVED***

  // Avoid code execution in load() via toString property
  // (still use its own toString for arrays, timestamps,
  // and whatever user schema extensions happen to have @@toStringTag)
  if (typeof keyNode === 'object' && _class(keyNode) === '[object Object]') ***REMOVED***
    keyNode = '[object Object]';
  ***REMOVED***


  keyNode = String(keyNode);

  if (_result === null) ***REMOVED***
    _result = ***REMOVED******REMOVED***;
  ***REMOVED***

  if (keyTag === 'tag:yaml.org,2002:merge') ***REMOVED***
    if (Array.isArray(valueNode)) ***REMOVED***
      for (index = 0, quantity = valueNode.length; index < quantity; index += 1) ***REMOVED***
        mergeMappings(state, _result, valueNode[index], overridableKeys);
      ***REMOVED***
    ***REMOVED*** else ***REMOVED***
      mergeMappings(state, _result, valueNode, overridableKeys);
    ***REMOVED***
  ***REMOVED*** else ***REMOVED***
    if (!state.json &&
        !_hasOwnProperty.call(overridableKeys, keyNode) &&
        _hasOwnProperty.call(_result, keyNode)) ***REMOVED***
      state.line = startLine || state.line;
      state.position = startPos || state.position;
      throwError(state, 'duplicated mapping key');
    ***REMOVED***
    _result[keyNode] = valueNode;
    delete overridableKeys[keyNode];
  ***REMOVED***

  return _result;
***REMOVED***

function readLineBreak(state) ***REMOVED***
  var ch;

  ch = state.input.charCodeAt(state.position);

  if (ch === 0x0A/* LF */) ***REMOVED***
    state.position++;
  ***REMOVED*** else if (ch === 0x0D/* CR */) ***REMOVED***
    state.position++;
    if (state.input.charCodeAt(state.position) === 0x0A/* LF */) ***REMOVED***
      state.position++;
    ***REMOVED***
  ***REMOVED*** else ***REMOVED***
    throwError(state, 'a line break is expected');
  ***REMOVED***

  state.line += 1;
  state.lineStart = state.position;
***REMOVED***

function skipSeparationSpace(state, allowComments, checkIndent) ***REMOVED***
  var lineBreaks = 0,
      ch = state.input.charCodeAt(state.position);

  while (ch !== 0) ***REMOVED***
    while (is_WHITE_SPACE(ch)) ***REMOVED***
      ch = state.input.charCodeAt(++state.position);
    ***REMOVED***

    if (allowComments && ch === 0x23/* # */) ***REMOVED***
      do ***REMOVED***
        ch = state.input.charCodeAt(++state.position);
      ***REMOVED*** while (ch !== 0x0A/* LF */ && ch !== 0x0D/* CR */ && ch !== 0);
    ***REMOVED***

    if (is_EOL(ch)) ***REMOVED***
      readLineBreak(state);

      ch = state.input.charCodeAt(state.position);
      lineBreaks++;
      state.lineIndent = 0;

      while (ch === 0x20/* Space */) ***REMOVED***
        state.lineIndent++;
        ch = state.input.charCodeAt(++state.position);
      ***REMOVED***
    ***REMOVED*** else ***REMOVED***
      break;
    ***REMOVED***
  ***REMOVED***

  if (checkIndent !== -1 && lineBreaks !== 0 && state.lineIndent < checkIndent) ***REMOVED***
    throwWarning(state, 'deficient indentation');
  ***REMOVED***

  return lineBreaks;
***REMOVED***

function testDocumentSeparator(state) ***REMOVED***
  var _position = state.position,
      ch;

  ch = state.input.charCodeAt(_position);

  // Condition state.position === state.lineStart is tested
  // in parent on each call, for efficiency. No needs to test here again.
  if ((ch === 0x2D/* - */ || ch === 0x2E/* . */) &&
      ch === state.input.charCodeAt(_position + 1) &&
      ch === state.input.charCodeAt(_position + 2)) ***REMOVED***

    _position += 3;

    ch = state.input.charCodeAt(_position);

    if (ch === 0 || is_WS_OR_EOL(ch)) ***REMOVED***
      return true;
    ***REMOVED***
  ***REMOVED***

  return false;
***REMOVED***

function writeFoldedLines(state, count) ***REMOVED***
  if (count === 1) ***REMOVED***
    state.result += ' ';
  ***REMOVED*** else if (count > 1) ***REMOVED***
    state.result += common.repeat('\n', count - 1);
  ***REMOVED***
***REMOVED***


function readPlainScalar(state, nodeIndent, withinFlowCollection) ***REMOVED***
  var preceding,
      following,
      captureStart,
      captureEnd,
      hasPendingContent,
      _line,
      _lineStart,
      _lineIndent,
      _kind = state.kind,
      _result = state.result,
      ch;

  ch = state.input.charCodeAt(state.position);

  if (is_WS_OR_EOL(ch)      ||
      is_FLOW_INDICATOR(ch) ||
      ch === 0x23/* # */    ||
      ch === 0x26/* & */    ||
      ch === 0x2A/* * */    ||
      ch === 0x21/* ! */    ||
      ch === 0x7C/* | */    ||
      ch === 0x3E/* > */    ||
      ch === 0x27/* ' */    ||
      ch === 0x22/* " */    ||
      ch === 0x25/* % */    ||
      ch === 0x40/* @ */    ||
      ch === 0x60/* ` */) ***REMOVED***
    return false;
  ***REMOVED***

  if (ch === 0x3F/* ? */ || ch === 0x2D/* - */) ***REMOVED***
    following = state.input.charCodeAt(state.position + 1);

    if (is_WS_OR_EOL(following) ||
        withinFlowCollection && is_FLOW_INDICATOR(following)) ***REMOVED***
      return false;
    ***REMOVED***
  ***REMOVED***

  state.kind = 'scalar';
  state.result = '';
  captureStart = captureEnd = state.position;
  hasPendingContent = false;

  while (ch !== 0) ***REMOVED***
    if (ch === 0x3A/* : */) ***REMOVED***
      following = state.input.charCodeAt(state.position + 1);

      if (is_WS_OR_EOL(following) ||
          withinFlowCollection && is_FLOW_INDICATOR(following)) ***REMOVED***
        break;
      ***REMOVED***

    ***REMOVED*** else if (ch === 0x23/* # */) ***REMOVED***
      preceding = state.input.charCodeAt(state.position - 1);

      if (is_WS_OR_EOL(preceding)) ***REMOVED***
        break;
      ***REMOVED***

    ***REMOVED*** else if ((state.position === state.lineStart && testDocumentSeparator(state)) ||
               withinFlowCollection && is_FLOW_INDICATOR(ch)) ***REMOVED***
      break;

    ***REMOVED*** else if (is_EOL(ch)) ***REMOVED***
      _line = state.line;
      _lineStart = state.lineStart;
      _lineIndent = state.lineIndent;
      skipSeparationSpace(state, false, -1);

      if (state.lineIndent >= nodeIndent) ***REMOVED***
        hasPendingContent = true;
        ch = state.input.charCodeAt(state.position);
        continue;
      ***REMOVED*** else ***REMOVED***
        state.position = captureEnd;
        state.line = _line;
        state.lineStart = _lineStart;
        state.lineIndent = _lineIndent;
        break;
      ***REMOVED***
    ***REMOVED***

    if (hasPendingContent) ***REMOVED***
      captureSegment(state, captureStart, captureEnd, false);
      writeFoldedLines(state, state.line - _line);
      captureStart = captureEnd = state.position;
      hasPendingContent = false;
    ***REMOVED***

    if (!is_WHITE_SPACE(ch)) ***REMOVED***
      captureEnd = state.position + 1;
    ***REMOVED***

    ch = state.input.charCodeAt(++state.position);
  ***REMOVED***

  captureSegment(state, captureStart, captureEnd, false);

  if (state.result) ***REMOVED***
    return true;
  ***REMOVED***

  state.kind = _kind;
  state.result = _result;
  return false;
***REMOVED***

function readSingleQuotedScalar(state, nodeIndent) ***REMOVED***
  var ch,
      captureStart, captureEnd;

  ch = state.input.charCodeAt(state.position);

  if (ch !== 0x27/* ' */) ***REMOVED***
    return false;
  ***REMOVED***

  state.kind = 'scalar';
  state.result = '';
  state.position++;
  captureStart = captureEnd = state.position;

  while ((ch = state.input.charCodeAt(state.position)) !== 0) ***REMOVED***
    if (ch === 0x27/* ' */) ***REMOVED***
      captureSegment(state, captureStart, state.position, true);
      ch = state.input.charCodeAt(++state.position);

      if (ch === 0x27/* ' */) ***REMOVED***
        captureStart = state.position;
        state.position++;
        captureEnd = state.position;
      ***REMOVED*** else ***REMOVED***
        return true;
      ***REMOVED***

    ***REMOVED*** else if (is_EOL(ch)) ***REMOVED***
      captureSegment(state, captureStart, captureEnd, true);
      writeFoldedLines(state, skipSeparationSpace(state, false, nodeIndent));
      captureStart = captureEnd = state.position;

    ***REMOVED*** else if (state.position === state.lineStart && testDocumentSeparator(state)) ***REMOVED***
      throwError(state, 'unexpected end of the document within a single quoted scalar');

    ***REMOVED*** else ***REMOVED***
      state.position++;
      captureEnd = state.position;
    ***REMOVED***
  ***REMOVED***

  throwError(state, 'unexpected end of the stream within a single quoted scalar');
***REMOVED***

function readDoubleQuotedScalar(state, nodeIndent) ***REMOVED***
  var captureStart,
      captureEnd,
      hexLength,
      hexResult,
      tmp,
      ch;

  ch = state.input.charCodeAt(state.position);

  if (ch !== 0x22/* " */) ***REMOVED***
    return false;
  ***REMOVED***

  state.kind = 'scalar';
  state.result = '';
  state.position++;
  captureStart = captureEnd = state.position;

  while ((ch = state.input.charCodeAt(state.position)) !== 0) ***REMOVED***
    if (ch === 0x22/* " */) ***REMOVED***
      captureSegment(state, captureStart, state.position, true);
      state.position++;
      return true;

    ***REMOVED*** else if (ch === 0x5C/* \ */) ***REMOVED***
      captureSegment(state, captureStart, state.position, true);
      ch = state.input.charCodeAt(++state.position);

      if (is_EOL(ch)) ***REMOVED***
        skipSeparationSpace(state, false, nodeIndent);

        // TODO: rework to inline fn with no type cast?
      ***REMOVED*** else if (ch < 256 && simpleEscapeCheck[ch]) ***REMOVED***
        state.result += simpleEscapeMap[ch];
        state.position++;

      ***REMOVED*** else if ((tmp = escapedHexLen(ch)) > 0) ***REMOVED***
        hexLength = tmp;
        hexResult = 0;

        for (; hexLength > 0; hexLength--) ***REMOVED***
          ch = state.input.charCodeAt(++state.position);

          if ((tmp = fromHexCode(ch)) >= 0) ***REMOVED***
            hexResult = (hexResult << 4) + tmp;

          ***REMOVED*** else ***REMOVED***
            throwError(state, 'expected hexadecimal character');
          ***REMOVED***
        ***REMOVED***

        state.result += charFromCodepoint(hexResult);

        state.position++;

      ***REMOVED*** else ***REMOVED***
        throwError(state, 'unknown escape sequence');
      ***REMOVED***

      captureStart = captureEnd = state.position;

    ***REMOVED*** else if (is_EOL(ch)) ***REMOVED***
      captureSegment(state, captureStart, captureEnd, true);
      writeFoldedLines(state, skipSeparationSpace(state, false, nodeIndent));
      captureStart = captureEnd = state.position;

    ***REMOVED*** else if (state.position === state.lineStart && testDocumentSeparator(state)) ***REMOVED***
      throwError(state, 'unexpected end of the document within a double quoted scalar');

    ***REMOVED*** else ***REMOVED***
      state.position++;
      captureEnd = state.position;
    ***REMOVED***
  ***REMOVED***

  throwError(state, 'unexpected end of the stream within a double quoted scalar');
***REMOVED***

function readFlowCollection(state, nodeIndent) ***REMOVED***
  var readNext = true,
      _line,
      _tag     = state.tag,
      _result,
      _anchor  = state.anchor,
      following,
      terminator,
      isPair,
      isExplicitPair,
      isMapping,
      overridableKeys = ***REMOVED******REMOVED***,
      keyNode,
      keyTag,
      valueNode,
      ch;

  ch = state.input.charCodeAt(state.position);

  if (ch === 0x5B/* [ */) ***REMOVED***
    terminator = 0x5D;/* ] */
    isMapping = false;
    _result = [];
  ***REMOVED*** else if (ch === 0x7B/* ***REMOVED*** */) ***REMOVED***
    terminator = 0x7D;/* ***REMOVED*** */
    isMapping = true;
    _result = ***REMOVED******REMOVED***;
  ***REMOVED*** else ***REMOVED***
    return false;
  ***REMOVED***

  if (state.anchor !== null) ***REMOVED***
    state.anchorMap[state.anchor] = _result;
  ***REMOVED***

  ch = state.input.charCodeAt(++state.position);

  while (ch !== 0) ***REMOVED***
    skipSeparationSpace(state, true, nodeIndent);

    ch = state.input.charCodeAt(state.position);

    if (ch === terminator) ***REMOVED***
      state.position++;
      state.tag = _tag;
      state.anchor = _anchor;
      state.kind = isMapping ? 'mapping' : 'sequence';
      state.result = _result;
      return true;
    ***REMOVED*** else if (!readNext) ***REMOVED***
      throwError(state, 'missed comma between flow collection entries');
    ***REMOVED***

    keyTag = keyNode = valueNode = null;
    isPair = isExplicitPair = false;

    if (ch === 0x3F/* ? */) ***REMOVED***
      following = state.input.charCodeAt(state.position + 1);

      if (is_WS_OR_EOL(following)) ***REMOVED***
        isPair = isExplicitPair = true;
        state.position++;
        skipSeparationSpace(state, true, nodeIndent);
      ***REMOVED***
    ***REMOVED***

    _line = state.line;
    composeNode(state, nodeIndent, CONTEXT_FLOW_IN, false, true);
    keyTag = state.tag;
    keyNode = state.result;
    skipSeparationSpace(state, true, nodeIndent);

    ch = state.input.charCodeAt(state.position);

    if ((isExplicitPair || state.line === _line) && ch === 0x3A/* : */) ***REMOVED***
      isPair = true;
      ch = state.input.charCodeAt(++state.position);
      skipSeparationSpace(state, true, nodeIndent);
      composeNode(state, nodeIndent, CONTEXT_FLOW_IN, false, true);
      valueNode = state.result;
    ***REMOVED***

    if (isMapping) ***REMOVED***
      storeMappingPair(state, _result, overridableKeys, keyTag, keyNode, valueNode);
    ***REMOVED*** else if (isPair) ***REMOVED***
      _result.push(storeMappingPair(state, null, overridableKeys, keyTag, keyNode, valueNode));
    ***REMOVED*** else ***REMOVED***
      _result.push(keyNode);
    ***REMOVED***

    skipSeparationSpace(state, true, nodeIndent);

    ch = state.input.charCodeAt(state.position);

    if (ch === 0x2C/* , */) ***REMOVED***
      readNext = true;
      ch = state.input.charCodeAt(++state.position);
    ***REMOVED*** else ***REMOVED***
      readNext = false;
    ***REMOVED***
  ***REMOVED***

  throwError(state, 'unexpected end of the stream within a flow collection');
***REMOVED***

function readBlockScalar(state, nodeIndent) ***REMOVED***
  var captureStart,
      folding,
      chomping       = CHOMPING_CLIP,
      didReadContent = false,
      detectedIndent = false,
      textIndent     = nodeIndent,
      emptyLines     = 0,
      atMoreIndented = false,
      tmp,
      ch;

  ch = state.input.charCodeAt(state.position);

  if (ch === 0x7C/* | */) ***REMOVED***
    folding = false;
  ***REMOVED*** else if (ch === 0x3E/* > */) ***REMOVED***
    folding = true;
  ***REMOVED*** else ***REMOVED***
    return false;
  ***REMOVED***

  state.kind = 'scalar';
  state.result = '';

  while (ch !== 0) ***REMOVED***
    ch = state.input.charCodeAt(++state.position);

    if (ch === 0x2B/* + */ || ch === 0x2D/* - */) ***REMOVED***
      if (CHOMPING_CLIP === chomping) ***REMOVED***
        chomping = (ch === 0x2B/* + */) ? CHOMPING_KEEP : CHOMPING_STRIP;
      ***REMOVED*** else ***REMOVED***
        throwError(state, 'repeat of a chomping mode identifier');
      ***REMOVED***

    ***REMOVED*** else if ((tmp = fromDecimalCode(ch)) >= 0) ***REMOVED***
      if (tmp === 0) ***REMOVED***
        throwError(state, 'bad explicit indentation width of a block scalar; it cannot be less than one');
      ***REMOVED*** else if (!detectedIndent) ***REMOVED***
        textIndent = nodeIndent + tmp - 1;
        detectedIndent = true;
      ***REMOVED*** else ***REMOVED***
        throwError(state, 'repeat of an indentation width identifier');
      ***REMOVED***

    ***REMOVED*** else ***REMOVED***
      break;
    ***REMOVED***
  ***REMOVED***

  if (is_WHITE_SPACE(ch)) ***REMOVED***
    do ***REMOVED*** ch = state.input.charCodeAt(++state.position); ***REMOVED***
    while (is_WHITE_SPACE(ch));

    if (ch === 0x23/* # */) ***REMOVED***
      do ***REMOVED*** ch = state.input.charCodeAt(++state.position); ***REMOVED***
      while (!is_EOL(ch) && (ch !== 0));
    ***REMOVED***
  ***REMOVED***

  while (ch !== 0) ***REMOVED***
    readLineBreak(state);
    state.lineIndent = 0;

    ch = state.input.charCodeAt(state.position);

    while ((!detectedIndent || state.lineIndent < textIndent) &&
           (ch === 0x20/* Space */)) ***REMOVED***
      state.lineIndent++;
      ch = state.input.charCodeAt(++state.position);
    ***REMOVED***

    if (!detectedIndent && state.lineIndent > textIndent) ***REMOVED***
      textIndent = state.lineIndent;
    ***REMOVED***

    if (is_EOL(ch)) ***REMOVED***
      emptyLines++;
      continue;
    ***REMOVED***

    // End of the scalar.
    if (state.lineIndent < textIndent) ***REMOVED***

      // Perform the chomping.
      if (chomping === CHOMPING_KEEP) ***REMOVED***
        state.result += common.repeat('\n', didReadContent ? 1 + emptyLines : emptyLines);
      ***REMOVED*** else if (chomping === CHOMPING_CLIP) ***REMOVED***
        if (didReadContent) ***REMOVED*** // i.e. only if the scalar is not empty.
          state.result += '\n';
        ***REMOVED***
      ***REMOVED***

      // Break this `while` cycle and go to the funciton's epilogue.
      break;
    ***REMOVED***

    // Folded style: use fancy rules to handle line breaks.
    if (folding) ***REMOVED***

      // Lines starting with white space characters (more-indented lines) are not folded.
      if (is_WHITE_SPACE(ch)) ***REMOVED***
        atMoreIndented = true;
        // except for the first content line (cf. Example 8.1)
        state.result += common.repeat('\n', didReadContent ? 1 + emptyLines : emptyLines);

      // End of more-indented block.
      ***REMOVED*** else if (atMoreIndented) ***REMOVED***
        atMoreIndented = false;
        state.result += common.repeat('\n', emptyLines + 1);

      // Just one line break - perceive as the same line.
      ***REMOVED*** else if (emptyLines === 0) ***REMOVED***
        if (didReadContent) ***REMOVED*** // i.e. only if we have already read some scalar content.
          state.result += ' ';
        ***REMOVED***

      // Several line breaks - perceive as different lines.
      ***REMOVED*** else ***REMOVED***
        state.result += common.repeat('\n', emptyLines);
      ***REMOVED***

    // Literal style: just add exact number of line breaks between content lines.
    ***REMOVED*** else ***REMOVED***
      // Keep all line breaks except the header line break.
      state.result += common.repeat('\n', didReadContent ? 1 + emptyLines : emptyLines);
    ***REMOVED***

    didReadContent = true;
    detectedIndent = true;
    emptyLines = 0;
    captureStart = state.position;

    while (!is_EOL(ch) && (ch !== 0)) ***REMOVED***
      ch = state.input.charCodeAt(++state.position);
    ***REMOVED***

    captureSegment(state, captureStart, state.position, false);
  ***REMOVED***

  return true;
***REMOVED***

function readBlockSequence(state, nodeIndent) ***REMOVED***
  var _line,
      _tag      = state.tag,
      _anchor   = state.anchor,
      _result   = [],
      following,
      detected  = false,
      ch;

  if (state.anchor !== null) ***REMOVED***
    state.anchorMap[state.anchor] = _result;
  ***REMOVED***

  ch = state.input.charCodeAt(state.position);

  while (ch !== 0) ***REMOVED***

    if (ch !== 0x2D/* - */) ***REMOVED***
      break;
    ***REMOVED***

    following = state.input.charCodeAt(state.position + 1);

    if (!is_WS_OR_EOL(following)) ***REMOVED***
      break;
    ***REMOVED***

    detected = true;
    state.position++;

    if (skipSeparationSpace(state, true, -1)) ***REMOVED***
      if (state.lineIndent <= nodeIndent) ***REMOVED***
        _result.push(null);
        ch = state.input.charCodeAt(state.position);
        continue;
      ***REMOVED***
    ***REMOVED***

    _line = state.line;
    composeNode(state, nodeIndent, CONTEXT_BLOCK_IN, false, true);
    _result.push(state.result);
    skipSeparationSpace(state, true, -1);

    ch = state.input.charCodeAt(state.position);

    if ((state.line === _line || state.lineIndent > nodeIndent) && (ch !== 0)) ***REMOVED***
      throwError(state, 'bad indentation of a sequence entry');
    ***REMOVED*** else if (state.lineIndent < nodeIndent) ***REMOVED***
      break;
    ***REMOVED***
  ***REMOVED***

  if (detected) ***REMOVED***
    state.tag = _tag;
    state.anchor = _anchor;
    state.kind = 'sequence';
    state.result = _result;
    return true;
  ***REMOVED***
  return false;
***REMOVED***

function readBlockMapping(state, nodeIndent, flowIndent) ***REMOVED***
  var following,
      allowCompact,
      _line,
      _pos,
      _tag          = state.tag,
      _anchor       = state.anchor,
      _result       = ***REMOVED******REMOVED***,
      overridableKeys = ***REMOVED******REMOVED***,
      keyTag        = null,
      keyNode       = null,
      valueNode     = null,
      atExplicitKey = false,
      detected      = false,
      ch;

  if (state.anchor !== null) ***REMOVED***
    state.anchorMap[state.anchor] = _result;
  ***REMOVED***

  ch = state.input.charCodeAt(state.position);

  while (ch !== 0) ***REMOVED***
    following = state.input.charCodeAt(state.position + 1);
    _line = state.line; // Save the current line.
    _pos = state.position;

    //
    // Explicit notation case. There are two separate blocks:
    // first for the key (denoted by "?") and second for the value (denoted by ":")
    //
    if ((ch === 0x3F/* ? */ || ch === 0x3A/* : */) && is_WS_OR_EOL(following)) ***REMOVED***

      if (ch === 0x3F/* ? */) ***REMOVED***
        if (atExplicitKey) ***REMOVED***
          storeMappingPair(state, _result, overridableKeys, keyTag, keyNode, null);
          keyTag = keyNode = valueNode = null;
        ***REMOVED***

        detected = true;
        atExplicitKey = true;
        allowCompact = true;

      ***REMOVED*** else if (atExplicitKey) ***REMOVED***
        // i.e. 0x3A/* : */ === character after the explicit key.
        atExplicitKey = false;
        allowCompact = true;

      ***REMOVED*** else ***REMOVED***
        throwError(state, 'incomplete explicit mapping pair; a key node is missed; or followed by a non-tabulated empty line');
      ***REMOVED***

      state.position += 1;
      ch = following;

    //
    // Implicit notation case. Flow-style node as the key first, then ":", and the value.
    //
    ***REMOVED*** else if (composeNode(state, flowIndent, CONTEXT_FLOW_OUT, false, true)) ***REMOVED***

      if (state.line === _line) ***REMOVED***
        ch = state.input.charCodeAt(state.position);

        while (is_WHITE_SPACE(ch)) ***REMOVED***
          ch = state.input.charCodeAt(++state.position);
        ***REMOVED***

        if (ch === 0x3A/* : */) ***REMOVED***
          ch = state.input.charCodeAt(++state.position);

          if (!is_WS_OR_EOL(ch)) ***REMOVED***
            throwError(state, 'a whitespace character is expected after the key-value separator within a block mapping');
          ***REMOVED***

          if (atExplicitKey) ***REMOVED***
            storeMappingPair(state, _result, overridableKeys, keyTag, keyNode, null);
            keyTag = keyNode = valueNode = null;
          ***REMOVED***

          detected = true;
          atExplicitKey = false;
          allowCompact = false;
          keyTag = state.tag;
          keyNode = state.result;

        ***REMOVED*** else if (detected) ***REMOVED***
          throwError(state, 'can not read an implicit mapping pair; a colon is missed');

        ***REMOVED*** else ***REMOVED***
          state.tag = _tag;
          state.anchor = _anchor;
          return true; // Keep the result of `composeNode`.
        ***REMOVED***

      ***REMOVED*** else if (detected) ***REMOVED***
        throwError(state, 'can not read a block mapping entry; a multiline key may not be an implicit key');

      ***REMOVED*** else ***REMOVED***
        state.tag = _tag;
        state.anchor = _anchor;
        return true; // Keep the result of `composeNode`.
      ***REMOVED***

    ***REMOVED*** else ***REMOVED***
      break; // Reading is done. Go to the epilogue.
    ***REMOVED***

    //
    // Common reading code for both explicit and implicit notations.
    //
    if (state.line === _line || state.lineIndent > nodeIndent) ***REMOVED***
      if (composeNode(state, nodeIndent, CONTEXT_BLOCK_OUT, true, allowCompact)) ***REMOVED***
        if (atExplicitKey) ***REMOVED***
          keyNode = state.result;
        ***REMOVED*** else ***REMOVED***
          valueNode = state.result;
        ***REMOVED***
      ***REMOVED***

      if (!atExplicitKey) ***REMOVED***
        storeMappingPair(state, _result, overridableKeys, keyTag, keyNode, valueNode, _line, _pos);
        keyTag = keyNode = valueNode = null;
      ***REMOVED***

      skipSeparationSpace(state, true, -1);
      ch = state.input.charCodeAt(state.position);
    ***REMOVED***

    if (state.lineIndent > nodeIndent && (ch !== 0)) ***REMOVED***
      throwError(state, 'bad indentation of a mapping entry');
    ***REMOVED*** else if (state.lineIndent < nodeIndent) ***REMOVED***
      break;
    ***REMOVED***
  ***REMOVED***

  //
  // Epilogue.
  //

  // Special case: last mapping's node contains only the key in explicit notation.
  if (atExplicitKey) ***REMOVED***
    storeMappingPair(state, _result, overridableKeys, keyTag, keyNode, null);
  ***REMOVED***

  // Expose the resulting mapping.
  if (detected) ***REMOVED***
    state.tag = _tag;
    state.anchor = _anchor;
    state.kind = 'mapping';
    state.result = _result;
  ***REMOVED***

  return detected;
***REMOVED***

function readTagProperty(state) ***REMOVED***
  var _position,
      isVerbatim = false,
      isNamed    = false,
      tagHandle,
      tagName,
      ch;

  ch = state.input.charCodeAt(state.position);

  if (ch !== 0x21/* ! */) return false;

  if (state.tag !== null) ***REMOVED***
    throwError(state, 'duplication of a tag property');
  ***REMOVED***

  ch = state.input.charCodeAt(++state.position);

  if (ch === 0x3C/* < */) ***REMOVED***
    isVerbatim = true;
    ch = state.input.charCodeAt(++state.position);

  ***REMOVED*** else if (ch === 0x21/* ! */) ***REMOVED***
    isNamed = true;
    tagHandle = '!!';
    ch = state.input.charCodeAt(++state.position);

  ***REMOVED*** else ***REMOVED***
    tagHandle = '!';
  ***REMOVED***

  _position = state.position;

  if (isVerbatim) ***REMOVED***
    do ***REMOVED*** ch = state.input.charCodeAt(++state.position); ***REMOVED***
    while (ch !== 0 && ch !== 0x3E/* > */);

    if (state.position < state.length) ***REMOVED***
      tagName = state.input.slice(_position, state.position);
      ch = state.input.charCodeAt(++state.position);
    ***REMOVED*** else ***REMOVED***
      throwError(state, 'unexpected end of the stream within a verbatim tag');
    ***REMOVED***
  ***REMOVED*** else ***REMOVED***
    while (ch !== 0 && !is_WS_OR_EOL(ch)) ***REMOVED***

      if (ch === 0x21/* ! */) ***REMOVED***
        if (!isNamed) ***REMOVED***
          tagHandle = state.input.slice(_position - 1, state.position + 1);

          if (!PATTERN_TAG_HANDLE.test(tagHandle)) ***REMOVED***
            throwError(state, 'named tag handle cannot contain such characters');
          ***REMOVED***

          isNamed = true;
          _position = state.position + 1;
        ***REMOVED*** else ***REMOVED***
          throwError(state, 'tag suffix cannot contain exclamation marks');
        ***REMOVED***
      ***REMOVED***

      ch = state.input.charCodeAt(++state.position);
    ***REMOVED***

    tagName = state.input.slice(_position, state.position);

    if (PATTERN_FLOW_INDICATORS.test(tagName)) ***REMOVED***
      throwError(state, 'tag suffix cannot contain flow indicator characters');
    ***REMOVED***
  ***REMOVED***

  if (tagName && !PATTERN_TAG_URI.test(tagName)) ***REMOVED***
    throwError(state, 'tag name cannot contain such characters: ' + tagName);
  ***REMOVED***

  if (isVerbatim) ***REMOVED***
    state.tag = tagName;

  ***REMOVED*** else if (_hasOwnProperty.call(state.tagMap, tagHandle)) ***REMOVED***
    state.tag = state.tagMap[tagHandle] + tagName;

  ***REMOVED*** else if (tagHandle === '!') ***REMOVED***
    state.tag = '!' + tagName;

  ***REMOVED*** else if (tagHandle === '!!') ***REMOVED***
    state.tag = 'tag:yaml.org,2002:' + tagName;

  ***REMOVED*** else ***REMOVED***
    throwError(state, 'undeclared tag handle "' + tagHandle + '"');
  ***REMOVED***

  return true;
***REMOVED***

function readAnchorProperty(state) ***REMOVED***
  var _position,
      ch;

  ch = state.input.charCodeAt(state.position);

  if (ch !== 0x26/* & */) return false;

  if (state.anchor !== null) ***REMOVED***
    throwError(state, 'duplication of an anchor property');
  ***REMOVED***

  ch = state.input.charCodeAt(++state.position);
  _position = state.position;

  while (ch !== 0 && !is_WS_OR_EOL(ch) && !is_FLOW_INDICATOR(ch)) ***REMOVED***
    ch = state.input.charCodeAt(++state.position);
  ***REMOVED***

  if (state.position === _position) ***REMOVED***
    throwError(state, 'name of an anchor node must contain at least one character');
  ***REMOVED***

  state.anchor = state.input.slice(_position, state.position);
  return true;
***REMOVED***

function readAlias(state) ***REMOVED***
  var _position, alias,
      ch;

  ch = state.input.charCodeAt(state.position);

  if (ch !== 0x2A/* * */) return false;

  ch = state.input.charCodeAt(++state.position);
  _position = state.position;

  while (ch !== 0 && !is_WS_OR_EOL(ch) && !is_FLOW_INDICATOR(ch)) ***REMOVED***
    ch = state.input.charCodeAt(++state.position);
  ***REMOVED***

  if (state.position === _position) ***REMOVED***
    throwError(state, 'name of an alias node must contain at least one character');
  ***REMOVED***

  alias = state.input.slice(_position, state.position);

  if (!state.anchorMap.hasOwnProperty(alias)) ***REMOVED***
    throwError(state, 'unidentified alias "' + alias + '"');
  ***REMOVED***

  state.result = state.anchorMap[alias];
  skipSeparationSpace(state, true, -1);
  return true;
***REMOVED***

function composeNode(state, parentIndent, nodeContext, allowToSeek, allowCompact) ***REMOVED***
  var allowBlockStyles,
      allowBlockScalars,
      allowBlockCollections,
      indentStatus = 1, // 1: this>parent, 0: this=parent, -1: this<parent
      atNewLine  = false,
      hasContent = false,
      typeIndex,
      typeQuantity,
      type,
      flowIndent,
      blockIndent;

  if (state.listener !== null) ***REMOVED***
    state.listener('open', state);
  ***REMOVED***

  state.tag    = null;
  state.anchor = null;
  state.kind   = null;
  state.result = null;

  allowBlockStyles = allowBlockScalars = allowBlockCollections =
    CONTEXT_BLOCK_OUT === nodeContext ||
    CONTEXT_BLOCK_IN  === nodeContext;

  if (allowToSeek) ***REMOVED***
    if (skipSeparationSpace(state, true, -1)) ***REMOVED***
      atNewLine = true;

      if (state.lineIndent > parentIndent) ***REMOVED***
        indentStatus = 1;
      ***REMOVED*** else if (state.lineIndent === parentIndent) ***REMOVED***
        indentStatus = 0;
      ***REMOVED*** else if (state.lineIndent < parentIndent) ***REMOVED***
        indentStatus = -1;
      ***REMOVED***
    ***REMOVED***
  ***REMOVED***

  if (indentStatus === 1) ***REMOVED***
    while (readTagProperty(state) || readAnchorProperty(state)) ***REMOVED***
      if (skipSeparationSpace(state, true, -1)) ***REMOVED***
        atNewLine = true;
        allowBlockCollections = allowBlockStyles;

        if (state.lineIndent > parentIndent) ***REMOVED***
          indentStatus = 1;
        ***REMOVED*** else if (state.lineIndent === parentIndent) ***REMOVED***
          indentStatus = 0;
        ***REMOVED*** else if (state.lineIndent < parentIndent) ***REMOVED***
          indentStatus = -1;
        ***REMOVED***
      ***REMOVED*** else ***REMOVED***
        allowBlockCollections = false;
      ***REMOVED***
    ***REMOVED***
  ***REMOVED***

  if (allowBlockCollections) ***REMOVED***
    allowBlockCollections = atNewLine || allowCompact;
  ***REMOVED***

  if (indentStatus === 1 || CONTEXT_BLOCK_OUT === nodeContext) ***REMOVED***
    if (CONTEXT_FLOW_IN === nodeContext || CONTEXT_FLOW_OUT === nodeContext) ***REMOVED***
      flowIndent = parentIndent;
    ***REMOVED*** else ***REMOVED***
      flowIndent = parentIndent + 1;
    ***REMOVED***

    blockIndent = state.position - state.lineStart;

    if (indentStatus === 1) ***REMOVED***
      if (allowBlockCollections &&
          (readBlockSequence(state, blockIndent) ||
           readBlockMapping(state, blockIndent, flowIndent)) ||
          readFlowCollection(state, flowIndent)) ***REMOVED***
        hasContent = true;
      ***REMOVED*** else ***REMOVED***
        if ((allowBlockScalars && readBlockScalar(state, flowIndent)) ||
            readSingleQuotedScalar(state, flowIndent) ||
            readDoubleQuotedScalar(state, flowIndent)) ***REMOVED***
          hasContent = true;

        ***REMOVED*** else if (readAlias(state)) ***REMOVED***
          hasContent = true;

          if (state.tag !== null || state.anchor !== null) ***REMOVED***
            throwError(state, 'alias node should not have any properties');
          ***REMOVED***

        ***REMOVED*** else if (readPlainScalar(state, flowIndent, CONTEXT_FLOW_IN === nodeContext)) ***REMOVED***
          hasContent = true;

          if (state.tag === null) ***REMOVED***
            state.tag = '?';
          ***REMOVED***
        ***REMOVED***

        if (state.anchor !== null) ***REMOVED***
          state.anchorMap[state.anchor] = state.result;
        ***REMOVED***
      ***REMOVED***
    ***REMOVED*** else if (indentStatus === 0) ***REMOVED***
      // Special case: block sequences are allowed to have same indentation level as the parent.
      // http://www.yaml.org/spec/1.2/spec.html#id2799784
      hasContent = allowBlockCollections && readBlockSequence(state, blockIndent);
    ***REMOVED***
  ***REMOVED***

  if (state.tag !== null && state.tag !== '!') ***REMOVED***
    if (state.tag === '?') ***REMOVED***
      for (typeIndex = 0, typeQuantity = state.implicitTypes.length; typeIndex < typeQuantity; typeIndex += 1) ***REMOVED***
        type = state.implicitTypes[typeIndex];

        // Implicit resolving is not allowed for non-scalar types, and '?'
        // non-specific tag is only assigned to plain scalars. So, it isn't
        // needed to check for 'kind' conformity.

        if (type.resolve(state.result)) ***REMOVED*** // `state.result` updated in resolver if matched
          state.result = type.construct(state.result);
          state.tag = type.tag;
          if (state.anchor !== null) ***REMOVED***
            state.anchorMap[state.anchor] = state.result;
          ***REMOVED***
          break;
        ***REMOVED***
      ***REMOVED***
    ***REMOVED*** else if (_hasOwnProperty.call(state.typeMap[state.kind || 'fallback'], state.tag)) ***REMOVED***
      type = state.typeMap[state.kind || 'fallback'][state.tag];

      if (state.result !== null && type.kind !== state.kind) ***REMOVED***
        throwError(state, 'unacceptable node kind for !<' + state.tag + '> tag; it should be "' + type.kind + '", not "' + state.kind + '"');
      ***REMOVED***

      if (!type.resolve(state.result)) ***REMOVED*** // `state.result` updated in resolver if matched
        throwError(state, 'cannot resolve a node with !<' + state.tag + '> explicit tag');
      ***REMOVED*** else ***REMOVED***
        state.result = type.construct(state.result);
        if (state.anchor !== null) ***REMOVED***
          state.anchorMap[state.anchor] = state.result;
        ***REMOVED***
      ***REMOVED***
    ***REMOVED*** else ***REMOVED***
      throwError(state, 'unknown tag !<' + state.tag + '>');
    ***REMOVED***
  ***REMOVED***

  if (state.listener !== null) ***REMOVED***
    state.listener('close', state);
  ***REMOVED***
  return state.tag !== null ||  state.anchor !== null || hasContent;
***REMOVED***

function readDocument(state) ***REMOVED***
  var documentStart = state.position,
      _position,
      directiveName,
      directiveArgs,
      hasDirectives = false,
      ch;

  state.version = null;
  state.checkLineBreaks = state.legacy;
  state.tagMap = ***REMOVED******REMOVED***;
  state.anchorMap = ***REMOVED******REMOVED***;

  while ((ch = state.input.charCodeAt(state.position)) !== 0) ***REMOVED***
    skipSeparationSpace(state, true, -1);

    ch = state.input.charCodeAt(state.position);

    if (state.lineIndent > 0 || ch !== 0x25/* % */) ***REMOVED***
      break;
    ***REMOVED***

    hasDirectives = true;
    ch = state.input.charCodeAt(++state.position);
    _position = state.position;

    while (ch !== 0 && !is_WS_OR_EOL(ch)) ***REMOVED***
      ch = state.input.charCodeAt(++state.position);
    ***REMOVED***

    directiveName = state.input.slice(_position, state.position);
    directiveArgs = [];

    if (directiveName.length < 1) ***REMOVED***
      throwError(state, 'directive name must not be less than one character in length');
    ***REMOVED***

    while (ch !== 0) ***REMOVED***
      while (is_WHITE_SPACE(ch)) ***REMOVED***
        ch = state.input.charCodeAt(++state.position);
      ***REMOVED***

      if (ch === 0x23/* # */) ***REMOVED***
        do ***REMOVED*** ch = state.input.charCodeAt(++state.position); ***REMOVED***
        while (ch !== 0 && !is_EOL(ch));
        break;
      ***REMOVED***

      if (is_EOL(ch)) break;

      _position = state.position;

      while (ch !== 0 && !is_WS_OR_EOL(ch)) ***REMOVED***
        ch = state.input.charCodeAt(++state.position);
      ***REMOVED***

      directiveArgs.push(state.input.slice(_position, state.position));
    ***REMOVED***

    if (ch !== 0) readLineBreak(state);

    if (_hasOwnProperty.call(directiveHandlers, directiveName)) ***REMOVED***
      directiveHandlers[directiveName](state, directiveName, directiveArgs);
    ***REMOVED*** else ***REMOVED***
      throwWarning(state, 'unknown document directive "' + directiveName + '"');
    ***REMOVED***
  ***REMOVED***

  skipSeparationSpace(state, true, -1);

  if (state.lineIndent === 0 &&
      state.input.charCodeAt(state.position)     === 0x2D/* - */ &&
      state.input.charCodeAt(state.position + 1) === 0x2D/* - */ &&
      state.input.charCodeAt(state.position + 2) === 0x2D/* - */) ***REMOVED***
    state.position += 3;
    skipSeparationSpace(state, true, -1);

  ***REMOVED*** else if (hasDirectives) ***REMOVED***
    throwError(state, 'directives end mark is expected');
  ***REMOVED***

  composeNode(state, state.lineIndent - 1, CONTEXT_BLOCK_OUT, false, true);
  skipSeparationSpace(state, true, -1);

  if (state.checkLineBreaks &&
      PATTERN_NON_ASCII_LINE_BREAKS.test(state.input.slice(documentStart, state.position))) ***REMOVED***
    throwWarning(state, 'non-ASCII line breaks are interpreted as content');
  ***REMOVED***

  state.documents.push(state.result);

  if (state.position === state.lineStart && testDocumentSeparator(state)) ***REMOVED***

    if (state.input.charCodeAt(state.position) === 0x2E/* . */) ***REMOVED***
      state.position += 3;
      skipSeparationSpace(state, true, -1);
    ***REMOVED***
    return;
  ***REMOVED***

  if (state.position < (state.length - 1)) ***REMOVED***
    throwError(state, 'end of the stream or a document separator is expected');
  ***REMOVED*** else ***REMOVED***
    return;
  ***REMOVED***
***REMOVED***


function loadDocuments(input, options) ***REMOVED***
  input = String(input);
  options = options || ***REMOVED******REMOVED***;

  if (input.length !== 0) ***REMOVED***

    // Add tailing `\n` if not exists
    if (input.charCodeAt(input.length - 1) !== 0x0A/* LF */ &&
        input.charCodeAt(input.length - 1) !== 0x0D/* CR */) ***REMOVED***
      input += '\n';
    ***REMOVED***

    // Strip BOM
    if (input.charCodeAt(0) === 0xFEFF) ***REMOVED***
      input = input.slice(1);
    ***REMOVED***
  ***REMOVED***

  var state = new State(input, options);

  // Use 0 as string terminator. That significantly simplifies bounds check.
  state.input += '\0';

  while (state.input.charCodeAt(state.position) === 0x20/* Space */) ***REMOVED***
    state.lineIndent += 1;
    state.position += 1;
  ***REMOVED***

  while (state.position < (state.length - 1)) ***REMOVED***
    readDocument(state);
  ***REMOVED***

  return state.documents;
***REMOVED***


function loadAll(input, iterator, options) ***REMOVED***
  var documents = loadDocuments(input, options), index, length;

  if (typeof iterator !== 'function') ***REMOVED***
    return documents;
  ***REMOVED***

  for (index = 0, length = documents.length; index < length; index += 1) ***REMOVED***
    iterator(documents[index]);
  ***REMOVED***
***REMOVED***


function load(input, options) ***REMOVED***
  var documents = loadDocuments(input, options);

  if (documents.length === 0) ***REMOVED***
    /*eslint-disable no-undefined*/
    return undefined;
  ***REMOVED*** else if (documents.length === 1) ***REMOVED***
    return documents[0];
  ***REMOVED***
  throw new YAMLException('expected a single document in the stream, but found more');
***REMOVED***


function safeLoadAll(input, output, options) ***REMOVED***
  if (typeof output === 'function') ***REMOVED***
    loadAll(input, output, common.extend(***REMOVED*** schema: DEFAULT_SAFE_SCHEMA ***REMOVED***, options));
  ***REMOVED*** else ***REMOVED***
    return loadAll(input, common.extend(***REMOVED*** schema: DEFAULT_SAFE_SCHEMA ***REMOVED***, options));
  ***REMOVED***
***REMOVED***


function safeLoad(input, options) ***REMOVED***
  return load(input, common.extend(***REMOVED*** schema: DEFAULT_SAFE_SCHEMA ***REMOVED***, options));
***REMOVED***


module.exports.loadAll     = loadAll;
module.exports.load        = load;
module.exports.safeLoadAll = safeLoadAll;
module.exports.safeLoad    = safeLoad;
