// eslint-plugin-node/lib/rules/no-hide-core-modules.js
/**
 * @author Toru Nagashima
 * See LICENSE file in root directory for full license.
 *
 * @deprecated since v4.2.0
 * This rule was based on an invalid assumption.
 * No meaning.
 */
"use strict"

const path = require("path")
const resolve = require("resolve")
const getPackageJson = require("../util/get-package-json")
const mergeVisitorsInPlace = require("../util/merge-visitors-in-place")
const visitImport = require("../util/visit-import")
const visitRequire = require("../util/visit-require")

const CORE_MODULES = new Set([
    "assert",
    "buffer",
    "child_process",
    "cluster",
    "console",
    "constants",
    "crypto",
    "dgram",
    "dns",
    /* "domain", */ "events",
    "fs",
    "http",
    "https",
    "module",
    "net",
    "os",
    "path",
    /* "punycode", */ "querystring",
    "readline",
    "repl",
    "stream",
    "string_decoder",
    "timers",
    "tls",
    "tty",
    "url",
    "util",
    "vm",
    "zlib",
])
const BACK_SLASH = /\\/gu

module.exports = ***REMOVED***
    meta: ***REMOVED***
        docs: ***REMOVED***
            description:
                "disallow third-party modules which are hiding core modules",
            category: "Possible Errors",
            recommended: false,
            url:
                "https://github.com/mysticatea/eslint-plugin-node/blob/v10.0.0/docs/rules/no-hide-core-modules.md",
        ***REMOVED***,
        type: "problem",
        deprecated: true,
        fixable: null,
        schema: [
            ***REMOVED***
                type: "object",
                properties: ***REMOVED***
                    allow: ***REMOVED***
                        type: "array",
                        items: ***REMOVED*** enum: Array.from(CORE_MODULES) ***REMOVED***,
                        additionalItems: false,
                        uniqueItems: true,
                    ***REMOVED***,
                    ignoreDirectDependencies: ***REMOVED*** type: "boolean" ***REMOVED***,
                    ignoreIndirectDependencies: ***REMOVED*** type: "boolean" ***REMOVED***,
                ***REMOVED***,
                additionalProperties: false,
            ***REMOVED***,
        ],
    ***REMOVED***,
    create(context) ***REMOVED***
        if (context.getFilename() === "<input>") ***REMOVED***
            return ***REMOVED******REMOVED***
        ***REMOVED***
        const filePath = path.resolve(context.getFilename())
        const dirPath = path.dirname(filePath)
        const packageJson = getPackageJson(filePath)
        const deps = new Set(
            [].concat(
                Object.keys((packageJson && packageJson.dependencies) || ***REMOVED******REMOVED***),
                Object.keys((packageJson && packageJson.devDependencies) || ***REMOVED******REMOVED***)
            )
        )
        const options = context.options[0] || ***REMOVED******REMOVED***
        const allow = options.allow || []
        const ignoreDirectDependencies = Boolean(
            options.ignoreDirectDependencies
        )
        const ignoreIndirectDependencies = Boolean(
            options.ignoreIndirectDependencies
        )
        const targets = []

        return [
            visitImport(context, ***REMOVED*** includeCore: true ***REMOVED***, importTargets =>
                targets.push(...importTargets)
            ),
            visitRequire(context, ***REMOVED*** includeCore: true ***REMOVED***, requireTargets =>
                targets.push(...requireTargets)
            ),
            ***REMOVED***
                "Program:exit"() ***REMOVED***
                    for (const target of targets.filter(
                        t =>
                            CORE_MODULES.has(t.moduleName) &&
                            t.moduleName === t.name
                    )) ***REMOVED***
                        const name = target.moduleName
                        const allowed =
                            allow.indexOf(name) !== -1 ||
                            (ignoreDirectDependencies && deps.has(name)) ||
                            (ignoreIndirectDependencies && !deps.has(name))

                        if (allowed) ***REMOVED***
                            continue
                        ***REMOVED***

                        let resolved = ""
                        try ***REMOVED***
                            resolved = resolve.sync(`$***REMOVED***name***REMOVED***/`, ***REMOVED***
                                basedir: dirPath,
                            ***REMOVED***)
                        ***REMOVED*** catch (_error) ***REMOVED***
                            continue
                        ***REMOVED***

                        context.report(***REMOVED***
                            node: target.node,
                            loc: target.node.loc,
                            message:
                                "Unexpected import of third-party module '***REMOVED******REMOVED***name***REMOVED******REMOVED***'.",
                            data: ***REMOVED***
                                name: path
                                    .relative(dirPath, resolved)
                                    .replace(BACK_SLASH, "/"),
                            ***REMOVED***,
                        ***REMOVED***)
                    ***REMOVED***
                ***REMOVED***,
            ***REMOVED***,
        ].reduce(
            (mergedVisitor, thisVisitor) =>
                mergeVisitorsInPlace(mergedVisitor, thisVisitor),
            ***REMOVED******REMOVED***
        )
    ***REMOVED***,
***REMOVED***
