// eslint-plugin-node/lib/rules/shebang.js
/**
 * @author Toru Nagashima
 * See LICENSE file in root directory for full license.
 */
"use strict"

const path = require("path")
const getConvertPath = require("../util/get-convert-path")
const getPackageJson = require("../util/get-package-json")

const NODE_SHEBANG = "#!/usr/bin/env node\n"
const SHEBANG_PATTERN = /^(#!.+?)?(\r)?\n/u
const NODE_SHEBANG_PATTERN = /#!\/usr\/bin\/env node(?: [^\r\n]+?)?\n/u

function simulateNodeResolutionAlgorithm(filePath, binField) ***REMOVED***
    const possibilities = [filePath]
    let newFilePath = filePath.replace(/\.js$/u, "")
    possibilities.push(newFilePath)
    newFilePath = newFilePath.replace(/[/\\]index$/u, "")
    possibilities.push(newFilePath)
    return possibilities.includes(binField)
***REMOVED***

/**
 * Checks whether or not a given path is a `bin` file.
 *
 * @param ***REMOVED***string***REMOVED*** filePath - A file path to check.
 * @param ***REMOVED***string|object|undefined***REMOVED*** binField - A value of the `bin` field of `package.json`.
 * @param ***REMOVED***string***REMOVED*** basedir - A directory path that `package.json` exists.
 * @returns ***REMOVED***boolean***REMOVED*** `true` if the file is a `bin` file.
 */
function isBinFile(filePath, binField, basedir) ***REMOVED***
    if (!binField) ***REMOVED***
        return false
    ***REMOVED***
    if (typeof binField === "string") ***REMOVED***
        return simulateNodeResolutionAlgorithm(
            filePath,
            path.resolve(basedir, binField)
        )
    ***REMOVED***
    return Object.keys(binField).some(key =>
        simulateNodeResolutionAlgorithm(
            filePath,
            path.resolve(basedir, binField[key])
        )
    )
***REMOVED***

/**
 * Gets the shebang line (includes a line ending) from a given code.
 *
 * @param ***REMOVED***SourceCode***REMOVED*** sourceCode - A source code object to check.
 * @returns ***REMOVED******REMOVED***length: number, bom: boolean, shebang: string, cr: boolean***REMOVED******REMOVED***
 *      shebang's information.
 *      `retv.shebang` is an empty string if shebang doesn't exist.
 */
function getShebangInfo(sourceCode) ***REMOVED***
    const m = SHEBANG_PATTERN.exec(sourceCode.text)

    return ***REMOVED***
        bom: sourceCode.hasBOM,
        cr: Boolean(m && m[2]),
        length: (m && m[0].length) || 0,
        shebang: (m && m[1] && `$***REMOVED***m[1]***REMOVED***\n`) || "",
    ***REMOVED***
***REMOVED***

module.exports = ***REMOVED***
    meta: ***REMOVED***
        docs: ***REMOVED***
            description: "suggest correct usage of shebang",
            category: "Possible Errors",
            recommended: true,
            url:
                "https://github.com/mysticatea/eslint-plugin-node/blob/v10.0.0/docs/rules/shebang.md",
        ***REMOVED***,
        type: "problem",
        fixable: "code",
        schema: [
            ***REMOVED***
                type: "object",
                properties: ***REMOVED***
                    //
                    convertPath: getConvertPath.schema,
                ***REMOVED***,
                additionalProperties: false,
            ***REMOVED***,
        ],
    ***REMOVED***,
    create(context) ***REMOVED***
        const sourceCode = context.getSourceCode()
        let filePath = context.getFilename()
        if (filePath === "<input>") ***REMOVED***
            return ***REMOVED******REMOVED***
        ***REMOVED***
        filePath = path.resolve(filePath)

        const p = getPackageJson(filePath)
        if (!p) ***REMOVED***
            return ***REMOVED******REMOVED***
        ***REMOVED***

        const basedir = path.dirname(p.filePath)
        filePath = path.join(
            basedir,
            getConvertPath(context)(
                path.relative(basedir, filePath).replace(/\\/gu, "/")
            )
        )

        const needsShebang = isBinFile(filePath, p.bin, basedir)
        const info = getShebangInfo(sourceCode)

        return ***REMOVED***
            Program(node) ***REMOVED***
                if (
                    needsShebang
                        ? NODE_SHEBANG_PATTERN.test(info.shebang)
                        : !info.shebang
                ) ***REMOVED***
                    // Good the shebang target.
                    // Checks BOM and \r.
                    if (needsShebang && info.bom) ***REMOVED***
                        context.report(***REMOVED***
                            node,
                            message: "This file must not have Unicode BOM.",
                            fix(fixer) ***REMOVED***
                                return fixer.removeRange([-1, 0])
                            ***REMOVED***,
                        ***REMOVED***)
                    ***REMOVED***
                    if (needsShebang && info.cr) ***REMOVED***
                        context.report(***REMOVED***
                            node,
                            message:
                                "This file must have Unix linebreaks (LF).",
                            fix(fixer) ***REMOVED***
                                const index = sourceCode.text.indexOf("\r")
                                return fixer.removeRange([index, index + 1])
                            ***REMOVED***,
                        ***REMOVED***)
                    ***REMOVED***
                ***REMOVED*** else if (needsShebang) ***REMOVED***
                    // Shebang is lacking.
                    context.report(***REMOVED***
                        node,
                        message:
                            'This file needs shebang "#!/usr/bin/env node".',
                        fix(fixer) ***REMOVED***
                            return fixer.replaceTextRange(
                                [-1, info.length],
                                NODE_SHEBANG
                            )
                        ***REMOVED***,
                    ***REMOVED***)
                ***REMOVED*** else ***REMOVED***
                    // Shebang is extra.
                    context.report(***REMOVED***
                        node,
                        message: "This file needs no shebang.",
                        fix(fixer) ***REMOVED***
                            return fixer.removeRange([0, info.length])
                        ***REMOVED***,
                    ***REMOVED***)
                ***REMOVED***
            ***REMOVED***,
        ***REMOVED***
    ***REMOVED***,
***REMOVED***
