// eslint-plugin-react/lib/util/props.js
/**
 * @fileoverview Utility functions for props
 */

'use strict';

const astUtil = require('./ast');

/**
 * Checks if the Identifier node passed in looks like a propTypes declaration.
 * @param ***REMOVED***ASTNode***REMOVED*** node The node to check. Must be an Identifier node.
 * @returns ***REMOVED***Boolean***REMOVED*** `true` if the node is a propTypes declaration, `false` if not
 */
function isPropTypesDeclaration(node) ***REMOVED***
  if (node && node.type === 'ClassProperty') ***REMOVED***
    // Flow support
    if (node.typeAnnotation && node.key.name === 'props') ***REMOVED***
      return true;
    ***REMOVED***
  ***REMOVED***
  return astUtil.getPropertyName(node) === 'propTypes';
***REMOVED***

/**
 * Checks if the node passed in looks like a contextTypes declaration.
 * @param ***REMOVED***ASTNode***REMOVED*** node The node to check.
 * @returns ***REMOVED***Boolean***REMOVED*** `true` if the node is a contextTypes declaration, `false` if not
 */
function isContextTypesDeclaration(node) ***REMOVED***
  if (node && node.type === 'ClassProperty') ***REMOVED***
    // Flow support
    if (node.typeAnnotation && node.key.name === 'context') ***REMOVED***
      return true;
    ***REMOVED***
  ***REMOVED***
  return astUtil.getPropertyName(node) === 'contextTypes';
***REMOVED***

/**
 * Checks if the node passed in looks like a contextType declaration.
 * @param ***REMOVED***ASTNode***REMOVED*** node The node to check.
 * @returns ***REMOVED***Boolean***REMOVED*** `true` if the node is a contextType declaration, `false` if not
 */
function isContextTypeDeclaration(node) ***REMOVED***
  return astUtil.getPropertyName(node) === 'contextType';
***REMOVED***

/**
 * Checks if the node passed in looks like a childContextTypes declaration.
 * @param ***REMOVED***ASTNode***REMOVED*** node The node to check.
 * @returns ***REMOVED***Boolean***REMOVED*** `true` if the node is a childContextTypes declaration, `false` if not
 */
function isChildContextTypesDeclaration(node) ***REMOVED***
  return astUtil.getPropertyName(node) === 'childContextTypes';
***REMOVED***

/**
 * Checks if the Identifier node passed in looks like a defaultProps declaration.
 * @param ***REMOVED***ASTNode***REMOVED*** node The node to check. Must be an Identifier node.
 * @returns ***REMOVED***Boolean***REMOVED*** `true` if the node is a defaultProps declaration, `false` if not
 */
function isDefaultPropsDeclaration(node) ***REMOVED***
  const propName = astUtil.getPropertyName(node);
  return (propName === 'defaultProps' || propName === 'getDefaultProps');
***REMOVED***

/**
 * Checks if we are declaring a display name
 * @param ***REMOVED***ASTNode***REMOVED*** node The AST node being checked.
 * @returns ***REMOVED***Boolean***REMOVED*** True if we are declaring a display name, false if not.
 */
function isDisplayNameDeclaration(node) ***REMOVED***
  switch (node.type) ***REMOVED***
    case 'ClassProperty':
      return node.key && node.key.name === 'displayName';
    case 'Identifier':
      return node.name === 'displayName';
    case 'Literal':
      return node.value === 'displayName';
    default:
      return false;
  ***REMOVED***
***REMOVED***

/**
 * Checks if the PropTypes MemberExpression node passed in declares a required propType.
 * @param ***REMOVED***ASTNode***REMOVED*** propTypeExpression node to check. Must be a `PropTypes` MemberExpression.
 * @returns ***REMOVED***Boolean***REMOVED*** `true` if this PropType is required, `false` if not.
 */
function isRequiredPropType(propTypeExpression) ***REMOVED***
  return propTypeExpression.type === 'MemberExpression' && propTypeExpression.property.name === 'isRequired';
***REMOVED***

module.exports = ***REMOVED***
  isPropTypesDeclaration,
  isContextTypesDeclaration,
  isContextTypeDeclaration,
  isChildContextTypesDeclaration,
  isDefaultPropsDeclaration,
  isDisplayNameDeclaration,
  isRequiredPropType
***REMOVED***;
