// rxjs/_esm5/internal/observable/combineLatest.js
/** PURE_IMPORTS_START tslib,_util_isScheduler,_util_isArray,_OuterSubscriber,_util_subscribeToResult,_fromArray PURE_IMPORTS_END */
import * as tslib_1 from "tslib";
import ***REMOVED*** isScheduler ***REMOVED*** from '../util/isScheduler';
import ***REMOVED*** isArray ***REMOVED*** from '../util/isArray';
import ***REMOVED*** OuterSubscriber ***REMOVED*** from '../OuterSubscriber';
import ***REMOVED*** subscribeToResult ***REMOVED*** from '../util/subscribeToResult';
import ***REMOVED*** fromArray ***REMOVED*** from './fromArray';
var NONE = ***REMOVED******REMOVED***;
export function combineLatest() ***REMOVED***
    var observables = [];
    for (var _i = 0; _i < arguments.length; _i++) ***REMOVED***
        observables[_i] = arguments[_i];
    ***REMOVED***
    var resultSelector = null;
    var scheduler = null;
    if (isScheduler(observables[observables.length - 1])) ***REMOVED***
        scheduler = observables.pop();
    ***REMOVED***
    if (typeof observables[observables.length - 1] === 'function') ***REMOVED***
        resultSelector = observables.pop();
    ***REMOVED***
    if (observables.length === 1 && isArray(observables[0])) ***REMOVED***
        observables = observables[0];
    ***REMOVED***
    return fromArray(observables, scheduler).lift(new CombineLatestOperator(resultSelector));
***REMOVED***
var CombineLatestOperator = /*@__PURE__*/ (function () ***REMOVED***
    function CombineLatestOperator(resultSelector) ***REMOVED***
        this.resultSelector = resultSelector;
    ***REMOVED***
    CombineLatestOperator.prototype.call = function (subscriber, source) ***REMOVED***
        return source.subscribe(new CombineLatestSubscriber(subscriber, this.resultSelector));
    ***REMOVED***;
    return CombineLatestOperator;
***REMOVED***());
export ***REMOVED*** CombineLatestOperator ***REMOVED***;
var CombineLatestSubscriber = /*@__PURE__*/ (function (_super) ***REMOVED***
    tslib_1.__extends(CombineLatestSubscriber, _super);
    function CombineLatestSubscriber(destination, resultSelector) ***REMOVED***
        var _this = _super.call(this, destination) || this;
        _this.resultSelector = resultSelector;
        _this.active = 0;
        _this.values = [];
        _this.observables = [];
        return _this;
    ***REMOVED***
    CombineLatestSubscriber.prototype._next = function (observable) ***REMOVED***
        this.values.push(NONE);
        this.observables.push(observable);
    ***REMOVED***;
    CombineLatestSubscriber.prototype._complete = function () ***REMOVED***
        var observables = this.observables;
        var len = observables.length;
        if (len === 0) ***REMOVED***
            this.destination.complete();
        ***REMOVED***
        else ***REMOVED***
            this.active = len;
            this.toRespond = len;
            for (var i = 0; i < len; i++) ***REMOVED***
                var observable = observables[i];
                this.add(subscribeToResult(this, observable, observable, i));
            ***REMOVED***
        ***REMOVED***
    ***REMOVED***;
    CombineLatestSubscriber.prototype.notifyComplete = function (unused) ***REMOVED***
        if ((this.active -= 1) === 0) ***REMOVED***
            this.destination.complete();
        ***REMOVED***
    ***REMOVED***;
    CombineLatestSubscriber.prototype.notifyNext = function (outerValue, innerValue, outerIndex, innerIndex, innerSub) ***REMOVED***
        var values = this.values;
        var oldVal = values[outerIndex];
        var toRespond = !this.toRespond
            ? 0
            : oldVal === NONE ? --this.toRespond : this.toRespond;
        values[outerIndex] = innerValue;
        if (toRespond === 0) ***REMOVED***
            if (this.resultSelector) ***REMOVED***
                this._tryResultSelector(values);
            ***REMOVED***
            else ***REMOVED***
                this.destination.next(values.slice());
            ***REMOVED***
        ***REMOVED***
    ***REMOVED***;
    CombineLatestSubscriber.prototype._tryResultSelector = function (values) ***REMOVED***
        var result;
        try ***REMOVED***
            result = this.resultSelector.apply(this, values);
        ***REMOVED***
        catch (err) ***REMOVED***
            this.destination.error(err);
            return;
        ***REMOVED***
        this.destination.next(result);
    ***REMOVED***;
    return CombineLatestSubscriber;
***REMOVED***(OuterSubscriber));
export ***REMOVED*** CombineLatestSubscriber ***REMOVED***;
