// discord.js/typings/discord.js-test.ts
/// <reference path='index.d.ts' />

import ***REMOVED*** Collector, Message, CollectorFilter, Client, CollectorHandler, MessageReaction, Collection, User, ReactionCollectorOptions, Snowflake ***REMOVED*** from 'discord.js';

const client = new Client(***REMOVED***
	disableEveryone: false,
	disabledEvents: ['GUILD_MEMBER_ADD']
***REMOVED***);

client.on('message', (message) => ***REMOVED***
	if (message.content === 'hello') ***REMOVED***
		message.channel.sendMessage('o/');
	***REMOVED***

	const collector: ReactionCollector = new ReactionCollector(message,
		(reaction: MessageReaction) => reaction.emoji.toString() === '👌',
		***REMOVED*** time: 30e3 ***REMOVED***);
	collector.on('end', collected => console.log(collected));
***REMOVED***);

client.login('dsfsd754.4fds4f68d4f6sd46f4s.7878easfdsgdfFDSIJIO');

export class TestCollector extends Collector<Snowflake, Message> ***REMOVED***
	public filter: CollectorFilter;
	public constructor(client: Client, filter: CollectorFilter) ***REMOVED***
		super(client, filter);
	***REMOVED***

	public handle(message: Message): CollectorHandler<Snowflake, Message> ***REMOVED***
		return ***REMOVED*** key: message.id, value: message ***REMOVED***;
	***REMOVED***

	public cleanup(): void ***REMOVED******REMOVED***
	public postCheck(): null ***REMOVED*** return null; ***REMOVED***
***REMOVED***

class ReactionCollector extends Collector<Snowflake, MessageReaction> ***REMOVED***
	public message: Message;
	public users: Collection<Snowflake, User>;
	public total: number;
	public options: ReactionCollectorOptions;
	public constructor(message: Message, filter: CollectorFilter, options?: ReactionCollectorOptions) ***REMOVED***
		super(message.client, filter, options || ***REMOVED******REMOVED***);
		this.message = message;
		this.users = new Collection<Snowflake, User>();
		this.total = 0;
		this.client.on('messageReactionAdd', this.listener);
	***REMOVED***

	public handle(reaction: MessageReaction): CollectorHandler<Snowflake, MessageReaction> ***REMOVED***
		if (reaction.message.id !== this.message.id) ***REMOVED*** return null; ***REMOVED***
		return ***REMOVED***
			key: reaction.emoji.id || reaction.emoji.name,
			value: reaction
		***REMOVED***;
	***REMOVED***

	public postCheck(reaction: MessageReaction, user: User): string ***REMOVED***
		this.users.set(user.id, user);
		if (this.options.max && ++this.total >= this.options.max) ***REMOVED*** return 'limit'; ***REMOVED***
		if (this.options.maxEmojis && this.collected.size >= this.options.maxEmojis) ***REMOVED*** return 'emojiLimit'; ***REMOVED***
		if (this.options.maxUsers && this.users.size >= this.options.maxUsers) ***REMOVED*** return 'userLimit'; ***REMOVED***
		return null;
	***REMOVED***

	public cleanup(): void ***REMOVED***
		this.client.removeListener('messageReactionAdd', this.listener as any);
	***REMOVED***
***REMOVED***
