// axios/lib/adapters/xhr.js
'use strict';

var utils = require('./../utils');
var settle = require('./../core/settle');
var buildURL = require('./../helpers/buildURL');
var buildFullPath = require('../core/buildFullPath');
var parseHeaders = require('./../helpers/parseHeaders');
var isURLSameOrigin = require('./../helpers/isURLSameOrigin');
var createError = require('../core/createError');

module.exports = function xhrAdapter(config) ***REMOVED***
  return new Promise(function dispatchXhrRequest(resolve, reject) ***REMOVED***
    var requestData = config.data;
    var requestHeaders = config.headers;

    if (utils.isFormData(requestData)) ***REMOVED***
      delete requestHeaders['Content-Type']; // Let the browser set it
    ***REMOVED***

    var request = new XMLHttpRequest();

    // HTTP basic authentication
    if (config.auth) ***REMOVED***
      var username = config.auth.username || '';
      var password = config.auth.password || '';
      requestHeaders.Authorization = 'Basic ' + btoa(username + ':' + password);
    ***REMOVED***

    var fullPath = buildFullPath(config.baseURL, config.url);
    request.open(config.method.toUpperCase(), buildURL(fullPath, config.params, config.paramsSerializer), true);

    // Set the request timeout in MS
    request.timeout = config.timeout;

    // Listen for ready state
    request.onreadystatechange = function handleLoad() ***REMOVED***
      if (!request || request.readyState !== 4) ***REMOVED***
        return;
      ***REMOVED***

      // The request errored out and we didn't get a response, this will be
      // handled by onerror instead
      // With one exception: request that using file: protocol, most browsers
      // will return status as 0 even though it's a successful request
      if (request.status === 0 && !(request.responseURL && request.responseURL.indexOf('file:') === 0)) ***REMOVED***
        return;
      ***REMOVED***

      // Prepare the response
      var responseHeaders = 'getAllResponseHeaders' in request ? parseHeaders(request.getAllResponseHeaders()) : null;
      var responseData = !config.responseType || config.responseType === 'text' ? request.responseText : request.response;
      var response = ***REMOVED***
        data: responseData,
        status: request.status,
        statusText: request.statusText,
        headers: responseHeaders,
        config: config,
        request: request
      ***REMOVED***;

      settle(resolve, reject, response);

      // Clean up request
      request = null;
    ***REMOVED***;

    // Handle browser request cancellation (as opposed to a manual cancellation)
    request.onabort = function handleAbort() ***REMOVED***
      if (!request) ***REMOVED***
        return;
      ***REMOVED***

      reject(createError('Request aborted', config, 'ECONNABORTED', request));

      // Clean up request
      request = null;
    ***REMOVED***;

    // Handle low level network errors
    request.onerror = function handleError() ***REMOVED***
      // Real errors are hidden from us by the browser
      // onerror should only fire if it's a network error
      reject(createError('Network Error', config, null, request));

      // Clean up request
      request = null;
    ***REMOVED***;

    // Handle timeout
    request.ontimeout = function handleTimeout() ***REMOVED***
      var timeoutErrorMessage = 'timeout of ' + config.timeout + 'ms exceeded';
      if (config.timeoutErrorMessage) ***REMOVED***
        timeoutErrorMessage = config.timeoutErrorMessage;
      ***REMOVED***
      reject(createError(timeoutErrorMessage, config, 'ECONNABORTED',
        request));

      // Clean up request
      request = null;
    ***REMOVED***;

    // Add xsrf header
    // This is only done if running in a standard browser environment.
    // Specifically not if we're in a web worker, or react-native.
    if (utils.isStandardBrowserEnv()) ***REMOVED***
      var cookies = require('./../helpers/cookies');

      // Add xsrf header
      var xsrfValue = (config.withCredentials || isURLSameOrigin(fullPath)) && config.xsrfCookieName ?
        cookies.read(config.xsrfCookieName) :
        undefined;

      if (xsrfValue) ***REMOVED***
        requestHeaders[config.xsrfHeaderName] = xsrfValue;
      ***REMOVED***
    ***REMOVED***

    // Add headers to the request
    if ('setRequestHeader' in request) ***REMOVED***
      utils.forEach(requestHeaders, function setRequestHeader(val, key) ***REMOVED***
        if (typeof requestData === 'undefined' && key.toLowerCase() === 'content-type') ***REMOVED***
          // Remove Content-Type if data is undefined
          delete requestHeaders[key];
        ***REMOVED*** else ***REMOVED***
          // Otherwise add header to the request
          request.setRequestHeader(key, val);
        ***REMOVED***
      ***REMOVED***);
    ***REMOVED***

    // Add withCredentials to request if needed
    if (!utils.isUndefined(config.withCredentials)) ***REMOVED***
      request.withCredentials = !!config.withCredentials;
    ***REMOVED***

    // Add responseType to request if needed
    if (config.responseType) ***REMOVED***
      try ***REMOVED***
        request.responseType = config.responseType;
      ***REMOVED*** catch (e) ***REMOVED***
        // Expected DOMException thrown by browsers not compatible XMLHttpRequest Level 2.
        // But, this can be suppressed for 'json' type as it can be parsed by default 'transformResponse' function.
        if (config.responseType !== 'json') ***REMOVED***
          throw e;
        ***REMOVED***
      ***REMOVED***
    ***REMOVED***

    // Handle progress if needed
    if (typeof config.onDownloadProgress === 'function') ***REMOVED***
      request.addEventListener('progress', config.onDownloadProgress);
    ***REMOVED***

    // Not all browsers support upload events
    if (typeof config.onUploadProgress === 'function' && request.upload) ***REMOVED***
      request.upload.addEventListener('progress', config.onUploadProgress);
    ***REMOVED***

    if (config.cancelToken) ***REMOVED***
      // Handle cancellation
      config.cancelToken.promise.then(function onCanceled(cancel) ***REMOVED***
        if (!request) ***REMOVED***
          return;
        ***REMOVED***

        request.abort();
        reject(cancel);
        // Clean up request
        request = null;
      ***REMOVED***);
    ***REMOVED***

    if (requestData === undefined) ***REMOVED***
      requestData = null;
    ***REMOVED***

    // Send the request
    request.send(requestData);
  ***REMOVED***);
***REMOVED***;
