// sprintf-js/gruntfile.js
module.exports = function(grunt) ***REMOVED***
    grunt.initConfig(***REMOVED***
        pkg: grunt.file.readJSON("package.json"),

        uglify: ***REMOVED***
            options: ***REMOVED***
                banner: "/*! <%= pkg.name %> | <%= pkg.author %> | <%= pkg.license %> */\n",
                sourceMap: true
            ***REMOVED***,
            build: ***REMOVED***
                files: [
                    ***REMOVED***
                        src: "src/sprintf.js",
                        dest: "dist/sprintf.min.js"
                    ***REMOVED***,
                    ***REMOVED***
                        src: "src/angular-sprintf.js",
                        dest: "dist/angular-sprintf.min.js"
                    ***REMOVED***
                ]
            ***REMOVED***
        ***REMOVED***,

        watch: ***REMOVED***
            js: ***REMOVED***
                files: "src/*.js",
                tasks: ["uglify"]
            ***REMOVED***
        ***REMOVED***
    ***REMOVED***)

    grunt.loadNpmTasks("grunt-contrib-uglify")
    grunt.loadNpmTasks("grunt-contrib-watch")

    grunt.registerTask("default", ["uglify", "watch"])
***REMOVED***
