// react-is/umd/react-is.development.js
/** @license React v16.12.0
 * react-is.development.js
 *
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

'use strict';

(function (global, factory) ***REMOVED***
	typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports) :
	typeof define === 'function' && define.amd ? define(['exports'], factory) :
	(factory((global.ReactIs = ***REMOVED******REMOVED***)));
***REMOVED***(this, (function (exports) ***REMOVED*** 'use strict';

// The Symbol used to tag the ReactElement-like types. If there is no native Symbol
// nor polyfill, then a plain number is used for performance.
var hasSymbol = typeof Symbol === 'function' && Symbol.for;
var REACT_ELEMENT_TYPE = hasSymbol ? Symbol.for('react.element') : 0xeac7;
var REACT_PORTAL_TYPE = hasSymbol ? Symbol.for('react.portal') : 0xeaca;
var REACT_FRAGMENT_TYPE = hasSymbol ? Symbol.for('react.fragment') : 0xeacb;
var REACT_STRICT_MODE_TYPE = hasSymbol ? Symbol.for('react.strict_mode') : 0xeacc;
var REACT_PROFILER_TYPE = hasSymbol ? Symbol.for('react.profiler') : 0xead2;
var REACT_PROVIDER_TYPE = hasSymbol ? Symbol.for('react.provider') : 0xeacd;
var REACT_CONTEXT_TYPE = hasSymbol ? Symbol.for('react.context') : 0xeace; // TODO: We don't use AsyncMode or ConcurrentMode anymore. They were temporary
// (unstable) APIs that have been removed. Can we remove the symbols?

var REACT_ASYNC_MODE_TYPE = hasSymbol ? Symbol.for('react.async_mode') : 0xeacf;
var REACT_CONCURRENT_MODE_TYPE = hasSymbol ? Symbol.for('react.concurrent_mode') : 0xeacf;
var REACT_FORWARD_REF_TYPE = hasSymbol ? Symbol.for('react.forward_ref') : 0xead0;
var REACT_SUSPENSE_TYPE = hasSymbol ? Symbol.for('react.suspense') : 0xead1;
var REACT_SUSPENSE_LIST_TYPE = hasSymbol ? Symbol.for('react.suspense_list') : 0xead8;
var REACT_MEMO_TYPE = hasSymbol ? Symbol.for('react.memo') : 0xead3;
var REACT_LAZY_TYPE = hasSymbol ? Symbol.for('react.lazy') : 0xead4;
var REACT_FUNDAMENTAL_TYPE = hasSymbol ? Symbol.for('react.fundamental') : 0xead5;
var REACT_RESPONDER_TYPE = hasSymbol ? Symbol.for('react.responder') : 0xead6;
var REACT_SCOPE_TYPE = hasSymbol ? Symbol.for('react.scope') : 0xead7;

function isValidElementType(type) ***REMOVED***
  return typeof type === 'string' || typeof type === 'function' || // Note: its typeof might be other than 'symbol' or 'number' if it's a polyfill.
  type === REACT_FRAGMENT_TYPE || type === REACT_CONCURRENT_MODE_TYPE || type === REACT_PROFILER_TYPE || type === REACT_STRICT_MODE_TYPE || type === REACT_SUSPENSE_TYPE || type === REACT_SUSPENSE_LIST_TYPE || typeof type === 'object' && type !== null && (type.$$typeof === REACT_LAZY_TYPE || type.$$typeof === REACT_MEMO_TYPE || type.$$typeof === REACT_PROVIDER_TYPE || type.$$typeof === REACT_CONTEXT_TYPE || type.$$typeof === REACT_FORWARD_REF_TYPE || type.$$typeof === REACT_FUNDAMENTAL_TYPE || type.$$typeof === REACT_RESPONDER_TYPE || type.$$typeof === REACT_SCOPE_TYPE);
***REMOVED***

/**
 * Forked from fbjs/warning:
 * https://github.com/facebook/fbjs/blob/e66ba20ad5be433eb54423f2b097d829324d9de6/packages/fbjs/src/__forks__/warning.js
 *
 * Only change is we use console.warn instead of console.error,
 * and do nothing when 'console' is not supported.
 * This really simplifies the code.
 * ---
 * Similar to invariant but only logs a warning if the condition is not met.
 * This can be used to log issues in development environments in critical
 * paths. Removing the logging code for production environments will keep the
 * same logic and follow the same code paths.
 */
var lowPriorityWarningWithoutStack = function () ***REMOVED******REMOVED***;

***REMOVED***
  var printWarning = function (format) ***REMOVED***
    for (var _len = arguments.length, args = new Array(_len > 1 ? _len - 1 : 0), _key = 1; _key < _len; _key++) ***REMOVED***
      args[_key - 1] = arguments[_key];
    ***REMOVED***

    var argIndex = 0;
    var message = 'Warning: ' + format.replace(/%s/g, function () ***REMOVED***
      return args[argIndex++];
    ***REMOVED***);

    if (typeof console !== 'undefined') ***REMOVED***
      console.warn(message);
    ***REMOVED***

    try ***REMOVED***
      // --- Welcome to debugging React ---
      // This error was thrown as a convenience so that you can use this stack
      // to find the callsite that caused this warning to fire.
      throw new Error(message);
    ***REMOVED*** catch (x) ***REMOVED******REMOVED***
  ***REMOVED***;

  lowPriorityWarningWithoutStack = function (condition, format) ***REMOVED***
    if (format === undefined) ***REMOVED***
      throw new Error('`lowPriorityWarningWithoutStack(condition, format, ...args)` requires a warning ' + 'message argument');
    ***REMOVED***

    if (!condition) ***REMOVED***
      for (var _len2 = arguments.length, args = new Array(_len2 > 2 ? _len2 - 2 : 0), _key2 = 2; _key2 < _len2; _key2++) ***REMOVED***
        args[_key2 - 2] = arguments[_key2];
      ***REMOVED***

      printWarning.apply(void 0, [format].concat(args));
    ***REMOVED***
  ***REMOVED***;
***REMOVED***

var lowPriorityWarningWithoutStack$1 = lowPriorityWarningWithoutStack;

function typeOf(object) ***REMOVED***
  if (typeof object === 'object' && object !== null) ***REMOVED***
    var $$typeof = object.$$typeof;

    switch ($$typeof) ***REMOVED***
      case REACT_ELEMENT_TYPE:
        var type = object.type;

        switch (type) ***REMOVED***
          case REACT_ASYNC_MODE_TYPE:
          case REACT_CONCURRENT_MODE_TYPE:
          case REACT_FRAGMENT_TYPE:
          case REACT_PROFILER_TYPE:
          case REACT_STRICT_MODE_TYPE:
          case REACT_SUSPENSE_TYPE:
            return type;

          default:
            var $$typeofType = type && type.$$typeof;

            switch ($$typeofType) ***REMOVED***
              case REACT_CONTEXT_TYPE:
              case REACT_FORWARD_REF_TYPE:
              case REACT_LAZY_TYPE:
              case REACT_MEMO_TYPE:
              case REACT_PROVIDER_TYPE:
                return $$typeofType;

              default:
                return $$typeof;
            ***REMOVED***

        ***REMOVED***

      case REACT_PORTAL_TYPE:
        return $$typeof;
    ***REMOVED***
  ***REMOVED***

  return undefined;
***REMOVED*** // AsyncMode is deprecated along with isAsyncMode

var AsyncMode = REACT_ASYNC_MODE_TYPE;
var ConcurrentMode = REACT_CONCURRENT_MODE_TYPE;
var ContextConsumer = REACT_CONTEXT_TYPE;
var ContextProvider = REACT_PROVIDER_TYPE;
var Element = REACT_ELEMENT_TYPE;
var ForwardRef = REACT_FORWARD_REF_TYPE;
var Fragment = REACT_FRAGMENT_TYPE;
var Lazy = REACT_LAZY_TYPE;
var Memo = REACT_MEMO_TYPE;
var Portal = REACT_PORTAL_TYPE;
var Profiler = REACT_PROFILER_TYPE;
var StrictMode = REACT_STRICT_MODE_TYPE;
var Suspense = REACT_SUSPENSE_TYPE;
var hasWarnedAboutDeprecatedIsAsyncMode = false; // AsyncMode should be deprecated

function isAsyncMode(object) ***REMOVED***
  ***REMOVED***
    if (!hasWarnedAboutDeprecatedIsAsyncMode) ***REMOVED***
      hasWarnedAboutDeprecatedIsAsyncMode = true;
      lowPriorityWarningWithoutStack$1(false, 'The ReactIs.isAsyncMode() alias has been deprecated, ' + 'and will be removed in React 17+. Update your code to use ' + 'ReactIs.isConcurrentMode() instead. It has the exact same API.');
    ***REMOVED***
  ***REMOVED***

  return isConcurrentMode(object) || typeOf(object) === REACT_ASYNC_MODE_TYPE;
***REMOVED***
function isConcurrentMode(object) ***REMOVED***
  return typeOf(object) === REACT_CONCURRENT_MODE_TYPE;
***REMOVED***
function isContextConsumer(object) ***REMOVED***
  return typeOf(object) === REACT_CONTEXT_TYPE;
***REMOVED***
function isContextProvider(object) ***REMOVED***
  return typeOf(object) === REACT_PROVIDER_TYPE;
***REMOVED***
function isElement(object) ***REMOVED***
  return typeof object === 'object' && object !== null && object.$$typeof === REACT_ELEMENT_TYPE;
***REMOVED***
function isForwardRef(object) ***REMOVED***
  return typeOf(object) === REACT_FORWARD_REF_TYPE;
***REMOVED***
function isFragment(object) ***REMOVED***
  return typeOf(object) === REACT_FRAGMENT_TYPE;
***REMOVED***
function isLazy(object) ***REMOVED***
  return typeOf(object) === REACT_LAZY_TYPE;
***REMOVED***
function isMemo(object) ***REMOVED***
  return typeOf(object) === REACT_MEMO_TYPE;
***REMOVED***
function isPortal(object) ***REMOVED***
  return typeOf(object) === REACT_PORTAL_TYPE;
***REMOVED***
function isProfiler(object) ***REMOVED***
  return typeOf(object) === REACT_PROFILER_TYPE;
***REMOVED***
function isStrictMode(object) ***REMOVED***
  return typeOf(object) === REACT_STRICT_MODE_TYPE;
***REMOVED***
function isSuspense(object) ***REMOVED***
  return typeOf(object) === REACT_SUSPENSE_TYPE;
***REMOVED***

exports.typeOf = typeOf;
exports.AsyncMode = AsyncMode;
exports.ConcurrentMode = ConcurrentMode;
exports.ContextConsumer = ContextConsumer;
exports.ContextProvider = ContextProvider;
exports.Element = Element;
exports.ForwardRef = ForwardRef;
exports.Fragment = Fragment;
exports.Lazy = Lazy;
exports.Memo = Memo;
exports.Portal = Portal;
exports.Profiler = Profiler;
exports.StrictMode = StrictMode;
exports.Suspense = Suspense;
exports.isValidElementType = isValidElementType;
exports.isAsyncMode = isAsyncMode;
exports.isConcurrentMode = isConcurrentMode;
exports.isContextConsumer = isContextConsumer;
exports.isContextProvider = isContextProvider;
exports.isElement = isElement;
exports.isForwardRef = isForwardRef;
exports.isFragment = isFragment;
exports.isLazy = isLazy;
exports.isMemo = isMemo;
exports.isPortal = isPortal;
exports.isProfiler = isProfiler;
exports.isStrictMode = isStrictMode;
exports.isSuspense = isSuspense;

Object.defineProperty(exports, '__esModule', ***REMOVED*** value: true ***REMOVED***);

***REMOVED***)));
