// mute-stream/mute.js
var Stream = require('stream')

module.exports = MuteStream

// var out = new MuteStream(process.stdout)
// argument auto-pipes
function MuteStream (opts) ***REMOVED***
  Stream.apply(this)
  opts = opts || ***REMOVED******REMOVED***
  this.writable = this.readable = true
  this.muted = false
  this.on('pipe', this._onpipe)
  this.replace = opts.replace

  // For readline-type situations
  // This much at the start of a line being redrawn after a ctrl char
  // is seen (such as backspace) won't be redrawn as the replacement
  this._prompt = opts.prompt || null
  this._hadControl = false
***REMOVED***

MuteStream.prototype = Object.create(Stream.prototype)

Object.defineProperty(MuteStream.prototype, 'constructor', ***REMOVED***
  value: MuteStream,
  enumerable: false
***REMOVED***)

MuteStream.prototype.mute = function () ***REMOVED***
  this.muted = true
***REMOVED***

MuteStream.prototype.unmute = function () ***REMOVED***
  this.muted = false
***REMOVED***

Object.defineProperty(MuteStream.prototype, '_onpipe', ***REMOVED***
  value: onPipe,
  enumerable: false,
  writable: true,
  configurable: true
***REMOVED***)

function onPipe (src) ***REMOVED***
  this._src = src
***REMOVED***

Object.defineProperty(MuteStream.prototype, 'isTTY', ***REMOVED***
  get: getIsTTY,
  set: setIsTTY,
  enumerable: true,
  configurable: true
***REMOVED***)

function getIsTTY () ***REMOVED***
  return( (this._dest) ? this._dest.isTTY
        : (this._src) ? this._src.isTTY
        : false
        )
***REMOVED***

// basically just get replace the getter/setter with a regular value
function setIsTTY (isTTY) ***REMOVED***
  Object.defineProperty(this, 'isTTY', ***REMOVED***
    value: isTTY,
    enumerable: true,
    writable: true,
    configurable: true
  ***REMOVED***)
***REMOVED***

Object.defineProperty(MuteStream.prototype, 'rows', ***REMOVED***
  get: function () ***REMOVED***
    return( this._dest ? this._dest.rows
          : this._src ? this._src.rows
          : undefined )
  ***REMOVED***, enumerable: true, configurable: true ***REMOVED***)

Object.defineProperty(MuteStream.prototype, 'columns', ***REMOVED***
  get: function () ***REMOVED***
    return( this._dest ? this._dest.columns
          : this._src ? this._src.columns
          : undefined )
  ***REMOVED***, enumerable: true, configurable: true ***REMOVED***)


MuteStream.prototype.pipe = function (dest, options) ***REMOVED***
  this._dest = dest
  return Stream.prototype.pipe.call(this, dest, options)
***REMOVED***

MuteStream.prototype.pause = function () ***REMOVED***
  if (this._src) return this._src.pause()
***REMOVED***

MuteStream.prototype.resume = function () ***REMOVED***
  if (this._src) return this._src.resume()
***REMOVED***

MuteStream.prototype.write = function (c) ***REMOVED***
  if (this.muted) ***REMOVED***
    if (!this.replace) return true
    if (c.match(/^\u001b/)) ***REMOVED***
      if(c.indexOf(this._prompt) === 0) ***REMOVED***
        c = c.substr(this._prompt.length);
        c = c.replace(/./g, this.replace);
        c = this._prompt + c;
      ***REMOVED***
      this._hadControl = true
      return this.emit('data', c)
    ***REMOVED*** else ***REMOVED***
      if (this._prompt && this._hadControl &&
          c.indexOf(this._prompt) === 0) ***REMOVED***
        this._hadControl = false
        this.emit('data', this._prompt)
        c = c.substr(this._prompt.length)
      ***REMOVED***
      c = c.toString().replace(/./g, this.replace)
    ***REMOVED***
  ***REMOVED***
  this.emit('data', c)
***REMOVED***

MuteStream.prototype.end = function (c) ***REMOVED***
  if (this.muted) ***REMOVED***
    if (c && this.replace) ***REMOVED***
      c = c.toString().replace(/./g, this.replace)
    ***REMOVED*** else ***REMOVED***
      c = null
    ***REMOVED***
  ***REMOVED***
  if (c) this.emit('data', c)
  this.emit('end')
***REMOVED***

function proxy (fn) ***REMOVED*** return function () ***REMOVED***
  var d = this._dest
  var s = this._src
  if (d && d[fn]) d[fn].apply(d, arguments)
  if (s && s[fn]) s[fn].apply(s, arguments)
***REMOVED******REMOVED***

MuteStream.prototype.destroy = proxy('destroy')
MuteStream.prototype.destroySoon = proxy('destroySoon')
MuteStream.prototype.close = proxy('close')
