// eslint-scope/lib/referencer.js
/*
  Copyright (C) 2015 Yusuke Suzuki <utatane.tea@gmail.com>

  Redistribution and use in source and binary forms, with or without
  modification, are permitted provided that the following conditions are met:

    * Redistributions of source code must retain the above copyright
      notice, this list of conditions and the following disclaimer.
    * Redistributions in binary form must reproduce the above copyright
      notice, this list of conditions and the following disclaimer in the
      documentation and/or other materials provided with the distribution.

  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
  ARE DISCLAIMED. IN NO EVENT SHALL <COPYRIGHT HOLDER> BE LIABLE FOR ANY
  DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
  (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
  ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
  (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
  THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
*/
"use strict";

/* eslint-disable no-underscore-dangle */
/* eslint-disable no-undefined */

const Syntax = require("estraverse").Syntax;
const esrecurse = require("esrecurse");
const Reference = require("./reference");
const Variable = require("./variable");
const PatternVisitor = require("./pattern-visitor");
const definition = require("./definition");
const assert = require("assert");

const ParameterDefinition = definition.ParameterDefinition;
const Definition = definition.Definition;

/**
 * Traverse identifier in pattern
 * @param ***REMOVED***Object***REMOVED*** options - options
 * @param ***REMOVED***pattern***REMOVED*** rootPattern - root pattern
 * @param ***REMOVED***Refencer***REMOVED*** referencer - referencer
 * @param ***REMOVED***callback***REMOVED*** callback - callback
 * @returns ***REMOVED***void***REMOVED***
 */
function traverseIdentifierInPattern(options, rootPattern, referencer, callback) ***REMOVED***

    // Call the callback at left hand identifier nodes, and Collect right hand nodes.
    const visitor = new PatternVisitor(options, rootPattern, callback);

    visitor.visit(rootPattern);

    // Process the right hand nodes recursively.
    if (referencer !== null && referencer !== undefined) ***REMOVED***
        visitor.rightHandNodes.forEach(referencer.visit, referencer);
    ***REMOVED***
***REMOVED***

// Importing ImportDeclaration.
// http://people.mozilla.org/~jorendorff/es6-draft.html#sec-moduledeclarationinstantiation
// https://github.com/estree/estree/blob/master/es6.md#importdeclaration
// FIXME: Now, we don't create module environment, because the context is
// implementation dependent.

class Importer extends esrecurse.Visitor ***REMOVED***
    constructor(declaration, referencer) ***REMOVED***
        super(null, referencer.options);
        this.declaration = declaration;
        this.referencer = referencer;
    ***REMOVED***

    visitImport(id, specifier) ***REMOVED***
        this.referencer.visitPattern(id, pattern => ***REMOVED***
            this.referencer.currentScope().__define(pattern,
                new Definition(
                    Variable.ImportBinding,
                    pattern,
                    specifier,
                    this.declaration,
                    null,
                    null
                ));
        ***REMOVED***);
    ***REMOVED***

    ImportNamespaceSpecifier(node) ***REMOVED***
        const local = (node.local || node.id);

        if (local) ***REMOVED***
            this.visitImport(local, node);
        ***REMOVED***
    ***REMOVED***

    ImportDefaultSpecifier(node) ***REMOVED***
        const local = (node.local || node.id);

        this.visitImport(local, node);
    ***REMOVED***

    ImportSpecifier(node) ***REMOVED***
        const local = (node.local || node.id);

        if (node.name) ***REMOVED***
            this.visitImport(node.name, node);
        ***REMOVED*** else ***REMOVED***
            this.visitImport(local, node);
        ***REMOVED***
    ***REMOVED***
***REMOVED***

// Referencing variables and creating bindings.
class Referencer extends esrecurse.Visitor ***REMOVED***
    constructor(options, scopeManager) ***REMOVED***
        super(null, options);
        this.options = options;
        this.scopeManager = scopeManager;
        this.parent = null;
        this.isInnerMethodDefinition = false;
    ***REMOVED***

    currentScope() ***REMOVED***
        return this.scopeManager.__currentScope;
    ***REMOVED***

    close(node) ***REMOVED***
        while (this.currentScope() && node === this.currentScope().block) ***REMOVED***
            this.scopeManager.__currentScope = this.currentScope().__close(this.scopeManager);
        ***REMOVED***
    ***REMOVED***

    pushInnerMethodDefinition(isInnerMethodDefinition) ***REMOVED***
        const previous = this.isInnerMethodDefinition;

        this.isInnerMethodDefinition = isInnerMethodDefinition;
        return previous;
    ***REMOVED***

    popInnerMethodDefinition(isInnerMethodDefinition) ***REMOVED***
        this.isInnerMethodDefinition = isInnerMethodDefinition;
    ***REMOVED***

    referencingDefaultValue(pattern, assignments, maybeImplicitGlobal, init) ***REMOVED***
        const scope = this.currentScope();

        assignments.forEach(assignment => ***REMOVED***
            scope.__referencing(
                pattern,
                Reference.WRITE,
                assignment.right,
                maybeImplicitGlobal,
                pattern !== assignment.left,
                init
            );
        ***REMOVED***);
    ***REMOVED***

    visitPattern(node, options, callback) ***REMOVED***
        let visitPatternOptions = options;
        let visitPatternCallback = callback;

        if (typeof options === "function") ***REMOVED***
            visitPatternCallback = options;
            visitPatternOptions = ***REMOVED*** processRightHandNodes: false ***REMOVED***;
        ***REMOVED***

        traverseIdentifierInPattern(
            this.options,
            node,
            visitPatternOptions.processRightHandNodes ? this : null,
            visitPatternCallback
        );
    ***REMOVED***

    visitFunction(node) ***REMOVED***
        let i, iz;

        // FunctionDeclaration name is defined in upper scope
        // NOTE: Not referring variableScope. It is intended.
        // Since
        //  in ES5, FunctionDeclaration should be in FunctionBody.
        //  in ES6, FunctionDeclaration should be block scoped.

        if (node.type === Syntax.FunctionDeclaration) ***REMOVED***

            // id is defined in upper scope
            this.currentScope().__define(node.id,
                new Definition(
                    Variable.FunctionName,
                    node.id,
                    node,
                    null,
                    null,
                    null
                ));
        ***REMOVED***

        // FunctionExpression with name creates its special scope;
        // FunctionExpressionNameScope.
        if (node.type === Syntax.FunctionExpression && node.id) ***REMOVED***
            this.scopeManager.__nestFunctionExpressionNameScope(node);
        ***REMOVED***

        // Consider this function is in the MethodDefinition.
        this.scopeManager.__nestFunctionScope(node, this.isInnerMethodDefinition);

        const that = this;

        /**
         * Visit pattern callback
         * @param ***REMOVED***pattern***REMOVED*** pattern - pattern
         * @param ***REMOVED***Object***REMOVED*** info - info
         * @returns ***REMOVED***void***REMOVED***
         */
        function visitPatternCallback(pattern, info) ***REMOVED***
            that.currentScope().__define(pattern,
                new ParameterDefinition(
                    pattern,
                    node,
                    i,
                    info.rest
                ));

            that.referencingDefaultValue(pattern, info.assignments, null, true);
        ***REMOVED***

        // Process parameter declarations.
        for (i = 0, iz = node.params.length; i < iz; ++i) ***REMOVED***
            this.visitPattern(node.params[i], ***REMOVED*** processRightHandNodes: true ***REMOVED***, visitPatternCallback);
        ***REMOVED***

        // if there's a rest argument, add that
        if (node.rest) ***REMOVED***
            this.visitPattern(***REMOVED***
                type: "RestElement",
                argument: node.rest
            ***REMOVED***, pattern => ***REMOVED***
                this.currentScope().__define(pattern,
                    new ParameterDefinition(
                        pattern,
                        node,
                        node.params.length,
                        true
                    ));
            ***REMOVED***);
        ***REMOVED***

        // In TypeScript there are a number of function-like constructs which have no body,
        // so check it exists before traversing
        if (node.body) ***REMOVED***

            // Skip BlockStatement to prevent creating BlockStatement scope.
            if (node.body.type === Syntax.BlockStatement) ***REMOVED***
                this.visitChildren(node.body);
            ***REMOVED*** else ***REMOVED***
                this.visit(node.body);
            ***REMOVED***
        ***REMOVED***

        this.close(node);
    ***REMOVED***

    visitClass(node) ***REMOVED***
        if (node.type === Syntax.ClassDeclaration) ***REMOVED***
            this.currentScope().__define(node.id,
                new Definition(
                    Variable.ClassName,
                    node.id,
                    node,
                    null,
                    null,
                    null
                ));
        ***REMOVED***

        this.visit(node.superClass);

        this.scopeManager.__nestClassScope(node);

        if (node.id) ***REMOVED***
            this.currentScope().__define(node.id,
                new Definition(
                    Variable.ClassName,
                    node.id,
                    node
                ));
        ***REMOVED***
        this.visit(node.body);

        this.close(node);
    ***REMOVED***

    visitProperty(node) ***REMOVED***
        let previous;

        if (node.computed) ***REMOVED***
            this.visit(node.key);
        ***REMOVED***

        const isMethodDefinition = node.type === Syntax.MethodDefinition;

        if (isMethodDefinition) ***REMOVED***
            previous = this.pushInnerMethodDefinition(true);
        ***REMOVED***
        this.visit(node.value);
        if (isMethodDefinition) ***REMOVED***
            this.popInnerMethodDefinition(previous);
        ***REMOVED***
    ***REMOVED***

    visitForIn(node) ***REMOVED***
        if (node.left.type === Syntax.VariableDeclaration && node.left.kind !== "var") ***REMOVED***
            this.scopeManager.__nestForScope(node);
        ***REMOVED***

        if (node.left.type === Syntax.VariableDeclaration) ***REMOVED***
            this.visit(node.left);
            this.visitPattern(node.left.declarations[0].id, pattern => ***REMOVED***
                this.currentScope().__referencing(pattern, Reference.WRITE, node.right, null, true, true);
            ***REMOVED***);
        ***REMOVED*** else ***REMOVED***
            this.visitPattern(node.left, ***REMOVED*** processRightHandNodes: true ***REMOVED***, (pattern, info) => ***REMOVED***
                let maybeImplicitGlobal = null;

                if (!this.currentScope().isStrict) ***REMOVED***
                    maybeImplicitGlobal = ***REMOVED***
                        pattern,
                        node
                    ***REMOVED***;
                ***REMOVED***
                this.referencingDefaultValue(pattern, info.assignments, maybeImplicitGlobal, false);
                this.currentScope().__referencing(pattern, Reference.WRITE, node.right, maybeImplicitGlobal, true, false);
            ***REMOVED***);
        ***REMOVED***
        this.visit(node.right);
        this.visit(node.body);

        this.close(node);
    ***REMOVED***

    visitVariableDeclaration(variableTargetScope, type, node, index) ***REMOVED***

        const decl = node.declarations[index];
        const init = decl.init;

        this.visitPattern(decl.id, ***REMOVED*** processRightHandNodes: true ***REMOVED***, (pattern, info) => ***REMOVED***
            variableTargetScope.__define(
                pattern,
                new Definition(
                    type,
                    pattern,
                    decl,
                    node,
                    index,
                    node.kind
                )
            );

            this.referencingDefaultValue(pattern, info.assignments, null, true);
            if (init) ***REMOVED***
                this.currentScope().__referencing(pattern, Reference.WRITE, init, null, !info.topLevel, true);
            ***REMOVED***
        ***REMOVED***);
    ***REMOVED***

    AssignmentExpression(node) ***REMOVED***
        if (PatternVisitor.isPattern(node.left)) ***REMOVED***
            if (node.operator === "=") ***REMOVED***
                this.visitPattern(node.left, ***REMOVED*** processRightHandNodes: true ***REMOVED***, (pattern, info) => ***REMOVED***
                    let maybeImplicitGlobal = null;

                    if (!this.currentScope().isStrict) ***REMOVED***
                        maybeImplicitGlobal = ***REMOVED***
                            pattern,
                            node
                        ***REMOVED***;
                    ***REMOVED***
                    this.referencingDefaultValue(pattern, info.assignments, maybeImplicitGlobal, false);
                    this.currentScope().__referencing(pattern, Reference.WRITE, node.right, maybeImplicitGlobal, !info.topLevel, false);
                ***REMOVED***);
            ***REMOVED*** else ***REMOVED***
                this.currentScope().__referencing(node.left, Reference.RW, node.right);
            ***REMOVED***
        ***REMOVED*** else ***REMOVED***
            this.visit(node.left);
        ***REMOVED***
        this.visit(node.right);
    ***REMOVED***

    CatchClause(node) ***REMOVED***
        this.scopeManager.__nestCatchScope(node);

        this.visitPattern(node.param, ***REMOVED*** processRightHandNodes: true ***REMOVED***, (pattern, info) => ***REMOVED***
            this.currentScope().__define(pattern,
                new Definition(
                    Variable.CatchClause,
                    node.param,
                    node,
                    null,
                    null,
                    null
                ));
            this.referencingDefaultValue(pattern, info.assignments, null, true);
        ***REMOVED***);
        this.visit(node.body);

        this.close(node);
    ***REMOVED***

    Program(node) ***REMOVED***
        this.scopeManager.__nestGlobalScope(node);

        if (this.scopeManager.__isNodejsScope()) ***REMOVED***

            // Force strictness of GlobalScope to false when using node.js scope.
            this.currentScope().isStrict = false;
            this.scopeManager.__nestFunctionScope(node, false);
        ***REMOVED***

        if (this.scopeManager.__isES6() && this.scopeManager.isModule()) ***REMOVED***
            this.scopeManager.__nestModuleScope(node);
        ***REMOVED***

        if (this.scopeManager.isStrictModeSupported() && this.scopeManager.isImpliedStrict()) ***REMOVED***
            this.currentScope().isStrict = true;
        ***REMOVED***

        this.visitChildren(node);
        this.close(node);
    ***REMOVED***

    Identifier(node) ***REMOVED***
        this.currentScope().__referencing(node);
    ***REMOVED***

    UpdateExpression(node) ***REMOVED***
        if (PatternVisitor.isPattern(node.argument)) ***REMOVED***
            this.currentScope().__referencing(node.argument, Reference.RW, null);
        ***REMOVED*** else ***REMOVED***
            this.visitChildren(node);
        ***REMOVED***
    ***REMOVED***

    MemberExpression(node) ***REMOVED***
        this.visit(node.object);
        if (node.computed) ***REMOVED***
            this.visit(node.property);
        ***REMOVED***
    ***REMOVED***

    Property(node) ***REMOVED***
        this.visitProperty(node);
    ***REMOVED***

    MethodDefinition(node) ***REMOVED***
        this.visitProperty(node);
    ***REMOVED***

    BreakStatement() ***REMOVED******REMOVED*** // eslint-disable-line class-methods-use-this

    ContinueStatement() ***REMOVED******REMOVED*** // eslint-disable-line class-methods-use-this

    LabeledStatement(node) ***REMOVED***
        this.visit(node.body);
    ***REMOVED***

    ForStatement(node) ***REMOVED***

        // Create ForStatement declaration.
        // NOTE: In ES6, ForStatement dynamically generates
        // per iteration environment. However, escope is
        // a static analyzer, we only generate one scope for ForStatement.
        if (node.init && node.init.type === Syntax.VariableDeclaration && node.init.kind !== "var") ***REMOVED***
            this.scopeManager.__nestForScope(node);
        ***REMOVED***

        this.visitChildren(node);

        this.close(node);
    ***REMOVED***

    ClassExpression(node) ***REMOVED***
        this.visitClass(node);
    ***REMOVED***

    ClassDeclaration(node) ***REMOVED***
        this.visitClass(node);
    ***REMOVED***

    CallExpression(node) ***REMOVED***

        // Check this is direct call to eval
        if (!this.scopeManager.__ignoreEval() && node.callee.type === Syntax.Identifier && node.callee.name === "eval") ***REMOVED***

            // NOTE: This should be `variableScope`. Since direct eval call always creates Lexical environment and
            // let / const should be enclosed into it. Only VariableDeclaration affects on the caller's environment.
            this.currentScope().variableScope.__detectEval();
        ***REMOVED***
        this.visitChildren(node);
    ***REMOVED***

    BlockStatement(node) ***REMOVED***
        if (this.scopeManager.__isES6()) ***REMOVED***
            this.scopeManager.__nestBlockScope(node);
        ***REMOVED***

        this.visitChildren(node);

        this.close(node);
    ***REMOVED***

    ThisExpression() ***REMOVED***
        this.currentScope().variableScope.__detectThis();
    ***REMOVED***

    WithStatement(node) ***REMOVED***
        this.visit(node.object);

        // Then nest scope for WithStatement.
        this.scopeManager.__nestWithScope(node);

        this.visit(node.body);

        this.close(node);
    ***REMOVED***

    VariableDeclaration(node) ***REMOVED***
        const variableTargetScope = (node.kind === "var") ? this.currentScope().variableScope : this.currentScope();

        for (let i = 0, iz = node.declarations.length; i < iz; ++i) ***REMOVED***
            const decl = node.declarations[i];

            this.visitVariableDeclaration(variableTargetScope, Variable.Variable, node, i);
            if (decl.init) ***REMOVED***
                this.visit(decl.init);
            ***REMOVED***
        ***REMOVED***
    ***REMOVED***

    // sec 13.11.8
    SwitchStatement(node) ***REMOVED***
        this.visit(node.discriminant);

        if (this.scopeManager.__isES6()) ***REMOVED***
            this.scopeManager.__nestSwitchScope(node);
        ***REMOVED***

        for (let i = 0, iz = node.cases.length; i < iz; ++i) ***REMOVED***
            this.visit(node.cases[i]);
        ***REMOVED***

        this.close(node);
    ***REMOVED***

    FunctionDeclaration(node) ***REMOVED***
        this.visitFunction(node);
    ***REMOVED***

    FunctionExpression(node) ***REMOVED***
        this.visitFunction(node);
    ***REMOVED***

    ForOfStatement(node) ***REMOVED***
        this.visitForIn(node);
    ***REMOVED***

    ForInStatement(node) ***REMOVED***
        this.visitForIn(node);
    ***REMOVED***

    ArrowFunctionExpression(node) ***REMOVED***
        this.visitFunction(node);
    ***REMOVED***

    ImportDeclaration(node) ***REMOVED***
        assert(this.scopeManager.__isES6() && this.scopeManager.isModule(), "ImportDeclaration should appear when the mode is ES6 and in the module context.");

        const importer = new Importer(node, this);

        importer.visit(node);
    ***REMOVED***

    visitExportDeclaration(node) ***REMOVED***
        if (node.source) ***REMOVED***
            return;
        ***REMOVED***
        if (node.declaration) ***REMOVED***
            this.visit(node.declaration);
            return;
        ***REMOVED***

        this.visitChildren(node);
    ***REMOVED***

    ExportDeclaration(node) ***REMOVED***
        this.visitExportDeclaration(node);
    ***REMOVED***

    ExportNamedDeclaration(node) ***REMOVED***
        this.visitExportDeclaration(node);
    ***REMOVED***

    ExportSpecifier(node) ***REMOVED***
        const local = (node.id || node.local);

        this.visit(local);
    ***REMOVED***

    MetaProperty() ***REMOVED*** // eslint-disable-line class-methods-use-this

        // do nothing.
    ***REMOVED***
***REMOVED***

module.exports = Referencer;

/* vim: set sw=4 ts=4 et tw=80 : */
