// eslint-scope/lib/scope.js
/*
  Copyright (C) 2015 Yusuke Suzuki <utatane.tea@gmail.com>

  Redistribution and use in source and binary forms, with or without
  modification, are permitted provided that the following conditions are met:

    * Redistributions of source code must retain the above copyright
      notice, this list of conditions and the following disclaimer.
    * Redistributions in binary form must reproduce the above copyright
      notice, this list of conditions and the following disclaimer in the
      documentation and/or other materials provided with the distribution.

  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
  ARE DISCLAIMED. IN NO EVENT SHALL <COPYRIGHT HOLDER> BE LIABLE FOR ANY
  DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
  (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
  ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
  (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
  THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
*/
"use strict";

/* eslint-disable no-underscore-dangle */
/* eslint-disable no-undefined */

const Syntax = require("estraverse").Syntax;

const Reference = require("./reference");
const Variable = require("./variable");
const Definition = require("./definition").Definition;
const assert = require("assert");

/**
 * Test if scope is struct
 * @param ***REMOVED***Scope***REMOVED*** scope - scope
 * @param ***REMOVED***Block***REMOVED*** block - block
 * @param ***REMOVED***boolean***REMOVED*** isMethodDefinition - is method definiton
 * @param ***REMOVED***boolean***REMOVED*** useDirective - use directive
 * @returns ***REMOVED***boolean***REMOVED*** is strict scope
 */
function isStrictScope(scope, block, isMethodDefinition, useDirective) ***REMOVED***
    let body;

    // When upper scope is exists and strict, inner scope is also strict.
    if (scope.upper && scope.upper.isStrict) ***REMOVED***
        return true;
    ***REMOVED***

    if (isMethodDefinition) ***REMOVED***
        return true;
    ***REMOVED***

    if (scope.type === "class" || scope.type === "module") ***REMOVED***
        return true;
    ***REMOVED***

    if (scope.type === "block" || scope.type === "switch") ***REMOVED***
        return false;
    ***REMOVED***

    if (scope.type === "function") ***REMOVED***
        if (block.type === Syntax.ArrowFunctionExpression && block.body.type !== Syntax.BlockStatement) ***REMOVED***
            return false;
        ***REMOVED***

        if (block.type === Syntax.Program) ***REMOVED***
            body = block;
        ***REMOVED*** else ***REMOVED***
            body = block.body;
        ***REMOVED***

        if (!body) ***REMOVED***
            return false;
        ***REMOVED***
    ***REMOVED*** else if (scope.type === "global") ***REMOVED***
        body = block;
    ***REMOVED*** else ***REMOVED***
        return false;
    ***REMOVED***

    // Search 'use strict' directive.
    if (useDirective) ***REMOVED***
        for (let i = 0, iz = body.body.length; i < iz; ++i) ***REMOVED***
            const stmt = body.body[i];

            if (stmt.type !== Syntax.DirectiveStatement) ***REMOVED***
                break;
            ***REMOVED***
            if (stmt.raw === "\"use strict\"" || stmt.raw === "'use strict'") ***REMOVED***
                return true;
            ***REMOVED***
        ***REMOVED***
    ***REMOVED*** else ***REMOVED***
        for (let i = 0, iz = body.body.length; i < iz; ++i) ***REMOVED***
            const stmt = body.body[i];

            if (stmt.type !== Syntax.ExpressionStatement) ***REMOVED***
                break;
            ***REMOVED***
            const expr = stmt.expression;

            if (expr.type !== Syntax.Literal || typeof expr.value !== "string") ***REMOVED***
                break;
            ***REMOVED***
            if (expr.raw !== null && expr.raw !== undefined) ***REMOVED***
                if (expr.raw === "\"use strict\"" || expr.raw === "'use strict'") ***REMOVED***
                    return true;
                ***REMOVED***
            ***REMOVED*** else ***REMOVED***
                if (expr.value === "use strict") ***REMOVED***
                    return true;
                ***REMOVED***
            ***REMOVED***
        ***REMOVED***
    ***REMOVED***
    return false;
***REMOVED***

/**
 * Register scope
 * @param ***REMOVED***ScopeManager***REMOVED*** scopeManager - scope manager
 * @param ***REMOVED***Scope***REMOVED*** scope - scope
 * @returns ***REMOVED***void***REMOVED***
 */
function registerScope(scopeManager, scope) ***REMOVED***
    scopeManager.scopes.push(scope);

    const scopes = scopeManager.__nodeToScope.get(scope.block);

    if (scopes) ***REMOVED***
        scopes.push(scope);
    ***REMOVED*** else ***REMOVED***
        scopeManager.__nodeToScope.set(scope.block, [scope]);
    ***REMOVED***
***REMOVED***

/**
 * Should be statically
 * @param ***REMOVED***Object***REMOVED*** def - def
 * @returns ***REMOVED***boolean***REMOVED*** should be statically
 */
function shouldBeStatically(def) ***REMOVED***
    return (
        (def.type === Variable.ClassName) ||
        (def.type === Variable.Variable && def.parent.kind !== "var")
    );
***REMOVED***

/**
 * @class Scope
 */
class Scope ***REMOVED***
    constructor(scopeManager, type, upperScope, block, isMethodDefinition) ***REMOVED***

        /**
         * One of 'module', 'block', 'switch', 'function', 'catch', 'with', 'function', 'class', 'global'.
         * @member ***REMOVED***String***REMOVED*** Scope#type
         */
        this.type = type;

        /**
         * The scoped ***REMOVED***@link Variable***REMOVED***s of this scope, as <code>***REMOVED*** Variable.name
         * : Variable ***REMOVED***</code>.
         * @member ***REMOVED***Map***REMOVED*** Scope#set
         */
        this.set = new Map();

        /**
         * The tainted variables of this scope, as <code>***REMOVED*** Variable.name :
         * boolean ***REMOVED***</code>.
         * @member ***REMOVED***Map***REMOVED*** Scope#taints */
        this.taints = new Map();

        /**
         * Generally, through the lexical scoping of JS you can always know
         * which variable an identifier in the source code refers to. There are
         * a few exceptions to this rule. With 'global' and 'with' scopes you
         * can only decide at runtime which variable a reference refers to.
         * Moreover, if 'eval()' is used in a scope, it might introduce new
         * bindings in this or its parent scopes.
         * All those scopes are considered 'dynamic'.
         * @member ***REMOVED***boolean***REMOVED*** Scope#dynamic
         */
        this.dynamic = this.type === "global" || this.type === "with";

        /**
         * A reference to the scope-defining syntax node.
         * @member ***REMOVED***espree.Node***REMOVED*** Scope#block
         */
        this.block = block;

        /**
         * The ***REMOVED***@link Reference|references***REMOVED*** that are not resolved with this scope.
         * @member ***REMOVED***Reference[]***REMOVED*** Scope#through
         */
        this.through = [];

        /**
         * The scoped ***REMOVED***@link Variable***REMOVED***s of this scope. In the case of a
         * 'function' scope this includes the automatic argument <em>arguments</em> as
         * its first element, as well as all further formal arguments.
         * @member ***REMOVED***Variable[]***REMOVED*** Scope#variables
         */
        this.variables = [];

        /**
         * Any variable ***REMOVED***@link Reference|reference***REMOVED*** found in this scope. This
         * includes occurrences of local variables as well as variables from
         * parent scopes (including the global scope). For local variables
         * this also includes defining occurrences (like in a 'var' statement).
         * In a 'function' scope this does not include the occurrences of the
         * formal parameter in the parameter list.
         * @member ***REMOVED***Reference[]***REMOVED*** Scope#references
         */
        this.references = [];

        /**
         * For 'global' and 'function' scopes, this is a self-reference. For
         * other scope types this is the <em>variableScope</em> value of the
         * parent scope.
         * @member ***REMOVED***Scope***REMOVED*** Scope#variableScope
         */
        this.variableScope =
            (this.type === "global" || this.type === "function" || this.type === "module") ? this : upperScope.variableScope;

        /**
         * Whether this scope is created by a FunctionExpression.
         * @member ***REMOVED***boolean***REMOVED*** Scope#functionExpressionScope
         */
        this.functionExpressionScope = false;

        /**
         * Whether this is a scope that contains an 'eval()' invocation.
         * @member ***REMOVED***boolean***REMOVED*** Scope#directCallToEvalScope
         */
        this.directCallToEvalScope = false;

        /**
         * @member ***REMOVED***boolean***REMOVED*** Scope#thisFound
         */
        this.thisFound = false;

        this.__left = [];

        /**
         * Reference to the parent ***REMOVED***@link Scope|scope***REMOVED***.
         * @member ***REMOVED***Scope***REMOVED*** Scope#upper
         */
        this.upper = upperScope;

        /**
         * Whether 'use strict' is in effect in this scope.
         * @member ***REMOVED***boolean***REMOVED*** Scope#isStrict
         */
        this.isStrict = isStrictScope(this, block, isMethodDefinition, scopeManager.__useDirective());

        /**
         * List of nested ***REMOVED***@link Scope***REMOVED***s.
         * @member ***REMOVED***Scope[]***REMOVED*** Scope#childScopes
         */
        this.childScopes = [];
        if (this.upper) ***REMOVED***
            this.upper.childScopes.push(this);
        ***REMOVED***

        this.__declaredVariables = scopeManager.__declaredVariables;

        registerScope(scopeManager, this);
    ***REMOVED***

    __shouldStaticallyClose(scopeManager) ***REMOVED***
        return (!this.dynamic || scopeManager.__isOptimistic());
    ***REMOVED***

    __shouldStaticallyCloseForGlobal(ref) ***REMOVED***

        // On global scope, let/const/class declarations should be resolved statically.
        const name = ref.identifier.name;

        if (!this.set.has(name)) ***REMOVED***
            return false;
        ***REMOVED***

        const variable = this.set.get(name);
        const defs = variable.defs;

        return defs.length > 0 && defs.every(shouldBeStatically);
    ***REMOVED***

    __staticCloseRef(ref) ***REMOVED***
        if (!this.__resolve(ref)) ***REMOVED***
            this.__delegateToUpperScope(ref);
        ***REMOVED***
    ***REMOVED***

    __dynamicCloseRef(ref) ***REMOVED***

        // notify all names are through to global
        let current = this;

        do ***REMOVED***
            current.through.push(ref);
            current = current.upper;
        ***REMOVED*** while (current);
    ***REMOVED***

    __globalCloseRef(ref) ***REMOVED***

        // let/const/class declarations should be resolved statically.
        // others should be resolved dynamically.
        if (this.__shouldStaticallyCloseForGlobal(ref)) ***REMOVED***
            this.__staticCloseRef(ref);
        ***REMOVED*** else ***REMOVED***
            this.__dynamicCloseRef(ref);
        ***REMOVED***
    ***REMOVED***

    __close(scopeManager) ***REMOVED***
        let closeRef;

        if (this.__shouldStaticallyClose(scopeManager)) ***REMOVED***
            closeRef = this.__staticCloseRef;
        ***REMOVED*** else if (this.type !== "global") ***REMOVED***
            closeRef = this.__dynamicCloseRef;
        ***REMOVED*** else ***REMOVED***
            closeRef = this.__globalCloseRef;
        ***REMOVED***

        // Try Resolving all references in this scope.
        for (let i = 0, iz = this.__left.length; i < iz; ++i) ***REMOVED***
            const ref = this.__left[i];

            closeRef.call(this, ref);
        ***REMOVED***
        this.__left = null;

        return this.upper;
    ***REMOVED***

    // To override by function scopes.
    // References in default parameters isn't resolved to variables which are in their function body.
    __isValidResolution(ref, variable) ***REMOVED*** // eslint-disable-line class-methods-use-this, no-unused-vars
        return true;
    ***REMOVED***

    __resolve(ref) ***REMOVED***
        const name = ref.identifier.name;

        if (!this.set.has(name)) ***REMOVED***
            return false;
        ***REMOVED***
        const variable = this.set.get(name);

        if (!this.__isValidResolution(ref, variable)) ***REMOVED***
            return false;
        ***REMOVED***
        variable.references.push(ref);
        variable.stack = variable.stack && ref.from.variableScope === this.variableScope;
        if (ref.tainted) ***REMOVED***
            variable.tainted = true;
            this.taints.set(variable.name, true);
        ***REMOVED***
        ref.resolved = variable;

        return true;
    ***REMOVED***

    __delegateToUpperScope(ref) ***REMOVED***
        if (this.upper) ***REMOVED***
            this.upper.__left.push(ref);
        ***REMOVED***
        this.through.push(ref);
    ***REMOVED***

    __addDeclaredVariablesOfNode(variable, node) ***REMOVED***
        if (node === null || node === undefined) ***REMOVED***
            return;
        ***REMOVED***

        let variables = this.__declaredVariables.get(node);

        if (variables === null || variables === undefined) ***REMOVED***
            variables = [];
            this.__declaredVariables.set(node, variables);
        ***REMOVED***
        if (variables.indexOf(variable) === -1) ***REMOVED***
            variables.push(variable);
        ***REMOVED***
    ***REMOVED***

    __defineGeneric(name, set, variables, node, def) ***REMOVED***
        let variable;

        variable = set.get(name);
        if (!variable) ***REMOVED***
            variable = new Variable(name, this);
            set.set(name, variable);
            variables.push(variable);
        ***REMOVED***

        if (def) ***REMOVED***
            variable.defs.push(def);
            this.__addDeclaredVariablesOfNode(variable, def.node);
            this.__addDeclaredVariablesOfNode(variable, def.parent);
        ***REMOVED***
        if (node) ***REMOVED***
            variable.identifiers.push(node);
        ***REMOVED***
    ***REMOVED***

    __define(node, def) ***REMOVED***
        if (node && node.type === Syntax.Identifier) ***REMOVED***
            this.__defineGeneric(
                node.name,
                this.set,
                this.variables,
                node,
                def
            );
        ***REMOVED***
    ***REMOVED***

    __referencing(node, assign, writeExpr, maybeImplicitGlobal, partial, init) ***REMOVED***

        // because Array element may be null
        if (!node || node.type !== Syntax.Identifier) ***REMOVED***
            return;
        ***REMOVED***

        // Specially handle like `this`.
        if (node.name === "super") ***REMOVED***
            return;
        ***REMOVED***

        const ref = new Reference(node, this, assign || Reference.READ, writeExpr, maybeImplicitGlobal, !!partial, !!init);

        this.references.push(ref);
        this.__left.push(ref);
    ***REMOVED***

    __detectEval() ***REMOVED***
        let current = this;

        this.directCallToEvalScope = true;
        do ***REMOVED***
            current.dynamic = true;
            current = current.upper;
        ***REMOVED*** while (current);
    ***REMOVED***

    __detectThis() ***REMOVED***
        this.thisFound = true;
    ***REMOVED***

    __isClosed() ***REMOVED***
        return this.__left === null;
    ***REMOVED***

    /**
     * returns resolved ***REMOVED***Reference***REMOVED***
     * @method Scope#resolve
     * @param ***REMOVED***Espree.Identifier***REMOVED*** ident - identifier to be resolved.
     * @returns ***REMOVED***Reference***REMOVED*** reference
     */
    resolve(ident) ***REMOVED***
        let ref, i, iz;

        assert(this.__isClosed(), "Scope should be closed.");
        assert(ident.type === Syntax.Identifier, "Target should be identifier.");
        for (i = 0, iz = this.references.length; i < iz; ++i) ***REMOVED***
            ref = this.references[i];
            if (ref.identifier === ident) ***REMOVED***
                return ref;
            ***REMOVED***
        ***REMOVED***
        return null;
    ***REMOVED***

    /**
     * returns this scope is static
     * @method Scope#isStatic
     * @returns ***REMOVED***boolean***REMOVED*** static
     */
    isStatic() ***REMOVED***
        return !this.dynamic;
    ***REMOVED***

    /**
     * returns this scope has materialized arguments
     * @method Scope#isArgumentsMaterialized
     * @returns ***REMOVED***boolean***REMOVED*** arguemnts materialized
     */
    isArgumentsMaterialized() ***REMOVED*** // eslint-disable-line class-methods-use-this
        return true;
    ***REMOVED***

    /**
     * returns this scope has materialized `this` reference
     * @method Scope#isThisMaterialized
     * @returns ***REMOVED***boolean***REMOVED*** this materialized
     */
    isThisMaterialized() ***REMOVED*** // eslint-disable-line class-methods-use-this
        return true;
    ***REMOVED***

    isUsedName(name) ***REMOVED***
        if (this.set.has(name)) ***REMOVED***
            return true;
        ***REMOVED***
        for (let i = 0, iz = this.through.length; i < iz; ++i) ***REMOVED***
            if (this.through[i].identifier.name === name) ***REMOVED***
                return true;
            ***REMOVED***
        ***REMOVED***
        return false;
    ***REMOVED***
***REMOVED***

class GlobalScope extends Scope ***REMOVED***
    constructor(scopeManager, block) ***REMOVED***
        super(scopeManager, "global", null, block, false);
        this.implicit = ***REMOVED***
            set: new Map(),
            variables: [],

            /**
            * List of ***REMOVED***@link Reference***REMOVED***s that are left to be resolved (i.e. which
            * need to be linked to the variable they refer to).
            * @member ***REMOVED***Reference[]***REMOVED*** Scope#implicit#left
            */
            left: []
        ***REMOVED***;
    ***REMOVED***

    __close(scopeManager) ***REMOVED***
        const implicit = [];

        for (let i = 0, iz = this.__left.length; i < iz; ++i) ***REMOVED***
            const ref = this.__left[i];

            if (ref.__maybeImplicitGlobal && !this.set.has(ref.identifier.name)) ***REMOVED***
                implicit.push(ref.__maybeImplicitGlobal);
            ***REMOVED***
        ***REMOVED***

        // create an implicit global variable from assignment expression
        for (let i = 0, iz = implicit.length; i < iz; ++i) ***REMOVED***
            const info = implicit[i];

            this.__defineImplicit(info.pattern,
                new Definition(
                    Variable.ImplicitGlobalVariable,
                    info.pattern,
                    info.node,
                    null,
                    null,
                    null
                ));

        ***REMOVED***

        this.implicit.left = this.__left;

        return super.__close(scopeManager);
    ***REMOVED***

    __defineImplicit(node, def) ***REMOVED***
        if (node && node.type === Syntax.Identifier) ***REMOVED***
            this.__defineGeneric(
                node.name,
                this.implicit.set,
                this.implicit.variables,
                node,
                def
            );
        ***REMOVED***
    ***REMOVED***
***REMOVED***

class ModuleScope extends Scope ***REMOVED***
    constructor(scopeManager, upperScope, block) ***REMOVED***
        super(scopeManager, "module", upperScope, block, false);
    ***REMOVED***
***REMOVED***

class FunctionExpressionNameScope extends Scope ***REMOVED***
    constructor(scopeManager, upperScope, block) ***REMOVED***
        super(scopeManager, "function-expression-name", upperScope, block, false);
        this.__define(block.id,
            new Definition(
                Variable.FunctionName,
                block.id,
                block,
                null,
                null,
                null
            ));
        this.functionExpressionScope = true;
    ***REMOVED***
***REMOVED***

class CatchScope extends Scope ***REMOVED***
    constructor(scopeManager, upperScope, block) ***REMOVED***
        super(scopeManager, "catch", upperScope, block, false);
    ***REMOVED***
***REMOVED***

class WithScope extends Scope ***REMOVED***
    constructor(scopeManager, upperScope, block) ***REMOVED***
        super(scopeManager, "with", upperScope, block, false);
    ***REMOVED***

    __close(scopeManager) ***REMOVED***
        if (this.__shouldStaticallyClose(scopeManager)) ***REMOVED***
            return super.__close(scopeManager);
        ***REMOVED***

        for (let i = 0, iz = this.__left.length; i < iz; ++i) ***REMOVED***
            const ref = this.__left[i];

            ref.tainted = true;
            this.__delegateToUpperScope(ref);
        ***REMOVED***
        this.__left = null;

        return this.upper;
    ***REMOVED***
***REMOVED***

class BlockScope extends Scope ***REMOVED***
    constructor(scopeManager, upperScope, block) ***REMOVED***
        super(scopeManager, "block", upperScope, block, false);
    ***REMOVED***
***REMOVED***

class SwitchScope extends Scope ***REMOVED***
    constructor(scopeManager, upperScope, block) ***REMOVED***
        super(scopeManager, "switch", upperScope, block, false);
    ***REMOVED***
***REMOVED***

class FunctionScope extends Scope ***REMOVED***
    constructor(scopeManager, upperScope, block, isMethodDefinition) ***REMOVED***
        super(scopeManager, "function", upperScope, block, isMethodDefinition);

        // section 9.2.13, FunctionDeclarationInstantiation.
        // NOTE Arrow functions never have an arguments objects.
        if (this.block.type !== Syntax.ArrowFunctionExpression) ***REMOVED***
            this.__defineArguments();
        ***REMOVED***
    ***REMOVED***

    isArgumentsMaterialized() ***REMOVED***

        // TODO(Constellation)
        // We can more aggressive on this condition like this.
        //
        // function t() ***REMOVED***
        //     // arguments of t is always hidden.
        //     function arguments() ***REMOVED***
        //     ***REMOVED***
        // ***REMOVED***
        if (this.block.type === Syntax.ArrowFunctionExpression) ***REMOVED***
            return false;
        ***REMOVED***

        if (!this.isStatic()) ***REMOVED***
            return true;
        ***REMOVED***

        const variable = this.set.get("arguments");

        assert(variable, "Always have arguments variable.");
        return variable.tainted || variable.references.length !== 0;
    ***REMOVED***

    isThisMaterialized() ***REMOVED***
        if (!this.isStatic()) ***REMOVED***
            return true;
        ***REMOVED***
        return this.thisFound;
    ***REMOVED***

    __defineArguments() ***REMOVED***
        this.__defineGeneric(
            "arguments",
            this.set,
            this.variables,
            null,
            null
        );
        this.taints.set("arguments", true);
    ***REMOVED***

    // References in default parameters isn't resolved to variables which are in their function body.
    //     const x = 1
    //     function f(a = x) ***REMOVED*** // This `x` is resolved to the `x` in the outer scope.
    //         const x = 2
    //         console.log(a)
    //     ***REMOVED***
    __isValidResolution(ref, variable) ***REMOVED***

        // If `options.nodejsScope` is true, `this.block` becomes a Program node.
        if (this.block.type === "Program") ***REMOVED***
            return true;
        ***REMOVED***

        const bodyStart = this.block.body.range[0];

        // It's invalid resolution in the following case:
        return !(
            variable.scope === this &&
            ref.identifier.range[0] < bodyStart && // the reference is in the parameter part.
            variable.defs.every(d => d.name.range[0] >= bodyStart) // the variable is in the body.
        );
    ***REMOVED***
***REMOVED***

class ForScope extends Scope ***REMOVED***
    constructor(scopeManager, upperScope, block) ***REMOVED***
        super(scopeManager, "for", upperScope, block, false);
    ***REMOVED***
***REMOVED***

class ClassScope extends Scope ***REMOVED***
    constructor(scopeManager, upperScope, block) ***REMOVED***
        super(scopeManager, "class", upperScope, block, false);
    ***REMOVED***
***REMOVED***

module.exports = ***REMOVED***
    Scope,
    GlobalScope,
    ModuleScope,
    FunctionExpressionNameScope,
    CatchScope,
    WithScope,
    BlockScope,
    SwitchScope,
    FunctionScope,
    ForScope,
    ClassScope
***REMOVED***;

/* vim: set sw=4 ts=4 et tw=80 : */
