// rxjs/internal/scheduler/VirtualTimeScheduler.js
"use strict";
var __extends = (this && this.__extends) || (function () ***REMOVED***
    var extendStatics = function (d, b) ***REMOVED***
        extendStatics = Object.setPrototypeOf ||
            (***REMOVED*** __proto__: [] ***REMOVED*** instanceof Array && function (d, b) ***REMOVED*** d.__proto__ = b; ***REMOVED***) ||
            function (d, b) ***REMOVED*** for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; ***REMOVED***;
        return extendStatics(d, b);
    ***REMOVED***
    return function (d, b) ***REMOVED***
        extendStatics(d, b);
        function __() ***REMOVED*** this.constructor = d; ***REMOVED***
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    ***REMOVED***;
***REMOVED***)();
Object.defineProperty(exports, "__esModule", ***REMOVED*** value: true ***REMOVED***);
var AsyncAction_1 = require("./AsyncAction");
var AsyncScheduler_1 = require("./AsyncScheduler");
var VirtualTimeScheduler = (function (_super) ***REMOVED***
    __extends(VirtualTimeScheduler, _super);
    function VirtualTimeScheduler(SchedulerAction, maxFrames) ***REMOVED***
        if (SchedulerAction === void 0) ***REMOVED*** SchedulerAction = VirtualAction; ***REMOVED***
        if (maxFrames === void 0) ***REMOVED*** maxFrames = Number.POSITIVE_INFINITY; ***REMOVED***
        var _this = _super.call(this, SchedulerAction, function () ***REMOVED*** return _this.frame; ***REMOVED***) || this;
        _this.maxFrames = maxFrames;
        _this.frame = 0;
        _this.index = -1;
        return _this;
    ***REMOVED***
    VirtualTimeScheduler.prototype.flush = function () ***REMOVED***
        var _a = this, actions = _a.actions, maxFrames = _a.maxFrames;
        var error, action;
        while ((action = actions[0]) && action.delay <= maxFrames) ***REMOVED***
            actions.shift();
            this.frame = action.delay;
            if (error = action.execute(action.state, action.delay)) ***REMOVED***
                break;
            ***REMOVED***
        ***REMOVED***
        if (error) ***REMOVED***
            while (action = actions.shift()) ***REMOVED***
                action.unsubscribe();
            ***REMOVED***
            throw error;
        ***REMOVED***
    ***REMOVED***;
    VirtualTimeScheduler.frameTimeFactor = 10;
    return VirtualTimeScheduler;
***REMOVED***(AsyncScheduler_1.AsyncScheduler));
exports.VirtualTimeScheduler = VirtualTimeScheduler;
var VirtualAction = (function (_super) ***REMOVED***
    __extends(VirtualAction, _super);
    function VirtualAction(scheduler, work, index) ***REMOVED***
        if (index === void 0) ***REMOVED*** index = scheduler.index += 1; ***REMOVED***
        var _this = _super.call(this, scheduler, work) || this;
        _this.scheduler = scheduler;
        _this.work = work;
        _this.index = index;
        _this.active = true;
        _this.index = scheduler.index = index;
        return _this;
    ***REMOVED***
    VirtualAction.prototype.schedule = function (state, delay) ***REMOVED***
        if (delay === void 0) ***REMOVED*** delay = 0; ***REMOVED***
        if (!this.id) ***REMOVED***
            return _super.prototype.schedule.call(this, state, delay);
        ***REMOVED***
        this.active = false;
        var action = new VirtualAction(this.scheduler, this.work);
        this.add(action);
        return action.schedule(state, delay);
    ***REMOVED***;
    VirtualAction.prototype.requestAsyncId = function (scheduler, id, delay) ***REMOVED***
        if (delay === void 0) ***REMOVED*** delay = 0; ***REMOVED***
        this.delay = scheduler.frame + delay;
        var actions = scheduler.actions;
        actions.push(this);
        actions.sort(VirtualAction.sortActions);
        return true;
    ***REMOVED***;
    VirtualAction.prototype.recycleAsyncId = function (scheduler, id, delay) ***REMOVED***
        if (delay === void 0) ***REMOVED*** delay = 0; ***REMOVED***
        return undefined;
    ***REMOVED***;
    VirtualAction.prototype._execute = function (state, delay) ***REMOVED***
        if (this.active === true) ***REMOVED***
            return _super.prototype._execute.call(this, state, delay);
        ***REMOVED***
    ***REMOVED***;
    VirtualAction.sortActions = function (a, b) ***REMOVED***
        if (a.delay === b.delay) ***REMOVED***
            if (a.index === b.index) ***REMOVED***
                return 0;
            ***REMOVED***
            else if (a.index > b.index) ***REMOVED***
                return 1;
            ***REMOVED***
            else ***REMOVED***
                return -1;
            ***REMOVED***
        ***REMOVED***
        else if (a.delay > b.delay) ***REMOVED***
            return 1;
        ***REMOVED***
        else ***REMOVED***
            return -1;
        ***REMOVED***
    ***REMOVED***;
    return VirtualAction;
***REMOVED***(AsyncAction_1.AsyncAction));
exports.VirtualAction = VirtualAction;
