// levn/lib/coerce.js
// Generated by LiveScript 1.2.0
(function()***REMOVED***
  var parsedTypeCheck, types, toString$ = ***REMOVED******REMOVED***.toString;
  parsedTypeCheck = require('type-check').parsedTypeCheck;
  types = ***REMOVED***
    '*': function(it)***REMOVED***
      switch (toString$.call(it).slice(8, -1)) ***REMOVED***
      case 'Array':
        return coerceType(it, ***REMOVED***
          type: 'Array'
        ***REMOVED***);
      case 'Object':
        return coerceType(it, ***REMOVED***
          type: 'Object'
        ***REMOVED***);
      default:
        return ***REMOVED***
          type: 'Just',
          value: coerceTypes(it, [
            ***REMOVED***
              type: 'Undefined'
            ***REMOVED***, ***REMOVED***
              type: 'Null'
            ***REMOVED***, ***REMOVED***
              type: 'NaN'
            ***REMOVED***, ***REMOVED***
              type: 'Boolean'
            ***REMOVED***, ***REMOVED***
              type: 'Number'
            ***REMOVED***, ***REMOVED***
              type: 'Date'
            ***REMOVED***, ***REMOVED***
              type: 'RegExp'
            ***REMOVED***, ***REMOVED***
              type: 'Array'
            ***REMOVED***, ***REMOVED***
              type: 'Object'
            ***REMOVED***, ***REMOVED***
              type: 'String'
            ***REMOVED***
          ], ***REMOVED***
            explicit: true
          ***REMOVED***)
        ***REMOVED***;
      ***REMOVED***
    ***REMOVED***,
    Undefined: function(it)***REMOVED***
      if (it === 'undefined' || it === void 8) ***REMOVED***
        return ***REMOVED***
          type: 'Just',
          value: void 8
        ***REMOVED***;
      ***REMOVED*** else ***REMOVED***
        return ***REMOVED***
          type: 'Nothing'
        ***REMOVED***;
      ***REMOVED***
    ***REMOVED***,
    Null: function(it)***REMOVED***
      if (it === 'null') ***REMOVED***
        return ***REMOVED***
          type: 'Just',
          value: null
        ***REMOVED***;
      ***REMOVED*** else ***REMOVED***
        return ***REMOVED***
          type: 'Nothing'
        ***REMOVED***;
      ***REMOVED***
    ***REMOVED***,
    NaN: function(it)***REMOVED***
      if (it === 'NaN') ***REMOVED***
        return ***REMOVED***
          type: 'Just',
          value: NaN
        ***REMOVED***;
      ***REMOVED*** else ***REMOVED***
        return ***REMOVED***
          type: 'Nothing'
        ***REMOVED***;
      ***REMOVED***
    ***REMOVED***,
    Boolean: function(it)***REMOVED***
      if (it === 'true') ***REMOVED***
        return ***REMOVED***
          type: 'Just',
          value: true
        ***REMOVED***;
      ***REMOVED*** else if (it === 'false') ***REMOVED***
        return ***REMOVED***
          type: 'Just',
          value: false
        ***REMOVED***;
      ***REMOVED*** else ***REMOVED***
        return ***REMOVED***
          type: 'Nothing'
        ***REMOVED***;
      ***REMOVED***
    ***REMOVED***,
    Number: function(it)***REMOVED***
      return ***REMOVED***
        type: 'Just',
        value: +it
      ***REMOVED***;
    ***REMOVED***,
    Int: function(it)***REMOVED***
      return ***REMOVED***
        type: 'Just',
        value: parseInt(it)
      ***REMOVED***;
    ***REMOVED***,
    Float: function(it)***REMOVED***
      return ***REMOVED***
        type: 'Just',
        value: parseFloat(it)
      ***REMOVED***;
    ***REMOVED***,
    Date: function(value, options)***REMOVED***
      var that;
      if (that = /^\#(.*)\#$/.exec(value)) ***REMOVED***
        return ***REMOVED***
          type: 'Just',
          value: new Date(+that[1] || that[1])
        ***REMOVED***;
      ***REMOVED*** else if (options.explicit) ***REMOVED***
        return ***REMOVED***
          type: 'Nothing'
        ***REMOVED***;
      ***REMOVED*** else ***REMOVED***
        return ***REMOVED***
          type: 'Just',
          value: new Date(+value || value)
        ***REMOVED***;
      ***REMOVED***
    ***REMOVED***,
    RegExp: function(value, options)***REMOVED***
      var that;
      if (that = /^\/(.*)\/([gimy]*)$/.exec(value)) ***REMOVED***
        return ***REMOVED***
          type: 'Just',
          value: new RegExp(that[1], that[2])
        ***REMOVED***;
      ***REMOVED*** else if (options.explicit) ***REMOVED***
        return ***REMOVED***
          type: 'Nothing'
        ***REMOVED***;
      ***REMOVED*** else ***REMOVED***
        return ***REMOVED***
          type: 'Just',
          value: new RegExp(value)
        ***REMOVED***;
      ***REMOVED***
    ***REMOVED***,
    Array: function(it)***REMOVED***
      return coerceArray(it, ***REMOVED***
        of: [***REMOVED***
          type: '*'
        ***REMOVED***]
      ***REMOVED***);
    ***REMOVED***,
    Object: function(it)***REMOVED***
      return coerceFields(it, ***REMOVED***
        of: ***REMOVED******REMOVED***
      ***REMOVED***);
    ***REMOVED***,
    String: function(it)***REMOVED***
      var that;
      if (toString$.call(it).slice(8, -1) !== 'String') ***REMOVED***
        return ***REMOVED***
          type: 'Nothing'
        ***REMOVED***;
      ***REMOVED***
      if (that = it.match(/^'(.*)'$/)) ***REMOVED***
        return ***REMOVED***
          type: 'Just',
          value: that[1]
        ***REMOVED***;
      ***REMOVED*** else if (that = it.match(/^"(.*)"$/)) ***REMOVED***
        return ***REMOVED***
          type: 'Just',
          value: that[1]
        ***REMOVED***;
      ***REMOVED*** else ***REMOVED***
        return ***REMOVED***
          type: 'Just',
          value: it
        ***REMOVED***;
      ***REMOVED***
    ***REMOVED***
  ***REMOVED***;
  function coerceArray(node, type)***REMOVED***
    var typeOf, element;
    if (toString$.call(node).slice(8, -1) !== 'Array') ***REMOVED***
      return ***REMOVED***
        type: 'Nothing'
      ***REMOVED***;
    ***REMOVED***
    typeOf = type.of;
    return ***REMOVED***
      type: 'Just',
      value: (function()***REMOVED***
        var i$, ref$, len$, results$ = [];
        for (i$ = 0, len$ = (ref$ = node).length; i$ < len$; ++i$) ***REMOVED***
          element = ref$[i$];
          results$.push(coerceTypes(element, typeOf));
        ***REMOVED***
        return results$;
      ***REMOVED***())
    ***REMOVED***;
  ***REMOVED***
  function coerceTuple(node, type)***REMOVED***
    var result, i$, ref$, len$, i, types, that;
    if (toString$.call(node).slice(8, -1) !== 'Array') ***REMOVED***
      return ***REMOVED***
        type: 'Nothing'
      ***REMOVED***;
    ***REMOVED***
    result = [];
    for (i$ = 0, len$ = (ref$ = type.of).length; i$ < len$; ++i$) ***REMOVED***
      i = i$;
      types = ref$[i$];
      if (that = coerceTypes(node[i], types)) ***REMOVED***
        result.push(that);
      ***REMOVED***
    ***REMOVED***
    return ***REMOVED***
      type: 'Just',
      value: result
    ***REMOVED***;
  ***REMOVED***
  function coerceFields(node, type)***REMOVED***
    var typeOf, key, value;
    if (toString$.call(node).slice(8, -1) !== 'Object') ***REMOVED***
      return ***REMOVED***
        type: 'Nothing'
      ***REMOVED***;
    ***REMOVED***
    typeOf = type.of;
    return ***REMOVED***
      type: 'Just',
      value: (function()***REMOVED***
        var ref$, results$ = ***REMOVED******REMOVED***;
        for (key in ref$ = node) ***REMOVED***
          value = ref$[key];
          results$[key] = coerceTypes(value, typeOf[key] || [***REMOVED***
            type: '*'
          ***REMOVED***]);
        ***REMOVED***
        return results$;
      ***REMOVED***())
    ***REMOVED***;
  ***REMOVED***
  function coerceType(node, typeObj, options)***REMOVED***
    var type, structure, coerceFunc;
    type = typeObj.type, structure = typeObj.structure;
    if (type) ***REMOVED***
      coerceFunc = types[type];
      return coerceFunc(node, options);
    ***REMOVED*** else ***REMOVED***
      switch (structure) ***REMOVED***
      case 'array':
        return coerceArray(node, typeObj);
      case 'tuple':
        return coerceTuple(node, typeObj);
      case 'fields':
        return coerceFields(node, typeObj);
      ***REMOVED***
    ***REMOVED***
  ***REMOVED***
  function coerceTypes(node, types, options)***REMOVED***
    var i$, len$, type, ref$, valueType, value;
    for (i$ = 0, len$ = types.length; i$ < len$; ++i$) ***REMOVED***
      type = types[i$];
      ref$ = coerceType(node, type, options), valueType = ref$.type, value = ref$.value;
      if (valueType === 'Nothing') ***REMOVED***
        continue;
      ***REMOVED***
      if (parsedTypeCheck([type], value)) ***REMOVED***
        return value;
      ***REMOVED***
    ***REMOVED***
    throw new Error("Value " + JSON.stringify(node) + " does not type check against " + JSON.stringify(types) + ".");
  ***REMOVED***
  module.exports = coerceTypes;
***REMOVED***).call(this);
