// levn/lib/parse.js
// Generated by LiveScript 1.2.0
(function()***REMOVED***
  var reject, special, tokenRegex;
  reject = require('prelude-ls').reject;
  function consumeOp(tokens, op)***REMOVED***
    if (tokens[0] === op) ***REMOVED***
      return tokens.shift();
    ***REMOVED*** else ***REMOVED***
      throw new Error("Expected '" + op + "', but got '" + tokens[0] + "' instead in " + JSON.stringify(tokens) + ".");
    ***REMOVED***
  ***REMOVED***
  function maybeConsumeOp(tokens, op)***REMOVED***
    if (tokens[0] === op) ***REMOVED***
      return tokens.shift();
    ***REMOVED***
  ***REMOVED***
  function consumeList(tokens, delimiters, hasDelimiters)***REMOVED***
    var result;
    if (hasDelimiters) ***REMOVED***
      consumeOp(tokens, delimiters[0]);
    ***REMOVED***
    result = [];
    while (tokens.length && tokens[0] !== delimiters[1]) ***REMOVED***
      result.push(consumeElement(tokens));
      maybeConsumeOp(tokens, ',');
    ***REMOVED***
    if (hasDelimiters) ***REMOVED***
      consumeOp(tokens, delimiters[1]);
    ***REMOVED***
    return result;
  ***REMOVED***
  function consumeArray(tokens, hasDelimiters)***REMOVED***
    return consumeList(tokens, ['[', ']'], hasDelimiters);
  ***REMOVED***
  function consumeTuple(tokens, hasDelimiters)***REMOVED***
    return consumeList(tokens, ['(', ')'], hasDelimiters);
  ***REMOVED***
  function consumeFields(tokens, hasDelimiters)***REMOVED***
    var result, key;
    if (hasDelimiters) ***REMOVED***
      consumeOp(tokens, '***REMOVED***');
    ***REMOVED***
    result = ***REMOVED******REMOVED***;
    while (tokens.length && (!hasDelimiters || tokens[0] !== '***REMOVED***')) ***REMOVED***
      key = tokens.shift();
      consumeOp(tokens, ':');
      result[key] = consumeElement(tokens);
      maybeConsumeOp(tokens, ',');
    ***REMOVED***
    if (hasDelimiters) ***REMOVED***
      consumeOp(tokens, '***REMOVED***');
    ***REMOVED***
    return result;
  ***REMOVED***
  function consumeElement(tokens)***REMOVED***
    switch (tokens[0]) ***REMOVED***
    case '[':
      return consumeArray(tokens, true);
    case '(':
      return consumeTuple(tokens, true);
    case '***REMOVED***':
      return consumeFields(tokens, true);
    default:
      return tokens.shift();
    ***REMOVED***
  ***REMOVED***
  function consumeTopLevel(tokens, types)***REMOVED***
    var ref$, type, structure, origTokens, result, finalResult, x$, y$;
    ref$ = types[0], type = ref$.type, structure = ref$.structure;
    origTokens = tokens.concat();
    if (types.length === 1 && (structure || (type === 'Array' || type === 'Object'))) ***REMOVED***
      result = structure === 'array' || type === 'Array'
        ? consumeArray(tokens, tokens[0] === '[')
        : structure === 'tuple'
          ? consumeTuple(tokens, tokens[0] === '(')
          : consumeFields(tokens, tokens[0] === '***REMOVED***');
      finalResult = tokens.length ? consumeElement(structure === 'array' || type === 'Array'
        ? (x$ = origTokens, x$.unshift('['), x$.push(']'), x$)
        : (y$ = origTokens, y$.unshift('('), y$.push(')'), y$)) : result;
    ***REMOVED*** else ***REMOVED***
      finalResult = consumeElement(tokens);
    ***REMOVED***
    if (tokens.length && origTokens.length) ***REMOVED***
      throw new Error("Unable to parse " + JSON.stringify(origTokens) + " of type " + JSON.stringify(types) + ".");
    ***REMOVED*** else ***REMOVED***
      return finalResult;
    ***REMOVED***
  ***REMOVED***
  special = /\[\]\(\)***REMOVED******REMOVED***:,/.source;
  tokenRegex = RegExp('("(?:[^"]|\\\\")*")|(\'(?:[^\']|\\\\\')*\')|(#.*#)|(/(?:\\\\/|[^/])*/[gimy]*)|([' + special + '])|([^\\s' + special + ']+)|\\s*');
  module.exports = function(string, types)***REMOVED***
    var tokens, node;
    tokens = reject(function(it)***REMOVED***
      return !it || /^\s+$/.test(it);
    ***REMOVED***, string.split(tokenRegex));
    node = consumeTopLevel(tokens, types);
    if (!node) ***REMOVED***
      throw new Error("Error parsing '" + string + "'.");
    ***REMOVED***
    return node;
  ***REMOVED***;
***REMOVED***).call(this);
