// levn/lib/parse-string.js
// Generated by LiveScript 1.4.0
(function()***REMOVED***
  var reject, special, tokenRegex;
  reject = require('prelude-ls').reject;
  function consumeOp(tokens, op)***REMOVED***
    if (tokens[0] === op) ***REMOVED***
      return tokens.shift();
    ***REMOVED*** else ***REMOVED***
      throw new Error("Expected '" + op + "', but got '" + tokens[0] + "' instead in " + JSON.stringify(tokens) + ".");
    ***REMOVED***
  ***REMOVED***
  function maybeConsumeOp(tokens, op)***REMOVED***
    if (tokens[0] === op) ***REMOVED***
      return tokens.shift();
    ***REMOVED***
  ***REMOVED***
  function consumeList(tokens, arg$, hasDelimiters)***REMOVED***
    var open, close, result, untilTest;
    open = arg$[0], close = arg$[1];
    if (hasDelimiters) ***REMOVED***
      consumeOp(tokens, open);
    ***REMOVED***
    result = [];
    untilTest = "," + (hasDelimiters ? close : '');
    while (tokens.length && (hasDelimiters && tokens[0] !== close)) ***REMOVED***
      result.push(consumeElement(tokens, untilTest));
      maybeConsumeOp(tokens, ',');
    ***REMOVED***
    if (hasDelimiters) ***REMOVED***
      consumeOp(tokens, close);
    ***REMOVED***
    return result;
  ***REMOVED***
  function consumeArray(tokens, hasDelimiters)***REMOVED***
    return consumeList(tokens, ['[', ']'], hasDelimiters);
  ***REMOVED***
  function consumeTuple(tokens, hasDelimiters)***REMOVED***
    return consumeList(tokens, ['(', ')'], hasDelimiters);
  ***REMOVED***
  function consumeFields(tokens, hasDelimiters)***REMOVED***
    var result, untilTest, key;
    if (hasDelimiters) ***REMOVED***
      consumeOp(tokens, '***REMOVED***');
    ***REMOVED***
    result = ***REMOVED******REMOVED***;
    untilTest = "," + (hasDelimiters ? '***REMOVED***' : '');
    while (tokens.length && (!hasDelimiters || tokens[0] !== '***REMOVED***')) ***REMOVED***
      key = consumeValue(tokens, ':');
      consumeOp(tokens, ':');
      result[key] = consumeElement(tokens, untilTest);
      maybeConsumeOp(tokens, ',');
    ***REMOVED***
    if (hasDelimiters) ***REMOVED***
      consumeOp(tokens, '***REMOVED***');
    ***REMOVED***
    return result;
  ***REMOVED***
  function consumeValue(tokens, untilTest)***REMOVED***
    var out;
    untilTest == null && (untilTest = '');
    out = '';
    while (tokens.length && -1 === untilTest.indexOf(tokens[0])) ***REMOVED***
      out += tokens.shift();
    ***REMOVED***
    return out;
  ***REMOVED***
  function consumeElement(tokens, untilTest)***REMOVED***
    switch (tokens[0]) ***REMOVED***
    case '[':
      return consumeArray(tokens, true);
    case '(':
      return consumeTuple(tokens, true);
    case '***REMOVED***':
      return consumeFields(tokens, true);
    default:
      return consumeValue(tokens, untilTest);
    ***REMOVED***
  ***REMOVED***
  function consumeTopLevel(tokens, types, options)***REMOVED***
    var ref$, type, structure, origTokens, result, finalResult, x$, y$;
    ref$ = types[0], type = ref$.type, structure = ref$.structure;
    origTokens = tokens.concat();
    if (!options.explicit && types.length === 1 && ((!type && structure) || (type === 'Array' || type === 'Object'))) ***REMOVED***
      result = structure === 'array' || type === 'Array'
        ? consumeArray(tokens, tokens[0] === '[')
        : structure === 'tuple'
          ? consumeTuple(tokens, tokens[0] === '(')
          : consumeFields(tokens, tokens[0] === '***REMOVED***');
      finalResult = tokens.length ? consumeElement(structure === 'array' || type === 'Array'
        ? (x$ = origTokens, x$.unshift('['), x$.push(']'), x$)
        : (y$ = origTokens, y$.unshift('('), y$.push(')'), y$)) : result;
    ***REMOVED*** else ***REMOVED***
      finalResult = consumeElement(tokens);
    ***REMOVED***
    return finalResult;
  ***REMOVED***
  special = /\[\]\(\)***REMOVED******REMOVED***:,/.source;
  tokenRegex = RegExp('("(?:\\\\"|[^"])*")|(\'(?:\\\\\'|[^\'])*\')|(/(?:\\\\/|[^/])*/[a-zA-Z]*)|(#.*#)|([' + special + '])|([^\\s' + special + '](?:\\s*[^\\s' + special + ']+)*)|\\s*');
  module.exports = function(types, string, options)***REMOVED***
    var tokens, node;
    options == null && (options = ***REMOVED******REMOVED***);
    if (!options.explicit && types.length === 1 && types[0].type === 'String') ***REMOVED***
      return "'" + string.replace(/\\'/g, "\\\\'") + "'";
    ***REMOVED***
    tokens = reject(not$, string.split(tokenRegex));
    node = consumeTopLevel(tokens, types, options);
    if (!node) ***REMOVED***
      throw new Error("Error parsing '" + string + "'.");
    ***REMOVED***
    return node;
  ***REMOVED***;
  function not$(x)***REMOVED*** return !x; ***REMOVED***
***REMOVED***).call(this);
