// discord.js/src/util/Collection.js
const util = require('util');

/**
 * A Map with additional utility methods. This is used throughout discord.js rather than Arrays for anything that has
 * an ID, for significantly improved performance and ease-of-use.
 * @extends ***REMOVED***Map***REMOVED***
 */
class Collection extends Map ***REMOVED***
  constructor(iterable) ***REMOVED***
    super(iterable);

    /**
     * Cached array for the `array()` method - will be reset to `null` whenever `set()` or `delete()` are called
     * @name Collection#_array
     * @type ***REMOVED***?Array***REMOVED***
     * @private
     */
    Object.defineProperty(this, '_array', ***REMOVED*** value: null, writable: true, configurable: true ***REMOVED***);

    /**
     * Cached array for the `keyArray()` method - will be reset to `null` whenever `set()` or `delete()` are called
     * @name Collection#_keyArray
     * @type ***REMOVED***?Array***REMOVED***
     * @private
     */
    Object.defineProperty(this, '_keyArray', ***REMOVED*** value: null, writable: true, configurable: true ***REMOVED***);
  ***REMOVED***

  set(key, val) ***REMOVED***
    this._array = null;
    this._keyArray = null;
    return super.set(key, val);
  ***REMOVED***

  delete(key) ***REMOVED***
    this._array = null;
    this._keyArray = null;
    return super.delete(key);
  ***REMOVED***

  /**
   * Creates an ordered array of the values of this collection, and caches it internally. The array will only be
   * reconstructed if an item is added to or removed from the collection, or if you change the length of the array
   * itself. If you don't want this caching behavior, use `[...collection.values()]` or
   * `Array.from(collection.values())` instead.
   * @returns ***REMOVED***Array***REMOVED***
   */
  array() ***REMOVED***
    if (!this._array || this._array.length !== this.size) this._array = [...this.values()];
    return this._array;
  ***REMOVED***

  /**
   * Creates an ordered array of the keys of this collection, and caches it internally. The array will only be
   * reconstructed if an item is added to or removed from the collection, or if you change the length of the array
   * itself. If you don't want this caching behavior, use `[...collection.keys()]` or
   * `Array.from(collection.keys())` instead.
   * @returns ***REMOVED***Array***REMOVED***
   */
  keyArray() ***REMOVED***
    if (!this._keyArray || this._keyArray.length !== this.size) this._keyArray = [...this.keys()];
    return this._keyArray;
  ***REMOVED***

  /**
   * Obtains the first value(s) in this collection.
   * @param ***REMOVED***number***REMOVED*** [count] Number of values to obtain from the beginning
   * @returns ***REMOVED****|Array<*>***REMOVED*** The single value if `count` is undefined, or an array of values of `count` length
   */
  first(count) ***REMOVED***
    if (count === undefined) return this.values().next().value;
    if (typeof count !== 'number') throw new TypeError('The count must be a number.');
    if (!Number.isInteger(count) || count < 1) throw new RangeError('The count must be an integer greater than 0.');
    count = Math.min(this.size, count);
    const arr = new Array(count);
    const iter = this.values();
    for (let i = 0; i < count; i++) arr[i] = iter.next().value;
    return arr;
  ***REMOVED***

  /**
   * Obtains the first key(s) in this collection.
   * @param ***REMOVED***number***REMOVED*** [count] Number of keys to obtain from the beginning
   * @returns ***REMOVED****|Array<*>***REMOVED*** The single key if `count` is undefined, or an array of keys of `count` length
   */
  firstKey(count) ***REMOVED***
    if (count === undefined) return this.keys().next().value;
    if (typeof count !== 'number') throw new TypeError('The count must be a number.');
    if (!Number.isInteger(count) || count < 1) throw new RangeError('The count must be an integer greater than 0.');
    count = Math.min(this.size, count);
    const arr = new Array(count);
    const iter = this.keys();
    for (let i = 0; i < count; i++) arr[i] = iter.next().value;
    return arr;
  ***REMOVED***

  /**
   * Obtains the last value(s) in this collection. This relies on ***REMOVED***@link Collection#array***REMOVED***, and thus the caching
   * mechanism applies here as well.
   * @param ***REMOVED***number***REMOVED*** [count] Number of values to obtain from the end
   * @returns ***REMOVED****|Array<*>***REMOVED*** The single value if `count` is undefined, or an array of values of `count` length
   */
  last(count) ***REMOVED***
    const arr = this.array();
    if (count === undefined) return arr[arr.length - 1];
    if (typeof count !== 'number') throw new TypeError('The count must be a number.');
    if (!Number.isInteger(count) || count < 1) throw new RangeError('The count must be an integer greater than 0.');
    return arr.slice(-count);
  ***REMOVED***

  /**
   * Obtains the last key(s) in this collection. This relies on ***REMOVED***@link Collection#keyArray***REMOVED***, and thus the caching
   * mechanism applies here as well.
   * @param ***REMOVED***number***REMOVED*** [count] Number of keys to obtain from the end
   * @returns ***REMOVED****|Array<*>***REMOVED*** The single key if `count` is undefined, or an array of keys of `count` length
   */
  lastKey(count) ***REMOVED***
    const arr = this.keyArray();
    if (count === undefined) return arr[arr.length - 1];
    if (typeof count !== 'number') throw new TypeError('The count must be a number.');
    if (!Number.isInteger(count) || count < 1) throw new RangeError('The count must be an integer greater than 0.');
    return arr.slice(-count);
  ***REMOVED***

  /**
   * Obtains random value(s) from this collection. This relies on ***REMOVED***@link Collection#array***REMOVED***, and thus the caching
   * mechanism applies here as well.
   * @param ***REMOVED***number***REMOVED*** [count] Number of values to obtain randomly
   * @returns ***REMOVED****|Array<*>***REMOVED*** The single value if `count` is undefined, or an array of values of `count` length
   */
  random(count) ***REMOVED***
    let arr = this.array();
    if (count === undefined) return arr[Math.floor(Math.random() * arr.length)];
    if (typeof count !== 'number') throw new TypeError('The count must be a number.');
    if (!Number.isInteger(count) || count < 1) throw new RangeError('The count must be an integer greater than 0.');
    if (arr.length === 0) return [];
    const rand = new Array(count);
    arr = arr.slice();
    for (let i = 0; i < count; i++) rand[i] = arr.splice(Math.floor(Math.random() * arr.length), 1)[0];
    return rand;
  ***REMOVED***

  /**
   * Obtains random key(s) from this collection. This relies on ***REMOVED***@link Collection#keyArray***REMOVED***, and thus the caching
   * mechanism applies here as well.
   * @param ***REMOVED***number***REMOVED*** [count] Number of keys to obtain randomly
   * @returns ***REMOVED****|Array<*>***REMOVED*** The single key if `count` is undefined, or an array of keys of `count` length
   */
  randomKey(count) ***REMOVED***
    let arr = this.keyArray();
    if (count === undefined) return arr[Math.floor(Math.random() * arr.length)];
    if (typeof count !== 'number') throw new TypeError('The count must be a number.');
    if (!Number.isInteger(count) || count < 1) throw new RangeError('The count must be an integer greater than 0.');
    if (arr.length === 0) return [];
    const rand = new Array(count);
    arr = arr.slice();
    for (let i = 0; i < count; i++) rand[i] = arr.splice(Math.floor(Math.random() * arr.length), 1)[0];
    return rand;
  ***REMOVED***

  /**
   * Searches for all items where their specified property's value is identical to the given value
   * (`item[prop] === value`).
   * @param ***REMOVED***string***REMOVED*** prop The property to test against
   * @param ***REMOVED*******REMOVED*** value The expected value
   * @returns ***REMOVED***Array***REMOVED***
   * @deprecated
   * @example
   * collection.findAll('username', 'Bob');
   */
  findAll(prop, value) ***REMOVED***
    if (typeof prop !== 'string') throw new TypeError('Key must be a string.');
    if (typeof value === 'undefined') throw new Error('Value must be specified.');
    const results = [];
    for (const item of this.values()) ***REMOVED***
      if (item[prop] === value) results.push(item);
    ***REMOVED***
    return results;
  ***REMOVED***

  /**
   * Searches for a single item where its specified property's value is identical to the given value
   * (`item[prop] === value`), or the given function returns a truthy value. In the latter case, this is identical to
   * [Array.find()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/find).
   * <warn>All collections used in Discord.js are mapped using their `id` property, and if you want to find by id you
   * should use the `get` method. See
   * [MDN](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map/get) for details.</warn>
   * @param ***REMOVED***string|Function***REMOVED*** propOrFn The property to test against, or the function to test with
   * @param ***REMOVED*******REMOVED*** [value] The expected value - only applicable and required if using a property for the first argument
   * @returns ***REMOVED*******REMOVED***
   * @example
   * collection.find('username', 'Bob');
   * @example
   * collection.find(val => val.username === 'Bob');
   */
  find(propOrFn, value) ***REMOVED***
    if (typeof propOrFn === 'string') ***REMOVED***
      if (typeof value === 'undefined') throw new Error('Value must be specified.');
      for (const item of this.values()) ***REMOVED***
        if (item[propOrFn] === value) return item;
      ***REMOVED***
      return null;
    ***REMOVED*** else if (typeof propOrFn === 'function') ***REMOVED***
      for (const [key, val] of this) ***REMOVED***
        if (propOrFn(val, key, this)) return val;
      ***REMOVED***
      return null;
    ***REMOVED*** else ***REMOVED***
      throw new Error('First argument must be a property string or a function.');
    ***REMOVED***
  ***REMOVED***

  /* eslint-disable max-len */
  /**
   * Searches for the key of a single item where its specified property's value is identical to the given value
   * (`item[prop] === value`), or the given function returns a truthy value. In the latter case, this is identical to
   * [Array.findIndex()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/findIndex).
   * @param ***REMOVED***string|Function***REMOVED*** propOrFn The property to test against, or the function to test with
   * @param ***REMOVED*******REMOVED*** [value] The expected value - only applicable and required if using a property for the first argument
   * @returns ***REMOVED*******REMOVED***
   * @example
   * collection.findKey('username', 'Bob');
   * @example
   * collection.findKey(val => val.username === 'Bob');
   */
  findKey(propOrFn, value) ***REMOVED***
    /* eslint-enable max-len */
    if (typeof propOrFn === 'string') ***REMOVED***
      if (typeof value === 'undefined') throw new Error('Value must be specified.');
      for (const [key, val] of this) ***REMOVED***
        if (val[propOrFn] === value) return key;
      ***REMOVED***
      return null;
    ***REMOVED*** else if (typeof propOrFn === 'function') ***REMOVED***
      for (const [key, val] of this) ***REMOVED***
        if (propOrFn(val, key, this)) return key;
      ***REMOVED***
      return null;
    ***REMOVED*** else ***REMOVED***
      throw new Error('First argument must be a property string or a function.');
    ***REMOVED***
  ***REMOVED***

  /**
   * Searches for the existence of a single item where its specified property's value is identical to the given value
   * (`item[prop] === value`).
   * <warn>Do not use this to check for an item by its ID. Instead, use `collection.has(id)`. See
   * [MDN](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map/has) for details.</warn>
   * @param ***REMOVED***string***REMOVED*** prop The property to test against
   * @param ***REMOVED*******REMOVED*** value The expected value
   * @returns ***REMOVED***boolean***REMOVED***
   * @deprecated
   * @example
   * if (collection.exists('username', 'Bob')) ***REMOVED***
   *  console.log('user here!');
   * ***REMOVED***
   */
  exists(prop, value) ***REMOVED***
    return Boolean(this.find(prop, value));
  ***REMOVED***

  /**
   * Removes entries that satisfy the provided filter function.
   * @param ***REMOVED***Function***REMOVED*** fn Function used to test (should return a boolean)
   * @param ***REMOVED***Object***REMOVED*** [thisArg] Value to use as `this` when executing function
   * @returns ***REMOVED***number***REMOVED*** The number of removed entries
   */
  sweep(fn, thisArg) ***REMOVED***
    if (thisArg) fn = fn.bind(thisArg);
    const previousSize = this.size;
    for (const [key, val] of this) ***REMOVED***
      if (fn(val, key, this)) this.delete(key);
    ***REMOVED***
    return previousSize - this.size;
  ***REMOVED***

  /**
   * Identical to
   * [Array.filter()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/filter),
   * but returns a Collection instead of an Array.
   * @param ***REMOVED***Function***REMOVED*** fn Function used to test (should return a boolean)
   * @param ***REMOVED***Object***REMOVED*** [thisArg] Value to use as `this` when executing function
   * @returns ***REMOVED***Collection***REMOVED***
   */
  filter(fn, thisArg) ***REMOVED***
    if (thisArg) fn = fn.bind(thisArg);
    const results = new Collection();
    for (const [key, val] of this) ***REMOVED***
      if (fn(val, key, this)) results.set(key, val);
    ***REMOVED***
    return results;
  ***REMOVED***

  /**
   * Identical to
   * [Array.filter()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/filter).
   * @param ***REMOVED***Function***REMOVED*** fn Function used to test (should return a boolean)
   * @param ***REMOVED***Object***REMOVED*** [thisArg] Value to use as `this` when executing function
   * @returns ***REMOVED***Array***REMOVED***
   * @deprecated
   */
  filterArray(fn, thisArg) ***REMOVED***
    if (thisArg) fn = fn.bind(thisArg);
    const results = [];
    for (const [key, val] of this) ***REMOVED***
      if (fn(val, key, this)) results.push(val);
    ***REMOVED***
    return results;
  ***REMOVED***

  /**
   * Partitions the collection into two collections where the first collection
   * contains the items that passed and the second contains the items that failed.
   * @param ***REMOVED***Function***REMOVED*** fn Function used to test (should return a boolean)
   * @param ***REMOVED*******REMOVED*** [thisArg] Value to use as `this` when executing function
   * @returns ***REMOVED***Collection[]***REMOVED***
   * @example const [big, small] = collection.partition(guild => guild.memberCount > 250);
   */
  partition(fn, thisArg) ***REMOVED***
    if (typeof thisArg !== 'undefined') fn = fn.bind(thisArg);
    const results = [new Collection(), new Collection()];
    for (const [key, val] of this) ***REMOVED***
      if (fn(val, key, this)) ***REMOVED***
        results[0].set(key, val);
      ***REMOVED*** else ***REMOVED***
        results[1].set(key, val);
      ***REMOVED***
    ***REMOVED***
    return results;
  ***REMOVED***

  /**
   * Identical to
   * [Array.map()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/map).
   * @param ***REMOVED***Function***REMOVED*** fn Function that produces an element of the new array, taking three arguments
   * @param ***REMOVED*******REMOVED*** [thisArg] Value to use as `this` when executing function
   * @returns ***REMOVED***Array***REMOVED***
   */
  map(fn, thisArg) ***REMOVED***
    if (thisArg) fn = fn.bind(thisArg);
    const arr = new Array(this.size);
    let i = 0;
    for (const [key, val] of this) arr[i++] = fn(val, key, this);
    return arr;
  ***REMOVED***

  /**
   * Identical to
   * [Array.some()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/some).
   * @param ***REMOVED***Function***REMOVED*** fn Function used to test (should return a boolean)
   * @param ***REMOVED***Object***REMOVED*** [thisArg] Value to use as `this` when executing function
   * @returns ***REMOVED***boolean***REMOVED***
   */
  some(fn, thisArg) ***REMOVED***
    if (thisArg) fn = fn.bind(thisArg);
    for (const [key, val] of this) ***REMOVED***
      if (fn(val, key, this)) return true;
    ***REMOVED***
    return false;
  ***REMOVED***

  /**
   * Identical to
   * [Array.every()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/every).
   * @param ***REMOVED***Function***REMOVED*** fn Function used to test (should return a boolean)
   * @param ***REMOVED***Object***REMOVED*** [thisArg] Value to use as `this` when executing function
   * @returns ***REMOVED***boolean***REMOVED***
   */
  every(fn, thisArg) ***REMOVED***
    if (thisArg) fn = fn.bind(thisArg);
    for (const [key, val] of this) ***REMOVED***
      if (!fn(val, key, this)) return false;
    ***REMOVED***
    return true;
  ***REMOVED***

  /**
   * Identical to
   * [Array.reduce()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/reduce).
   * @param ***REMOVED***Function***REMOVED*** fn Function used to reduce, taking four arguments; `accumulator`, `currentValue`, `currentKey`,
   * and `collection`
   * @param ***REMOVED*******REMOVED*** [initialValue] Starting value for the accumulator
   * @returns ***REMOVED*******REMOVED***
   */
  reduce(fn, initialValue) ***REMOVED***
    let accumulator;
    if (typeof initialValue !== 'undefined') ***REMOVED***
      accumulator = initialValue;
      for (const [key, val] of this) accumulator = fn(accumulator, val, key, this);
    ***REMOVED*** else ***REMOVED***
      let first = true;
      for (const [key, val] of this) ***REMOVED***
        if (first) ***REMOVED***
          accumulator = val;
          first = false;
          continue;
        ***REMOVED***
        accumulator = fn(accumulator, val, key, this);
      ***REMOVED***
    ***REMOVED***
    return accumulator;
  ***REMOVED***

  /**
   * Identical to
   * [Map.forEach()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map/forEach),
   * but returns the collection instead of undefined.
   * @param ***REMOVED***Function***REMOVED*** fn Function to execute for each element
   * @param ***REMOVED*******REMOVED*** [thisArg] Value to use as `this` when executing function
   * @returns ***REMOVED***Collection***REMOVED***
   * @example
   * collection
   *  .tap(user => console.log(user.username))
   *  .filter(user => user.bot)
   *  .tap(user => console.log(user.username));
   */
  tap(fn, thisArg) ***REMOVED***
    this.forEach(fn, thisArg);
    return this;
  ***REMOVED***

  /**
   * Creates an identical shallow copy of this collection.
   * @returns ***REMOVED***Collection***REMOVED***
   * @example const newColl = someColl.clone();
   */
  clone() ***REMOVED***
    return new this.constructor(this);
  ***REMOVED***

  /**
   * Combines this collection with others into a new collection. None of the source collections are modified.
   * @param ***REMOVED***...Collection***REMOVED*** collections Collections to merge
   * @returns ***REMOVED***Collection***REMOVED***
   * @example const newColl = someColl.concat(someOtherColl, anotherColl, ohBoyAColl);
   */
  concat(...collections) ***REMOVED***
    const newColl = this.clone();
    for (const coll of collections) ***REMOVED***
      for (const [key, val] of coll) newColl.set(key, val);
    ***REMOVED***
    return newColl;
  ***REMOVED***

  /**
   * Calls the `delete()` method on all items that have it.
   * @returns ***REMOVED***Promise[]***REMOVED***
   */
  deleteAll() ***REMOVED***
    const returns = [];
    for (const item of this.values()) ***REMOVED***
      if (item.delete) returns.push(item.delete());
    ***REMOVED***
    return returns;
  ***REMOVED***

  /**
   * Checks if this collection shares identical key-value pairings with another.
   * This is different to checking for equality using equal-signs, because
   * the collections may be different objects, but contain the same data.
   * @param ***REMOVED***Collection***REMOVED*** collection Collection to compare with
   * @returns ***REMOVED***boolean***REMOVED*** Whether the collections have identical contents
   */
  equals(collection) ***REMOVED***
    if (!collection) return false;
    if (this === collection) return true;
    if (this.size !== collection.size) return false;
    return !this.find((value, key) => ***REMOVED***
      const testVal = collection.get(key);
      return testVal !== value || (testVal === undefined && !collection.has(key));
    ***REMOVED***);
  ***REMOVED***

  /**
   * The sort() method sorts the elements of a collection in place and returns the collection.
   * The sort is not necessarily stable. The default sort order is according to string Unicode code points.
   * @param ***REMOVED***Function***REMOVED*** [compareFunction] Specifies a function that defines the sort order.
   * if omitted, the collection is sorted according to each character's Unicode code point value,
   * according to the string conversion of each element.
   * @returns ***REMOVED***Collection***REMOVED***
   */
  sort(compareFunction = (x, y) => +(x > y) || +(x === y) - 1) ***REMOVED***
    return new Collection([...this.entries()].sort((a, b) => compareFunction(a[1], b[1], a[0], b[0])));
  ***REMOVED***
***REMOVED***

Collection.prototype.findAll =
  util.deprecate(Collection.prototype.findAll, 'Collection#findAll: use Collection#filter instead');

Collection.prototype.filterArray =
  util.deprecate(Collection.prototype.filterArray, 'Collection#filterArray: use Collection#filter instead');

Collection.prototype.exists =
  util.deprecate(Collection.prototype.exists, 'Collection#exists: use Collection#some instead');

Collection.prototype.find = function find(propOrFn, value) ***REMOVED***
  if (typeof propOrFn === 'string') ***REMOVED***
    process.emitWarning('Collection#find: pass a function instead', 'DeprecationWarning');
    if (typeof value === 'undefined') throw new Error('Value must be specified.');
    for (const item of this.values()) ***REMOVED***
      if (item[propOrFn] === value) return item;
    ***REMOVED***
    return null;
  ***REMOVED*** else if (typeof propOrFn === 'function') ***REMOVED***
    for (const [key, val] of this) ***REMOVED***
      if (propOrFn(val, key, this)) return val;
    ***REMOVED***
    return null;
  ***REMOVED*** else ***REMOVED***
    throw new Error('First argument must be a property string or a function.');
  ***REMOVED***
***REMOVED***;

Collection.prototype.findKey = function findKey(propOrFn, value) ***REMOVED***
  if (typeof propOrFn === 'string') ***REMOVED***
    process.emitWarning('Collection#findKey: pass a function instead', 'DeprecationWarning');
    if (typeof value === 'undefined') throw new Error('Value must be specified.');
    for (const [key, val] of this) ***REMOVED***
      if (val[propOrFn] === value) return key;
    ***REMOVED***
    return null;
  ***REMOVED*** else if (typeof propOrFn === 'function') ***REMOVED***
    for (const [key, val] of this) ***REMOVED***
      if (propOrFn(val, key, this)) return key;
    ***REMOVED***
    return null;
  ***REMOVED*** else ***REMOVED***
    throw new Error('First argument must be a property string or a function.');
  ***REMOVED***
***REMOVED***;

module.exports = Collection;
