// discord.js/src/util/MessageFlags.js
const BitField = require('./BitField');

/**
 * Data structure that makes it easy to interact with an ***REMOVED***@link Message#flags***REMOVED*** bitfield.
 * @extends ***REMOVED***BitField***REMOVED***
 */
class MessageFlags extends BitField ***REMOVED******REMOVED***

/**
 * Data that can be resolved to give a permission number. This can be:
 * * A string (see ***REMOVED***@link MessageFlags.FLAGS***REMOVED***)
 * * A message flag
 * * An instance of MessageFlags
 * * An array of MessageFlagsResolvable
 * @typedef ***REMOVED***string|number|MessageFlags|MessageFlagsResolvable[]***REMOVED*** MessageFlagsResolvable
 */

/**
 * Numeric message flags. All available properties:
 * * `CROSSPOSTED`
 * * `IS_CROSSPOST`
 * * `SUPPRESS_EMBEDS`
 * * `SOURCE_MESSAGE_DELETED`
 * * `URGENT`
 * @type ***REMOVED***Object***REMOVED***
 * @see ***REMOVED***@link https://discordapp.com/developers/docs/resources/channel#message-object-message-flags***REMOVED***
 */
MessageFlags.FLAGS = ***REMOVED***
  CROSSPOSTED: 1 << 0,
  IS_CROSSPOST: 1 << 1,
  SUPPRESS_EMBEDS: 1 << 2,
  SOURCE_MESSAGE_DELETED: 1 << 3,
  URGENT: 1 << 4,
***REMOVED***;

module.exports = MessageFlags;
