// discord.js/src/util/Permissions.js
const BitField = require('./BitField');
const util = require('util');

/**
 * Data structure that makes it easy to interact with a permission bitfield. All ***REMOVED***@link GuildMember***REMOVED***s have a set of
 * permissions in their guild, and each channel in the guild may also have ***REMOVED***@link PermissionOverwrites***REMOVED*** for the member
 * that override their default permissions.
 * @extends ***REMOVED***BitField***REMOVED***
 */
class Permissions extends BitField ***REMOVED***
  /**
   * @param ***REMOVED***GuildMember***REMOVED*** [member] Member the permissions are for **(deprecated)**
   * @param ***REMOVED***number|PermissionResolvable***REMOVED*** permissions Permissions or bitfield to read from
   */
  constructor(member, permissions) ***REMOVED***
    super(typeof member === 'object' && !(member instanceof Array) ? permissions : member);

    Object.defineProperty(this, '_member', ***REMOVED***
      writable: true,
      value: typeof member === 'object' && !(member instanceof Array) ? member : null,
    ***REMOVED***);
  ***REMOVED***

  /**
     * Member the permissions are for
     * @type ***REMOVED***GuildMember***REMOVED***
     * @deprecated
     */
  get member() ***REMOVED***
    return this._member;
  ***REMOVED***

  set member(value) ***REMOVED***
    this._member = value;
  ***REMOVED***

  /**
   * Bitfield of the packed permissions
   * @type ***REMOVED***number***REMOVED***
   * @see ***REMOVED***@link Permissions#bitfield***REMOVED***
   * @deprecated
   * @readonly
   */
  get raw() ***REMOVED***
    return this.bitfield;
  ***REMOVED***

  set raw(raw) ***REMOVED***
    this.bitfield = raw;
  ***REMOVED***

  /**
   * Checks whether the bitfield has a permission, or any of multiple permissions.
   * @param ***REMOVED***PermissionResolvable***REMOVED*** permission Permission(s) to check for
   * @param ***REMOVED***boolean***REMOVED*** [checkAdmin=true] Whether to allow the administrator permission to override
   * @returns ***REMOVED***boolean***REMOVED***
   */
  any(permission, checkAdmin = true) ***REMOVED***
    return (checkAdmin && super.has(this.constructor.FLAGS.ADMINISTRATOR)) || super.any(permission);
  ***REMOVED***

  /**
   * Checks whether the bitfield has a permission, or multiple permissions.
   * @param ***REMOVED***PermissionResolvable***REMOVED*** permission Permission(s) to check for
   * @param ***REMOVED***boolean***REMOVED*** [checkAdmin=true] Whether to allow the administrator permission to override
   * @returns ***REMOVED***boolean***REMOVED***
   */
  has(permission, checkAdmin = true) ***REMOVED***
    return (checkAdmin && super.has(this.constructor.FLAGS.ADMINISTRATOR)) || super.has(permission);
  ***REMOVED***

  /**
   * Checks whether the user has a certain permission, e.g. `READ_MESSAGES`.
   * @param ***REMOVED***PermissionResolvable***REMOVED*** permission The permission to check for
   * @param ***REMOVED***boolean***REMOVED*** [explicit=false] Whether to require the user to explicitly have the exact permission
   * @returns ***REMOVED***boolean***REMOVED***
   * @see ***REMOVED***@link Permissions#has***REMOVED***
   * @deprecated
   */
  hasPermission(permission, explicit = false) ***REMOVED***
    return this.has(permission, !explicit);
  ***REMOVED***

  /**
   * Checks whether the user has all specified permissions.
   * @param ***REMOVED***PermissionResolvable***REMOVED*** permissions The permissions to check for
   * @param ***REMOVED***boolean***REMOVED*** [explicit=false] Whether to require the user to explicitly have the exact permissions
   * @returns ***REMOVED***boolean***REMOVED***
   * @see ***REMOVED***@link Permissions#has***REMOVED***
   * @deprecated
   */
  hasPermissions(permissions, explicit = false) ***REMOVED***
    return this.has(permissions, !explicit);
  ***REMOVED***

  /**
   * Checks whether the user has all specified permissions, and lists any missing permissions.
   * @param ***REMOVED***PermissionResolvable***REMOVED*** permissions The permissions to check for
   * @param ***REMOVED***boolean***REMOVED*** [explicit=false] Whether to require the user to explicitly have the exact permissions
   * @returns ***REMOVED***PermissionResolvable***REMOVED***
   * @see ***REMOVED***@link Permissions#missing***REMOVED***
   * @deprecated
   */
  missingPermissions(permissions, explicit = false) ***REMOVED***
    return this.missing(permissions, !explicit);
  ***REMOVED***
***REMOVED***

/**
 * Data that can be resolved to give a permission number. This can be:
 * * A string (see ***REMOVED***@link Permissions.FLAGS***REMOVED***)
 * * A permission number
 * @typedef ***REMOVED***string|number|Permissions|PermissionResolvable[]***REMOVED*** PermissionResolvable
 */

/**
 * Numeric permission flags. All available properties:
 * - `ADMINISTRATOR` (implicitly has *all* permissions, and bypasses all channel overwrites)
 * - `CREATE_INSTANT_INVITE` (create invitations to the guild)
 * - `KICK_MEMBERS`
 * - `BAN_MEMBERS`
 * - `MANAGE_CHANNELS` (edit and reorder channels)
 * - `MANAGE_GUILD` (edit the guild information, region, etc.)
 * - `ADD_REACTIONS` (add new reactions to messages)
 * - `VIEW_AUDIT_LOG`
 * - `PRIORITY_SPEAKER`
 * - `STREAM`
 * - `VIEW_CHANNEL`
 * - `READ_MESSAGES` **(deprecated)**
 * - `SEND_MESSAGES`
 * - `SEND_TTS_MESSAGES`
 * - `MANAGE_MESSAGES` (delete messages and reactions)
 * - `EMBED_LINKS` (links posted will have a preview embedded)
 * - `ATTACH_FILES`
 * - `READ_MESSAGE_HISTORY` (view messages that were posted prior to opening Discord)
 * - `MENTION_EVERYONE`
 * - `USE_EXTERNAL_EMOJIS` (use emojis from different guilds)
 * - `EXTERNAL_EMOJIS` **(deprecated)**
 * - `CONNECT` (connect to a voice channel)
 * - `SPEAK` (speak in a voice channel)
 * - `MUTE_MEMBERS` (mute members across all voice channels)
 * - `DEAFEN_MEMBERS` (deafen members across all voice channels)
 * - `MOVE_MEMBERS` (move members between voice channels)
 * - `USE_VAD` (use voice activity detection)
 * - `CHANGE_NICKNAME`
 * - `MANAGE_NICKNAMES` (change other members' nicknames)
 * - `MANAGE_ROLES`
 * - `MANAGE_ROLES_OR_PERMISSIONS` **(deprecated)**
 * - `MANAGE_WEBHOOKS`
 * - `MANAGE_EMOJIS`
 * @type ***REMOVED***Object***REMOVED***
 * @see ***REMOVED***@link https://discordapp.com/developers/docs/topics/permissions***REMOVED***
 */
Permissions.FLAGS = ***REMOVED***
  CREATE_INSTANT_INVITE: 1 << 0,
  KICK_MEMBERS: 1 << 1,
  BAN_MEMBERS: 1 << 2,
  ADMINISTRATOR: 1 << 3,
  MANAGE_CHANNELS: 1 << 4,
  MANAGE_GUILD: 1 << 5,
  ADD_REACTIONS: 1 << 6,
  VIEW_AUDIT_LOG: 1 << 7,
  PRIORITY_SPEAKER: 1 << 8,
  STREAM: 1 << 9,

  VIEW_CHANNEL: 1 << 10,
  READ_MESSAGES: 1 << 10,
  SEND_MESSAGES: 1 << 11,
  SEND_TTS_MESSAGES: 1 << 12,
  MANAGE_MESSAGES: 1 << 13,
  EMBED_LINKS: 1 << 14,
  ATTACH_FILES: 1 << 15,
  READ_MESSAGE_HISTORY: 1 << 16,
  MENTION_EVERYONE: 1 << 17,
  EXTERNAL_EMOJIS: 1 << 18,
  USE_EXTERNAL_EMOJIS: 1 << 18,

  CONNECT: 1 << 20,
  SPEAK: 1 << 21,
  MUTE_MEMBERS: 1 << 22,
  DEAFEN_MEMBERS: 1 << 23,
  MOVE_MEMBERS: 1 << 24,
  USE_VAD: 1 << 25,

  CHANGE_NICKNAME: 1 << 26,
  MANAGE_NICKNAMES: 1 << 27,
  MANAGE_ROLES: 1 << 28,
  MANAGE_ROLES_OR_PERMISSIONS: 1 << 28,
  MANAGE_WEBHOOKS: 1 << 29,
  MANAGE_EMOJIS: 1 << 30,
***REMOVED***;

/**
 * Bitfield representing every permission combined
 * @type ***REMOVED***number***REMOVED***
 */
Permissions.ALL = Object.keys(Permissions.FLAGS).reduce((all, p) => all | Permissions.FLAGS[p], 0);

/**
 * Bitfield representing the default permissions for users
 * @type ***REMOVED***number***REMOVED***
 */
Permissions.DEFAULT = 104324673;

/**
 * @class EvaluatedPermissions
 * @classdesc The final evaluated permissions for a member in a channel
 * @see ***REMOVED***@link Permissions***REMOVED***
 * @deprecated
 */

Permissions.prototype.hasPermission = util.deprecate(Permissions.prototype.hasPermission,
  'EvaluatedPermissions#hasPermission is deprecated, use Permissions#has instead');
Permissions.prototype.hasPermissions = util.deprecate(Permissions.prototype.hasPermissions,
  'EvaluatedPermissions#hasPermissions is deprecated, use Permissions#has instead');
Permissions.prototype.missingPermissions = util.deprecate(Permissions.prototype.missingPermissions,
  'EvaluatedPermissions#missingPermissions is deprecated, use Permissions#missing instead');
Object.defineProperty(Permissions.prototype, 'raw', ***REMOVED***
  get: util
    .deprecate(Object.getOwnPropertyDescriptor(Permissions.prototype, 'raw').get,
      'EvaluatedPermissions#raw is deprecated use Permissions#bitfield instead'),
  set: util.deprecate(Object.getOwnPropertyDescriptor(Permissions.prototype, 'raw').set,
    'EvaluatedPermissions#raw is deprecated use Permissions#bitfield instead'),
***REMOVED***);
Object.defineProperty(Permissions.prototype, 'member', ***REMOVED***
  get: util
    .deprecate(Object.getOwnPropertyDescriptor(Permissions.prototype, 'member').get,
      'EvaluatedPermissions#member is deprecated'),
  set: util
    .deprecate(Object.getOwnPropertyDescriptor(Permissions.prototype, 'member').set,
      'EvaluatedPermissions#member is deprecated'),
***REMOVED***);

module.exports = Permissions;
