// eslint-plugin-node/lib/rules/prefer-global/console.js
/**
 * @author Toru Nagashima
 * See LICENSE file in root directory for full license.
 */
"use strict"

const ***REMOVED*** READ ***REMOVED*** = require("eslint-utils")
const checkForPreferGlobal = require("../../util/check-prefer-global")

const trackMap = ***REMOVED***
    globals: ***REMOVED***
        console: ***REMOVED*** [READ]: true ***REMOVED***,
    ***REMOVED***,
    modules: ***REMOVED***
        console: ***REMOVED*** [READ]: true ***REMOVED***,
    ***REMOVED***,
***REMOVED***

module.exports = ***REMOVED***
    meta: ***REMOVED***
        docs: ***REMOVED***
            description: 'enforce either `console` or `require("console")`',
            category: "Stylistic Issues",
            recommended: false,
            url:
                "https://github.com/mysticatea/eslint-plugin-node/blob/v10.0.0/docs/rules/prefer-global/console.md",
        ***REMOVED***,
        type: "suggestion",
        fixable: null,
        schema: [***REMOVED*** enum: ["always", "never"] ***REMOVED***],
        messages: ***REMOVED***
            preferGlobal:
                "Unexpected use of 'require(\"console\")'. Use the global variable 'console' instead.",
            preferModule:
                "Unexpected use of the global variable 'console'. Use 'require(\"console\")' instead.",
        ***REMOVED***,
    ***REMOVED***,

    create(context) ***REMOVED***
        return ***REMOVED***
            "Program:exit"() ***REMOVED***
                checkForPreferGlobal(context, trackMap)
            ***REMOVED***,
        ***REMOVED***
    ***REMOVED***,
***REMOVED***
