// es-abstract/test/helpers/values.js
'use strict';

var assign = require('../../helpers/assign');

var hasSymbols = require('has-symbols')();

var coercibleObject = ***REMOVED*** valueOf: function () ***REMOVED*** return 3; ***REMOVED***, toString: function () ***REMOVED*** return 42; ***REMOVED*** ***REMOVED***;
var coercibleFnObject = ***REMOVED***
	valueOf: function () ***REMOVED*** return function valueOfFn() ***REMOVED******REMOVED***; ***REMOVED***,
	toString: function () ***REMOVED*** return 42; ***REMOVED***
***REMOVED***;
var valueOfOnlyObject = ***REMOVED*** valueOf: function () ***REMOVED*** return 4; ***REMOVED***, toString: function () ***REMOVED*** return ***REMOVED******REMOVED***; ***REMOVED*** ***REMOVED***;
var toStringOnlyObject = ***REMOVED*** valueOf: function () ***REMOVED*** return ***REMOVED******REMOVED***; ***REMOVED***, toString: function () ***REMOVED*** return 7; ***REMOVED*** ***REMOVED***;
var uncoercibleObject = ***REMOVED*** valueOf: function () ***REMOVED*** return ***REMOVED******REMOVED***; ***REMOVED***, toString: function () ***REMOVED*** return ***REMOVED******REMOVED***; ***REMOVED*** ***REMOVED***;
var uncoercibleFnObject = ***REMOVED***
	valueOf: function () ***REMOVED*** return function valueOfFn() ***REMOVED******REMOVED***; ***REMOVED***,
	toString: function () ***REMOVED*** return function toStrFn() ***REMOVED******REMOVED***; ***REMOVED***
***REMOVED***;
var objects = [***REMOVED******REMOVED***, coercibleObject, coercibleFnObject, toStringOnlyObject, valueOfOnlyObject];
var nullPrimitives = [undefined, null];
var nonIntegerNumbers = [-1.3, 0.2, 1.8, 1 / 3];
var zeroes = [0, -0];
var infinities = [Infinity, -Infinity];
var numbers = zeroes.concat([42], infinities, nonIntegerNumbers);
var strings = ['', 'foo', 'a\uD83D\uDCA9c'];
var booleans = [true, false];
var symbols = hasSymbols ? [Symbol.iterator, Symbol('foo')] : [];
var nonSymbolPrimitives = [].concat(nullPrimitives, booleans, strings, numbers);
var nonNumberPrimitives = [].concat(nullPrimitives, booleans, strings, symbols);
var nonNullPrimitives = [].concat(booleans, strings, numbers, symbols);
var nonUndefinedPrimitives = [].concat(null, nonNullPrimitives);
var nonStrings = [].concat(nullPrimitives, booleans, numbers, symbols, objects);
var primitives = [].concat(nullPrimitives, nonNullPrimitives);
var nonPropertyKeys = [].concat(nullPrimitives, booleans, numbers, objects);
var propertyKeys = [].concat(strings, symbols);
var nonBooleans = [].concat(nullPrimitives, strings, symbols, numbers, objects);
var falsies = [].concat(nullPrimitives, false, '', 0, -0, NaN);
var truthies = [].concat(true, 'foo', 42, symbols, objects);
var timestamps = [].concat(0, 946713600000, 1546329600000);
var nonFunctions = [].concat(primitives, objects, [42]);
var nonArrays = [].concat(nonFunctions);

var descriptors = ***REMOVED***
	configurable: function (descriptor) ***REMOVED***
		return assign(assign(***REMOVED******REMOVED***, descriptor), ***REMOVED*** '[[Configurable]]': true ***REMOVED***);
	***REMOVED***,
	nonConfigurable: function (descriptor) ***REMOVED***
		return assign(assign(***REMOVED******REMOVED***, descriptor), ***REMOVED*** '[[Configurable]]': false ***REMOVED***);
	***REMOVED***,
	enumerable: function (descriptor) ***REMOVED***
		return assign(assign(***REMOVED******REMOVED***, descriptor), ***REMOVED*** '[[Enumerable]]': true ***REMOVED***);
	***REMOVED***,
	nonEnumerable: function (descriptor) ***REMOVED***
		return assign(assign(***REMOVED******REMOVED***, descriptor), ***REMOVED*** '[[Enumerable]]': false ***REMOVED***);
	***REMOVED***,
	writable: function (descriptor) ***REMOVED***
		return assign(assign(***REMOVED******REMOVED***, descriptor), ***REMOVED*** '[[Writable]]': true ***REMOVED***);
	***REMOVED***,
	nonWritable: function (descriptor) ***REMOVED***
		return assign(assign(***REMOVED******REMOVED***, descriptor), ***REMOVED*** '[[Writable]]': false ***REMOVED***);
	***REMOVED***
***REMOVED***;

module.exports = ***REMOVED***
	coercibleObject: coercibleObject,
	coercibleFnObject: coercibleFnObject,
	valueOfOnlyObject: valueOfOnlyObject,
	toStringOnlyObject: toStringOnlyObject,
	uncoercibleObject: uncoercibleObject,
	uncoercibleFnObject: uncoercibleFnObject,
	objects: objects,
	nonFunctions: nonFunctions,
	nonArrays: nonArrays,
	nullPrimitives: nullPrimitives,
	numbers: numbers,
	zeroes: zeroes,
	infinities: infinities,
	strings: strings,
	booleans: booleans,
	symbols: symbols,
	hasSymbols: hasSymbols,
	nonSymbolPrimitives: nonSymbolPrimitives,
	nonNumberPrimitives: nonNumberPrimitives,
	nonNullPrimitives: nonNullPrimitives,
	nonUndefinedPrimitives: nonUndefinedPrimitives,
	nonStrings: nonStrings,
	nonNumbers: nonNumberPrimitives.concat(objects),
	nonIntegerNumbers: nonIntegerNumbers,
	primitives: primitives,
	nonPropertyKeys: nonPropertyKeys,
	propertyKeys: propertyKeys,
	nonBooleans: nonBooleans,
	falsies: falsies,
	truthies: truthies,
	timestamps: timestamps,
	bothDescriptor: function () ***REMOVED***
		return ***REMOVED*** '[[Get]]': function () ***REMOVED******REMOVED***, '[[Value]]': true ***REMOVED***;
	***REMOVED***,
	bothDescriptorWritable: function () ***REMOVED***
		return descriptors.writable(***REMOVED*** '[[Get]]': function () ***REMOVED******REMOVED*** ***REMOVED***);
	***REMOVED***,
	accessorDescriptor: function (value) ***REMOVED***
		return descriptors.enumerable(descriptors.configurable(***REMOVED***
			'[[Get]]': function get() ***REMOVED*** return value; ***REMOVED***
		***REMOVED***));
	***REMOVED***,
	mutatorDescriptor: function () ***REMOVED***
		return descriptors.enumerable(descriptors.configurable(***REMOVED***
			'[[Set]]': function () ***REMOVED******REMOVED***
		***REMOVED***));
	***REMOVED***,
	dataDescriptor: function (value) ***REMOVED***
		return descriptors.nonWritable(***REMOVED***
			'[[Value]]': arguments.length > 0 ? value : 42
		***REMOVED***);
	***REMOVED***,
	genericDescriptor: function () ***REMOVED***
		return descriptors.configurable(descriptors.nonEnumerable());
	***REMOVED***,
	descriptors: descriptors
***REMOVED***;
