// jsx-ast-utils/src/values/expressions/index.js
import Literal from '../Literal';
import JSXElement from '../JSXElement';
import Identifier from './Identifier';
import TaggedTemplateExpression from './TaggedTemplateExpression';
import TemplateLiteral from './TemplateLiteral';
import FunctionExpression from './FunctionExpression';
import LogicalExpression from './LogicalExpression';
import MemberExpression from './MemberExpression';
import OptionalMemberExpression from './OptionalMemberExpression';
import CallExpression from './CallExpression';
import UnaryExpression from './UnaryExpression';
import ThisExpression from './ThisExpression';
import ConditionalExpression from './ConditionalExpression';
import BinaryExpression from './BinaryExpression';
import ObjectExpression from './ObjectExpression';
import NewExpression from './NewExpression';
import UpdateExpression from './UpdateExpression';
import ArrayExpression from './ArrayExpression';
import BindExpression from './BindExpression';
import SpreadElement from './SpreadElement';
import TypeCastExpression from './TypeCastExpression';

// Composition map of types to their extractor functions.
const TYPES = ***REMOVED***
  Identifier,
  Literal,
  JSXElement,
  TaggedTemplateExpression,
  TemplateLiteral,
  ArrowFunctionExpression: FunctionExpression,
  FunctionExpression,
  LogicalExpression,
  MemberExpression,
  OptionalMemberExpression,
  CallExpression,
  UnaryExpression,
  ThisExpression,
  ConditionalExpression,
  BinaryExpression,
  ObjectExpression,
  NewExpression,
  UpdateExpression,
  ArrayExpression,
  BindExpression,
  SpreadElement,
  TypeCastExpression,
***REMOVED***;

const noop = () => null;

const errorMessage = expression => `The prop value with an expression type of $***REMOVED***expression***REMOVED*** could not be resolved. Please file issue to get this fixed immediately.`;

/**
 * This function maps an AST value node
 * to its correct extractor function for its
 * given type.
 *
 * This will map correctly for *all* possible expression types.
 *
 * @param - value - AST Value object with type `JSXExpressionContainer`
 * @returns The extracted value.
 */
export default function extract(value) ***REMOVED***
  // Value will not have the expression property when we recurse.
  // The type for expression on ArrowFunctionExpression is a boolean.
  let expression;
  if (
    typeof value.expression !== 'boolean'
    && value.expression
  ) ***REMOVED***
    expression = value.expression; // eslint-disable-line prefer-destructuring
  ***REMOVED*** else ***REMOVED***
    expression = value;
  ***REMOVED***
  let ***REMOVED*** type ***REMOVED*** = expression;

  while (type === 'TSNonNullExpression' || type === 'TSAsExpression') ***REMOVED***
    (***REMOVED*** type ***REMOVED*** = expression);
    if (expression.expression) ***REMOVED***
      (***REMOVED*** expression ***REMOVED*** = expression);
    ***REMOVED***
  ***REMOVED***

  if (TYPES[type] === undefined) ***REMOVED***
    // eslint-disable-next-line no-console
    console.error(errorMessage(type));
    return null;
  ***REMOVED***

  return TYPES[type](expression);
***REMOVED***

// Composition map of types to their extractor functions to handle literals.
const LITERAL_TYPES = Object.assign(***REMOVED******REMOVED***, TYPES, ***REMOVED***
  Literal: (value) => ***REMOVED***
    const extractedVal = TYPES.Literal.call(undefined, value);
    const isNull = extractedVal === null;
    // This will be convention for attributes that have null
    // value explicitly defined (<div prop=***REMOVED***null***REMOVED*** /> maps to 'null').
    return isNull ? 'null' : extractedVal;
  ***REMOVED***,
  Identifier: (value) => ***REMOVED***
    const isUndefined = TYPES.Identifier.call(undefined, value) === undefined;
    return isUndefined ? undefined : null;
  ***REMOVED***,
  JSXElement: noop,
  ArrowFunctionExpression: noop,
  FunctionExpression: noop,
  LogicalExpression: noop,
  MemberExpression: noop,
  OptionalMemberExpression: noop,
  CallExpression: noop,
  UnaryExpression: (value) => ***REMOVED***
    const extractedVal = TYPES.UnaryExpression.call(undefined, value);
    return extractedVal === undefined ? null : extractedVal;
  ***REMOVED***,
  UpdateExpression: (value) => ***REMOVED***
    const extractedVal = TYPES.UpdateExpression.call(undefined, value);
    return extractedVal === undefined ? null : extractedVal;
  ***REMOVED***,
  ThisExpression: noop,
  ConditionalExpression: noop,
  BinaryExpression: noop,
  ObjectExpression: noop,
  NewExpression: noop,
  ArrayExpression: (value) => ***REMOVED***
    const extractedVal = TYPES.ArrayExpression.call(undefined, value);
    return extractedVal.filter(val => val !== null);
  ***REMOVED***,
  BindExpression: noop,
  SpreadElement: noop,
  TSNonNullExpression: noop,
  TSAsExpression: noop,
  TypeCastExpression: noop,
***REMOVED***);

/**
 * This function maps an AST value node
 * to its correct extractor function for its
 * given type.
 *
 * This will map correctly for *some* possible types that map to literals.
 *
 * @param - value - AST Value object with type `JSXExpressionContainer`
 * @returns The extracted value.
 */
export function extractLiteral(value) ***REMOVED***
  // Value will not have the expression property when we recurse.
  const expression = value.expression || value;
  const ***REMOVED*** type ***REMOVED*** = expression;

  if (LITERAL_TYPES[type] === undefined) ***REMOVED***
    // eslint-disable-next-line no-console
    console.error(errorMessage(type));
    return null;
  ***REMOVED***

  return LITERAL_TYPES[type](expression);
***REMOVED***
