// discord.js/src/structures/DMChannel.js
const Channel = require('./Channel');
const TextBasedChannel = require('./interfaces/TextBasedChannel');
const Collection = require('../util/Collection');

/**
 * Represents a direct message channel between two users.
 * @extends ***REMOVED***Channel***REMOVED***
 * @implements ***REMOVED***TextBasedChannel***REMOVED***
 */
class DMChannel extends Channel ***REMOVED***
  constructor(client, data) ***REMOVED***
    super(client, data);
    this.type = 'dm';
    this.messages = new Collection();
    this._typing = new Map();
  ***REMOVED***

  setup(data) ***REMOVED***
    super.setup(data);

    /**
     * The recipient on the other end of the DM
     * @type ***REMOVED***User***REMOVED***
     */
    this.recipient = this.client.dataManager.newUser(data.recipients[0]);

    /**
     * The ID of the last message in the channel, if one was sent
     * @type ***REMOVED***?Snowflake***REMOVED***
     */
    this.lastMessageID = data.last_message_id;

    /**
     * The timestamp when the last pinned message was pinned, if there was one
     * @type ***REMOVED***?number***REMOVED***
     */
    this.lastPinTimestamp = data.last_pin_timestamp ? new Date(data.last_pin_timestamp).getTime() : null;
  ***REMOVED***

  /**
   * When concatenated with a string, this automatically concatenates the recipient's mention instead of the
   * DM channel object.
   * @returns ***REMOVED***string***REMOVED***
   */
  toString() ***REMOVED***
    return this.recipient.toString();
  ***REMOVED***

  // These are here only for documentation purposes - they are implemented by TextBasedChannel
  /* eslint-disable no-empty-function */
  get lastPinAt() ***REMOVED******REMOVED***
  send() ***REMOVED******REMOVED***
  sendMessage() ***REMOVED******REMOVED***
  sendEmbed() ***REMOVED******REMOVED***
  sendFile() ***REMOVED******REMOVED***
  sendFiles() ***REMOVED******REMOVED***
  sendCode() ***REMOVED******REMOVED***
  fetchMessage() ***REMOVED******REMOVED***
  fetchMessages() ***REMOVED******REMOVED***
  fetchPinnedMessages() ***REMOVED******REMOVED***
  search() ***REMOVED******REMOVED***
  startTyping() ***REMOVED******REMOVED***
  stopTyping() ***REMOVED******REMOVED***
  get typing() ***REMOVED******REMOVED***
  get typingCount() ***REMOVED******REMOVED***
  createCollector() ***REMOVED******REMOVED***
  createMessageCollector() ***REMOVED******REMOVED***
  awaitMessages() ***REMOVED******REMOVED***
  // Doesn't work on DM channels; bulkDelete() ***REMOVED******REMOVED***
  acknowledge() ***REMOVED******REMOVED***
  _cacheMessage() ***REMOVED******REMOVED***
***REMOVED***

TextBasedChannel.applyToClass(DMChannel, true, ['bulkDelete']);

module.exports = DMChannel;
