// discord.js/src/structures/GuildAuditLogs.js
const Collection = require('../util/Collection');
const Snowflake = require('../util/Snowflake');
const Webhook = require('./Webhook');
const Integration = require('./Integration');
const Invite = require('./Invite');

/**
 * The target type of an entry, e.g. `GUILD`. Here are the available types:
 * * GUILD
 * * CHANNEL
 * * USER
 * * ROLE
 * * INVITE
 * * WEBHOOK
 * * EMOJI
 * * MESSAGE
 * * INTEGRATION
 * @typedef ***REMOVED***string***REMOVED*** AuditLogTargetType
 */

/**
 * Key mirror of all available audit log targets.
 * @name GuildAuditLogs.Targets
 * @type ***REMOVED***AuditLogTargetType***REMOVED***
 */
const Targets = ***REMOVED***
  ALL: 'ALL',
  GUILD: 'GUILD',
  CHANNEL: 'CHANNEL',
  USER: 'USER',
  ROLE: 'ROLE',
  INVITE: 'INVITE',
  WEBHOOK: 'WEBHOOK',
  EMOJI: 'EMOJI',
  MESSAGE: 'MESSAGE',
  INTEGRATION: 'INTEGRATION',
  UNKNOWN: 'UNKNOWN',
***REMOVED***;

/**
 * The action of an entry. Here are the available actions:
 * * ALL: null
 * * GUILD_UPDATE: 1
 * * CHANNEL_CREATE: 10
 * * CHANNEL_UPDATE: 11
 * * CHANNEL_DELETE: 12
 * * CHANNEL_OVERWRITE_CREATE: 13
 * * CHANNEL_OVERWRITE_UPDATE: 14
 * * CHANNEL_OVERWRITE_DELETE: 15
 * * MEMBER_KICK: 20
 * * MEMBER_PRUNE: 21
 * * MEMBER_BAN_ADD: 22
 * * MEMBER_BAN_REMOVE: 23
 * * MEMBER_UPDATE: 24
 * * MEMBER_ROLE_UPDATE: 25
 * * MEMBER_MOVE: 26
 * * MEMBER_DISCONNECT: 27
 * * BOT_ADD: 28,
 * * ROLE_CREATE: 30
 * * ROLE_UPDATE: 31
 * * ROLE_DELETE: 32
 * * INVITE_CREATE: 40
 * * INVITE_UPDATE: 41
 * * INVITE_DELETE: 42
 * * WEBHOOK_CREATE: 50
 * * WEBHOOK_UPDATE: 51
 * * WEBHOOK_DELETE: 52
 * * EMOJI_CREATE: 60
 * * EMOJI_UPDATE: 61
 * * EMOJI_DELETE: 62
 * * MESSAGE_DELETE: 72
 * * MESSAGE_BULK_DELETE: 73
 * * MESSAGE_PIN: 74
 * * MESSAGE_UNPIN: 75
 * * INTEGRATION_CREATE: 80
 * * INTEGRATION_UPDATE: 81
 * * INTEGRATION_DELETE: 82
 * @typedef ***REMOVED***?number|string***REMOVED*** AuditLogAction
 */

/**
 * All available actions keyed under their names to their numeric values.
 * @name GuildAuditLogs.Actions
 * @type ***REMOVED***AuditLogAction***REMOVED***
 */
const Actions = ***REMOVED***
  ALL: null,
  GUILD_UPDATE: 1,
  CHANNEL_CREATE: 10,
  CHANNEL_UPDATE: 11,
  CHANNEL_DELETE: 12,
  CHANNEL_OVERWRITE_CREATE: 13,
  CHANNEL_OVERWRITE_UPDATE: 14,
  CHANNEL_OVERWRITE_DELETE: 15,
  MEMBER_KICK: 20,
  MEMBER_PRUNE: 21,
  MEMBER_BAN_ADD: 22,
  MEMBER_BAN_REMOVE: 23,
  MEMBER_UPDATE: 24,
  MEMBER_ROLE_UPDATE: 25,
  MEMBER_MOVE: 26,
  MEMBER_DISCONNECT: 27,
  BOT_ADD: 28,
  ROLE_CREATE: 30,
  ROLE_UPDATE: 31,
  ROLE_DELETE: 32,
  INVITE_CREATE: 40,
  INVITE_UPDATE: 41,
  INVITE_DELETE: 42,
  WEBHOOK_CREATE: 50,
  WEBHOOK_UPDATE: 51,
  WEBHOOK_DELETE: 52,
  EMOJI_CREATE: 60,
  EMOJI_UPDATE: 61,
  EMOJI_DELETE: 62,
  MESSAGE_DELETE: 72,
  MESSAGE_BULK_DELETE: 73,
  MESSAGE_PIN: 74,
  MESSAGE_UNPIN: 75,
  INTEGRATION_CREATE: 80,
  INTEGRATION_UPDATE: 81,
  INTEGRATION_DELETE: 82,
***REMOVED***;


/**
 * Audit logs entries are held in this class.
 */
class GuildAuditLogs ***REMOVED***
  constructor(guild, data) ***REMOVED***
    if (data.users) for (const user of data.users) guild.client.dataManager.newUser(user);

    /**
     * Cached webhooks
     * @type ***REMOVED***Collection<Snowflake, Webhook>***REMOVED***
     * @private
     */
    this.webhooks = new Collection();
    if (data.webhooks) ***REMOVED***
      for (const hook of data.webhooks) ***REMOVED***
        this.webhooks.set(hook.id, new Webhook(guild.client, hook));
      ***REMOVED***
    ***REMOVED***

    /**
     * Cached integrations
     * @type ***REMOVED***Collection<Snowflake, Integration>***REMOVED***
     * @private
     */
    this.integrations = new Collection();
    if (data.integrations) ***REMOVED***
      for (const integration of data.integrations) ***REMOVED***
        this.integrations.set(integration.id, new Integration(guild.client, integration, guild));
      ***REMOVED***
    ***REMOVED***

    /**
     * The entries for this guild's audit logs
     * @type ***REMOVED***Collection<Snowflake, GuildAuditLogsEntry>***REMOVED***
     */
    this.entries = new Collection();
    for (const item of data.audit_log_entries) ***REMOVED***
      const entry = new GuildAuditLogsEntry(this, guild, item);
      this.entries.set(entry.id, entry);
    ***REMOVED***
  ***REMOVED***

  /**
   * Handles possible promises for entry targets.
   * @returns ***REMOVED***Promise<GuildAuditLogs>***REMOVED***
   */
  static build(...args) ***REMOVED***
    const logs = new GuildAuditLogs(...args);
    return Promise.all(logs.entries.map(e => e.target)).then(() => logs);
  ***REMOVED***

  /**
   * The target of an entry. It can be one of:
   * * A guild
   * * A user
   * * A role
   * * An emoji
   * * An invite
   * * A webhook
   * * An integration
   * * An object with an id key if target was deleted
   * * An object where the keys represent either the new value or the old value
   * @typedef ***REMOVED***?Object|Guild|User|Role|Emoji|Invite|Webhook|Integration***REMOVED*** AuditLogEntryTarget
   */

  /**
   * Find target type from entry action.
   * @param ***REMOVED***number***REMOVED*** target The action target
   * @returns ***REMOVED***?string***REMOVED***
   */
  static targetType(target) ***REMOVED***
    if (target < 10) return Targets.GUILD;
    if (target < 20) return Targets.CHANNEL;
    if (target < 30) return Targets.USER;
    if (target < 40) return Targets.ROLE;
    if (target < 50) return Targets.INVITE;
    if (target < 60) return Targets.WEBHOOK;
    if (target < 70) return Targets.EMOJI;
    if (target < 80) return Targets.MESSAGE;
    if (target < 90) return Targets.INTEGRATION;
    return null;
  ***REMOVED***

  /**
   * The action type of an entry, e.g. `CREATE`. Here are the available types:
   * * CREATE
   * * DELETE
   * * UPDATE
   * * ALL
   * @typedef ***REMOVED***string***REMOVED*** AuditLogActionType
   */


  /**
   * Finds the action type from the entry action.
   * @param ***REMOVED***AuditLogAction***REMOVED*** action The action target
   * @returns ***REMOVED***AuditLogActionType***REMOVED***
   */
  static actionType(action) ***REMOVED***
    if ([
      Actions.CHANNEL_CREATE,
      Actions.CHANNEL_OVERWRITE_CREATE,
      Actions.MEMBER_BAN_REMOVE,
      Actions.BOT_ADD,
      Actions.ROLE_CREATE,
      Actions.INVITE_CREATE,
      Actions.WEBHOOK_CREATE,
      Actions.EMOJI_CREATE,
      Actions.MESSAGE_PIN,
      Actions.INTEGRATION_CREATE,
    ].includes(action)) return 'CREATE';

    if ([
      Actions.CHANNEL_DELETE,
      Actions.CHANNEL_OVERWRITE_DELETE,
      Actions.MEMBER_KICK,
      Actions.MEMBER_PRUNE,
      Actions.MEMBER_BAN_ADD,
      Actions.MEMBER_DISCONNECT,
      Actions.ROLE_DELETE,
      Actions.INVITE_DELETE,
      Actions.WEBHOOK_DELETE,
      Actions.EMOJI_DELETE,
      Actions.MESSAGE_DELETE,
      Actions.MESSAGE_BULK_DELETE,
      Actions.MESSAGE_UNPIN,
      Actions.INTEGRATION_DELETE,
    ].includes(action)) return 'DELETE';

    if ([
      Actions.GUILD_UPDATE,
      Actions.CHANNEL_UPDATE,
      Actions.CHANNEL_OVERWRITE_UPDATE,
      Actions.MEMBER_UPDATE,
      Actions.MEMBER_ROLE_UPDATE,
      Actions.MEMBER_MOVE,
      Actions.ROLE_UPDATE,
      Actions.INVITE_UPDATE,
      Actions.WEBHOOK_UPDATE,
      Actions.EMOJI_UPDATE,
      Actions.INTEGRATION_UPDATE,
    ].includes(action)) return 'UPDATE';

    return 'ALL';
  ***REMOVED***
***REMOVED***

/**
 * Audit logs entry.
 */
class GuildAuditLogsEntry ***REMOVED***
  // eslint-disable-next-line complexity
  constructor(logs, guild, data) ***REMOVED***
    const targetType = GuildAuditLogs.targetType(data.action_type);
    /**
     * The target type of this entry
     * @type ***REMOVED***AuditLogTargetType***REMOVED***
     */
    this.targetType = targetType;

    /**
     * The action type of this entry
     * @type ***REMOVED***AuditLogActionType***REMOVED***
     */
    this.actionType = GuildAuditLogs.actionType(data.action_type);

    /**
     * Specific action type of this entry in its string representation
     * @type ***REMOVED***AuditLogAction***REMOVED***
     */
    this.action = Object.keys(Actions).find(k => Actions[k] === data.action_type);

    /**
     * The reason of this entry
     * @type ***REMOVED***?string***REMOVED***
     */
    this.reason = data.reason || null;

    /**
     * The user that executed this entry
     * @type ***REMOVED***User***REMOVED***
     */
    this.executor = guild.client.users.get(data.user_id);

    /**
     * An entry in the audit log representing a specific change.
     * @typedef ***REMOVED***object***REMOVED*** AuditLogChange
     * @property ***REMOVED***string***REMOVED*** key The property that was changed, e.g. `nick` for nickname changes
     * @property ***REMOVED*******REMOVED*** [old] The old value of the change, e.g. for nicknames, the old nickname
     * @property ***REMOVED*******REMOVED*** [new] The new value of the change, e.g. for nicknames, the new nickname
     */

    /**
     * Specific property changes
     * @type ***REMOVED***AuditLogChange[]***REMOVED***
     */
    this.changes = data.changes ? data.changes.map(c => (***REMOVED*** key: c.key, old: c.old_value, new: c.new_value ***REMOVED***)) : null;

    /**
     * The ID of this entry
     * @type ***REMOVED***Snowflake***REMOVED***
     */
    this.id = data.id;

    /**
     * Any extra data from the entry
     * @type ***REMOVED***?Object|Role|GuildMember***REMOVED***
     */
    this.extra = null;
    switch (data.action_type) ***REMOVED***
      case Actions.MEMBER_PRUNE:
        this.extra = ***REMOVED***
          removed: Number(data.options.members_removed),
          days: Number(data.options.delete_member_days),
        ***REMOVED***;
        break;

      case Actions.MEMBER_MOVE:
      case Actions.MESSAGE_DELETE:
      case Actions.MESSAGE_BULK_DELETE:
        this.extra = ***REMOVED***
          channel: guild.channels.get(data.options.channel_id) || ***REMOVED*** id: data.options.channel_id ***REMOVED***,
          count: Number(data.options.count),
        ***REMOVED***;
        break;

      case Actions.MESSAGE_PIN:
      case Actions.MESSAGE_UNPIN:
        this.extra = ***REMOVED***
          channel: guild.client.channels.get(data.options.channel_id) || ***REMOVED*** id: data.options.channel_id ***REMOVED***,
          messageID: data.options.message_id,
        ***REMOVED***;
        break;

      case Actions.MEMBER_DISCONNECT:
        this.extra = ***REMOVED***
          count: Number(data.options.count),
        ***REMOVED***;
        break;

      case Actions.CHANNEL_OVERWRITE_CREATE:
      case Actions.CHANNEL_OVERWRITE_UPDATE:
      case Actions.CHANNEL_OVERWRITE_DELETE:
        switch (data.options.type) ***REMOVED***
          case 'member':
            this.extra = guild.members.get(data.options.id) ||
              ***REMOVED*** id: data.options.id, type: 'member' ***REMOVED***;
            break;
          case 'role':
            this.extra = guild.roles.get(data.options.id) ||
              ***REMOVED*** id: data.options.id, name: data.options.role_name, type: 'role' ***REMOVED***;
            break;
          default:
            break;
        ***REMOVED***
        break;

      default:
        break;
    ***REMOVED***

    /**
     * The target of this entry
     * @type ***REMOVED***AuditLogEntryTarget***REMOVED***
     */
    this.target = null;
    if (targetType === Targets.UNKNOWN) ***REMOVED***
      this.changes.reduce((o, c) => ***REMOVED***
        o[c.key] = c.new || c.old;
        return o;
      ***REMOVED***, ***REMOVED******REMOVED***);
      this.target.id = data.target_id;
      // MEMBER_DISCONNECT and similar types do not provide a target_id.
    ***REMOVED*** else if (targetType === Targets.USER && data.target_id) ***REMOVED***
      this.target = guild.client.users.get(data.target_id);
    ***REMOVED*** else if (targetType === Targets.GUILD) ***REMOVED***
      this.target = guild.client.guilds.get(data.target_id);
    ***REMOVED*** else if (targetType === Targets.WEBHOOK) ***REMOVED***
      this.target = logs.webhooks.get(data.target_id) ||
        new Webhook(guild.client,
          this.changes.reduce((o, c) => ***REMOVED***
            o[c.key] = c.new || c.old;
            return o;
          ***REMOVED***, ***REMOVED***
            id: data.target_id,
            guild_id: guild.id,
          ***REMOVED***));
    ***REMOVED*** else if (targetType === Targets.INVITE) ***REMOVED***
      const changes = this.changes.reduce((o, c) => ***REMOVED***
        o[c.key] = c.new || c.old;
        return o;
      ***REMOVED***, ***REMOVED***
        id: data.target_id,
        guild,
      ***REMOVED***);
      changes.channel = ***REMOVED*** id: changes.channel_id ***REMOVED***;
      this.target = new Invite(guild.client, changes);
    ***REMOVED*** else if (targetType === Targets.MESSAGE) ***REMOVED***
      // Discord sends a channel id for the MESSAGE_BULK_DELETE action type.
      this.target = data.action_type === Actions.MESSAGE_BULK_DELETE ?
        guild.channels.get(data.target_id) || ***REMOVED*** id: data.target_id ***REMOVED*** :
        guild.client.users.get(data.target_id);
    ***REMOVED*** else if (targetType === Targets.INTEGRATION) ***REMOVED***
      this.target = logs.integrations.get(data.target_id) ||
        new Integration(guild.client, this.changes.reduce((o, c) => ***REMOVED***
          o[c.key] = c.new || c.old;
          return o;
        ***REMOVED***, ***REMOVED*** id: data.target_id ***REMOVED***), guild);
    ***REMOVED*** else if (data.target_id) ***REMOVED***
      this.target = guild[`$***REMOVED***targetType.toLowerCase()***REMOVED***s`].get(data.target_id) || ***REMOVED*** id: data.target_id ***REMOVED***;
    ***REMOVED***
  ***REMOVED***

  /**
   * The timestamp this entry was created at
   * @type ***REMOVED***number***REMOVED***
   * @readonly
   */
  get createdTimestamp() ***REMOVED***
    return Snowflake.deconstruct(this.id).timestamp;
  ***REMOVED***

  /**
   * The time this entry was created
   * @type ***REMOVED***Date***REMOVED***
   * @readonly
   */
  get createdAt() ***REMOVED***
    return new Date(this.createdTimestamp);
  ***REMOVED***
***REMOVED***

GuildAuditLogs.Actions = Actions;
GuildAuditLogs.Targets = Targets;
GuildAuditLogs.Entry = GuildAuditLogsEntry;

module.exports = GuildAuditLogs;
