// discord.js/src/structures/MessageAttachment.js
const ***REMOVED*** basename ***REMOVED*** = require('path');

/**
 * Represents an attachment in a message.
 */
class MessageAttachment ***REMOVED***
  constructor(message, data) ***REMOVED***
    /**
     * The client that instantiated this MessageAttachment
     * @name MessageAttachment#client
     * @type ***REMOVED***Client***REMOVED***
     * @readonly
     */
    Object.defineProperty(this, 'client', ***REMOVED*** value: message.client ***REMOVED***);

    /**
     * The message this attachment is part of
     * @type ***REMOVED***Message***REMOVED***
     */
    this.message = message;

    this.setup(data);
  ***REMOVED***

  setup(data) ***REMOVED***
    /**
     * The ID of this attachment
     * @type ***REMOVED***Snowflake***REMOVED***
     */
    this.id = data.id;

    /**
     * The file name of this attachment
     * @type ***REMOVED***string***REMOVED***
     */
    this.filename = data.filename;

    /**
     * The size of this attachment in bytes
     * @type ***REMOVED***number***REMOVED***
     */
    this.filesize = data.size;

    /**
     * The URL to this attachment
     * @type ***REMOVED***string***REMOVED***
     */
    this.url = data.url;

    /**
     * The Proxy URL to this attachment
     * @type ***REMOVED***string***REMOVED***
     */
    this.proxyURL = data.proxy_url;

    /**
     * The height of this attachment (if an image)
     * @type ***REMOVED***?number***REMOVED***
     */
    this.height = data.height;

    /**
     * The width of this attachment (if an image)
     * @type ***REMOVED***?number***REMOVED***
     */
    this.width = data.width;
  ***REMOVED***

  /**
   * Whether or not this attachment has been marked as a spoiler
   * @type ***REMOVED***boolean***REMOVED***
   * @readonly
   */
  get spoiler() ***REMOVED***
    return basename(this.url).startsWith('SPOILER_');
  ***REMOVED***
***REMOVED***

module.exports = MessageAttachment;
