// discord.js/src/structures/MessageEmbed.js
/**
 * Represents an embed in a message (image/video preview, rich embed, etc.)
 * <info>This class is only used for *received* embeds. If you wish to send one, use the ***REMOVED***@link RichEmbed***REMOVED*** class.</info>
 */
class MessageEmbed ***REMOVED***
  constructor(message, data) ***REMOVED***
    /**
     * The client that instantiated this embed
     * @name MessageEmbed#client
     * @type ***REMOVED***Client***REMOVED***
     * @readonly
     */
    Object.defineProperty(this, 'client', ***REMOVED*** value: message.client ***REMOVED***);

    /**
     * The message this embed is part of
     * @type ***REMOVED***Message***REMOVED***
     */
    this.message = message;

    this.setup(data);
  ***REMOVED***

  setup(data) ***REMOVED***
    /**
     * The type of this embed
     * @type ***REMOVED***string***REMOVED***
     */
    this.type = data.type;

    /**
     * The title of this embed
     * @type ***REMOVED***?string***REMOVED***
     */
    this.title = data.title;

    /**
     * The description of this embed
     * @type ***REMOVED***?string***REMOVED***
     */
    this.description = data.description;

    /**
     * The URL of this embed
     * @type ***REMOVED***string***REMOVED***
     */
    this.url = data.url;

    /**
     * The color of the embed
     * @type ***REMOVED***number***REMOVED***
     */
    this.color = data.color;

    /**
     * The fields of this embed
     * @type ***REMOVED***MessageEmbedField[]***REMOVED***
     */
    this.fields = [];
    if (data.fields) for (const field of data.fields) this.fields.push(new MessageEmbedField(this, field));

    /**
     * The timestamp of this embed
     * @type ***REMOVED***number***REMOVED***
     */
    this.timestamp = data.timestamp;

    /**
     * The thumbnail of this embed
     * @type ***REMOVED***?MessageEmbedThumbnail***REMOVED***
     */
    this.thumbnail = data.thumbnail ? new MessageEmbedThumbnail(this, data.thumbnail) : null;

    /**
     * The image of this embed
     * @type ***REMOVED***?MessageEmbedImage***REMOVED***
     */
    this.image = data.image ? new MessageEmbedImage(this, data.image) : null;

    /**
     * The video of this embed
     * @type ***REMOVED***?MessageEmbedVideo***REMOVED***
     */
    this.video = data.video ? new MessageEmbedVideo(this, data.video) : null;

    /**
     * The author of this embed
     * @type ***REMOVED***?MessageEmbedAuthor***REMOVED***
     */
    this.author = data.author ? new MessageEmbedAuthor(this, data.author) : null;

    /**
     * The provider of this embed
     * @type ***REMOVED***?MessageEmbedProvider***REMOVED***
     */
    this.provider = data.provider ? new MessageEmbedProvider(this, data.provider) : null;

    /**
     * The footer of this embed
     * @type ***REMOVED***?MessageEmbedFooter***REMOVED***
     */
    this.footer = data.footer ? new MessageEmbedFooter(this, data.footer) : null;
  ***REMOVED***

  /**
   * The date this embed was created
   * @type ***REMOVED***Date***REMOVED***
   * @readonly
   */
  get createdAt() ***REMOVED***
    return new Date(this.createdTimestamp);
  ***REMOVED***

  /**
   * The hexadecimal version of the embed color, with a leading hash
   * @type ***REMOVED***?string***REMOVED***
   * @readonly
   */
  get hexColor() ***REMOVED***
    if (!this.color) return null;
    let col = this.color.toString(16);
    while (col.length < 6) col = `0$***REMOVED***col***REMOVED***`;
    return `#$***REMOVED***col***REMOVED***`;
  ***REMOVED***
***REMOVED***

/**
 * Represents a thumbnail for a message embed.
 */
class MessageEmbedThumbnail ***REMOVED***
  constructor(embed, data) ***REMOVED***
    /**
     * The embed this thumbnail is part of
     * @type ***REMOVED***MessageEmbed***REMOVED***
     */
    this.embed = embed;

    this.setup(data);
  ***REMOVED***

  setup(data) ***REMOVED***
    /**
     * The URL for this thumbnail
     * @type ***REMOVED***string***REMOVED***
     */
    this.url = data.url;

    /**
     * The Proxy URL for this thumbnail
     * @type ***REMOVED***string***REMOVED***
     */
    this.proxyURL = data.proxy_url;

    /**
     * The height of the thumbnail
     * @type ***REMOVED***number***REMOVED***
     */
    this.height = data.height;

    /**
     * The width of the thumbnail
     * @type ***REMOVED***number***REMOVED***
     */
    this.width = data.width;
  ***REMOVED***
***REMOVED***

/**
 * Represents an image for a message embed.
 */
class MessageEmbedImage ***REMOVED***
  constructor(embed, data) ***REMOVED***
    /**
     * The embed this image is part of
     * @type ***REMOVED***MessageEmbed***REMOVED***
     */
    this.embed = embed;

    this.setup(data);
  ***REMOVED***

  setup(data) ***REMOVED***
    /**
     * The URL for this image
     * @type ***REMOVED***string***REMOVED***
     */
    this.url = data.url;

    /**
     * The Proxy URL for this image
     * @type ***REMOVED***string***REMOVED***
     */
    this.proxyURL = data.proxy_url;

    /**
     * The height of the image
     * @type ***REMOVED***number***REMOVED***
     */
    this.height = data.height;

    /**
     * The width of the image
     * @type ***REMOVED***number***REMOVED***
     */
    this.width = data.width;
  ***REMOVED***
***REMOVED***

/**
 * Represents a video for a message embed.
 */
class MessageEmbedVideo ***REMOVED***
  constructor(embed, data) ***REMOVED***
    /**
     * The embed this video is part of
     * @type ***REMOVED***MessageEmbed***REMOVED***
     */
    this.embed = embed;

    this.setup(data);
  ***REMOVED***

  setup(data) ***REMOVED***
    /**
     * The source URL for this video
     * @type ***REMOVED***string***REMOVED***
     */
    this.url = data.url;

    /**
     * The height of the video
     * @type ***REMOVED***number***REMOVED***
     */
    this.height = data.height;

    /**
     * The width of the video
     * @type ***REMOVED***number***REMOVED***
     */
    this.width = data.width;
  ***REMOVED***
***REMOVED***

/**
 * Represents a provider for a message embed.
 */
class MessageEmbedProvider ***REMOVED***
  constructor(embed, data) ***REMOVED***
    /**
     * The embed this provider is part of
     * @type ***REMOVED***MessageEmbed***REMOVED***
     */
    this.embed = embed;

    this.setup(data);
  ***REMOVED***

  setup(data) ***REMOVED***
    /**
     * The name of this provider
     * @type ***REMOVED***string***REMOVED***
     */
    this.name = data.name;

    /**
     * The URL of this provider
     * @type ***REMOVED***string***REMOVED***
     */
    this.url = data.url;
  ***REMOVED***
***REMOVED***

/**
 * Represents an author for a message embed.
 */
class MessageEmbedAuthor ***REMOVED***
  constructor(embed, data) ***REMOVED***
    /**
     * The embed this author is part of
     * @type ***REMOVED***MessageEmbed***REMOVED***
     */
    this.embed = embed;

    this.setup(data);
  ***REMOVED***

  setup(data) ***REMOVED***
    /**
     * The name of this author
     * @type ***REMOVED***string***REMOVED***
     */
    this.name = data.name;

    /**
     * The URL of this author
     * @type ***REMOVED***string***REMOVED***
     */
    this.url = data.url;

    /**
     * The icon URL of this author
     * @type ***REMOVED***string***REMOVED***
     */
    this.iconURL = data.icon_url;
  ***REMOVED***
***REMOVED***

/**
 * Represents a field for a message embed.
 */
class MessageEmbedField ***REMOVED***
  constructor(embed, data) ***REMOVED***
    /**
     * The embed this footer is part of
     * @type ***REMOVED***MessageEmbed***REMOVED***
     */
    this.embed = embed;

    this.setup(data);
  ***REMOVED***

  setup(data) ***REMOVED***
    /**
     * The name of this field
     * @type ***REMOVED***string***REMOVED***
     */
    this.name = data.name;

    /**
     * The value of this field
     * @type ***REMOVED***string***REMOVED***
     */
    this.value = data.value;

    /**
     * If this field is displayed inline
     * @type ***REMOVED***boolean***REMOVED***
     */
    this.inline = data.inline;
  ***REMOVED***
***REMOVED***

/**
 * Represents the footer of a message embed.
 */
class MessageEmbedFooter ***REMOVED***
  constructor(embed, data) ***REMOVED***
    /**
     * The embed this footer is part of
     * @type ***REMOVED***MessageEmbed***REMOVED***
     */
    this.embed = embed;

    this.setup(data);
  ***REMOVED***

  setup(data) ***REMOVED***
    /**
     * The text in this footer
     * @type ***REMOVED***string***REMOVED***
     */
    this.text = data.text;

    /**
     * The icon URL of this footer
     * @type ***REMOVED***string***REMOVED***
     */
    this.iconURL = data.icon_url;

    /**
     * The proxy icon URL of this footer
     * @type ***REMOVED***string***REMOVED***
     */
    this.proxyIconUrl = data.proxy_icon_url;
  ***REMOVED***
***REMOVED***

MessageEmbed.Thumbnail = MessageEmbedThumbnail;
MessageEmbed.Image = MessageEmbedImage;
MessageEmbed.Video = MessageEmbedVideo;
MessageEmbed.Provider = MessageEmbedProvider;
MessageEmbed.Author = MessageEmbedAuthor;
MessageEmbed.Field = MessageEmbedField;
MessageEmbed.Footer = MessageEmbedFooter;

module.exports = MessageEmbed;
