// discord.js/src/structures/MessageMentions.js
const Collection = require('../util/Collection');
const ***REMOVED*** ChannelTypes ***REMOVED*** = require('../util/Constants');

/**
 * Keeps track of mentions in a ***REMOVED***@link Message***REMOVED***.
 */
class MessageMentions ***REMOVED***
  constructor(message, users, roles, everyone, crosspostedChannels) ***REMOVED***
    /**
     * Whether `@everyone` or `@here` were mentioned
     * @type ***REMOVED***boolean***REMOVED***
     */
    this.everyone = Boolean(everyone);

    if (users) ***REMOVED***
      if (users instanceof Collection) ***REMOVED***
        /**
         * Any users that were mentioned
         * <info>Order as received from the API, not as they appear in the message content</info>
         * @type ***REMOVED***Collection<Snowflake, User>***REMOVED***
         */
        this.users = new Collection(users);
      ***REMOVED*** else ***REMOVED***
        this.users = new Collection();
        for (const mention of users) ***REMOVED***
          let user = message.client.users.get(mention.id);
          if (!user) user = message.client.dataManager.newUser(mention);
          this.users.set(user.id, user);
          if (mention.member && message.guild && !message.guild.members.has(mention.id)) ***REMOVED***
            message.guild._addMember(Object.assign(mention.member, ***REMOVED*** user ***REMOVED***), false);
          ***REMOVED***
        ***REMOVED***
      ***REMOVED***
    ***REMOVED*** else ***REMOVED***
      this.users = new Collection();
    ***REMOVED***

    if (roles) ***REMOVED***
      if (roles instanceof Collection) ***REMOVED***
        /**
         * Any roles that were mentioned
         * <info>Order as received from the API, not as they appear in the message content</
         * @type ***REMOVED***Collection<Snowflake, Role>***REMOVED***
         */
        this.roles = new Collection(roles);
      ***REMOVED*** else ***REMOVED***
        this.roles = new Collection();
        for (const mention of roles) ***REMOVED***
          const role = message.channel.guild.roles.get(mention);
          if (role) this.roles.set(role.id, role);
        ***REMOVED***
      ***REMOVED***
    ***REMOVED*** else ***REMOVED***
      this.roles = new Collection();
    ***REMOVED***

    /**
     * Content of the message
     * @type ***REMOVED***Message***REMOVED***
     * @private
     */
    this._content = message.content;

    /**
     * The client the message is from
     * @type ***REMOVED***Client***REMOVED***
     * @private
     */
    this._client = message.client;

    /**
     * The guild the message is in
     * @type ***REMOVED***?Guild***REMOVED***
     * @private
     */
    this._guild = message.channel.guild;

    /**
     * Cached members for ***REMOVED***@MessageMention#members***REMOVED***
     * @type ***REMOVED***?Collection<Snowflake, GuildMember>***REMOVED***
     * @private
     */
    this._members = null;

    /**
     * Cached channels for ***REMOVED***@MessageMention#channels***REMOVED***
     * @type ***REMOVED***?Collection<Snowflake, GuildChannel>***REMOVED***
     * @private
     */
    this._channels = null;

    /**
     * Crossposted channel data.
     * @typedef ***REMOVED***Object***REMOVED*** CrosspostedChannel
     * @property ***REMOVED***Snowflake***REMOVED*** channelID ID of the mentioned channel
     * @property ***REMOVED***Snowflake***REMOVED*** guildID ID of the guild that has the channel
     * @property ***REMOVED***string***REMOVED*** type Type of the channel
     * @property ***REMOVED***string***REMOVED*** name Name of the channel
     */

    if (crosspostedChannels) ***REMOVED***
      if (crosspostedChannels instanceof Collection) ***REMOVED***
        /**
        * A collection of crossposted channels
        * @type ***REMOVED***Collection<Snowflake, CrosspostedChannel>***REMOVED***
        */
        this.crosspostedChannels = new Collection(crosspostedChannels);
      ***REMOVED*** else ***REMOVED***
        this.crosspostedChannels = new Collection();
        const channelTypes = Object.keys(ChannelTypes);
        for (const d of crosspostedChannels) ***REMOVED***
          const type = channelTypes[d.type];
          this.crosspostedChannels.set(d.id, ***REMOVED***
            channelID: d.id,
            guildID: d.guild_id,
            type: type ? type.toLowerCase() : 'unknown',
            name: d.name,
          ***REMOVED***);
        ***REMOVED***
      ***REMOVED***
    ***REMOVED*** else ***REMOVED***
      this.crosspostedChannels = new Collection();
    ***REMOVED***
  ***REMOVED***

  /**
   * Any members that were mentioned (only in ***REMOVED***@link TextChannel***REMOVED***s)
   * <info>Order as received from the API, not as they appear in the message content</
   * @type ***REMOVED***?Collection<Snowflake, GuildMember>***REMOVED***
   * @readonly
   */
  get members() ***REMOVED***
    if (this._members) return this._members;
    if (!this._guild) return null;
    this._members = new Collection();
    this.users.forEach(user => ***REMOVED***
      const member = this._guild.member(user);
      if (member) this._members.set(member.user.id, member);
    ***REMOVED***);
    return this._members;
  ***REMOVED***

  /**
   * Any channels that were mentioned
   * <info>Order as they appear first in the message content</info>
   * @type ***REMOVED***Collection<Snowflake, GuildChannel>***REMOVED***
   * @readonly
   */
  get channels() ***REMOVED***
    if (this._channels) return this._channels;
    this._channels = new Collection();
    let matches;
    while ((matches = this.constructor.CHANNELS_PATTERN.exec(this._content)) !== null) ***REMOVED***
      const chan = this._client.channels.get(matches[1]);
      if (chan) this._channels.set(chan.id, chan);
    ***REMOVED***
    return this._channels;
  ***REMOVED***
***REMOVED***

/**
 * Regular expression that globally matches `@everyone` and `@here`
 * @type ***REMOVED***RegExp***REMOVED***
 */
MessageMentions.EVERYONE_PATTERN = /@(everyone|here)/g;

/**
 * Regular expression that globally matches user mentions like `<@81440962496172032>`
 * @type ***REMOVED***RegExp***REMOVED***
 */
MessageMentions.USERS_PATTERN = /<@!?[0-9]+>/g;

/**
 * Regular expression that globally matches role mentions like `<@&297577916114403338>`
 * @type ***REMOVED***RegExp***REMOVED***
 */
MessageMentions.ROLES_PATTERN = /<@&[0-9]+>/g;

/**
 * Regular expression that globally matches channel mentions like `<#222079895583457280>`
 * @type ***REMOVED***RegExp***REMOVED***
 */
MessageMentions.CHANNELS_PATTERN = /<#([0-9]+)>/g;

module.exports = MessageMentions;
