// discord.js/src/structures/OAuth2Application.js
const Snowflake = require('../util/Snowflake');
const Team = require('./Team');
const util = require('util');

/**
 * Represents an OAuth2 Application.
 */
class OAuth2Application ***REMOVED***
  constructor(client, data) ***REMOVED***
    /**
     * The client that instantiated the application
     * @name OAuth2Application#client
     * @type ***REMOVED***Client***REMOVED***
     * @readonly
     */
    Object.defineProperty(this, 'client', ***REMOVED*** value: client ***REMOVED***);

    this.setup(data);
  ***REMOVED***

  setup(data) ***REMOVED***
    /**
     * The ID of the app
     * @type ***REMOVED***Snowflake***REMOVED***
     */
    this.id = data.id;

    /**
     * The name of the app
     * @type ***REMOVED***string***REMOVED***
     */
    this.name = data.name;

    /**
     * The app's description
     * @type ***REMOVED***string***REMOVED***
     */
    this.description = data.description;

    /**
     * The app's icon hash
     * @type ***REMOVED***?string***REMOVED***
     */
    this.icon = data.icon;

    /**
     * The app's icon URL
     * @type ***REMOVED***string***REMOVED***
     */
    this.iconURL = `https://cdn.discordapp.com/app-icons/$***REMOVED***this.id***REMOVED***/$***REMOVED***this.icon***REMOVED***.jpg`;

    /**
     * The app's RPC origins
     * @type ***REMOVED***?string[]***REMOVED***
     */
    this.rpcOrigins = data.rpc_origins;

    /**
     * The app's redirect URIs
     * @type ***REMOVED***string[]***REMOVED***
     */
    this.redirectURIs = data.redirect_uris;

    /**
     * If this app's bot requires a code grant when using the OAuth2 flow
     * @type ***REMOVED***boolean***REMOVED***
     */
    this.botRequireCodeGrant = data.bot_require_code_grant;

    /**
     * If this app's bot is public
     * @type ***REMOVED***boolean***REMOVED***
     */
    this.botPublic = data.bot_public;

    /**
     * If this app can use rpc
     * @type ***REMOVED***boolean***REMOVED***
     */
    this.rpcApplicationState = data.rpc_application_state;

    /**
     * Object containing basic info about this app's bot
     * @type ***REMOVED***Object***REMOVED***
     */
    this.bot = data.bot;

    /**
     * The flags for the app
     * @type ***REMOVED***number***REMOVED***
     */
    this.flags = data.flags;

    /**
     * OAuth2 secret for the application
     * @type ***REMOVED***boolean***REMOVED***
     */
    this.secret = data.secret;

    if (data.owner) ***REMOVED***
      /**
       * The owner of this OAuth application
       * @type ***REMOVED***?User***REMOVED***
       */
      this.owner = this.client.dataManager.newUser(data.owner);
    ***REMOVED***

    /**
     * The owning team of this OAuth application
     * <info>In v12.0.0 this property moves to `Team#owner`.</info>
     * @type ***REMOVED***?Team***REMOVED***
     * @deprecated
     */
    this.team = data.team ? new Team(this.client, data.team) : null;
  ***REMOVED***

  /**
   * The timestamp the app was created at
   * @type ***REMOVED***number***REMOVED***
   * @readonly
   */
  get createdTimestamp() ***REMOVED***
    return Snowflake.deconstruct(this.id).timestamp;
  ***REMOVED***

  /**
   * The time the app was created
   * @type ***REMOVED***Date***REMOVED***
   * @readonly
   */
  get createdAt() ***REMOVED***
    return new Date(this.createdTimestamp);
  ***REMOVED***

  /**
   * Reset the app's secret and bot token.
   * <warn>This is only available when using a user account.</warn>
   * @returns ***REMOVED***OAuth2Application***REMOVED***
   * @deprecated
   */
  reset() ***REMOVED***
    return this.client.rest.methods.resetApplication(this.id);
  ***REMOVED***

  /**
   * When concatenated with a string, this automatically concatenates the app name rather than the app object.
   * @returns ***REMOVED***string***REMOVED***
   */
  toString() ***REMOVED***
    return this.name;
  ***REMOVED***
***REMOVED***

OAuth2Application.prototype.reset =
  util.deprecate(OAuth2Application.prototype.reset, 'OAuth2Application#reset: userbot methods will be removed');

module.exports = OAuth2Application;
