// discord.js/src/structures/Presence.js
const ***REMOVED*** ActivityFlags, Endpoints ***REMOVED*** = require('../util/Constants');
const ReactionEmoji = require('./ReactionEmoji');

/**
 * The status of this presence:
 * * **`online`** - user is online
 * * **`idle`** - user is AFK
 * * **`offline`** - user is offline or invisible
 * * **`dnd`** - user is in Do Not Disturb
 * @typedef ***REMOVED***string***REMOVED*** PresenceStatus
 */

/**
 * The status of this presence:
 * * **`online`** - user is online
 * * **`idle`** - user is AFK
 * * **`dnd`** - user is in Do Not Disturb
 * @typedef ***REMOVED***string***REMOVED*** ClientPresenceStatus
 */

/**
 * Represents a user's presence.
 */
class Presence ***REMOVED***
  constructor(data = ***REMOVED******REMOVED***, client) ***REMOVED***
    /**
     * The client that instantiated this
     * @name Presence#client
     * @type ***REMOVED***Client***REMOVED***
     * @readonly
     */
    Object.defineProperty(this, 'client', ***REMOVED*** value: client ***REMOVED***);

    this.update(data);
  ***REMOVED***

  update(data) ***REMOVED***
    /**
     * The status of this presence:
     * @type ***REMOVED***PresenceStatus***REMOVED***
     */
    this.status = data.status || this.status || 'offline';

    /**
     * The game that the user is playing
     * @type ***REMOVED***?Game***REMOVED***
     * @deprecated
     */
    this.game = data.game ? new Game(data.game, this) : null;

    if (data.activities) ***REMOVED***
      /**
       * The activities of this presence
       * @type ***REMOVED***Game[]***REMOVED***
       */
      this.activities = data.activities.map(activity => new Game(activity, this));
    ***REMOVED*** else if (data.activity || data.game) ***REMOVED***
      this.activities = [new Game(data.activity || data.game, this)];
    ***REMOVED*** else ***REMOVED***
      this.activities = [];
    ***REMOVED***

    /**
     * The devices this presence is on
     * @type ***REMOVED***?Object***REMOVED***
     * @property ***REMOVED***?ClientPresenceStatus***REMOVED*** web The current presence in the web application
     * @property ***REMOVED***?ClientPresenceStatus***REMOVED*** mobile The current presence in the mobile application
     * @property ***REMOVED***?ClientPresenceStatus***REMOVED*** desktop The current presence in the desktop application
     */
    this.clientStatus = data.client_status || null;
  ***REMOVED***

  /**
   * Whether this presence is equal to another
   * @param ***REMOVED***Presence***REMOVED*** presence The presence to compare with
   * @returns ***REMOVED***boolean***REMOVED***
   */
  equals(presence) ***REMOVED***
    return this === presence || (
      presence &&
      this.status === presence.status &&
      this.activities.length === presence.activities.length &&
      this.activities.every((activity, index) => activity.equals(presence.activities[index])) &&
      this.clientStatus.web === presence.clientStatus.web &&
      this.clientStatus.mobile === presence.clientStatus.mobile &&
      this.clientStatus.desktop === presence.clientStatus.desktop
    );
  ***REMOVED***
***REMOVED***

/**
 * Represents a game that is part of a user's presence.
 */
class Game ***REMOVED***
  constructor(data, presence) ***REMOVED***
    Object.defineProperty(this, 'presence', ***REMOVED*** value: presence ***REMOVED***);

    /**
     * The name of the game being played
     * @type ***REMOVED***string***REMOVED***
     */
    this.name = data.name;

    /**
     * The type of the game status, its possible values:
     * - 0: Playing
     * - 1: Streaming
     * - 2: Listening
     * - 3: Watching
     * @type ***REMOVED***number***REMOVED***
     */
    this.type = data.type;

    /**
     * If the game is being streamed, a link to the stream
     * @type ***REMOVED***?string***REMOVED***
     */
    this.url = data.url || null;

    /**
     * Details about the activity
     * @type ***REMOVED***?string***REMOVED***
     */
    this.details = data.details || null;

    /**
     * State of the activity
     * @type ***REMOVED***?string***REMOVED***
     */
    this.state = data.state || null;

    /**
     * Application ID associated with this activity
     * @type ***REMOVED***?Snowflake***REMOVED***
     */
    this.applicationID = data.application_id || null;

    /**
     * Timestamps for the activity
     * @type ***REMOVED***?Object***REMOVED***
     * @prop ***REMOVED***?Date***REMOVED*** start When the activity started
     * @prop ***REMOVED***?Date***REMOVED*** end When the activity will end
     */
    this.timestamps = data.timestamps ? ***REMOVED***
      start: data.timestamps.start ? new Date(Number(data.timestamps.start)) : null,
      end: data.timestamps.end ? new Date(Number(data.timestamps.end)) : null,
    ***REMOVED*** : null;

    /**
     * Party of the activity
     * @type ***REMOVED***?Object***REMOVED***
     * @prop ***REMOVED***?string***REMOVED*** id ID of the party
     * @prop ***REMOVED***number[]***REMOVED*** size Size of the party as `[current, max]`
     */
    this.party = data.party || null;

    /**
     * Assets for rich presence
     * @type ***REMOVED***?RichPresenceAssets***REMOVED***
     */
    this.assets = data.assets ? new RichPresenceAssets(this, data.assets) : null;

    if (data.emoji) ***REMOVED***
      /**
       * Emoji for a custom activity
       * <warn>There is no `reaction` property for this emoji.</warn>
       * @type ***REMOVED***?ReactionEmoji***REMOVED***
       */
      this.emoji = new ReactionEmoji(***REMOVED*** message: ***REMOVED*** client: this.presence.client ***REMOVED*** ***REMOVED***, data.emoji);
      this.emoji.reaction = null;
    ***REMOVED*** else ***REMOVED***
      this.emoji = null;
    ***REMOVED***


    /**
     * Creation date of the activity
     * @type ***REMOVED***number***REMOVED***
     */
    this.createdTimestamp = new Date(data.created_at).getTime();

    this.syncID = data.sync_id;
    this._flags = data.flags;
  ***REMOVED***

  /**
   * The time the activity was created at
   * @type ***REMOVED***Date***REMOVED***
   * @readonly
   */
  get createdAt() ***REMOVED***
    return new Date(this.createdTimestamp);
  ***REMOVED***

  /**
   * Flags that describe the activity
   * @type ***REMOVED***ActivityFlags[]***REMOVED***
   */
  get flags() ***REMOVED***
    const flags = [];
    for (const [name, flag] of Object.entries(ActivityFlags)) ***REMOVED***
      if ((this._flags & flag) === flag) flags.push(name);
    ***REMOVED***
    return flags;
  ***REMOVED***

  /**
   * Whether or not the game is being streamed
   * @type ***REMOVED***boolean***REMOVED***
   * @readonly
   */
  get streaming() ***REMOVED***
    return this.type === 1;
  ***REMOVED***

  /**
   * When concatenated with a string, this automatically returns the game's name instead of the Game object.
   * @returns ***REMOVED***string***REMOVED***
   */
  toString() ***REMOVED***
    return this.name;
  ***REMOVED***

  /**
   * Whether this game is equal to another game
   * @param ***REMOVED***Game***REMOVED*** game The game to compare with
   * @returns ***REMOVED***boolean***REMOVED***
   */
  equals(game) ***REMOVED***
    return this === game || (
      game &&
      this.name === game.name &&
      this.type === game.type &&
      this.url === game.url
    );
  ***REMOVED***
***REMOVED***

/**
 * Assets for a rich presence
 */
class RichPresenceAssets ***REMOVED***
  constructor(game, assets) ***REMOVED***
    Object.defineProperty(this, 'game', ***REMOVED*** value: game ***REMOVED***);

    /**
     * Hover text for the large image
     * @type ***REMOVED***?string***REMOVED***
     */
    this.largeText = assets.large_text || null;

    /**
     * Hover text for the small image
     * @type ***REMOVED***?string***REMOVED***
     */
    this.smallText = assets.small_text || null;

    /**
     * ID of the large image asset
     * @type ***REMOVED***?Snowflake***REMOVED***
     */
    this.largeImage = assets.large_image || null;

    /**
     * ID of the small image asset
     * @type ***REMOVED***?Snowflake***REMOVED***
     */
    this.smallImage = assets.small_image || null;
  ***REMOVED***

  /**
   * The URL of the small image asset
   * @type ***REMOVED***?string***REMOVED***
   * @readonly
   */
  get smallImageURL() ***REMOVED***
    if (!this.smallImage) return null;
    return Endpoints.CDN(this.game.presence.client.options.http.cdn)
      .AppAsset(this.game.applicationID, this.smallImage);
  ***REMOVED***

  /**
   * The URL of the large image asset
   * @type ***REMOVED***?string***REMOVED***
   * @readonly
   */
  get largeImageURL() ***REMOVED***
    if (!this.largeImage) return null;
    if (/^spotify:/.test(this.largeImage)) ***REMOVED***
      return `https://i.scdn.co/image/$***REMOVED***this.largeImage.slice(8)***REMOVED***`;
    ***REMOVED*** else if (/^twitch:/.test(this.largeImage)) ***REMOVED***
      return `https://static-cdn.jtvnw.net/previews-ttv/live_user_$***REMOVED***this.largeImage.slice(7)***REMOVED***.png`;
    ***REMOVED***
    return Endpoints.CDN(this.game.presence.client.options.http.cdn)
      .AppAsset(this.game.applicationID, this.largeImage);
  ***REMOVED***
***REMOVED***

exports.Presence = Presence;
exports.Game = Game;
exports.RichPresenceAssets = RichPresenceAssets;
