// discord.js/src/structures/Team.js
const Snowflake = require('../util/Snowflake');
const Collection = require('../util/Collection');
const TeamMember = require('./TeamMember');
const Constants = require('../util/Constants');

/**
 * Represents a Client OAuth2 Application Team.
 */
class Team ***REMOVED***
  constructor(client, data) ***REMOVED***
    /**
     * The client that instantiated the team
     * @name Team#client
     * @type ***REMOVED***Client***REMOVED***
     * @readonly
     */
    Object.defineProperty(this, 'client', ***REMOVED*** value: client ***REMOVED***);

    this._patch(data);
  ***REMOVED***

  _patch(data) ***REMOVED***
    /**
     * The ID of the Team
     * @type ***REMOVED***Snowflake***REMOVED***
     */
    this.id = data.id;

    /**
     * The name of the Team
     * @type ***REMOVED***string***REMOVED***
     */
    this.name = data.name;

    /**
     * The Team's icon hash
     * @type ***REMOVED***?string***REMOVED***
     */
    this.icon = data.icon || null;

    /**
     * The Team's owner id
     * @type ***REMOVED***?string***REMOVED***
     */
    this.ownerID = data.owner_user_id || null;

    /**
     * The Team's members
     * @type ***REMOVED***Collection<Snowflake, TeamMember>***REMOVED***
     */
    this.members = new Collection();

    for (const memberData of data.members) ***REMOVED***
      const member = new TeamMember(this.client, this, memberData);
      this.members.set(member.id, member);
    ***REMOVED***
  ***REMOVED***

  /**
   * The owner of the team
   * @type ***REMOVED***?TeamMember***REMOVED***
   * @readonly
   */
  get owner() ***REMOVED***
    return this.members.get(this.ownerID) || null;
  ***REMOVED***

  /**
   * The timestamp the team was created at
   * @type ***REMOVED***number***REMOVED***
   * @readonly
   */
  get createdTimestamp() ***REMOVED***
    return Snowflake.deconstruct(this.id).timestamp;
  ***REMOVED***

  /**
   * The time the team was created at
   * @type ***REMOVED***Date***REMOVED***
   * @readonly
   */
  get createdAt() ***REMOVED***
    return new Date(this.createdTimestamp);
  ***REMOVED***

  /**
   * A link to the teams's icon.
   * @type ***REMOVED***?string***REMOVED***
   * @readonly
   */
  get iconURL() ***REMOVED***
    if (!this.icon) return null;
    return Constants.Endpoints.CDN(this.client.options.http.cdn).TeamIcon(this.id, this.icon);
  ***REMOVED***

  /**
   * When concatenated with a string, this automatically returns the Team's name instead of the
   * Team object.
   * @returns ***REMOVED***string***REMOVED***
   * @example
   * // Logs: Team name: My Team
   * console.log(`Team name: $***REMOVED***team***REMOVED***`);
   */
  toString() ***REMOVED***
    return this.name;
  ***REMOVED***
***REMOVED***

module.exports = Team;
