// discord.js/src/structures/User.js
const TextBasedChannel = require('./interfaces/TextBasedChannel');
const Constants = require('../util/Constants');
const Presence = require('./Presence').Presence;
const Snowflake = require('../util/Snowflake');
const util = require('util');

/**
 * Represents a user on Discord.
 * @implements ***REMOVED***TextBasedChannel***REMOVED***
 */
class User ***REMOVED***
  constructor(client, data) ***REMOVED***
    /**
     * The client that created the instance of the user
     * @name User#client
     * @type ***REMOVED***Client***REMOVED***
     * @readonly
     */
    Object.defineProperty(this, 'client', ***REMOVED*** value: client ***REMOVED***);

    if (data) this.setup(data);
  ***REMOVED***

  setup(data) ***REMOVED***
    /**
     * The ID of the user
     * @type ***REMOVED***Snowflake***REMOVED***
     */
    this.id = data.id;

    /**
     * The username of the user
     * @type ***REMOVED***string***REMOVED***
     */
    this.username = data.username;

    /**
     * A discriminator based on username for the user
     * @type ***REMOVED***string***REMOVED***
     */
    this.discriminator = data.discriminator;

    /**
     * The ID of the user's avatar
     * @type ***REMOVED***string***REMOVED***
     */
    this.avatar = data.avatar;

    /**
     * Whether or not the user is a bot
     * @type ***REMOVED***boolean***REMOVED***
     */
    this.bot = Boolean(data.bot);

    /**
     * Whether this is an Official Discord System user (part of the urgent message system)
     * @type ***REMOVED***?boolean***REMOVED***
     * @name User#system
     */
    if (typeof data.system !== 'undefined') this.system = Boolean(data.system);

    /**
     * The ID of the last message sent by the user, if one was sent
     * @type ***REMOVED***?Snowflake***REMOVED***
     */
    this.lastMessageID = null;

    /**
     * The Message object of the last message sent by the user, if one was sent
     * @type ***REMOVED***?Message***REMOVED***
     */
    this.lastMessage = null;
  ***REMOVED***

  patch(data) ***REMOVED***
    for (const prop of ['id', 'username', 'discriminator', 'avatar', 'bot']) ***REMOVED***
      if (typeof data[prop] !== 'undefined') this[prop] = data[prop];
    ***REMOVED***
    if (data.token) this.client.token = data.token;
  ***REMOVED***

  /**
   * The timestamp the user was created at
   * @type ***REMOVED***number***REMOVED***
   * @readonly
   */
  get createdTimestamp() ***REMOVED***
    return Snowflake.deconstruct(this.id).timestamp;
  ***REMOVED***

  /**
   * The time the user was created
   * @type ***REMOVED***Date***REMOVED***
   * @readonly
   */
  get createdAt() ***REMOVED***
    return new Date(this.createdTimestamp);
  ***REMOVED***

  /**
   * The presence of this user
   * @type ***REMOVED***Presence***REMOVED***
   * @readonly
   */
  get presence() ***REMOVED***
    if (this.client.presences.has(this.id)) return this.client.presences.get(this.id);
    for (const guild of this.client.guilds.values()) ***REMOVED***
      if (guild.presences.has(this.id)) return guild.presences.get(this.id);
    ***REMOVED***
    return new Presence(undefined, this.client);
  ***REMOVED***

  /**
   * A link to the user's avatar
   * @type ***REMOVED***?string***REMOVED***
   * @readonly
   */
  get avatarURL() ***REMOVED***
    if (!this.avatar) return null;
    return Constants.Endpoints.User(this).Avatar(this.client.options.http.cdn, this.avatar);
  ***REMOVED***

  /**
   * A link to the user's default avatar
   * @type ***REMOVED***string***REMOVED***
   * @readonly
   */
  get defaultAvatarURL() ***REMOVED***
    const avatars = Object.keys(Constants.DefaultAvatars);
    const avatar = avatars[this.discriminator % avatars.length];
    return Constants.Endpoints.CDN(this.client.options.http.host).Asset(`$***REMOVED***Constants.DefaultAvatars[avatar]***REMOVED***.png`);
  ***REMOVED***

  /**
   * A link to the user's avatar if they have one. Otherwise a link to their default avatar will be returned
   * @type ***REMOVED***string***REMOVED***
   * @readonly
   */
  get displayAvatarURL() ***REMOVED***
    return this.avatarURL || this.defaultAvatarURL;
  ***REMOVED***

  /**
   * The Discord "tag" (e.g. `hydrabolt#0001`) for this user
   * @type ***REMOVED***string***REMOVED***
   * @readonly
   */
  get tag() ***REMOVED***
    return `$***REMOVED***this.username***REMOVED***#$***REMOVED***this.discriminator***REMOVED***`;
  ***REMOVED***

  /**
   * The note that is set for the user
   * <warn>This is only available when using a user account.</warn>
   * @type ***REMOVED***?string***REMOVED***
   * @readonly
   * @deprecated
   */
  get note() ***REMOVED***
    return this.client.user.notes.get(this.id) || null;
  ***REMOVED***

  /**
   * Check whether the user is typing in a channel.
   * @param ***REMOVED***ChannelResolvable***REMOVED*** channel The channel to check in
   * @returns ***REMOVED***boolean***REMOVED***
   */
  typingIn(channel) ***REMOVED***
    channel = this.client.resolver.resolveChannel(channel);
    return channel._typing.has(this.id);
  ***REMOVED***

  /**
   * Get the time that the user started typing.
   * @param ***REMOVED***ChannelResolvable***REMOVED*** channel The channel to get the time in
   * @returns ***REMOVED***?Date***REMOVED***
   */
  typingSinceIn(channel) ***REMOVED***
    channel = this.client.resolver.resolveChannel(channel);
    return channel._typing.has(this.id) ? new Date(channel._typing.get(this.id).since) : null;
  ***REMOVED***

  /**
   * Get the amount of time the user has been typing in a channel for (in milliseconds), or -1 if they're not typing.
   * @param ***REMOVED***ChannelResolvable***REMOVED*** channel The channel to get the time in
   * @returns ***REMOVED***number***REMOVED***
   */
  typingDurationIn(channel) ***REMOVED***
    channel = this.client.resolver.resolveChannel(channel);
    return channel._typing.has(this.id) ? channel._typing.get(this.id).elapsedTime : -1;
  ***REMOVED***

  /**
   * The DM between the client's user and this user
   * @type ***REMOVED***?DMChannel***REMOVED***
   * @readonly
   */
  get dmChannel() ***REMOVED***
    return this.client.channels.find(c => c.type === 'dm' && c.recipient.id === this.id);
  ***REMOVED***

  /**
   * Creates a DM channel between the client and the user.
   * @returns ***REMOVED***Promise<DMChannel>***REMOVED***
   */
  createDM() ***REMOVED***
    return this.client.rest.methods.createDM(this);
  ***REMOVED***

  /**
   * Deletes a DM channel (if one exists) between the client and the user. Resolves with the channel if successful.
   * @returns ***REMOVED***Promise<DMChannel>***REMOVED***
   */
  deleteDM() ***REMOVED***
    return this.client.rest.methods.deleteChannel(this);
  ***REMOVED***

  /**
   * Sends a friend request to the user.
   * <warn>This is only available when using a user account.</warn>
   * @returns ***REMOVED***Promise<User>***REMOVED***
   * @deprecated
   */
  addFriend() ***REMOVED***
    return this.client.rest.methods.addFriend(this);
  ***REMOVED***

  /**
   * Removes the user from your friends.
   * <warn>This is only available when using a user account.</warn>
   * @returns ***REMOVED***Promise<User>***REMOVED***
   * @deprecated
   */
  removeFriend() ***REMOVED***
    return this.client.rest.methods.removeFriend(this);
  ***REMOVED***

  /**
   * Blocks the user.
   * <warn>This is only available when using a user account.</warn>
   * @returns ***REMOVED***Promise<User>***REMOVED***
   * @deprecated
   */
  block() ***REMOVED***
    return this.client.rest.methods.blockUser(this);
  ***REMOVED***

  /**
   * Unblocks the user.
   * <warn>This is only available when using a user account.</warn>
   * @returns ***REMOVED***Promise<User>***REMOVED***
   * @deprecated
   */
  unblock() ***REMOVED***
    return this.client.rest.methods.unblockUser(this);
  ***REMOVED***

  /**
   * Get the profile of the user.
   * <warn>This is only available when using a user account.</warn>
   * @returns ***REMOVED***Promise<UserProfile>***REMOVED***
   * @deprecated
   */
  fetchProfile() ***REMOVED***
    return this.client.rest.methods.fetchUserProfile(this);
  ***REMOVED***

  /**
   * Sets a note for the user.
   * <warn>This is only available when using a user account.</warn>
   * @param ***REMOVED***string***REMOVED*** note The note to set for the user
   * @returns ***REMOVED***Promise<User>***REMOVED***
   * @deprecated
   */
  setNote(note) ***REMOVED***
    return this.client.rest.methods.setNote(this, note);
  ***REMOVED***

  /**
   * Checks if the user is equal to another. It compares ID, username, discriminator, avatar, and bot flags.
   * It is recommended to compare equality by using `user.id === user2.id` unless you want to compare all properties.
   * @param ***REMOVED***User***REMOVED*** user User to compare with
   * @returns ***REMOVED***boolean***REMOVED***
   */
  equals(user) ***REMOVED***
    let equal = user &&
      this.id === user.id &&
      this.username === user.username &&
      this.discriminator === user.discriminator &&
      this.avatar === user.avatar &&
      this.bot === Boolean(user.bot);

    return equal;
  ***REMOVED***

  /**
   * When concatenated with a string, this automatically concatenates the user's mention instead of the User object.
   * @returns ***REMOVED***string***REMOVED***
   * @example
   * // logs: Hello from <@123456789>!
   * console.log(`Hello from $***REMOVED***user***REMOVED***!`);
   */
  toString() ***REMOVED***
    return `<@$***REMOVED***this.id***REMOVED***>`;
  ***REMOVED***

  // These are here only for documentation purposes - they are implemented by TextBasedChannel
  /* eslint-disable no-empty-function */
  send() ***REMOVED******REMOVED***
  sendMessage() ***REMOVED******REMOVED***
  sendEmbed() ***REMOVED******REMOVED***
  sendFile() ***REMOVED******REMOVED***
  sendCode() ***REMOVED******REMOVED***
***REMOVED***

TextBasedChannel.applyToClass(User);

User.prototype.block =
  util.deprecate(User.prototype.block, 'User#block: userbot methods will be removed');

User.prototype.unblock =
  util.deprecate(User.prototype.unblock, 'User#unblock: userbot methods will be removed');

User.prototype.addFriend =
  util.deprecate(User.prototype.addFriend, 'User#addFriend: userbot methods will be removed');

User.prototype.removeFriend =
  util.deprecate(User.prototype.removeFriend, 'User#removeFriend: userbot methods will be removed');

User.prototype.setNote =
  util.deprecate(User.prototype.setNote, 'User#setNote, userbot methods will be removed');

User.prototype.fetchProfile =
  util.deprecate(User.prototype.fetchProfile, 'User#fetchProfile: userbot methods will be removed');

module.exports = User;
