// discord.js/src/structures/UserProfile.js
const Collection = require('../util/Collection');
const UserConnection = require('./UserConnection');

/**
 * Represents a user's profile on Discord.
 */
class UserProfile ***REMOVED***
  constructor(user, data) ***REMOVED***
    /**
     * The owner of the profile
     * @type ***REMOVED***User***REMOVED***
     */
    this.user = user;

    /**
     * The client that created the instance of the UserProfile
     * @name UserProfile#client
     * @type ***REMOVED***Client***REMOVED***
     * @readonly
     */
    Object.defineProperty(this, 'client', ***REMOVED*** value: user.client ***REMOVED***);

    /**
     * The guilds that the client user and the user share
     * @type ***REMOVED***Collection<Snowflake, Guild>***REMOVED***
     */
    this.mutualGuilds = new Collection();

    /**
     * The user's connections
     * @type ***REMOVED***Collection<Snowflake, UserConnection>***REMOVED***
     */
    this.connections = new Collection();

    this.setup(data);
  ***REMOVED***

  setup(data) ***REMOVED***
    /**
     * If the user has Discord Premium
     * @type ***REMOVED***boolean***REMOVED***
     */
    this.premium = data.premium;

    /**
     * The date since which the user has had Discord Premium
     * @type ***REMOVED***?Date***REMOVED***
     */
    this.premiumSince = data.premium_since ? new Date(data.premium_since) : null;

    for (const guild of data.mutual_guilds) ***REMOVED***
      if (this.client.guilds.has(guild.id)) ***REMOVED***
        this.mutualGuilds.set(guild.id, this.client.guilds.get(guild.id));
      ***REMOVED***
    ***REMOVED***
    for (const connection of data.connected_accounts) ***REMOVED***
      this.connections.set(connection.id, new UserConnection(this.user, connection));
    ***REMOVED***
  ***REMOVED***
***REMOVED***

module.exports = UserProfile;
