// discord.js/src/structures/UserConnection.js
/**
 * Represents a user connection (or "platform identity").
 */
class UserConnection ***REMOVED***
  constructor(user, data) ***REMOVED***
    /**
     * The user that owns the connection
     * @type ***REMOVED***User***REMOVED***
     */
    this.user = user;

    this.setup(data);
  ***REMOVED***

  setup(data) ***REMOVED***
    /**
     * The type of the connection
     * @type ***REMOVED***string***REMOVED***
     */
    this.type = data.type;

    /**
     * The username of the connection account
     * @type ***REMOVED***string***REMOVED***
     */
    this.name = data.name;

    /**
     * The id of the connection account
     * @type ***REMOVED***string***REMOVED***
     */
    this.id = data.id;

    /**
     * Whether the connection is revoked
     * @type ***REMOVED***boolean***REMOVED***
     */
    this.revoked = data.revoked;

    /**
     * Partial server integrations (not yet implemented)
     * @type ***REMOVED***Object[]***REMOVED***
     */
    this.integrations = data.integrations;
  ***REMOVED***
***REMOVED***

module.exports = UserConnection;
