// discord.js/src/structures/VoiceRegion.js
/**
 * Represents a Discord voice region for guilds.
 */
class VoiceRegion ***REMOVED***
  constructor(data) ***REMOVED***
    /**
     * The ID of the region
     * @type ***REMOVED***string***REMOVED***
     */
    this.id = data.id;

    /**
     * Name of the region
     * @type ***REMOVED***string***REMOVED***
     */
    this.name = data.name;

    /**
     * Whether the region is VIP-only
     * @type ***REMOVED***boolean***REMOVED***
     */
    this.vip = data.vip;

    /**
     * Whether the region is deprecated
     * @type ***REMOVED***boolean***REMOVED***
     */
    this.deprecated = data.deprecated;

    /**
     * Whether the region is optimal
     * @type ***REMOVED***boolean***REMOVED***
     */
    this.optimal = data.optimal;

    /**
     * Whether the region is custom
     * @type ***REMOVED***boolean***REMOVED***
     */
    this.custom = data.custom;

    /**
     * A sample hostname for what a connection might look like
     * @type ***REMOVED***string***REMOVED***
     */
    this.sampleHostname = data.sample_hostname;
  ***REMOVED***
***REMOVED***

module.exports = VoiceRegion;
