// inquirer/lib/utils/screen-manager.js
'use strict';
var _ = require('lodash');
var util = require('./readline');
var cliWidth = require('cli-width');
var stripAnsi = require('strip-ansi');
var stringWidth = require('string-width');

function height(content) ***REMOVED***
  return content.split('\n').length;
***REMOVED***

function lastLine(content) ***REMOVED***
  return _.last(content.split('\n'));
***REMOVED***

class ScreenManager ***REMOVED***
  constructor(rl) ***REMOVED***
    // These variables are keeping information to allow correct prompt re-rendering
    this.height = 0;
    this.extraLinesUnderPrompt = 0;

    this.rl = rl;
  ***REMOVED***

  render(content, bottomContent) ***REMOVED***
    this.rl.output.unmute();
    this.clean(this.extraLinesUnderPrompt);

    /**
     * Write message to screen and setPrompt to control backspace
     */

    var promptLine = lastLine(content);
    var rawPromptLine = stripAnsi(promptLine);

    // Remove the rl.line from our prompt. We can't rely on the content of
    // rl.line (mainly because of the password prompt), so just rely on it's
    // length.
    var prompt = rawPromptLine;
    if (this.rl.line.length) ***REMOVED***
      prompt = prompt.slice(0, -this.rl.line.length);
    ***REMOVED***

    this.rl.setPrompt(prompt);

    // SetPrompt will change cursor position, now we can get correct value
    var cursorPos = this.rl._getCursorPos();
    var width = this.normalizedCliWidth();

    content = this.forceLineReturn(content, width);
    if (bottomContent) ***REMOVED***
      bottomContent = this.forceLineReturn(bottomContent, width);
    ***REMOVED***

    // Manually insert an extra line if we're at the end of the line.
    // This prevent the cursor from appearing at the beginning of the
    // current line.
    if (rawPromptLine.length % width === 0) ***REMOVED***
      content += '\n';
    ***REMOVED***

    var fullContent = content + (bottomContent ? '\n' + bottomContent : '');
    this.rl.output.write(fullContent);

    /**
     * Re-adjust the cursor at the correct position.
     */

    // We need to consider parts of the prompt under the cursor as part of the bottom
    // content in order to correctly cleanup and re-render.
    var promptLineUpDiff = Math.floor(rawPromptLine.length / width) - cursorPos.rows;
    var bottomContentHeight =
      promptLineUpDiff + (bottomContent ? height(bottomContent) : 0);
    if (bottomContentHeight > 0) ***REMOVED***
      util.up(this.rl, bottomContentHeight);
    ***REMOVED***

    // Reset cursor at the beginning of the line
    util.left(this.rl, stringWidth(lastLine(fullContent)));

    // Adjust cursor on the right
    if (cursorPos.cols > 0) ***REMOVED***
      util.right(this.rl, cursorPos.cols);
    ***REMOVED***

    /**
     * Set up state for next re-rendering
     */
    this.extraLinesUnderPrompt = bottomContentHeight;
    this.height = height(fullContent);

    this.rl.output.mute();
  ***REMOVED***

  clean(extraLines) ***REMOVED***
    if (extraLines > 0) ***REMOVED***
      util.down(this.rl, extraLines);
    ***REMOVED***

    util.clearLine(this.rl, this.height);
  ***REMOVED***

  done() ***REMOVED***
    this.rl.setPrompt('');
    this.rl.output.unmute();
    this.rl.output.write('\n');
  ***REMOVED***

  releaseCursor() ***REMOVED***
    if (this.extraLinesUnderPrompt > 0) ***REMOVED***
      util.down(this.rl, this.extraLinesUnderPrompt);
    ***REMOVED***
  ***REMOVED***

  normalizedCliWidth() ***REMOVED***
    var width = cliWidth(***REMOVED***
      defaultWidth: 80,
      output: this.rl.output
    ***REMOVED***);
    return width;
  ***REMOVED***

  breakLines(lines, width) ***REMOVED***
    // Break lines who're longer than the cli width so we can normalize the natural line
    // returns behavior across terminals.
    width = width || this.normalizedCliWidth();
    var regex = new RegExp('(?:(?:\\033[[0-9;]*m)*.?)***REMOVED***1,' + width + '***REMOVED***', 'g');
    return lines.map(line => ***REMOVED***
      var chunk = line.match(regex);
      // Last match is always empty
      chunk.pop();
      return chunk || '';
    ***REMOVED***);
  ***REMOVED***

  forceLineReturn(content, width) ***REMOVED***
    width = width || this.normalizedCliWidth();
    return _.flatten(this.breakLines(content.split('\n'), width)).join('\n');
  ***REMOVED***
***REMOVED***

module.exports = ScreenManager;
