// rxjs/src/internal/observable/dom/MiscJSDoc.ts
import ***REMOVED*** Subscriber ***REMOVED*** from '../../Subscriber';
import ***REMOVED*** AjaxResponse ***REMOVED*** from './AjaxObservable';

/**
 * @see ***REMOVED***@link ajax***REMOVED***
 *
 * @interface
 * @name AjaxRequest
 * @noimport true
 */
export class AjaxRequestDoc ***REMOVED***
  /**
   * @type ***REMOVED***string***REMOVED***
   */
  url: string = '';
  /**
   * @type ***REMOVED***number***REMOVED***
   */
  body: any = 0;
  /**
   * @type ***REMOVED***string***REMOVED***
   */
  user: string = '';
  /**
   * @type ***REMOVED***boolean***REMOVED***
   */
  async: boolean = false;
  /**
   * @type ***REMOVED***string***REMOVED***
   */
  method: string = '';
  /**
   * @type ***REMOVED***Object***REMOVED***
   */
  headers: Object = null;
  /**
   * @type ***REMOVED***number***REMOVED***
   */
  timeout: number = 0;
  /**
   * @type ***REMOVED***string***REMOVED***
   */
  password: string = '';
  /**
   * @type ***REMOVED***boolean***REMOVED***
   */
  hasContent: boolean = false;
  /**
   * @type ***REMOVED***boolean***REMOVED***
   */
  crossDomain: boolean = false;
  /**
   * @type ***REMOVED***boolean***REMOVED***
   */
  withCredentials: boolean = false;
  /**
   * @return ***REMOVED***XMLHttpRequest***REMOVED***
   */
  createXHR(): XMLHttpRequest ***REMOVED***
    return null;
  ***REMOVED***
  /**
   * @type ***REMOVED***Subscriber***REMOVED***
   */
  progressSubscriber: Subscriber<any> = null;
  /**
   * @param ***REMOVED***AjaxResponse***REMOVED*** response
   * @return ***REMOVED***T***REMOVED***
   */
  resultSelector<T>(response: AjaxResponse): T ***REMOVED***
    return null;
  ***REMOVED***
  /**
   * @type ***REMOVED***string***REMOVED***
   */
  responseType: string = '';
***REMOVED***
