// tmp/lib/tmp.js
/*!
 * Tmp
 *
 * Copyright (c) 2011-2017 KARASZI Istvan <github@spam.raszi.hu>
 *
 * MIT Licensed
 */

/*
 * Module dependencies.
 */
const fs = require('fs');
const path = require('path');
const crypto = require('crypto');
const osTmpDir = require('os-tmpdir');
const _c = process.binding('constants');

/*
 * The working inner variables.
 */
const
  /**
   * The temporary directory.
   * @type ***REMOVED***string***REMOVED***
   */
  tmpDir = osTmpDir(),

  // the random characters to choose from
  RANDOM_CHARS = '0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz',

  TEMPLATE_PATTERN = /XXXXXX/,

  DEFAULT_TRIES = 3,

  CREATE_FLAGS = (_c.O_CREAT || _c.fs.O_CREAT) | (_c.O_EXCL || _c.fs.O_EXCL) | (_c.O_RDWR || _c.fs.O_RDWR),

  EBADF = _c.EBADF || _c.os.errno.EBADF,
  ENOENT = _c.ENOENT || _c.os.errno.ENOENT,

  DIR_MODE = 448 /* 0o700 */,
  FILE_MODE = 384 /* 0o600 */,

  // this will hold the objects need to be removed on exit
  _removeObjects = [];

var
  _gracefulCleanup = false,
  _uncaughtException = false;

/**
 * Random name generator based on crypto.
 * Adapted from http://blog.tompawlak.org/how-to-generate-random-values-nodejs-javascript
 *
 * @param ***REMOVED***number***REMOVED*** howMany
 * @returns ***REMOVED***string***REMOVED*** the generated random name
 * @private
 */
function _randomChars(howMany) ***REMOVED***
  var
    value = [],
    rnd = null;

  // make sure that we do not fail because we ran out of entropy
  try ***REMOVED***
    rnd = crypto.randomBytes(howMany);
  ***REMOVED*** catch (e) ***REMOVED***
    rnd = crypto.pseudoRandomBytes(howMany);
  ***REMOVED***

  for (var i = 0; i < howMany; i++) ***REMOVED***
    value.push(RANDOM_CHARS[rnd[i] % RANDOM_CHARS.length]);
  ***REMOVED***

  return value.join('');
***REMOVED***

/**
 * Checks whether the `obj` parameter is defined or not.
 *
 * @param ***REMOVED***Object***REMOVED*** obj
 * @returns ***REMOVED***boolean***REMOVED*** true if the object is undefined
 * @private
 */
function _isUndefined(obj) ***REMOVED***
  return typeof obj === 'undefined';
***REMOVED***

/**
 * Parses the function arguments.
 *
 * This function helps to have optional arguments.
 *
 * @param ***REMOVED***(Options|Function)***REMOVED*** options
 * @param ***REMOVED***Function***REMOVED*** callback
 * @returns ***REMOVED***Array***REMOVED*** parsed arguments
 * @private
 */
function _parseArguments(options, callback) ***REMOVED***
  if (typeof options == 'function') ***REMOVED***
    return [callback || ***REMOVED******REMOVED***, options];
  ***REMOVED***

  if (_isUndefined(options)) ***REMOVED***
    return [***REMOVED******REMOVED***, callback];
  ***REMOVED***

  return [options, callback];
***REMOVED***

/**
 * Generates a new temporary name.
 *
 * @param ***REMOVED***Object***REMOVED*** opts
 * @returns ***REMOVED***string***REMOVED*** the new random name according to opts
 * @private
 */
function _generateTmpName(opts) ***REMOVED***
  if (opts.name) ***REMOVED***
    return path.join(opts.dir || tmpDir, opts.name);
  ***REMOVED***

  // mkstemps like template
  if (opts.template) ***REMOVED***
    return opts.template.replace(TEMPLATE_PATTERN, _randomChars(6));
  ***REMOVED***

  // prefix and postfix
  const name = [
    opts.prefix || 'tmp-',
    process.pid,
    _randomChars(12),
    opts.postfix || ''
  ].join('');

  return path.join(opts.dir || tmpDir, name);
***REMOVED***

/**
 * Gets a temporary file name.
 *
 * @param ***REMOVED***(Options|tmpNameCallback)***REMOVED*** options options or callback
 * @param ***REMOVED***?tmpNameCallback***REMOVED*** callback the callback function
 */
function tmpName(options, callback) ***REMOVED***
  var
    args = _parseArguments(options, callback),
    opts = args[0],
    cb = args[1],
    tries = opts.name ? 1 : opts.tries || DEFAULT_TRIES;

  if (isNaN(tries) || tries < 0)
    return cb(new Error('Invalid tries'));

  if (opts.template && !opts.template.match(TEMPLATE_PATTERN))
    return cb(new Error('Invalid template provided'));

  (function _getUniqueName() ***REMOVED***
    const name = _generateTmpName(opts);

    // check whether the path exists then retry if needed
    fs.stat(name, function (err) ***REMOVED***
      if (!err) ***REMOVED***
        if (tries-- > 0) return _getUniqueName();

        return cb(new Error('Could not get a unique tmp filename, max tries reached ' + name));
      ***REMOVED***

      cb(null, name);
    ***REMOVED***);
  ***REMOVED***());
***REMOVED***

/**
 * Synchronous version of tmpName.
 *
 * @param ***REMOVED***Object***REMOVED*** options
 * @returns ***REMOVED***string***REMOVED*** the generated random name
 * @throws ***REMOVED***Error***REMOVED*** if the options are invalid or could not generate a filename
 */
function tmpNameSync(options) ***REMOVED***
  var
    args = _parseArguments(options),
    opts = args[0],
    tries = opts.name ? 1 : opts.tries || DEFAULT_TRIES;

  if (isNaN(tries) || tries < 0)
    throw new Error('Invalid tries');

  if (opts.template && !opts.template.match(TEMPLATE_PATTERN))
    throw new Error('Invalid template provided');

  do ***REMOVED***
    const name = _generateTmpName(opts);
    try ***REMOVED***
      fs.statSync(name);
    ***REMOVED*** catch (e) ***REMOVED***
      return name;
    ***REMOVED***
  ***REMOVED*** while (tries-- > 0);

  throw new Error('Could not get a unique tmp filename, max tries reached');
***REMOVED***

/**
 * Creates and opens a temporary file.
 *
 * @param ***REMOVED***(Options|fileCallback)***REMOVED*** options the config options or the callback function
 * @param ***REMOVED***?fileCallback***REMOVED*** callback
 */
function file(options, callback) ***REMOVED***
  var
    args = _parseArguments(options, callback),
    opts = args[0],
    cb = args[1];

  opts.postfix = (_isUndefined(opts.postfix)) ? '.tmp' : opts.postfix;

  // gets a temporary filename
  tmpName(opts, function _tmpNameCreated(err, name) ***REMOVED***
    if (err) return cb(err);

    // create and open the file
    fs.open(name, CREATE_FLAGS, opts.mode || FILE_MODE, function _fileCreated(err, fd) ***REMOVED***
      if (err) return cb(err);

      if (opts.discardDescriptor) ***REMOVED***
        return fs.close(fd, function _discardCallback(err) ***REMOVED***
          if (err) ***REMOVED***
            // Low probability, and the file exists, so this could be
            // ignored.  If it isn't we certainly need to unlink the
            // file, and if that fails too its error is more
            // important.
            try ***REMOVED***
              fs.unlinkSync(name);
            ***REMOVED*** catch (e) ***REMOVED***
              if (!isENOENT(e)) ***REMOVED***
                err = e;
              ***REMOVED***
            ***REMOVED***
            return cb(err);
          ***REMOVED***
          cb(null, name, undefined, _prepareTmpFileRemoveCallback(name, -1, opts));
        ***REMOVED***);
      ***REMOVED***
      if (opts.detachDescriptor) ***REMOVED***
        return cb(null, name, fd, _prepareTmpFileRemoveCallback(name, -1, opts));
      ***REMOVED***
      cb(null, name, fd, _prepareTmpFileRemoveCallback(name, fd, opts));
    ***REMOVED***);
  ***REMOVED***);
***REMOVED***

/**
 * Synchronous version of file.
 *
 * @param ***REMOVED***Options***REMOVED*** options
 * @returns ***REMOVED***FileSyncObject***REMOVED*** object consists of name, fd and removeCallback
 * @throws ***REMOVED***Error***REMOVED*** if cannot create a file
 */
function fileSync(options) ***REMOVED***
  var
    args = _parseArguments(options),
    opts = args[0];

  opts.postfix = opts.postfix || '.tmp';

  const discardOrDetachDescriptor = opts.discardDescriptor || opts.detachDescriptor;
  const name = tmpNameSync(opts);
  var fd = fs.openSync(name, CREATE_FLAGS, opts.mode || FILE_MODE);
  if (opts.discardDescriptor) ***REMOVED***
    fs.closeSync(fd); 
    fd = undefined;
  ***REMOVED***

  return ***REMOVED***
    name: name,
    fd: fd,
    removeCallback: _prepareTmpFileRemoveCallback(name, discardOrDetachDescriptor ? -1 : fd, opts)
  ***REMOVED***;
***REMOVED***

/**
 * Removes files and folders in a directory recursively.
 *
 * @param ***REMOVED***string***REMOVED*** root
 * @private
 */
function _rmdirRecursiveSync(root) ***REMOVED***
  const dirs = [root];

  do ***REMOVED***
    var
      dir = dirs.pop(),
      deferred = false,
      files = fs.readdirSync(dir);

    for (var i = 0, length = files.length; i < length; i++) ***REMOVED***
      var
        file = path.join(dir, files[i]),
        stat = fs.lstatSync(file); // lstat so we don't recurse into symlinked directories

      if (stat.isDirectory()) ***REMOVED***
        if (!deferred) ***REMOVED***
          deferred = true;
          dirs.push(dir);
        ***REMOVED***
        dirs.push(file);
      ***REMOVED*** else ***REMOVED***
        fs.unlinkSync(file);
      ***REMOVED***
    ***REMOVED***

    if (!deferred) ***REMOVED***
      fs.rmdirSync(dir);
    ***REMOVED***
  ***REMOVED*** while (dirs.length !== 0);
***REMOVED***

/**
 * Creates a temporary directory.
 *
 * @param ***REMOVED***(Options|dirCallback)***REMOVED*** options the options or the callback function
 * @param ***REMOVED***?dirCallback***REMOVED*** callback
 */
function dir(options, callback) ***REMOVED***
  var
    args = _parseArguments(options, callback),
    opts = args[0],
    cb = args[1];

  // gets a temporary filename
  tmpName(opts, function _tmpNameCreated(err, name) ***REMOVED***
    if (err) return cb(err);

    // create the directory
    fs.mkdir(name, opts.mode || DIR_MODE, function _dirCreated(err) ***REMOVED***
      if (err) return cb(err);

      cb(null, name, _prepareTmpDirRemoveCallback(name, opts));
    ***REMOVED***);
  ***REMOVED***);
***REMOVED***

/**
 * Synchronous version of dir.
 *
 * @param ***REMOVED***Options***REMOVED*** options
 * @returns ***REMOVED***DirSyncObject***REMOVED*** object consists of name and removeCallback
 * @throws ***REMOVED***Error***REMOVED*** if it cannot create a directory
 */
function dirSync(options) ***REMOVED***
  var
    args = _parseArguments(options),
    opts = args[0];

  const name = tmpNameSync(opts);
  fs.mkdirSync(name, opts.mode || DIR_MODE);

  return ***REMOVED***
    name: name,
    removeCallback: _prepareTmpDirRemoveCallback(name, opts)
  ***REMOVED***;
***REMOVED***

/**
 * Prepares the callback for removal of the temporary file.
 *
 * @param ***REMOVED***string***REMOVED*** name the path of the file
 * @param ***REMOVED***number***REMOVED*** fd file descriptor
 * @param ***REMOVED***Object***REMOVED*** opts
 * @returns ***REMOVED***fileCallback***REMOVED***
 * @private
 */
function _prepareTmpFileRemoveCallback(name, fd, opts) ***REMOVED***
  const removeCallback = _prepareRemoveCallback(function _removeCallback(fdPath) ***REMOVED***
    try ***REMOVED***
      if (0 <= fdPath[0]) ***REMOVED***
        fs.closeSync(fdPath[0]);
      ***REMOVED***
    ***REMOVED***
    catch (e) ***REMOVED***
      // under some node/windows related circumstances, a temporary file
      // may have not be created as expected or the file was already closed
      // by the user, in which case we will simply ignore the error
      if (!isEBADF(e) && !isENOENT(e)) ***REMOVED***
        // reraise any unanticipated error
        throw e;
      ***REMOVED***
    ***REMOVED***
    try ***REMOVED***
      fs.unlinkSync(fdPath[1]);
    ***REMOVED***
    catch (e) ***REMOVED***
      if (!isENOENT(e)) ***REMOVED***
        // reraise any unanticipated error
        throw e;
      ***REMOVED***
    ***REMOVED***
  ***REMOVED***, [fd, name]);

  if (!opts.keep) ***REMOVED***
    _removeObjects.unshift(removeCallback);
  ***REMOVED***

  return removeCallback;
***REMOVED***

/**
 * Prepares the callback for removal of the temporary directory.
 *
 * @param ***REMOVED***string***REMOVED*** name
 * @param ***REMOVED***Object***REMOVED*** opts
 * @returns ***REMOVED***Function***REMOVED*** the callback
 * @private
 */
function _prepareTmpDirRemoveCallback(name, opts) ***REMOVED***
  const removeFunction = opts.unsafeCleanup ? _rmdirRecursiveSync : fs.rmdirSync.bind(fs);
  const removeCallback = _prepareRemoveCallback(removeFunction, name);

  if (!opts.keep) ***REMOVED***
    _removeObjects.unshift(removeCallback);
  ***REMOVED***

  return removeCallback;
***REMOVED***

/**
 * Creates a guarded function wrapping the removeFunction call.
 *
 * @param ***REMOVED***Function***REMOVED*** removeFunction
 * @param ***REMOVED***Object***REMOVED*** arg
 * @returns ***REMOVED***Function***REMOVED***
 * @private
 */
function _prepareRemoveCallback(removeFunction, arg) ***REMOVED***
  var called = false;

  return function _cleanupCallback(next) ***REMOVED***
    if (!called) ***REMOVED***
      const index = _removeObjects.indexOf(_cleanupCallback);
      if (index >= 0) ***REMOVED***
        _removeObjects.splice(index, 1);
      ***REMOVED***

      called = true;
      removeFunction(arg);
    ***REMOVED***

    if (next) next(null);
  ***REMOVED***;
***REMOVED***

/**
 * The garbage collector.
 *
 * @private
 */
function _garbageCollector() ***REMOVED***
  if (_uncaughtException && !_gracefulCleanup) ***REMOVED***
    return;
  ***REMOVED***

  // the function being called removes itself from _removeObjects,
  // loop until _removeObjects is empty
  while (_removeObjects.length) ***REMOVED***
    try ***REMOVED***
      _removeObjects[0].call(null);
    ***REMOVED*** catch (e) ***REMOVED***
      // already removed?
    ***REMOVED***
  ***REMOVED***
***REMOVED***

/**
 * Helper for testing against EBADF to compensate changes made to Node 7.x under Windows.
 */
function isEBADF(error) ***REMOVED***
  return isExpectedError(error, -EBADF, 'EBADF');
***REMOVED***

/**
 * Helper for testing against ENOENT to compensate changes made to Node 7.x under Windows.
 */
function isENOENT(error) ***REMOVED***
  return isExpectedError(error, -ENOENT, 'ENOENT');
***REMOVED***

/**
 * Helper to determine whether the expected error code matches the actual code and errno,
 * which will differ between the supported node versions.
 *
 * - Node >= 7.0:
 *   error.code ***REMOVED***String***REMOVED***
 *   error.errno ***REMOVED***String|Number***REMOVED*** any numerical value will be negated
 *
 * - Node >= 6.0 < 7.0:
 *   error.code ***REMOVED***String***REMOVED***
 *   error.errno ***REMOVED***Number***REMOVED*** negated
 *
 * - Node >= 4.0 < 6.0: introduces SystemError
 *   error.code ***REMOVED***String***REMOVED***
 *   error.errno ***REMOVED***Number***REMOVED*** negated
 *
 * - Node >= 0.10 < 4.0:
 *   error.code ***REMOVED***Number***REMOVED*** negated
 *   error.errno n/a
 */
function isExpectedError(error, code, errno) ***REMOVED***
  return error.code == code || error.code == errno;
***REMOVED***

/**
 * Sets the graceful cleanup.
 *
 * Also removes the created files and directories when an uncaught exception occurs.
 */
function setGracefulCleanup() ***REMOVED***
  _gracefulCleanup = true;
***REMOVED***

const version = process.versions.node.split('.').map(function (value) ***REMOVED***
  return parseInt(value, 10);
***REMOVED***);

if (version[0] === 0 && (version[1] < 9 || version[1] === 9 && version[2] < 5)) ***REMOVED***
  process.addListener('uncaughtException', function _uncaughtExceptionThrown(err) ***REMOVED***
    _uncaughtException = true;
    _garbageCollector();

    throw err;
  ***REMOVED***);
***REMOVED***

process.addListener('exit', function _exit(code) ***REMOVED***
  if (code) _uncaughtException = true;
  _garbageCollector();
***REMOVED***);

/**
 * Configuration options.
 *
 * @typedef ***REMOVED***Object***REMOVED*** Options
 * @property ***REMOVED***?number***REMOVED*** tries the number of tries before give up the name generation
 * @property ***REMOVED***?string***REMOVED*** template the "mkstemp" like filename template
 * @property ***REMOVED***?string***REMOVED*** name fix name
 * @property ***REMOVED***?string***REMOVED*** dir the tmp directory to use
 * @property ***REMOVED***?string***REMOVED*** prefix prefix for the generated name
 * @property ***REMOVED***?string***REMOVED*** postfix postfix for the generated name
 */

/**
 * @typedef ***REMOVED***Object***REMOVED*** FileSyncObject
 * @property ***REMOVED***string***REMOVED*** name the name of the file
 * @property ***REMOVED***string***REMOVED*** fd the file descriptor
 * @property ***REMOVED***fileCallback***REMOVED*** removeCallback the callback function to remove the file
 */

/**
 * @typedef ***REMOVED***Object***REMOVED*** DirSyncObject
 * @property ***REMOVED***string***REMOVED*** name the name of the directory
 * @property ***REMOVED***fileCallback***REMOVED*** removeCallback the callback function to remove the directory
 */

/**
 * @callback tmpNameCallback
 * @param ***REMOVED***?Error***REMOVED*** err the error object if anything goes wrong
 * @param ***REMOVED***string***REMOVED*** name the temporary file name
 */

/**
 * @callback fileCallback
 * @param ***REMOVED***?Error***REMOVED*** err the error object if anything goes wrong
 * @param ***REMOVED***string***REMOVED*** name the temporary file name
 * @param ***REMOVED***number***REMOVED*** fd the file descriptor
 * @param ***REMOVED***cleanupCallback***REMOVED*** fn the cleanup callback function
 */

/**
 * @callback dirCallback
 * @param ***REMOVED***?Error***REMOVED*** err the error object if anything goes wrong
 * @param ***REMOVED***string***REMOVED*** name the temporary file name
 * @param ***REMOVED***cleanupCallback***REMOVED*** fn the cleanup callback function
 */

/**
 * Removes the temporary created file or directory.
 *
 * @callback cleanupCallback
 * @param ***REMOVED***simpleCallback***REMOVED*** [next] function to call after entry was removed
 */

/**
 * Callback function for function composition.
 * @see ***REMOVED***@link https://github.com/raszi/node-tmp/issues/57|raszi/node-tmp#57***REMOVED***
 *
 * @callback simpleCallback
 */

// exporting all the needed methods
module.exports.tmpdir = tmpDir;

module.exports.dir = dir;
module.exports.dirSync = dirSync;

module.exports.file = file;
module.exports.fileSync = fileSync;

module.exports.tmpName = tmpName;
module.exports.tmpNameSync = tmpNameSync;

module.exports.setGracefulCleanup = setGracefulCleanup;
