// eslint/conf/config-schema.js
/**
 * @fileoverview Defines a schema for configs.
 * @author Sylvan Mably
 */

"use strict";

const baseConfigProperties = ***REMOVED***
    env: ***REMOVED*** type: "object" ***REMOVED***,
    extends: ***REMOVED*** $ref: "#/definitions/stringOrStrings" ***REMOVED***,
    globals: ***REMOVED*** type: "object" ***REMOVED***,
    overrides: ***REMOVED***
        type: "array",
        items: ***REMOVED*** $ref: "#/definitions/overrideConfig" ***REMOVED***,
        additionalItems: false
    ***REMOVED***,
    parser: ***REMOVED*** type: ["string", "null"] ***REMOVED***,
    parserOptions: ***REMOVED*** type: "object" ***REMOVED***,
    plugins: ***REMOVED*** type: "array" ***REMOVED***,
    processor: ***REMOVED*** type: "string" ***REMOVED***,
    rules: ***REMOVED*** type: "object" ***REMOVED***,
    settings: ***REMOVED*** type: "object" ***REMOVED***,
    noInlineConfig: ***REMOVED*** type: "boolean" ***REMOVED***,
    reportUnusedDisableDirectives: ***REMOVED*** type: "boolean" ***REMOVED***,

    ecmaFeatures: ***REMOVED*** type: "object" ***REMOVED*** // deprecated; logs a warning when used
***REMOVED***;

const configSchema = ***REMOVED***
    definitions: ***REMOVED***
        stringOrStrings: ***REMOVED***
            oneOf: [
                ***REMOVED*** type: "string" ***REMOVED***,
                ***REMOVED***
                    type: "array",
                    items: ***REMOVED*** type: "string" ***REMOVED***,
                    additionalItems: false
                ***REMOVED***
            ]
        ***REMOVED***,
        stringOrStringsRequired: ***REMOVED***
            oneOf: [
                ***REMOVED*** type: "string" ***REMOVED***,
                ***REMOVED***
                    type: "array",
                    items: ***REMOVED*** type: "string" ***REMOVED***,
                    additionalItems: false,
                    minItems: 1
                ***REMOVED***
            ]
        ***REMOVED***,

        // Config at top-level.
        objectConfig: ***REMOVED***
            type: "object",
            properties: ***REMOVED***
                root: ***REMOVED*** type: "boolean" ***REMOVED***,
                ...baseConfigProperties
            ***REMOVED***,
            additionalProperties: false
        ***REMOVED***,

        // Config in `overrides`.
        overrideConfig: ***REMOVED***
            type: "object",
            properties: ***REMOVED***
                excludedFiles: ***REMOVED*** $ref: "#/definitions/stringOrStrings" ***REMOVED***,
                files: ***REMOVED*** $ref: "#/definitions/stringOrStringsRequired" ***REMOVED***,
                ...baseConfigProperties
            ***REMOVED***,
            required: ["files"],
            additionalProperties: false
        ***REMOVED***
    ***REMOVED***,

    $ref: "#/definitions/objectConfig"
***REMOVED***;

module.exports = configSchema;
