// eslint/conf/environments.js
/**
 * @fileoverview Defines environment settings and globals.
 * @author Elan Shanker
 */
"use strict";

//------------------------------------------------------------------------------
// Requirements
//------------------------------------------------------------------------------

const globals = require("globals");

//------------------------------------------------------------------------------
// Helpers
//------------------------------------------------------------------------------

/**
 * Get the object that has differentce.
 * @param ***REMOVED***Record<string,boolean>***REMOVED*** current The newer object.
 * @param ***REMOVED***Record<string,boolean>***REMOVED*** prev The older object.
 * @returns ***REMOVED***Record<string,boolean>***REMOVED*** The difference object.
 */
function getDiff(current, prev) ***REMOVED***
    const retv = ***REMOVED******REMOVED***;

    for (const [key, value] of Object.entries(current)) ***REMOVED***
        if (!Object.hasOwnProperty.call(prev, key)) ***REMOVED***
            retv[key] = value;
        ***REMOVED***
    ***REMOVED***

    return retv;
***REMOVED***

const newGlobals2015 = getDiff(globals.es2015, globals.es5); // 19 variables such as Promise, Map, ...
const newGlobals2017 = ***REMOVED***
    Atomics: false,
    SharedArrayBuffer: false
***REMOVED***;
const newGlobals2020 = ***REMOVED***
    BigInt: false,
    BigInt64Array: false,
    BigUint64Array: false
***REMOVED***;

//------------------------------------------------------------------------------
// Public Interface
//------------------------------------------------------------------------------

/** @type ***REMOVED***Map<string, import("../lib/shared/types").Environment>***REMOVED*** */
module.exports = new Map(Object.entries(***REMOVED***

    // Language
    builtin: ***REMOVED***
        globals: globals.es5
    ***REMOVED***,
    es6: ***REMOVED***
        globals: newGlobals2015,
        parserOptions: ***REMOVED***
            ecmaVersion: 6
        ***REMOVED***
    ***REMOVED***,
    es2015: ***REMOVED***
        globals: newGlobals2015,
        parserOptions: ***REMOVED***
            ecmaVersion: 6
        ***REMOVED***
    ***REMOVED***,
    es2017: ***REMOVED***
        globals: ***REMOVED*** ...newGlobals2015, ...newGlobals2017 ***REMOVED***,
        parserOptions: ***REMOVED***
            ecmaVersion: 8
        ***REMOVED***
    ***REMOVED***,
    es2020: ***REMOVED***
        globals: ***REMOVED*** ...newGlobals2015, ...newGlobals2017, ...newGlobals2020 ***REMOVED***,
        parserOptions: ***REMOVED***
            ecmaVersion: 11
        ***REMOVED***
    ***REMOVED***,

    // Platforms
    browser: ***REMOVED***
        globals: globals.browser
    ***REMOVED***,
    node: ***REMOVED***
        globals: globals.node,
        parserOptions: ***REMOVED***
            ecmaFeatures: ***REMOVED***
                globalReturn: true
            ***REMOVED***
        ***REMOVED***
    ***REMOVED***,
    "shared-node-browser": ***REMOVED***
        globals: globals["shared-node-browser"]
    ***REMOVED***,
    worker: ***REMOVED***
        globals: globals.worker
    ***REMOVED***,
    serviceworker: ***REMOVED***
        globals: globals.serviceworker
    ***REMOVED***,

    // Frameworks
    commonjs: ***REMOVED***
        globals: globals.commonjs,
        parserOptions: ***REMOVED***
            ecmaFeatures: ***REMOVED***
                globalReturn: true
            ***REMOVED***
        ***REMOVED***
    ***REMOVED***,
    amd: ***REMOVED***
        globals: globals.amd
    ***REMOVED***,
    mocha: ***REMOVED***
        globals: globals.mocha
    ***REMOVED***,
    jasmine: ***REMOVED***
        globals: globals.jasmine
    ***REMOVED***,
    jest: ***REMOVED***
        globals: globals.jest
    ***REMOVED***,
    phantomjs: ***REMOVED***
        globals: globals.phantomjs
    ***REMOVED***,
    jquery: ***REMOVED***
        globals: globals.jquery
    ***REMOVED***,
    qunit: ***REMOVED***
        globals: globals.qunit
    ***REMOVED***,
    prototypejs: ***REMOVED***
        globals: globals.prototypejs
    ***REMOVED***,
    shelljs: ***REMOVED***
        globals: globals.shelljs
    ***REMOVED***,
    meteor: ***REMOVED***
        globals: globals.meteor
    ***REMOVED***,
    mongo: ***REMOVED***
        globals: globals.mongo
    ***REMOVED***,
    protractor: ***REMOVED***
        globals: globals.protractor
    ***REMOVED***,
    applescript: ***REMOVED***
        globals: globals.applescript
    ***REMOVED***,
    nashorn: ***REMOVED***
        globals: globals.nashorn
    ***REMOVED***,
    atomtest: ***REMOVED***
        globals: globals.atomtest
    ***REMOVED***,
    embertest: ***REMOVED***
        globals: globals.embertest
    ***REMOVED***,
    webextensions: ***REMOVED***
        globals: globals.webextensions
    ***REMOVED***,
    greasemonkey: ***REMOVED***
        globals: globals.greasemonkey
    ***REMOVED***
***REMOVED***));
