// graceful-fs/graceful-fs.js
var fs = require('fs')
var polyfills = require('./polyfills.js')
var legacy = require('./legacy-streams.js')
var clone = require('./clone.js')

var util = require('util')

/* istanbul ignore next - node 0.x polyfill */
var gracefulQueue
var previousSymbol

/* istanbul ignore else - node 0.x polyfill */
if (typeof Symbol === 'function' && typeof Symbol.for === 'function') ***REMOVED***
  gracefulQueue = Symbol.for('graceful-fs.queue')
  // This is used in testing by future versions
  previousSymbol = Symbol.for('graceful-fs.previous')
***REMOVED*** else ***REMOVED***
  gracefulQueue = '___graceful-fs.queue'
  previousSymbol = '___graceful-fs.previous'
***REMOVED***

function noop () ***REMOVED******REMOVED***

var debug = noop
if (util.debuglog)
  debug = util.debuglog('gfs4')
else if (/\bgfs4\b/i.test(process.env.NODE_DEBUG || ''))
  debug = function() ***REMOVED***
    var m = util.format.apply(util, arguments)
    m = 'GFS4: ' + m.split(/\n/).join('\nGFS4: ')
    console.error(m)
  ***REMOVED***

// Once time initialization
if (!global[gracefulQueue]) ***REMOVED***
  // This queue can be shared by multiple loaded instances
  var queue = []
  Object.defineProperty(global, gracefulQueue, ***REMOVED***
    get: function() ***REMOVED***
      return queue
    ***REMOVED***
  ***REMOVED***)

  // Patch fs.close/closeSync to shared queue version, because we need
  // to retry() whenever a close happens *anywhere* in the program.
  // This is essential when multiple graceful-fs instances are
  // in play at the same time.
  fs.close = (function (fs$close) ***REMOVED***
    function close (fd, cb) ***REMOVED***
      return fs$close.call(fs, fd, function (err) ***REMOVED***
        // This function uses the graceful-fs shared queue
        if (!err) ***REMOVED***
          retry()
        ***REMOVED***

        if (typeof cb === 'function')
          cb.apply(this, arguments)
      ***REMOVED***)
    ***REMOVED***

    Object.defineProperty(close, previousSymbol, ***REMOVED***
      value: fs$close
    ***REMOVED***)
    return close
  ***REMOVED***)(fs.close)

  fs.closeSync = (function (fs$closeSync) ***REMOVED***
    function closeSync (fd) ***REMOVED***
      // This function uses the graceful-fs shared queue
      fs$closeSync.apply(fs, arguments)
      retry()
    ***REMOVED***

    Object.defineProperty(closeSync, previousSymbol, ***REMOVED***
      value: fs$closeSync
    ***REMOVED***)
    return closeSync
  ***REMOVED***)(fs.closeSync)

  if (/\bgfs4\b/i.test(process.env.NODE_DEBUG || '')) ***REMOVED***
    process.on('exit', function() ***REMOVED***
      debug(global[gracefulQueue])
      require('assert').equal(global[gracefulQueue].length, 0)
    ***REMOVED***)
  ***REMOVED***
***REMOVED***

module.exports = patch(clone(fs))
if (process.env.TEST_GRACEFUL_FS_GLOBAL_PATCH && !fs.__patched) ***REMOVED***
    module.exports = patch(fs)
    fs.__patched = true;
***REMOVED***

function patch (fs) ***REMOVED***
  // Everything that references the open() function needs to be in here
  polyfills(fs)
  fs.gracefulify = patch

  fs.createReadStream = createReadStream
  fs.createWriteStream = createWriteStream
  var fs$readFile = fs.readFile
  fs.readFile = readFile
  function readFile (path, options, cb) ***REMOVED***
    if (typeof options === 'function')
      cb = options, options = null

    return go$readFile(path, options, cb)

    function go$readFile (path, options, cb) ***REMOVED***
      return fs$readFile(path, options, function (err) ***REMOVED***
        if (err && (err.code === 'EMFILE' || err.code === 'ENFILE'))
          enqueue([go$readFile, [path, options, cb]])
        else ***REMOVED***
          if (typeof cb === 'function')
            cb.apply(this, arguments)
          retry()
        ***REMOVED***
      ***REMOVED***)
    ***REMOVED***
  ***REMOVED***

  var fs$writeFile = fs.writeFile
  fs.writeFile = writeFile
  function writeFile (path, data, options, cb) ***REMOVED***
    if (typeof options === 'function')
      cb = options, options = null

    return go$writeFile(path, data, options, cb)

    function go$writeFile (path, data, options, cb) ***REMOVED***
      return fs$writeFile(path, data, options, function (err) ***REMOVED***
        if (err && (err.code === 'EMFILE' || err.code === 'ENFILE'))
          enqueue([go$writeFile, [path, data, options, cb]])
        else ***REMOVED***
          if (typeof cb === 'function')
            cb.apply(this, arguments)
          retry()
        ***REMOVED***
      ***REMOVED***)
    ***REMOVED***
  ***REMOVED***

  var fs$appendFile = fs.appendFile
  if (fs$appendFile)
    fs.appendFile = appendFile
  function appendFile (path, data, options, cb) ***REMOVED***
    if (typeof options === 'function')
      cb = options, options = null

    return go$appendFile(path, data, options, cb)

    function go$appendFile (path, data, options, cb) ***REMOVED***
      return fs$appendFile(path, data, options, function (err) ***REMOVED***
        if (err && (err.code === 'EMFILE' || err.code === 'ENFILE'))
          enqueue([go$appendFile, [path, data, options, cb]])
        else ***REMOVED***
          if (typeof cb === 'function')
            cb.apply(this, arguments)
          retry()
        ***REMOVED***
      ***REMOVED***)
    ***REMOVED***
  ***REMOVED***

  var fs$readdir = fs.readdir
  fs.readdir = readdir
  function readdir (path, options, cb) ***REMOVED***
    var args = [path]
    if (typeof options !== 'function') ***REMOVED***
      args.push(options)
    ***REMOVED*** else ***REMOVED***
      cb = options
    ***REMOVED***
    args.push(go$readdir$cb)

    return go$readdir(args)

    function go$readdir$cb (err, files) ***REMOVED***
      if (files && files.sort)
        files.sort()

      if (err && (err.code === 'EMFILE' || err.code === 'ENFILE'))
        enqueue([go$readdir, [args]])

      else ***REMOVED***
        if (typeof cb === 'function')
          cb.apply(this, arguments)
        retry()
      ***REMOVED***
    ***REMOVED***
  ***REMOVED***

  function go$readdir (args) ***REMOVED***
    return fs$readdir.apply(fs, args)
  ***REMOVED***

  if (process.version.substr(0, 4) === 'v0.8') ***REMOVED***
    var legStreams = legacy(fs)
    ReadStream = legStreams.ReadStream
    WriteStream = legStreams.WriteStream
  ***REMOVED***

  var fs$ReadStream = fs.ReadStream
  if (fs$ReadStream) ***REMOVED***
    ReadStream.prototype = Object.create(fs$ReadStream.prototype)
    ReadStream.prototype.open = ReadStream$open
  ***REMOVED***

  var fs$WriteStream = fs.WriteStream
  if (fs$WriteStream) ***REMOVED***
    WriteStream.prototype = Object.create(fs$WriteStream.prototype)
    WriteStream.prototype.open = WriteStream$open
  ***REMOVED***

  Object.defineProperty(fs, 'ReadStream', ***REMOVED***
    get: function () ***REMOVED***
      return ReadStream
    ***REMOVED***,
    set: function (val) ***REMOVED***
      ReadStream = val
    ***REMOVED***,
    enumerable: true,
    configurable: true
  ***REMOVED***)
  Object.defineProperty(fs, 'WriteStream', ***REMOVED***
    get: function () ***REMOVED***
      return WriteStream
    ***REMOVED***,
    set: function (val) ***REMOVED***
      WriteStream = val
    ***REMOVED***,
    enumerable: true,
    configurable: true
  ***REMOVED***)

  // legacy names
  var FileReadStream = ReadStream
  Object.defineProperty(fs, 'FileReadStream', ***REMOVED***
    get: function () ***REMOVED***
      return FileReadStream
    ***REMOVED***,
    set: function (val) ***REMOVED***
      FileReadStream = val
    ***REMOVED***,
    enumerable: true,
    configurable: true
  ***REMOVED***)
  var FileWriteStream = WriteStream
  Object.defineProperty(fs, 'FileWriteStream', ***REMOVED***
    get: function () ***REMOVED***
      return FileWriteStream
    ***REMOVED***,
    set: function (val) ***REMOVED***
      FileWriteStream = val
    ***REMOVED***,
    enumerable: true,
    configurable: true
  ***REMOVED***)

  function ReadStream (path, options) ***REMOVED***
    if (this instanceof ReadStream)
      return fs$ReadStream.apply(this, arguments), this
    else
      return ReadStream.apply(Object.create(ReadStream.prototype), arguments)
  ***REMOVED***

  function ReadStream$open () ***REMOVED***
    var that = this
    open(that.path, that.flags, that.mode, function (err, fd) ***REMOVED***
      if (err) ***REMOVED***
        if (that.autoClose)
          that.destroy()

        that.emit('error', err)
      ***REMOVED*** else ***REMOVED***
        that.fd = fd
        that.emit('open', fd)
        that.read()
      ***REMOVED***
    ***REMOVED***)
  ***REMOVED***

  function WriteStream (path, options) ***REMOVED***
    if (this instanceof WriteStream)
      return fs$WriteStream.apply(this, arguments), this
    else
      return WriteStream.apply(Object.create(WriteStream.prototype), arguments)
  ***REMOVED***

  function WriteStream$open () ***REMOVED***
    var that = this
    open(that.path, that.flags, that.mode, function (err, fd) ***REMOVED***
      if (err) ***REMOVED***
        that.destroy()
        that.emit('error', err)
      ***REMOVED*** else ***REMOVED***
        that.fd = fd
        that.emit('open', fd)
      ***REMOVED***
    ***REMOVED***)
  ***REMOVED***

  function createReadStream (path, options) ***REMOVED***
    return new fs.ReadStream(path, options)
  ***REMOVED***

  function createWriteStream (path, options) ***REMOVED***
    return new fs.WriteStream(path, options)
  ***REMOVED***

  var fs$open = fs.open
  fs.open = open
  function open (path, flags, mode, cb) ***REMOVED***
    if (typeof mode === 'function')
      cb = mode, mode = null

    return go$open(path, flags, mode, cb)

    function go$open (path, flags, mode, cb) ***REMOVED***
      return fs$open(path, flags, mode, function (err, fd) ***REMOVED***
        if (err && (err.code === 'EMFILE' || err.code === 'ENFILE'))
          enqueue([go$open, [path, flags, mode, cb]])
        else ***REMOVED***
          if (typeof cb === 'function')
            cb.apply(this, arguments)
          retry()
        ***REMOVED***
      ***REMOVED***)
    ***REMOVED***
  ***REMOVED***

  return fs
***REMOVED***

function enqueue (elem) ***REMOVED***
  debug('ENQUEUE', elem[0].name, elem[1])
  global[gracefulQueue].push(elem)
***REMOVED***

function retry () ***REMOVED***
  var elem = global[gracefulQueue].shift()
  if (elem) ***REMOVED***
    debug('RETRY', elem[0].name, elem[1])
    elem[0].apply(null, elem[1])
  ***REMOVED***
***REMOVED***
