// rxjs/_esm5/internal/util/Immediate.js
/** PURE_IMPORTS_START  PURE_IMPORTS_END */
var nextHandle = 1;
var RESOLVED = /*@__PURE__*/ (function () ***REMOVED*** return /*@__PURE__*/ Promise.resolve(); ***REMOVED***)();
var activeHandles = ***REMOVED******REMOVED***;
function findAndClearHandle(handle) ***REMOVED***
    if (handle in activeHandles) ***REMOVED***
        delete activeHandles[handle];
        return true;
    ***REMOVED***
    return false;
***REMOVED***
export var Immediate = ***REMOVED***
    setImmediate: function (cb) ***REMOVED***
        var handle = nextHandle++;
        activeHandles[handle] = true;
        RESOLVED.then(function () ***REMOVED*** return findAndClearHandle(handle) && cb(); ***REMOVED***);
        return handle;
    ***REMOVED***,
    clearImmediate: function (handle) ***REMOVED***
        findAndClearHandle(handle);
    ***REMOVED***,
***REMOVED***;
export var TestTools = ***REMOVED***
    pending: function () ***REMOVED***
        return Object.keys(activeHandles).length;
    ***REMOVED***
***REMOVED***;
