// eslint/lib/linter/code-path-analysis/code-path-analyzer.js
/**
 * @fileoverview A class of the code path analyzer.
 * @author Toru Nagashima
 */

"use strict";

//------------------------------------------------------------------------------
// Requirements
//------------------------------------------------------------------------------

const assert = require("assert"),
    ***REMOVED*** breakableTypePattern ***REMOVED*** = require("../../shared/ast-utils"),
    CodePath = require("./code-path"),
    CodePathSegment = require("./code-path-segment"),
    IdGenerator = require("./id-generator"),
    debug = require("./debug-helpers");

//------------------------------------------------------------------------------
// Helpers
//------------------------------------------------------------------------------

/**
 * Checks whether or not a given node is a `case` node (not `default` node).
 *
 * @param ***REMOVED***ASTNode***REMOVED*** node - A `SwitchCase` node to check.
 * @returns ***REMOVED***boolean***REMOVED*** `true` if the node is a `case` node (not `default` node).
 */
function isCaseNode(node) ***REMOVED***
    return Boolean(node.test);
***REMOVED***

/**
 * Checks whether the given logical operator is taken into account for the code
 * path analysis.
 *
 * @param ***REMOVED***string***REMOVED*** operator - The operator found in the LogicalExpression node
 * @returns ***REMOVED***boolean***REMOVED*** `true` if the operator is "&&" or "||"
 */
function isHandledLogicalOperator(operator) ***REMOVED***
    return operator === "&&" || operator === "||";
***REMOVED***

/**
 * Gets the label if the parent node of a given node is a LabeledStatement.
 *
 * @param ***REMOVED***ASTNode***REMOVED*** node - A node to get.
 * @returns ***REMOVED***string|null***REMOVED*** The label or `null`.
 */
function getLabel(node) ***REMOVED***
    if (node.parent.type === "LabeledStatement") ***REMOVED***
        return node.parent.label.name;
    ***REMOVED***
    return null;
***REMOVED***

/**
 * Checks whether or not a given logical expression node goes different path
 * between the `true` case and the `false` case.
 *
 * @param ***REMOVED***ASTNode***REMOVED*** node - A node to check.
 * @returns ***REMOVED***boolean***REMOVED*** `true` if the node is a test of a choice statement.
 */
function isForkingByTrueOrFalse(node) ***REMOVED***
    const parent = node.parent;

    switch (parent.type) ***REMOVED***
        case "ConditionalExpression":
        case "IfStatement":
        case "WhileStatement":
        case "DoWhileStatement":
        case "ForStatement":
            return parent.test === node;

        case "LogicalExpression":
            return isHandledLogicalOperator(parent.operator);

        default:
            return false;
    ***REMOVED***
***REMOVED***

/**
 * Gets the boolean value of a given literal node.
 *
 * This is used to detect infinity loops (e.g. `while (true) ***REMOVED******REMOVED***`).
 * Statements preceded by an infinity loop are unreachable if the loop didn't
 * have any `break` statement.
 *
 * @param ***REMOVED***ASTNode***REMOVED*** node - A node to get.
 * @returns ***REMOVED***boolean|undefined***REMOVED*** a boolean value if the node is a Literal node,
 *   otherwise `undefined`.
 */
function getBooleanValueIfSimpleConstant(node) ***REMOVED***
    if (node.type === "Literal") ***REMOVED***
        return Boolean(node.value);
    ***REMOVED***
    return void 0;
***REMOVED***

/**
 * Checks that a given identifier node is a reference or not.
 *
 * This is used to detect the first throwable node in a `try` block.
 *
 * @param ***REMOVED***ASTNode***REMOVED*** node - An Identifier node to check.
 * @returns ***REMOVED***boolean***REMOVED*** `true` if the node is a reference.
 */
function isIdentifierReference(node) ***REMOVED***
    const parent = node.parent;

    switch (parent.type) ***REMOVED***
        case "LabeledStatement":
        case "BreakStatement":
        case "ContinueStatement":
        case "ArrayPattern":
        case "RestElement":
        case "ImportSpecifier":
        case "ImportDefaultSpecifier":
        case "ImportNamespaceSpecifier":
        case "CatchClause":
            return false;

        case "FunctionDeclaration":
        case "FunctionExpression":
        case "ArrowFunctionExpression":
        case "ClassDeclaration":
        case "ClassExpression":
        case "VariableDeclarator":
            return parent.id !== node;

        case "Property":
        case "MethodDefinition":
            return (
                parent.key !== node ||
                parent.computed ||
                parent.shorthand
            );

        case "AssignmentPattern":
            return parent.key !== node;

        default:
            return true;
    ***REMOVED***
***REMOVED***

/**
 * Updates the current segment with the head segment.
 * This is similar to local branches and tracking branches of git.
 *
 * To separate the current and the head is in order to not make useless segments.
 *
 * In this process, both "onCodePathSegmentStart" and "onCodePathSegmentEnd"
 * events are fired.
 *
 * @param ***REMOVED***CodePathAnalyzer***REMOVED*** analyzer - The instance.
 * @param ***REMOVED***ASTNode***REMOVED*** node - The current AST node.
 * @returns ***REMOVED***void***REMOVED***
 */
function forwardCurrentToHead(analyzer, node) ***REMOVED***
    const codePath = analyzer.codePath;
    const state = CodePath.getState(codePath);
    const currentSegments = state.currentSegments;
    const headSegments = state.headSegments;
    const end = Math.max(currentSegments.length, headSegments.length);
    let i, currentSegment, headSegment;

    // Fires leaving events.
    for (i = 0; i < end; ++i) ***REMOVED***
        currentSegment = currentSegments[i];
        headSegment = headSegments[i];

        if (currentSegment !== headSegment && currentSegment) ***REMOVED***
            debug.dump(`onCodePathSegmentEnd $***REMOVED***currentSegment.id***REMOVED***`);

            if (currentSegment.reachable) ***REMOVED***
                analyzer.emitter.emit(
                    "onCodePathSegmentEnd",
                    currentSegment,
                    node
                );
            ***REMOVED***
        ***REMOVED***
    ***REMOVED***

    // Update state.
    state.currentSegments = headSegments;

    // Fires entering events.
    for (i = 0; i < end; ++i) ***REMOVED***
        currentSegment = currentSegments[i];
        headSegment = headSegments[i];

        if (currentSegment !== headSegment && headSegment) ***REMOVED***
            debug.dump(`onCodePathSegmentStart $***REMOVED***headSegment.id***REMOVED***`);

            CodePathSegment.markUsed(headSegment);
            if (headSegment.reachable) ***REMOVED***
                analyzer.emitter.emit(
                    "onCodePathSegmentStart",
                    headSegment,
                    node
                );
            ***REMOVED***
        ***REMOVED***
    ***REMOVED***

***REMOVED***

/**
 * Updates the current segment with empty.
 * This is called at the last of functions or the program.
 *
 * @param ***REMOVED***CodePathAnalyzer***REMOVED*** analyzer - The instance.
 * @param ***REMOVED***ASTNode***REMOVED*** node - The current AST node.
 * @returns ***REMOVED***void***REMOVED***
 */
function leaveFromCurrentSegment(analyzer, node) ***REMOVED***
    const state = CodePath.getState(analyzer.codePath);
    const currentSegments = state.currentSegments;

    for (let i = 0; i < currentSegments.length; ++i) ***REMOVED***
        const currentSegment = currentSegments[i];

        debug.dump(`onCodePathSegmentEnd $***REMOVED***currentSegment.id***REMOVED***`);
        if (currentSegment.reachable) ***REMOVED***
            analyzer.emitter.emit(
                "onCodePathSegmentEnd",
                currentSegment,
                node
            );
        ***REMOVED***
    ***REMOVED***

    state.currentSegments = [];
***REMOVED***

/**
 * Updates the code path due to the position of a given node in the parent node
 * thereof.
 *
 * For example, if the node is `parent.consequent`, this creates a fork from the
 * current path.
 *
 * @param ***REMOVED***CodePathAnalyzer***REMOVED*** analyzer - The instance.
 * @param ***REMOVED***ASTNode***REMOVED*** node - The current AST node.
 * @returns ***REMOVED***void***REMOVED***
 */
function preprocess(analyzer, node) ***REMOVED***
    const codePath = analyzer.codePath;
    const state = CodePath.getState(codePath);
    const parent = node.parent;

    switch (parent.type) ***REMOVED***
        case "LogicalExpression":
            if (
                parent.right === node &&
                isHandledLogicalOperator(parent.operator)
            ) ***REMOVED***
                state.makeLogicalRight();
            ***REMOVED***
            break;

        case "ConditionalExpression":
        case "IfStatement":

            /*
             * Fork if this node is at `consequent`/`alternate`.
             * `popForkContext()` exists at `IfStatement:exit` and
             * `ConditionalExpression:exit`.
             */
            if (parent.consequent === node) ***REMOVED***
                state.makeIfConsequent();
            ***REMOVED*** else if (parent.alternate === node) ***REMOVED***
                state.makeIfAlternate();
            ***REMOVED***
            break;

        case "SwitchCase":
            if (parent.consequent[0] === node) ***REMOVED***
                state.makeSwitchCaseBody(false, !parent.test);
            ***REMOVED***
            break;

        case "TryStatement":
            if (parent.handler === node) ***REMOVED***
                state.makeCatchBlock();
            ***REMOVED*** else if (parent.finalizer === node) ***REMOVED***
                state.makeFinallyBlock();
            ***REMOVED***
            break;

        case "WhileStatement":
            if (parent.test === node) ***REMOVED***
                state.makeWhileTest(getBooleanValueIfSimpleConstant(node));
            ***REMOVED*** else ***REMOVED***
                assert(parent.body === node);
                state.makeWhileBody();
            ***REMOVED***
            break;

        case "DoWhileStatement":
            if (parent.body === node) ***REMOVED***
                state.makeDoWhileBody();
            ***REMOVED*** else ***REMOVED***
                assert(parent.test === node);
                state.makeDoWhileTest(getBooleanValueIfSimpleConstant(node));
            ***REMOVED***
            break;

        case "ForStatement":
            if (parent.test === node) ***REMOVED***
                state.makeForTest(getBooleanValueIfSimpleConstant(node));
            ***REMOVED*** else if (parent.update === node) ***REMOVED***
                state.makeForUpdate();
            ***REMOVED*** else if (parent.body === node) ***REMOVED***
                state.makeForBody();
            ***REMOVED***
            break;

        case "ForInStatement":
        case "ForOfStatement":
            if (parent.left === node) ***REMOVED***
                state.makeForInOfLeft();
            ***REMOVED*** else if (parent.right === node) ***REMOVED***
                state.makeForInOfRight();
            ***REMOVED*** else ***REMOVED***
                assert(parent.body === node);
                state.makeForInOfBody();
            ***REMOVED***
            break;

        case "AssignmentPattern":

            /*
             * Fork if this node is at `right`.
             * `left` is executed always, so it uses the current path.
             * `popForkContext()` exists at `AssignmentPattern:exit`.
             */
            if (parent.right === node) ***REMOVED***
                state.pushForkContext();
                state.forkBypassPath();
                state.forkPath();
            ***REMOVED***
            break;

        default:
            break;
    ***REMOVED***
***REMOVED***

/**
 * Updates the code path due to the type of a given node in entering.
 *
 * @param ***REMOVED***CodePathAnalyzer***REMOVED*** analyzer - The instance.
 * @param ***REMOVED***ASTNode***REMOVED*** node - The current AST node.
 * @returns ***REMOVED***void***REMOVED***
 */
function processCodePathToEnter(analyzer, node) ***REMOVED***
    let codePath = analyzer.codePath;
    let state = codePath && CodePath.getState(codePath);
    const parent = node.parent;

    switch (node.type) ***REMOVED***
        case "Program":
        case "FunctionDeclaration":
        case "FunctionExpression":
        case "ArrowFunctionExpression":
            if (codePath) ***REMOVED***

                // Emits onCodePathSegmentStart events if updated.
                forwardCurrentToHead(analyzer, node);
                debug.dumpState(node, state, false);
            ***REMOVED***

            // Create the code path of this scope.
            codePath = analyzer.codePath = new CodePath(
                analyzer.idGenerator.next(),
                codePath,
                analyzer.onLooped
            );
            state = CodePath.getState(codePath);

            // Emits onCodePathStart events.
            debug.dump(`onCodePathStart $***REMOVED***codePath.id***REMOVED***`);
            analyzer.emitter.emit("onCodePathStart", codePath, node);
            break;

        case "LogicalExpression":
            if (isHandledLogicalOperator(node.operator)) ***REMOVED***
                state.pushChoiceContext(
                    node.operator,
                    isForkingByTrueOrFalse(node)
                );
            ***REMOVED***
            break;

        case "ConditionalExpression":
        case "IfStatement":
            state.pushChoiceContext("test", false);
            break;

        case "SwitchStatement":
            state.pushSwitchContext(
                node.cases.some(isCaseNode),
                getLabel(node)
            );
            break;

        case "TryStatement":
            state.pushTryContext(Boolean(node.finalizer));
            break;

        case "SwitchCase":

            /*
             * Fork if this node is after the 2st node in `cases`.
             * It's similar to `else` blocks.
             * The next `test` node is processed in this path.
             */
            if (parent.discriminant !== node && parent.cases[0] !== node) ***REMOVED***
                state.forkPath();
            ***REMOVED***
            break;

        case "WhileStatement":
        case "DoWhileStatement":
        case "ForStatement":
        case "ForInStatement":
        case "ForOfStatement":
            state.pushLoopContext(node.type, getLabel(node));
            break;

        case "LabeledStatement":
            if (!breakableTypePattern.test(node.body.type)) ***REMOVED***
                state.pushBreakContext(false, node.label.name);
            ***REMOVED***
            break;

        default:
            break;
    ***REMOVED***

    // Emits onCodePathSegmentStart events if updated.
    forwardCurrentToHead(analyzer, node);
    debug.dumpState(node, state, false);
***REMOVED***

/**
 * Updates the code path due to the type of a given node in leaving.
 *
 * @param ***REMOVED***CodePathAnalyzer***REMOVED*** analyzer - The instance.
 * @param ***REMOVED***ASTNode***REMOVED*** node - The current AST node.
 * @returns ***REMOVED***void***REMOVED***
 */
function processCodePathToExit(analyzer, node) ***REMOVED***
    const codePath = analyzer.codePath;
    const state = CodePath.getState(codePath);
    let dontForward = false;

    switch (node.type) ***REMOVED***
        case "IfStatement":
        case "ConditionalExpression":
            state.popChoiceContext();
            break;

        case "LogicalExpression":
            if (isHandledLogicalOperator(node.operator)) ***REMOVED***
                state.popChoiceContext();
            ***REMOVED***
            break;

        case "SwitchStatement":
            state.popSwitchContext();
            break;

        case "SwitchCase":

            /*
             * This is the same as the process at the 1st `consequent` node in
             * `preprocess` function.
             * Must do if this `consequent` is empty.
             */
            if (node.consequent.length === 0) ***REMOVED***
                state.makeSwitchCaseBody(true, !node.test);
            ***REMOVED***
            if (state.forkContext.reachable) ***REMOVED***
                dontForward = true;
            ***REMOVED***
            break;

        case "TryStatement":
            state.popTryContext();
            break;

        case "BreakStatement":
            forwardCurrentToHead(analyzer, node);
            state.makeBreak(node.label && node.label.name);
            dontForward = true;
            break;

        case "ContinueStatement":
            forwardCurrentToHead(analyzer, node);
            state.makeContinue(node.label && node.label.name);
            dontForward = true;
            break;

        case "ReturnStatement":
            forwardCurrentToHead(analyzer, node);
            state.makeReturn();
            dontForward = true;
            break;

        case "ThrowStatement":
            forwardCurrentToHead(analyzer, node);
            state.makeThrow();
            dontForward = true;
            break;

        case "Identifier":
            if (isIdentifierReference(node)) ***REMOVED***
                state.makeFirstThrowablePathInTryBlock();
                dontForward = true;
            ***REMOVED***
            break;

        case "CallExpression":
        case "ImportExpression":
        case "MemberExpression":
        case "NewExpression":
            state.makeFirstThrowablePathInTryBlock();
            break;

        case "WhileStatement":
        case "DoWhileStatement":
        case "ForStatement":
        case "ForInStatement":
        case "ForOfStatement":
            state.popLoopContext();
            break;

        case "AssignmentPattern":
            state.popForkContext();
            break;

        case "LabeledStatement":
            if (!breakableTypePattern.test(node.body.type)) ***REMOVED***
                state.popBreakContext();
            ***REMOVED***
            break;

        default:
            break;
    ***REMOVED***

    // Emits onCodePathSegmentStart events if updated.
    if (!dontForward) ***REMOVED***
        forwardCurrentToHead(analyzer, node);
    ***REMOVED***
    debug.dumpState(node, state, true);
***REMOVED***

/**
 * Updates the code path to finalize the current code path.
 *
 * @param ***REMOVED***CodePathAnalyzer***REMOVED*** analyzer - The instance.
 * @param ***REMOVED***ASTNode***REMOVED*** node - The current AST node.
 * @returns ***REMOVED***void***REMOVED***
 */
function postprocess(analyzer, node) ***REMOVED***
    switch (node.type) ***REMOVED***
        case "Program":
        case "FunctionDeclaration":
        case "FunctionExpression":
        case "ArrowFunctionExpression": ***REMOVED***
            let codePath = analyzer.codePath;

            // Mark the current path as the final node.
            CodePath.getState(codePath).makeFinal();

            // Emits onCodePathSegmentEnd event of the current segments.
            leaveFromCurrentSegment(analyzer, node);

            // Emits onCodePathEnd event of this code path.
            debug.dump(`onCodePathEnd $***REMOVED***codePath.id***REMOVED***`);
            analyzer.emitter.emit("onCodePathEnd", codePath, node);
            debug.dumpDot(codePath);

            codePath = analyzer.codePath = analyzer.codePath.upper;
            if (codePath) ***REMOVED***
                debug.dumpState(node, CodePath.getState(codePath), true);
            ***REMOVED***
            break;
        ***REMOVED***

        default:
            break;
    ***REMOVED***
***REMOVED***

//------------------------------------------------------------------------------
// Public Interface
//------------------------------------------------------------------------------

/**
 * The class to analyze code paths.
 * This class implements the EventGenerator interface.
 */
class CodePathAnalyzer ***REMOVED***

    /**
     * @param ***REMOVED***EventGenerator***REMOVED*** eventGenerator - An event generator to wrap.
     */
    constructor(eventGenerator) ***REMOVED***
        this.original = eventGenerator;
        this.emitter = eventGenerator.emitter;
        this.codePath = null;
        this.idGenerator = new IdGenerator("s");
        this.currentNode = null;
        this.onLooped = this.onLooped.bind(this);
    ***REMOVED***

    /**
     * Does the process to enter a given AST node.
     * This updates state of analysis and calls `enterNode` of the wrapped.
     *
     * @param ***REMOVED***ASTNode***REMOVED*** node - A node which is entering.
     * @returns ***REMOVED***void***REMOVED***
     */
    enterNode(node) ***REMOVED***
        this.currentNode = node;

        // Updates the code path due to node's position in its parent node.
        if (node.parent) ***REMOVED***
            preprocess(this, node);
        ***REMOVED***

        /*
         * Updates the code path.
         * And emits onCodePathStart/onCodePathSegmentStart events.
         */
        processCodePathToEnter(this, node);

        // Emits node events.
        this.original.enterNode(node);

        this.currentNode = null;
    ***REMOVED***

    /**
     * Does the process to leave a given AST node.
     * This updates state of analysis and calls `leaveNode` of the wrapped.
     *
     * @param ***REMOVED***ASTNode***REMOVED*** node - A node which is leaving.
     * @returns ***REMOVED***void***REMOVED***
     */
    leaveNode(node) ***REMOVED***
        this.currentNode = node;

        /*
         * Updates the code path.
         * And emits onCodePathStart/onCodePathSegmentStart events.
         */
        processCodePathToExit(this, node);

        // Emits node events.
        this.original.leaveNode(node);

        // Emits the last onCodePathStart/onCodePathSegmentStart events.
        postprocess(this, node);

        this.currentNode = null;
    ***REMOVED***

    /**
     * This is called on a code path looped.
     * Then this raises a looped event.
     *
     * @param ***REMOVED***CodePathSegment***REMOVED*** fromSegment - A segment of prev.
     * @param ***REMOVED***CodePathSegment***REMOVED*** toSegment - A segment of next.
     * @returns ***REMOVED***void***REMOVED***
     */
    onLooped(fromSegment, toSegment) ***REMOVED***
        if (fromSegment.reachable && toSegment.reachable) ***REMOVED***
            debug.dump(`onCodePathSegmentLoop $***REMOVED***fromSegment.id***REMOVED*** -> $***REMOVED***toSegment.id***REMOVED***`);
            this.emitter.emit(
                "onCodePathSegmentLoop",
                fromSegment,
                toSegment,
                this.currentNode
            );
        ***REMOVED***
    ***REMOVED***
***REMOVED***

module.exports = CodePathAnalyzer;
