// optionator/lib/util.js
// Generated by LiveScript 1.6.0
(function()***REMOVED***
  var prelude, map, sortBy, fl, closestString, nameToRaw, dasherize, naturalJoin;
  prelude = require('prelude-ls'), map = prelude.map, sortBy = prelude.sortBy;
  fl = require('fast-levenshtein');
  closestString = function(possibilities, input)***REMOVED***
    var distances, ref$, string, distance;
    if (!possibilities.length) ***REMOVED***
      return;
    ***REMOVED***
    distances = map(function(it)***REMOVED***
      var ref$, longer, shorter;
      ref$ = input.length > it.length
        ? [input, it]
        : [it, input], longer = ref$[0], shorter = ref$[1];
      return ***REMOVED***
        string: it,
        distance: fl.get(longer, shorter)
      ***REMOVED***;
    ***REMOVED***)(
    possibilities);
    ref$ = sortBy(function(it)***REMOVED***
      return it.distance;
    ***REMOVED***, distances)[0], string = ref$.string, distance = ref$.distance;
    return string;
  ***REMOVED***;
  nameToRaw = function(name)***REMOVED***
    if (name.length === 1 || name === 'NUM') ***REMOVED***
      return "-" + name;
    ***REMOVED*** else ***REMOVED***
      return "--" + name;
    ***REMOVED***
  ***REMOVED***;
  dasherize = function(string)***REMOVED***
    if (/^[A-Z]/.test(string)) ***REMOVED***
      return string;
    ***REMOVED*** else ***REMOVED***
      return prelude.dasherize(string);
    ***REMOVED***
  ***REMOVED***;
  naturalJoin = function(array)***REMOVED***
    if (array.length < 3) ***REMOVED***
      return array.join(' or ');
    ***REMOVED*** else ***REMOVED***
      return array.slice(0, -1).join(', ') + ", or " + array[array.length - 1];
    ***REMOVED***
  ***REMOVED***;
  module.exports = ***REMOVED***
    closestString: closestString,
    nameToRaw: nameToRaw,
    dasherize: dasherize,
    naturalJoin: naturalJoin
  ***REMOVED***;
***REMOVED***).call(this);
