// hosted-git-info/git-host.js
'use strict'
var gitHosts = require('./git-host-info.js')
/* eslint-disable node/no-deprecated-api */

// copy-pasta util._extend from node's source, to avoid pulling
// the whole util module into peoples' webpack bundles.
/* istanbul ignore next */
var extend = Object.assign || function _extend (target, source) ***REMOVED***
  // Don't do anything if source isn't an object
  if (source === null || typeof source !== 'object') return target

  var keys = Object.keys(source)
  var i = keys.length
  while (i--) ***REMOVED***
    target[keys[i]] = source[keys[i]]
  ***REMOVED***
  return target
***REMOVED***

module.exports = GitHost
function GitHost (type, user, auth, project, committish, defaultRepresentation, opts) ***REMOVED***
  var gitHostInfo = this
  gitHostInfo.type = type
  Object.keys(gitHosts[type]).forEach(function (key) ***REMOVED***
    gitHostInfo[key] = gitHosts[type][key]
  ***REMOVED***)
  gitHostInfo.user = user
  gitHostInfo.auth = auth
  gitHostInfo.project = project
  gitHostInfo.committish = committish
  gitHostInfo.default = defaultRepresentation
  gitHostInfo.opts = opts || ***REMOVED******REMOVED***
***REMOVED***

GitHost.prototype.hash = function () ***REMOVED***
  return this.committish ? '#' + this.committish : ''
***REMOVED***

GitHost.prototype._fill = function (template, opts) ***REMOVED***
  if (!template) return
  var vars = extend(***REMOVED******REMOVED***, opts)
  vars.path = vars.path ? vars.path.replace(/^[/]+/g, '') : ''
  opts = extend(extend(***REMOVED******REMOVED***, this.opts), opts)
  var self = this
  Object.keys(this).forEach(function (key) ***REMOVED***
    if (self[key] != null && vars[key] == null) vars[key] = self[key]
  ***REMOVED***)
  var rawAuth = vars.auth
  var rawcommittish = vars.committish
  var rawFragment = vars.fragment
  var rawPath = vars.path
  var rawProject = vars.project
  Object.keys(vars).forEach(function (key) ***REMOVED***
    var value = vars[key]
    if ((key === 'path' || key === 'project') && typeof value === 'string') ***REMOVED***
      vars[key] = value.split('/').map(function (pathComponent) ***REMOVED***
        return encodeURIComponent(pathComponent)
      ***REMOVED***).join('/')
    ***REMOVED*** else ***REMOVED***
      vars[key] = encodeURIComponent(value)
    ***REMOVED***
  ***REMOVED***)
  vars['auth@'] = rawAuth ? rawAuth + '@' : ''
  vars['#fragment'] = rawFragment ? '#' + this.hashformat(rawFragment) : ''
  vars.fragment = vars.fragment ? vars.fragment : ''
  vars['#path'] = rawPath ? '#' + this.hashformat(rawPath) : ''
  vars['/path'] = vars.path ? '/' + vars.path : ''
  vars.projectPath = rawProject.split('/').map(encodeURIComponent).join('/')
  if (opts.noCommittish) ***REMOVED***
    vars['#committish'] = ''
    vars['/tree/committish'] = ''
    vars['/committish'] = ''
    vars.committish = ''
  ***REMOVED*** else ***REMOVED***
    vars['#committish'] = rawcommittish ? '#' + rawcommittish : ''
    vars['/tree/committish'] = vars.committish
      ? '/' + vars.treepath + '/' + vars.committish
      : ''
    vars['/committish'] = vars.committish ? '/' + vars.committish : ''
    vars.committish = vars.committish || 'master'
  ***REMOVED***
  var res = template
  Object.keys(vars).forEach(function (key) ***REMOVED***
    res = res.replace(new RegExp('[***REMOVED***]' + key + '[***REMOVED***]', 'g'), vars[key])
  ***REMOVED***)
  if (opts.noGitPlus) ***REMOVED***
    return res.replace(/^git[+]/, '')
  ***REMOVED*** else ***REMOVED***
    return res
  ***REMOVED***
***REMOVED***

GitHost.prototype.ssh = function (opts) ***REMOVED***
  return this._fill(this.sshtemplate, opts)
***REMOVED***

GitHost.prototype.sshurl = function (opts) ***REMOVED***
  return this._fill(this.sshurltemplate, opts)
***REMOVED***

GitHost.prototype.browse = function (P, F, opts) ***REMOVED***
  if (typeof P === 'string') ***REMOVED***
    if (typeof F !== 'string') ***REMOVED***
      opts = F
      F = null
    ***REMOVED***
    return this._fill(this.browsefiletemplate, extend(***REMOVED***
      fragment: F,
      path: P
    ***REMOVED***, opts))
  ***REMOVED*** else ***REMOVED***
    return this._fill(this.browsetemplate, P)
  ***REMOVED***
***REMOVED***

GitHost.prototype.docs = function (opts) ***REMOVED***
  return this._fill(this.docstemplate, opts)
***REMOVED***

GitHost.prototype.bugs = function (opts) ***REMOVED***
  return this._fill(this.bugstemplate, opts)
***REMOVED***

GitHost.prototype.https = function (opts) ***REMOVED***
  return this._fill(this.httpstemplate, opts)
***REMOVED***

GitHost.prototype.git = function (opts) ***REMOVED***
  return this._fill(this.gittemplate, opts)
***REMOVED***

GitHost.prototype.shortcut = function (opts) ***REMOVED***
  return this._fill(this.shortcuttemplate, opts)
***REMOVED***

GitHost.prototype.path = function (opts) ***REMOVED***
  return this._fill(this.pathtemplate, opts)
***REMOVED***

GitHost.prototype.tarball = function (opts_) ***REMOVED***
  var opts = extend(***REMOVED******REMOVED***, opts_, ***REMOVED*** noCommittish: false ***REMOVED***)
  return this._fill(this.tarballtemplate, opts)
***REMOVED***

GitHost.prototype.file = function (P, opts) ***REMOVED***
  return this._fill(this.filetemplate, extend(***REMOVED*** path: P ***REMOVED***, opts))
***REMOVED***

GitHost.prototype.getDefaultRepresentation = function () ***REMOVED***
  return this.default
***REMOVED***

GitHost.prototype.toString = function (opts) ***REMOVED***
  if (this.default && typeof this[this.default] === 'function') return this[this.default](opts)
  return this.sshurl(opts)
***REMOVED***
