// tslint-eslint-rules/dist/rules/braceStyleRule.js
"use strict";
Object.defineProperty(exports, "__esModule", ***REMOVED*** value: true ***REMOVED***);
var tslib_1 = require("tslib");
var ts = require("typescript");
var Lint = require("tslint");
var OPTION_1TBS = '1tbs';
var OPTION_ALLMAN = 'allman';
var OPTION_STROUSTRUP = 'stroustrup';
var BraceStyle;
(function (BraceStyle) ***REMOVED***
    BraceStyle[BraceStyle["OneTBS"] = 0] = "OneTBS";
    BraceStyle[BraceStyle["Allman"] = 1] = "Allman";
    BraceStyle[BraceStyle["Stroustrup"] = 2] = "Stroustrup";
***REMOVED***)(BraceStyle || (BraceStyle = ***REMOVED******REMOVED***));
var Rule = (function (_super) ***REMOVED***
    tslib_1.__extends(Rule, _super);
    function Rule() ***REMOVED***
        return _super !== null && _super.apply(this, arguments) || this;
    ***REMOVED***
    Rule.prototype.apply = function (sourceFile) ***REMOVED***
        var walker = new BraceStyleWalker(sourceFile, this.getOptions());
        return this.applyWithWalker(walker);
    ***REMOVED***;
    Rule.FAILURE_STRING = ***REMOVED***
        open: 'Opening curly brace does not appear on the same line as controlling statement.',
        openAllman: 'Opening curly brace appears on the same line as controlling statement.',
        body: 'Statement inside of curly braces should be on next line.',
        close: 'Closing curly brace does not appear on the same line as the subsequent block.',
        closeSingle: 'Closing curly brace should be on the same line as opening curly brace or on the line after the previous block.',
        closeStroustrupAllman: 'Closing curly brace appears on the same line as the subsequent block.'
    ***REMOVED***;
    return Rule;
***REMOVED***(Lint.Rules.AbstractRule));
exports.Rule = Rule;
var BraceStyleWalker = (function (_super) ***REMOVED***
    tslib_1.__extends(BraceStyleWalker, _super);
    function BraceStyleWalker(sourceFile, options) ***REMOVED***
        var _this = _super.call(this, sourceFile, options) || this;
        _this.allowSingleLine = false;
        if (_this.hasOption(OPTION_1TBS)) ***REMOVED***
            _this.braceStyle = BraceStyle.OneTBS;
        ***REMOVED***
        else if (_this.hasOption(OPTION_ALLMAN)) ***REMOVED***
            _this.braceStyle = BraceStyle.Allman;
        ***REMOVED***
        else if (_this.hasOption(OPTION_STROUSTRUP)) ***REMOVED***
            _this.braceStyle = BraceStyle.Stroustrup;
        ***REMOVED***
        else ***REMOVED***
        ***REMOVED***
        _this.allowSingleLine = _this.getOptions()[1] && _this.getOptions()[1].allowSingleLine;
        return _this;
    ***REMOVED***
    BraceStyleWalker.prototype.visitTryStatement = function (tryStatement) ***REMOVED***
        var _this = this;
        _super.prototype.visitTryStatement.call(this, tryStatement);
        var checkTryStatementError = function (node) ***REMOVED***
            var previousNode = _this.getPreviousNode(tryStatement.getChildren(), node);
            var openingBracketError = _this.areOnSameLine(previousNode, node) !== (_this.braceStyle === BraceStyle.OneTBS);
            if (_this.allowSingleLine && _this.getStartPosition(node).line === _this.getEndPosition(tryStatement).line) ***REMOVED***
                return;
            ***REMOVED***
            if (openingBracketError) ***REMOVED***
                var failureString = _this.braceStyle === BraceStyle.OneTBS ? Rule.FAILURE_STRING.open : Rule.FAILURE_STRING.openAllman;
                _this.addFailure(_this.createFailure(node.getStart(), node.getWidth(), failureString));
            ***REMOVED***
        ***REMOVED***;
        var catchClause = tryStatement.catchClause;
        if (catchClause) ***REMOVED***
            checkTryStatementError(catchClause);
        ***REMOVED***
        var finallyBlock = tryStatement.finallyBlock;
        if (finallyBlock) ***REMOVED***
            checkTryStatementError(finallyBlock);
        ***REMOVED***
    ***REMOVED***;
    BraceStyleWalker.prototype.visitIfStatement = function (ifStatement) ***REMOVED***
        _super.prototype.visitIfStatement.call(this, ifStatement);
        var elseKeyword = ifStatement.getChildren().filter(function (ch) ***REMOVED*** return ch.kind === ts.SyntaxKind.ElseKeyword; ***REMOVED***).shift();
        if (!elseKeyword) ***REMOVED***
            return;
        ***REMOVED***
        var previousNode = ifStatement.getChildren()[ifStatement.getChildren().indexOf(elseKeyword) - 1];
        var openingBracketError = this.areOnSameLine(previousNode, elseKeyword) !== (this.braceStyle === BraceStyle.OneTBS);
        if (this.allowSingleLine && this.getStartPosition(elseKeyword).line === this.getEndPosition(ifStatement).line) ***REMOVED***
            return;
        ***REMOVED***
        if (!ifStatement.getChildren().some(function (ch) ***REMOVED*** return ch.kind === ts.SyntaxKind.Block; ***REMOVED***)) ***REMOVED***
            return;
        ***REMOVED***
        if (openingBracketError) ***REMOVED***
            var failureString = this.braceStyle === BraceStyle.OneTBS ? Rule.FAILURE_STRING.open : Rule.FAILURE_STRING.openAllman;
            this.addFailure(this.createFailure(elseKeyword.getStart(), elseKeyword.getWidth(), failureString));
        ***REMOVED***
    ***REMOVED***;
    BraceStyleWalker.prototype.visitBlock = function (block) ***REMOVED***
        _super.prototype.visitBlock.call(this, block);
        if (this.allowSingleLine && this.getStartPosition(block).line === this.getEndPosition(block).line) ***REMOVED***
            return;
        ***REMOVED***
        var blockChildren = block.getChildren();
        var openingCurlyBrace = blockChildren.filter(function (ch) ***REMOVED*** return ch.kind === ts.SyntaxKind.OpenBraceToken; ***REMOVED***).shift();
        var closingCurlyBrace = blockChildren.filter(function (ch) ***REMOVED*** return ch.kind === ts.SyntaxKind.CloseBraceToken; ***REMOVED***).pop();
        var syntaxList = blockChildren.filter(function (ch) ***REMOVED*** return ch.kind === ts.SyntaxKind.SyntaxList; ***REMOVED***).shift();
        var parentChildren = block.parent ? block.parent.getChildren() : [];
        var blockPreviousNode = parentChildren[parentChildren.indexOf(block) - 1];
        if (!openingCurlyBrace || !closingCurlyBrace || !syntaxList || !blockPreviousNode) ***REMOVED***
            return;
        ***REMOVED***
        var openingBracketError = this.areOnSameLine(blockPreviousNode, block) === (this.braceStyle === BraceStyle.Allman);
        if (openingBracketError) ***REMOVED***
            var failureString = this.braceStyle === BraceStyle.Allman ? Rule.FAILURE_STRING.openAllman : Rule.FAILURE_STRING.open;
            this.addFailure(this.createFailure(openingCurlyBrace.getStart(), openingCurlyBrace.getWidth(), failureString));
        ***REMOVED***
        if (syntaxList.getChildCount() > 0) ***REMOVED***
            var bodyError = this.areOnSameLine(openingCurlyBrace, syntaxList);
            if (bodyError) ***REMOVED***
                this.addFailure(this.createFailure(syntaxList.getStart(), syntaxList.getWidth(), Rule.FAILURE_STRING.body));
            ***REMOVED***
            var nodeBeforeClosingBracket = syntaxList.getChildren()[syntaxList.getChildren().length - 1];
            var closingBracketError = this.areOnSameLine(nodeBeforeClosingBracket, closingCurlyBrace);
            if (closingBracketError) ***REMOVED***
                this.addFailure(this.createFailure(closingCurlyBrace.getStart(), closingCurlyBrace.getWidth(), Rule.FAILURE_STRING.closeSingle));
            ***REMOVED***
        ***REMOVED***
    ***REMOVED***;
    BraceStyleWalker.prototype.areOnSameLine = function (node, nextNode) ***REMOVED***
        return this.getEndPosition(node).line === this.getStartPosition(nextNode).line;
    ***REMOVED***;
    BraceStyleWalker.prototype.getStartPosition = function (node) ***REMOVED***
        return node.getSourceFile().getLineAndCharacterOfPosition(node.getStart());
    ***REMOVED***;
    BraceStyleWalker.prototype.getEndPosition = function (node) ***REMOVED***
        return node.getSourceFile().getLineAndCharacterOfPosition(node.getEnd());
    ***REMOVED***;
    BraceStyleWalker.prototype.getPreviousNode = function (children, node) ***REMOVED***
        var position = children.indexOf(node) - 1;
        while (position >= 0) ***REMOVED***
            if (children[position].kind === ts.SyntaxKind.Block || children[position].kind === ts.SyntaxKind.CatchClause) ***REMOVED***
                break;
            ***REMOVED***
            position -= 1;
        ***REMOVED***
        return children[position];
    ***REMOVED***;
    return BraceStyleWalker;
***REMOVED***(Lint.RuleWalker));
