// tslint-eslint-rules/dist/rules/noControlRegexRule.js
"use strict";
Object.defineProperty(exports, "__esModule", ***REMOVED*** value: true ***REMOVED***);
var tslib_1 = require("tslib");
var ts = require("typescript");
var Lint = require("tslint");
var Rule = (function (_super) ***REMOVED***
    tslib_1.__extends(Rule, _super);
    function Rule() ***REMOVED***
        return _super !== null && _super.apply(this, arguments) || this;
    ***REMOVED***
    Rule.prototype.apply = function (sourceFile) ***REMOVED***
        var walker = new NoControlRegexWalker(sourceFile, this.getOptions());
        return this.applyWithWalker(walker);
    ***REMOVED***;
    Rule.FAILURE_STRING = 'unexpected control character in regular expression';
    return Rule;
***REMOVED***(Lint.Rules.AbstractRule));
exports.Rule = Rule;
var NoControlRegexWalker = (function (_super) ***REMOVED***
    tslib_1.__extends(NoControlRegexWalker, _super);
    function NoControlRegexWalker() ***REMOVED***
        return _super !== null && _super.apply(this, arguments) || this;
    ***REMOVED***
    NoControlRegexWalker.prototype.visitRegularExpressionLiteral = function (node) ***REMOVED***
        this.validateControlRegex(node);
        _super.prototype.visitRegularExpressionLiteral.call(this, node);
    ***REMOVED***;
    NoControlRegexWalker.prototype.visitNewExpression = function (node) ***REMOVED***
        if (node.expression.getText() === 'RegExp') ***REMOVED***
            this.visitRegularExpressionFunction(node);
        ***REMOVED***
        _super.prototype.visitNewExpression.call(this, node);
    ***REMOVED***;
    NoControlRegexWalker.prototype.visitCallExpression = function (node) ***REMOVED***
        if (node.expression.getText() === 'RegExp') ***REMOVED***
            this.visitRegularExpressionFunction(node);
        ***REMOVED***
        _super.prototype.visitCallExpression.call(this, node);
    ***REMOVED***;
    NoControlRegexWalker.prototype.visitRegularExpressionFunction = function (node) ***REMOVED***
        if (node.arguments && node.arguments.length > 0 && node.arguments[0].kind === ts.SyntaxKind.StringLiteral) ***REMOVED***
            this.validateControlRegex(node.arguments[0]);
        ***REMOVED***
    ***REMOVED***;
    NoControlRegexWalker.prototype.validateControlRegex = function (node) ***REMOVED***
        if (/[\x00-\x1f]/.test(node.text)) ***REMOVED***
            this.addFailure(this.createFailure(node.getStart(), node.getWidth(), Rule.FAILURE_STRING));
        ***REMOVED***
    ***REMOVED***;
    return NoControlRegexWalker;
***REMOVED***(Lint.RuleWalker));
