// tslint-eslint-rules/dist/rules/noExAssignRule.js
"use strict";
Object.defineProperty(exports, "__esModule", ***REMOVED*** value: true ***REMOVED***);
var tslib_1 = require("tslib");
var ts = require("typescript");
var Lint = require("tslint");
var token_1 = require("../support/token");
var Rule = (function (_super) ***REMOVED***
    tslib_1.__extends(Rule, _super);
    function Rule() ***REMOVED***
        return _super !== null && _super.apply(this, arguments) || this;
    ***REMOVED***
    Rule.prototype.apply = function (sourceFile) ***REMOVED***
        var walker = new NoExAssignWalker(sourceFile, this.getOptions());
        return this.applyWithWalker(walker);
    ***REMOVED***;
    Rule.FAILURE_STRING = 'do not assign to the exception parameter';
    return Rule;
***REMOVED***(Lint.Rules.AbstractRule));
exports.Rule = Rule;
var NoExAssignWalker = (function (_super) ***REMOVED***
    tslib_1.__extends(NoExAssignWalker, _super);
    function NoExAssignWalker() ***REMOVED***
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.isInCatchClause = false;
        return _this;
    ***REMOVED***
    NoExAssignWalker.prototype.visitCatchClause = function (node) ***REMOVED***
        this.variableNode = node.variableDeclaration;
        this.isInCatchClause = true;
        _super.prototype.visitCatchClause.call(this, node);
        this.isInCatchClause = false;
        delete this.variableNode;
    ***REMOVED***;
    NoExAssignWalker.prototype.visitBinaryExpression = function (node) ***REMOVED***
        var _this = this;
        if (this.isInCatchClause) ***REMOVED***
            if (!token_1.isAssignmentToken(node.operatorToken)) ***REMOVED***
                return;
            ***REMOVED***
            if (this.variableNode &&
                this.variableNode.name.getText() === node.left.getText() &&
                node.left.kind === ts.SyntaxKind.Identifier) ***REMOVED***
                this.addFailure(this.createFailure(node.getStart(), node.getWidth(), Rule.FAILURE_STRING));
            ***REMOVED***
            else if (node.left.kind === ts.SyntaxKind.ArrayLiteralExpression) ***REMOVED***
                var els = node.left.elements;
                if (els.some(function (el) ***REMOVED*** return !!_this.variableNode && el.getText() === _this.variableNode.getText(); ***REMOVED***)) ***REMOVED***
                    this.addFailure(this.createFailure(node.getStart(), node.getWidth(), Rule.FAILURE_STRING));
                ***REMOVED***
            ***REMOVED***
        ***REMOVED***
        _super.prototype.visitBinaryExpression.call(this, node);
    ***REMOVED***;
    return NoExAssignWalker;
***REMOVED***(Lint.RuleWalker));
