// tslint-eslint-rules/dist/rules/noExtraBooleanCastRule.js
"use strict";
Object.defineProperty(exports, "__esModule", ***REMOVED*** value: true ***REMOVED***);
var tslib_1 = require("tslib");
var ts = require("typescript");
var Lint = require("tslint");
var Rule = (function (_super) ***REMOVED***
    tslib_1.__extends(Rule, _super);
    function Rule() ***REMOVED***
        return _super !== null && _super.apply(this, arguments) || this;
    ***REMOVED***
    Rule.prototype.apply = function (sourceFile) ***REMOVED***
        var walker = new NoExtraBooleanCastWalker(sourceFile, this.getOptions());
        return this.applyWithWalker(walker);
    ***REMOVED***;
    Rule.FAILURE_STRING = ***REMOVED***
        if: 'redundant double negation in an if statement condition',
        do: 'redundant double negation in a do while loop condition',
        while: 'redundant double negation in a while loop condition',
        ternaryif: 'redundant double negation in a ternary condition',
        for: 'redundant double negation in a for loop condition',
        unaryCast: 'redundant multiple negation',
        objectCast: 'redundant double negation in call to Boolean()',
        newCast: 'redundant double negation in Boolean constructor call'
    ***REMOVED***;
    return Rule;
***REMOVED***(Lint.Rules.AbstractRule));
exports.Rule = Rule;
var NoExtraBooleanCastWalker = (function (_super) ***REMOVED***
    tslib_1.__extends(NoExtraBooleanCastWalker, _super);
    function NoExtraBooleanCastWalker() ***REMOVED***
        return _super !== null && _super.apply(this, arguments) || this;
    ***REMOVED***
    NoExtraBooleanCastWalker.prototype.visitPrefixUnaryExpression = function (node) ***REMOVED***
        this.validateNoExtraBoolean(node);
        _super.prototype.visitPrefixUnaryExpression.call(this, node);
    ***REMOVED***;
    NoExtraBooleanCastWalker.prototype.validateNoExtraBoolean = function (node) ***REMOVED***
        if (!node.parent || !node.parent.parent) ***REMOVED***
            return;
        ***REMOVED***
        var parent = node.parent;
        var grandparent = parent.parent;
        if (node.operator !== ts.SyntaxKind.ExclamationToken ||
            parent.kind !== ts.SyntaxKind.PrefixUnaryExpression ||
            parent.operator !== ts.SyntaxKind.ExclamationToken ||
            !grandparent) ***REMOVED***
            return;
        ***REMOVED***
        if (grandparent.kind === ts.SyntaxKind.BinaryExpression) ***REMOVED***
            grandparent = grandparent.parent;
        ***REMOVED***
        if (!grandparent) ***REMOVED***
            return;
        ***REMOVED***
        if (grandparent.kind === ts.SyntaxKind.IfStatement) ***REMOVED***
            this.addFailure(this.createFailure(node.getStart(), node.getWidth(), Rule.FAILURE_STRING.if));
        ***REMOVED***
        else if (grandparent.kind === ts.SyntaxKind.DoStatement) ***REMOVED***
            this.addFailure(this.createFailure(node.getStart(), node.getWidth(), Rule.FAILURE_STRING.do));
        ***REMOVED***
        else if (grandparent.kind === ts.SyntaxKind.WhileStatement) ***REMOVED***
            this.addFailure(this.createFailure(node.getStart(), node.getWidth(), Rule.FAILURE_STRING.while));
        ***REMOVED***
        else if (grandparent.kind === ts.SyntaxKind.ConditionalExpression && parent === grandparent.condition) ***REMOVED***
            this.addFailure(this.createFailure(node.getStart(), node.getWidth(), Rule.FAILURE_STRING.ternaryif));
        ***REMOVED***
        else if (grandparent.kind === ts.SyntaxKind.ForStatement && parent === grandparent.condition) ***REMOVED***
            this.addFailure(this.createFailure(node.getStart(), node.getWidth(), Rule.FAILURE_STRING.for));
        ***REMOVED***
        else if (grandparent.kind === ts.SyntaxKind.PrefixUnaryExpression && grandparent.operator === ts.SyntaxKind.ExclamationToken) ***REMOVED***
            this.addFailure(this.createFailure(node.getStart(), node.getWidth(), Rule.FAILURE_STRING.unaryCast));
        ***REMOVED***
        else if (grandparent.kind === ts.SyntaxKind.CallExpression && /^Boolean/.test(grandparent.getText())) ***REMOVED***
            this.addFailure(this.createFailure(node.getStart(), node.getWidth(), Rule.FAILURE_STRING.objectCast));
        ***REMOVED***
        else if (grandparent.kind === ts.SyntaxKind.NewExpression && /^new Boolean/.test(grandparent.getText())) ***REMOVED***
            this.addFailure(this.createFailure(node.getStart(), node.getWidth(), Rule.FAILURE_STRING.newCast));
        ***REMOVED***
    ***REMOVED***;
    return NoExtraBooleanCastWalker;
***REMOVED***(Lint.RuleWalker));
