// tslint-eslint-rules/dist/rules/noInvalidRegexpRule.js
"use strict";
Object.defineProperty(exports, "__esModule", ***REMOVED*** value: true ***REMOVED***);
var tslib_1 = require("tslib");
var ts = require("typescript");
var Lint = require("tslint");
var Rule = (function (_super) ***REMOVED***
    tslib_1.__extends(Rule, _super);
    function Rule() ***REMOVED***
        return _super !== null && _super.apply(this, arguments) || this;
    ***REMOVED***
    Rule.prototype.apply = function (sourceFile) ***REMOVED***
        var walker = new NoInvalidRegexpWalker(sourceFile, this.getOptions());
        return this.applyWithWalker(walker);
    ***REMOVED***;
    return Rule;
***REMOVED***(Lint.Rules.AbstractRule));
exports.Rule = Rule;
var NoInvalidRegexpWalker = (function (_super) ***REMOVED***
    tslib_1.__extends(NoInvalidRegexpWalker, _super);
    function NoInvalidRegexpWalker() ***REMOVED***
        return _super !== null && _super.apply(this, arguments) || this;
    ***REMOVED***
    NoInvalidRegexpWalker.prototype.visitNewExpression = function (node) ***REMOVED***
        this.validateInvalidRegExp(node);
        _super.prototype.visitNewExpression.call(this, node);
    ***REMOVED***;
    NoInvalidRegexpWalker.prototype.visitCallExpression = function (node) ***REMOVED***
        this.validateInvalidRegExp(node);
        _super.prototype.visitCallExpression.call(this, node);
    ***REMOVED***;
    NoInvalidRegexpWalker.prototype.validateInvalidRegExp = function (node) ***REMOVED***
        if (node.expression.getText() === 'RegExp') ***REMOVED***
            var args = node.arguments;
            if (args && args.length > 0 && args[0].kind === ts.SyntaxKind.StringLiteral) ***REMOVED***
                var expr = args[0].text;
                var flags = args.length > 1 && args[1].kind === ts.SyntaxKind.StringLiteral ? args[1].text : undefined;
                try ***REMOVED***
                    new RegExp(expr, flags);
                ***REMOVED***
                catch (e) ***REMOVED***
                    this.addFailure(this.createFailure(node.getStart(), node.getWidth(), e.message));
                ***REMOVED***
            ***REMOVED***
        ***REMOVED***
    ***REMOVED***;
    return NoInvalidRegexpWalker;
***REMOVED***(Lint.RuleWalker));
