// tslint-eslint-rules/dist/rules/noInnerDeclarationsRule.js
"use strict";
Object.defineProperty(exports, "__esModule", ***REMOVED*** value: true ***REMOVED***);
var tslib_1 = require("tslib");
var ts = require("typescript");
var Lint = require("tslint");
var Rule = (function (_super) ***REMOVED***
    tslib_1.__extends(Rule, _super);
    function Rule() ***REMOVED***
        return _super !== null && _super.apply(this, arguments) || this;
    ***REMOVED***
    Rule.prototype.apply = function (sourceFile) ***REMOVED***
        var walker = new NoInnerDeclarationsWalker(sourceFile, this.getOptions());
        return this.applyWithWalker(walker);
    ***REMOVED***;
    return Rule;
***REMOVED***(Lint.Rules.AbstractRule));
exports.Rule = Rule;
var NoInnerDeclarationsWalker = (function (_super) ***REMOVED***
    tslib_1.__extends(NoInnerDeclarationsWalker, _super);
    function NoInnerDeclarationsWalker() ***REMOVED***
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.VALID_PARENT_TYPES = [
            ts.SyntaxKind.SourceFile,
            ts.SyntaxKind.FunctionDeclaration,
            ts.SyntaxKind.FunctionExpression,
            ts.SyntaxKind.ArrowFunction,
            ts.SyntaxKind.MethodDeclaration,
            ts.SyntaxKind.ModuleDeclaration,
            ts.SyntaxKind.Constructor
        ];
        return _this;
    ***REMOVED***
    NoInnerDeclarationsWalker.prototype.visitFunctionDeclaration = function (node) ***REMOVED***
        this.validateInnerDeclaration(node);
        _super.prototype.visitFunctionDeclaration.call(this, node);
    ***REMOVED***;
    NoInnerDeclarationsWalker.prototype.visitVariableStatement = function (node) ***REMOVED***
        if (this.hasOption('both') && node.declarationList.getFirstToken().kind === ts.SyntaxKind.VarKeyword) ***REMOVED***
            this.validateInnerDeclaration(node);
        ***REMOVED***
        _super.prototype.visitVariableStatement.call(this, node);
    ***REMOVED***;
    NoInnerDeclarationsWalker.prototype.validateInnerDeclaration = function (node) ***REMOVED***
        var body = this.nearestBody(node);
        var isValid = (body.isSourceFile && body.distance === 1) || body.distance === 2;
        if (!isValid) ***REMOVED***
            var decl = node.kind === ts.SyntaxKind.FunctionDeclaration ? 'function' : 'variable';
            var root = body.isSourceFile ? 'program' : 'function body';
            this.addFailure(this.createFailure(node.getStart(), node.getWidth(), "move " + decl + " declaration to " + root + " root"));
        ***REMOVED***
    ***REMOVED***;
    NoInnerDeclarationsWalker.prototype.nearestBody = function (node) ***REMOVED***
        var ancestor = node.parent;
        var generation = 1;
        while (ancestor && this.VALID_PARENT_TYPES.indexOf(ancestor.kind) === -1) ***REMOVED***
            generation++;
            ancestor = ancestor.parent;
        ***REMOVED***
        return ***REMOVED***
            isSourceFile: (ancestor && ancestor.kind === ts.SyntaxKind.SourceFile) || !ancestor,
            distance: generation
        ***REMOVED***;
    ***REMOVED***;
    return NoInnerDeclarationsWalker;
***REMOVED***(Lint.RuleWalker));
