// tslint-eslint-rules/dist/rules/noMultiSpacesRule.js
"use strict";
Object.defineProperty(exports, "__esModule", ***REMOVED*** value: true ***REMOVED***);
var tslib_1 = require("tslib");
var ts = require("typescript");
var Lint = require("tslint");
var Rule = (function (_super) ***REMOVED***
    tslib_1.__extends(Rule, _super);
    function Rule() ***REMOVED***
        return _super !== null && _super.apply(this, arguments) || this;
    ***REMOVED***
    Rule.prototype.apply = function (sourceFile) ***REMOVED***
        var walker = new NoMultiSpacesWalker(sourceFile, this.getOptions());
        return this.applyWithWalker(walker);
    ***REMOVED***;
    return Rule;
***REMOVED***(Lint.Rules.AbstractRule));
exports.Rule = Rule;
var NoMultiSpacesWalker = (function (_super) ***REMOVED***
    tslib_1.__extends(NoMultiSpacesWalker, _super);
    function NoMultiSpacesWalker(sourceFile, options) ***REMOVED***
        var _a;
        var _this = _super.call(this, sourceFile, options) || this;
        _this.EXCEPTION_MAP = (_a = ***REMOVED******REMOVED***,
            _a[ts.SyntaxKind.VariableDeclaration] = 'VariableDeclaration',
            _a[ts.SyntaxKind.PropertyAssignment] = 'PropertyAssignment',
            _a[ts.SyntaxKind.BinaryExpression] = 'BinaryExpression',
            _a);
        _this.STRING_TYPES = [
            ts.SyntaxKind.NoSubstitutionTemplateLiteral,
            ts.SyntaxKind.LastTemplateToken,
            ts.SyntaxKind.StringLiteral
        ];
        _this.exceptions = ***REMOVED******REMOVED***;
        _this.targets = [];
        _this.targetNode = ***REMOVED******REMOVED***;
        _this.targetIndex = 0;
        var opt = _this.getOptions();
        _this.src = sourceFile.getFullText();
        if (opt.length) ***REMOVED***
            _this.exceptions = opt[0].exceptions || ***REMOVED******REMOVED***;
        ***REMOVED***
        if (_this.exceptions['PropertyAssignment'] === undefined) ***REMOVED***
            _this.exceptions['PropertyAssignment'] = true;
        ***REMOVED***
        var pattern = /[^\n\r\u2028\u2029\t ].? ***REMOVED***2,***REMOVED***/g;
        while (pattern.test(_this.src)) ***REMOVED***
            _this.targets.push(pattern.lastIndex);
            _this.targetNode[pattern.lastIndex] = sourceFile;
        ***REMOVED***
        _this.lastNode = sourceFile.getLastToken();
        return _this;
    ***REMOVED***
    NoMultiSpacesWalker.prototype.inRange = function (x, range) ***REMOVED***
        return x >= range[0] && x <= range[1];
    ***REMOVED***;
    NoMultiSpacesWalker.prototype.warn = function (value, pos, node) ***REMOVED***
        var msg = "Multiple spaces found before '" + value + "'.";
        var exceptionName = this.EXCEPTION_MAP[node.parent.kind];
        var report = true;
        var start = node.getFullStart() - 1;
        var previousChar = this.src.substring(start, start + 1);
        if (exceptionName && this.exceptions[exceptionName]) ***REMOVED***
            if (previousChar !== ',') ***REMOVED***
                report = false;
            ***REMOVED***
        ***REMOVED***
        if (previousChar === ':') ***REMOVED***
            var crt = node.parent;
            while (crt.kind !== ts.SyntaxKind.SourceFile) ***REMOVED***
                crt = crt.parent;
                if (crt.kind === ts.SyntaxKind.PropertyAssignment) ***REMOVED***
                    if (this.exceptions['PropertyAssignment']) ***REMOVED***
                        report = false;
                    ***REMOVED***
                    break;
                ***REMOVED***
            ***REMOVED***
        ***REMOVED***
        if (report) ***REMOVED***
            this.addFailure(this.createFailure(pos, value.length, msg));
        ***REMOVED***
    ***REMOVED***;
    NoMultiSpacesWalker.prototype.walkChildren = function (node) ***REMOVED***
        var _this = this;
        var range = [node.getStart(), node.getEnd()];
        for (var i = this.targetIndex, len = this.targets.length, target = void 0; i < len; i++) ***REMOVED***
            target = this.targets[i];
            if (this.inRange(target, range)) ***REMOVED***
                this.targetNode[target] = node;
            ***REMOVED***
            if (range[0] > this.targets[this.targetIndex]) ***REMOVED***
                this.targetIndex++;
            ***REMOVED***
        ***REMOVED***
        if (node === this.lastNode) ***REMOVED***
            this.targets.forEach(function (target) ***REMOVED***
                var valid = _this.targetNode[target];
                if (target === valid.getStart()) ***REMOVED***
                    _this.warn(valid.getText(), target, valid);
                ***REMOVED***
                else if (target === valid.getEnd() - 1 && _this.STRING_TYPES.indexOf(valid.kind) === -1) ***REMOVED***
                    var endChar = _this.src.substring(target, valid.getEnd());
                    _this.warn(endChar, target, valid);
                ***REMOVED***
                else ***REMOVED***
                    if (_this.src.charAt(target) !== '\n' && valid.kind !== ts.SyntaxKind.SourceFile) ***REMOVED***
                    ***REMOVED***
                ***REMOVED***
            ***REMOVED***);
        ***REMOVED***
        var children = node.getChildren();
        for (var index in children) ***REMOVED***
            this.visitNode(children[index]);
        ***REMOVED***
    ***REMOVED***;
    return NoMultiSpacesWalker;
***REMOVED***(Lint.RuleWalker));
