// tslint-eslint-rules/dist/rules/noUnexpectedMultilineRule.js
"use strict";
Object.defineProperty(exports, "__esModule", ***REMOVED*** value: true ***REMOVED***);
var tslib_1 = require("tslib");
var ts = require("typescript");
var Lint = require("tslint");
var Rule = (function (_super) ***REMOVED***
    tslib_1.__extends(Rule, _super);
    function Rule() ***REMOVED***
        return _super !== null && _super.apply(this, arguments) || this;
    ***REMOVED***
    Rule.prototype.apply = function (sourceFile) ***REMOVED***
        var walker = new NoUnexpectedMultilineWalker(sourceFile, this.getOptions());
        return this.applyWithWalker(walker);
    ***REMOVED***;
    Rule.FAILURE_STRING = ***REMOVED***
        func: 'unexpected newline between function and ( of function call',
        prop: 'unexpected newline between object and [ of property access',
        template: 'unexpected newline between template tag and template literal'
    ***REMOVED***;
    return Rule;
***REMOVED***(Lint.Rules.AbstractRule));
exports.Rule = Rule;
var NoUnexpectedMultilineWalker = (function (_super) ***REMOVED***
    tslib_1.__extends(NoUnexpectedMultilineWalker, _super);
    function NoUnexpectedMultilineWalker() ***REMOVED***
        return _super !== null && _super.apply(this, arguments) || this;
    ***REMOVED***
    NoUnexpectedMultilineWalker.prototype.visitCallExpression = function (node) ***REMOVED***
        var firstLeftParen = node.getChildren().filter(function (ch) ***REMOVED*** return ch.kind === ts.SyntaxKind.OpenParenToken; ***REMOVED***)[0];
        if (this.isBreakBefore(firstLeftParen)) ***REMOVED***
            this.addFailure(this.createFailure(node.getStart(), node.getWidth(), this.getMessage(node)));
        ***REMOVED***
        _super.prototype.visitCallExpression.call(this, node);
    ***REMOVED***;
    NoUnexpectedMultilineWalker.prototype.visitElementAccessExpression = function (node) ***REMOVED***
        var firstLeftSquareBracket = node.getChildren().filter(function (ch) ***REMOVED*** return ch.kind === ts.SyntaxKind.OpenBracketToken; ***REMOVED***)[0];
        if (this.isBreakBefore(firstLeftSquareBracket)) ***REMOVED***
            this.addFailure(this.createFailure(node.getStart(), node.getWidth(), this.getMessage(node)));
        ***REMOVED***
        _super.prototype.visitElementAccessExpression.call(this, node);
    ***REMOVED***;
    NoUnexpectedMultilineWalker.prototype.visitNode = function (node) ***REMOVED***
        if (node.kind === ts.SyntaxKind.TaggedTemplateExpression) ***REMOVED***
            var children = node.getChildren();
            var tag = children.filter(function (ch) ***REMOVED*** return ch.kind === ts.SyntaxKind.Identifier; ***REMOVED***)[0];
            var tagIndex = children.indexOf(tag);
            if (tag && children[tagIndex + 1]) ***REMOVED***
                var template = children[tagIndex + 1];
                if (this.isBreakBefore(template)) ***REMOVED***
                    this.addFailure(this.createFailure(node.getStart(), node.getWidth(), this.getMessage(node)));
                ***REMOVED***
            ***REMOVED***
        ***REMOVED***
        _super.prototype.visitNode.call(this, node);
    ***REMOVED***;
    NoUnexpectedMultilineWalker.prototype.isBreakBefore = function (node) ***REMOVED***
        if (node.parent) ***REMOVED***
            var children = node.parent.getChildren();
            var nodeIndex = children.indexOf(node);
            if (nodeIndex > 0) ***REMOVED***
                var nodeLine = this.getStartPosition(node).line;
                var previousNodeLine = this.getEndPosition(children[nodeIndex - 1]).line;
                if (nodeLine !== previousNodeLine) ***REMOVED***
                    return true;
                ***REMOVED***
            ***REMOVED***
        ***REMOVED***
        return false;
    ***REMOVED***;
    NoUnexpectedMultilineWalker.prototype.getMessage = function (node) ***REMOVED***
        switch (node.kind) ***REMOVED***
            case ts.SyntaxKind.CallExpression:
                return Rule.FAILURE_STRING.func;
            case ts.SyntaxKind.ElementAccessExpression:
                return Rule.FAILURE_STRING.prop;
            case ts.SyntaxKind.TaggedTemplateExpression:
                return Rule.FAILURE_STRING.template;
            default:
                throw 'Unexpected node type: ' + ts.SyntaxKind[node.kind];
        ***REMOVED***
    ***REMOVED***;
    NoUnexpectedMultilineWalker.prototype.getStartPosition = function (node) ***REMOVED***
        return node.getSourceFile().getLineAndCharacterOfPosition(node.getStart());
    ***REMOVED***;
    NoUnexpectedMultilineWalker.prototype.getEndPosition = function (node) ***REMOVED***
        return node.getSourceFile().getLineAndCharacterOfPosition(node.getEnd());
    ***REMOVED***;
    return NoUnexpectedMultilineWalker;
***REMOVED***(Lint.RuleWalker));
