// tslint-eslint-rules/dist/rules/objectCurlySpacingRule.js
"use strict";
Object.defineProperty(exports, "__esModule", ***REMOVED*** value: true ***REMOVED***);
var tslib_1 = require("tslib");
var ts = require("typescript");
var Lint = require("tslint");
var OPTION_ALWAYS = 'always';
var Rule = (function (_super) ***REMOVED***
    tslib_1.__extends(Rule, _super);
    function Rule() ***REMOVED***
        return _super !== null && _super.apply(this, arguments) || this;
    ***REMOVED***
    Rule.prototype.apply = function (sourceFile) ***REMOVED***
        var walker = new ObjectCurlySpacingWalker(sourceFile, this.getOptions());
        return this.applyWithWalker(walker);
    ***REMOVED***;
    Rule.FAILURE_STRING = ***REMOVED***
        always: ***REMOVED***
            start: "A space is required after '***REMOVED***'",
            end: "A space is required before '***REMOVED***'"
        ***REMOVED***,
        never: ***REMOVED***
            start: "There should be no space after '***REMOVED***'",
            end: "There should be no space before '***REMOVED***'"
        ***REMOVED***
    ***REMOVED***;
    return Rule;
***REMOVED***(Lint.Rules.AbstractRule));
exports.Rule = Rule;
var ObjectCurlySpacingWalker = (function (_super) ***REMOVED***
    tslib_1.__extends(ObjectCurlySpacingWalker, _super);
    function ObjectCurlySpacingWalker(sourceFile, options) ***REMOVED***
        var _this = _super.call(this, sourceFile, options) || this;
        _this.always = _this.hasOption(OPTION_ALWAYS) || (_this.getOptions() && _this.getOptions().length === 0);
        var opt = _this.getOptions();
        _this.exceptions = opt[1] || ***REMOVED******REMOVED***;
        if (_this.exceptions.arraysInObjects === undefined) ***REMOVED***
            _this.exceptions.arraysInObjects = _this.always;
        ***REMOVED***
        if (_this.exceptions.objectsInObjects === undefined) ***REMOVED***
            _this.exceptions.objectsInObjects = _this.always;
        ***REMOVED***
        return _this;
    ***REMOVED***
    ObjectCurlySpacingWalker.prototype.visitNode = function (node) ***REMOVED***
        var bracedKind = [
            ts.SyntaxKind.ObjectLiteralExpression,
            ts.SyntaxKind.ObjectBindingPattern,
            ts.SyntaxKind.NamedImports,
            ts.SyntaxKind.NamedExports
        ];
        if (bracedKind.indexOf(node.kind) > -1) ***REMOVED***
            this.checkSpacingInsideBraces(node);
        ***REMOVED***
        _super.prototype.visitNode.call(this, node);
    ***REMOVED***;
    ObjectCurlySpacingWalker.prototype.checkSpacingInsideBraces = function (node) ***REMOVED***
        var text = node.getText();
        if (text.indexOf('\n') !== -1 || /^\***REMOVED***\s*\***REMOVED***$/.test(text)) ***REMOVED***
            return;
        ***REMOVED***
        var endsWithObjectLiteral = false;
        var endsWithArrayLiteral = false;
        if (node.getChildren().length === 3) ***REMOVED***
            var contents = node.getChildren()[1].getChildren();
            if (contents.length > 0) ***REMOVED***
                var lastElement = contents[contents.length - 1];
                if (lastElement.kind === ts.SyntaxKind.PropertyAssignment || lastElement.kind === ts.SyntaxKind.BindingElement) ***REMOVED***
                    var value = lastElement.getChildren();
                    if (value.length === 3) ***REMOVED***
                        endsWithObjectLiteral = value[2].kind === ts.SyntaxKind.ObjectLiteralExpression || value[2].kind === ts.SyntaxKind.ObjectBindingPattern;
                        endsWithArrayLiteral = value[2].kind === ts.SyntaxKind.ArrayLiteralExpression;
                    ***REMOVED***
                ***REMOVED***
            ***REMOVED***
        ***REMOVED***
        var leadingSpace = text.match(/^\***REMOVED***(\s***REMOVED***0,2***REMOVED***)/)[1].length;
        if (this.always) ***REMOVED***
            if (leadingSpace === 0) ***REMOVED***
                var fix = Lint.Replacement.appendText(node.getStart() + 1, ' ');
                this.addFailure(this.createFailure(node.getStart(), 1, Rule.FAILURE_STRING.always.start, fix));
            ***REMOVED***
        ***REMOVED***
        else ***REMOVED***
            if (leadingSpace > 0) ***REMOVED***
                var fix = Lint.Replacement.deleteText(node.getStart() + 1, leadingSpace);
                this.addFailure(this.createFailure(node.getStart(), 1, Rule.FAILURE_STRING.never.start, fix));
            ***REMOVED***
        ***REMOVED***
        var trailingSpace = text.match(/(\s***REMOVED***0,2***REMOVED***)***REMOVED***$/)[1].length;
        var arrayExceptionApplies = this.always !== this.exceptions.arraysInObjects && endsWithArrayLiteral;
        var objectExceptionApplies = this.always !== this.exceptions.objectsInObjects && endsWithObjectLiteral;
        var spaceRequired = arrayExceptionApplies || objectExceptionApplies ? !this.always : this.always;
        if (spaceRequired) ***REMOVED***
            if (trailingSpace === 0) ***REMOVED***
                var fix = Lint.Replacement.appendText(node.getEnd() - 1, ' ');
                this.addFailure(this.createFailure(node.getEnd() - 1, 1, Rule.FAILURE_STRING.always.end, fix));
            ***REMOVED***
        ***REMOVED***
        else ***REMOVED***
            if (trailingSpace > 0) ***REMOVED***
                var fix = Lint.Replacement.deleteText(node.getEnd() - trailingSpace - 1, trailingSpace);
                this.addFailure(this.createFailure(node.getEnd() - 1, 1, Rule.FAILURE_STRING.never.end, fix));
            ***REMOVED***
        ***REMOVED***
    ***REMOVED***;
    return ObjectCurlySpacingWalker;
***REMOVED***(Lint.RuleWalker));
