// tslint-eslint-rules/dist/rules/terArrowSpacingRule.js
"use strict";
Object.defineProperty(exports, "__esModule", ***REMOVED*** value: true ***REMOVED***);
var tslib_1 = require("tslib");
var Lint = require("tslint");
var RULE_NAME = 'ter-arrow-spacing';
var Rule = (function (_super) ***REMOVED***
    tslib_1.__extends(Rule, _super);
    function Rule() ***REMOVED***
        return _super !== null && _super.apply(this, arguments) || this;
    ***REMOVED***
    Rule.prototype.apply = function (sourceFile) ***REMOVED***
        var walker = new RuleWalker(sourceFile, this.getOptions());
        return this.applyWithWalker(walker);
    ***REMOVED***;
    Rule.metadata = ***REMOVED***
        ruleName: RULE_NAME,
        description: 'require space before/after arrow function\'s arrow',
        rationale: Lint.Utils.dedent(templateObject_1 || (templateObject_1 = tslib_1.__makeTemplateObject(["\n      This rule normalizes the style of spacing before/after an arrow function\u2019s arrow(`=>`).\n      "], ["\n      This rule normalizes the style of spacing before/after an arrow function\u2019s arrow(\\`=>\\`).\n      "]))),
        optionsDescription: Lint.Utils.dedent(templateObject_2 || (templateObject_2 = tslib_1.__makeTemplateObject(["\n      This rule takes an object argument with `before` and `after` properties, each with a\n      Boolean value.\n\n      The default configuration is `***REMOVED*** \"before\": true, \"after\": true ***REMOVED***`.\n\n      `true` means there should be one or more spaces and `false` means no spaces.\n      "], ["\n      This rule takes an object argument with \\`before\\` and \\`after\\` properties, each with a\n      Boolean value.\n\n      The default configuration is \\`***REMOVED*** \"before\": true, \"after\": true ***REMOVED***\\`.\n\n      \\`true\\` means there should be one or more spaces and \\`false\\` means no spaces.\n      "]))),
        options: ***REMOVED***
            type: 'array',
            items: [***REMOVED***
                    type: 'object',
                    properties: ***REMOVED***
                        before: ***REMOVED***
                            type: 'boolean'
                        ***REMOVED***,
                        after: ***REMOVED***
                            type: 'boolean'
                        ***REMOVED***
                    ***REMOVED***,
                    additionalProperties: false
                ***REMOVED***],
            maxLength: 1
        ***REMOVED***,
        optionExamples: [
            Lint.Utils.dedent(templateObject_3 || (templateObject_3 = tslib_1.__makeTemplateObject(["\n        \"", "\": [true]\n        "], ["\n        \"", "\": [true]\n        "])), RULE_NAME),
            Lint.Utils.dedent(templateObject_4 || (templateObject_4 = tslib_1.__makeTemplateObject(["\n        \"", "\": [true, ***REMOVED***\n          \"before\": false,\n          \"after\": false\n        ***REMOVED***]\n        "], ["\n        \"", "\": [true, ***REMOVED***\n          \"before\": false,\n          \"after\": false\n        ***REMOVED***]\n        "])), RULE_NAME)
        ],
        typescriptOnly: false,
        type: 'style'
    ***REMOVED***;
    return Rule;
***REMOVED***(Lint.Rules.AbstractRule));
exports.Rule = Rule;
var RuleWalker = (function (_super) ***REMOVED***
    tslib_1.__extends(RuleWalker, _super);
    function RuleWalker(sourceFile, options) ***REMOVED***
        var _this = _super.call(this, sourceFile, options) || this;
        _this.before = true;
        _this.after = true;
        var opt = _this.getOptions();
        if (opt[0]) ***REMOVED***
            _this.before = opt[0].before !== false;
            _this.after = opt[0].after !== false;
        ***REMOVED***
        _this.srcFile = sourceFile;
        _this.srcText = sourceFile.getFullText();
        return _this;
    ***REMOVED***
    RuleWalker.prototype.visitArrowFunction = function (node) ***REMOVED***
        var arrow = node.equalsGreaterThanToken;
        var arrowStart = arrow.getStart(this.srcFile);
        var bodyStart = node.body.getStart(this.srcFile);
        var space = ***REMOVED***
            before: /\s/.test(this.srcText[arrowStart - 1]),
            after: /\s/.test(this.srcText[arrow.end])
        ***REMOVED***;
        if (this.before) ***REMOVED***
            if (!space.before) ***REMOVED***
                var fix = Lint.Replacement.appendText(arrowStart, ' ');
                this.report(arrow, 'Missing', 'before', fix);
            ***REMOVED***
        ***REMOVED***
        else ***REMOVED***
            if (space.before) ***REMOVED***
                var spaces = arrowStart - arrow.getFullStart();
                var fix = Lint.Replacement.deleteText(arrowStart - spaces, spaces);
                this.report(arrow, 'Unexpected', 'before', fix);
            ***REMOVED***
        ***REMOVED***
        if (this.after) ***REMOVED***
            if (!space.after) ***REMOVED***
                var fix = Lint.Replacement.appendText(arrow.end, ' ');
                this.report(arrow, 'Missing', 'after', fix);
            ***REMOVED***
        ***REMOVED***
        else ***REMOVED***
            if (space.after) ***REMOVED***
                var fix = Lint.Replacement.deleteText(arrow.end, bodyStart - arrow.end);
                this.report(arrow, 'Unexpected', 'after', fix);
            ***REMOVED***
        ***REMOVED***
        _super.prototype.visitArrowFunction.call(this, node);
    ***REMOVED***;
    RuleWalker.prototype.report = function (arrowToken, status, place, fix) ***REMOVED***
        var failure = this.createFailure(arrowToken.getStart(this.srcFile), arrowToken.getWidth(this.srcFile), status + " space " + place + " =>.", fix);
        this.addFailure(failure);
    ***REMOVED***;
    return RuleWalker;
***REMOVED***(Lint.RuleWalker));
var templateObject_1, templateObject_2, templateObject_3, templateObject_4;
