// tslint-eslint-rules/dist/rules/terFuncCallSpacingRule.js
"use strict";
Object.defineProperty(exports, "__esModule", ***REMOVED*** value: true ***REMOVED***);
var tslib_1 = require("tslib");
var ts = require("typescript");
var Lint = require("tslint");
var RULE_NAME = 'ter-func-call-spacing';
var ALWAYS = 'always';
var MISSING_SPACE = 'Missing space between function name and paren.';
var UNEXPECTED_SPACE = 'Unexpected space between function name and paren.';
var UNEXPECTED_NEWLINE = 'Unexpected newline between function name and paren.';
var Rule = (function (_super) ***REMOVED***
    tslib_1.__extends(Rule, _super);
    function Rule() ***REMOVED***
        return _super !== null && _super.apply(this, arguments) || this;
    ***REMOVED***
    Rule.prototype.apply = function (sourceFile) ***REMOVED***
        var options = ***REMOVED***
            expectSpace: false,
            spacePattern: /\s/
        ***REMOVED***;
        var userOptions = this.getOptions().ruleArguments;
        if (userOptions[0] === ALWAYS) ***REMOVED***
            options.expectSpace = true;
            if (userOptions[1] !== undefined && userOptions[1].allowNewlines) ***REMOVED***
                options.spacePattern = /[ \t\r\n\u2028\u2029]/;
            ***REMOVED***
            else ***REMOVED***
                options.spacePattern = /[ \t]/;
            ***REMOVED***
        ***REMOVED***
        var walker = new RuleWalker(sourceFile, RULE_NAME, options);
        return this.applyWithWalker(walker);
    ***REMOVED***;
    Rule.metadata = ***REMOVED***
        ruleName: RULE_NAME,
        hasFix: true,
        description: 'require or disallow spacing between function identifiers and their invocations',
        rationale: Lint.Utils.dedent(templateObject_1 || (templateObject_1 = tslib_1.__makeTemplateObject(["\n      This rule will enforce consistency of spacing in function calls,\n      by disallowing or requiring one or more spaces before the open paren.\n      "], ["\n      This rule will enforce consistency of spacing in function calls,\n      by disallowing or requiring one or more spaces before the open paren.\n      "]))),
        optionsDescription: Lint.Utils.dedent(templateObject_2 || (templateObject_2 = tslib_1.__makeTemplateObject(["\n      This rule has a string option:\n\n      * `\"never\"` (default) disallows space between the function name and the opening parenthesis.\n      * `\"always\"` requires space between the function name and the opening parenthesis.\n\n      Further, in `\"always\"` mode, a second object option is available that contains a single boolean `allowNewlines` property.\n      "], ["\n      This rule has a string option:\n\n      * \\`\"never\"\\` (default) disallows space between the function name and the opening parenthesis.\n      * \\`\"always\"\\` requires space between the function name and the opening parenthesis.\n\n      Further, in \\`\"always\"\\` mode, a second object option is available that contains a single boolean \\`allowNewlines\\` property.\n      "]))),
        options: ***REMOVED***
            type: 'array',
            items: [
                ***REMOVED***
                    enum: ['always', 'never']
                ***REMOVED***,
                ***REMOVED***
                    type: 'object',
                    properties: ***REMOVED***
                        allowNewlines: ***REMOVED***
                            type: 'boolean'
                        ***REMOVED***
                    ***REMOVED***,
                    additionalProperties: false
                ***REMOVED***
            ],
            minItems: 0,
            maxItems: 2
        ***REMOVED***,
        optionExamples: [
            Lint.Utils.dedent(templateObject_3 || (templateObject_3 = tslib_1.__makeTemplateObject(["\n        \"", "\": [true]\n        "], ["\n        \"", "\": [true]\n        "])), RULE_NAME),
            Lint.Utils.dedent(templateObject_4 || (templateObject_4 = tslib_1.__makeTemplateObject(["\n        \"", "\": [true, \"always\"]\n        "], ["\n        \"", "\": [true, \"always\"]\n        "])), RULE_NAME),
            Lint.Utils.dedent(templateObject_5 || (templateObject_5 = tslib_1.__makeTemplateObject(["\n        \"", "\": [true, \"always\", ***REMOVED*** allowNewlines: true ***REMOVED***]\n        "], ["\n        \"", "\": [true, \"always\", ***REMOVED*** allowNewlines: true ***REMOVED***]\n        "])), RULE_NAME)
        ],
        typescriptOnly: false,
        type: 'style'
    ***REMOVED***;
    return Rule;
***REMOVED***(Lint.Rules.AbstractRule));
exports.Rule = Rule;
var RuleWalker = (function (_super) ***REMOVED***
    tslib_1.__extends(RuleWalker, _super);
    function RuleWalker(sourceFile, ruleName, options) ***REMOVED***
        var _this = _super.call(this, sourceFile, ruleName, options) || this;
        _this.sourceText = sourceFile.getFullText();
        return _this;
    ***REMOVED***
    RuleWalker.prototype.walk = function (sourceFile) ***REMOVED***
        var _this = this;
        var cb = function (node) ***REMOVED***
            if (node.kind === ts.SyntaxKind.NewExpression) ***REMOVED***
                _this.visitNewExpression(node);
            ***REMOVED***
            else if (node.kind === ts.SyntaxKind.CallExpression) ***REMOVED***
                _this.visitCallExpression(node);
            ***REMOVED***
            else if (node.kind >= ts.SyntaxKind.FirstTypeNode && node.kind <= ts.SyntaxKind.LastTypeNode) ***REMOVED***
                return;
            ***REMOVED***
            return ts.forEachChild(node, cb);
        ***REMOVED***;
        return ts.forEachChild(sourceFile, cb);
    ***REMOVED***;
    RuleWalker.prototype.visitNewExpression = function (node) ***REMOVED***
        this.checkWhitespaceAfterExpression(node.expression, node.typeArguments, node.arguments);
    ***REMOVED***;
    RuleWalker.prototype.visitCallExpression = function (node) ***REMOVED***
        this.checkWhitespaceAfterExpression(node.expression, node.typeArguments, node.arguments);
    ***REMOVED***;
    RuleWalker.prototype.checkWhitespaceAfterExpression = function (expression, typeArguments, funcArguments) ***REMOVED***
        if (funcArguments !== undefined) ***REMOVED***
            var start = void 0;
            if (typeArguments !== undefined) ***REMOVED***
                start = typeArguments.end + 1;
            ***REMOVED***
            else ***REMOVED***
                start = expression.getEnd();
            ***REMOVED***
            this.checkWhitespaceBetween(start, funcArguments.pos - 1);
        ***REMOVED***
    ***REMOVED***;
    RuleWalker.prototype.checkWhitespaceBetween = function (start, end) ***REMOVED***
        var whitespace = this.sourceText.substring(start, end);
        if (this.options.spacePattern.test(whitespace)) ***REMOVED***
            if (!this.options.expectSpace) ***REMOVED***
                var fix = Lint.Replacement.deleteText(start, whitespace.length);
                var failureMessage = this.failureMessageForUnexpectedWhitespace(whitespace);
                this.addFailureAt(start, whitespace.length, failureMessage, fix);
            ***REMOVED***
        ***REMOVED***
        else if (this.options.expectSpace) ***REMOVED***
            var fix = Lint.Replacement.appendText(start, ' ');
            this.addFailureAt(start, 1, MISSING_SPACE, fix);
        ***REMOVED***
    ***REMOVED***;
    RuleWalker.prototype.failureMessageForUnexpectedWhitespace = function (whitespace) ***REMOVED***
        if (/[\r\n]/.test(whitespace)) ***REMOVED***
            return UNEXPECTED_NEWLINE;
        ***REMOVED***
        else ***REMOVED***
            return UNEXPECTED_SPACE;
        ***REMOVED***
    ***REMOVED***;
    return RuleWalker;
***REMOVED***(Lint.AbstractWalker));
var templateObject_1, templateObject_2, templateObject_3, templateObject_4, templateObject_5;
