// tslint-eslint-rules/dist/rules/terNoSparseArraysRule.js
"use strict";
Object.defineProperty(exports, "__esModule", ***REMOVED*** value: true ***REMOVED***);
var tslib_1 = require("tslib");
var ts = require("typescript");
var Lint = require("tslint");
var RULE_NAME = 'ter-no-sparse-arrays';
var Rule = (function (_super) ***REMOVED***
    tslib_1.__extends(Rule, _super);
    function Rule() ***REMOVED***
        return _super !== null && _super.apply(this, arguments) || this;
    ***REMOVED***
    Rule.prototype.apply = function (sourceFile) ***REMOVED***
        var walker = new NoSparseArraysWalker(sourceFile, this.getOptions());
        return this.applyWithWalker(walker);
    ***REMOVED***;
    Rule.metadata = ***REMOVED***
        ruleName: RULE_NAME,
        description: 'disallow sparse arrays (recommended)',
        rationale: Lint.Utils.dedent(templateObject_1 || (templateObject_1 = tslib_1.__makeTemplateObject(["\n      Invalid or irregular whitespace causes issues with ECMAScript 5 parsers and also makes code\n      harder to debug in a similar nature to mixed tabs and spaces.\n      "], ["\n      Invalid or irregular whitespace causes issues with ECMAScript 5 parsers and also makes code\n      harder to debug in a similar nature to mixed tabs and spaces.\n      "]))),
        optionsDescription: '',
        options: ***REMOVED******REMOVED***,
        optionExamples: [
            Lint.Utils.dedent(templateObject_2 || (templateObject_2 = tslib_1.__makeTemplateObject(["\n        \"", "\": [true]\n        "], ["\n        \"", "\": [true]\n        "])), RULE_NAME)
        ],
        typescriptOnly: false,
        type: 'typescript'
    ***REMOVED***;
    Rule.FAILURE_STRING = 'unexpected comma in middle of array';
    return Rule;
***REMOVED***(Lint.Rules.AbstractRule));
exports.Rule = Rule;
var NoSparseArraysWalker = (function (_super) ***REMOVED***
    tslib_1.__extends(NoSparseArraysWalker, _super);
    function NoSparseArraysWalker() ***REMOVED***
        return _super !== null && _super.apply(this, arguments) || this;
    ***REMOVED***
    NoSparseArraysWalker.prototype.visitArrayLiteralExpression = function (node) ***REMOVED***
        this.validateNoSparseArray(node);
        _super.prototype.visitArrayLiteralExpression.call(this, node);
    ***REMOVED***;
    NoSparseArraysWalker.prototype.validateNoSparseArray = function (node) ***REMOVED***
        var hasEmptySlot = node.elements.some(function (el) ***REMOVED*** return el.kind === ts.SyntaxKind.OmittedExpression; ***REMOVED***);
        if (hasEmptySlot) ***REMOVED***
            this.addFailure(this.createFailure(node.getStart(), node.getWidth(), Rule.FAILURE_STRING));
        ***REMOVED***
    ***REMOVED***;
    return NoSparseArraysWalker;
***REMOVED***(Lint.RuleWalker));
var templateObject_1, templateObject_2;
