// tslint-eslint-rules/dist/rules/terNoSelfCompareRule.js
"use strict";
Object.defineProperty(exports, "__esModule", ***REMOVED*** value: true ***REMOVED***);
var tslib_1 = require("tslib");
var ts = require("typescript");
var Lint = require("tslint");
var RULE_NAME = 'ter-no-self-compare';
var Rule = (function (_super) ***REMOVED***
    tslib_1.__extends(Rule, _super);
    function Rule() ***REMOVED***
        return _super !== null && _super.apply(this, arguments) || this;
    ***REMOVED***
    Rule.prototype.apply = function (sourceFile) ***REMOVED***
        return this.applyWithWalker(new NoSelfCompareWalker(sourceFile, this.ruleName, new Set(this.ruleArguments.map(String))));
    ***REMOVED***;
    Rule.FAILURE_STRING = 'Comparing to itself is potentially pointless.';
    Rule.metadata = ***REMOVED***
        ruleName: RULE_NAME,
        hasFix: false,
        description: 'disallow comparisons where both sides are exactly the same',
        rationale: 'Comparing a variable against itself is usually an error, ' +
            'either a typo or refactoring error. It is confusing to the reader ' +
            'and may potentially introduce a runtime error.',
        optionsDescription: '',
        options: ***REMOVED******REMOVED***,
        optionExamples: [
            Lint.Utils.dedent(templateObject_1 || (templateObject_1 = tslib_1.__makeTemplateObject(["\n        \"", "\": true\n        "], ["\n        \"", "\": true\n        "])), RULE_NAME)
        ],
        typescriptOnly: false,
        type: 'maintainability'
    ***REMOVED***;
    return Rule;
***REMOVED***(Lint.Rules.AbstractRule));
exports.Rule = Rule;
var NoSelfCompareWalker = (function (_super) ***REMOVED***
    tslib_1.__extends(NoSelfCompareWalker, _super);
    function NoSelfCompareWalker() ***REMOVED***
        return _super !== null && _super.apply(this, arguments) || this;
    ***REMOVED***
    NoSelfCompareWalker.prototype.isComparisonOperator = function (node) ***REMOVED***
        var operators = new Set([
            ts.SyntaxKind.EqualsEqualsEqualsToken,
            ts.SyntaxKind.EqualsEqualsToken,
            ts.SyntaxKind.ExclamationEqualsEqualsToken,
            ts.SyntaxKind.ExclamationEqualsToken,
            ts.SyntaxKind.GreaterThanToken,
            ts.SyntaxKind.LessThanToken,
            ts.SyntaxKind.GreaterThanEqualsToken,
            ts.SyntaxKind.LessThanEqualsToken
        ]);
        return operators.has(node.operatorToken.kind);
    ***REMOVED***;
    NoSelfCompareWalker.prototype.hasSameToken = function (left, right) ***REMOVED***
        return left.kind === right.kind && left.getText() === right.getText();
    ***REMOVED***;
    NoSelfCompareWalker.prototype.walk = function (sourceFile) ***REMOVED***
        var _this = this;
        var cb = function (node) ***REMOVED***
            if (ts.isBinaryExpression(node)) ***REMOVED***
                var nodeExpr = node;
                if (_this.isComparisonOperator(nodeExpr) &&
                    _this.hasSameToken(nodeExpr.left, nodeExpr.right)) ***REMOVED***
                    _this.addFailureAt(nodeExpr.operatorToken.getStart(), nodeExpr.operatorToken.getWidth(), Rule.FAILURE_STRING);
                ***REMOVED***
            ***REMOVED***
            else ***REMOVED***
                return ts.forEachChild(node, cb);
            ***REMOVED***
        ***REMOVED***;
        return ts.forEachChild(sourceFile, cb);
    ***REMOVED***;
    return NoSelfCompareWalker;
***REMOVED***(Lint.AbstractWalker));
var templateObject_1;
