// tslint-eslint-rules/dist/rules/terNewlineAfterVarRule.js
"use strict";
Object.defineProperty(exports, "__esModule", ***REMOVED*** value: true ***REMOVED***);
var tslib_1 = require("tslib");
var ts = require("typescript");
var Lint = require("tslint");
var RULE_NAME = 'ter-newline-after-var';
var EXPECTED_BLANK_LINE_MESSAGE = 'Expected blank line after variable declarations.';
var UNEXPECTED_BLANK_LINE_MESSAGE = 'Unexpected blank line after variable declarations.';
var Rule = (function (_super) ***REMOVED***
    tslib_1.__extends(Rule, _super);
    function Rule() ***REMOVED***
        return _super !== null && _super.apply(this, arguments) || this;
    ***REMOVED***
    Rule.prototype.formatOptions = function (_a) ***REMOVED***
        var alwaysOrNever = _a[0];
        return ***REMOVED***
            always: alwaysOrNever !== 'never'
        ***REMOVED***;
    ***REMOVED***;
    Rule.prototype.apply = function (sourceFile) ***REMOVED***
        var opt = this.formatOptions(this.ruleArguments);
        var walker = new RuleWalker(sourceFile, this.ruleName, opt);
        return this.applyWithWalker(walker);
    ***REMOVED***;
    Rule.metadata = ***REMOVED***
        ruleName: RULE_NAME,
        hasFix: true,
        description: 'require or disallow an empty line after variable declarations',
        rationale: Lint.Utils.dedent(templateObject_1 || (templateObject_1 = tslib_1.__makeTemplateObject(["\n      This rule enforces a coding style where empty lines are required or disallowed after `var`, `let`, or `const`\n      statements to achieve a consistent coding style across the project.\n      "], ["\n      This rule enforces a coding style where empty lines are required or disallowed after \\`var\\`, \\`let\\`, or \\`const\\`\n      statements to achieve a consistent coding style across the project.\n      "]))),
        optionsDescription: Lint.Utils.dedent(templateObject_2 || (templateObject_2 = tslib_1.__makeTemplateObject(["\n      This rule has a string option:\n\n      * `\"always\"` (default) requires an empty line after `var`, `let`, or `const`.\n        Comments on a line directly after var statements are treated like additional var statements.\n      * `\"never\"` disallows empty lines after `var`, `let`, or `const`.\n      "], ["\n      This rule has a string option:\n\n      * \\`\"always\"\\` (default) requires an empty line after \\`var\\`, \\`let\\`, or \\`const\\`.\n        Comments on a line directly after var statements are treated like additional var statements.\n      * \\`\"never\"\\` disallows empty lines after \\`var\\`, \\`let\\`, or \\`const\\`.\n      "]))),
        options: ***REMOVED***
            type: 'array',
            items: [***REMOVED***
                    enum: ['always', 'never']
                ***REMOVED***],
            maxLength: 1
        ***REMOVED***,
        optionExamples: [
            Lint.Utils.dedent(templateObject_3 || (templateObject_3 = tslib_1.__makeTemplateObject(["\n        \"", "\": [true]\n        "], ["\n        \"", "\": [true]\n        "])), RULE_NAME),
            Lint.Utils.dedent(templateObject_4 || (templateObject_4 = tslib_1.__makeTemplateObject(["\n        \"", "\": [true, \"always\"]\n        "], ["\n        \"", "\": [true, \"always\"]\n        "])), RULE_NAME),
            Lint.Utils.dedent(templateObject_5 || (templateObject_5 = tslib_1.__makeTemplateObject(["\n        \"", "\": [true, \"never\"]\n        "], ["\n        \"", "\": [true, \"never\"]\n        "])), RULE_NAME)
        ],
        typescriptOnly: false,
        type: 'style'
    ***REMOVED***;
    return Rule;
***REMOVED***(Lint.Rules.AbstractRule));
exports.Rule = Rule;
var RuleWalker = (function (_super) ***REMOVED***
    tslib_1.__extends(RuleWalker, _super);
    function RuleWalker() ***REMOVED***
        return _super !== null && _super.apply(this, arguments) || this;
    ***REMOVED***
    RuleWalker.prototype.walk = function (sourceFile) ***REMOVED***
        var _this = this;
        this.sourceFileText = sourceFile.getFullText();
        var onNode = function (node) ***REMOVED***
            var _a = _this, lastVariableStatementNode = _a.lastVariableStatementNode, sourceFileText = _a.sourceFileText;
            if (node.kind === ts.SyntaxKind.VariableStatement) ***REMOVED***
                _this.lastVariableStatementNode = node;
                return;
            ***REMOVED***
            if (node.kind === ts.SyntaxKind.EndOfFileToken) ***REMOVED***
                _this.lastVariableStatementNode = undefined;
                return;
            ***REMOVED***
            if (lastVariableStatementNode) ***REMOVED***
                var unexpectedLineFixes = [];
                var expectedLineFixes = [];
                var isNewLineRequired = _this.options.always;
                var expectedLinePos = lastVariableStatementNode.end;
                var newLinesCount = 0;
                for (var i = lastVariableStatementNode.end; i < node.end; i++) ***REMOVED***
                    var code = sourceFileText.charCodeAt(i);
                    if (code === 10) ***REMOVED***
                        newLinesCount++;
                        if (!isNewLineRequired && newLinesCount > 1) ***REMOVED***
                            unexpectedLineFixes.push(Lint.Replacement.deleteText(i, 1));
                        ***REMOVED***
                    ***REMOVED***
                    else if (code !== 9 && code !== 13 && code !== 32) ***REMOVED***
                        var leadingComments = ts.getLeadingCommentRanges("\n" + sourceFileText.slice(i), 0);
                        var lastLeadingComment = leadingComments && leadingComments.pop();
                        if (lastLeadingComment && (!isNewLineRequired || (isNewLineRequired && newLinesCount < 2))) ***REMOVED***
                            newLinesCount = 0;
                            expectedLinePos = i - 1 + lastLeadingComment.end;
                            i = expectedLinePos - 1;
                        ***REMOVED***
                        else ***REMOVED***
                            if (isNewLineRequired && newLinesCount < 2) ***REMOVED***
                                expectedLineFixes.push(Lint.Replacement.appendText(expectedLinePos, '\n'));
                            ***REMOVED***
                            break;
                        ***REMOVED***
                    ***REMOVED***
                ***REMOVED***
                if (isNewLineRequired && expectedLineFixes[0]) ***REMOVED***
                    _this.addFailureAt(lastVariableStatementNode.getStart(), 1, EXPECTED_BLANK_LINE_MESSAGE, expectedLineFixes);
                ***REMOVED***
                else if (unexpectedLineFixes[0]) ***REMOVED***
                    _this.addFailureAt(lastVariableStatementNode.getStart(), 1, UNEXPECTED_BLANK_LINE_MESSAGE, unexpectedLineFixes);
                ***REMOVED***
                _this.lastVariableStatementNode = undefined;
            ***REMOVED***
            return ts.forEachChild(node, onNode);
        ***REMOVED***;
        return ts.forEachChild(sourceFile, onNode);
    ***REMOVED***;
    return RuleWalker;
***REMOVED***(Lint.AbstractWalker));
var templateObject_1, templateObject_2, templateObject_3, templateObject_4, templateObject_5;
