// tslint-eslint-rules/dist/rules/terPaddedBlocksRule.js
"use strict";
Object.defineProperty(exports, "__esModule", ***REMOVED*** value: true ***REMOVED***);
var tslib_1 = require("tslib");
var ts = require("typescript");
var Lint = require("tslint");
var tsutils_1 = require("tsutils");
var RULE_NAME = 'ter-padded-blocks';
var OPTION_ALWAYS = 'always';
var Rule = (function (_super) ***REMOVED***
    tslib_1.__extends(Rule, _super);
    function Rule() ***REMOVED***
        return _super !== null && _super.apply(this, arguments) || this;
    ***REMOVED***
    Rule.prototype.formatOptions = function (ruleArguments) ***REMOVED***
        var config = ruleArguments[0] || OPTION_ALWAYS;
        if (typeof (config) === 'string') ***REMOVED***
            var always = config === OPTION_ALWAYS;
            return ***REMOVED***
                blocks: always,
                classes: always,
                switches: always
            ***REMOVED***;
        ***REMOVED***
        return ***REMOVED***
            blocks: config['blocks'] && config['blocks'] === OPTION_ALWAYS,
            classes: config['classes'] && config['classes'] === OPTION_ALWAYS,
            switches: config['switches'] && config['switches'] === OPTION_ALWAYS
        ***REMOVED***;
    ***REMOVED***;
    Rule.prototype.apply = function (sourceFile) ***REMOVED***
        var opt = this.formatOptions(this.ruleArguments);
        var walker = new RuleWalker(sourceFile, this.ruleName, opt);
        return this.applyWithWalker(walker);
    ***REMOVED***;
    Rule.metadata = ***REMOVED***
        ruleName: RULE_NAME,
        hasFix: false,
        description: 'enforces consistent empty line padding within blocks',
        rationale: Lint.Utils.dedent(templateObject_1 || (templateObject_1 = tslib_1.__makeTemplateObject(["\n      Some style guides require block statements to start and end with blank\n      lines. The goal is to improve readability by visually separating the\n      block content and the surrounding code.\n      "], ["\n      Some style guides require block statements to start and end with blank\n      lines. The goal is to improve readability by visually separating the\n      block content and the surrounding code.\n      "]))),
        optionsDescription: 'This rule has one option, which can be a string option or an object option',
        options: ***REMOVED***
            type: 'array',
            items: [
                ***REMOVED***
                    enum: ['always', 'never']
                ***REMOVED***,
                ***REMOVED***
                    type: 'object',
                    properties: ***REMOVED***
                        blocks: ***REMOVED***
                            enum: ['always', 'never']
                        ***REMOVED***,
                        classes: ***REMOVED***
                            enum: ['always', 'never']
                        ***REMOVED***,
                        switches: ***REMOVED***
                            enum: ['always', 'never']
                        ***REMOVED***
                    ***REMOVED***,
                    additionalProperties: false
                ***REMOVED***
            ],
            maxLength: 1
        ***REMOVED***,
        optionExamples: [
            Lint.Utils.dedent(templateObject_2 || (templateObject_2 = tslib_1.__makeTemplateObject(["\n        \"", "\": [true]\n        "], ["\n        \"", "\": [true]\n        "])), RULE_NAME),
            Lint.Utils.dedent(templateObject_3 || (templateObject_3 = tslib_1.__makeTemplateObject(["\n        \"", "\": [true, \"always\"]\n        "], ["\n        \"", "\": [true, \"always\"]\n        "])), RULE_NAME),
            Lint.Utils.dedent(templateObject_4 || (templateObject_4 = tslib_1.__makeTemplateObject(["\n        \"", "\": [true, \"never\"]\n        "], ["\n        \"", "\": [true, \"never\"]\n        "])), RULE_NAME),
            Lint.Utils.dedent(templateObject_5 || (templateObject_5 = tslib_1.__makeTemplateObject(["\n        \"", "\": [true, ***REMOVED*** \"blocks\": \"always\" ***REMOVED***]\n        "], ["\n        \"", "\": [true, ***REMOVED*** \"blocks\": \"always\" ***REMOVED***]\n        "])), RULE_NAME),
            Lint.Utils.dedent(templateObject_6 || (templateObject_6 = tslib_1.__makeTemplateObject(["\n        \"", "\": [true, ***REMOVED*** \"blocks\": \"never\" ***REMOVED***]\n        "], ["\n        \"", "\": [true, ***REMOVED*** \"blocks\": \"never\" ***REMOVED***]\n        "])), RULE_NAME)
        ],
        typescriptOnly: false,
        type: 'style'
    ***REMOVED***;
    Rule.FAILURE_STRING = ***REMOVED***
        always: 'Block must be padded by blank lines.',
        never: 'Block must not be padded by blank lines.'
    ***REMOVED***;
    return Rule;
***REMOVED***(Lint.Rules.AbstractRule));
exports.Rule = Rule;
var RuleWalker = (function (_super) ***REMOVED***
    tslib_1.__extends(RuleWalker, _super);
    function RuleWalker() ***REMOVED***
        return _super !== null && _super.apply(this, arguments) || this;
    ***REMOVED***
    RuleWalker.prototype.walk = function (sourceFile) ***REMOVED***
        var _this = this;
        sourceFile.forEachChild(function (node) ***REMOVED*** return _this.processNode(node); ***REMOVED***);
    ***REMOVED***;
    RuleWalker.prototype.processNode = function (node) ***REMOVED***
        var _this = this;
        switch (node.kind) ***REMOVED***
            case ts.SyntaxKind.Block:
            case ts.SyntaxKind.ClassDeclaration:
            case ts.SyntaxKind.CaseBlock:
                this.checkPadding(node);
        ***REMOVED***
        node.forEachChild(function (child) ***REMOVED*** return _this.processNode(child); ***REMOVED***);
    ***REMOVED***;
    RuleWalker.prototype.getParts = function (node) ***REMOVED***
        var openBrace, body, closeBrace;
        node.getChildren().forEach(function (child) ***REMOVED***
            if (child.kind === ts.SyntaxKind.OpenBraceToken) ***REMOVED***
                openBrace = child;
            ***REMOVED***
            else if (child.kind === ts.SyntaxKind.SyntaxList) ***REMOVED***
                body = child;
            ***REMOVED***
            else if (child.kind === ts.SyntaxKind.CloseBraceToken) ***REMOVED***
                closeBrace = child;
            ***REMOVED***
        ***REMOVED***);
        return ***REMOVED***
            openBrace: openBrace,
            body: body,
            closeBrace: closeBrace
        ***REMOVED***;
    ***REMOVED***;
    RuleWalker.prototype.getPositions = function (node) ***REMOVED***
        var _a = this.getParts(node), openBrace = _a.openBrace, body = _a.body, closeBrace = _a.closeBrace;
        var firstChild = body.getChildAt(0);
        var lastChild = body.getChildAt(body.getChildCount() - 1);
        var positions = ***REMOVED***
            openPosition: openBrace.getStart(this.sourceFile),
            openLine: this.getLine(openBrace.getStart(this.sourceFile)),
            closePosition: closeBrace.getEnd(),
            closeLine: this.getLine(closeBrace.getEnd())
        ***REMOVED***;
        if (firstChild) ***REMOVED***
            positions.firstChildPosition = firstChild.getStart(this.sourceFile);
            positions.firstChildLine = this.getLine(positions.firstChildPosition);
        ***REMOVED***
        if (lastChild) ***REMOVED***
            positions.lastChildPosition = lastChild.getEnd();
            positions.lastChildLine = this.getLine(positions.lastChildPosition);
        ***REMOVED***
        return positions;
    ***REMOVED***;
    RuleWalker.prototype.checkPadding = function (node) ***REMOVED***
        var _this = this;
        var paddingAllowed = this.options.blocks;
        if (node.kind === ts.SyntaxKind.ClassDeclaration) ***REMOVED***
            paddingAllowed = this.options.classes;
        ***REMOVED***
        else if (node.parent && node.parent.kind === ts.SyntaxKind.SwitchStatement) ***REMOVED***
            paddingAllowed = this.options.switches;
        ***REMOVED***
        if (paddingAllowed === undefined) ***REMOVED***
            return;
        ***REMOVED***
        var positions = this.getPositions(node);
        var openBraceReplacement = ***REMOVED***
            from: positions.openPosition + 1,
            to: positions.firstChildPosition || positions.closePosition
        ***REMOVED***;
        var closeBraceReplacement = ***REMOVED***
            from: positions.lastChildPosition || positions.openPosition,
            to: positions.closePosition - 1
        ***REMOVED***;
        var comments = [];
        tsutils_1.forEachComment(node, function (_fullText, comment) ***REMOVED***
            if (comment.pos > positions.openPosition && comment.pos < positions.closePosition) ***REMOVED***
                var commentLineEnd = _this.getLine(comment.end);
                if (commentLineEnd > positions.openLine) ***REMOVED***
                    comments.push(comment);
                ***REMOVED***
                else if (commentLineEnd === positions.openLine) ***REMOVED***
                    openBraceReplacement.from = comment.end;
                ***REMOVED***
            ***REMOVED***
        ***REMOVED***);
        if (comments.length > 0) ***REMOVED***
            var firstCommentLine = this.getLine(comments[0].pos);
            var lastCommentLine = this.getLine(comments[comments.length - 1].end);
            if (!positions.firstChildLine || firstCommentLine < positions.firstChildLine) ***REMOVED***
                positions.firstChildLine = firstCommentLine;
                positions.firstChildPosition = comments[0].pos;
                openBraceReplacement.to = positions.firstChildPosition;
            ***REMOVED***
            if (!positions.lastChildLine || lastCommentLine >= positions.lastChildLine) ***REMOVED***
                positions.lastChildLine = lastCommentLine;
                positions.lastChildPosition = comments[comments.length - 1].end;
                closeBraceReplacement.from = positions.lastChildPosition;
            ***REMOVED***
        ***REMOVED***
        if (this.getLine(openBraceReplacement.from) !== this.getLine(openBraceReplacement.to)) ***REMOVED***
            openBraceReplacement.to = this.getPosition(this.getLine(openBraceReplacement.to));
        ***REMOVED***
        if (this.getLine(closeBraceReplacement.from) !== this.getLine(closeBraceReplacement.to)) ***REMOVED***
            closeBraceReplacement.to = this.getPosition(this.getLine(closeBraceReplacement.to));
        ***REMOVED***
        if (positions.firstChildLine === undefined && positions.lastChildLine === undefined) ***REMOVED***
            return;
        ***REMOVED***
        var openPadded = false;
        if (positions.firstChildLine !== undefined) ***REMOVED***
            openPadded = positions.firstChildLine - positions.openLine > 1;
        ***REMOVED***
        else ***REMOVED***
            openPadded = positions.closeLine - positions.openLine > 1;
        ***REMOVED***
        var closePadded = false;
        if (positions.lastChildLine !== undefined) ***REMOVED***
            closePadded = positions.closeLine - positions.lastChildLine > 1;
        ***REMOVED***
        else ***REMOVED***
            closePadded = positions.closeLine - positions.openLine > 1;
        ***REMOVED***
        if (paddingAllowed ? !openPadded : openPadded) ***REMOVED***
            var openFix = Lint.Replacement.replaceFromTo(openBraceReplacement.from, openBraceReplacement.to, paddingAllowed ? '\n\n' : '\n');
            this.addFailure(positions.openPosition, positions.openPosition + 1, paddingAllowed ? Rule.FAILURE_STRING.always : Rule.FAILURE_STRING.never, openFix);
        ***REMOVED***
        if (paddingAllowed ? !closePadded : closePadded) ***REMOVED***
            var closeFix = Lint.Replacement.replaceFromTo(closeBraceReplacement.from, closeBraceReplacement.to, paddingAllowed ? '\n\n' : '\n');
            this.addFailure(positions.closePosition - 1, positions.closePosition, paddingAllowed ? Rule.FAILURE_STRING.always : Rule.FAILURE_STRING.never, closeFix);
        ***REMOVED***
    ***REMOVED***;
    RuleWalker.prototype.getLine = function (pos) ***REMOVED***
        return this.sourceFile.getLineAndCharacterOfPosition(pos).line;
    ***REMOVED***;
    RuleWalker.prototype.getPosition = function (line) ***REMOVED***
        return this.sourceFile.getPositionOfLineAndCharacter(line, 0);
    ***REMOVED***;
    return RuleWalker;
***REMOVED***(Lint.AbstractWalker));
var templateObject_1, templateObject_2, templateObject_3, templateObject_4, templateObject_5, templateObject_6;
