// rxjs/internal/operators/bufferWhen.js
"use strict";
var __extends = (this && this.__extends) || (function () ***REMOVED***
    var extendStatics = function (d, b) ***REMOVED***
        extendStatics = Object.setPrototypeOf ||
            (***REMOVED*** __proto__: [] ***REMOVED*** instanceof Array && function (d, b) ***REMOVED*** d.__proto__ = b; ***REMOVED***) ||
            function (d, b) ***REMOVED*** for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; ***REMOVED***;
        return extendStatics(d, b);
    ***REMOVED***
    return function (d, b) ***REMOVED***
        extendStatics(d, b);
        function __() ***REMOVED*** this.constructor = d; ***REMOVED***
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    ***REMOVED***;
***REMOVED***)();
Object.defineProperty(exports, "__esModule", ***REMOVED*** value: true ***REMOVED***);
var Subscription_1 = require("../Subscription");
var OuterSubscriber_1 = require("../OuterSubscriber");
var subscribeToResult_1 = require("../util/subscribeToResult");
function bufferWhen(closingSelector) ***REMOVED***
    return function (source) ***REMOVED***
        return source.lift(new BufferWhenOperator(closingSelector));
    ***REMOVED***;
***REMOVED***
exports.bufferWhen = bufferWhen;
var BufferWhenOperator = (function () ***REMOVED***
    function BufferWhenOperator(closingSelector) ***REMOVED***
        this.closingSelector = closingSelector;
    ***REMOVED***
    BufferWhenOperator.prototype.call = function (subscriber, source) ***REMOVED***
        return source.subscribe(new BufferWhenSubscriber(subscriber, this.closingSelector));
    ***REMOVED***;
    return BufferWhenOperator;
***REMOVED***());
var BufferWhenSubscriber = (function (_super) ***REMOVED***
    __extends(BufferWhenSubscriber, _super);
    function BufferWhenSubscriber(destination, closingSelector) ***REMOVED***
        var _this = _super.call(this, destination) || this;
        _this.closingSelector = closingSelector;
        _this.subscribing = false;
        _this.openBuffer();
        return _this;
    ***REMOVED***
    BufferWhenSubscriber.prototype._next = function (value) ***REMOVED***
        this.buffer.push(value);
    ***REMOVED***;
    BufferWhenSubscriber.prototype._complete = function () ***REMOVED***
        var buffer = this.buffer;
        if (buffer) ***REMOVED***
            this.destination.next(buffer);
        ***REMOVED***
        _super.prototype._complete.call(this);
    ***REMOVED***;
    BufferWhenSubscriber.prototype._unsubscribe = function () ***REMOVED***
        this.buffer = null;
        this.subscribing = false;
    ***REMOVED***;
    BufferWhenSubscriber.prototype.notifyNext = function (outerValue, innerValue, outerIndex, innerIndex, innerSub) ***REMOVED***
        this.openBuffer();
    ***REMOVED***;
    BufferWhenSubscriber.prototype.notifyComplete = function () ***REMOVED***
        if (this.subscribing) ***REMOVED***
            this.complete();
        ***REMOVED***
        else ***REMOVED***
            this.openBuffer();
        ***REMOVED***
    ***REMOVED***;
    BufferWhenSubscriber.prototype.openBuffer = function () ***REMOVED***
        var closingSubscription = this.closingSubscription;
        if (closingSubscription) ***REMOVED***
            this.remove(closingSubscription);
            closingSubscription.unsubscribe();
        ***REMOVED***
        var buffer = this.buffer;
        if (this.buffer) ***REMOVED***
            this.destination.next(buffer);
        ***REMOVED***
        this.buffer = [];
        var closingNotifier;
        try ***REMOVED***
            var closingSelector = this.closingSelector;
            closingNotifier = closingSelector();
        ***REMOVED***
        catch (err) ***REMOVED***
            return this.error(err);
        ***REMOVED***
        closingSubscription = new Subscription_1.Subscription();
        this.closingSubscription = closingSubscription;
        this.add(closingSubscription);
        this.subscribing = true;
        closingSubscription.add(subscribeToResult_1.subscribeToResult(this, closingNotifier));
        this.subscribing = false;
    ***REMOVED***;
    return BufferWhenSubscriber;
***REMOVED***(OuterSubscriber_1.OuterSubscriber));
