// rxjs/internal/operators/count.js
"use strict";
var __extends = (this && this.__extends) || (function () ***REMOVED***
    var extendStatics = function (d, b) ***REMOVED***
        extendStatics = Object.setPrototypeOf ||
            (***REMOVED*** __proto__: [] ***REMOVED*** instanceof Array && function (d, b) ***REMOVED*** d.__proto__ = b; ***REMOVED***) ||
            function (d, b) ***REMOVED*** for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; ***REMOVED***;
        return extendStatics(d, b);
    ***REMOVED***
    return function (d, b) ***REMOVED***
        extendStatics(d, b);
        function __() ***REMOVED*** this.constructor = d; ***REMOVED***
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    ***REMOVED***;
***REMOVED***)();
Object.defineProperty(exports, "__esModule", ***REMOVED*** value: true ***REMOVED***);
var Subscriber_1 = require("../Subscriber");
function count(predicate) ***REMOVED***
    return function (source) ***REMOVED*** return source.lift(new CountOperator(predicate, source)); ***REMOVED***;
***REMOVED***
exports.count = count;
var CountOperator = (function () ***REMOVED***
    function CountOperator(predicate, source) ***REMOVED***
        this.predicate = predicate;
        this.source = source;
    ***REMOVED***
    CountOperator.prototype.call = function (subscriber, source) ***REMOVED***
        return source.subscribe(new CountSubscriber(subscriber, this.predicate, this.source));
    ***REMOVED***;
    return CountOperator;
***REMOVED***());
var CountSubscriber = (function (_super) ***REMOVED***
    __extends(CountSubscriber, _super);
    function CountSubscriber(destination, predicate, source) ***REMOVED***
        var _this = _super.call(this, destination) || this;
        _this.predicate = predicate;
        _this.source = source;
        _this.count = 0;
        _this.index = 0;
        return _this;
    ***REMOVED***
    CountSubscriber.prototype._next = function (value) ***REMOVED***
        if (this.predicate) ***REMOVED***
            this._tryPredicate(value);
        ***REMOVED***
        else ***REMOVED***
            this.count++;
        ***REMOVED***
    ***REMOVED***;
    CountSubscriber.prototype._tryPredicate = function (value) ***REMOVED***
        var result;
        try ***REMOVED***
            result = this.predicate(value, this.index++, this.source);
        ***REMOVED***
        catch (err) ***REMOVED***
            this.destination.error(err);
            return;
        ***REMOVED***
        if (result) ***REMOVED***
            this.count++;
        ***REMOVED***
    ***REMOVED***;
    CountSubscriber.prototype._complete = function () ***REMOVED***
        this.destination.next(this.count);
        this.destination.complete();
    ***REMOVED***;
    return CountSubscriber;
***REMOVED***(Subscriber_1.Subscriber));
