// rxjs/internal/operators/debounceTime.js
"use strict";
var __extends = (this && this.__extends) || (function () ***REMOVED***
    var extendStatics = function (d, b) ***REMOVED***
        extendStatics = Object.setPrototypeOf ||
            (***REMOVED*** __proto__: [] ***REMOVED*** instanceof Array && function (d, b) ***REMOVED*** d.__proto__ = b; ***REMOVED***) ||
            function (d, b) ***REMOVED*** for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; ***REMOVED***;
        return extendStatics(d, b);
    ***REMOVED***
    return function (d, b) ***REMOVED***
        extendStatics(d, b);
        function __() ***REMOVED*** this.constructor = d; ***REMOVED***
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    ***REMOVED***;
***REMOVED***)();
Object.defineProperty(exports, "__esModule", ***REMOVED*** value: true ***REMOVED***);
var Subscriber_1 = require("../Subscriber");
var async_1 = require("../scheduler/async");
function debounceTime(dueTime, scheduler) ***REMOVED***
    if (scheduler === void 0) ***REMOVED*** scheduler = async_1.async; ***REMOVED***
    return function (source) ***REMOVED*** return source.lift(new DebounceTimeOperator(dueTime, scheduler)); ***REMOVED***;
***REMOVED***
exports.debounceTime = debounceTime;
var DebounceTimeOperator = (function () ***REMOVED***
    function DebounceTimeOperator(dueTime, scheduler) ***REMOVED***
        this.dueTime = dueTime;
        this.scheduler = scheduler;
    ***REMOVED***
    DebounceTimeOperator.prototype.call = function (subscriber, source) ***REMOVED***
        return source.subscribe(new DebounceTimeSubscriber(subscriber, this.dueTime, this.scheduler));
    ***REMOVED***;
    return DebounceTimeOperator;
***REMOVED***());
var DebounceTimeSubscriber = (function (_super) ***REMOVED***
    __extends(DebounceTimeSubscriber, _super);
    function DebounceTimeSubscriber(destination, dueTime, scheduler) ***REMOVED***
        var _this = _super.call(this, destination) || this;
        _this.dueTime = dueTime;
        _this.scheduler = scheduler;
        _this.debouncedSubscription = null;
        _this.lastValue = null;
        _this.hasValue = false;
        return _this;
    ***REMOVED***
    DebounceTimeSubscriber.prototype._next = function (value) ***REMOVED***
        this.clearDebounce();
        this.lastValue = value;
        this.hasValue = true;
        this.add(this.debouncedSubscription = this.scheduler.schedule(dispatchNext, this.dueTime, this));
    ***REMOVED***;
    DebounceTimeSubscriber.prototype._complete = function () ***REMOVED***
        this.debouncedNext();
        this.destination.complete();
    ***REMOVED***;
    DebounceTimeSubscriber.prototype.debouncedNext = function () ***REMOVED***
        this.clearDebounce();
        if (this.hasValue) ***REMOVED***
            var lastValue = this.lastValue;
            this.lastValue = null;
            this.hasValue = false;
            this.destination.next(lastValue);
        ***REMOVED***
    ***REMOVED***;
    DebounceTimeSubscriber.prototype.clearDebounce = function () ***REMOVED***
        var debouncedSubscription = this.debouncedSubscription;
        if (debouncedSubscription !== null) ***REMOVED***
            this.remove(debouncedSubscription);
            debouncedSubscription.unsubscribe();
            this.debouncedSubscription = null;
        ***REMOVED***
    ***REMOVED***;
    return DebounceTimeSubscriber;
***REMOVED***(Subscriber_1.Subscriber));
function dispatchNext(subscriber) ***REMOVED***
    subscriber.debouncedNext();
***REMOVED***
