// rxjs/internal/operators/defaultIfEmpty.js
"use strict";
var __extends = (this && this.__extends) || (function () ***REMOVED***
    var extendStatics = function (d, b) ***REMOVED***
        extendStatics = Object.setPrototypeOf ||
            (***REMOVED*** __proto__: [] ***REMOVED*** instanceof Array && function (d, b) ***REMOVED*** d.__proto__ = b; ***REMOVED***) ||
            function (d, b) ***REMOVED*** for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; ***REMOVED***;
        return extendStatics(d, b);
    ***REMOVED***
    return function (d, b) ***REMOVED***
        extendStatics(d, b);
        function __() ***REMOVED*** this.constructor = d; ***REMOVED***
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    ***REMOVED***;
***REMOVED***)();
Object.defineProperty(exports, "__esModule", ***REMOVED*** value: true ***REMOVED***);
var Subscriber_1 = require("../Subscriber");
function defaultIfEmpty(defaultValue) ***REMOVED***
    if (defaultValue === void 0) ***REMOVED*** defaultValue = null; ***REMOVED***
    return function (source) ***REMOVED*** return source.lift(new DefaultIfEmptyOperator(defaultValue)); ***REMOVED***;
***REMOVED***
exports.defaultIfEmpty = defaultIfEmpty;
var DefaultIfEmptyOperator = (function () ***REMOVED***
    function DefaultIfEmptyOperator(defaultValue) ***REMOVED***
        this.defaultValue = defaultValue;
    ***REMOVED***
    DefaultIfEmptyOperator.prototype.call = function (subscriber, source) ***REMOVED***
        return source.subscribe(new DefaultIfEmptySubscriber(subscriber, this.defaultValue));
    ***REMOVED***;
    return DefaultIfEmptyOperator;
***REMOVED***());
var DefaultIfEmptySubscriber = (function (_super) ***REMOVED***
    __extends(DefaultIfEmptySubscriber, _super);
    function DefaultIfEmptySubscriber(destination, defaultValue) ***REMOVED***
        var _this = _super.call(this, destination) || this;
        _this.defaultValue = defaultValue;
        _this.isEmpty = true;
        return _this;
    ***REMOVED***
    DefaultIfEmptySubscriber.prototype._next = function (value) ***REMOVED***
        this.isEmpty = false;
        this.destination.next(value);
    ***REMOVED***;
    DefaultIfEmptySubscriber.prototype._complete = function () ***REMOVED***
        if (this.isEmpty) ***REMOVED***
            this.destination.next(this.defaultValue);
        ***REMOVED***
        this.destination.complete();
    ***REMOVED***;
    return DefaultIfEmptySubscriber;
***REMOVED***(Subscriber_1.Subscriber));
