// rxjs/internal/operators/every.js
"use strict";
var __extends = (this && this.__extends) || (function () ***REMOVED***
    var extendStatics = function (d, b) ***REMOVED***
        extendStatics = Object.setPrototypeOf ||
            (***REMOVED*** __proto__: [] ***REMOVED*** instanceof Array && function (d, b) ***REMOVED*** d.__proto__ = b; ***REMOVED***) ||
            function (d, b) ***REMOVED*** for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; ***REMOVED***;
        return extendStatics(d, b);
    ***REMOVED***
    return function (d, b) ***REMOVED***
        extendStatics(d, b);
        function __() ***REMOVED*** this.constructor = d; ***REMOVED***
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    ***REMOVED***;
***REMOVED***)();
Object.defineProperty(exports, "__esModule", ***REMOVED*** value: true ***REMOVED***);
var Subscriber_1 = require("../Subscriber");
function every(predicate, thisArg) ***REMOVED***
    return function (source) ***REMOVED*** return source.lift(new EveryOperator(predicate, thisArg, source)); ***REMOVED***;
***REMOVED***
exports.every = every;
var EveryOperator = (function () ***REMOVED***
    function EveryOperator(predicate, thisArg, source) ***REMOVED***
        this.predicate = predicate;
        this.thisArg = thisArg;
        this.source = source;
    ***REMOVED***
    EveryOperator.prototype.call = function (observer, source) ***REMOVED***
        return source.subscribe(new EverySubscriber(observer, this.predicate, this.thisArg, this.source));
    ***REMOVED***;
    return EveryOperator;
***REMOVED***());
var EverySubscriber = (function (_super) ***REMOVED***
    __extends(EverySubscriber, _super);
    function EverySubscriber(destination, predicate, thisArg, source) ***REMOVED***
        var _this = _super.call(this, destination) || this;
        _this.predicate = predicate;
        _this.thisArg = thisArg;
        _this.source = source;
        _this.index = 0;
        _this.thisArg = thisArg || _this;
        return _this;
    ***REMOVED***
    EverySubscriber.prototype.notifyComplete = function (everyValueMatch) ***REMOVED***
        this.destination.next(everyValueMatch);
        this.destination.complete();
    ***REMOVED***;
    EverySubscriber.prototype._next = function (value) ***REMOVED***
        var result = false;
        try ***REMOVED***
            result = this.predicate.call(this.thisArg, value, this.index++, this.source);
        ***REMOVED***
        catch (err) ***REMOVED***
            this.destination.error(err);
            return;
        ***REMOVED***
        if (!result) ***REMOVED***
            this.notifyComplete(false);
        ***REMOVED***
    ***REMOVED***;
    EverySubscriber.prototype._complete = function () ***REMOVED***
        this.notifyComplete(true);
    ***REMOVED***;
    return EverySubscriber;
***REMOVED***(Subscriber_1.Subscriber));
