// rxjs/internal/operators/mergeScan.js
"use strict";
var __extends = (this && this.__extends) || (function () ***REMOVED***
    var extendStatics = function (d, b) ***REMOVED***
        extendStatics = Object.setPrototypeOf ||
            (***REMOVED*** __proto__: [] ***REMOVED*** instanceof Array && function (d, b) ***REMOVED*** d.__proto__ = b; ***REMOVED***) ||
            function (d, b) ***REMOVED*** for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; ***REMOVED***;
        return extendStatics(d, b);
    ***REMOVED***
    return function (d, b) ***REMOVED***
        extendStatics(d, b);
        function __() ***REMOVED*** this.constructor = d; ***REMOVED***
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    ***REMOVED***;
***REMOVED***)();
Object.defineProperty(exports, "__esModule", ***REMOVED*** value: true ***REMOVED***);
var subscribeToResult_1 = require("../util/subscribeToResult");
var OuterSubscriber_1 = require("../OuterSubscriber");
var InnerSubscriber_1 = require("../InnerSubscriber");
function mergeScan(accumulator, seed, concurrent) ***REMOVED***
    if (concurrent === void 0) ***REMOVED*** concurrent = Number.POSITIVE_INFINITY; ***REMOVED***
    return function (source) ***REMOVED*** return source.lift(new MergeScanOperator(accumulator, seed, concurrent)); ***REMOVED***;
***REMOVED***
exports.mergeScan = mergeScan;
var MergeScanOperator = (function () ***REMOVED***
    function MergeScanOperator(accumulator, seed, concurrent) ***REMOVED***
        this.accumulator = accumulator;
        this.seed = seed;
        this.concurrent = concurrent;
    ***REMOVED***
    MergeScanOperator.prototype.call = function (subscriber, source) ***REMOVED***
        return source.subscribe(new MergeScanSubscriber(subscriber, this.accumulator, this.seed, this.concurrent));
    ***REMOVED***;
    return MergeScanOperator;
***REMOVED***());
exports.MergeScanOperator = MergeScanOperator;
var MergeScanSubscriber = (function (_super) ***REMOVED***
    __extends(MergeScanSubscriber, _super);
    function MergeScanSubscriber(destination, accumulator, acc, concurrent) ***REMOVED***
        var _this = _super.call(this, destination) || this;
        _this.accumulator = accumulator;
        _this.acc = acc;
        _this.concurrent = concurrent;
        _this.hasValue = false;
        _this.hasCompleted = false;
        _this.buffer = [];
        _this.active = 0;
        _this.index = 0;
        return _this;
    ***REMOVED***
    MergeScanSubscriber.prototype._next = function (value) ***REMOVED***
        if (this.active < this.concurrent) ***REMOVED***
            var index = this.index++;
            var destination = this.destination;
            var ish = void 0;
            try ***REMOVED***
                var accumulator = this.accumulator;
                ish = accumulator(this.acc, value, index);
            ***REMOVED***
            catch (e) ***REMOVED***
                return destination.error(e);
            ***REMOVED***
            this.active++;
            this._innerSub(ish, value, index);
        ***REMOVED***
        else ***REMOVED***
            this.buffer.push(value);
        ***REMOVED***
    ***REMOVED***;
    MergeScanSubscriber.prototype._innerSub = function (ish, value, index) ***REMOVED***
        var innerSubscriber = new InnerSubscriber_1.InnerSubscriber(this, value, index);
        var destination = this.destination;
        destination.add(innerSubscriber);
        var innerSubscription = subscribeToResult_1.subscribeToResult(this, ish, undefined, undefined, innerSubscriber);
        if (innerSubscription !== innerSubscriber) ***REMOVED***
            destination.add(innerSubscription);
        ***REMOVED***
    ***REMOVED***;
    MergeScanSubscriber.prototype._complete = function () ***REMOVED***
        this.hasCompleted = true;
        if (this.active === 0 && this.buffer.length === 0) ***REMOVED***
            if (this.hasValue === false) ***REMOVED***
                this.destination.next(this.acc);
            ***REMOVED***
            this.destination.complete();
        ***REMOVED***
        this.unsubscribe();
    ***REMOVED***;
    MergeScanSubscriber.prototype.notifyNext = function (outerValue, innerValue, outerIndex, innerIndex, innerSub) ***REMOVED***
        var destination = this.destination;
        this.acc = innerValue;
        this.hasValue = true;
        destination.next(innerValue);
    ***REMOVED***;
    MergeScanSubscriber.prototype.notifyComplete = function (innerSub) ***REMOVED***
        var buffer = this.buffer;
        var destination = this.destination;
        destination.remove(innerSub);
        this.active--;
        if (buffer.length > 0) ***REMOVED***
            this._next(buffer.shift());
        ***REMOVED***
        else if (this.active === 0 && this.hasCompleted) ***REMOVED***
            if (this.hasValue === false) ***REMOVED***
                this.destination.next(this.acc);
            ***REMOVED***
            this.destination.complete();
        ***REMOVED***
    ***REMOVED***;
    return MergeScanSubscriber;
***REMOVED***(OuterSubscriber_1.OuterSubscriber));
exports.MergeScanSubscriber = MergeScanSubscriber;
