// rxjs/internal/operators/multicast.js
"use strict";
Object.defineProperty(exports, "__esModule", ***REMOVED*** value: true ***REMOVED***);
var ConnectableObservable_1 = require("../observable/ConnectableObservable");
function multicast(subjectOrSubjectFactory, selector) ***REMOVED***
    return function multicastOperatorFunction(source) ***REMOVED***
        var subjectFactory;
        if (typeof subjectOrSubjectFactory === 'function') ***REMOVED***
            subjectFactory = subjectOrSubjectFactory;
        ***REMOVED***
        else ***REMOVED***
            subjectFactory = function subjectFactory() ***REMOVED***
                return subjectOrSubjectFactory;
            ***REMOVED***;
        ***REMOVED***
        if (typeof selector === 'function') ***REMOVED***
            return source.lift(new MulticastOperator(subjectFactory, selector));
        ***REMOVED***
        var connectable = Object.create(source, ConnectableObservable_1.connectableObservableDescriptor);
        connectable.source = source;
        connectable.subjectFactory = subjectFactory;
        return connectable;
    ***REMOVED***;
***REMOVED***
exports.multicast = multicast;
var MulticastOperator = (function () ***REMOVED***
    function MulticastOperator(subjectFactory, selector) ***REMOVED***
        this.subjectFactory = subjectFactory;
        this.selector = selector;
    ***REMOVED***
    MulticastOperator.prototype.call = function (subscriber, source) ***REMOVED***
        var selector = this.selector;
        var subject = this.subjectFactory();
        var subscription = selector(subject).subscribe(subscriber);
        subscription.add(source.subscribe(subject));
        return subscription;
    ***REMOVED***;
    return MulticastOperator;
***REMOVED***());
exports.MulticastOperator = MulticastOperator;
