// rxjs/internal/operators/pairwise.js
"use strict";
var __extends = (this && this.__extends) || (function () ***REMOVED***
    var extendStatics = function (d, b) ***REMOVED***
        extendStatics = Object.setPrototypeOf ||
            (***REMOVED*** __proto__: [] ***REMOVED*** instanceof Array && function (d, b) ***REMOVED*** d.__proto__ = b; ***REMOVED***) ||
            function (d, b) ***REMOVED*** for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; ***REMOVED***;
        return extendStatics(d, b);
    ***REMOVED***
    return function (d, b) ***REMOVED***
        extendStatics(d, b);
        function __() ***REMOVED*** this.constructor = d; ***REMOVED***
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    ***REMOVED***;
***REMOVED***)();
Object.defineProperty(exports, "__esModule", ***REMOVED*** value: true ***REMOVED***);
var Subscriber_1 = require("../Subscriber");
function pairwise() ***REMOVED***
    return function (source) ***REMOVED*** return source.lift(new PairwiseOperator()); ***REMOVED***;
***REMOVED***
exports.pairwise = pairwise;
var PairwiseOperator = (function () ***REMOVED***
    function PairwiseOperator() ***REMOVED***
    ***REMOVED***
    PairwiseOperator.prototype.call = function (subscriber, source) ***REMOVED***
        return source.subscribe(new PairwiseSubscriber(subscriber));
    ***REMOVED***;
    return PairwiseOperator;
***REMOVED***());
var PairwiseSubscriber = (function (_super) ***REMOVED***
    __extends(PairwiseSubscriber, _super);
    function PairwiseSubscriber(destination) ***REMOVED***
        var _this = _super.call(this, destination) || this;
        _this.hasPrev = false;
        return _this;
    ***REMOVED***
    PairwiseSubscriber.prototype._next = function (value) ***REMOVED***
        var pair;
        if (this.hasPrev) ***REMOVED***
            pair = [this.prev, value];
        ***REMOVED***
        else ***REMOVED***
            this.hasPrev = true;
        ***REMOVED***
        this.prev = value;
        if (pair) ***REMOVED***
            this.destination.next(pair);
        ***REMOVED***
    ***REMOVED***;
    return PairwiseSubscriber;
***REMOVED***(Subscriber_1.Subscriber));
