// rxjs/internal/operators/retry.js
"use strict";
var __extends = (this && this.__extends) || (function () ***REMOVED***
    var extendStatics = function (d, b) ***REMOVED***
        extendStatics = Object.setPrototypeOf ||
            (***REMOVED*** __proto__: [] ***REMOVED*** instanceof Array && function (d, b) ***REMOVED*** d.__proto__ = b; ***REMOVED***) ||
            function (d, b) ***REMOVED*** for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; ***REMOVED***;
        return extendStatics(d, b);
    ***REMOVED***
    return function (d, b) ***REMOVED***
        extendStatics(d, b);
        function __() ***REMOVED*** this.constructor = d; ***REMOVED***
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    ***REMOVED***;
***REMOVED***)();
Object.defineProperty(exports, "__esModule", ***REMOVED*** value: true ***REMOVED***);
var Subscriber_1 = require("../Subscriber");
function retry(count) ***REMOVED***
    if (count === void 0) ***REMOVED*** count = -1; ***REMOVED***
    return function (source) ***REMOVED*** return source.lift(new RetryOperator(count, source)); ***REMOVED***;
***REMOVED***
exports.retry = retry;
var RetryOperator = (function () ***REMOVED***
    function RetryOperator(count, source) ***REMOVED***
        this.count = count;
        this.source = source;
    ***REMOVED***
    RetryOperator.prototype.call = function (subscriber, source) ***REMOVED***
        return source.subscribe(new RetrySubscriber(subscriber, this.count, this.source));
    ***REMOVED***;
    return RetryOperator;
***REMOVED***());
var RetrySubscriber = (function (_super) ***REMOVED***
    __extends(RetrySubscriber, _super);
    function RetrySubscriber(destination, count, source) ***REMOVED***
        var _this = _super.call(this, destination) || this;
        _this.count = count;
        _this.source = source;
        return _this;
    ***REMOVED***
    RetrySubscriber.prototype.error = function (err) ***REMOVED***
        if (!this.isStopped) ***REMOVED***
            var _a = this, source = _a.source, count = _a.count;
            if (count === 0) ***REMOVED***
                return _super.prototype.error.call(this, err);
            ***REMOVED***
            else if (count > -1) ***REMOVED***
                this.count = count - 1;
            ***REMOVED***
            source.subscribe(this._unsubscribeAndRecycle());
        ***REMOVED***
    ***REMOVED***;
    return RetrySubscriber;
***REMOVED***(Subscriber_1.Subscriber));
