// rxjs/internal/operators/scan.js
"use strict";
var __extends = (this && this.__extends) || (function () ***REMOVED***
    var extendStatics = function (d, b) ***REMOVED***
        extendStatics = Object.setPrototypeOf ||
            (***REMOVED*** __proto__: [] ***REMOVED*** instanceof Array && function (d, b) ***REMOVED*** d.__proto__ = b; ***REMOVED***) ||
            function (d, b) ***REMOVED*** for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; ***REMOVED***;
        return extendStatics(d, b);
    ***REMOVED***
    return function (d, b) ***REMOVED***
        extendStatics(d, b);
        function __() ***REMOVED*** this.constructor = d; ***REMOVED***
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    ***REMOVED***;
***REMOVED***)();
Object.defineProperty(exports, "__esModule", ***REMOVED*** value: true ***REMOVED***);
var Subscriber_1 = require("../Subscriber");
function scan(accumulator, seed) ***REMOVED***
    var hasSeed = false;
    if (arguments.length >= 2) ***REMOVED***
        hasSeed = true;
    ***REMOVED***
    return function scanOperatorFunction(source) ***REMOVED***
        return source.lift(new ScanOperator(accumulator, seed, hasSeed));
    ***REMOVED***;
***REMOVED***
exports.scan = scan;
var ScanOperator = (function () ***REMOVED***
    function ScanOperator(accumulator, seed, hasSeed) ***REMOVED***
        if (hasSeed === void 0) ***REMOVED*** hasSeed = false; ***REMOVED***
        this.accumulator = accumulator;
        this.seed = seed;
        this.hasSeed = hasSeed;
    ***REMOVED***
    ScanOperator.prototype.call = function (subscriber, source) ***REMOVED***
        return source.subscribe(new ScanSubscriber(subscriber, this.accumulator, this.seed, this.hasSeed));
    ***REMOVED***;
    return ScanOperator;
***REMOVED***());
var ScanSubscriber = (function (_super) ***REMOVED***
    __extends(ScanSubscriber, _super);
    function ScanSubscriber(destination, accumulator, _seed, hasSeed) ***REMOVED***
        var _this = _super.call(this, destination) || this;
        _this.accumulator = accumulator;
        _this._seed = _seed;
        _this.hasSeed = hasSeed;
        _this.index = 0;
        return _this;
    ***REMOVED***
    Object.defineProperty(ScanSubscriber.prototype, "seed", ***REMOVED***
        get: function () ***REMOVED***
            return this._seed;
        ***REMOVED***,
        set: function (value) ***REMOVED***
            this.hasSeed = true;
            this._seed = value;
        ***REMOVED***,
        enumerable: true,
        configurable: true
    ***REMOVED***);
    ScanSubscriber.prototype._next = function (value) ***REMOVED***
        if (!this.hasSeed) ***REMOVED***
            this.seed = value;
            this.destination.next(value);
        ***REMOVED***
        else ***REMOVED***
            return this._tryNext(value);
        ***REMOVED***
    ***REMOVED***;
    ScanSubscriber.prototype._tryNext = function (value) ***REMOVED***
        var index = this.index++;
        var result;
        try ***REMOVED***
            result = this.accumulator(this.seed, value, index);
        ***REMOVED***
        catch (err) ***REMOVED***
            this.destination.error(err);
        ***REMOVED***
        this.seed = result;
        this.destination.next(result);
    ***REMOVED***;
    return ScanSubscriber;
***REMOVED***(Subscriber_1.Subscriber));
