// rxjs/internal/operators/switchMap.js
"use strict";
var __extends = (this && this.__extends) || (function () ***REMOVED***
    var extendStatics = function (d, b) ***REMOVED***
        extendStatics = Object.setPrototypeOf ||
            (***REMOVED*** __proto__: [] ***REMOVED*** instanceof Array && function (d, b) ***REMOVED*** d.__proto__ = b; ***REMOVED***) ||
            function (d, b) ***REMOVED*** for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; ***REMOVED***;
        return extendStatics(d, b);
    ***REMOVED***
    return function (d, b) ***REMOVED***
        extendStatics(d, b);
        function __() ***REMOVED*** this.constructor = d; ***REMOVED***
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    ***REMOVED***;
***REMOVED***)();
Object.defineProperty(exports, "__esModule", ***REMOVED*** value: true ***REMOVED***);
var OuterSubscriber_1 = require("../OuterSubscriber");
var InnerSubscriber_1 = require("../InnerSubscriber");
var subscribeToResult_1 = require("../util/subscribeToResult");
var map_1 = require("./map");
var from_1 = require("../observable/from");
function switchMap(project, resultSelector) ***REMOVED***
    if (typeof resultSelector === 'function') ***REMOVED***
        return function (source) ***REMOVED*** return source.pipe(switchMap(function (a, i) ***REMOVED*** return from_1.from(project(a, i)).pipe(map_1.map(function (b, ii) ***REMOVED*** return resultSelector(a, b, i, ii); ***REMOVED***)); ***REMOVED***)); ***REMOVED***;
    ***REMOVED***
    return function (source) ***REMOVED*** return source.lift(new SwitchMapOperator(project)); ***REMOVED***;
***REMOVED***
exports.switchMap = switchMap;
var SwitchMapOperator = (function () ***REMOVED***
    function SwitchMapOperator(project) ***REMOVED***
        this.project = project;
    ***REMOVED***
    SwitchMapOperator.prototype.call = function (subscriber, source) ***REMOVED***
        return source.subscribe(new SwitchMapSubscriber(subscriber, this.project));
    ***REMOVED***;
    return SwitchMapOperator;
***REMOVED***());
var SwitchMapSubscriber = (function (_super) ***REMOVED***
    __extends(SwitchMapSubscriber, _super);
    function SwitchMapSubscriber(destination, project) ***REMOVED***
        var _this = _super.call(this, destination) || this;
        _this.project = project;
        _this.index = 0;
        return _this;
    ***REMOVED***
    SwitchMapSubscriber.prototype._next = function (value) ***REMOVED***
        var result;
        var index = this.index++;
        try ***REMOVED***
            result = this.project(value, index);
        ***REMOVED***
        catch (error) ***REMOVED***
            this.destination.error(error);
            return;
        ***REMOVED***
        this._innerSub(result, value, index);
    ***REMOVED***;
    SwitchMapSubscriber.prototype._innerSub = function (result, value, index) ***REMOVED***
        var innerSubscription = this.innerSubscription;
        if (innerSubscription) ***REMOVED***
            innerSubscription.unsubscribe();
        ***REMOVED***
        var innerSubscriber = new InnerSubscriber_1.InnerSubscriber(this, value, index);
        var destination = this.destination;
        destination.add(innerSubscriber);
        this.innerSubscription = subscribeToResult_1.subscribeToResult(this, result, undefined, undefined, innerSubscriber);
        if (this.innerSubscription !== innerSubscriber) ***REMOVED***
            destination.add(this.innerSubscription);
        ***REMOVED***
    ***REMOVED***;
    SwitchMapSubscriber.prototype._complete = function () ***REMOVED***
        var innerSubscription = this.innerSubscription;
        if (!innerSubscription || innerSubscription.closed) ***REMOVED***
            _super.prototype._complete.call(this);
        ***REMOVED***
        this.unsubscribe();
    ***REMOVED***;
    SwitchMapSubscriber.prototype._unsubscribe = function () ***REMOVED***
        this.innerSubscription = null;
    ***REMOVED***;
    SwitchMapSubscriber.prototype.notifyComplete = function (innerSub) ***REMOVED***
        var destination = this.destination;
        destination.remove(innerSub);
        this.innerSubscription = null;
        if (this.isStopped) ***REMOVED***
            _super.prototype._complete.call(this);
        ***REMOVED***
    ***REMOVED***;
    SwitchMapSubscriber.prototype.notifyNext = function (outerValue, innerValue, outerIndex, innerIndex, innerSub) ***REMOVED***
        this.destination.next(innerValue);
    ***REMOVED***;
    return SwitchMapSubscriber;
***REMOVED***(OuterSubscriber_1.OuterSubscriber));
