// rxjs/internal/operators/throttle.js
"use strict";
var __extends = (this && this.__extends) || (function () ***REMOVED***
    var extendStatics = function (d, b) ***REMOVED***
        extendStatics = Object.setPrototypeOf ||
            (***REMOVED*** __proto__: [] ***REMOVED*** instanceof Array && function (d, b) ***REMOVED*** d.__proto__ = b; ***REMOVED***) ||
            function (d, b) ***REMOVED*** for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; ***REMOVED***;
        return extendStatics(d, b);
    ***REMOVED***
    return function (d, b) ***REMOVED***
        extendStatics(d, b);
        function __() ***REMOVED*** this.constructor = d; ***REMOVED***
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    ***REMOVED***;
***REMOVED***)();
Object.defineProperty(exports, "__esModule", ***REMOVED*** value: true ***REMOVED***);
var OuterSubscriber_1 = require("../OuterSubscriber");
var subscribeToResult_1 = require("../util/subscribeToResult");
exports.defaultThrottleConfig = ***REMOVED***
    leading: true,
    trailing: false
***REMOVED***;
function throttle(durationSelector, config) ***REMOVED***
    if (config === void 0) ***REMOVED*** config = exports.defaultThrottleConfig; ***REMOVED***
    return function (source) ***REMOVED*** return source.lift(new ThrottleOperator(durationSelector, config.leading, config.trailing)); ***REMOVED***;
***REMOVED***
exports.throttle = throttle;
var ThrottleOperator = (function () ***REMOVED***
    function ThrottleOperator(durationSelector, leading, trailing) ***REMOVED***
        this.durationSelector = durationSelector;
        this.leading = leading;
        this.trailing = trailing;
    ***REMOVED***
    ThrottleOperator.prototype.call = function (subscriber, source) ***REMOVED***
        return source.subscribe(new ThrottleSubscriber(subscriber, this.durationSelector, this.leading, this.trailing));
    ***REMOVED***;
    return ThrottleOperator;
***REMOVED***());
var ThrottleSubscriber = (function (_super) ***REMOVED***
    __extends(ThrottleSubscriber, _super);
    function ThrottleSubscriber(destination, durationSelector, _leading, _trailing) ***REMOVED***
        var _this = _super.call(this, destination) || this;
        _this.destination = destination;
        _this.durationSelector = durationSelector;
        _this._leading = _leading;
        _this._trailing = _trailing;
        _this._hasValue = false;
        return _this;
    ***REMOVED***
    ThrottleSubscriber.prototype._next = function (value) ***REMOVED***
        this._hasValue = true;
        this._sendValue = value;
        if (!this._throttled) ***REMOVED***
            if (this._leading) ***REMOVED***
                this.send();
            ***REMOVED***
            else ***REMOVED***
                this.throttle(value);
            ***REMOVED***
        ***REMOVED***
    ***REMOVED***;
    ThrottleSubscriber.prototype.send = function () ***REMOVED***
        var _a = this, _hasValue = _a._hasValue, _sendValue = _a._sendValue;
        if (_hasValue) ***REMOVED***
            this.destination.next(_sendValue);
            this.throttle(_sendValue);
        ***REMOVED***
        this._hasValue = false;
        this._sendValue = null;
    ***REMOVED***;
    ThrottleSubscriber.prototype.throttle = function (value) ***REMOVED***
        var duration = this.tryDurationSelector(value);
        if (!!duration) ***REMOVED***
            this.add(this._throttled = subscribeToResult_1.subscribeToResult(this, duration));
        ***REMOVED***
    ***REMOVED***;
    ThrottleSubscriber.prototype.tryDurationSelector = function (value) ***REMOVED***
        try ***REMOVED***
            return this.durationSelector(value);
        ***REMOVED***
        catch (err) ***REMOVED***
            this.destination.error(err);
            return null;
        ***REMOVED***
    ***REMOVED***;
    ThrottleSubscriber.prototype.throttlingDone = function () ***REMOVED***
        var _a = this, _throttled = _a._throttled, _trailing = _a._trailing;
        if (_throttled) ***REMOVED***
            _throttled.unsubscribe();
        ***REMOVED***
        this._throttled = null;
        if (_trailing) ***REMOVED***
            this.send();
        ***REMOVED***
    ***REMOVED***;
    ThrottleSubscriber.prototype.notifyNext = function (outerValue, innerValue, outerIndex, innerIndex, innerSub) ***REMOVED***
        this.throttlingDone();
    ***REMOVED***;
    ThrottleSubscriber.prototype.notifyComplete = function () ***REMOVED***
        this.throttlingDone();
    ***REMOVED***;
    return ThrottleSubscriber;
***REMOVED***(OuterSubscriber_1.OuterSubscriber));
