// rxjs/internal/operators/window.js
"use strict";
var __extends = (this && this.__extends) || (function () ***REMOVED***
    var extendStatics = function (d, b) ***REMOVED***
        extendStatics = Object.setPrototypeOf ||
            (***REMOVED*** __proto__: [] ***REMOVED*** instanceof Array && function (d, b) ***REMOVED*** d.__proto__ = b; ***REMOVED***) ||
            function (d, b) ***REMOVED*** for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; ***REMOVED***;
        return extendStatics(d, b);
    ***REMOVED***
    return function (d, b) ***REMOVED***
        extendStatics(d, b);
        function __() ***REMOVED*** this.constructor = d; ***REMOVED***
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    ***REMOVED***;
***REMOVED***)();
Object.defineProperty(exports, "__esModule", ***REMOVED*** value: true ***REMOVED***);
var Subject_1 = require("../Subject");
var OuterSubscriber_1 = require("../OuterSubscriber");
var subscribeToResult_1 = require("../util/subscribeToResult");
function window(windowBoundaries) ***REMOVED***
    return function windowOperatorFunction(source) ***REMOVED***
        return source.lift(new WindowOperator(windowBoundaries));
    ***REMOVED***;
***REMOVED***
exports.window = window;
var WindowOperator = (function () ***REMOVED***
    function WindowOperator(windowBoundaries) ***REMOVED***
        this.windowBoundaries = windowBoundaries;
    ***REMOVED***
    WindowOperator.prototype.call = function (subscriber, source) ***REMOVED***
        var windowSubscriber = new WindowSubscriber(subscriber);
        var sourceSubscription = source.subscribe(windowSubscriber);
        if (!sourceSubscription.closed) ***REMOVED***
            windowSubscriber.add(subscribeToResult_1.subscribeToResult(windowSubscriber, this.windowBoundaries));
        ***REMOVED***
        return sourceSubscription;
    ***REMOVED***;
    return WindowOperator;
***REMOVED***());
var WindowSubscriber = (function (_super) ***REMOVED***
    __extends(WindowSubscriber, _super);
    function WindowSubscriber(destination) ***REMOVED***
        var _this = _super.call(this, destination) || this;
        _this.window = new Subject_1.Subject();
        destination.next(_this.window);
        return _this;
    ***REMOVED***
    WindowSubscriber.prototype.notifyNext = function (outerValue, innerValue, outerIndex, innerIndex, innerSub) ***REMOVED***
        this.openWindow();
    ***REMOVED***;
    WindowSubscriber.prototype.notifyError = function (error, innerSub) ***REMOVED***
        this._error(error);
    ***REMOVED***;
    WindowSubscriber.prototype.notifyComplete = function (innerSub) ***REMOVED***
        this._complete();
    ***REMOVED***;
    WindowSubscriber.prototype._next = function (value) ***REMOVED***
        this.window.next(value);
    ***REMOVED***;
    WindowSubscriber.prototype._error = function (err) ***REMOVED***
        this.window.error(err);
        this.destination.error(err);
    ***REMOVED***;
    WindowSubscriber.prototype._complete = function () ***REMOVED***
        this.window.complete();
        this.destination.complete();
    ***REMOVED***;
    WindowSubscriber.prototype._unsubscribe = function () ***REMOVED***
        this.window = null;
    ***REMOVED***;
    WindowSubscriber.prototype.openWindow = function () ***REMOVED***
        var prevWindow = this.window;
        if (prevWindow) ***REMOVED***
            prevWindow.complete();
        ***REMOVED***
        var destination = this.destination;
        var newWindow = this.window = new Subject_1.Subject();
        destination.next(newWindow);
    ***REMOVED***;
    return WindowSubscriber;
***REMOVED***(OuterSubscriber_1.OuterSubscriber));
